<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Biology 723: Statistical Computing for Biologists</title>
  <meta name="description" content="Hands-on materials for Bio 723 course at Duke University.">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Biology 723: Statistical Computing for Biologists" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Biology 723: Statistical Computing for Biologists" />
  
  <meta name="twitter:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

<meta name="author" content="Paul M. Magwene">


<meta name="date" content="2018-04-18">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="randomization-jackknife-and-bootstrap.html">
<link rel="next" href="canonical-variates-analysis.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#how-to-use-these-lecture-notes"><i class="fa fa-check"></i><b>1.1</b> How to use these lecture notes</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html"><i class="fa fa-check"></i><b>2</b> Getting Started with R</a><ul>
<li class="chapter" data-level="2.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-r"><i class="fa fa-check"></i><b>2.1</b> What is R?</a></li>
<li class="chapter" data-level="2.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-rstudio"><i class="fa fa-check"></i><b>2.2</b> What is RStudio?</a></li>
<li class="chapter" data-level="2.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#entering-commands-in-the-console"><i class="fa fa-check"></i><b>2.3</b> Entering commands in the console</a></li>
<li class="chapter" data-level="2.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#comments"><i class="fa fa-check"></i><b>2.4</b> Comments</a></li>
<li class="chapter" data-level="2.5" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#using-r-as-a-calculator"><i class="fa fa-check"></i><b>2.5</b> Using R as a Calculator</a><ul>
<li class="chapter" data-level="2.5.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#common-mathematical-functions"><i class="fa fa-check"></i><b>2.5.1</b> Common mathematical functions</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#numeric-data-types-in-r"><i class="fa fa-check"></i><b>2.6</b> Numeric data types in R</a></li>
<li class="chapter" data-level="2.7" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#variable-assignment-in-r"><i class="fa fa-check"></i><b>2.7</b> Variable assignment in R</a><ul>
<li class="chapter" data-level="2.7.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#valid-variable-names"><i class="fa fa-check"></i><b>2.7.1</b> Valid variable names</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#working-with-vectors-in-r"><i class="fa fa-check"></i><b>2.8</b> Working with Vectors in R</a><ul>
<li class="chapter" data-level="2.8.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#vector-arithmetic"><i class="fa fa-check"></i><b>2.8.1</b> Vector Arithmetic</a></li>
<li class="chapter" data-level="2.8.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#vector-recycling"><i class="fa fa-check"></i><b>2.8.2</b> Vector recycling</a></li>
<li class="chapter" data-level="2.8.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#simple-statistical-functions-for-numeric-vectors"><i class="fa fa-check"></i><b>2.8.3</b> Simple statistical functions for numeric vectors</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#the-r-help-system"><i class="fa fa-check"></i><b>2.9</b> The R Help System</a><ul>
<li class="chapter" data-level="2.9.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#getting-help-from-the-console"><i class="fa fa-check"></i><b>2.9.1</b> Getting help from the console</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#packages"><i class="fa fa-check"></i><b>2.10</b> Packages</a><ul>
<li class="chapter" data-level="2.10.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-console"><i class="fa fa-check"></i><b>2.10.1</b> Installing packages from the console</a></li>
<li class="chapter" data-level="2.10.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#install-the-tidyverse-package"><i class="fa fa-check"></i><b>2.10.2</b> Install the tidyverse package</a></li>
<li class="chapter" data-level="2.10.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-rstudio-dialog"><i class="fa fa-check"></i><b>2.10.3</b> Installing packages from the RStudio dialog</a></li>
<li class="chapter" data-level="2.10.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#install-the-stringr-package"><i class="fa fa-check"></i><b>2.10.4</b> Install the stringr package</a></li>
<li class="chapter" data-level="2.10.5" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#loading-packages-with-the-library-function"><i class="fa fa-check"></i><b>2.10.5</b> Loading packages with the <code>library()</code> function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html"><i class="fa fa-check"></i><b>3</b> R Basics: Data types and data structures</a><ul>
<li class="chapter" data-level="3.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-boolean-values"><i class="fa fa-check"></i><b>3.1</b> Logical (Boolean) values</a><ul>
<li class="chapter" data-level="3.1.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#comparison-operators-applied-to-numeric-vectors"><i class="fa fa-check"></i><b>3.1.1</b> Comparison operators applied to numeric vectors</a></li>
<li class="chapter" data-level="3.1.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-vectors"><i class="fa fa-check"></i><b>3.1.2</b> Logical vectors</a></li>
<li class="chapter" data-level="3.1.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-operators"><i class="fa fa-check"></i><b>3.1.3</b> Logical operators</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#character-strings"><i class="fa fa-check"></i><b>3.2</b> Character strings</a><ul>
<li class="chapter" data-level="3.2.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#the-stringr-package"><i class="fa fa-check"></i><b>3.2.1</b> The <code>stringr</code> package</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#vectors"><i class="fa fa-check"></i><b>3.3</b> Vectors</a><ul>
<li class="chapter" data-level="3.3.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#indexing-vectors"><i class="fa fa-check"></i><b>3.3.1</b> Indexing Vectors</a></li>
<li class="chapter" data-level="3.3.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#combining-indexing-and-comparison-of-vectors"><i class="fa fa-check"></i><b>3.3.2</b> Combining Indexing and Comparison of Vectors</a></li>
<li class="chapter" data-level="3.3.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#vector-manipulation"><i class="fa fa-check"></i><b>3.3.3</b> Vector manipulation</a></li>
<li class="chapter" data-level="3.3.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#vectors-from-regular-sequences"><i class="fa fa-check"></i><b>3.3.4</b> Vectors from regular sequences</a></li>
<li class="chapter" data-level="3.3.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#additional-functions-for-working-with-vectors"><i class="fa fa-check"></i><b>3.3.5</b> Additional functions for working with vectors</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#lists"><i class="fa fa-check"></i><b>3.4</b> Lists</a><ul>
<li class="chapter" data-level="3.4.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#length-and-type-of-lists"><i class="fa fa-check"></i><b>3.4.1</b> Length and type of lists</a></li>
<li class="chapter" data-level="3.4.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#indexing-lists"><i class="fa fa-check"></i><b>3.4.2</b> Indexing lists</a></li>
<li class="chapter" data-level="3.4.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#naming-list-elements"><i class="fa fa-check"></i><b>3.4.3</b> Naming list elements</a></li>
<li class="chapter" data-level="3.4.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#the-operator"><i class="fa fa-check"></i><b>3.4.4</b> The <code>$</code> operator</a></li>
<li class="chapter" data-level="3.4.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#changing-and-adding-lists-items"><i class="fa fa-check"></i><b>3.4.5</b> Changing and adding lists items</a></li>
<li class="chapter" data-level="3.4.6" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#combining-lists"><i class="fa fa-check"></i><b>3.4.6</b> Combining lists</a></li>
<li class="chapter" data-level="3.4.7" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#converting-lists-to-vectors"><i class="fa fa-check"></i><b>3.4.7</b> Converting lists to vectors</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#data-frames"><i class="fa fa-check"></i><b>3.5</b> Data frames</a><ul>
<li class="chapter" data-level="3.5.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#creating-a-data-frame"><i class="fa fa-check"></i><b>3.5.1</b> Creating a data frame</a></li>
<li class="chapter" data-level="3.5.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#type-and-class-for-data-frames"><i class="fa fa-check"></i><b>3.5.2</b> Type and class for data frames</a></li>
<li class="chapter" data-level="3.5.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#length-and-dimension-for-data-frames"><i class="fa fa-check"></i><b>3.5.3</b> Length and dimension for data frames</a></li>
<li class="chapter" data-level="3.5.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#indexing-and-accessing-data-frames"><i class="fa fa-check"></i><b>3.5.4</b> Indexing and accessing data frames</a></li>
<li class="chapter" data-level="3.5.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-indexing-of-data-frames"><i class="fa fa-check"></i><b>3.5.5</b> Logical indexing of data frames</a></li>
<li class="chapter" data-level="3.5.6" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#adding-columns-to-a-data-frame"><i class="fa fa-check"></i><b>3.5.6</b> Adding columns to a data frame</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html"><i class="fa fa-check"></i><b>4</b> R Markdown and R Notebooks</a><ul>
<li class="chapter" data-level="4.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#r-notebooks"><i class="fa fa-check"></i><b>4.1</b> R Notebooks</a></li>
<li class="chapter" data-level="4.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#creating-an-r-notebook"><i class="fa fa-check"></i><b>4.2</b> Creating an R Notebook</a></li>
<li class="chapter" data-level="4.3" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#the-default-r-notebook-template"><i class="fa fa-check"></i><b>4.3</b> The default R Notebook template</a></li>
<li class="chapter" data-level="4.4" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-and-non-code-blocks"><i class="fa fa-check"></i><b>4.4</b> Code and Non-code blocks</a><ul>
<li class="chapter" data-level="4.4.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#non-code-blocks"><i class="fa fa-check"></i><b>4.4.1</b> Non-code blocks</a></li>
<li class="chapter" data-level="4.4.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-blocks"><i class="fa fa-check"></i><b>4.4.2</b> Code blocks</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-a-code-chunk"><i class="fa fa-check"></i><b>4.5</b> Running a code chunk</a></li>
<li class="chapter" data-level="4.6" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-all-code-chunks-above"><i class="fa fa-check"></i><b>4.6</b> Running all code chunks above</a></li>
<li class="chapter" data-level="4.7" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#knitting-an-r-markdown-to-html"><i class="fa fa-check"></i><b>4.7</b> “Knitting an” R Markdown to HTML</a></li>
<li class="chapter" data-level="4.8" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#sharing-your-reproducible-r-notebook"><i class="fa fa-check"></i><b>4.8</b> Sharing your reproducible R Notebook</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html"><i class="fa fa-check"></i><b>5</b> Introduction to ggplot2</a><ul>
<li class="chapter" data-level="5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#loading-ggplot2"><i class="fa fa-check"></i><b>5.1</b> Loading ggplot2</a></li>
<li class="chapter" data-level="5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#example-data-set-andersons-iris-data"><i class="fa fa-check"></i><b>5.2</b> Example data set: Anderson’s Iris Data</a></li>
<li class="chapter" data-level="5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#template-for-single-layer-plots-in-ggplot2"><i class="fa fa-check"></i><b>5.3</b> Template for single layer plots in ggplot2</a></li>
<li class="chapter" data-level="5.4" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#an-aside-about-function-arguments"><i class="fa fa-check"></i><b>5.4</b> An aside about function arguments</a></li>
<li class="chapter" data-level="5.5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#strip-plots"><i class="fa fa-check"></i><b>5.5</b> Strip plots</a><ul>
<li class="chapter" data-level="5.5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#jittering-data"><i class="fa fa-check"></i><b>5.5.1</b> Jittering data</a></li>
<li class="chapter" data-level="5.5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-categorical-information"><i class="fa fa-check"></i><b>5.5.2</b> Adding categorical information</a></li>
<li class="chapter" data-level="5.5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#rotating-plot-coordinates"><i class="fa fa-check"></i><b>5.5.3</b> Rotating plot coordinates</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#histograms"><i class="fa fa-check"></i><b>5.6</b> Histograms</a><ul>
<li class="chapter" data-level="5.6.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#variations-on-histograms-when-considering-categorical-data"><i class="fa fa-check"></i><b>5.6.1</b> Variations on histograms when considering categorical data</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#faceting-to-depict-categorical-information"><i class="fa fa-check"></i><b>5.7</b> Faceting to depict categorical information</a></li>
<li class="chapter" data-level="5.8" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots"><i class="fa fa-check"></i><b>5.8</b> Density plots</a></li>
<li class="chapter" data-level="5.9" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-or-beanplot"><i class="fa fa-check"></i><b>5.9</b> Violin or Beanplot</a></li>
<li class="chapter" data-level="5.10" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplots"><i class="fa fa-check"></i><b>5.10</b> Boxplots</a></li>
<li class="chapter" data-level="5.11" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#building-complex-visualizations-with-layers"><i class="fa fa-check"></i><b>5.11</b> Building complex visualizations with layers</a></li>
<li class="chapter" data-level="5.12" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#useful-combination-plots"><i class="fa fa-check"></i><b>5.12</b> Useful combination plots</a><ul>
<li class="chapter" data-level="5.12.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplot-plus-strip-plot"><i class="fa fa-check"></i><b>5.12.1</b> Boxplot plus strip plot</a></li>
<li class="chapter" data-level="5.12.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#setting-shared-aesthetics"><i class="fa fa-check"></i><b>5.12.2</b> Setting shared aesthetics</a></li>
</ul></li>
<li class="chapter" data-level="5.13" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot-layers-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>5.13</b> ggplot layers can be assigned to variables</a><ul>
<li class="chapter" data-level="5.13.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-plot-plus-strip-plot"><i class="fa fa-check"></i><b>5.13.1</b> Violin plot plus strip plot</a></li>
</ul></li>
<li class="chapter" data-level="5.14" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-titles-and-tweaking-axis-labels"><i class="fa fa-check"></i><b>5.14</b> Adding titles and tweaking axis labels</a></li>
<li class="chapter" data-level="5.15" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot2-themes"><i class="fa fa-check"></i><b>5.15</b> ggplot2 themes</a><ul>
<li class="chapter" data-level="5.15.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#further-customization-with-ggplot2theme"><i class="fa fa-check"></i><b>5.15.1</b> Further customization with <code>ggplot2::theme</code></a></li>
</ul></li>
<li class="chapter" data-level="5.16" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#other-aspects-of-ggplots-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>5.16</b> Other aspects of ggplots can be assigned to variables</a></li>
<li class="chapter" data-level="5.17" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-plots"><i class="fa fa-check"></i><b>5.17</b> Bivariate plots</a><ul>
<li class="chapter" data-level="5.17.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#scatter-plots"><i class="fa fa-check"></i><b>5.17.1</b> Scatter plots</a></li>
</ul></li>
<li class="chapter" data-level="5.18" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-density-plots"><i class="fa fa-check"></i><b>5.18</b> Bivariate density plots</a></li>
<li class="chapter" data-level="5.19" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#combining-scatter-plots-and-density-plots-with-categorical-information"><i class="fa fa-check"></i><b>5.19</b> Combining Scatter Plots and Density Plots with Categorical Information</a></li>
<li class="chapter" data-level="5.20" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots-with-fill"><i class="fa fa-check"></i><b>5.20</b> Density plots with fill</a></li>
<li class="chapter" data-level="5.21" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#d-bin-and-hex-plots"><i class="fa fa-check"></i><b>5.21</b> 2D bin and hex plots</a></li>
<li class="chapter" data-level="5.22" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#the-cowplot-package"><i class="fa fa-check"></i><b>5.22</b> The <code>cowplot</code> package</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html"><i class="fa fa-check"></i><b>6</b> Introduction to the <code>dplyr</code> package</a><ul>
<li class="chapter" data-level="6.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#libraries"><i class="fa fa-check"></i><b>6.1</b> Libraries</a></li>
<li class="chapter" data-level="6.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#reading-data-with-the-readr-package"><i class="fa fa-check"></i><b>6.2</b> Reading data with the <code>readr</code> package</a><ul>
<li class="chapter" data-level="6.2.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#example-data-nc-births"><i class="fa fa-check"></i><b>6.2.1</b> Example data: NC Births</a></li>
<li class="chapter" data-level="6.2.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#reading-excel-files"><i class="fa fa-check"></i><b>6.2.2</b> Reading Excel files</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#a-note-on-tibbles"><i class="fa fa-check"></i><b>6.3</b> A note on “tibbles”</a></li>
<li class="chapter" data-level="6.4" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#data-filtering-and-transformation-with-dplyr"><i class="fa fa-check"></i><b>6.4</b> Data filtering and transformation with <code>dplyr</code></a></li>
<li class="chapter" data-level="6.5" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#dplyrs-verbs"><i class="fa fa-check"></i><b>6.5</b> dplyr’s “verbs”</a><ul>
<li class="chapter" data-level="6.5.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#select"><i class="fa fa-check"></i><b>6.5.1</b> <code>select</code></a></li>
<li class="chapter" data-level="6.5.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#filter"><i class="fa fa-check"></i><b>6.5.2</b> <code>filter</code></a></li>
<li class="chapter" data-level="6.5.3" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#mutate"><i class="fa fa-check"></i><b>6.5.3</b> <code>mutate</code></a></li>
<li class="chapter" data-level="6.5.4" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#arrange"><i class="fa fa-check"></i><b>6.5.4</b> <code>arrange</code></a></li>
<li class="chapter" data-level="6.5.5" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#summarize"><i class="fa fa-check"></i><b>6.5.5</b> <code>summarize</code></a></li>
<li class="chapter" data-level="6.5.6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#group_by"><i class="fa fa-check"></i><b>6.5.6</b> <code>group_by</code></a></li>
<li class="chapter" data-level="6.5.7" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#combining-grouping-and-summarizing"><i class="fa fa-check"></i><b>6.5.7</b> Combining grouping and summarizing</a></li>
<li class="chapter" data-level="6.5.8" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#scoped-variants-of-mutate-and-summarize"><i class="fa fa-check"></i><b>6.5.8</b> Scoped variants of <code>mutate</code> and <code>summarize</code></a></li>
<li class="chapter" data-level="6.5.9" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#combining-summarize-with-grouping-aesthetics-in-ggplot2"><i class="fa fa-check"></i><b>6.5.9</b> Combining summarize with grouping aesthetics in <code>ggplot2</code></a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#pipes"><i class="fa fa-check"></i><b>6.6</b> Pipes</a><ul>
<li class="chapter" data-level="6.6.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#install-and-load-magrittr"><i class="fa fa-check"></i><b>6.6.1</b> Install and load <code>magrittr</code></a></li>
<li class="chapter" data-level="6.6.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-basic-pipe-operator"><i class="fa fa-check"></i><b>6.6.2</b> The basic pipe operator</a></li>
<li class="chapter" data-level="6.6.3" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#an-example-without-pipes"><i class="fa fa-check"></i><b>6.6.3</b> An example without pipes</a></li>
<li class="chapter" data-level="6.6.4" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-same-example-using-pipes"><i class="fa fa-check"></i><b>6.6.4</b> The same example using pipes</a></li>
<li class="chapter" data-level="6.6.5" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#assigning-the-output-of-a-statement-involving-pipes-to-a-variable"><i class="fa fa-check"></i><b>6.6.5</b> Assigning the output of a statement involving pipes to a variable</a></li>
<li class="chapter" data-level="6.6.6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#compound-assignment-pipe-operator"><i class="fa fa-check"></i><b>6.6.6</b> Compound assignment pipe operator</a></li>
<li class="chapter" data-level="6.6.7" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-dot-operator-with-pipes"><i class="fa fa-check"></i><b>6.6.7</b> The dot operator with pipes</a></li>
<li class="chapter" data-level="6.6.8" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-exposition-pipe-operator"><i class="fa fa-check"></i><b>6.6.8</b> The exposition pipe operator</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>7</b> Data wrangling</a><ul>
<li class="chapter" data-level="7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#libraries-1"><i class="fa fa-check"></i><b>7.1</b> Libraries</a></li>
<li class="chapter" data-level="7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#data"><i class="fa fa-check"></i><b>7.2</b> Data</a></li>
<li class="chapter" data-level="7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#renaming-data-frame-columms"><i class="fa fa-check"></i><b>7.3</b> Renaming data frame columms</a></li>
<li class="chapter" data-level="7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-unneeded-columns"><i class="fa fa-check"></i><b>7.4</b> Dropping unneeded columns</a><ul>
<li class="chapter" data-level="7.4.1" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-all-empty-columns"><i class="fa fa-check"></i><b>7.4.1</b> Finding all empty columns</a></li>
<li class="chapter" data-level="7.4.2" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-columns-by-matching-names"><i class="fa fa-check"></i><b>7.4.2</b> Dropping columns by matching names</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="data-wrangling.html"><a href="data-wrangling.html#reshaping-data-with-tidyr"><i class="fa fa-check"></i><b>7.5</b> Reshaping data with tidyr</a><ul>
<li class="chapter" data-level="7.5.1" data-path="data-wrangling.html"><a href="data-wrangling.html#wide-to-long-conversions-using-tidyrgather"><i class="fa fa-check"></i><b>7.5.1</b> Wide to long conversions using <code>tidyr::gather</code></a></li>
<li class="chapter" data-level="7.5.2" data-path="data-wrangling.html"><a href="data-wrangling.html#extracting-information-from-combined-variables-using-tidyrextract"><i class="fa fa-check"></i><b>7.5.2</b> Extracting information from combined variables using <code>tidyr::extract</code></a></li>
<li class="chapter" data-level="7.5.3" data-path="data-wrangling.html"><a href="data-wrangling.html#subsetting-rows"><i class="fa fa-check"></i><b>7.5.3</b> Subsetting rows</a></li>
<li class="chapter" data-level="7.5.4" data-path="data-wrangling.html"><a href="data-wrangling.html#splitting-columns"><i class="fa fa-check"></i><b>7.5.4</b> Splitting columns</a></li>
<li class="chapter" data-level="7.5.5" data-path="data-wrangling.html"><a href="data-wrangling.html#combining-data-frame-rows"><i class="fa fa-check"></i><b>7.5.5</b> Combining data frame rows</a></li>
<li class="chapter" data-level="7.5.6" data-path="data-wrangling.html"><a href="data-wrangling.html#sorting-data-frame-rows"><i class="fa fa-check"></i><b>7.5.6</b> Sorting data frame rows</a></li>
<li class="chapter" data-level="7.5.7" data-path="data-wrangling.html"><a href="data-wrangling.html#collapsing-columns-using-tidyrunite"><i class="fa fa-check"></i><b>7.5.7</b> Collapsing columns using <code>tidyr::unite</code></a></li>
<li class="chapter" data-level="7.5.8" data-path="data-wrangling.html"><a href="data-wrangling.html#splitting-columns-using-tidyrseparate"><i class="fa fa-check"></i><b>7.5.8</b> Splitting columns using <code>tidyr::separate</code></a></li>
<li class="chapter" data-level="7.5.9" data-path="data-wrangling.html"><a href="data-wrangling.html#long-to-wide-conversion-using-tidyrspread"><i class="fa fa-check"></i><b>7.5.9</b> Long-to-wide conversion using <code>tidyr::spread</code></a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="data-wrangling.html"><a href="data-wrangling.html#using-your-tidy-data"><i class="fa fa-check"></i><b>7.6</b> Using your tidy data</a><ul>
<li class="chapter" data-level="7.6.1" data-path="data-wrangling.html"><a href="data-wrangling.html#visualizing-gene-expression-time-series"><i class="fa fa-check"></i><b>7.6.1</b> Visualizing gene expression time series</a></li>
<li class="chapter" data-level="7.6.2" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-the-most-variable-genes"><i class="fa fa-check"></i><b>7.6.2</b> Finding the most variable genes</a></li>
<li class="chapter" data-level="7.6.3" data-path="data-wrangling.html"><a href="data-wrangling.html#exploring-bivariate-relationships-using-wide-data"><i class="fa fa-check"></i><b>7.6.3</b> Exploring bivariate relationships using “wide” data</a></li>
<li class="chapter" data-level="7.6.4" data-path="data-wrangling.html"><a href="data-wrangling.html#calculating-correlations-between-genes"><i class="fa fa-check"></i><b>7.6.4</b> Calculating correlations between genes</a></li>
<li class="chapter" data-level="7.6.5" data-path="data-wrangling.html"><a href="data-wrangling.html#adding-new-columns-and-combining-filtered-data-frames"><i class="fa fa-check"></i><b>7.6.5</b> Adding new columns and combining filtered data frames</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="data-wrangling.html"><a href="data-wrangling.html#heat-maps"><i class="fa fa-check"></i><b>7.7</b> Heat maps</a><ul>
<li class="chapter" data-level="7.7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#rcolorbrewer"><i class="fa fa-check"></i><b>7.7.1</b> RColorBrewer</a></li>
<li class="chapter" data-level="7.7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#creating-a-heat-map-using-geom_tile"><i class="fa fa-check"></i><b>7.7.2</b> Creating a heat map using <code>geom_tile</code></a></li>
<li class="chapter" data-level="7.7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#a-fancy-figure"><i class="fa fa-check"></i><b>7.7.3</b> A “fancy” figure</a></li>
<li class="chapter" data-level="7.7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#sorting-and-visualizing-genes-by-time-of-maximum-expression"><i class="fa fa-check"></i><b>7.7.4</b> Sorting and visualizing genes by time of maximum expression</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html"><i class="fa fa-check"></i><b>8</b> Vectors and Functions</a><ul>
<li class="chapter" data-level="8.1" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#libraries-2"><i class="fa fa-check"></i><b>8.1</b> Libraries</a></li>
<li class="chapter" data-level="8.2" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#vector-mathematics-in-r"><i class="fa fa-check"></i><b>8.2</b> Vector Mathematics in R</a></li>
<li class="chapter" data-level="8.3" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#writing-your-own-functions-in-r"><i class="fa fa-check"></i><b>8.3</b> Writing Your Own Functions in R</a></li>
<li class="chapter" data-level="8.4" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#function-arguments"><i class="fa fa-check"></i><b>8.4</b> Function arguments</a></li>
<li class="chapter" data-level="8.5" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#writing-functions-with-optional-arguments"><i class="fa fa-check"></i><b>8.5</b> Writing functions with optional arguments</a></li>
<li class="chapter" data-level="8.6" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#putting-r-functions-in-scripts"><i class="fa fa-check"></i><b>8.6</b> Putting R functions in Scripts</a></li>
<li class="chapter" data-level="8.7" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#simple-statistics-in-vector-form"><i class="fa fa-check"></i><b>8.7</b> Simple statistics in vector form</a><ul>
<li class="chapter" data-level="8.7.1" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#mean"><i class="fa fa-check"></i><b>8.7.1</b> Mean</a></li>
<li class="chapter" data-level="8.7.2" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#variance-and-standard-deviation"><i class="fa fa-check"></i><b>8.7.2</b> Variance and standard deviation</a></li>
<li class="chapter" data-level="8.7.3" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#covariance-and-correlation"><i class="fa fa-check"></i><b>8.7.3</b> Covariance and correlation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html"><i class="fa fa-check"></i><b>9</b> Bivariate Linear Regression</a><ul>
<li class="chapter" data-level="9.1" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#regression-terminology"><i class="fa fa-check"></i><b>9.1</b> Regression terminology</a></li>
<li class="chapter" data-level="9.2" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#libraries-3"><i class="fa fa-check"></i><b>9.2</b> Libraries</a></li>
<li class="chapter" data-level="9.3" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#illustrating-linear-regression-with-simulated-data"><i class="fa fa-check"></i><b>9.3</b> Illustrating linear regression with simulated data</a></li>
<li class="chapter" data-level="9.4" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#specifying-regression-models-in-r"><i class="fa fa-check"></i><b>9.4</b> Specifying Regression Models in R</a><ul>
<li class="chapter" data-level="9.4.1" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#fitted-values"><i class="fa fa-check"></i><b>9.4.1</b> Fitted values</a></li>
<li class="chapter" data-level="9.4.2" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#getting-the-model-coefficients"><i class="fa fa-check"></i><b>9.4.2</b> Getting the model coefficients</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#residuals"><i class="fa fa-check"></i><b>9.5</b> Residuals</a></li>
<li class="chapter" data-level="9.6" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#regression-as-sum-of-squares-decomposition"><i class="fa fa-check"></i><b>9.6</b> Regression as sum-of-squares decomposition</a></li>
<li class="chapter" data-level="9.7" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#variance-explained-by-a-regression-model"><i class="fa fa-check"></i><b>9.7</b> Variance “explained” by a regression model</a></li>
<li class="chapter" data-level="9.8" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broom-a-library-for-converting-model-results-into-data-frames"><i class="fa fa-check"></i><b>9.8</b> Broom: a library for converting model results into data frames</a><ul>
<li class="chapter" data-level="9.8.1" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broomtidy"><i class="fa fa-check"></i><b>9.8.1</b> <code>broom::tidy</code></a></li>
<li class="chapter" data-level="9.8.2" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broomaugment"><i class="fa fa-check"></i><b>9.8.2</b> <code>broom::augment</code></a></li>
<li class="chapter" data-level="9.8.3" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broomglance"><i class="fa fa-check"></i><b>9.8.3</b> <code>broom::glance</code></a></li>
</ul></li>
<li class="chapter" data-level="9.9" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#bivariate-linear-regression-on-a-real-world-data-set"><i class="fa fa-check"></i><b>9.9</b> Bivariate linear regression on a real world data set</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="matrices-in-r.html"><a href="matrices-in-r.html"><i class="fa fa-check"></i><b>10</b> Matrices in R</a><ul>
<li class="chapter" data-level="10.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#creating-matrices-in-r"><i class="fa fa-check"></i><b>10.1</b> Creating matrices in R</a></li>
<li class="chapter" data-level="10.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-arithmetic-operations-in-r"><i class="fa fa-check"></i><b>10.2</b> Matrix arithmetic operations in R</a></li>
<li class="chapter" data-level="10.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#descriptive-statistics-as-matrix-functions"><i class="fa fa-check"></i><b>10.3</b> Descriptive statistics as matrix functions</a><ul>
<li class="chapter" data-level="10.3.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#mean-vector-and-matrix"><i class="fa fa-check"></i><b>10.3.1</b> Mean vector and matrix</a></li>
<li class="chapter" data-level="10.3.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#deviation-matrix"><i class="fa fa-check"></i><b>10.3.2</b> Deviation matrix</a></li>
<li class="chapter" data-level="10.3.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#covariance-matrix"><i class="fa fa-check"></i><b>10.3.3</b> Covariance matrix</a></li>
<li class="chapter" data-level="10.3.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#correlation-matrix"><i class="fa fa-check"></i><b>10.3.4</b> Correlation matrix</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-inverse"><i class="fa fa-check"></i><b>10.4</b> Matrix Inverse</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="multiple-regression.html"><a href="multiple-regression.html"><i class="fa fa-check"></i><b>11</b> Multiple Regression</a><ul>
<li class="chapter" data-level="11.1" data-path="multiple-regression.html"><a href="multiple-regression.html#new-libraries-to-install"><i class="fa fa-check"></i><b>11.1</b> New Libraries to install</a></li>
<li class="chapter" data-level="11.2" data-path="multiple-regression.html"><a href="multiple-regression.html#libraries-4"><i class="fa fa-check"></i><b>11.2</b> Libraries</a></li>
<li class="chapter" data-level="11.3" data-path="multiple-regression.html"><a href="multiple-regression.html#examplar-data"><i class="fa fa-check"></i><b>11.3</b> Examplar data</a></li>
<li class="chapter" data-level="11.4" data-path="multiple-regression.html"><a href="multiple-regression.html#data-exploration"><i class="fa fa-check"></i><b>11.4</b> Data exploration</a></li>
<li class="chapter" data-level="11.5" data-path="multiple-regression.html"><a href="multiple-regression.html#d-plots"><i class="fa fa-check"></i><b>11.5</b> 3D Plots</a><ul>
<li class="chapter" data-level="11.5.1" data-path="multiple-regression.html"><a href="multiple-regression.html#scatterplot3d"><i class="fa fa-check"></i><b>11.5.1</b> scatterplot3d</a></li>
<li class="chapter" data-level="11.5.2" data-path="multiple-regression.html"><a href="multiple-regression.html#rgl"><i class="fa fa-check"></i><b>11.5.2</b> rgl</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="multiple-regression.html"><a href="multiple-regression.html#fitting-the-regression-model"><i class="fa fa-check"></i><b>11.6</b> Fitting the regression model</a><ul>
<li class="chapter" data-level="11.6.1" data-path="multiple-regression.html"><a href="multiple-regression.html#visualizing-the-regression-model-in-scatterplot3d"><i class="fa fa-check"></i><b>11.6.1</b> Visualizing the regression model in scatterplot3d</a></li>
</ul></li>
<li class="chapter" data-level="11.7" data-path="multiple-regression.html"><a href="multiple-regression.html#interpretting-the-regression-model"><i class="fa fa-check"></i><b>11.7</b> Interpretting the regression model</a></li>
<li class="chapter" data-level="11.8" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-vector-geometry-of-the-regression-model"><i class="fa fa-check"></i><b>11.8</b> Exploring the Vector Geometry of the Regression Model</a></li>
<li class="chapter" data-level="11.9" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-residuals-from-the-model-fit"><i class="fa fa-check"></i><b>11.9</b> Exploring the Residuals from the Model Fit</a></li>
<li class="chapter" data-level="11.10" data-path="multiple-regression.html"><a href="multiple-regression.html#an-alternate-model"><i class="fa fa-check"></i><b>11.10</b> An alternate model</a></li>
<li class="chapter" data-level="11.11" data-path="multiple-regression.html"><a href="multiple-regression.html#fitting-a-curvilinear-model-using-lm"><i class="fa fa-check"></i><b>11.11</b> Fitting a curvilinear model using lm()</a></li>
<li class="chapter" data-level="11.12" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-impact-of-nearly-collinear-predictors-on-regression"><i class="fa fa-check"></i><b>11.12</b> Exploring the impact of nearly collinear predictors on regression</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html"><i class="fa fa-check"></i><b>12</b> Principal Components Analysis</a><ul>
<li class="chapter" data-level="12.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#libraries-5"><i class="fa fa-check"></i><b>12.1</b> Libraries</a></li>
<li class="chapter" data-level="12.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#matrices-as-linear-transformations"><i class="fa fa-check"></i><b>12.2</b> Matrices as linear transformations</a></li>
<li class="chapter" data-level="12.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#eigenanalysis-in-r"><i class="fa fa-check"></i><b>12.3</b> Eigenanalysis in R</a></li>
<li class="chapter" data-level="12.4" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#principal-components-analysis-in-r"><i class="fa fa-check"></i><b>12.4</b> Principal Components Analysis in R</a><ul>
<li class="chapter" data-level="12.4.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#bioenv-dataset"><i class="fa fa-check"></i><b>12.4.1</b> Bioenv dataset</a></li>
<li class="chapter" data-level="12.4.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pca-of-the-bioenv-dataset"><i class="fa fa-check"></i><b>12.4.2</b> PCA of the Bioenv dataset</a></li>
<li class="chapter" data-level="12.4.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#calculating-factor-loadings"><i class="fa fa-check"></i><b>12.4.3</b> Calculating Factor Loadings</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#drawing-figures-to-represent-pca"><i class="fa fa-check"></i><b>12.5</b> Drawing Figures to Represent PCA</a><ul>
<li class="chapter" data-level="12.5.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pc-score-plots"><i class="fa fa-check"></i><b>12.5.1</b> PC Score Plots</a></li>
<li class="chapter" data-level="12.5.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#simultaneous-depiction-of-observations-and-variables-in-the-pc-space"><i class="fa fa-check"></i><b>12.5.2</b> Simultaneous Depiction of Observations and Variables in the PC Space</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html"><i class="fa fa-check"></i><b>13</b> Simulating Sampling Distributions</a><ul>
<li class="chapter" data-level="13.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#libraries-6"><i class="fa fa-check"></i><b>13.1</b> Libraries</a></li>
<li class="chapter" data-level="13.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#data-set-simulated-male-heights"><i class="fa fa-check"></i><b>13.2</b> Data set: Simulated male heights</a><ul>
<li class="chapter" data-level="13.2.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#seeding-the-pseudo-random-number-generator"><i class="fa fa-check"></i><b>13.2.1</b> Seeding the pseudo-random number generator</a></li>
<li class="chapter" data-level="13.2.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#generating-the-simulated-population"><i class="fa fa-check"></i><b>13.2.2</b> Generating the simulated population</a></li>
<li class="chapter" data-level="13.2.3" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#properties-of-the-simulated-population"><i class="fa fa-check"></i><b>13.2.3</b> Properties of the simulated population</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#random-sampling-from-the-simulated-population"><i class="fa fa-check"></i><b>13.3</b> Random sampling from the simulated population</a><ul>
<li class="chapter" data-level="13.3.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#another-random-sample"><i class="fa fa-check"></i><b>13.3.1</b> Another random sample</a></li>
<li class="chapter" data-level="13.3.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#simulating-the-generation-of-many-random-samples"><i class="fa fa-check"></i><b>13.3.2</b> Simulating the generation of many random samples</a></li>
<li class="chapter" data-level="13.3.3" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#a-function-to-estimate-statistics-of-interest-in-a-random-sample"><i class="fa fa-check"></i><b>13.3.3</b> A function to estimate statistics of interest in a random sample</a></li>
<li class="chapter" data-level="13.3.4" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#generating-statistics-for-many-random-samples"><i class="fa fa-check"></i><b>13.3.4</b> Generating statistics for many random samples</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#simulated-sampling-distribution-of-the-mean"><i class="fa fa-check"></i><b>13.4</b> Simulated sampling distribution of the mean</a><ul>
<li class="chapter" data-level="13.4.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#sampling-distributions-for-different-sample-sizes"><i class="fa fa-check"></i><b>13.4.1</b> Sampling distributions for different sample sizes</a></li>
<li class="chapter" data-level="13.4.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#discussion-of-trends-for-sampling-distributions-of-different-sample-sizes"><i class="fa fa-check"></i><b>13.4.2</b> Discussion of trends for sampling distributions of different sample sizes</a></li>
</ul></li>
<li class="chapter" data-level="13.5" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#standard-error-of-the-mean"><i class="fa fa-check"></i><b>13.5</b> Standard Error of the Mean</a></li>
<li class="chapter" data-level="13.6" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#sampling-distribution-of-the-standard-deviation"><i class="fa fa-check"></i><b>13.6</b> Sampling Distribution of the Standard Deviation</a></li>
<li class="chapter" data-level="13.7" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#what-happens-to-the-sampling-distribution-of-the-mean-and-standard-deviation-when-our-sample-size-is-small"><i class="fa fa-check"></i><b>13.7</b> What happens to the sampling distribution of the mean and standard deviation when our sample size is small?</a><ul>
<li class="chapter" data-level="13.7.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#for-small-samples-sample-standard-deviations-systematically-underestimate-the-population-standard-deviation"><i class="fa fa-check"></i><b>13.7.1</b> For small samples, sample standard deviations systematically underestimate the population standard deviation</a></li>
<li class="chapter" data-level="13.7.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#underestimates-of-the-standard-deviation-given-small-n-lead-to-understimates-of-the-se-of-the-mean"><i class="fa fa-check"></i><b>13.7.2</b> Underestimates of the standard deviation given small <span class="math inline">\(n\)</span> lead to understimates of the SE of the mean</a></li>
</ul></li>
<li class="chapter" data-level="13.8" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#the-t-distribution-is-the-appropriate-distribution-for-describing-the-sampling-distribution-of-the-mean-when-n-is-small"><i class="fa fa-check"></i><b>13.8</b> The t-distribution is the appropriate distribution for describing the sampling distribution of the mean when <span class="math inline">\(n\)</span> is small</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html"><i class="fa fa-check"></i><b>14</b> Simulating confidence intervals</a><ul>
<li class="chapter" data-level="14.1" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#libraries-7"><i class="fa fa-check"></i><b>14.1</b> Libraries</a></li>
<li class="chapter" data-level="14.2" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#confidence-intervals"><i class="fa fa-check"></i><b>14.2</b> Confidence Intervals</a></li>
<li class="chapter" data-level="14.3" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#calibrating-confidence-intervals"><i class="fa fa-check"></i><b>14.3</b> Calibrating confidence intervals</a></li>
<li class="chapter" data-level="14.4" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#standard-formulation-for-confidence-intervals"><i class="fa fa-check"></i><b>14.4</b> Standard formulation for confidence intervals</a></li>
<li class="chapter" data-level="14.5" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#simulating-confidence-intervals-for-the-mean"><i class="fa fa-check"></i><b>14.5</b> Simulating confidence intervals for the mean</a></li>
<li class="chapter" data-level="14.6" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#sample-estimate-of-the-standard-error-of-the-mean"><i class="fa fa-check"></i><b>14.6</b> Sample estimate of the standard error of the mean</a></li>
<li class="chapter" data-level="14.7" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#ci-of-mean-simulation"><i class="fa fa-check"></i><b>14.7</b> CI of mean simulation</a></li>
<li class="chapter" data-level="14.8" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#generating-a-table-of-cis-and-corresponding-margins-of-error"><i class="fa fa-check"></i><b>14.8</b> Generating a table of CIs and corresponding margins of error</a></li>
<li class="chapter" data-level="14.9" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#interpreting-our-simulation"><i class="fa fa-check"></i><b>14.9</b> Interpreting our simulation</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html"><i class="fa fa-check"></i><b>15</b> Randomization, Jackknife, and Bootstrap</a><ul>
<li class="chapter" data-level="15.1" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#libraries-8"><i class="fa fa-check"></i><b>15.1</b> Libraries</a></li>
<li class="chapter" data-level="15.2" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#randomization"><i class="fa fa-check"></i><b>15.2</b> Randomization</a></li>
<li class="chapter" data-level="15.3" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#using-randomization-to-test-for-a-difference-in-means"><i class="fa fa-check"></i><b>15.3</b> Using randomization to test for a difference in means</a></li>
<li class="chapter" data-level="15.4" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#using-randomization-to-test-for-equality-of-variances"><i class="fa fa-check"></i><b>15.4</b> Using randomization to test for equality of variances</a></li>
<li class="chapter" data-level="15.5" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#jackknifing-in-r"><i class="fa fa-check"></i><b>15.5</b> Jackknifing in R</a></li>
<li class="chapter" data-level="15.6" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#bootstrapping-in-r"><i class="fa fa-check"></i><b>15.6</b> Bootstrapping in R</a><ul>
<li class="chapter" data-level="15.6.1" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#a-more-sophisticated-application-of-the-bootstrap"><i class="fa fa-check"></i><b>15.6.1</b> A more sophisticated application of the bootstrap</a></li>
<li class="chapter" data-level="15.6.2" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#about-logistic-regression"><i class="fa fa-check"></i><b>15.6.2</b> About logistic regression</a></li>
<li class="chapter" data-level="15.6.3" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#bumpus-house-sparrow-data-set"><i class="fa fa-check"></i><b>15.6.3</b> Bumpus house sparrow data set</a></li>
<li class="chapter" data-level="15.6.4" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#predicting-survival-as-a-function-of-body-weight"><i class="fa fa-check"></i><b>15.6.4</b> Predicting survival as a function of body weight</a></li>
<li class="chapter" data-level="15.6.5" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#fitting-the-logistic-regression"><i class="fa fa-check"></i><b>15.6.5</b> Fitting the logistic regression</a></li>
<li class="chapter" data-level="15.6.6" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#defining-an-appropriate-function-for-boot"><i class="fa fa-check"></i><b>15.6.6</b> Defining an appropriate function for <code>boot</code></a></li>
<li class="chapter" data-level="15.6.7" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#calculating-bootstrap-confidence-intervals"><i class="fa fa-check"></i><b>15.6.7</b> Calculating bootstrap confidence intervals</a></li>
<li class="chapter" data-level="15.6.8" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#visualizing-bootstrap-confidence-intervals"><i class="fa fa-check"></i><b>15.6.8</b> Visualizing bootstrap confidence intervals</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="16" data-path="anova.html"><a href="anova.html"><i class="fa fa-check"></i><b>16</b> ANOVA</a><ul>
<li class="chapter" data-level="16.1" data-path="anova.html"><a href="anova.html#libraries-9"><i class="fa fa-check"></i><b>16.1</b> Libraries</a></li>
<li class="chapter" data-level="16.2" data-path="anova.html"><a href="anova.html#example-data-set-the-effect-of-light-treatments-on-circadian-rhythms"><i class="fa fa-check"></i><b>16.2</b> Example data set: The effect of light treatments on circadian rhythms</a></li>
<li class="chapter" data-level="16.3" data-path="anova.html"><a href="anova.html#visualizing-the-data"><i class="fa fa-check"></i><b>16.3</b> Visualizing the data</a></li>
<li class="chapter" data-level="16.4" data-path="anova.html"><a href="anova.html#analysis-of-variance"><i class="fa fa-check"></i><b>16.4</b> Analysis of Variance</a><ul>
<li class="chapter" data-level="16.4.1" data-path="anova.html"><a href="anova.html#hypotheses-for-anova"><i class="fa fa-check"></i><b>16.4.1</b> Hypotheses for ANOVA</a></li>
<li class="chapter" data-level="16.4.2" data-path="anova.html"><a href="anova.html#anova-assumptions"><i class="fa fa-check"></i><b>16.4.2</b> ANOVA, assumptions</a></li>
<li class="chapter" data-level="16.4.3" data-path="anova.html"><a href="anova.html#anova-key-idea"><i class="fa fa-check"></i><b>16.4.3</b> ANOVA, key idea</a></li>
<li class="chapter" data-level="16.4.4" data-path="anova.html"><a href="anova.html#partioning-of-sum-of-squares"><i class="fa fa-check"></i><b>16.4.4</b> Partioning of sum of squares</a></li>
</ul></li>
<li class="chapter" data-level="16.5" data-path="anova.html"><a href="anova.html#the-aov-function"><i class="fa fa-check"></i><b>16.5</b> The <code>aov</code> function</a></li>
<li class="chapter" data-level="16.6" data-path="anova.html"><a href="anova.html#anova-calculations-step-by-step"><i class="fa fa-check"></i><b>16.6</b> ANOVA calculations: Step-by-step</a><ul>
<li class="chapter" data-level="16.6.1" data-path="anova.html"><a href="anova.html#total-sum-of-squares"><i class="fa fa-check"></i><b>16.6.1</b> Total sum of squares</a></li>
<li class="chapter" data-level="16.6.2" data-path="anova.html"><a href="anova.html#group-sum-of-squares-and-mean-square"><i class="fa fa-check"></i><b>16.6.2</b> Group sum of squares and mean square</a></li>
<li class="chapter" data-level="16.6.3" data-path="anova.html"><a href="anova.html#error-sum-of-squares-and-mean-square"><i class="fa fa-check"></i><b>16.6.3</b> Error sum of squares and mean square</a></li>
<li class="chapter" data-level="16.6.4" data-path="anova.html"><a href="anova.html#calculating-the-f-statistic"><i class="fa fa-check"></i><b>16.6.4</b> Calculating the F-statistic</a></li>
<li class="chapter" data-level="16.6.5" data-path="anova.html"><a href="anova.html#the-f-distribution"><i class="fa fa-check"></i><b>16.6.5</b> The F-distribution</a></li>
<li class="chapter" data-level="16.6.6" data-path="anova.html"><a href="anova.html#critical-values-of-the-f-distribution"><i class="fa fa-check"></i><b>16.6.6</b> Critical values of the F-distribution</a></li>
</ul></li>
<li class="chapter" data-level="16.7" data-path="anova.html"><a href="anova.html#combined-visualization"><i class="fa fa-check"></i><b>16.7</b> Combined visualization</a></li>
<li class="chapter" data-level="16.8" data-path="anova.html"><a href="anova.html#which-pairs-of-group-means-are-different"><i class="fa fa-check"></i><b>16.8</b> Which pairs of group means are different?</a><ul>
<li class="chapter" data-level="16.8.1" data-path="anova.html"><a href="anova.html#tukey-kramer-test"><i class="fa fa-check"></i><b>16.8.1</b> Tukey-Kramer test</a></li>
</ul></li>
<li class="chapter" data-level="16.9" data-path="anova.html"><a href="anova.html#anova-as-regression"><i class="fa fa-check"></i><b>16.9</b> ANOVA as Regression</a><ul>
<li class="chapter" data-level="16.9.1" data-path="anova.html"><a href="anova.html#treating-a-variable-as-a-factor"><i class="fa fa-check"></i><b>16.9.1</b> Treating a variable as a factor</a></li>
<li class="chapter" data-level="16.9.2" data-path="anova.html"><a href="anova.html#factor-coding"><i class="fa fa-check"></i><b>16.9.2</b> Factor coding</a></li>
<li class="chapter" data-level="16.9.3" data-path="anova.html"><a href="anova.html#fitting-the-anova-using-lm"><i class="fa fa-check"></i><b>16.9.3</b> Fitting the ANOVA using <code>lm</code></a></li>
<li class="chapter" data-level="16.9.4" data-path="anova.html"><a href="anova.html#changing-the-factor-coding"><i class="fa fa-check"></i><b>16.9.4</b> Changing the factor coding</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="17" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html"><i class="fa fa-check"></i><b>17</b> Canonical Variates Analysis</a><ul>
<li class="chapter" data-level="17.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#libraries-10"><i class="fa fa-check"></i><b>17.1</b> Libraries</a></li>
<li class="chapter" data-level="17.2" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#discriminant-analysis-in-r"><i class="fa fa-check"></i><b>17.2</b> Discriminant Analysis in R</a><ul>
<li class="chapter" data-level="17.2.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#shorthand-formulae-in-r"><i class="fa fa-check"></i><b>17.2.1</b> Shorthand Formulae in R</a></li>
<li class="chapter" data-level="17.2.2" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#working-with-the-output-of-lda"><i class="fa fa-check"></i><b>17.2.2</b> Working with the output of <code>lda()</code></a></li>
</ul></li>
<li class="chapter" data-level="17.3" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#estimating-confidence-regions-for-group-means-in-cva"><i class="fa fa-check"></i><b>17.3</b> Estimating confidence regions for group means in CVA</a><ul>
<li class="chapter" data-level="17.3.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#drawing-the-cva-confidence-regions"><i class="fa fa-check"></i><b>17.3.1</b> Drawing the CVA confidence regions</a></li>
</ul></li>
<li class="chapter" data-level="17.4" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#calculating-the-within-and-between-group-covariance-matrices"><i class="fa fa-check"></i><b>17.4</b> Calculating the Within and Between Group Covariance Matrices</a><ul>
<li class="chapter" data-level="17.4.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#recapitulating-the-cva-analysis-of-lda"><i class="fa fa-check"></i><b>17.4.1</b> Recapitulating the CVA analysis of <code>lda()</code></a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Biology 723: Statistical Computing for Biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="anova" class="section level1">
<h1><span class="header-section-number">Chapter 16</span> ANOVA</h1>
<div id="libraries-9" class="section level2">
<h2><span class="header-section-number">16.1</span> Libraries</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)
<span class="kw">library</span>(cowplot)
<span class="kw">library</span>(broom)
<span class="kw">library</span>(magrittr)</code></pre></div>
</div>
<div id="example-data-set-the-effect-of-light-treatments-on-circadian-rhythms" class="section level2">
<h2><span class="header-section-number">16.2</span> Example data set: The effect of light treatments on circadian rhythms</h2>
<p>Whitlock &amp; Schluter (Analysis of Biological Data) describe an examplar data set from a study designed to test the effects of light treatment on circadian rhythms (see Whitlock &amp; Schluter, Example 15.1).</p>
<ul>
<li><p>The investigators randomly assigned 22 individuals to one of three treatment groups and measured phase shifts in melatonin production. The treatment groups were:</p>
<ul>
<li>control group (8 individuals)</li>
<li>light applied on the back of the knee (7 individuals)</li>
<li>light applied to the eyes (7 individuals)</li>
</ul></li>
</ul>
<p>These data are available at: <a href="https://raw.githubusercontent.com/bio304-class/bio304-fall2017/master/datasets/ABD-circadian-rythms.csv">ABD-circadian-rythms.csv</a></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">circadian &lt;-<span class="st"> </span><span class="kw">read_csv</span>(<span class="st">&quot;https://raw.githubusercontent.com/bio304-class/bio304-fall2017/master/datasets/ABD-circadian-rythms.csv&quot;</span>)</code></pre></div>
<p>The data set has two columns, <code>treatment</code> and <code>shift</code>. <code>treatment</code> is a categorical variable indicating the particular light treatment that each individual received; <code>shift</code> is the measured phase shift in circadian cycles.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(circadian)
<span class="co"># A tibble: 6 x 2</span>
  treatment   shift
  <span class="op">&lt;</span>chr<span class="op">&gt;</span><span class="st">       </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="dv">1</span> control    <span class="fl">0.5300</span>
<span class="dv">2</span> control    <span class="fl">0.3600</span>
<span class="dv">3</span> control    <span class="fl">0.2000</span>
<span class="dv">4</span> control   <span class="op">-</span><span class="fl">0.3700</span>
<span class="dv">5</span> control   <span class="op">-</span><span class="fl">0.6000</span>
<span class="dv">6</span> control   <span class="op">-</span><span class="fl">0.6400</span></code></pre></div>
</div>
<div id="visualizing-the-data" class="section level2">
<h2><span class="header-section-number">16.3</span> Visualizing the data</h2>
<p>As we usually do, let’s start by visualizing the data. We’ll create a plot depicting the observations (as points), colored by treatment group, along with a set of vertical lines that indicate how an individual observation differs from the overall mean of the data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># pre-compute grand mean because we&#39;ll be using it repeatedly</span>
grand.mean &lt;-<span class="st"> </span><span class="kw">mean</span>(circadian<span class="op">$</span>shift)

total.plot &lt;-
<span class="st">  </span>circadian <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x=</span>treatment, <span class="dt">y=</span>shift, 
             <span class="dt">color=</span>treatment, <span class="dt">group=</span><span class="kw">row.names</span>(circadian))) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> grand.mean, <span class="dt">linetype=</span><span class="st">&#39;dashed&#39;</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">position =</span> <span class="kw">position_dodge</span>(<span class="fl">0.5</span>)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_linerange</span>(<span class="kw">aes</span>(<span class="dt">ymin =</span> grand.mean, <span class="dt">ymax =</span> shift), 
                 <span class="dt">position =</span> <span class="kw">position_dodge</span>(<span class="fl">0.5</span>)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">ylim</span>(<span class="op">-</span><span class="dv">3</span>,<span class="dv">1</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Treatment&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Phase shift (h)&quot;</span>,
       <span class="dt">title =</span> <span class="st">&quot;Deviation of observations around the grand mean (dashed line)&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">theme</span>(<span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">9</span>))

total.plot</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-455-1.png" width="576" /></p>
</div>
<div id="analysis-of-variance" class="section level2">
<h2><span class="header-section-number">16.4</span> Analysis of Variance</h2>
<p><span class="math inline">\(t\)</span>-tests are the standard approach for comparing means between two groups. When you want to compare means between more than two groups the standard approach is “Analysis of Variance” (ANOVA).</p>
<div id="hypotheses-for-anova" class="section level3">
<h3><span class="header-section-number">16.4.1</span> Hypotheses for ANOVA</h3>
<p>When using ANOVA to compare means, the null and alternative hypotheses are:</p>
<ul>
<li><span class="math inline">\(H_0\)</span>: The means of all the groups are equal</li>
<li><span class="math inline">\(H_A\)</span>: At least one of the means is different from the others</li>
</ul>
</div>
<div id="anova-assumptions" class="section level3">
<h3><span class="header-section-number">16.4.2</span> ANOVA, assumptions</h3>
<p>ANOVA assumes:</p>
<ul>
<li>The measurements in every group represent a random sample from the corresponding population</li>
<li>The varaible of interest is normally distributed</li>
<li>The variance is approximately the same in all the groups</li>
</ul>
</div>
<div id="anova-key-idea" class="section level3">
<h3><span class="header-section-number">16.4.3</span> ANOVA, key idea</h3>
<p>The key idea behind ANOVA is that:</p>
<ul>
<li><p>If the observations in each group are drawn from populations with equal means (and variances) then the variation <em>between</em> group means should be similar to the inter-individual variation <em>within groups</em>.</p></li>
<li><p>The test statistic used in ANOVA is designated <span class="math inline">\(F\)</span>, and is based on the ratio of two measures of variance, the “group mean square deviation” (<span class="math inline">\(\text{MS}_\text{groups}\)</span>; measures between group variance) and the “error mean square deviation” (<span class="math inline">\(\text{MS}_\text{error}\)</span>; measures within group variance):</p></li>
</ul>
<p><span class="math display">\[
F = \frac{\text{MS}_\text{group}}{\text{MS}_\text{error}}
\]</span></p>
<ul>
<li><p>Under the null hypothesis, the between group and within group variances are similar and thus the <span class="math inline">\(F\)</span> statistic should be approximately 1.</p></li>
<li><p>Large values of the <span class="math inline">\(F\)</span>-statistic means that the between group variance exceeds the within group variance, indicating that at least one of the means is different from the others</p></li>
</ul>
</div>
<div id="partioning-of-sum-of-squares" class="section level3">
<h3><span class="header-section-number">16.4.4</span> Partioning of sum of squares</h3>
<p>Another way to think about ANOVA is as a “partitioning of variance”. The total variance among all the individuals across groups can be decomposed into:</p>
<ol style="list-style-type: decimal">
<li>variance of the group means around the “grand mean”</li>
<li>variance of individuals around the group means.</li>
</ol>
<p>However, rather than using variance we use sums of square deviations around the respectives means (usually shortened to “sums of squares”).</p>
</div>
</div>
<div id="the-aov-function" class="section level2">
<h2><span class="header-section-number">16.5</span> The <code>aov</code> function</h2>
<p>As you would suspect, there is a built in R function to carry out ANOVA. This function is designated <code>aov()</code>. <code>aov</code> takes a formula style argument where the variable of interest is on the left, and the grouping variable indicated on the right.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">anova.circadian &lt;-<span class="st"> </span><span class="kw">aov</span>(shift <span class="op">~</span><span class="st"> </span>treatment, <span class="dt">data =</span> circadian)</code></pre></div>
<p>The <code>summary</code> function applied to the <code>aov</code> fit will print out a typical ANOVA table and calculate the associated P-value for the <span class="math inline">\(F\)</span> test statistic:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(anova.circadian)
            Df Sum Sq Mean Sq F value  <span class="kw">Pr</span>(<span class="op">&gt;</span>F)   
treatment    <span class="dv">2</span>  <span class="fl">7.224</span>   <span class="fl">3.612</span>   <span class="fl">7.289</span> <span class="fl">0.00447</span> <span class="op">**</span>
Residuals   <span class="dv">19</span>  <span class="fl">9.415</span>   <span class="fl">0.496</span>                   
<span class="op">---</span>
Signif. codes<span class="op">:</span><span class="st">  </span><span class="dv">0</span> <span class="st">&#39;***&#39;</span> <span class="fl">0.001</span> <span class="st">&#39;**&#39;</span> <span class="fl">0.01</span> <span class="st">&#39;*&#39;</span> <span class="fl">0.05</span> <span class="st">&#39;.&#39;</span> <span class="fl">0.1</span> <span class="st">&#39; &#39;</span> <span class="dv">1</span></code></pre></div>
<p>If you want the ANOVA table in a form you can compute with, the <code>broom::tidy</code> function comes in handy:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">tidy</span>(anova.circadian)
       term df    sumsq    meansq statistic     p.value
<span class="dv">1</span> treatment  <span class="dv">2</span> <span class="fl">7.224492</span> <span class="fl">3.6122459</span>  <span class="fl">7.289449</span> <span class="fl">0.004472271</span>
<span class="dv">2</span> Residuals <span class="dv">19</span> <span class="fl">9.415345</span> <span class="fl">0.4955445</span>        <span class="ot">NA</span>          <span class="ot">NA</span></code></pre></div>
</div>
<div id="anova-calculations-step-by-step" class="section level2">
<h2><span class="header-section-number">16.6</span> ANOVA calculations: Step-by-step</h2>
<p>The <code>aov()</code> function carries out all the ANOVA calculations behind the scenes. It’s useful to pull back the curtain and see how the various quantities are calculated.</p>
<div id="total-sum-of-squares" class="section level3">
<h3><span class="header-section-number">16.6.1</span> Total sum of squares</h3>
<p>We call the sum of the squared deviations around the grand mean the “total sum of sqaures” (<span class="math inline">\(SS_\text{total}\)</span>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># total sum of squares</span>
total.table &lt;-<span class="st"> </span>
<span class="st">  </span>circadian <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">sum.squares =</span> <span class="kw">sum</span>((shift <span class="op">-</span><span class="st"> </span>grand.mean)<span class="op">**</span><span class="dv">2</span>),
            <span class="dt">df =</span> <span class="kw">n</span>() <span class="op">-</span><span class="st"> </span><span class="dv">1</span>)

total.table
<span class="co"># A tibble: 1 x 2</span>
  sum.squares    df
        <span class="op">&lt;</span>dbl<span class="op">&gt;</span><span class="st"> </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="dv">1</span>       <span class="fl">16.64</span>   <span class="dv">21</span>.</code></pre></div>
<p>In vector terms, <span class="math inline">\(SS_\text{total}\)</span> is simply the squared length of the mean centered vector, which is easily calculated using the dot product:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">shift.ctrd &lt;-<span class="st"> </span>circadian<span class="op">$</span>shift <span class="op">-</span><span class="st"> </span><span class="kw">mean</span>(circadian<span class="op">$</span>shift)
sum.squares.total &lt;-<span class="st"> </span>shift.ctrd <span class="op">%*%</span><span class="st"> </span>shift.ctrd
sum.squares.total
         [,<span class="dv">1</span>]
[<span class="dv">1</span>,] <span class="fl">16.63984</span></code></pre></div>
</div>
<div id="group-sum-of-squares-and-mean-square" class="section level3">
<h3><span class="header-section-number">16.6.2</span> Group sum of squares and mean square</h3>
<p>Next we turn to variation of the group means around the grand mean. We use <code>group_by</code> and <code>summarize</code> to calculates group means and the group deviates (the difference between the group means and the grand mean):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">group.df &lt;-
<span class="st">  </span>circadian <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">group_by</span>(treatment) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">n =</span> <span class="kw">n</span>(), 
            <span class="dt">group.mean =</span> <span class="kw">mean</span>(shift), 
            <span class="dt">grand.mean =</span> grand.mean,
            <span class="dt">group.deviates =</span> group.mean <span class="op">-</span><span class="st"> </span>grand.mean)</code></pre></div>
<p>Let’s visualize the difference of the group means from the grand mean:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">group.plot &lt;-
<span class="st">  </span>group.df <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> treatment, <span class="dt">y =</span> group.mean, <span class="dt">color=</span>treatment)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_linerange</span>(<span class="kw">aes</span>(<span class="dt">ymin =</span> grand.mean, <span class="dt">ymax =</span> group.mean), <span class="dt">size=</span><span class="dv">2</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="dv">3</span>, <span class="dt">alpha =</span> <span class="fl">0.25</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> grand.mean, <span class="dt">linetype=</span><span class="st">&#39;dashed&#39;</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">ylim</span>(<span class="op">-</span><span class="dv">3</span>,<span class="dv">1</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Treatment&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Phase shift (h)&quot;</span>,
         <span class="dt">title =</span> <span class="st">&quot;Deviation of group means around the grand mean&quot;</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">9</span>))

group.plot</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-462-1.png" width="576" /></p>
<p>Now we calculate the group sum of squares (<span class="math inline">\(SS_\text{group}\)</span>) and the group mean square (<span class="math inline">\(MS_\text{group}\)</span>). This calculation takes into account the size of each group (for the group sum of squares) and the degrees of freedom associated with the number of groups (for the group mean square).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">group.table &lt;-
<span class="st">  </span>group.df <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">SS =</span> <span class="kw">sum</span>(n <span class="op">*</span><span class="st"> </span>group.deviates<span class="op">**</span><span class="dv">2</span>),
            <span class="dt">k =</span> <span class="kw">n</span>(),
            <span class="dt">df =</span> k<span class="op">-</span><span class="dv">1</span>,
            <span class="dt">MS =</span> SS<span class="op">/</span>df)</code></pre></div>
</div>
<div id="error-sum-of-squares-and-mean-square" class="section level3">
<h3><span class="header-section-number">16.6.3</span> Error sum of squares and mean square</h3>
<p>Next we turn to variation of the individual observations around the group means, which is the basis of the error sum of squares and mean square.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">error.df &lt;-
<span class="st">  </span>circadian <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">group_by</span>(treatment) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">group.mean =</span> <span class="kw">mean</span>(shift),
         <span class="dt">error.deviates =</span> shift <span class="op">-</span><span class="st"> </span>group.mean) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">SS =</span> <span class="kw">sum</span>(error.deviates<span class="op">**</span><span class="dv">2</span>),
            <span class="dt">n =</span> <span class="kw">n</span>())</code></pre></div>
<p>We can visualize these individual deviates around the group means as so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">error.plot &lt;-
<span class="st">  </span>circadian <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">group_by</span>(treatment) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">group.mean =</span> <span class="kw">mean</span>(shift)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> treatment, <span class="dt">y =</span> shift, <span class="dt">color =</span> treatment, <span class="dt">group=</span><span class="kw">row.names</span>(circadian))) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">y =</span> group.mean),<span class="dt">size=</span><span class="dv">3</span>,<span class="dt">alpha=</span><span class="fl">0.1</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_linerange</span>(<span class="kw">aes</span>(<span class="dt">ymin =</span> group.mean, <span class="dt">ymax =</span> shift), <span class="dt">position =</span> <span class="kw">position_dodge</span>(<span class="fl">0.5</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">ylim</span>(<span class="op">-</span><span class="dv">3</span>,<span class="dv">1</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Treatment&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Phase shift (h)&quot;</span>,
         <span class="dt">title =</span> <span class="st">&quot;Deviation of observations around the groups means&quot;</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">size=</span><span class="dv">9</span>))

error.plot</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-465-1.png" width="576" /></p>
<p>Now we calculate the error sum of squares (<span class="math inline">\(SS_\text{error}\)</span>) and the error mean square (<span class="math inline">\(MS_\text{error}\)</span>). Here the degrees of freedom is the total sample size minus the number of groups.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">error.table &lt;-
<span class="st">  </span>error.df <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">SS =</span> <span class="kw">sum</span>(SS),
            <span class="dt">k =</span> <span class="kw">n</span>(),
            <span class="dt">N =</span> <span class="kw">sum</span>(n),
            <span class="dt">df =</span> N <span class="op">-</span><span class="st"> </span>k,
            <span class="dt">MS =</span> SS<span class="op">/</span>df)</code></pre></div>
</div>
<div id="calculating-the-f-statistic" class="section level3">
<h3><span class="header-section-number">16.6.4</span> Calculating the F-statistic</h3>
<p>Having calculated our estimates of between group variance and within group variance (<span class="math inline">\(MS_\text{group}\)</span> and $MS_) we’re now ready to calculate the <span class="math inline">\(F\)</span> test statistic.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">F.stat &lt;-<span class="st"> </span>group.table<span class="op">$</span>MS<span class="op">/</span>error.table<span class="op">$</span>MS
F.stat
[<span class="dv">1</span>] <span class="fl">7.289449</span></code></pre></div>
</div>
<div id="the-f-distribution" class="section level3">
<h3><span class="header-section-number">16.6.5</span> The F-distribution</h3>
<p>Our calculated F statistic is much larger than 1. To calculate the probability of observing an F-statistic this large or greater under the null hypothesis, we can need to examine the F-distribution.</p>
<p>The F-distribution has two degree of freedom parameters, indicating the degrees of freedom associated with the group variance and the degrees of freedom associated with the error variance. Here is an illustration of the F-distribution, <span class="math inline">\(F_{2,19}\)</span></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">f &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dv">0</span>, <span class="dv">10</span>, <span class="dt">length.out =</span> <span class="dv">200</span>)
df1 &lt;-<span class="st"> </span><span class="dv">2</span>
df2 &lt;-<span class="st"> </span><span class="dv">19</span>
f.density &lt;-<span class="st"> </span><span class="kw">df</span>(f, df1, df2)

<span class="kw">ggplot</span>(<span class="kw">data_frame</span>(<span class="dt">F =</span> f, <span class="dt">Density =</span> f.density), <span class="kw">aes</span>(<span class="dt">x =</span> F, <span class="dt">y =</span> Density)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&quot;F-distribution with df1=2, df2 = 19&quot;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-468-1.png" width="576" /></p>
<p>We can use the F distribution function, <code>pf()</code> to lookup the probability of getting a value of 7.2894487 or larger under the null hypothesis:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># degrees of freedom for group and error sum of squares</span>
df.group &lt;-<span class="st"> </span>group.table<span class="op">$</span>df
df.error &lt;-<span class="st"> </span>error.table<span class="op">$</span>df

<span class="co"># use the F distribution function </span>
<span class="kw">pf</span>(F.stat, df.group, df.error, <span class="dt">lower.tail =</span> <span class="ot">FALSE</span>)
[<span class="dv">1</span>] <span class="fl">0.004472271</span></code></pre></div>
</div>
<div id="critical-values-of-the-f-distribution" class="section level3">
<h3><span class="header-section-number">16.6.6</span> Critical values of the F-distribution</h3>
<p>If we wanted to know what the critical F value is for a corresponding type I error rate we can use the <code>qf()</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># the critical value of F for alpha = 0.05</span>
<span class="kw">qf</span>(<span class="fl">0.05</span>, df.group, df.error, <span class="dt">lower.tail =</span> <span class="ot">FALSE</span>)
[<span class="dv">1</span>] <span class="fl">3.521893</span></code></pre></div>
</div>
</div>
<div id="combined-visualization" class="section level2">
<h2><span class="header-section-number">16.7</span> Combined visualization</h2>
<p>We can combine our three plots created above into a single figure using <code>cowplot::plot_grid</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">combined.plot &lt;-<span class="kw">plot_grid</span>(total.plot, group.plot, error.plot, 
                          <span class="dt">labels =</span> <span class="kw">c</span>(<span class="st">&quot;A&quot;</span>,<span class="st">&quot;B&quot;</span>,<span class="st">&quot;C&quot;</span>), <span class="dt">nrow =</span> <span class="dv">1</span>)
combined.plot</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-471-1.png" width="1920" /></p>
</div>
<div id="which-pairs-of-group-means-are-different" class="section level2">
<h2><span class="header-section-number">16.8</span> Which pairs of group means are different?</h2>
<p>If an ANOVA indicates that at least one of the group means is different than the others, the next question is usually “which pairs are different?”. There are slighly different tests for what are called “planned” versus “unplanned” comparisons. Your textbook discusses the differences between these two types</p>
<p>Here we focus on a common test for unplanned comparisons, called the Tukey Honest Significant Differences test (referred to as the Tukey-Kramer test in your textbook). The Tukey HSD test controls for the “family-wise error rate”, meaning it tries to keep the overall false positive (Type I error) rate at a specified value.</p>
<div id="tukey-kramer-test" class="section level3">
<h3><span class="header-section-number">16.8.1</span> Tukey-Kramer test</h3>
<p>The function <code>TukeyHSD</code> implements the Tukey-Kramer test. The input to <code>TukeyHSD</code> is the fit from <code>aov</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">TukeyHSD</span>(anova.circadian)
  Tukey multiple comparisons of means
    <span class="dv">95</span>% family<span class="op">-</span>wise confidence level

Fit<span class="op">:</span><span class="st"> </span><span class="kw">aov</span>(<span class="dt">formula =</span> shift <span class="op">~</span><span class="st"> </span>treatment, <span class="dt">data =</span> circadian)

<span class="op">$</span>treatment
                    diff        lwr        upr     p adj
eyes<span class="op">-</span>control <span class="op">-</span><span class="fl">1.24267857</span> <span class="op">-</span><span class="fl">2.1682364</span> <span class="op">-</span><span class="fl">0.3171207</span> <span class="fl">0.0078656</span>
knee<span class="op">-</span>control <span class="op">-</span><span class="fl">0.02696429</span> <span class="op">-</span><span class="fl">0.9525222</span>  <span class="fl">0.8985936</span> <span class="fl">0.9969851</span>
knee<span class="op">-</span>eyes     <span class="fl">1.21571429</span>  <span class="fl">0.2598022</span>  <span class="fl">2.1716263</span> <span class="fl">0.0116776</span></code></pre></div>
<p>Here again, the <code>broom::tidy</code> function comes in handy:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">tidy</span>(<span class="kw">TukeyHSD</span>(anova.circadian))
       term   comparison    estimate   conf.low  conf.high adj.p.value
<span class="dv">1</span> treatment eyes<span class="op">-</span>control <span class="op">-</span><span class="fl">1.24267857</span> <span class="op">-</span><span class="fl">2.1682364</span> <span class="op">-</span><span class="fl">0.3171207</span> <span class="fl">0.007865623</span>
<span class="dv">2</span> treatment knee<span class="op">-</span>control <span class="op">-</span><span class="fl">0.02696429</span> <span class="op">-</span><span class="fl">0.9525222</span>  <span class="fl">0.8985936</span> <span class="fl">0.996985102</span>
<span class="dv">3</span> treatment    knee<span class="op">-</span>eyes  <span class="fl">1.21571429</span>  <span class="fl">0.2598022</span>  <span class="fl">2.1716263</span> <span class="fl">0.011677556</span></code></pre></div>
<p>The Tukey HSD test by default give us 95% confidence intervals for the differences in means between each pair of groups, and an associated P-value for the null hypothesis of equal means between pairs. Interpretting the results above, we see that we fail to reject the null hypothesis of equal means for the knee and control treatment groups (i.e. we have no statistical support to conclude they are different). However, we reject the null hypothesis for equality of means between control and eye treatments and between knee and eye treatements. We have evidence that light treatments applied to the eye cause a mean negative shift in the phase of melatonin production relative to control and knee treatment groups.</p>
</div>
</div>
<div id="anova-as-regression" class="section level2">
<h2><span class="header-section-number">16.9</span> ANOVA as Regression</h2>
<p>The <code>aov()</code> function we introduced earlier is a wrapper around the <code>lm()</code>. For complex models it can be useful to know how to specify the ANOVA model in the form of a regression. In order to carryout ANOVA using regression, we need to code the group structure in the form of a matrix we can use in the regression.</p>
<p>In lecture we described two simple coding systems – “dummy coding” and “effect coding.” We will illustrate <a href="https://stats.idre.ucla.edu/other/mult-pkg/faq/general/faqwhat-is-dummy-coding/">effect coding</a> here. These two websites – <a href="https://stats.idre.ucla.edu/r/library/r-library-contrast-coding-systems-for-categorical-variables/">UCLA Stats website</a> and <a href="http://faculty.cas.usf.edu/mbrannick/regression/anova1.html">USF course on regression</a> – have additional useful information about other methods for constructing coding matrices, and their uses and interpretation.</p>
<div id="treating-a-variable-as-a-factor" class="section level3">
<h3><span class="header-section-number">16.9.1</span> Treating a variable as a factor</h3>
<p>First, we need to make sure our treatment variable is viewed as a factor by R. We’ll create a new column, <code>treatment.factor</code>, to represent this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">circadian &lt;-
<span class="st">  </span>circadian <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">treatment.factor =</span> <span class="kw">as.factor</span>(treatment))</code></pre></div>
<p>When you create a factor, R automatically maps the different values of the variable to factor levels:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">levels</span>(circadian<span class="op">$</span>treatment.factor)
[<span class="dv">1</span>] <span class="st">&quot;control&quot;</span> <span class="st">&quot;eyes&quot;</span>    <span class="st">&quot;knee&quot;</span>   </code></pre></div>
<p>In the above case, “control” gets mapped to level 1, “eyes” to level 2, etc. #### Changing factor levels</p>
<p>As an aside, if we wanted to change the leveling, the function <code>fct_relevel</code> defined in the <code>forcats</code> package (part of the tidyverse) is useful:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">fct_relevel</span>(circadian<span class="op">$</span>treatment.factor,
            <span class="st">&quot;eyes&quot;</span>, <span class="st">&quot;knee&quot;</span>, <span class="st">&quot;control&quot;</span>)
 [<span class="dv">1</span>] control control control control control control control control
 [<span class="dv">9</span>] knee    knee    knee    knee    knee    knee    knee    eyes   
[<span class="dv">17</span>] eyes    eyes    eyes    eyes    eyes    eyes   
Levels<span class="op">:</span><span class="st"> </span>eyes knee control</code></pre></div>
<p>Now “eyes” is level 1, “knee” is level 2, etc. Note that <code>fct_level()</code> doesn’t changing the leveling of <code>circadian$treatment.factor</code> itself, but rather returns a new factor with the same values mapped to the new levels.</p>
<p>For our purposes treating the “control” group as the level 1 factor is convenient, so we’ll keep that leveling for now.</p>
</div>
<div id="factor-coding" class="section level3">
<h3><span class="header-section-number">16.9.2</span> Factor coding</h3>
<p>Factor variables have a default coding associated with them, corresponding to <a href="https://stats.idre.ucla.edu/other/mult-pkg/faq/general/faqwhat-is-dummy-coding/">“dummy coding”</a>. To view this, use the <code>contrasts()</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">contrasts</span>(circadian<span class="op">$</span>treatment.factor)
        eyes knee
control    <span class="dv">0</span>    <span class="dv">0</span>
eyes       <span class="dv">1</span>    <span class="dv">0</span>
knee       <span class="dv">0</span>    <span class="dv">1</span></code></pre></div>
<p>Using dummy coding, when we fit the regression each level of the factor will be compared to the “base level” (the factor with all zeros in the coding matrix).</p>
</div>
<div id="fitting-the-anova-using-lm" class="section level3">
<h3><span class="header-section-number">16.9.3</span> Fitting the ANOVA using <code>lm</code></h3>
<p>Now that we’ve encoded our grouping variable as a contrast matrix, we can fit the ANOVA using the <code>lm()</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">lm.circadian &lt;-<span class="st"> </span><span class="kw">lm</span>(shift <span class="op">~</span><span class="st"> </span>treatment.factor, <span class="dt">data =</span> circadian)</code></pre></div>
<p>We can then examine information about the model fit:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(lm.circadian)

Call<span class="op">:</span>
<span class="kw">lm</span>(<span class="dt">formula =</span> shift <span class="op">~</span><span class="st"> </span>treatment.factor, <span class="dt">data =</span> circadian)

Residuals<span class="op">:</span>
<span class="st">     </span>Min       1Q   Median       3Q      Max 
<span class="op">-</span><span class="fl">1.27857</span> <span class="op">-</span><span class="fl">0.36125</span>  <span class="fl">0.03857</span>  <span class="fl">0.61147</span>  <span class="fl">1.06571</span> 

Coefficients<span class="op">:</span>
<span class="st">                     </span>Estimate Std. Error t value <span class="kw">Pr</span>(<span class="op">&gt;</span><span class="er">|</span>t<span class="op">|</span>)   
(Intercept)          <span class="op">-</span><span class="fl">0.30875</span>    <span class="fl">0.24888</span>  <span class="op">-</span><span class="fl">1.241</span>  <span class="fl">0.22988</span>   
treatment.factoreyes <span class="op">-</span><span class="fl">1.24268</span>    <span class="fl">0.36433</span>  <span class="op">-</span><span class="fl">3.411</span>  <span class="fl">0.00293</span> <span class="op">**</span>
treatment.factorknee <span class="op">-</span><span class="fl">0.02696</span>    <span class="fl">0.36433</span>  <span class="op">-</span><span class="fl">0.074</span>  <span class="fl">0.94178</span>   
<span class="op">---</span>
Signif. codes<span class="op">:</span><span class="st">  </span><span class="dv">0</span> <span class="st">&#39;***&#39;</span> <span class="fl">0.001</span> <span class="st">&#39;**&#39;</span> <span class="fl">0.01</span> <span class="st">&#39;*&#39;</span> <span class="fl">0.05</span> <span class="st">&#39;.&#39;</span> <span class="fl">0.1</span> <span class="st">&#39; &#39;</span> <span class="dv">1</span>

Residual standard error<span class="op">:</span><span class="st"> </span><span class="fl">0.7039</span> on <span class="dv">19</span> degrees of freedom
Multiple R<span class="op">-</span>squared<span class="op">:</span><span class="st">  </span><span class="fl">0.4342</span>,    Adjusted R<span class="op">-</span>squared<span class="op">:</span><span class="st">  </span><span class="fl">0.3746</span> 
F<span class="op">-</span>statistic<span class="op">:</span><span class="st"> </span><span class="fl">7.289</span> on <span class="dv">2</span> and <span class="dv">19</span> DF,  p<span class="op">-</span>value<span class="op">:</span><span class="st"> </span><span class="fl">0.004472</span></code></pre></div>
<p>Key value for comparison with our earlier <code>aov()</code> generated results include the F-statistic, degrees of freedom, and corresponding p-values.</p>
<p>Let’s use <code>broom::tidy()</code> to focus on the coefficients of the model:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">broom<span class="op">::</span><span class="kw">tidy</span>(lm.circadian)
                  term    estimate std.error   statistic     p.value
<span class="dv">1</span>          (Intercept) <span class="op">-</span><span class="fl">0.30875000</span> <span class="fl">0.2488836</span> <span class="op">-</span><span class="fl">1.24053965</span> <span class="fl">0.229876818</span>
<span class="dv">2</span> treatment.factoreyes <span class="op">-</span><span class="fl">1.24267857</span> <span class="fl">0.3643283</span> <span class="op">-</span><span class="fl">3.41087562</span> <span class="fl">0.002931447</span>
<span class="dv">3</span> treatment.factorknee <span class="op">-</span><span class="fl">0.02696429</span> <span class="fl">0.3643283</span> <span class="op">-</span><span class="fl">0.07401095</span> <span class="fl">0.941775317</span></code></pre></div>
<p>When we use dummy coding, the intercept of the model corresponds to the mean of the variable of interest in the group used as the baseline (“control” in our present example):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># compare mean of control treatment observattions to intercept</span>
control.mean.shift &lt;-
<span class="st">  </span><span class="kw">filter</span>(circadian, treatment <span class="op">==</span><span class="st"> &quot;control&quot;</span>) <span class="op">%$%</span><span class="st"> </span>shift <span class="op">%&gt;%</span><span class="st"> </span>mean 

control.mean.shift
[<span class="dv">1</span>] <span class="op">-</span><span class="fl">0.30875</span></code></pre></div>
<p>The other coefficients are the differences in the means of the other treatment groups relative to the base treatment group. For example, the difference in the means of the “control” and “eyes” treatment groups is:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">circadian <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">group_by</span>(treatment) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">diff.from.control =</span> <span class="kw">mean</span>(shift) <span class="op">-</span><span class="st">  </span>control.mean.shift) 
<span class="co"># A tibble: 3 x 2</span>
  treatment diff.from.control
  <span class="op">&lt;</span>chr<span class="op">&gt;</span><span class="st">                 </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="dv">1</span> control             <span class="dv">0</span>.     
<span class="dv">2</span> eyes               <span class="op">-</span><span class="fl">1.243</span>  
<span class="dv">3</span> knee               <span class="op">-</span><span class="fl">0.02696</span></code></pre></div>
</div>
<div id="changing-the-factor-coding" class="section level3">
<h3><span class="header-section-number">16.9.4</span> Changing the factor coding</h3>
<p>R defines a number of functions – <code>contr.sum()</code>, <code>contr.helmert()</code>, <code>contr.treatment()</code> – for chainging the coding matrix associated with a factor.</p>
<p>If we wanted to use effect coding instead of dummy coding, we can use the <code>contr.sum()</code>. <code>contr.sum()</code> has a single required argument that specifies the number of levels of the categorical variable (three for our example):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">circadian[<span class="st">&quot;treatment.effect&quot;</span>] &lt;-<span class="st"> </span>circadian<span class="op">$</span>treatment.factor
<span class="kw">contrasts</span>(circadian<span class="op">$</span>treatment.effect) &lt;-<span class="st"> </span><span class="kw">contr.sum</span>(<span class="dv">3</span>)
<span class="kw">contrasts</span>(circadian<span class="op">$</span>treatment.effect)
        [,<span class="dv">1</span>] [,<span class="dv">2</span>]
control    <span class="dv">1</span>    <span class="dv">0</span>
eyes       <span class="dv">0</span>    <span class="dv">1</span>
knee      <span class="op">-</span><span class="dv">1</span>   <span class="op">-</span><span class="dv">1</span></code></pre></div>
<p>We again fit the linear model, this time using the effect coding:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">lm.effect &lt;-<span class="st"> </span><span class="kw">lm</span>(shift <span class="op">~</span><span class="st"> </span>treatment.effect, circadian)
<span class="kw">summary</span>(lm.effect)

Call<span class="op">:</span>
<span class="kw">lm</span>(<span class="dt">formula =</span> shift <span class="op">~</span><span class="st"> </span>treatment.effect, <span class="dt">data =</span> circadian)

Residuals<span class="op">:</span>
<span class="st">     </span>Min       1Q   Median       3Q      Max 
<span class="op">-</span><span class="fl">1.27857</span> <span class="op">-</span><span class="fl">0.36125</span>  <span class="fl">0.03857</span>  <span class="fl">0.61147</span>  <span class="fl">1.06571</span> 

Coefficients<span class="op">:</span>
<span class="st">                  </span>Estimate Std. Error t value <span class="kw">Pr</span>(<span class="op">&gt;</span><span class="er">|</span>t<span class="op">|</span>)    
(Intercept)        <span class="op">-</span><span class="fl">0.7320</span>     <span class="fl">0.1504</span>  <span class="op">-</span><span class="fl">4.867</span> <span class="fl">0.000107</span> <span class="op">**</span><span class="er">*</span>
treatment.effect1   <span class="fl">0.4232</span>     <span class="fl">0.2080</span>   <span class="fl">2.035</span> <span class="fl">0.056074</span> .  
treatment.effect2  <span class="op">-</span><span class="fl">0.8195</span>     <span class="fl">0.2150</span>  <span class="op">-</span><span class="fl">3.812</span> <span class="fl">0.001177</span> <span class="op">**</span><span class="st"> </span>
<span class="op">---</span>
Signif. codes<span class="op">:</span><span class="st">  </span><span class="dv">0</span> <span class="st">&#39;***&#39;</span> <span class="fl">0.001</span> <span class="st">&#39;**&#39;</span> <span class="fl">0.01</span> <span class="st">&#39;*&#39;</span> <span class="fl">0.05</span> <span class="st">&#39;.&#39;</span> <span class="fl">0.1</span> <span class="st">&#39; &#39;</span> <span class="dv">1</span>

Residual standard error<span class="op">:</span><span class="st"> </span><span class="fl">0.7039</span> on <span class="dv">19</span> degrees of freedom
Multiple R<span class="op">-</span>squared<span class="op">:</span><span class="st">  </span><span class="fl">0.4342</span>,    Adjusted R<span class="op">-</span>squared<span class="op">:</span><span class="st">  </span><span class="fl">0.3746</span> 
F<span class="op">-</span>statistic<span class="op">:</span><span class="st"> </span><span class="fl">7.289</span> on <span class="dv">2</span> and <span class="dv">19</span> DF,  p<span class="op">-</span>value<span class="op">:</span><span class="st"> </span><span class="fl">0.004472</span></code></pre></div>
<p>The F-statistic and associated p-value of the model stay the same, but now the coefficients have a different interpretation. The intercept is the mean of the group means (unweighted grand mean; equal to the “grand mean” if the number of observations is the same across groups), and the other coefficients are the difference between the means of each effect level and the unweighted grand mean.</p>
<p>Again it’s convenient to work with a tidy version of the coefficients</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">effect.coefficients &lt;-<span class="st"> </span>broom<span class="op">::</span><span class="kw">tidy</span>(lm.effect)

<span class="co"># remind ourselves of what the levels are and their coding</span>
<span class="kw">levels</span>(circadian<span class="op">$</span>treatment.effect)
[<span class="dv">1</span>] <span class="st">&quot;control&quot;</span> <span class="st">&quot;eyes&quot;</span>    <span class="st">&quot;knee&quot;</span>   
<span class="kw">contrasts</span>(circadian<span class="op">$</span>treatment.effect)
        [,<span class="dv">1</span>] [,<span class="dv">2</span>]
control    <span class="dv">1</span>    <span class="dv">0</span>
eyes       <span class="dv">0</span>    <span class="dv">1</span>
knee      <span class="op">-</span><span class="dv">1</span>   <span class="op">-</span><span class="dv">1</span></code></pre></div>
<p>The unweighted grand mean is given by:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">unweighted.grand.mean &lt;-<span class="st"> </span>effect.coefficients<span class="op">$</span>estimate[<span class="dv">1</span>]
unweighted.grand.mean
[<span class="dv">1</span>] <span class="op">-</span><span class="fl">0.7319643</span></code></pre></div>
<p>To calculate the effect of the control treatment relative to the unweighted grand mean, we do:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">effect.control &lt;-<span class="st"> </span>effect.coefficients<span class="op">$</span>estimate[<span class="dv">2</span>]
effect.control
[<span class="dv">1</span>] <span class="fl">0.4232143</span></code></pre></div>
<p>Similarly for the effect of they “eyes” treatment:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">effect.eyes &lt;-<span class="st"> </span>effect.coefficients<span class="op">$</span>estimate[<span class="dv">3</span>]
effect.eyes
[<span class="dv">1</span>] <span class="op">-</span><span class="fl">0.8194643</span></code></pre></div>
<p>Because the coefficients of the coding matrix are all -1 for the knees treatment, we need to multiply each coefficient by -1 and sum the values to get the effect of the “knees” treatment:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">effect.knees &lt;-<span class="st"> </span><span class="kw">sum</span>(<span class="op">-</span><span class="dv">1</span> <span class="op">*</span><span class="st"> </span>effect.coefficients<span class="op">$</span>estimate[<span class="dv">2</span><span class="op">:</span><span class="dv">3</span>])
effect.knees
[<span class="dv">1</span>] <span class="fl">0.39625</span></code></pre></div>
<p>We can reconstruct the group means from the unweighted grand mean and the group effects:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">unweighted.grand.mean <span class="op">+</span><span class="st"> </span>effect.control
[<span class="dv">1</span>] <span class="op">-</span><span class="fl">0.30875</span>
unweighted.grand.mean <span class="op">+</span><span class="st"> </span>effect.eyes
[<span class="dv">1</span>] <span class="op">-</span><span class="fl">1.551429</span>
unweighted.grand.mean <span class="op">+</span><span class="st"> </span>effect.knees
[<span class="dv">1</span>] <span class="op">-</span><span class="fl">0.3357143</span></code></pre></div>
<p>And for comparison:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">circadian <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">group_by</span>(treatment) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summarize</span>(<span class="dt">mean.shift =</span> <span class="kw">mean</span>(shift))
<span class="co"># A tibble: 3 x 2</span>
  treatment mean.shift
  <span class="op">&lt;</span>chr<span class="op">&gt;</span><span class="st">          </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="dv">1</span> control      <span class="op">-</span><span class="fl">0.3088</span>
<span class="dv">2</span> eyes         <span class="op">-</span><span class="fl">1.551</span> 
<span class="dv">3</span> knee         <span class="op">-</span><span class="fl">0.3357</span></code></pre></div>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="randomization-jackknife-and-bootstrap.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="canonical-variates-analysis.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
