<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Biology 723: Statistical Computing for Biologists</title>
  <meta name="description" content="Hands-on materials for Bio 723 course at Duke University.">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Biology 723: Statistical Computing for Biologists" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Biology 723: Statistical Computing for Biologists" />
  
  <meta name="twitter:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

<meta name="author" content="Paul M. Magwene">


<meta name="date" content="2018-04-04">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="introduction-to-the-dplyr-package.html">
<link rel="next" href="vectors-and-functions.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#how-to-use-these-lecture-notes"><i class="fa fa-check"></i><b>1.1</b> How to use these lecture notes</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html"><i class="fa fa-check"></i><b>2</b> Getting Started with R</a><ul>
<li class="chapter" data-level="2.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-r"><i class="fa fa-check"></i><b>2.1</b> What is R?</a></li>
<li class="chapter" data-level="2.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-rstudio"><i class="fa fa-check"></i><b>2.2</b> What is RStudio?</a></li>
<li class="chapter" data-level="2.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#entering-commands-in-the-console"><i class="fa fa-check"></i><b>2.3</b> Entering commands in the console</a></li>
<li class="chapter" data-level="2.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#comments"><i class="fa fa-check"></i><b>2.4</b> Comments</a></li>
<li class="chapter" data-level="2.5" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#using-r-as-a-calculator"><i class="fa fa-check"></i><b>2.5</b> Using R as a Calculator</a><ul>
<li class="chapter" data-level="2.5.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#common-mathematical-functions"><i class="fa fa-check"></i><b>2.5.1</b> Common mathematical functions</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#numeric-data-types-in-r"><i class="fa fa-check"></i><b>2.6</b> Numeric data types in R</a></li>
<li class="chapter" data-level="2.7" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#variable-assignment-in-r"><i class="fa fa-check"></i><b>2.7</b> Variable assignment in R</a><ul>
<li class="chapter" data-level="2.7.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#valid-variable-names"><i class="fa fa-check"></i><b>2.7.1</b> Valid variable names</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#working-with-vectors-in-r"><i class="fa fa-check"></i><b>2.8</b> Working with Vectors in R</a><ul>
<li class="chapter" data-level="2.8.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#vector-arithmetic"><i class="fa fa-check"></i><b>2.8.1</b> Vector Arithmetic</a></li>
<li class="chapter" data-level="2.8.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#vector-recycling"><i class="fa fa-check"></i><b>2.8.2</b> Vector recycling</a></li>
<li class="chapter" data-level="2.8.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#simple-statistical-functions-for-numeric-vectors"><i class="fa fa-check"></i><b>2.8.3</b> Simple statistical functions for numeric vectors</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#the-r-help-system"><i class="fa fa-check"></i><b>2.9</b> The R Help System</a><ul>
<li class="chapter" data-level="2.9.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#getting-help-from-the-console"><i class="fa fa-check"></i><b>2.9.1</b> Getting help from the console</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#packages"><i class="fa fa-check"></i><b>2.10</b> Packages</a><ul>
<li class="chapter" data-level="2.10.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-console"><i class="fa fa-check"></i><b>2.10.1</b> Installing packages from the console</a></li>
<li class="chapter" data-level="2.10.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#install-the-tidyverse-package"><i class="fa fa-check"></i><b>2.10.2</b> Install the tidyverse package</a></li>
<li class="chapter" data-level="2.10.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-rstudio-dialog"><i class="fa fa-check"></i><b>2.10.3</b> Installing packages from the RStudio dialog</a></li>
<li class="chapter" data-level="2.10.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#install-the-stringr-package"><i class="fa fa-check"></i><b>2.10.4</b> Install the stringr package</a></li>
<li class="chapter" data-level="2.10.5" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#loading-packages-with-the-library-function"><i class="fa fa-check"></i><b>2.10.5</b> Loading packages with the <code>library()</code> function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html"><i class="fa fa-check"></i><b>3</b> R Basics: Data types and data structures</a><ul>
<li class="chapter" data-level="3.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-boolean-values"><i class="fa fa-check"></i><b>3.1</b> Logical (Boolean) values</a><ul>
<li class="chapter" data-level="3.1.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#comparison-operators-applied-to-numeric-vectors"><i class="fa fa-check"></i><b>3.1.1</b> Comparison operators applied to numeric vectors</a></li>
<li class="chapter" data-level="3.1.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-vectors"><i class="fa fa-check"></i><b>3.1.2</b> Logical vectors</a></li>
<li class="chapter" data-level="3.1.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-operators"><i class="fa fa-check"></i><b>3.1.3</b> Logical operators</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#character-strings"><i class="fa fa-check"></i><b>3.2</b> Character strings</a><ul>
<li class="chapter" data-level="3.2.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#the-stringr-package"><i class="fa fa-check"></i><b>3.2.1</b> The <code>stringr</code> package</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#vectors"><i class="fa fa-check"></i><b>3.3</b> Vectors</a><ul>
<li class="chapter" data-level="3.3.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#indexing-vectors"><i class="fa fa-check"></i><b>3.3.1</b> Indexing Vectors</a></li>
<li class="chapter" data-level="3.3.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#combining-indexing-and-comparison-of-vectors"><i class="fa fa-check"></i><b>3.3.2</b> Combining Indexing and Comparison of Vectors</a></li>
<li class="chapter" data-level="3.3.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#vector-manipulation"><i class="fa fa-check"></i><b>3.3.3</b> Vector manipulation</a></li>
<li class="chapter" data-level="3.3.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#vectors-from-regular-sequences"><i class="fa fa-check"></i><b>3.3.4</b> Vectors from regular sequences</a></li>
<li class="chapter" data-level="3.3.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#additional-functions-for-working-with-vectors"><i class="fa fa-check"></i><b>3.3.5</b> Additional functions for working with vectors</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#lists"><i class="fa fa-check"></i><b>3.4</b> Lists</a><ul>
<li class="chapter" data-level="3.4.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#length-and-type-of-lists"><i class="fa fa-check"></i><b>3.4.1</b> Length and type of lists</a></li>
<li class="chapter" data-level="3.4.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#indexing-lists"><i class="fa fa-check"></i><b>3.4.2</b> Indexing lists</a></li>
<li class="chapter" data-level="3.4.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#naming-list-elements"><i class="fa fa-check"></i><b>3.4.3</b> Naming list elements</a></li>
<li class="chapter" data-level="3.4.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#the-operator"><i class="fa fa-check"></i><b>3.4.4</b> The <code>$</code> operator</a></li>
<li class="chapter" data-level="3.4.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#changing-and-adding-lists-items"><i class="fa fa-check"></i><b>3.4.5</b> Changing and adding lists items</a></li>
<li class="chapter" data-level="3.4.6" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#combining-lists"><i class="fa fa-check"></i><b>3.4.6</b> Combining lists</a></li>
<li class="chapter" data-level="3.4.7" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#converting-lists-to-vectors"><i class="fa fa-check"></i><b>3.4.7</b> Converting lists to vectors</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#data-frames"><i class="fa fa-check"></i><b>3.5</b> Data frames</a><ul>
<li class="chapter" data-level="3.5.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#creating-a-data-frame"><i class="fa fa-check"></i><b>3.5.1</b> Creating a data frame</a></li>
<li class="chapter" data-level="3.5.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#type-and-class-for-data-frames"><i class="fa fa-check"></i><b>3.5.2</b> Type and class for data frames</a></li>
<li class="chapter" data-level="3.5.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#length-and-dimension-for-data-frames"><i class="fa fa-check"></i><b>3.5.3</b> Length and dimension for data frames</a></li>
<li class="chapter" data-level="3.5.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#indexing-and-accessing-data-frames"><i class="fa fa-check"></i><b>3.5.4</b> Indexing and accessing data frames</a></li>
<li class="chapter" data-level="3.5.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-indexing-of-data-frames"><i class="fa fa-check"></i><b>3.5.5</b> Logical indexing of data frames</a></li>
<li class="chapter" data-level="3.5.6" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#adding-columns-to-a-data-frame"><i class="fa fa-check"></i><b>3.5.6</b> Adding columns to a data frame</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html"><i class="fa fa-check"></i><b>4</b> R Markdown and R Notebooks</a><ul>
<li class="chapter" data-level="4.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#r-notebooks"><i class="fa fa-check"></i><b>4.1</b> R Notebooks</a></li>
<li class="chapter" data-level="4.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#creating-an-r-notebook"><i class="fa fa-check"></i><b>4.2</b> Creating an R Notebook</a></li>
<li class="chapter" data-level="4.3" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#the-default-r-notebook-template"><i class="fa fa-check"></i><b>4.3</b> The default R Notebook template</a></li>
<li class="chapter" data-level="4.4" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-and-non-code-blocks"><i class="fa fa-check"></i><b>4.4</b> Code and Non-code blocks</a><ul>
<li class="chapter" data-level="4.4.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#non-code-blocks"><i class="fa fa-check"></i><b>4.4.1</b> Non-code blocks</a></li>
<li class="chapter" data-level="4.4.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-blocks"><i class="fa fa-check"></i><b>4.4.2</b> Code blocks</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-a-code-chunk"><i class="fa fa-check"></i><b>4.5</b> Running a code chunk</a></li>
<li class="chapter" data-level="4.6" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-all-code-chunks-above"><i class="fa fa-check"></i><b>4.6</b> Running all code chunks above</a></li>
<li class="chapter" data-level="4.7" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#knitting-an-r-markdown-to-html"><i class="fa fa-check"></i><b>4.7</b> “Knitting an” R Markdown to HTML</a></li>
<li class="chapter" data-level="4.8" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#sharing-your-reproducible-r-notebook"><i class="fa fa-check"></i><b>4.8</b> Sharing your reproducible R Notebook</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html"><i class="fa fa-check"></i><b>5</b> Introduction to ggplot2</a><ul>
<li class="chapter" data-level="5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#loading-ggplot2"><i class="fa fa-check"></i><b>5.1</b> Loading ggplot2</a></li>
<li class="chapter" data-level="5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#example-data-set-andersons-iris-data"><i class="fa fa-check"></i><b>5.2</b> Example data set: Anderson’s Iris Data</a></li>
<li class="chapter" data-level="5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#template-for-single-layer-plots-in-ggplot2"><i class="fa fa-check"></i><b>5.3</b> Template for single layer plots in ggplot2</a></li>
<li class="chapter" data-level="5.4" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#an-aside-about-function-arguments"><i class="fa fa-check"></i><b>5.4</b> An aside about function arguments</a></li>
<li class="chapter" data-level="5.5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#strip-plots"><i class="fa fa-check"></i><b>5.5</b> Strip plots</a><ul>
<li class="chapter" data-level="5.5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#jittering-data"><i class="fa fa-check"></i><b>5.5.1</b> Jittering data</a></li>
<li class="chapter" data-level="5.5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-categorical-information"><i class="fa fa-check"></i><b>5.5.2</b> Adding categorical information</a></li>
<li class="chapter" data-level="5.5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#rotating-plot-coordinates"><i class="fa fa-check"></i><b>5.5.3</b> Rotating plot coordinates</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#histograms"><i class="fa fa-check"></i><b>5.6</b> Histograms</a><ul>
<li class="chapter" data-level="5.6.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#variations-on-histograms-when-considering-categorical-data"><i class="fa fa-check"></i><b>5.6.1</b> Variations on histograms when considering categorical data</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#faceting-to-depict-categorical-information"><i class="fa fa-check"></i><b>5.7</b> Faceting to depict categorical information</a></li>
<li class="chapter" data-level="5.8" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots"><i class="fa fa-check"></i><b>5.8</b> Density plots</a></li>
<li class="chapter" data-level="5.9" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-or-beanplot"><i class="fa fa-check"></i><b>5.9</b> Violin or Beanplot</a></li>
<li class="chapter" data-level="5.10" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplots"><i class="fa fa-check"></i><b>5.10</b> Boxplots</a></li>
<li class="chapter" data-level="5.11" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#building-complex-visualizations-with-layers"><i class="fa fa-check"></i><b>5.11</b> Building complex visualizations with layers</a></li>
<li class="chapter" data-level="5.12" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#useful-combination-plots"><i class="fa fa-check"></i><b>5.12</b> Useful combination plots</a><ul>
<li class="chapter" data-level="5.12.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplot-plus-strip-plot"><i class="fa fa-check"></i><b>5.12.1</b> Boxplot plus strip plot</a></li>
<li class="chapter" data-level="5.12.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#setting-shared-aesthetics"><i class="fa fa-check"></i><b>5.12.2</b> Setting shared aesthetics</a></li>
</ul></li>
<li class="chapter" data-level="5.13" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot-layers-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>5.13</b> ggplot layers can be assigned to variables</a><ul>
<li class="chapter" data-level="5.13.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-plot-plus-strip-plot"><i class="fa fa-check"></i><b>5.13.1</b> Violin plot plus strip plot</a></li>
</ul></li>
<li class="chapter" data-level="5.14" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-titles-and-tweaking-axis-labels"><i class="fa fa-check"></i><b>5.14</b> Adding titles and tweaking axis labels</a></li>
<li class="chapter" data-level="5.15" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot2-themes"><i class="fa fa-check"></i><b>5.15</b> ggplot2 themes</a><ul>
<li class="chapter" data-level="5.15.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#further-customization-with-ggplot2theme"><i class="fa fa-check"></i><b>5.15.1</b> Further customization with <code>ggplot2::theme</code></a></li>
</ul></li>
<li class="chapter" data-level="5.16" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#other-aspects-of-ggplots-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>5.16</b> Other aspects of ggplots can be assigned to variables</a></li>
<li class="chapter" data-level="5.17" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-plots"><i class="fa fa-check"></i><b>5.17</b> Bivariate plots</a><ul>
<li class="chapter" data-level="5.17.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#scatter-plots"><i class="fa fa-check"></i><b>5.17.1</b> Scatter plots</a></li>
</ul></li>
<li class="chapter" data-level="5.18" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-density-plots"><i class="fa fa-check"></i><b>5.18</b> Bivariate density plots</a></li>
<li class="chapter" data-level="5.19" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#combining-scatter-plots-and-density-plots-with-categorical-information"><i class="fa fa-check"></i><b>5.19</b> Combining Scatter Plots and Density Plots with Categorical Information</a></li>
<li class="chapter" data-level="5.20" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots-with-fill"><i class="fa fa-check"></i><b>5.20</b> Density plots with fill</a></li>
<li class="chapter" data-level="5.21" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#d-bin-and-hex-plots"><i class="fa fa-check"></i><b>5.21</b> 2D bin and hex plots</a></li>
<li class="chapter" data-level="5.22" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#the-cowplot-package"><i class="fa fa-check"></i><b>5.22</b> The <code>cowplot</code> package</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html"><i class="fa fa-check"></i><b>6</b> Introduction to the <code>dplyr</code> package</a><ul>
<li class="chapter" data-level="6.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#libraries"><i class="fa fa-check"></i><b>6.1</b> Libraries</a></li>
<li class="chapter" data-level="6.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#reading-data-with-the-readr-package"><i class="fa fa-check"></i><b>6.2</b> Reading data with the <code>readr</code> package</a><ul>
<li class="chapter" data-level="6.2.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#example-data-nc-births"><i class="fa fa-check"></i><b>6.2.1</b> Example data: NC Births</a></li>
<li class="chapter" data-level="6.2.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#reading-excel-files"><i class="fa fa-check"></i><b>6.2.2</b> Reading Excel files</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#a-note-on-tibbles"><i class="fa fa-check"></i><b>6.3</b> A note on “tibbles”</a></li>
<li class="chapter" data-level="6.4" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#data-filtering-and-transformation-with-dplyr"><i class="fa fa-check"></i><b>6.4</b> Data filtering and transformation with <code>dplyr</code></a></li>
<li class="chapter" data-level="6.5" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#dplyrs-verbs"><i class="fa fa-check"></i><b>6.5</b> dplyr’s “verbs”</a><ul>
<li class="chapter" data-level="6.5.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#select"><i class="fa fa-check"></i><b>6.5.1</b> <code>select</code></a></li>
<li class="chapter" data-level="6.5.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#filter"><i class="fa fa-check"></i><b>6.5.2</b> <code>filter</code></a></li>
<li class="chapter" data-level="6.5.3" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#mutate"><i class="fa fa-check"></i><b>6.5.3</b> <code>mutate</code></a></li>
<li class="chapter" data-level="6.5.4" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#arrange"><i class="fa fa-check"></i><b>6.5.4</b> <code>arrange</code></a></li>
<li class="chapter" data-level="6.5.5" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#summarize"><i class="fa fa-check"></i><b>6.5.5</b> <code>summarize</code></a></li>
<li class="chapter" data-level="6.5.6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#group_by"><i class="fa fa-check"></i><b>6.5.6</b> <code>group_by</code></a></li>
<li class="chapter" data-level="6.5.7" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#combining-grouping-and-summarizing"><i class="fa fa-check"></i><b>6.5.7</b> Combining grouping and summarizing</a></li>
<li class="chapter" data-level="6.5.8" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#scoped-variants-of-mutate-and-summarize"><i class="fa fa-check"></i><b>6.5.8</b> Scoped variants of <code>mutate</code> and <code>summarize</code></a></li>
<li class="chapter" data-level="6.5.9" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#combining-summarize-with-grouping-aesthetics-in-ggplot2"><i class="fa fa-check"></i><b>6.5.9</b> Combining summarize with grouping aesthetics in <code>ggplot2</code></a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#pipes"><i class="fa fa-check"></i><b>6.6</b> Pipes</a><ul>
<li class="chapter" data-level="6.6.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#install-and-load-magrittr"><i class="fa fa-check"></i><b>6.6.1</b> Install and load <code>magrittr</code></a></li>
<li class="chapter" data-level="6.6.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-basic-pipe-operator"><i class="fa fa-check"></i><b>6.6.2</b> The basic pipe operator</a></li>
<li class="chapter" data-level="6.6.3" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#an-example-without-pipes"><i class="fa fa-check"></i><b>6.6.3</b> An example without pipes</a></li>
<li class="chapter" data-level="6.6.4" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-same-example-using-pipes"><i class="fa fa-check"></i><b>6.6.4</b> The same example using pipes</a></li>
<li class="chapter" data-level="6.6.5" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#assigning-the-output-of-a-statement-involving-pipes-to-a-variable"><i class="fa fa-check"></i><b>6.6.5</b> Assigning the output of a statement involving pipes to a variable</a></li>
<li class="chapter" data-level="6.6.6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#compound-assignment-pipe-operator"><i class="fa fa-check"></i><b>6.6.6</b> Compound assignment pipe operator</a></li>
<li class="chapter" data-level="6.6.7" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-dot-operator-with-pipes"><i class="fa fa-check"></i><b>6.6.7</b> The dot operator with pipes</a></li>
<li class="chapter" data-level="6.6.8" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-exposition-pipe-operator"><i class="fa fa-check"></i><b>6.6.8</b> The exposition pipe operator</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>7</b> Data wrangling</a><ul>
<li class="chapter" data-level="7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#libraries-1"><i class="fa fa-check"></i><b>7.1</b> Libraries</a></li>
<li class="chapter" data-level="7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#data"><i class="fa fa-check"></i><b>7.2</b> Data</a></li>
<li class="chapter" data-level="7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#renaming-data-frame-columms"><i class="fa fa-check"></i><b>7.3</b> Renaming data frame columms</a></li>
<li class="chapter" data-level="7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-unneeded-columns"><i class="fa fa-check"></i><b>7.4</b> Dropping unneeded columns</a><ul>
<li class="chapter" data-level="7.4.1" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-all-empty-columns"><i class="fa fa-check"></i><b>7.4.1</b> Finding all empty columns</a></li>
<li class="chapter" data-level="7.4.2" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-columns-by-matching-names"><i class="fa fa-check"></i><b>7.4.2</b> Dropping columns by matching names</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="data-wrangling.html"><a href="data-wrangling.html#reshaping-data-with-tidyr"><i class="fa fa-check"></i><b>7.5</b> Reshaping data with tidyr</a><ul>
<li class="chapter" data-level="7.5.1" data-path="data-wrangling.html"><a href="data-wrangling.html#wide-to-long-conversions-using-tidyrgather"><i class="fa fa-check"></i><b>7.5.1</b> Wide to long conversions using <code>tidyr::gather</code></a></li>
<li class="chapter" data-level="7.5.2" data-path="data-wrangling.html"><a href="data-wrangling.html#extracting-information-from-combined-variables-using-tidyrextract"><i class="fa fa-check"></i><b>7.5.2</b> Extracting information from combined variables using <code>tidyr::extract</code></a></li>
<li class="chapter" data-level="7.5.3" data-path="data-wrangling.html"><a href="data-wrangling.html#subsetting-rows"><i class="fa fa-check"></i><b>7.5.3</b> Subsetting rows</a></li>
<li class="chapter" data-level="7.5.4" data-path="data-wrangling.html"><a href="data-wrangling.html#splitting-columns"><i class="fa fa-check"></i><b>7.5.4</b> Splitting columns</a></li>
<li class="chapter" data-level="7.5.5" data-path="data-wrangling.html"><a href="data-wrangling.html#combining-data-frame-rows"><i class="fa fa-check"></i><b>7.5.5</b> Combining data frame rows</a></li>
<li class="chapter" data-level="7.5.6" data-path="data-wrangling.html"><a href="data-wrangling.html#sorting-data-frame-rows"><i class="fa fa-check"></i><b>7.5.6</b> Sorting data frame rows</a></li>
<li class="chapter" data-level="7.5.7" data-path="data-wrangling.html"><a href="data-wrangling.html#collapsing-columns-using-tidyrunite"><i class="fa fa-check"></i><b>7.5.7</b> Collapsing columns using <code>tidyr::unite</code></a></li>
<li class="chapter" data-level="7.5.8" data-path="data-wrangling.html"><a href="data-wrangling.html#splitting-columns-using-tidyrseparate"><i class="fa fa-check"></i><b>7.5.8</b> Splitting columns using <code>tidyr::separate</code></a></li>
<li class="chapter" data-level="7.5.9" data-path="data-wrangling.html"><a href="data-wrangling.html#long-to-wide-conversion-using-tidyrspread"><i class="fa fa-check"></i><b>7.5.9</b> Long-to-wide conversion using <code>tidyr::spread</code></a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="data-wrangling.html"><a href="data-wrangling.html#using-your-tidy-data"><i class="fa fa-check"></i><b>7.6</b> Using your tidy data</a><ul>
<li class="chapter" data-level="7.6.1" data-path="data-wrangling.html"><a href="data-wrangling.html#visualizing-gene-expression-time-series"><i class="fa fa-check"></i><b>7.6.1</b> Visualizing gene expression time series</a></li>
<li class="chapter" data-level="7.6.2" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-the-most-variable-genes"><i class="fa fa-check"></i><b>7.6.2</b> Finding the most variable genes</a></li>
<li class="chapter" data-level="7.6.3" data-path="data-wrangling.html"><a href="data-wrangling.html#exploring-bivariate-relationships-using-wide-data"><i class="fa fa-check"></i><b>7.6.3</b> Exploring bivariate relationships using “wide” data</a></li>
<li class="chapter" data-level="7.6.4" data-path="data-wrangling.html"><a href="data-wrangling.html#calculating-correlations-between-genes"><i class="fa fa-check"></i><b>7.6.4</b> Calculating correlations between genes</a></li>
<li class="chapter" data-level="7.6.5" data-path="data-wrangling.html"><a href="data-wrangling.html#adding-new-columns-and-combining-filtered-data-frames"><i class="fa fa-check"></i><b>7.6.5</b> Adding new columns and combining filtered data frames</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="data-wrangling.html"><a href="data-wrangling.html#heat-maps"><i class="fa fa-check"></i><b>7.7</b> Heat maps</a><ul>
<li class="chapter" data-level="7.7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#rcolorbrewer"><i class="fa fa-check"></i><b>7.7.1</b> RColorBrewer</a></li>
<li class="chapter" data-level="7.7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#creating-a-heat-map-using-geom_tile"><i class="fa fa-check"></i><b>7.7.2</b> Creating a heat map using <code>geom_tile</code></a></li>
<li class="chapter" data-level="7.7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#a-fancy-figure"><i class="fa fa-check"></i><b>7.7.3</b> A “fancy” figure</a></li>
<li class="chapter" data-level="7.7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#sorting-and-visualizing-genes-by-time-of-maximum-expression"><i class="fa fa-check"></i><b>7.7.4</b> Sorting and visualizing genes by time of maximum expression</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html"><i class="fa fa-check"></i><b>8</b> Vectors and Functions</a><ul>
<li class="chapter" data-level="8.1" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#libraries-2"><i class="fa fa-check"></i><b>8.1</b> Libraries</a></li>
<li class="chapter" data-level="8.2" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#vector-mathematics-in-r"><i class="fa fa-check"></i><b>8.2</b> Vector Mathematics in R</a></li>
<li class="chapter" data-level="8.3" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#writing-your-own-functions-in-r"><i class="fa fa-check"></i><b>8.3</b> Writing Your Own Functions in R</a></li>
<li class="chapter" data-level="8.4" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#function-arguments"><i class="fa fa-check"></i><b>8.4</b> Function arguments</a></li>
<li class="chapter" data-level="8.5" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#writing-functions-with-optional-arguments"><i class="fa fa-check"></i><b>8.5</b> Writing functions with optional arguments</a></li>
<li class="chapter" data-level="8.6" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#putting-r-functions-in-scripts"><i class="fa fa-check"></i><b>8.6</b> Putting R functions in Scripts</a></li>
<li class="chapter" data-level="8.7" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#simple-statistics-in-vector-form"><i class="fa fa-check"></i><b>8.7</b> Simple statistics in vector form</a><ul>
<li class="chapter" data-level="8.7.1" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#mean"><i class="fa fa-check"></i><b>8.7.1</b> Mean</a></li>
<li class="chapter" data-level="8.7.2" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#variance-and-standard-deviation"><i class="fa fa-check"></i><b>8.7.2</b> Variance and standard deviation</a></li>
<li class="chapter" data-level="8.7.3" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#covariance-and-correlation"><i class="fa fa-check"></i><b>8.7.3</b> Covariance and correlation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html"><i class="fa fa-check"></i><b>9</b> Bivariate Linear Regression</a><ul>
<li class="chapter" data-level="9.1" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#regression-terminology"><i class="fa fa-check"></i><b>9.1</b> Regression terminology</a></li>
<li class="chapter" data-level="9.2" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#libraries-3"><i class="fa fa-check"></i><b>9.2</b> Libraries</a></li>
<li class="chapter" data-level="9.3" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#illustrating-linear-regression-with-simulated-data"><i class="fa fa-check"></i><b>9.3</b> Illustrating linear regression with simulated data</a></li>
<li class="chapter" data-level="9.4" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#specifying-regression-models-in-r"><i class="fa fa-check"></i><b>9.4</b> Specifying Regression Models in R</a><ul>
<li class="chapter" data-level="9.4.1" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#fitted-values"><i class="fa fa-check"></i><b>9.4.1</b> Fitted values</a></li>
<li class="chapter" data-level="9.4.2" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#getting-the-model-coefficients"><i class="fa fa-check"></i><b>9.4.2</b> Getting the model coefficients</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#residuals"><i class="fa fa-check"></i><b>9.5</b> Residuals</a></li>
<li class="chapter" data-level="9.6" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#regression-as-sum-of-squares-decomposition"><i class="fa fa-check"></i><b>9.6</b> Regression as sum-of-squares decomposition</a></li>
<li class="chapter" data-level="9.7" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#variance-explained-by-a-regression-model"><i class="fa fa-check"></i><b>9.7</b> Variance “explained” by a regression model</a></li>
<li class="chapter" data-level="9.8" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broom-a-library-for-converting-model-results-into-data-frames"><i class="fa fa-check"></i><b>9.8</b> Broom: a library for converting model results into data frames</a><ul>
<li class="chapter" data-level="9.8.1" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broomtidy"><i class="fa fa-check"></i><b>9.8.1</b> <code>broom::tidy</code></a></li>
<li class="chapter" data-level="9.8.2" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broomaugment"><i class="fa fa-check"></i><b>9.8.2</b> <code>broom::augment</code></a></li>
<li class="chapter" data-level="9.8.3" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broomglance"><i class="fa fa-check"></i><b>9.8.3</b> <code>broom::glance</code></a></li>
</ul></li>
<li class="chapter" data-level="9.9" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#bivariate-linear-regression-on-a-real-world-data-set"><i class="fa fa-check"></i><b>9.9</b> Bivariate linear regression on a real world data set</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="matrices-in-r.html"><a href="matrices-in-r.html"><i class="fa fa-check"></i><b>10</b> Matrices in R</a><ul>
<li class="chapter" data-level="10.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#creating-matrices-in-r"><i class="fa fa-check"></i><b>10.1</b> Creating matrices in R</a></li>
<li class="chapter" data-level="10.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-arithmetic-operations-in-r"><i class="fa fa-check"></i><b>10.2</b> Matrix arithmetic operations in R</a></li>
<li class="chapter" data-level="10.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#descriptive-statistics-as-matrix-functions"><i class="fa fa-check"></i><b>10.3</b> Descriptive statistics as matrix functions</a><ul>
<li class="chapter" data-level="10.3.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#mean-vector-and-matrix"><i class="fa fa-check"></i><b>10.3.1</b> Mean vector and matrix</a></li>
<li class="chapter" data-level="10.3.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#deviation-matrix"><i class="fa fa-check"></i><b>10.3.2</b> Deviation matrix</a></li>
<li class="chapter" data-level="10.3.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#covariance-matrix"><i class="fa fa-check"></i><b>10.3.3</b> Covariance matrix</a></li>
<li class="chapter" data-level="10.3.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#correlation-matrix"><i class="fa fa-check"></i><b>10.3.4</b> Correlation matrix</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-inverse"><i class="fa fa-check"></i><b>10.4</b> Matrix Inverse</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="multiple-regression.html"><a href="multiple-regression.html"><i class="fa fa-check"></i><b>11</b> Multiple Regression</a><ul>
<li class="chapter" data-level="11.1" data-path="multiple-regression.html"><a href="multiple-regression.html#new-libraries-to-install"><i class="fa fa-check"></i><b>11.1</b> New Libraries to install</a></li>
<li class="chapter" data-level="11.2" data-path="multiple-regression.html"><a href="multiple-regression.html#libraries-4"><i class="fa fa-check"></i><b>11.2</b> Libraries</a></li>
<li class="chapter" data-level="11.3" data-path="multiple-regression.html"><a href="multiple-regression.html#examplar-data"><i class="fa fa-check"></i><b>11.3</b> Examplar data</a></li>
<li class="chapter" data-level="11.4" data-path="multiple-regression.html"><a href="multiple-regression.html#data-exploration"><i class="fa fa-check"></i><b>11.4</b> Data exploration</a></li>
<li class="chapter" data-level="11.5" data-path="multiple-regression.html"><a href="multiple-regression.html#d-plots"><i class="fa fa-check"></i><b>11.5</b> 3D Plots</a><ul>
<li class="chapter" data-level="11.5.1" data-path="multiple-regression.html"><a href="multiple-regression.html#scatterplot3d"><i class="fa fa-check"></i><b>11.5.1</b> scatterplot3d</a></li>
<li class="chapter" data-level="11.5.2" data-path="multiple-regression.html"><a href="multiple-regression.html#rgl"><i class="fa fa-check"></i><b>11.5.2</b> rgl</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="multiple-regression.html"><a href="multiple-regression.html#fitting-the-regression-model"><i class="fa fa-check"></i><b>11.6</b> Fitting the regression model</a><ul>
<li class="chapter" data-level="11.6.1" data-path="multiple-regression.html"><a href="multiple-regression.html#visualizing-the-regression-model-in-scatterplot3d"><i class="fa fa-check"></i><b>11.6.1</b> Visualizing the regression model in scatterplot3d</a></li>
</ul></li>
<li class="chapter" data-level="11.7" data-path="multiple-regression.html"><a href="multiple-regression.html#interpretting-the-regression-model"><i class="fa fa-check"></i><b>11.7</b> Interpretting the regression model</a></li>
<li class="chapter" data-level="11.8" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-vector-geometry-of-the-regression-model"><i class="fa fa-check"></i><b>11.8</b> Exploring the Vector Geometry of the Regression Model</a></li>
<li class="chapter" data-level="11.9" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-residuals-from-the-model-fit"><i class="fa fa-check"></i><b>11.9</b> Exploring the Residuals from the Model Fit</a></li>
<li class="chapter" data-level="11.10" data-path="multiple-regression.html"><a href="multiple-regression.html#an-alternate-model"><i class="fa fa-check"></i><b>11.10</b> An alternate model</a></li>
<li class="chapter" data-level="11.11" data-path="multiple-regression.html"><a href="multiple-regression.html#fitting-a-curvilinear-model-using-lm"><i class="fa fa-check"></i><b>11.11</b> Fitting a curvilinear model using lm()</a></li>
<li class="chapter" data-level="11.12" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-impact-of-nearly-collinear-predictors-on-regression"><i class="fa fa-check"></i><b>11.12</b> Exploring the impact of nearly collinear predictors on regression</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html"><i class="fa fa-check"></i><b>12</b> Principal Components Analysis</a><ul>
<li class="chapter" data-level="12.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#libraries-5"><i class="fa fa-check"></i><b>12.1</b> Libraries</a></li>
<li class="chapter" data-level="12.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#matrices-as-linear-transformations"><i class="fa fa-check"></i><b>12.2</b> Matrices as linear transformations</a></li>
<li class="chapter" data-level="12.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#eigenanalysis-in-r"><i class="fa fa-check"></i><b>12.3</b> Eigenanalysis in R</a></li>
<li class="chapter" data-level="12.4" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#principal-components-analysis-in-r"><i class="fa fa-check"></i><b>12.4</b> Principal Components Analysis in R</a><ul>
<li class="chapter" data-level="12.4.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#bioenv-dataset"><i class="fa fa-check"></i><b>12.4.1</b> Bioenv dataset</a></li>
<li class="chapter" data-level="12.4.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pca-of-the-bioenv-dataset"><i class="fa fa-check"></i><b>12.4.2</b> PCA of the Bioenv dataset</a></li>
<li class="chapter" data-level="12.4.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#calculating-factor-loadings"><i class="fa fa-check"></i><b>12.4.3</b> Calculating Factor Loadings</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#drawing-figures-to-represent-pca"><i class="fa fa-check"></i><b>12.5</b> Drawing Figures to Represent PCA</a><ul>
<li class="chapter" data-level="12.5.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pc-score-plots"><i class="fa fa-check"></i><b>12.5.1</b> PC Score Plots</a></li>
<li class="chapter" data-level="12.5.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#simultaneous-depiction-of-observations-and-variables-in-the-pc-space"><i class="fa fa-check"></i><b>12.5.2</b> Simultaneous Depiction of Observations and Variables in the PC Space</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html"><i class="fa fa-check"></i><b>13</b> Simulating Sampling Distributions</a><ul>
<li class="chapter" data-level="13.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#libraries-6"><i class="fa fa-check"></i><b>13.1</b> Libraries</a></li>
<li class="chapter" data-level="13.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#data-set-simulated-male-heights"><i class="fa fa-check"></i><b>13.2</b> Data set: Simulated male heights</a><ul>
<li class="chapter" data-level="13.2.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#seeding-the-pseudo-random-number-generator"><i class="fa fa-check"></i><b>13.2.1</b> Seeding the pseudo-random number generator</a></li>
<li class="chapter" data-level="13.2.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#generating-the-simulated-population"><i class="fa fa-check"></i><b>13.2.2</b> Generating the simulated population</a></li>
<li class="chapter" data-level="13.2.3" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#properties-of-the-simulated-population"><i class="fa fa-check"></i><b>13.2.3</b> Properties of the simulated population</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#random-sampling-from-the-simulated-population"><i class="fa fa-check"></i><b>13.3</b> Random sampling from the simulated population</a><ul>
<li class="chapter" data-level="13.3.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#another-random-sample"><i class="fa fa-check"></i><b>13.3.1</b> Another random sample</a></li>
<li class="chapter" data-level="13.3.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#simulating-the-generation-of-many-random-samples"><i class="fa fa-check"></i><b>13.3.2</b> Simulating the generation of many random samples</a></li>
<li class="chapter" data-level="13.3.3" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#a-function-to-estimate-statistics-of-interest-in-a-random-sample"><i class="fa fa-check"></i><b>13.3.3</b> A function to estimate statistics of interest in a random sample</a></li>
<li class="chapter" data-level="13.3.4" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#generating-statistics-for-many-random-samples"><i class="fa fa-check"></i><b>13.3.4</b> Generating statistics for many random samples</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#simulated-sampling-distribution-of-the-mean"><i class="fa fa-check"></i><b>13.4</b> Simulated sampling distribution of the mean</a><ul>
<li class="chapter" data-level="13.4.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#sampling-distributions-for-different-sample-sizes"><i class="fa fa-check"></i><b>13.4.1</b> Sampling distributions for different sample sizes</a></li>
<li class="chapter" data-level="13.4.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#discussion-of-trends-for-sampling-distributions-of-different-sample-sizes"><i class="fa fa-check"></i><b>13.4.2</b> Discussion of trends for sampling distributions of different sample sizes</a></li>
</ul></li>
<li class="chapter" data-level="13.5" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#standard-error-of-the-mean"><i class="fa fa-check"></i><b>13.5</b> Standard Error of the Mean</a></li>
<li class="chapter" data-level="13.6" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#sampling-distribution-of-the-standard-deviation"><i class="fa fa-check"></i><b>13.6</b> Sampling Distribution of the Standard Deviation</a></li>
<li class="chapter" data-level="13.7" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#what-happens-to-the-sampling-distribution-of-the-mean-and-standard-deviation-when-our-sample-size-is-small"><i class="fa fa-check"></i><b>13.7</b> What happens to the sampling distribution of the mean and standard deviation when our sample size is small?</a><ul>
<li class="chapter" data-level="13.7.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#for-small-samples-sample-standard-deviations-systematically-underestimate-the-population-standard-deviation"><i class="fa fa-check"></i><b>13.7.1</b> For small samples, sample standard deviations systematically underestimate the population standard deviation</a></li>
<li class="chapter" data-level="13.7.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#underestimates-of-the-standard-deviation-given-small-n-lead-to-understimates-of-the-se-of-the-mean"><i class="fa fa-check"></i><b>13.7.2</b> Underestimates of the standard deviation given small <span class="math inline">\(n\)</span> lead to understimates of the SE of the mean</a></li>
</ul></li>
<li class="chapter" data-level="13.8" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#the-t-distribution-is-the-appropriate-distribution-for-describing-the-sampling-distribution-of-the-mean-when-n-is-small"><i class="fa fa-check"></i><b>13.8</b> The t-distribution is the appropriate distribution for describing the sampling distribution of the mean when <span class="math inline">\(n\)</span> is small</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html"><i class="fa fa-check"></i><b>14</b> Simulating confidence intervals</a><ul>
<li class="chapter" data-level="14.1" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#libraries-7"><i class="fa fa-check"></i><b>14.1</b> Libraries</a></li>
<li class="chapter" data-level="14.2" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#confidence-intervals"><i class="fa fa-check"></i><b>14.2</b> Confidence Intervals</a></li>
<li class="chapter" data-level="14.3" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#calibrating-confidence-intervals"><i class="fa fa-check"></i><b>14.3</b> Calibrating confidence intervals</a></li>
<li class="chapter" data-level="14.4" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#standard-formulation-for-confidence-intervals"><i class="fa fa-check"></i><b>14.4</b> Standard formulation for confidence intervals</a></li>
<li class="chapter" data-level="14.5" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#simulating-confidence-intervals-for-the-mean"><i class="fa fa-check"></i><b>14.5</b> Simulating confidence intervals for the mean</a></li>
<li class="chapter" data-level="14.6" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#sample-estimate-of-the-standard-error-of-the-mean"><i class="fa fa-check"></i><b>14.6</b> Sample estimate of the standard error of the mean</a></li>
<li class="chapter" data-level="14.7" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#ci-of-mean-simulation"><i class="fa fa-check"></i><b>14.7</b> CI of mean simulation</a></li>
<li class="chapter" data-level="14.8" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#generating-a-table-of-cis-and-corresponding-margins-of-error"><i class="fa fa-check"></i><b>14.8</b> Generating a table of CIs and corresponding margins of error</a></li>
<li class="chapter" data-level="14.9" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#interpreting-our-simulation"><i class="fa fa-check"></i><b>14.9</b> Interpreting our simulation</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html"><i class="fa fa-check"></i><b>15</b> Randomization, Jackknife, and Bootstrap</a><ul>
<li class="chapter" data-level="15.1" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#libraries-8"><i class="fa fa-check"></i><b>15.1</b> Libraries</a></li>
<li class="chapter" data-level="15.2" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#randomization"><i class="fa fa-check"></i><b>15.2</b> Randomization</a></li>
<li class="chapter" data-level="15.3" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#using-randomization-to-test-for-a-difference-in-means"><i class="fa fa-check"></i><b>15.3</b> Using randomization to test for a difference in means</a></li>
<li class="chapter" data-level="15.4" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#using-randomization-to-test-for-equality-of-variances"><i class="fa fa-check"></i><b>15.4</b> Using randomization to test for equality of variances</a></li>
<li class="chapter" data-level="15.5" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#jackknifing-in-r"><i class="fa fa-check"></i><b>15.5</b> Jackknifing in R</a></li>
<li class="chapter" data-level="15.6" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#bootstrapping-in-r"><i class="fa fa-check"></i><b>15.6</b> Bootstrapping in R</a><ul>
<li class="chapter" data-level="15.6.1" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#a-more-sophisticated-application-of-the-bootstrap"><i class="fa fa-check"></i><b>15.6.1</b> A more sophisticated application of the bootstrap</a></li>
<li class="chapter" data-level="15.6.2" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#about-logistic-regression"><i class="fa fa-check"></i><b>15.6.2</b> About logistic regression</a></li>
<li class="chapter" data-level="15.6.3" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#bumpus-house-sparrow-data-set"><i class="fa fa-check"></i><b>15.6.3</b> Bumpus house sparrow data set</a></li>
<li class="chapter" data-level="15.6.4" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#predicting-survival-as-a-function-of-body-weight"><i class="fa fa-check"></i><b>15.6.4</b> Predicting survival as a function of body weight</a></li>
<li class="chapter" data-level="15.6.5" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#fitting-the-logistic-regression"><i class="fa fa-check"></i><b>15.6.5</b> Fitting the logistic regression</a></li>
<li class="chapter" data-level="15.6.6" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#defining-an-appropriate-function-for-boot"><i class="fa fa-check"></i><b>15.6.6</b> Defining an appropriate function for <code>boot</code></a></li>
<li class="chapter" data-level="15.6.7" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#calculating-bootstrap-confidence-intervals"><i class="fa fa-check"></i><b>15.6.7</b> Calculating bootstrap confidence intervals</a></li>
<li class="chapter" data-level="15.6.8" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#visualizing-bootstrap-confidence-intervals"><i class="fa fa-check"></i><b>15.6.8</b> Visualizing bootstrap confidence intervals</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Biology 723: Statistical Computing for Biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="data-wrangling" class="section level1">
<h1><span class="header-section-number">Chapter 7</span> Data wrangling</h1>
<p>In the real world you’ll often create a data set (or be given one) in a format that is less than ideal for analysis. This can happen for a number of reasons. For example, the data may have been recorded in a manner convenient for collection and visual inspection, but which does not work well for analysis and plotting. Or the data may be an amalgamation of multiple experiments, in which each of the experimenters used slightly different naming conventions. Or the data may have been produced by an instrument that produces output with a fixed format. Sometimes important experimental information is included in the column headers of a spreadsheet.</p>
<p>Whatever the case, we often find ourselves in the situation where we need to “wrangle” our data into a “tidy” format before we can proceed with visualization and analysis. The “R for Data Science” text discusses some desirable rules for “tidy” data in order to facilitate downstream analyses. These are:</p>
<ol style="list-style-type: decimal">
<li>Each variable must have its own column.</li>
<li>Each observation must have its own row.</li>
<li>Each value must have its own cell.</li>
</ol>
<p>In this lecture we’re going to walk through an extended example of wrangling some data into a “tidy” format.</p>
<div id="libraries-1" class="section level2">
<h2><span class="header-section-number">7.1</span> Libraries</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(magrittr)
<span class="kw">library</span>(stringr)
<span class="kw">library</span>(tidyverse)
<span class="kw">library</span>(cowplot)</code></pre></div>
</div>
<div id="data" class="section level2">
<h2><span class="header-section-number">7.2</span> Data</h2>
<p>To illustrate a standard data wrangling pipeline, we’re going to use a gene expression microarray data set, based on the following paper:</p>
<ul>
<li>Spellman PT, et al. 1998. Comprehensive identification of cell cycle-regulated genes of the yeast Saccharomyces cerevisiae by microarray hybridization. Mol Biol Cell 9(12): 3273-97.</li>
</ul>
<p>In this paper, Spellman and colleagues tried to identify all the genes in the yeast genome (&gt;6000 genes) that exhibited oscillatory behaviors suggestive of cell cycle regulation. To do so, they combined gene expression measurements from six different types of cell cycle synchronization experiments.</p>
<p><em>Download the Spellman data to your filesystem</em> from <a href="https://github.com/Bio723-class/example-datasets/blob/master/spellman-combined.txt">this link</a> (right-click the “Download” button and save to your Downloads folder or similar).</p>
<p>I suggest that once you download the data, you open it in a spreadsheet program (e.g. Excel) or use the RStudio Data Viewer to get a sense of what the data looks like.</p>
<p>Let’s load it into R, using the <code>read_tsv()</code> function, using the appropriate file path.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># the filepath may differ on your computer</span>
spellman &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;~/Downloads/spellman-combined.txt&quot;</span>)
Parsed with column specification<span class="op">:</span>
<span class="kw">cols</span>(
  <span class="dt">.default =</span> <span class="kw">col_double</span>(),
  <span class="dt">X1 =</span> <span class="kw">col_character</span>(),
  <span class="dt">clb =</span> <span class="kw">col_character</span>(),
  <span class="dt">alpha =</span> <span class="kw">col_character</span>(),
  <span class="dt">cdc15 =</span> <span class="kw">col_character</span>(),
  <span class="dt">cdc28 =</span> <span class="kw">col_character</span>(),
  <span class="dt">elu =</span> <span class="kw">col_character</span>()
)
See <span class="kw">spec</span>(...) <span class="cf">for</span> full column specifications.</code></pre></div>
<p>The initial dimenions of the data frame are:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(spellman)
[<span class="dv">1</span>] <span class="dv">6178</span>   <span class="dv">83</span></code></pre></div>
<p>The six types of cell cycle synchronization experiments are:</p>
<ol style="list-style-type: decimal">
<li>synchronization by alpha-factor = “alpha”</li>
<li>synchronization by cdc15 temperature sensitive mutants = “cdc15”</li>
<li>synchronization by cdc28 temperature sensitive mutants = “cdc28”</li>
<li>synchronization by elutration = “elu”</li>
<li>synchronization by cln3 mutatant strains = “cln3”</li>
<li>synchronization by clb2 mutant strains = “clb2”</li>
</ol>
</div>
<div id="renaming-data-frame-columms" class="section level2">
<h2><span class="header-section-number">7.3</span> Renaming data frame columms</h2>
<p>Notice that when we imported the data we got a warning message: <code>Missing column names filled in: 'X1' [1]</code>. In a data frame, every column must have a name. The first column of our data set did not have a name in the header, so <code>read_tsv</code> automatically gave it the name <code>X1</code>.</p>
<p>Our first task is to give the first column a more meaningful name. This column gives “systematic gene names” – a standardized naming scheme for genes in the yeast genome. We’ll use <code>dplyr::rename</code> to do rename <code>X1</code> to <code>gene</code>. Note that <code>rename</code> can take multiple arguments if you need to rename multiple columns simultaneously.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.clean &lt;-<span class="st"> </span>
<span class="st">  </span>spellman <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">rename</span>(<span class="dt">gene =</span> X1)</code></pre></div>
</div>
<div id="dropping-unneeded-columns" class="section level2">
<h2><span class="header-section-number">7.4</span> Dropping unneeded columns</h2>
<p>Take a look at the Spellman data again in your spreadsheet program (or the R Studio data viewer). You’ll notice there are some blank columns. These are simply visual organizing elements that the creator of the spreadsheet added to separate the different experiments that are included in the data set.</p>
<p>We can use <code>dplyr::select</code> to drop columns by prependingcolumn names with the negative sign:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># drop the alpha column keeping all others</span>
spellman.clean <span class="op">%&lt;&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>alpha)  </code></pre></div>
<p>Note that usually <code>select()</code> keeps only the variables you specify. However if the first expression is negative, <code>select</code> will instead automatically keep all variables, dropping only those you specify.</p>
<div id="finding-all-empty-columns" class="section level3">
<h3><span class="header-section-number">7.4.1</span> Finding all empty columns</h3>
<p>In the example above, we looked at the data and saw that the “alpha” column was empty, and thus dropped it. This worked because there are only a modest number of columns in the data frame in it’s initial form. However, if our data frame contained thousands of columns, this “look and see” procedure would not be efficient. Can we come up with a general solution for removing empty columns from a data frame?</p>
<p>When you load a data frame from a spreadsheet, empty cells are given the value <code>NA</code>. In previous class sessions we were introduced to the function <code>is.na()</code> which tests each value in a vector or data frame for whether it’s NA or not. We can count NA values in a vector by summing the output of <code>is.na()</code>. Conversely we can count the number of “not NA” items by using the negation operator (<code>!</code>):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># count number of NA values in the alpha0 column</span>
<span class="kw">sum</span>(<span class="kw">is.na</span>(spellman<span class="op">$</span>alpha0))
[<span class="dv">1</span>] <span class="dv">165</span>

<span class="co"># count number of values that are NOT NA in alpha0</span>
<span class="kw">sum</span>(<span class="op">!</span><span class="kw">is.na</span>(spellman<span class="op">$</span>alpha0))
[<span class="dv">1</span>] <span class="dv">6013</span></code></pre></div>
<p>This seems like it should get us close to a solution but <code>sum(is.na(..))</code> when applied to a data frame counts NAs across the entire data frame, not column-by-column.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># doesn&#39;t do what we hoped!</span>
<span class="kw">sum</span>(<span class="kw">is.na</span>(spellman))
[<span class="dv">1</span>] <span class="dv">59017</span></code></pre></div>
<p>If we want sums of NAs by column, we instead use the <code>colSums()</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get number of NAs by column</span>
<span class="kw">colSums</span>(<span class="kw">is.na</span>(spellman))
       X1    cln3<span class="op">-</span><span class="dv">1</span>    cln3<span class="op">-</span><span class="dv">2</span>       clb    clb2<span class="op">-</span><span class="dv">2</span>    clb2<span class="op">-</span><span class="dv">1</span>     alpha 
        <span class="dv">0</span>       <span class="dv">193</span>       <span class="dv">365</span>      <span class="dv">6178</span>       <span class="dv">454</span>       <span class="dv">142</span>      <span class="dv">6178</span> 
   alpha0    alpha7   alpha14   alpha21   alpha28   alpha35   alpha42 
      <span class="dv">165</span>       <span class="dv">525</span>       <span class="dv">191</span>       <span class="dv">312</span>       <span class="dv">267</span>       <span class="dv">207</span>       <span class="dv">123</span> 
  alpha49   alpha56   alpha63   alpha70   alpha77   alpha84   alpha91 
      <span class="dv">257</span>       <span class="dv">147</span>       <span class="dv">186</span>       <span class="dv">185</span>       <span class="dv">178</span>       <span class="dv">155</span>       <span class="dv">329</span> 
  alpha98  alpha105  alpha112  alpha119     cdc15  cdc15_<span class="dv">10</span>  cdc15_<span class="dv">30</span> 
      <span class="dv">209</span>       <span class="dv">174</span>       <span class="dv">222</span>       <span class="dv">251</span>      <span class="dv">6178</span>       <span class="dv">677</span>       <span class="dv">477</span> 
 cdc15_<span class="dv">50</span>  cdc15_<span class="dv">70</span>  cdc15_<span class="dv">80</span>  cdc15_<span class="dv">90</span> cdc15_<span class="dv">100</span> cdc15_<span class="dv">110</span> cdc15_<span class="dv">120</span> 
      <span class="dv">501</span>       <span class="dv">608</span>       <span class="dv">573</span>       <span class="dv">562</span>       <span class="dv">606</span>       <span class="dv">570</span>       <span class="dv">611</span> 
cdc15_<span class="dv">130</span> cdc15_<span class="dv">140</span> cdc15_<span class="dv">150</span> cdc15_<span class="dv">160</span> cdc15_<span class="dv">170</span> cdc15_<span class="dv">180</span> cdc15_<span class="dv">190</span> 
      <span class="dv">495</span>       <span class="dv">574</span>       <span class="dv">811</span>       <span class="dv">583</span>       <span class="dv">571</span>       <span class="dv">803</span>       <span class="dv">613</span> 
cdc15_<span class="dv">200</span> cdc15_<span class="dv">210</span> cdc15_<span class="dv">220</span> cdc15_<span class="dv">230</span> cdc15_<span class="dv">240</span> cdc15_<span class="dv">250</span> cdc15_<span class="dv">270</span> 
     <span class="dv">1014</span>       <span class="dv">573</span>       <span class="dv">741</span>       <span class="dv">596</span>       <span class="dv">847</span>       <span class="dv">379</span>       <span class="dv">537</span> 
cdc15_<span class="dv">290</span>     cdc28   cdc28_<span class="dv">0</span>  cdc28_<span class="dv">10</span>  cdc28_<span class="dv">20</span>  cdc28_<span class="dv">30</span>  cdc28_<span class="dv">40</span> 
      <span class="dv">426</span>      <span class="dv">6178</span>       <span class="dv">122</span>        <span class="dv">72</span>        <span class="dv">67</span>        <span class="dv">55</span>        <span class="dv">66</span> 
 cdc28_<span class="dv">50</span>  cdc28_<span class="dv">60</span>  cdc28_<span class="dv">70</span>  cdc28_<span class="dv">80</span>  cdc28_<span class="dv">90</span> cdc28_<span class="dv">100</span> cdc28_<span class="dv">110</span> 
       <span class="dv">56</span>        <span class="dv">82</span>        <span class="dv">84</span>        <span class="dv">75</span>       <span class="dv">237</span>       <span class="dv">165</span>       <span class="dv">319</span> 
cdc28_<span class="dv">120</span> cdc28_<span class="dv">130</span> cdc28_<span class="dv">140</span> cdc28_<span class="dv">150</span> cdc28_<span class="dv">160</span>       elu      elu0 
      <span class="dv">312</span>      <span class="dv">1439</span>      <span class="dv">2159</span>       <span class="dv">521</span>       <span class="dv">543</span>      <span class="dv">6178</span>       <span class="dv">122</span> 
    elu30     elu60     elu90    elu120    elu150    elu180    elu210 
      <span class="dv">153</span>       <span class="dv">175</span>       <span class="dv">132</span>       <span class="dv">103</span>       <span class="dv">119</span>       <span class="dv">111</span>       <span class="dv">118</span> 
   elu240    elu270    elu300    elu330    elu360    elu390 
      <span class="dv">131</span>       <span class="dv">110</span>       <span class="dv">112</span>       <span class="dv">112</span>       <span class="dv">156</span>       <span class="dv">114</span> </code></pre></div>
<p>Columns with <em>all</em> missing values can be more conveniently found by asking for those columns where the number of “not missing” values is zero:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get names of all columns for which all rows are NA</span>
<span class="co"># useing standard indexing</span>
<span class="kw">names</span>(spellman)[<span class="kw">colSums</span>(<span class="op">!</span><span class="kw">is.na</span>(spellman)) <span class="op">==</span><span class="st"> </span><span class="dv">0</span>]
[<span class="dv">1</span>] <span class="st">&quot;clb&quot;</span>   <span class="st">&quot;alpha&quot;</span> <span class="st">&quot;cdc15&quot;</span> <span class="st">&quot;cdc28&quot;</span> <span class="st">&quot;elu&quot;</span>  </code></pre></div>
<p>We can combine the <code>colSums(!is.na())</code> idiom with the <code>dplyr::select_if</code> function to quickly remove all empty columns as so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.clean <span class="op">%&lt;&gt;%</span>
<span class="st">  </span><span class="co"># keep ONLY the non-empty columns</span>
<span class="st">  </span><span class="kw">select_if</span>(<span class="kw">colSums</span>(<span class="op">!</span><span class="kw">is.na</span>(.)) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)</code></pre></div>
</div>
<div id="dropping-columns-by-matching-names" class="section level3">
<h3><span class="header-section-number">7.4.2</span> Dropping columns by matching names</h3>
<p>Only two time points from the cln3 and clb2 experiments were reported in the original publication. Since complete time series are unavailable for these two experimental conditions we will drop them from further consideration.</p>
<p><code>select()</code> can be called be called with a number of “helper function” (<code>?select_helpers</code>). Here we’ll illustrate the <code>matches()</code> helper function which matches column names to a “regular expression”. Regular expressions (also referred to as “regex” or “regexp”) are a way of specifying patterns in string. For the purposes of this document we’ll illustrate regexs by example; for a more detailed explanation of regular expressions see the the regex help(<code>?regex</code>) and the <a href="http://r4ds.had.co.nz/strings.html">Chapter on Strings in “R for Data Analysis”</a>:</p>
<p>Let’s see how to drop all the “cln3” and “clb2” columns from the data frame using <code>matches()</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.clean <span class="op">%&lt;&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span><span class="kw">matches</span>(<span class="st">&quot;cln3&quot;</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span><span class="kw">matches</span>(<span class="st">&quot;clb2&quot;</span>))</code></pre></div>
<p>If we wanted we could have collapsed our two match statements into one as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.clean <span class="op">%&lt;&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span><span class="kw">matches</span>(<span class="st">&quot;cln3|clb2&quot;</span>))</code></pre></div>
<p>In this second example, the character “|” is specifing an OR match within the regular expression, so this regular expression matches column names that contain “cln3” OR “clb2”.</p>
</div>
</div>
<div id="reshaping-data-with-tidyr" class="section level2">
<h2><span class="header-section-number">7.5</span> Reshaping data with tidyr</h2>
<p>The <code>tidyr</code> package provides functions for reshaping or tidying data frames. <code>tidyr</code> is yet another component of the tidyverse, and thus was loaded by the <code>library(tidyverse)</code>.</p>
<p>Today we’re going to look at two functions <code>tidyr::gather()</code> and <code>tidyr::extract()</code>, and how they can be combined with now familiar <code>dplyr</code> functions we’ve seen previously. The reading assignment for today’s class session covers a variety of other functions defined in <code>tidyr</code>.</p>
<p>The Spellman data, as I provided it to you, is in what we would call “wide” format. Each column (besides the <code>gene</code> column) corresponds to an experimental condition <em>and</em> time point. For example, “alpha0” is the alpha-factor experiment at time point 0 mins; “alpha7” is the alpha-factor experiment at time point 7 mins, etc. The cells within each column correspond to the expression of a corresponding gene (given by the first column which we renamed <code>gene</code>) in that particular experiment at that particular time point.</p>
<p>In every column (except “gene”), the cellsrepresents the same abstract property of interest – the expression of a gene of interest in a particular experiment/time point. Our first task will be to rearrange our “wide” data frame that consists of many different columns representing gene expression into a “long” data frame with just a single column representing expression. We’ll also create a new column to keep track of the which experiment and time point the measurement came from.</p>
<div id="wide-to-long-conversions-using-tidyrgather" class="section level3">
<h3><span class="header-section-number">7.5.1</span> Wide to long conversions using <code>tidyr::gather</code></h3>
<p><code>tidyr::gather()</code> takes multiple columns, and collapses them together into a smaller number of new columns. When using <code>gather()</code> you give the names of the <em>new </em>columns to create, as well as the names of any existing columns <code>gather</code> should <em>not</em> collect together.</p>
<p>Here we want to collapse all 73 or the expression columns – “alpha0” to “elu390” – into two columns: 1) a column to represent the expt/time point of the measurement, and 2) a column to represent the corresponding expression value. The only column we don’t want to touch is the <code>gene</code> column with our gene names.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># convert &quot;wide&quot; data to &quot;long&quot;</span>
spellman.long &lt;-
<span class="st">  </span>spellman.clean <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">gather</span>(expt.and.time, expression, <span class="op">-</span>gene)</code></pre></div>
<p>Take a moment to look at the data in the “long format”:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(spellman.long)
<span class="co"># A tibble: 6 x 3</span>
  gene    expt.and.time expression
  <span class="op">&lt;</span>chr<span class="op">&gt;</span><span class="st">   </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">              </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="dv">1</span> YAL001C alpha0           <span class="op">-</span><span class="fl">0.150</span> 
<span class="dv">2</span> YAL002W alpha0           <span class="op">-</span><span class="fl">0.110</span> 
<span class="dv">3</span> YAL003W alpha0           <span class="op">-</span><span class="fl">0.140</span> 
<span class="dv">4</span> YAL004W alpha0           <span class="op">-</span><span class="fl">0.0200</span>
<span class="dv">5</span> YAL005C alpha0           <span class="op">-</span><span class="fl">0.0500</span>
<span class="dv">6</span> YAL007C alpha0           <span class="op">-</span><span class="fl">0.600</span> </code></pre></div>
<p>And compare the dimensions of the wide data to the new data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(spellman.clean)  <span class="co"># for comparison</span>
[<span class="dv">1</span>] <span class="dv">6178</span>   <span class="dv">74</span>
<span class="kw">dim</span>(spellman.long)
[<span class="dv">1</span>] <span class="dv">450994</span>      <span class="dv">3</span></code></pre></div>
<p>As you see, we’ve gone from a data frame with 6178 rows and 74 columns (wide format), to a new data frame with 450994 rows and 3 columns (long format).</p>
</div>
<div id="extracting-information-from-combined-variables-using-tidyrextract" class="section level3">
<h3><span class="header-section-number">7.5.2</span> Extracting information from combined variables using <code>tidyr::extract</code></h3>
<p>The column <code>expt.and.time</code> violates one of our principles of tidy data: “Each variable must have its own column.”. This column conflates two different types of information – the experiment type and the time point of the measurement. Our next task is to split this information up into two new variables, which will help to facilitate downstream plotting and analysis.</p>
<p>One complicating factor is that the different experiments/time combinations have different naming conventions:</p>
<ul>
<li><p>The “alpha” and “elu” experiments are of the form “alpha0”, “alpha7”, “elu0”, “elu30”, etc. In this case, the first part of the string gives the experiment type (either alpha or elu) and the following digits give the time point.</p></li>
<li><p>In the “cdc15” and “cdc28” experiments the convention is slightly different; they are of the form “cdc15_0”, “cdc15_10”, “cdc28_0”, “cdc28_10”, etc. Here the part of the string before the underscore gives the experiment type, and the digits after the underscore give the time point.</p></li>
</ul>
<p>Because of the differences in naming conventions, we will find it easiest to break up <code>spellman.long</code> into a series of sub-data sets corresponding to each experiment type in order to extract out the experiment and time information. After processing each data subset separately, we will join the modified sub-data frames back together.</p>
</div>
<div id="subsetting-rows" class="section level3">
<h3><span class="header-section-number">7.5.3</span> Subsetting rows</h3>
<p>Let’s start by getting just the rows corresponding to the “alpha” experiment/times. Here we use <code>dplyr::filter</code> in combination with <code>stringr::str_detect</code> to get all those rows in which the <code>expt.and.time</code> variable contains the string “alpha”.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">alpha.long &lt;-<span class="st"> </span>
<span class="st">  </span>spellman.long <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(<span class="kw">str_detect</span>(expt.and.time, <span class="st">&quot;alpha&quot;</span>))

<span class="co"># look at the new data frame</span>
<span class="kw">dim</span>(alpha.long)
[<span class="dv">1</span>] <span class="dv">111204</span>      <span class="dv">3</span>
<span class="kw">head</span>(alpha.long, <span class="dt">n =</span> <span class="dv">10</span>)
<span class="co"># A tibble: 10 x 3</span>
   gene    expt.and.time expression
   <span class="op">&lt;</span>chr<span class="op">&gt;</span><span class="st">   </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">              </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="st"> </span><span class="dv">1</span> YAL001C alpha0           <span class="op">-</span><span class="fl">0.150</span> 
 <span class="dv">2</span> YAL002W alpha0           <span class="op">-</span><span class="fl">0.110</span> 
 <span class="dv">3</span> YAL003W alpha0           <span class="op">-</span><span class="fl">0.140</span> 
 <span class="dv">4</span> YAL004W alpha0           <span class="op">-</span><span class="fl">0.0200</span>
 <span class="dv">5</span> YAL005C alpha0           <span class="op">-</span><span class="fl">0.0500</span>
 <span class="dv">6</span> YAL007C alpha0           <span class="op">-</span><span class="fl">0.600</span> 
 <span class="dv">7</span> YAL008W alpha0           <span class="op">-</span><span class="fl">0.280</span> 
 <span class="dv">8</span> YAL009W alpha0           <span class="op">-</span><span class="fl">0.0300</span>
 <span class="dv">9</span> YAL010C alpha0           <span class="op">-</span><span class="fl">0.0500</span>
<span class="dv">10</span> YAL011W alpha0           <span class="op">-</span><span class="fl">0.310</span> </code></pre></div>
</div>
<div id="splitting-columns" class="section level3">
<h3><span class="header-section-number">7.5.4</span> Splitting columns</h3>
<p>Having subsetted the data, we can now split <code>expt.and.time</code> into two new variables – <code>expt</code> and <code>time</code>. To do this we use <code>tidyr::extract</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">alpha.long <span class="op">%&lt;&gt;%</span><span class="st"> </span>
<span class="st">  </span>tidyr<span class="op">::</span><span class="kw">extract</span>(expt.and.time,     <span class="co"># column we&#39;re extracting from</span>
                 <span class="kw">c</span>(<span class="st">&quot;expt&quot;</span>, <span class="st">&quot;time&quot;</span>), <span class="co"># new columns we&#39;re creating</span>
                 <span class="dt">regex=</span><span class="st">&quot;(alpha)([[:digit:]]+)&quot;</span>,  <span class="co"># regexp (see below)</span>
                 <span class="dt">convert=</span><span class="ot">TRUE</span>)      <span class="co"># automatically convert column types</span>

<span class="co"># </span><span class="al">NOTE</span><span class="co">: I&#39;m being explict about saying tidyr::extract because the</span>
<span class="co"># magrittr package defines a different extract function</span></code></pre></div>
<p>Let’s take a moment to look at the <code>regex</code> argument to <code>extract</code> – <code>regex=&quot;(alpha)([[:digit:]]+)&quot;</code>. The regex is specified as a character string. Each part we want to match <em>and</em> extract is surround by parentheses. In this case we have two sets of parentheses corresponding to the two matches we want to make. The first part of the regex is <code>(alpha)</code>; here we’re looking to make an exact match to the string “alpha”. The second part of the regex reads <code>([[:digit:]]+)</code>. <code>[[:digit:]]</code> indicates we’re looking for a numeric digit. The <code>+</code> after <code>[[:digit:]]</code> indicates that we want to match <em>one or more</em> digits (i.e. to get a match we need to find at least one digit, but more than one digit should also be a match).</p>
<p>Let’s take a look at the new version of <code>alpha.long</code> following application of <code>extract</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(alpha.long, <span class="dt">n =</span> <span class="dv">10</span>)
<span class="co"># A tibble: 10 x 4</span>
   gene    expt   time expression
   <span class="op">&lt;</span>chr<span class="op">&gt;</span><span class="st">   </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st"> </span><span class="er">&lt;</span>int<span class="op">&gt;</span><span class="st">      </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="st"> </span><span class="dv">1</span> YAL001C alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.150</span> 
 <span class="dv">2</span> YAL002W alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.110</span> 
 <span class="dv">3</span> YAL003W alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.140</span> 
 <span class="dv">4</span> YAL004W alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.0200</span>
 <span class="dv">5</span> YAL005C alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.0500</span>
 <span class="dv">6</span> YAL007C alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.600</span> 
 <span class="dv">7</span> YAL008W alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.280</span> 
 <span class="dv">8</span> YAL009W alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.0300</span>
 <span class="dv">9</span> YAL010C alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.0500</span>
<span class="dv">10</span> YAL011W alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.310</span> </code></pre></div>
<p>Notice our two new variables, both of which have appropriate types!</p>
<p>A data frame for the elutriation data can be created similarly:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">elu.long &lt;-
<span class="st">  </span>spellman.long <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(<span class="kw">str_detect</span>(expt.and.time, <span class="st">&quot;elu&quot;</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span>tidyr<span class="op">::</span><span class="kw">extract</span>(expt.and.time,     <span class="co"># column we&#39;re extracting from</span>
                 <span class="kw">c</span>(<span class="st">&quot;expt&quot;</span>, <span class="st">&quot;time&quot;</span>), <span class="co"># new columns we&#39;re creating</span>
                 <span class="dt">regex=</span><span class="st">&quot;(elu)([[:digit:]]+)&quot;</span>,  <span class="co"># regexp (see below)</span>
                 <span class="dt">convert=</span><span class="ot">TRUE</span>)      <span class="co"># automatically convert column types</span></code></pre></div>
<div id="a-fancier-regex-for-the-cdc-experiments" class="section level4">
<h4><span class="header-section-number">7.5.4.1</span> A fancier regex for the cdc experiments</h4>
<p>Now let’s process the cdc experiments (cdc15 and cdc28). As before we extract the corresponding rows of the data frame using <code>filter</code> and <code>str_detect</code>. We then split <code>expt.and.time</code> using <code>tidyr::extract</code>. In this case we carry out the two steps in a single code block using pipes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cdc.long &lt;-<span class="st"> </span>
<span class="st">  </span>spellman.long <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># both cdc15 and cdc28 contain &quot;cdc&quot; as a sub-string</span>
<span class="st">  </span><span class="kw">filter</span>(<span class="kw">str_detect</span>(expt.and.time, <span class="st">&quot;cdc&quot;</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span>tidyr<span class="op">::</span><span class="kw">extract</span>(expt.and.time, 
                 <span class="kw">c</span>(<span class="st">&quot;expt&quot;</span>, <span class="st">&quot;time&quot;</span>), 
                 <span class="dt">regex=</span><span class="st">&quot;(cdc15|cdc28)_([[:digit:]]+)&quot;</span>, <span class="co"># note the fancier regex</span>
                 <span class="dt">convert=</span><span class="ot">TRUE</span>)</code></pre></div>
<p>The regex – <code>&quot;(cdc15|cdc28)_([[:digit:]]+)&quot;</code> – is slightly fancier in this example. As before there are two parts we’re extracting: <code>(cdc15|cdc28)</code> and <code>([[:digit:]]+)</code>. The first parenthesized regexp is an “OR” – i.e. match “cdc15” <em>or</em> “cdc28”. The second parenthesized regexp is the same as we saw previously. Separating the two parenthesized regexps is an underscore (<code>_</code>). The underscore isn’t parenthesized because we only want to use it to make a match <em>not</em> to extract the corresponding match.</p>
</div>
</div>
<div id="combining-data-frame-rows" class="section level3">
<h3><span class="header-section-number">7.5.5</span> Combining data frame rows</h3>
<p>If you have two or more data frames with identical columns, the rows of the data frame can be combined using <code>rbind</code> (defined in the <code>base</code> package). For example, to reassemble the <code>alpha.long</code>, <code>elu.long</code>, and <code>cdc.long</code> data frames into a single data frame we do:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.final &lt;-<span class="st"> </span><span class="kw">rbind</span>(alpha.long, elu.long, cdc.long)

<span class="co"># check the dimensions of the new data frame</span>
<span class="kw">dim</span>(spellman.final)
[<span class="dv">1</span>] <span class="dv">450994</span>      <span class="dv">4</span></code></pre></div>
</div>
<div id="sorting-data-frame-rows" class="section level3">
<h3><span class="header-section-number">7.5.6</span> Sorting data frame rows</h3>
<p>Currently the <code>spellman.final</code> data frame is sorted by time point and experiment.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(spellman.final, <span class="dt">n =</span> <span class="dv">10</span>)
<span class="co"># A tibble: 10 x 4</span>
   gene    expt   time expression
   <span class="op">&lt;</span>chr<span class="op">&gt;</span><span class="st">   </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st"> </span><span class="er">&lt;</span>int<span class="op">&gt;</span><span class="st">      </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="st"> </span><span class="dv">1</span> YAL001C alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.150</span> 
 <span class="dv">2</span> YAL002W alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.110</span> 
 <span class="dv">3</span> YAL003W alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.140</span> 
 <span class="dv">4</span> YAL004W alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.0200</span>
 <span class="dv">5</span> YAL005C alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.0500</span>
 <span class="dv">6</span> YAL007C alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.600</span> 
 <span class="dv">7</span> YAL008W alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.280</span> 
 <span class="dv">8</span> YAL009W alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.0300</span>
 <span class="dv">9</span> YAL010C alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.0500</span>
<span class="dv">10</span> YAL011W alpha     <span class="dv">0</span>    <span class="op">-</span><span class="fl">0.310</span> </code></pre></div>
<p>It might be useful instead to sort by gene and experiment. To do this we can use <code>dplyr::arrange</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.final <span class="op">%&lt;&gt;%</span>
<span class="st">  </span><span class="kw">arrange</span>(gene, expt)

<span class="co"># look again at the rearranged data</span>
<span class="kw">head</span>(spellman.final, <span class="dt">n =</span> <span class="dv">10</span>)
<span class="co"># A tibble: 10 x 4</span>
   gene    expt   time expression
   <span class="op">&lt;</span>chr<span class="op">&gt;</span><span class="st">   </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st"> </span><span class="er">&lt;</span>int<span class="op">&gt;</span><span class="st">      </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="st"> </span><span class="dv">1</span> YAL001C alpha     <span class="dv">0</span>     <span class="op">-</span><span class="fl">0.150</span>
 <span class="dv">2</span> YAL001C alpha     <span class="dv">7</span>     <span class="op">-</span><span class="fl">0.150</span>
 <span class="dv">3</span> YAL001C alpha    <span class="dv">14</span>     <span class="op">-</span><span class="fl">0.210</span>
 <span class="dv">4</span> YAL001C alpha    <span class="dv">21</span>      <span class="fl">0.170</span>
 <span class="dv">5</span> YAL001C alpha    <span class="dv">28</span>     <span class="op">-</span><span class="fl">0.420</span>
 <span class="dv">6</span> YAL001C alpha    <span class="dv">35</span>     <span class="op">-</span><span class="fl">0.440</span>
 <span class="dv">7</span> YAL001C alpha    <span class="dv">42</span>     <span class="op">-</span><span class="fl">0.150</span>
 <span class="dv">8</span> YAL001C alpha    <span class="dv">49</span>      <span class="fl">0.240</span>
 <span class="dv">9</span> YAL001C alpha    <span class="dv">56</span>     <span class="op">-</span><span class="fl">0.100</span>
<span class="dv">10</span> YAL001C alpha    <span class="dv">63</span>     <span class="ot">NA</span>    </code></pre></div>
</div>
<div id="collapsing-columns-using-tidyrunite" class="section level3">
<h3><span class="header-section-number">7.5.7</span> Collapsing columns using <code>tidyr::unite</code></h3>
<p><code>unite</code> combines the data from multiple columns into one, separating the column data with a consistent separator character (the underscore character “_&quot; by default). For example, if we wanted to combined the experiment and time columns again (recall that we previously separated with with <code>extract</code>) we could do:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.recombined &lt;-
<span class="st">  </span>spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">unite</span>(expt.and.time, expt, time)  <span class="co"># give new column name first</span>
                                    <span class="co"># then give columns to combine</span>

<span class="kw">dim</span>(spellman.recombined)
[<span class="dv">1</span>] <span class="dv">450994</span>      <span class="dv">3</span>
<span class="kw">head</span>(spellman.recombined, <span class="dt">n =</span> <span class="dv">10</span>)
<span class="co"># A tibble: 10 x 3</span>
   gene    expt.and.time expression
   <span class="op">&lt;</span>chr<span class="op">&gt;</span><span class="st">   </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">              </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="st"> </span><span class="dv">1</span> YAL001C alpha_<span class="dv">0</span>           <span class="op">-</span><span class="fl">0.150</span>
 <span class="dv">2</span> YAL001C alpha_<span class="dv">7</span>           <span class="op">-</span><span class="fl">0.150</span>
 <span class="dv">3</span> YAL001C alpha_<span class="dv">14</span>          <span class="op">-</span><span class="fl">0.210</span>
 <span class="dv">4</span> YAL001C alpha_<span class="dv">21</span>           <span class="fl">0.170</span>
 <span class="dv">5</span> YAL001C alpha_<span class="dv">28</span>          <span class="op">-</span><span class="fl">0.420</span>
 <span class="dv">6</span> YAL001C alpha_<span class="dv">35</span>          <span class="op">-</span><span class="fl">0.440</span>
 <span class="dv">7</span> YAL001C alpha_<span class="dv">42</span>          <span class="op">-</span><span class="fl">0.150</span>
 <span class="dv">8</span> YAL001C alpha_<span class="dv">49</span>           <span class="fl">0.240</span>
 <span class="dv">9</span> YAL001C alpha_<span class="dv">56</span>          <span class="op">-</span><span class="fl">0.100</span>
<span class="dv">10</span> YAL001C alpha_<span class="dv">63</span>          <span class="ot">NA</span>    </code></pre></div>
</div>
<div id="splitting-columns-using-tidyrseparate" class="section level3">
<h3><span class="header-section-number">7.5.8</span> Splitting columns using <code>tidyr::separate</code></h3>
<p><code>separate</code> splits a column into multiple columns. It’s very much like <code>extract</code> except that it only takes a single regex for the separator, which by default is any non-alphabetic or non-numeric character. For example, to re-split <code>expt.and.time</code> in <code>spellman.recombined</code> we could do:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.respread &lt;-<span class="st"> </span>
<span class="st">  </span>spellman.recombined <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">separate</span>(expt.and.time, <span class="kw">c</span>(<span class="st">&quot;expt&quot;</span>, <span class="st">&quot;time&quot;</span>), <span class="dt">convert =</span> <span class="ot">TRUE</span>)

<span class="kw">dim</span>(spellman.respread)
[<span class="dv">1</span>] <span class="dv">450994</span>      <span class="dv">4</span>
<span class="kw">head</span>(spellman.respread,<span class="dt">n=</span> <span class="dv">10</span>)
<span class="co"># A tibble: 10 x 4</span>
   gene    expt   time expression
   <span class="op">&lt;</span>chr<span class="op">&gt;</span><span class="st">   </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st"> </span><span class="er">&lt;</span>int<span class="op">&gt;</span><span class="st">      </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="st"> </span><span class="dv">1</span> YAL001C alpha     <span class="dv">0</span>     <span class="op">-</span><span class="fl">0.150</span>
 <span class="dv">2</span> YAL001C alpha     <span class="dv">7</span>     <span class="op">-</span><span class="fl">0.150</span>
 <span class="dv">3</span> YAL001C alpha    <span class="dv">14</span>     <span class="op">-</span><span class="fl">0.210</span>
 <span class="dv">4</span> YAL001C alpha    <span class="dv">21</span>      <span class="fl">0.170</span>
 <span class="dv">5</span> YAL001C alpha    <span class="dv">28</span>     <span class="op">-</span><span class="fl">0.420</span>
 <span class="dv">6</span> YAL001C alpha    <span class="dv">35</span>     <span class="op">-</span><span class="fl">0.440</span>
 <span class="dv">7</span> YAL001C alpha    <span class="dv">42</span>     <span class="op">-</span><span class="fl">0.150</span>
 <span class="dv">8</span> YAL001C alpha    <span class="dv">49</span>      <span class="fl">0.240</span>
 <span class="dv">9</span> YAL001C alpha    <span class="dv">56</span>     <span class="op">-</span><span class="fl">0.100</span>
<span class="dv">10</span> YAL001C alpha    <span class="dv">63</span>     <span class="ot">NA</span>    </code></pre></div>
<p>Faced with the task of splitting columns, try <code>tidyr::separate</code> first and only consider <code>tidyr::extract</code> if you need to defines more sophisticated rules for splitting.</p>
</div>
<div id="long-to-wide-conversion-using-tidyrspread" class="section level3">
<h3><span class="header-section-number">7.5.9</span> Long-to-wide conversion using <code>tidyr::spread</code></h3>
<p><code>tidyr::spread</code> is the inverse of <code>tidyr::gather</code>. <code>gather()</code> took multiple columns and collapsed them together into a smaller number of new columns. The <code>tidyr</code> documentation calls this “collapsing into key-value pairs”. By contrast, <code>spread()</code> creates new columns by spreading “key-value pairs” (a column representing the “keys” and a column reprsenting the “values”) into multiple columns.</p>
<p>Here let’s use spread to use the gene names (the “key”) and expression measures (the “values”) to create a new data frame where the genes are the primary variables (columns) of the data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.gene.focused &lt;-
<span class="st">  </span>spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">spread</span>(gene, expression)

<span class="kw">dim</span>(spellman.gene.focused)
[<span class="dv">1</span>]   <span class="dv">73</span> <span class="dv">6180</span></code></pre></div>
<p>Both “long” and “wide” tidy versions of the same data are useful, depending on the specific analyses and/or visualiztions you want to accomplish. We’ll look at applications of both in the sections that follow.</p>
</div>
</div>
<div id="using-your-tidy-data" class="section level2">
<h2><span class="header-section-number">7.6</span> Using your tidy data</h2>
<p>Whew – that was a fair amount of work to tidy our data! But having done so we can now carry out a wide variety of very powerful analyses.</p>
<div id="visualizing-gene-expression-time-series" class="section level3">
<h3><span class="header-section-number">7.6.1</span> Visualizing gene expression time series</h3>
<p>Let’s start by walking through a series of visualizations of gene expression time series. Each plot will show the expression of one or more genes, at different time points, in one or more experimental conditions. Our initial visualizations exploit the “long” versions of the tidy data.</p>
<p>First a single gene in a single experimental condition:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(expt <span class="op">==</span><span class="st"> &quot;alpha&quot;</span>, gene <span class="op">==</span><span class="st"> &quot;YAL022C&quot;</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Expression of YAL022C&quot;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-222-1.png" width="576" /></p>
<p>We can easily modify the above code block to visualize the expression of multiple genes of interest:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes.of.interest &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;YAL022C&quot;</span>, <span class="st">&quot;YAR018C&quot;</span>, <span class="st">&quot;YGR188C&quot;</span>)

spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>genes.of.interest, expt <span class="op">==</span><span class="st"> &quot;alpha&quot;</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression, <span class="dt">color =</span> gene)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Normalized expression&quot;</span>,
         <span class="dt">title =</span> <span class="st">&quot;Expression of multiple genes</span><span class="ch">\n</span><span class="st">following synchronization by alpha factor&quot;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-223-1.png" width="576" /> By employing <code>facet_wrap()</code> we can visualize the relationship between this set of genes in each of the experiment types:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>genes.of.interest) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression, <span class="dt">color =</span> gene)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>expt) <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Normalized expression&quot;</span>,
         <span class="dt">title =</span> <span class="st">&quot;Expression of Multiple Genes</span><span class="ch">\n</span><span class="st">Across experiments&quot;</span>) </code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-224-1.png" width="576" /></p>
<p>The different experimental treatments were carried out for varying lengths of time due to the differences in their physiological effects. Plotting them all on the same time scale can obscure that patterns of oscillation we might be interested in, so let’s modify our code block so that plots that share the same y-axis, but have differently scaled x-axes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>genes.of.interest) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression, <span class="dt">color =</span> gene)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>expt, <span class="dt">scales =</span> <span class="st">&quot;free_x&quot;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Normalized expression&quot;</span>,
         <span class="dt">title =</span> <span class="st">&quot;Expression of Multiple Genes</span><span class="ch">\n</span><span class="st">Across experiments&quot;</span>) </code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-225-1.png" width="576" /></p>
</div>
<div id="finding-the-most-variable-genes" class="section level3">
<h3><span class="header-section-number">7.6.2</span> Finding the most variable genes</h3>
<p>When dealing with vary large data sets, one ad hoc filtering criteria that is often employed is to focus on those variables that exhibit that greatest variation. To do this, we first need to order our variables (genes) by their variation. Let’s see how we can accomplish this using our long data frame:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">by.variance &lt;-
<span class="st">  </span>spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(gene) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">expression.var =</span> <span class="kw">var</span>(expression, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">arrange</span>(<span class="kw">desc</span>(expression.var))

<span class="kw">head</span>(by.variance)
<span class="co"># A tibble: 6 x 2</span>
  gene    expression.var
  <span class="op">&lt;</span>chr<span class="op">&gt;</span><span class="st">            </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="dv">1</span> YLR286C           <span class="fl">2.16</span>
<span class="dv">2</span> YNR067C           <span class="fl">1.73</span>
<span class="dv">3</span> YNL327W           <span class="fl">1.65</span>
<span class="dv">4</span> YGL028C           <span class="fl">1.57</span>
<span class="dv">5</span> YHL028W           <span class="fl">1.52</span>
<span class="dv">6</span> YKL164C           <span class="fl">1.52</span></code></pre></div>
<p>The code above calculates variance but ignores the fact that we have different experimental conditions. To take into account the experimental design of the data at hand, let’s calculate the average variance across the experimental conditions:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">by.avg.variance &lt;-
<span class="st">  </span>spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(gene, expt) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">expression.var =</span> <span class="kw">var</span>(expression, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(gene) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">avg.expression.var =</span> <span class="kw">mean</span>(expression.var)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">arrange</span>(<span class="kw">desc</span>(avg.expression.var))

<span class="kw">head</span>(by.avg.variance)
<span class="co"># A tibble: 6 x 2</span>
  gene    avg.expression.var
  <span class="op">&lt;</span>chr<span class="op">&gt;</span><span class="st">                </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="dv">1</span> YFR014C               <span class="fl">3.58</span>
<span class="dv">2</span> YFR053C               <span class="fl">2.38</span>
<span class="dv">3</span> YBL032W               <span class="fl">2.30</span>
<span class="dv">4</span> YDR274C               <span class="fl">2.17</span>
<span class="dv">5</span> YLR286C               <span class="fl">2.13</span>
<span class="dv">6</span> YMR206W               <span class="fl">1.94</span></code></pre></div>
<p>Based on the average experession variance across experimental conditions, let’s get the names of the 1000 most variable genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">top.genes.1k &lt;-<span class="st"> </span>by.avg.variance[<span class="dv">1</span><span class="op">:</span><span class="dv">1000</span>,]<span class="op">$</span>gene

<span class="kw">head</span>(top.genes.1k)
[<span class="dv">1</span>] <span class="st">&quot;YFR014C&quot;</span> <span class="st">&quot;YFR053C&quot;</span> <span class="st">&quot;YBL032W&quot;</span> <span class="st">&quot;YDR274C&quot;</span> <span class="st">&quot;YLR286C&quot;</span> <span class="st">&quot;YMR206W&quot;</span></code></pre></div>
</div>
<div id="exploring-bivariate-relationships-using-wide-data" class="section level3">
<h3><span class="header-section-number">7.6.3</span> Exploring bivariate relationships using “wide” data</h3>
<p>The “long” version of our data frame proved useful for exploring how gene expression changed over time. By contrast, our “wide” data frame is more convient for exploring how pairs of genes covary together. For example, we can generate bivariate scatter plots depicting the relationship between two genes in the four different experimental conditions:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.gene.focused <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">is.na</span>(YAL022C) <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(YAR018C)) <span class="op">%&gt;%</span><span class="st">  </span><span class="co"># filter any missing values</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> YAL022C, <span class="dt">y =</span> YAR018C)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">aspect.ratio =</span> <span class="dv">1</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span>expt, <span class="dt">nrow =</span> <span class="dv">2</span>, <span class="dt">ncol =</span> <span class="dv">2</span>) </code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-229-1.png" width="576" /></p>
</div>
<div id="calculating-correlations-between-genes" class="section level3">
<h3><span class="header-section-number">7.6.4</span> Calculating correlations between genes</h3>
<p>Let’s explore more global bivariate relationships by calculating all the pairwise correlations between genes, focusing our attention on only the top 1000 most variable genes (as calculated previously).</p>
<p>Correlation is a measure of linear association between a pair of variables, and ranges from -1 to 1. A value near zero indicates the variables are uncorrelated (no linear association), while values approaching +1 indicate a strong positive association (the variables tend to get bigger or smaller together) while values near -1 indicate strong negative association (when one variable is larger, the other tends to be small).</p>
<p>First, let’s reduce our “wide” data frame to including only the top 1000 genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">top.1k.wide &lt;-
<span class="st">  </span>spellman.gene.focused <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># get expt and time column plus top 1000 genes</span>
<span class="st">  </span><span class="kw">select</span>(expt, time, top.genes.1k) 

<span class="kw">dim</span>(top.1k.wide)
[<span class="dv">1</span>]   <span class="dv">73</span> <span class="dv">1002</span></code></pre></div>
<p>Now, we calculate the correlations between all those genes. We need to drop two columns from the data frame, corresponding to “expt” and “time”, and we then use the <code>cor</code> function to calculate the pairwise correlations. Since there is missing data (indicated with NAs) we also need to tell the <code>cor</code> function to only use pairwise complete observations when calculating correlations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.cor &lt;-<span class="st"> </span>
<span class="st">  </span>top.1k.wide <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>expt, <span class="op">-</span>time)  <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># drop expt and time</span>
<span class="st">  </span><span class="kw">cor</span>(<span class="dt">use =</span> <span class="st">&quot;pairwise.complete.obs&quot;</span>)</code></pre></div>
<p>The correlation matrix is a square matrix with the number of rows and columns equal to the number of variables:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(spellman.cor)
[<span class="dv">1</span>] <span class="dv">1000</span> <span class="dv">1000</span></code></pre></div>
<p>To get the correlations with a gene of interest, we can index with the gene name on the rows of the correlation matrix. For example, to get all the correlations between the gene YAL022C and the other genes in the top 1000 set:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.cor[<span class="st">&quot;YAL022C&quot;</span>,]  
      YFR014C       YFR053C       YBL032W       YDR274C       YLR286C 
<span class="op">-</span><span class="fl">0.0286883534</span>  <span class="fl">0.1061796423</span>  <span class="fl">0.3143767286</span> <span class="op">-</span><span class="fl">0.3151280104</span> <span class="op">-</span><span class="fl">0.0844089765</span> 
      YMR206W       YCRX11W       YNR067C       YDR337W       YNL327W 
<span class="op">-</span><span class="fl">0.2115248731</span>  <span class="fl">0.1159781117</span>  <span class="fl">0.3065585116</span>  <span class="fl">0.4778333353</span>  <span class="fl">0.2868559029</span> 
      YCR104W       YMR218C       YGL028C       YHL028W       YER124C 
<span class="op">-</span><span class="fl">0.0440531822</span>  <span class="fl">0.0381755416</span>  <span class="fl">0.1330163451</span>  <span class="fl">0.4289999121</span>  <span class="fl">0.0359292260</span> 
      YGR117C       YKL164C       YGL037C       YBR090C       YCR100C 
 <span class="fl">0.0683951306</span>  <span class="fl">0.2656815642</span> <span class="op">-</span><span class="fl">0.0771173501</span>  <span class="fl">0.0833816882</span>  <span class="fl">0.0945617330</span> 
      YDL191W       YNL018C       YBR038W       YNR044W       YDL037C 
 <span class="fl">0.0114585067</span> <span class="op">-</span><span class="fl">0.1370538983</span>  <span class="fl">0.6767483573</span> <span class="op">-</span><span class="fl">0.0501138796</span>  <span class="fl">0.1399467371</span> 
      YBR054W       YHR143W       YNL296W       YBR009C       YGL055W 
 <span class="fl">0.5874136129</span> <span class="op">-</span><span class="fl">0.1017829045</span>  <span class="fl">0.2414025877</span> <span class="op">-</span><span class="fl">0.0815754480</span> <span class="op">-</span><span class="fl">0.1342443192</span> 
      YHR152W       YPR119W       YBR040W       YER070W       YKL125W 
 <span class="fl">0.5685684046</span>  <span class="fl">0.4944216448</span> <span class="op">-</span><span class="fl">0.2220889263</span> <span class="op">-</span><span class="fl">0.0625151935</span> <span class="op">-</span><span class="fl">0.2427881831</span> 
      YKL185W       YOR255W       YBL002W       YGR108W       YMR244W 
 <span class="fl">0.3534584753</span>  <span class="fl">0.0513256848</span> <span class="op">-</span><span class="fl">0.0288862913</span>  <span class="fl">0.5262281195</span> <span class="op">-</span><span class="fl">0.4018006560</span> 
      YPL061W       YNL160W       YDR033W       YPR149W       YER150W 
 <span class="fl">0.3761353441</span>  <span class="fl">0.4869534960</span>  <span class="fl">0.6454767678</span>  <span class="fl">0.4839791735</span> <span class="op">-</span><span class="fl">0.2357127137</span> 
      YPL256C       YNL030W       YCRX16C       YOL101C       YLR183C 
<span class="op">-</span><span class="fl">0.2039214092</span> <span class="op">-</span><span class="fl">0.1070530317</span>  <span class="fl">0.0638333407</span> <span class="op">-</span><span class="fl">0.2199840099</span> <span class="op">-</span><span class="fl">0.1128581074</span> 
      YBR051W       YMR297W       YHR160C       YPL163C       YMR032W 
<span class="op">-</span><span class="fl">0.3088116505</span>  <span class="fl">0.0430893117</span> <span class="op">-</span><span class="fl">0.2429794546</span> <span class="op">-</span><span class="fl">0.0212938980</span>  <span class="fl">0.5425463399</span> 
      YDR461W       YBR088C       YDR225W       YKR039W       YCR098C 
<span class="op">-</span><span class="fl">0.1908170286</span> <span class="op">-</span><span class="fl">0.1650108329</span> <span class="op">-</span><span class="fl">0.1242027461</span>  <span class="fl">0.2969274750</span> <span class="op">-</span><span class="fl">0.1891684344</span> 
      YKL163W       YGR087C       YBR158W       YOL007C       YMR058W 
 <span class="fl">0.2996797052</span> <span class="op">-</span><span class="fl">0.1152862904</span>  <span class="fl">0.1378974832</span> <span class="op">-</span><span class="fl">0.2469726169</span>  <span class="fl">0.3907602438</span> 
      YKR013W       YFR025C       YBR089W       YGR044C       YBR010W 
<span class="op">-</span><span class="fl">0.3376738415</span>  <span class="fl">0.1468963065</span> <span class="op">-</span><span class="fl">0.1181068439</span> <span class="op">-</span><span class="fl">0.0311776847</span> <span class="op">-</span><span class="fl">0.1542630984</span> 
      YBR092C       YMR001C       YNL020C       YPL158C       YDR224C 
 <span class="fl">0.6833823397</span>  <span class="fl">0.4793964162</span>  <span class="fl">0.3016275342</span>  <span class="fl">0.3375996437</span> <span class="op">-</span><span class="fl">0.1941701493</span> 
      YPL021W       YMR305C       YKL096W       YBL003C       YDL125C 
<span class="op">-</span><span class="fl">0.1035945658</span> <span class="op">-</span><span class="fl">0.0593409840</span>  <span class="fl">0.2244099811</span> <span class="op">-</span><span class="fl">0.1246019635</span>  <span class="fl">0.1947516335</span> 
      YNL002C       YBL073W       YGL089C       YGR189C       YDL003W 
 <span class="fl">0.0425075311</span>  <span class="fl">0.3602637335</span>  <span class="fl">0.1429082759</span>  <span class="fl">0.0162498806</span> <span class="op">-</span><span class="fl">0.2185164033</span> 
      YDR327W       YGR109C       YHR162W       YBL037W       YMR215W 
 <span class="fl">0.1421919404</span>  <span class="fl">0.0182997926</span> <span class="op">-</span><span class="fl">0.0552174064</span>  <span class="fl">0.1453925831</span>  <span class="fl">0.1600066978</span> 
      YML027W       YHR055C       YHR053C       YJL153C       YCL064C 
<span class="op">-</span><span class="fl">0.1771427528</span> <span class="op">-</span><span class="fl">0.3997746469</span> <span class="op">-</span><span class="fl">0.3781016372</span> <span class="op">-</span><span class="fl">0.2530899200</span>  <span class="fl">0.0023790988</span> 
      YBR072W       YLR308W       YPL127C       YMR226C       YOR247W 
<span class="op">-</span><span class="fl">0.2851121860</span>  <span class="fl">0.0133499022</span> <span class="op">-</span><span class="fl">0.1847228615</span>  <span class="fl">0.0776092853</span>  <span class="fl">0.3205073786</span> 
      YDL149W       YPR156C       YNL058C       YGL188C       YPL265W 
<span class="op">-</span><span class="fl">0.0256392164</span>  <span class="fl">0.5703507173</span>  <span class="fl">0.7406816267</span>  <span class="fl">0.0826286206</span>  <span class="fl">0.2582815082</span> 
      YLL026W       YCR021C       YJL158C       YHR178W       YDL039C 
<span class="op">-</span><span class="fl">0.1537408516</span> <span class="op">-</span><span class="fl">0.0739169450</span>  <span class="fl">0.3691184823</span>  <span class="fl">0.0458664853</span>  <span class="fl">0.1323709280</span> 
      YBR108W       YBL036C       YKL086W       YHR215W       YIL140W 
 <span class="fl">0.1310492504</span>  <span class="fl">0.2746395513</span> <span class="op">-</span><span class="fl">0.3122257240</span>  <span class="fl">0.4479302089</span> <span class="op">-</span><span class="fl">0.2289151564</span> 
      YBL035C       YDR055W       YMR205C       YNR009W       YMR232W 
<span class="op">-</span><span class="fl">0.0798156989</span>  <span class="fl">0.3615737685</span>  <span class="fl">0.5013289098</span> <span class="op">-</span><span class="fl">0.0741334683</span> <span class="op">-</span><span class="fl">0.2504976667</span> 
      YAL005C       YPL267W       YDR340W       YJR082C       YJL157C 
 <span class="fl">0.0773313597</span> <span class="op">-</span><span class="fl">0.2090532185</span>  <span class="fl">0.0391939977</span> <span class="op">-</span><span class="fl">0.1413973109</span>  <span class="fl">0.5375692469</span> 
      YNL279W       YCR007C       YLR190W       YBR071W       YCL040W 
<span class="op">-</span><span class="fl">0.0584967941</span>  <span class="fl">0.0694421074</span>  <span class="fl">0.5699897298</span>  <span class="fl">0.0834237477</span>  <span class="fl">0.0855301516</span> 
      YJL159W       YJL079C       YNL057W       YDL038C       YLR049C 
 <span class="fl">0.3292349327</span>  <span class="fl">0.3181403942</span>  <span class="fl">0.7078795864</span> <span class="op">-</span><span class="fl">0.0202989479</span>  <span class="fl">0.1469737443</span> 
      YMR199W       YBR110W       YMR250W       YBR067C       YAR071W 
<span class="op">-</span><span class="fl">0.0561794307</span>  <span class="fl">0.1870388826</span> <span class="op">-</span><span class="fl">0.0454105037</span>  <span class="fl">0.0406480035</span>  <span class="fl">0.5579207051</span> 
      YOR263C       YPL240C       YJL170C       YER001W       YMR107W 
 <span class="fl">0.2380858365</span> <span class="op">-</span><span class="fl">0.2140106688</span> <span class="op">-</span><span class="fl">0.0169081501</span> <span class="op">-</span><span class="fl">0.3153651235</span>  <span class="fl">0.0577110087</span> 
      YIL141W       YFL012W       YOR248W       YOR025W       YJR004C 
<span class="op">-</span><span class="fl">0.2273536880</span>  <span class="fl">0.0127273347</span>  <span class="fl">0.2596667783</span>  <span class="fl">0.4540887836</span>  <span class="fl">0.1078818448</span> 
      YNL078W       YML047C       YER081W       YGL032C       YDR475C 
 <span class="fl">0.4099708355</span> <span class="op">-</span><span class="fl">0.0357221844</span> <span class="op">-</span><span class="fl">0.3001572147</span> <span class="op">-</span><span class="fl">0.0896118230</span> <span class="op">-</span><span class="fl">0.1542293261</span> 
      YBL051C       YNL082W       YGL021W       YDR097C       YHR137W 
 <span class="fl">0.1206507331</span>  <span class="fl">0.0897341739</span>  <span class="fl">0.4312426547</span> <span class="op">-</span><span class="fl">0.1605522395</span>  <span class="fl">0.2964205629</span> 
      YNL300W       YNL031C       YBR202W       YFL066C       YHR217C 
<span class="op">-</span><span class="fl">0.2778721995</span> <span class="op">-</span><span class="fl">0.2167212809</span>  <span class="fl">0.6889264789</span>  <span class="fl">0.2399656802</span> <span class="op">-</span><span class="fl">0.1037777337</span> 
      YKR086W       YPL186C       YPR160W       YKL178C       YDR146C 
<span class="op">-</span><span class="fl">0.1532220801</span>  <span class="fl">0.0873683068</span>  <span class="fl">0.0853699755</span>  <span class="fl">0.2694210070</span>  <span class="fl">0.5687971948</span> 
      YOR264W       YFL014W       YLL024C       YIL123W       YAL067C 
 <span class="fl">0.2919668997</span> <span class="op">-</span><span class="fl">0.3532417042</span>  <span class="fl">0.0591293552</span>  <span class="fl">0.2781267792</span> <span class="op">-</span><span class="fl">0.2441177846</span> 
      YLR300W       YPL058C       YBR124W       YML058W       YLR079W 
 <span class="fl">0.2096401531</span>  <span class="fl">0.1924970262</span>  <span class="fl">0.1650570585</span>  <span class="fl">0.3104729553</span>  <span class="fl">0.1367913935</span> 
      YAL004W       YFL052W       YDR258C       YPR089W       YCR005C 
 <span class="fl">0.1192950712</span>  <span class="fl">0.1557951261</span> <span class="op">-</span><span class="fl">0.1856847015</span> <span class="op">-</span><span class="fl">0.2566943201</span>  <span class="fl">0.0515486602</span> 
      YOL132W       YOR382W       YOL090W       YOR315W       YOR074C 
<span class="op">-</span><span class="fl">0.0863263324</span> <span class="op">-</span><span class="fl">0.0182262059</span> <span class="op">-</span><span class="fl">0.2418699878</span>  <span class="fl">0.6261722221</span> <span class="op">-</span><span class="fl">0.0821462975</span> 
      YMR179W       YCL024W       YPL187W       YLR131C       YMR189W 
<span class="op">-</span><span class="fl">0.0343346047</span> <span class="op">-</span><span class="fl">0.1592876382</span>  <span class="fl">0.1555191336</span>  <span class="fl">0.4432887893</span>  <span class="fl">0.3938863771</span> 
      YIL066C       YPL121C       YMR031C       YDR278C       YKR077W 
<span class="op">-</span><span class="fl">0.1864322080</span> <span class="op">-</span><span class="fl">0.1361752489</span>  <span class="fl">0.3642248838</span>  <span class="fl">0.1425420496</span>  <span class="fl">0.0725875637</span> 
      YDL123W       YDL055C       YMR317W       YKR012C       YJL115W 
<span class="op">-</span><span class="fl">0.0255725388</span> <span class="op">-</span><span class="fl">0.0842054357</span>  <span class="fl">0.0261073152</span> <span class="op">-</span><span class="fl">0.2689016878</span> <span class="op">-</span><span class="fl">0.3354834652</span> 
      YCL022C       YLR178C       YNR069C       YPL062W       YDR215C 
<span class="op">-</span><span class="fl">0.1228209908</span> <span class="op">-</span><span class="fl">0.2706472546</span> <span class="op">-</span><span class="fl">0.0708256297</span>  <span class="fl">0.1775223075</span>  <span class="fl">0.0624842812</span> 
      YLR194C       YPL153C       YGR248W       YIL158W       YGL184C 
 <span class="fl">0.0574184113</span> <span class="op">-</span><span class="fl">0.1250398451</span> <span class="op">-</span><span class="fl">0.1519125497</span>  <span class="fl">0.3695412028</span> <span class="op">-</span><span class="fl">0.3261795208</span> 
      YOR313C       YDR451C       YML034W       YLR013W       YHR182W 
 <span class="fl">0.7087445727</span>  <span class="fl">0.1223004252</span>  <span class="fl">0.3747740962</span>  <span class="fl">0.1155198497</span>  <span class="fl">0.1153829979</span> 
      YDR171W       YDR380W       YDR355C       YHR066W       YLR413W 
<span class="op">-</span><span class="fl">0.2177169286</span>  <span class="fl">0.3983529588</span> <span class="op">-</span><span class="fl">0.0521061808</span>  <span class="fl">0.2570740438</span>  <span class="fl">0.1160772861</span> 
      YBL013W       YAR007C       YLR217W       YNL192W       YDL222C 
 <span class="fl">0.2255186281</span> <span class="op">-</span><span class="fl">0.1966626306</span> <span class="op">-</span><span class="fl">0.3889552994</span>  <span class="fl">0.2105866451</span> <span class="op">-</span><span class="fl">0.1964242333</span> 
      YGR227W       YPR031W       YBR078W       YNL066W       YFR015C 
 <span class="fl">0.2710207098</span> <span class="op">-</span><span class="fl">0.2431605366</span>  <span class="fl">0.4682509499</span>  <span class="fl">0.5015769081</span>  <span class="fl">0.1452400258</span> 
      YLR136C       YNL007C       YDL179W       YCL027W       YGL121C 
<span class="op">-</span><span class="fl">0.2171544492</span> <span class="op">-</span><span class="fl">0.0814742528</span>  <span class="fl">0.3428984751</span> <span class="op">-</span><span class="fl">0.0224610024</span> <span class="op">-</span><span class="fl">0.3288397352</span> 
      YJR132W       YCLX07W       YJR119C       YGL182C       YDL160C 
<span class="op">-</span><span class="fl">0.0521454890</span> <span class="op">-</span><span class="fl">0.0691820776</span> <span class="op">-</span><span class="fl">0.1514407636</span> <span class="op">-</span><span class="fl">0.1085282368</span>  <span class="fl">0.1211280108</span> 
      YML128C       YER053C       YDR269C       YDR535C       YHR210C 
 <span class="fl">0.0208170384</span>  <span class="fl">0.0131635837</span>  <span class="fl">0.0342189899</span>  <span class="fl">0.3190638098</span> <span class="op">-</span><span class="fl">0.2171551339</span> 
      YDL227C       YGR142W       YEL065W       YGL116W       YDL127W 
 <span class="fl">0.0349337766</span> <span class="op">-</span><span class="fl">0.1824083842</span>  <span class="fl">0.5020000273</span>  <span class="fl">0.5475455794</span> <span class="op">-</span><span class="fl">0.0955093008</span> 
      YGL238W       YLR142W       YBR083W       YHR140W       YNL036W 
 <span class="fl">0.3341567242</span> <span class="op">-</span><span class="fl">0.0070662369</span>  <span class="fl">0.2623249676</span>  <span class="fl">0.1395351562</span>  <span class="fl">0.1617281941</span> 
      YEL011W       YDR309C       YOR383C       YDR507C       YDR039C 
 <span class="fl">0.1664855038</span>  <span class="fl">0.0237342685</span>  <span class="fl">0.1406353361</span>  <span class="fl">0.2244170615</span>  <span class="fl">0.2026812879</span> 
      YCL042W       YPL165C       YGR221C       YBR138C       YLL065W 
 <span class="fl">0.0133317858</span> <span class="op">-</span><span class="fl">0.1657346436</span> <span class="op">-</span><span class="fl">0.1171675233</span>  <span class="fl">0.4017450077</span>  <span class="fl">0.2027991992</span> 
      YDL163W       YMR245W       YIL129C       YBR162C       YKL132C 
 <span class="fl">0.0184571398</span> <span class="op">-</span><span class="fl">0.0484193698</span> <span class="op">-</span><span class="fl">0.0004762559</span>  <span class="fl">0.4887184904</span> <span class="op">-</span><span class="fl">0.1394617233</span> 
      YML119W       YMR326C       YJR092W       YPL141C       YKL103C 
 <span class="fl">0.5686976677</span>  <span class="fl">0.1615925226</span>  <span class="fl">0.4984168377</span>  <span class="fl">0.4123613340</span> <span class="op">-</span><span class="fl">0.1781059131</span> 
      YPL136W       YLR109W       YML100W       YLR303W     YKL096W<span class="op">-</span>A 
<span class="op">-</span><span class="fl">0.0581992754</span> <span class="op">-</span><span class="fl">0.1487032106</span>  <span class="fl">0.0561414368</span> <span class="op">-</span><span class="fl">0.2176242658</span>  <span class="fl">0.3041131057</span> 
      YGL240W       YHR005C       YCL055W       YHR125W       YPR028W 
 <span class="fl">0.1172025071</span>  <span class="fl">0.2677456569</span> <span class="op">-</span><span class="fl">0.1810688595</span> <span class="op">-</span><span class="fl">0.3677622706</span>  <span class="fl">0.0704531199</span> 
      YJL051W       YKL172W       YJL078C       YGL230C       YHR023W 
 <span class="fl">0.4711181785</span> <span class="op">-</span><span class="fl">0.2721991245</span> <span class="op">-</span><span class="fl">0.0678252090</span> <span class="op">-</span><span class="fl">0.4688018634</span>  <span class="fl">0.5911430711</span> 
      YIL029C       YAR018C       YOR024W       YGR092W       YBR183W 
 <span class="fl">0.1402047310</span>  <span class="fl">0.6915321996</span>  <span class="fl">0.0876010033</span>  <span class="fl">0.6391449885</span>  <span class="fl">0.3918884624</span> 
      YHL021C       YDL164C       YBL108W       YAL054C       YMR011W 
 <span class="fl">0.0925680240</span>  <span class="fl">0.1199445452</span>  <span class="fl">0.4126670740</span>  <span class="fl">0.3895593119</span>  <span class="fl">0.2688081349</span> 
      YHR126C       YNL270C       YMR186W       YPL223C       YCL013W 
 <span class="fl">0.1049504910</span>  <span class="fl">0.0728111686</span> <span class="op">-</span><span class="fl">0.0454025754</span> <span class="op">-</span><span class="fl">0.0303845805</span>  <span class="fl">0.3194041454</span> 
      YDR300C       YGR152C       YJR150C      YCR097WA       YAL066W 
 <span class="fl">0.1400593164</span> <span class="op">-</span><span class="fl">0.2767921070</span>  <span class="fl">0.0965685215</span>  <span class="fl">0.2043400848</span> <span class="op">-</span><span class="fl">0.0170308913</span> 
      YJR010W       YPR018W       YJL116C       YIL025C       YMR210W 
<span class="op">-</span><span class="fl">0.2648643460</span> <span class="op">-</span><span class="fl">0.0732719999</span>  <span class="fl">0.0140673365</span> <span class="op">-</span><span class="fl">0.1236160878</span>  <span class="fl">0.1427579604</span> 
      YNL102W       YLR258W       YLR235C       YBL078C       YMR241W 
<span class="op">-</span><span class="fl">0.2677697358</span>  <span class="fl">0.0224981675</span> <span class="op">-</span><span class="fl">0.1399501977</span> <span class="op">-</span><span class="fl">0.0022903055</span>  <span class="fl">0.3729657962</span> 
      YOL084W       YPR096C       YBR093C       YDL083C       YGR008C 
<span class="op">-</span><span class="fl">0.0990741926</span> <span class="op">-</span><span class="fl">0.0743702551</span>  <span class="fl">0.5109771486</span> <span class="op">-</span><span class="fl">0.1311365154</span> <span class="op">-</span><span class="fl">0.0970744867</span> 
      YFL056C       YFR057W       YPL130W       YDR273W       YGR041W 
<span class="op">-</span><span class="fl">0.1323317470</span> <span class="op">-</span><span class="fl">0.2684767856</span> <span class="op">-</span><span class="fl">0.1524079708</span>  <span class="fl">0.0185778762</span>  <span class="fl">0.0776420058</span> 
      YNL283C       YPL036W       YJR128W       YLR327C       YKL177W 
<span class="op">-</span><span class="fl">0.1123323508</span>  <span class="fl">0.4532999780</span>  <span class="fl">0.2113244606</span> <span class="op">-</span><span class="fl">0.2673150709</span>  <span class="fl">0.2493318807</span> 
      YDL158C       YHR096C     YCR024C<span class="op">-</span>A       YLR313C       YNL289W 
<span class="op">-</span><span class="fl">0.2429262961</span> <span class="op">-</span><span class="fl">0.0638636653</span>  <span class="fl">0.6119968093</span> <span class="op">-</span><span class="fl">0.2012132262</span> <span class="op">-</span><span class="fl">0.1125286079</span> 
      YOR348C       YIL131C       YDL023C       YFL011W       YBR144C 
 <span class="fl">0.4959467801</span>  <span class="fl">0.0187661671</span> <span class="op">-</span><span class="fl">0.1012676099</span>  <span class="fl">0.4561695374</span>  <span class="fl">0.0681327565</span> 
      YGR279C       YCL014W       YHL018W       YIR019C       YOR314W 
 <span class="fl">0.4353506460</span>  <span class="fl">0.2193095468</span>  <span class="fl">0.1337808337</span> <span class="op">-</span><span class="fl">0.0915904284</span>  <span class="fl">0.6125448152</span> 
      YGL094C       YGR291C       YPL242C       YCL025C       YER079W 
<span class="op">-</span><span class="fl">0.0096790827</span>  <span class="fl">0.1849534110</span>  <span class="fl">0.6205437471</span>  <span class="fl">0.2481660517</span>  <span class="fl">0.0471726391</span> 
      YOR066W       YDR528W       YER103W       YOL155C       YKL221W 
 <span class="fl">0.2719683473</span>  <span class="fl">0.0134378978</span> <span class="op">-</span><span class="fl">0.2314315566</span> <span class="op">-</span><span class="fl">0.1025696700</span> <span class="op">-</span><span class="fl">0.2295842331</span> 
      YIL037C       YLR452C       YMR105C       YKR093W       YOR173W 
<span class="op">-</span><span class="fl">0.0957571965</span>  <span class="fl">0.0021728878</span> <span class="op">-</span><span class="fl">0.2328899895</span>  <span class="fl">0.4295286499</span> <span class="op">-</span><span class="fl">0.2057425788</span> 
      YNL004W       YDL009C       YNL145W       YER095W       YFR012W 
 <span class="fl">0.0957906792</span> <span class="op">-</span><span class="fl">0.0770058636</span>  <span class="fl">0.1853350841</span> <span class="op">-</span><span class="fl">0.1351268061</span> <span class="op">-</span><span class="fl">0.4075311215</span> 
      YLR164W       YGL008C       YHR164C       YAR037W       YPL123C 
 <span class="fl">0.1235426772</span>  <span class="fl">0.6250411101</span>  <span class="fl">0.1285009879</span>  <span class="fl">0.3117726969</span>  <span class="fl">0.0315596856</span> 
      YNL200C       YPL192C       YDR536W       YDR102C       YAR040C 
 <span class="fl">0.1368255160</span> <span class="op">-</span><span class="fl">0.1619905827</span>  <span class="fl">0.0300023802</span>  <span class="fl">0.1402814467</span>  <span class="fl">0.3978319390</span> 
      YDL239C       YLR274W       YMR173W       YOR153W       YLR119W 
 <span class="fl">0.1021506249</span>  <span class="fl">0.4038939444</span> <span class="op">-</span><span class="fl">0.4424445064</span>  <span class="fl">0.3004830136</span>  <span class="fl">0.0488849797</span> 
      YML054C       YDL022W       YBR106W       YDR010C       YGR151C 
<span class="op">-</span><span class="fl">0.2662514487</span> <span class="op">-</span><span class="fl">0.0356716913</span>  <span class="fl">0.4682243347</span> <span class="op">-</span><span class="fl">0.2448715036</span> <span class="op">-</span><span class="fl">0.2553008768</span> 
      YBR070C       YGR035C       YHR189W       YLR365W       YBR157C 
<span class="op">-</span><span class="fl">0.0741180746</span>  <span class="fl">0.0537186873</span>  <span class="fl">0.0081871349</span>  <span class="fl">0.0294451733</span>  <span class="fl">0.0711230451</span> 
      YER106W       YLR103C       YOL058W       YGR160W       YML084W 
<span class="op">-</span><span class="fl">0.1099489906</span> <span class="op">-</span><span class="fl">0.2249608855</span> <span class="op">-</span><span class="fl">0.1091161131</span> <span class="op">-</span><span class="fl">0.0421362911</span> <span class="op">-</span><span class="fl">0.1359581020</span> 
      YGL259W       YML123C       YOR374W       YPR142C       YOR237W 
 <span class="fl">0.0566424272</span>  <span class="fl">0.3249126268</span>  <span class="fl">0.0462493659</span> <span class="op">-</span><span class="fl">0.3923410121</span> <span class="op">-</span><span class="fl">0.1317747253</span> 
      YGR088W       YJL074C       YDL124W       YFL026W       YMR078C 
<span class="op">-</span><span class="fl">0.1783575251</span> <span class="op">-</span><span class="fl">0.0980414466</span> <span class="op">-</span><span class="fl">0.1632200366</span>  <span class="fl">0.0593077220</span>  <span class="fl">0.1238530280</span> 
      YOR052C       YLR214W       YBL021C       YFR023W       YGL170C 
 <span class="fl">0.1411149498</span>  <span class="fl">0.3308321431</span>  <span class="fl">0.3064699451</span> <span class="op">-</span><span class="fl">0.3133130397</span> <span class="op">-</span><span class="fl">0.3352409661</span> 
      YMR280C       YJL194W       YCL060C       YBL065W       YMR193W 
 <span class="fl">0.0736936253</span>  <span class="fl">0.3295477584</span> <span class="op">-</span><span class="fl">0.2162226073</span> <span class="op">-</span><span class="fl">0.0903457244</span> <span class="op">-</span><span class="fl">0.0337282249</span> 
      YBR267W       YDL214C       YIL009W       YGL033W       YLL028W 
<span class="op">-</span><span class="fl">0.1243396792</span> <span class="op">-</span><span class="fl">0.3096142437</span>  <span class="fl">0.2782212820</span> <span class="op">-</span><span class="fl">0.1952406345</span>  <span class="fl">0.3868741336</span> 
      YMR159C       YCL008C       YCL012W       YHR153C       YKL043W 
 <span class="fl">0.1033460465</span> <span class="op">-</span><span class="fl">0.1027070022</span>  <span class="fl">0.3211380078</span> <span class="op">-</span><span class="fl">0.1302678321</span>  <span class="fl">0.2947125490</span> 
      YOR338W       YOL017W       YDL117W       YNL173C       YEL035C 
<span class="op">-</span><span class="fl">0.3307116633</span> <span class="op">-</span><span class="fl">0.2127698409</span>  <span class="fl">0.4627775749</span>  <span class="fl">0.1741116877</span> <span class="op">-</span><span class="fl">0.1335662092</span> 
      YLR304C       YBR056W       YAL061W       YHR095W       YBR296C 
 <span class="fl">0.3332536733</span>  <span class="fl">0.2112324346</span>  <span class="fl">0.1182255826</span> <span class="op">-</span><span class="fl">0.1954111464</span> <span class="op">-</span><span class="fl">0.1298652913</span> 
      YGR273C       YHR109W       YFL051C       YKL045W       YPR175W 
<span class="op">-</span><span class="fl">0.0120357702</span>  <span class="fl">0.1036574707</span>  <span class="fl">0.2207478044</span> <span class="op">-</span><span class="fl">0.1552190451</span> <span class="op">-</span><span class="fl">0.1934776849</span> 
      YNL046W       YLR121C       YPL092W       YER091C       YDL024C 
 <span class="fl">0.4363353327</span> <span class="op">-</span><span class="fl">0.3590433809</span>  <span class="fl">0.2262713737</span> <span class="op">-</span><span class="fl">0.0283736282</span>  <span class="fl">0.0917133927</span> 
      YMR071C       YGR236C       YNL142W       YMR198W       YCL076W 
 <span class="fl">0.0020979607</span>  <span class="fl">0.0312216382</span>  <span class="fl">0.3403679728</span> <span class="op">-</span><span class="fl">0.0533251185</span>  <span class="fl">0.1024406429</span> 
      YNL194C       YIL100W       YCL075W       YDR281C       YNL065W 
<span class="op">-</span><span class="fl">0.0902833613</span> <span class="op">-</span><span class="fl">0.2444417136</span>  <span class="fl">0.1504961987</span> <span class="op">-</span><span class="fl">0.0842010022</span> <span class="op">-</span><span class="fl">0.1680388051</span> 
      YGR159C       YOR273C       YML033W       YBR285W       YMR195W 
<span class="op">-</span><span class="fl">0.0153855534</span>  <span class="fl">0.3050982660</span>  <span class="fl">0.3657611982</span> <span class="op">-</span><span class="fl">0.1781703060</span>  <span class="fl">0.1011029028</span> 
      YER054C       YJL114W       YNL318C       YKL152C       YMR203W 
<span class="op">-</span><span class="fl">0.1812507152</span> <span class="op">-</span><span class="fl">0.0832641390</span> <span class="op">-</span><span class="fl">0.0216772191</span> <span class="op">-</span><span class="fl">0.0994651792</span>  <span class="fl">0.3099329921</span> 
      YCR010C       YNR019W       YDR339C       YDL234C       YLR216C 
 <span class="fl">0.3959373799</span>  <span class="fl">0.2653578819</span> <span class="op">-</span><span class="fl">0.0108908279</span> <span class="op">-</span><span class="fl">0.0195281178</span> <span class="op">-</span><span class="fl">0.3299933655</span> 
      YHR185C       YKL026C       YDR382W       YOR220W       YPR203W 
<span class="op">-</span><span class="fl">0.0873826579</span> <span class="op">-</span><span class="fl">0.1841897437</span>  <span class="fl">0.2582009222</span> <span class="op">-</span><span class="fl">0.2035233516</span>  <span class="fl">0.0836167519</span> 
      YDR070C       YDL204W       YDR343C       YJR048W       YNL022C 
<span class="op">-</span><span class="fl">0.3517625690</span> <span class="op">-</span><span class="fl">0.1185897539</span>  <span class="fl">0.2976565559</span>  <span class="fl">0.1832709085</span>  <span class="fl">0.1361726556</span> 
      YAL003W       YNL195C       YER062C       YGR225W       YKL029C 
 <span class="fl">0.3182406985</span> <span class="op">-</span><span class="fl">0.1596760627</span> <span class="op">-</span><span class="fl">0.2816092594</span>  <span class="fl">0.1154407509</span> <span class="op">-</span><span class="fl">0.0771589517</span> 
      YOR386W       YDR019C       YJL190C       YJR160C       YMR316W 
<span class="op">-</span><span class="fl">0.0020657625</span>  <span class="fl">0.5192861196</span>  <span class="fl">0.1216579003</span> <span class="op">-</span><span class="fl">0.1018574457</span> <span class="op">-</span><span class="fl">0.3407524789</span> 
      YPR106W       YJR079W       YBR073W       YJL028W       YBL049W 
<span class="op">-</span><span class="fl">0.3100498543</span> <span class="op">-</span><span class="fl">0.0265760420</span> <span class="op">-</span><span class="fl">0.1146425926</span> <span class="op">-</span><span class="fl">0.1750640129</span> <span class="op">-</span><span class="fl">0.0022388058</span> 
      YDL182W       YBR169C       YOR298W       YCLX09W       YPL250C 
 <span class="fl">0.2450049918</span> <span class="op">-</span><span class="fl">0.2480494885</span>  <span class="fl">0.0345725531</span>  <span class="fl">0.2279446999</span>  <span class="fl">0.0763432824</span> 
     YCR097WB       YAL022C       YDL175C       YKR103W       YNL141W 
 <span class="fl">0.0753779687</span>  <span class="fl">1.0000000000</span> <span class="op">-</span><span class="fl">0.1240959120</span>  <span class="fl">0.3016612680</span> <span class="op">-</span><span class="fl">0.1536980256</span> 
      YGR143W       YPR039W       YNL205C       YJL166W       YNR066C 
 <span class="fl">0.4852912327</span> <span class="op">-</span><span class="fl">0.2907586159</span> <span class="op">-</span><span class="fl">0.1285507665</span> <span class="op">-</span><span class="fl">0.0451235373</span> <span class="op">-</span><span class="fl">0.0928617100</span> 
      YDL198C       YLL035W       YPR141C       YIL053W       YGL034C 
 <span class="fl">0.3311884934</span>  <span class="fl">0.0978175944</span> <span class="op">-</span><span class="fl">0.1693149309</span>  <span class="fl">0.0066941634</span>  <span class="fl">0.1588912747</span> 
      YLR463C       YJL033W       YHL049C       YKL047W       YOL126C 
 <span class="fl">0.0074153530</span> <span class="op">-</span><span class="fl">0.1210940898</span>  <span class="fl">0.1588113850</span> <span class="op">-</span><span class="fl">0.2683146636</span>  <span class="fl">0.4193621676</span> 
      YPL222W       YBL112C       YDR113C       YNL134C       YPL093W 
<span class="op">-</span><span class="fl">0.1489950300</span>  <span class="fl">0.2491406611</span> <span class="op">-</span><span class="fl">0.4204508484</span> <span class="op">-</span><span class="fl">0.0070737926</span>  <span class="fl">0.0850157509</span> 
      YGL189C       YBR031W       YMR003W       YKL108W       YPL171C 
<span class="op">-</span><span class="fl">0.2643915024</span>  <span class="fl">0.3868470149</span> <span class="op">-</span><span class="fl">0.2342296173</span> <span class="op">-</span><span class="fl">0.2277813945</span>  <span class="fl">0.1108765822</span> 
      YJL038C       YPR112C       YFR017C       YGR138C       YLR406C 
<span class="op">-</span><span class="fl">0.2477970346</span> <span class="op">-</span><span class="fl">0.2053491107</span> <span class="op">-</span><span class="fl">0.1075600539</span>  <span class="fl">0.0531462687</span>  <span class="fl">0.2080488697</span> 
      YLR438W       YFL064C       YOR307C       YER145C       YLR340W 
 <span class="fl">0.2413475155</span>  <span class="fl">0.1294974308</span>  <span class="fl">0.1508481729</span>  <span class="fl">0.2260111202</span>  <span class="fl">0.2427754945</span> 
      YDR501W       YOL019W       YNL112W       YOL158C       YJL007C 
<span class="op">-</span><span class="fl">0.2013473639</span> <span class="op">-</span><span class="fl">0.0400500840</span> <span class="op">-</span><span class="fl">0.0115696542</span>  <span class="fl">0.5575230739</span>  <span class="fl">0.0782228348</span> 
      YOR289W       YCR012W       YDL173W       YDL169C       YLR198C 
<span class="op">-</span><span class="fl">0.2760190155</span>  <span class="fl">0.3352891755</span>  <span class="fl">0.0806025364</span> <span class="op">-</span><span class="fl">0.0850153057</span> <span class="op">-</span><span class="fl">0.0730561029</span> 
      YOR305W       YLR126C       YNL333W       YLR255C       YDR012W 
 <span class="fl">0.1391350929</span> <span class="op">-</span><span class="fl">0.1987578299</span> <span class="op">-</span><span class="fl">0.3562597193</span> <span class="op">-</span><span class="fl">0.2468157722</span>  <span class="fl">0.0262580984</span> 
      YDL081C       YDL075W       YDL061C     YJL205C<span class="op">-</span>A       YDR418W 
 <span class="fl">0.2324101215</span> <span class="op">-</span><span class="fl">0.0384625714</span>  <span class="fl">0.2272642816</span> <span class="op">-</span><span class="fl">0.0239053749</span>  <span class="fl">0.0602411328</span> 
      YPR157W       YDL048C       YHR158C       YMR048W       YDR525W 
 <span class="fl">0.3213687055</span> <span class="op">-</span><span class="fl">0.0392957221</span>  <span class="fl">0.1000145753</span> <span class="op">-</span><span class="fl">0.0836989353</span>  <span class="fl">0.2432478969</span> 
      YGR043C       YFL033C       YDR015C       YJL164C       YDR313C 
<span class="op">-</span><span class="fl">0.2191209881</span> <span class="op">-</span><span class="fl">0.1497238138</span> <span class="op">-</span><span class="fl">0.1589843849</span>  <span class="fl">0.2642468242</span> <span class="op">-</span><span class="fl">0.1402294085</span> 
      YJL196C       YGR103W       YOL069W       YOR135C       YNL042W 
 <span class="fl">0.3980307062</span> <span class="op">-</span><span class="fl">0.1547829581</span>  <span class="fl">0.1698733701</span>  <span class="fl">0.2530762972</span> <span class="op">-</span><span class="fl">0.0350372913</span> 
      YBL023C       YGL031C       YLR058C       YOR293W       YLR004C 
 <span class="fl">0.5940455669</span> <span class="op">-</span><span class="fl">0.1824712639</span>  <span class="fl">0.3797759775</span> <span class="op">-</span><span class="fl">0.0561342506</span> <span class="op">-</span><span class="fl">0.0575076335</span> 
      YGR067C       YKR042W       YJR030C       YHL047C       YDL018C 
 <span class="fl">0.1599652429</span>  <span class="fl">0.2985125857</span> <span class="op">-</span><span class="fl">0.1929857093</span>  <span class="fl">0.0337705080</span> <span class="op">-</span><span class="fl">0.2350767640</span> 
      YPL146C       YHR107C       YGL158W       YER139C       YNL077W 
<span class="op">-</span><span class="fl">0.2165028601</span>  <span class="fl">0.1851828292</span> <span class="op">-</span><span class="fl">0.2156654873</span>  <span class="fl">0.1930677132</span> <span class="op">-</span><span class="fl">0.1554539742</span> 
      YEL039C       YDL096C       YLR048W       YKL151C       YGR290W 
 <span class="fl">0.1039234601</span>  <span class="fl">0.0111438988</span>  <span class="fl">0.1680572976</span>  <span class="fl">0.1590248890</span>  <span class="fl">0.0705863167</span> 
      YMR090W       YDR358W       YOR365C       YLR054C       YOR308C 
<span class="op">-</span><span class="fl">0.1645030988</span>  <span class="fl">0.1039372305</span> <span class="op">-</span><span class="fl">0.2050535783</span> <span class="op">-</span><span class="fl">0.0357656357</span> <span class="op">-</span><span class="fl">0.0040816230</span> 
      YDL213C       YPL088W       YPR192W       YPL090C       YCR089W 
 <span class="fl">0.0181380487</span> <span class="op">-</span><span class="fl">0.1925753294</span>  <span class="fl">0.0635971187</span>  <span class="fl">0.2252144371</span>  <span class="fl">0.1628330050</span> 
      YMR096W       YOR136W       YJL045W       YMR196W       YHL037C 
 <span class="fl">0.2450831594</span>  <span class="fl">0.1803068929</span> <span class="op">-</span><span class="fl">0.2400930169</span>  <span class="fl">0.2877922806</span> <span class="op">-</span><span class="fl">0.1808651526</span> 
      YOR376W       YOR214C       YNR056C       YDR453C       YOL012C 
<span class="op">-</span><span class="fl">0.1205740953</span>  <span class="fl">0.1262113062</span>  <span class="fl">0.0003107212</span>  <span class="fl">0.1417024260</span>  <span class="fl">0.2321485275</span> 
      YKL035W       YIL106W       YJL144W       YFL037W       YLR124W 
 <span class="fl">0.0746189175</span>  <span class="fl">0.3642290395</span> <span class="op">-</span><span class="fl">0.2234743890</span>  <span class="fl">0.2888710643</span> <span class="op">-</span><span class="fl">0.0095850121</span> 
      YCR085W       YIL167W       YKL109W       YOL109W       YCR065W 
<span class="op">-</span><span class="fl">0.0080854253</span>  <span class="fl">0.4019825678</span>  <span class="fl">0.1109732616</span> <span class="op">-</span><span class="fl">0.0422770563</span> <span class="op">-</span><span class="fl">0.0325633470</span> 
      YPR174C       YML060W       YLR223C       YKL142W       YIL136W 
<span class="op">-</span><span class="fl">0.1314829841</span> <span class="op">-</span><span class="fl">0.2614855100</span> <span class="op">-</span><span class="fl">0.2288183441</span> <span class="op">-</span><span class="fl">0.2203866250</span> <span class="op">-</span><span class="fl">0.1513205079</span> 
      YNL262W       YER135C       YDR523C       YLR333C       YGR055W 
<span class="op">-</span><span class="fl">0.1320123817</span> <span class="op">-</span><span class="fl">0.1670435697</span> <span class="op">-</span><span class="fl">0.0016627085</span> <span class="op">-</span><span class="fl">0.0164488754</span>  <span class="fl">0.0450438634</span> 
      YPR006C       YKL009W       YER080W       YPL132W       YHR010W 
 <span class="fl">0.3674955086</span> <span class="op">-</span><span class="fl">0.1655186083</span>  <span class="fl">0.0165046294</span>  <span class="fl">0.1025320698</span> <span class="op">-</span><span class="fl">0.0624608752</span> 
      YCR013C       YHR061C       YBR116C       YBL064C       YCR101C 
 <span class="fl">0.3081825848</span> <span class="op">-</span><span class="fl">0.2679556927</span> <span class="op">-</span><span class="fl">0.1047680021</span>  <span class="fl">0.0124850069</span>  <span class="fl">0.2761553832</span> 
      YPR002W       YNL208W       YDR218C       YLL059C       YIR017C 
 <span class="fl">0.3361093933</span> <span class="op">-</span><span class="fl">0.1466419697</span> <span class="op">-</span><span class="fl">0.0789702963</span> <span class="op">-</span><span class="fl">0.2117711575</span> <span class="op">-</span><span class="fl">0.3276105997</span> 
      YOR100C       YPR158W       YDL077C       YDR018C       YNL037C 
 <span class="fl">0.2712857388</span> <span class="op">-</span><span class="fl">0.0932927073</span>  <span class="fl">0.0207133859</span> <span class="op">-</span><span class="fl">0.0640440715</span>  <span class="fl">0.2562815942</span> 
      YNL015W       YJL187C       YPL004C       YOR027W       YBL012C 
<span class="op">-</span><span class="fl">0.1604511549</span> <span class="op">-</span><span class="fl">0.1880758696</span>  <span class="fl">0.3163450676</span> <span class="op">-</span><span class="fl">0.2365835100</span>  <span class="fl">0.0172012731</span> 
      YER185W       YDR450W       YGR286C       YMR246W       YOR256C 
<span class="op">-</span><span class="fl">0.0832803307</span>  <span class="fl">0.1317957856</span> <span class="op">-</span><span class="fl">0.3111905571</span> <span class="op">-</span><span class="fl">0.0811341820</span>  <span class="fl">0.2417881618</span> 
      YMR095C       YIL102C       YLR056W       YBR126C       YBR019C 
<span class="op">-</span><span class="fl">0.0777476458</span>  <span class="fl">0.1405349762</span>  <span class="fl">0.2377638841</span> <span class="op">-</span><span class="fl">0.1378886696</span> <span class="op">-</span><span class="fl">0.0349753295</span> 
      YCL074W       YPR019W       YJL172W       YJL219W       YPR078C 
 <span class="fl">0.0079256996</span>  <span class="fl">0.5614892319</span>  <span class="fl">0.4842547457</span>  <span class="fl">0.1099039408</span> <span class="op">-</span><span class="fl">0.1054934363</span> 
      YMR323W       YPL012W       YBR250W       YGL225W       YHR149C 
 <span class="fl">0.2141990627</span>  <span class="fl">0.0393315430</span> <span class="op">-</span><span class="fl">0.1256912901</span>  <span class="fl">0.1983062627</span>  <span class="fl">0.0393305365</span> 
      YJL092W       YDR025W       YEL060C       YLR307W       YLR110C 
 <span class="fl">0.0935392211</span>  <span class="fl">0.1553105916</span> <span class="op">-</span><span class="fl">0.4405397392</span> <span class="op">-</span><span class="fl">0.1001015527</span>  <span class="fl">0.3764655382</span> 
      YAL038W       YDR285W       YNL339C       YNL309W       YBL109W 
 <span class="fl">0.0285008124</span> <span class="op">-</span><span class="fl">0.1349643377</span>  <span class="fl">0.1435032852</span> <span class="op">-</span><span class="fl">0.2637472003</span>  <span class="fl">0.0410542849</span> 
      YLR249W       YGL062W       YLR247C       YER111C       YGR188C 
 <span class="fl">0.3008056507</span>  <span class="fl">0.0746277873</span>  <span class="fl">0.1453713581</span>  <span class="fl">0.0201345875</span> <span class="op">-</span><span class="fl">0.4563998053</span> 
      YPR120C       YLR441C       YCR039C       YAL062W       YBR184W 
<span class="op">-</span><span class="fl">0.0950955961</span> <span class="op">-</span><span class="fl">0.1116331712</span> <span class="op">-</span><span class="fl">0.1841309439</span> <span class="op">-</span><span class="fl">0.3663836098</span> <span class="op">-</span><span class="fl">0.1603599444</span> 
      YDL176W       YPL191C       YJR148W       YBR189W       YER153C 
 <span class="fl">0.0547663691</span> <span class="op">-</span><span class="fl">0.2478675194</span>  <span class="fl">0.0348375119</span>  <span class="fl">0.1356618612</span> <span class="op">-</span><span class="fl">0.1333441548</span> 
      YMR072W       YNR014W       YKL091C       YBR191W       YOL160W 
 <span class="fl">0.1756777470</span> <span class="op">-</span><span class="fl">0.1466754765</span>  <span class="fl">0.0330376908</span>  <span class="fl">0.2145967754</span> <span class="op">-</span><span class="fl">0.3732563375</span> 
      YBR243C       YIL082W       YGR059W       YER003C       YAR062W 
 <span class="fl">0.1797519974</span> <span class="op">-</span><span class="fl">0.0898942274</span> <span class="op">-</span><span class="fl">0.0427629476</span> <span class="op">-</span><span class="fl">0.1643077836</span>  <span class="fl">0.3069126262</span> 
      YBR118W       YIL119C       YAL018C       YAL065C       YEL032W 
 <span class="fl">0.2332684874</span>  <span class="fl">0.2085215197</span>  <span class="fl">0.0307067440</span>  <span class="fl">0.1628335162</span>  <span class="fl">0.4245075135</span> 
      YKL067W       YIL117C       YOR114W       YAL025C       YML042W 
<span class="op">-</span><span class="fl">0.3060411291</span> <span class="op">-</span><span class="fl">0.1239500941</span> <span class="op">-</span><span class="fl">0.3350434771</span> <span class="op">-</span><span class="fl">0.1149068147</span>  <span class="fl">0.2715623526</span> 
      YNL009W       YFR026C       YMR076C       YCL026C       YGL090W 
 <span class="fl">0.0273621001</span> <span class="op">-</span><span class="fl">0.0216969397</span>  <span class="fl">0.0008008518</span> <span class="op">-</span><span class="fl">0.0259257059</span>  <span class="fl">0.1106609829</span> 
      YLR344W       YML110C       YLR061W       YPL044C       YDR250C 
<span class="op">-</span><span class="fl">0.0922629314</span>  <span class="fl">0.0668334030</span> <span class="op">-</span><span class="fl">0.0499234633</span> <span class="op">-</span><span class="fl">0.1368485113</span>  <span class="fl">0.0865348072</span> 
      YCR031C       YPR124W       YNL276C       YGR086C       YNL126W 
 <span class="fl">0.1483802379</span>  <span class="fl">0.1861815567</span> <span class="op">-</span><span class="fl">0.1558542529</span>  <span class="fl">0.1466462551</span> <span class="op">-</span><span class="fl">0.0573266366</span> 
      YGR148C       YJR109C       YEL075C       YNL269W       YMR269W 
<span class="op">-</span><span class="fl">0.1912792665</span>  <span class="fl">0.3148074083</span>  <span class="fl">0.0937848388</span>  <span class="fl">0.0140046343</span> <span class="op">-</span><span class="fl">0.2136840076</span> 
      YDR191W       YAR008W       YFL059W       YBR214W       YBL113C 
 <span class="fl">0.3766034739</span> <span class="op">-</span><span class="fl">0.0395493093</span> <span class="op">-</span><span class="fl">0.2813204810</span> <span class="op">-</span><span class="fl">0.1502025520</span>  <span class="fl">0.2717806327</span> 
      YDL130W       YGL138C       YIL057C       YGR280C       YJL037W 
 <span class="fl">0.1389708034</span>  <span class="fl">0.1018121335</span>  <span class="fl">0.1345948557</span> <span class="op">-</span><span class="fl">0.2309113396</span> <span class="op">-</span><span class="fl">0.2252513411</span> 
      YBR230C       YHR218W       YCL048W       YMR015C       YKR011C 
 <span class="fl">0.2296307774</span>  <span class="fl">0.1706201864</span>  <span class="fl">0.1398425833</span>  <span class="fl">0.2787421167</span> <span class="op">-</span><span class="fl">0.0678805200</span> 
      YHL045W       YGL204C       YGR224W       YHR097C       YNL073W 
<span class="op">-</span><span class="fl">0.0019326745</span> <span class="op">-</span><span class="fl">0.1968206948</span>  <span class="fl">0.4021516370</span> <span class="op">-</span><span class="fl">0.3395960036</span>  <span class="fl">0.2141287635</span> 
      YHR018C       YCL062W       YJL216C       YOR058C       YJL214W 
 <span class="fl">0.1111401000</span>  <span class="fl">0.0730572532</span>  <span class="fl">0.2183090917</span>  <span class="fl">0.1394356932</span>  <span class="fl">0.3977814168</span> 
      YPR204W       YOL023W       YPR077C       YBR247C       YPR150W 
 <span class="fl">0.1936245905</span>  <span class="fl">0.0734747017</span> <span class="op">-</span><span class="fl">0.0802740414</span> <span class="op">-</span><span class="fl">0.0472740025</span> <span class="op">-</span><span class="fl">0.1418420626</span> 
      YLR197W       YKL078W       YMR145C       YKL116C       YOR083W 
<span class="op">-</span><span class="fl">0.0621797609</span> <span class="op">-</span><span class="fl">0.1471203549</span>  <span class="fl">0.5079851829</span>  <span class="fl">0.3097069037</span>  <span class="fl">0.0589035001</span> 
      YER131W       YOL016C       YJL089W       YDR044W       YAL060W 
<span class="op">-</span><span class="fl">0.1640640905</span> <span class="op">-</span><span class="fl">0.3002781310</span>  <span class="fl">0.2127703663</span> <span class="op">-</span><span class="fl">0.1655642307</span>  <span class="fl">0.3858333051</span> 
      YFL053W       YBR188C       YHL041W       YJR095W       YPL054W 
 <span class="fl">0.3128835919</span>  <span class="fl">0.0121671863</span>  <span class="fl">0.2153828146</span>  <span class="fl">0.3985896266</span> <span class="op">-</span><span class="fl">0.5092294056</span> 
      YHR110W       YOR303W       YOL118C       YOR065W       YAR009C 
<span class="op">-</span><span class="fl">0.3716775946</span>  <span class="fl">0.2042701847</span>  <span class="fl">0.1568516725</span>  <span class="fl">0.1254604970</span>  <span class="fl">0.0333297210</span> 
      YBR241C       YMR239C       YMR321C       YMR128W       YIR027C 
<span class="op">-</span><span class="fl">0.0417405866</span>  <span class="fl">0.1720322649</span>  <span class="fl">0.2572081000</span> <span class="op">-</span><span class="fl">0.2078147453</span> <span class="op">-</span><span class="fl">0.3185905595</span> 
      YPL281C       YGR014W       YCR091W       YLR029C       YKL066W 
 <span class="fl">0.0789068758</span>  <span class="fl">0.1111947556</span>  <span class="fl">0.2210304214</span> <span class="op">-</span><span class="fl">0.0449059940</span> <span class="op">-</span><span class="fl">0.1380473715</span> 
      YIL104C       YKL001C       YCL061C       YNR074C       YOL059W 
<span class="op">-</span><span class="fl">0.1320158357</span> <span class="op">-</span><span class="fl">0.2889322399</span> <span class="op">-</span><span class="fl">0.2208711543</span>  <span class="fl">0.2593177399</span>  <span class="fl">0.1264937448</span> 
      YLL061W       YPL266W       YER065C       YLR383W       YBL111C 
<span class="op">-</span><span class="fl">0.1118581554</span> <span class="op">-</span><span class="fl">0.2152943487</span>  <span class="fl">0.2556167100</span> <span class="op">-</span><span class="fl">0.2864827411</span>  <span class="fl">0.2764584016</span> 
      YHR022C       YDR516C       YCRX09C       YLR466W       YGR084C 
 <span class="fl">0.2179272644</span>  <span class="fl">0.1560488279</span>  <span class="fl">0.0815185074</span>  <span class="fl">0.1490697953</span> <span class="op">-</span><span class="fl">0.0462873093</span> 
      YIL015W       YPL057C       YKR091W       YBR209W       YPR029C 
 <span class="fl">0.1361778150</span> <span class="op">-</span><span class="fl">0.1479436060</span> <span class="op">-</span><span class="fl">0.1077362566</span>  <span class="fl">0.3769450912</span> <span class="op">-</span><span class="fl">0.1373041150</span> 
      YLR372W       YOR312C       YCR009C       YDR399W       YOR134W 
 <span class="fl">0.0936782038</span> <span class="op">-</span><span class="fl">0.1037106095</span>  <span class="fl">0.1197879116</span>  <span class="fl">0.0393015887</span> <span class="op">-</span><span class="fl">0.0149337658</span> 
      YER032W       YLR081W       YBR117C       YDR042C       YDR136C 
 <span class="fl">0.0485483705</span>  <span class="fl">0.2648434351</span> <span class="op">-</span><span class="fl">0.2761839624</span> <span class="op">-</span><span class="fl">0.0985121147</span> <span class="op">-</span><span class="fl">0.0268150407</span> 
      YAR043C       YGL096W       YAL051W       YDR297W       YKL101W 
 <span class="fl">0.3569783388</span> <span class="op">-</span><span class="fl">0.3181403369</span> <span class="op">-</span><span class="fl">0.3109300965</span> <span class="op">-</span><span class="fl">0.1080769976</span> <span class="op">-</span><span class="fl">0.1120278976</span> 
      YEL054C       YLR374C       YOR183W       YJR146W       YLR342W 
 <span class="fl">0.3746970226</span> <span class="op">-</span><span class="fl">0.2273675049</span>  <span class="fl">0.1132929216</span> <span class="op">-</span><span class="fl">0.0502010724</span> <span class="op">-</span><span class="fl">0.0241602473</span> 
      YPL142C       YKL090W       YOL150C       YNL312W       YPR032W 
 <span class="fl">0.4649178256</span> <span class="op">-</span><span class="fl">0.1329417315</span>  <span class="fl">0.3283427655</span> <span class="op">-</span><span class="fl">0.1676677425</span>  <span class="fl">0.3867057830</span> 
      YBR266C       YPR030W       YJL088W       YCR034W       YCL067C 
<span class="op">-</span><span class="fl">0.1669415781</span>  <span class="fl">0.0148142909</span>  <span class="fl">0.0787666250</span>  <span class="fl">0.2940652651</span> <span class="op">-</span><span class="fl">0.2236403322</span> 
      YKR076W       YJL215C       YOR394W       YLR467W       YOR206W 
<span class="op">-</span><span class="fl">0.1381714108</span>  <span class="fl">0.2448770304</span> <span class="op">-</span><span class="fl">0.2251683095</span>  <span class="fl">0.1681982698</span> <span class="op">-</span><span class="fl">0.1961140804</span> 
      YPL087W       YOR391C       YDR256C       YPL033C       YDR374C 
 <span class="fl">0.0133421522</span> <span class="op">-</span><span class="fl">0.2183401873</span>  <span class="fl">0.1225832295</span> <span class="op">-</span><span class="fl">0.2482778086</span> <span class="op">-</span><span class="fl">0.0917347639</span> 
      YER102W       YNR001C       YIL175W       YKL008C       YGR039W 
<span class="op">-</span><span class="fl">0.0300505616</span>  <span class="fl">0.2550147886</span> <span class="op">-</span><span class="fl">0.1248154091</span> <span class="op">-</span><span class="fl">0.2530985066</span> <span class="op">-</span><span class="fl">0.1923337778</span> 
      YER110C       YER043C       YBL063W       YMR049C       YDL228C 
 <span class="fl">0.3046050770</span>  <span class="fl">0.3105959406</span> <span class="op">-</span><span class="fl">0.2034666934</span>  <span class="fl">0.0069776775</span>  <span class="fl">0.3571656370</span> 
      YHR209W       YKR041W       YPL102C       YOL014W       YER190W 
 <span class="fl">0.2620384201</span>  <span class="fl">0.0371119315</span> <span class="op">-</span><span class="fl">0.0279897043</span>  <span class="fl">0.1056049701</span>  <span class="fl">0.1255554679</span> 
      YBL075C       YOR029W       YDR046C       YJR037W       YDR040C 
 <span class="fl">0.0390750261</span>  <span class="fl">0.2139944224</span>  <span class="fl">0.3951934418</span>  <span class="fl">0.1008045420</span>  <span class="fl">0.0319597826</span> 
      YAR064W       YLR167W       YLR045C       YOR344C       YAL028W 
 <span class="fl">0.1819505950</span> <span class="op">-</span><span class="fl">0.0673020849</span> <span class="op">-</span><span class="fl">0.0379580312</span> <span class="op">-</span><span class="fl">0.1131816750</span>  <span class="fl">0.1237026718</span> 
      YHR139C       YLR236C       YKL153W       YCL059C       YLR212C 
<span class="op">-</span><span class="fl">0.2757014954</span>  <span class="fl">0.0006889656</span>  <span class="fl">0.0447586493</span> <span class="op">-</span><span class="fl">0.1629682607</span>  <span class="fl">0.0685096540</span> 
      YLL046C       YLR149C       YPR151C       YOR388C       YGL209W 
<span class="op">-</span><span class="fl">0.1207887412</span>  <span class="fl">0.0729031919</span> <span class="op">-</span><span class="fl">0.2505535151</span> <span class="op">-</span><span class="fl">0.1051335029</span>  <span class="fl">0.2395956063</span> 
      YOR347C       YDR214W       YBR238C       YLR162W       YCL030C 
 <span class="fl">0.1959635756</span> <span class="op">-</span><span class="fl">0.0266453978</span>  <span class="fl">0.1969892259</span> <span class="op">-</span><span class="fl">0.1630248849</span>  <span class="fl">0.1645320045</span> 
      YAR073W       YIL126W       YML045W       YCL023C       YGL255W 
 <span class="fl">0.4345615934</span> <span class="op">-</span><span class="fl">0.1723045257</span> <span class="op">-</span><span class="fl">0.2460875611</span> <span class="op">-</span><span class="fl">0.1200342196</span>  <span class="fl">0.3274183304</span> 
      YPR035W       YNL117W       YJR047C       YGL103W       YJR008W 
<span class="op">-</span><span class="fl">0.2518503203</span>  <span class="fl">0.2071469205</span>  <span class="fl">0.1600389474</span> <span class="op">-</span><span class="fl">0.0645014970</span> <span class="op">-</span><span class="fl">0.1135789649</span> 
      YJL195C       YPR201W       YLR318W       YDR446W       YNL175C 
 <span class="fl">0.3378482028</span>  <span class="fl">0.0555207851</span>  <span class="fl">0.0499707855</span> <span class="op">-</span><span class="fl">0.0394964115</span>  <span class="fl">0.1571383157</span> 
      YPR047W       YLL002W       YGR245C       YKL030W       YDL063C 
 <span class="fl">0.3028269956</span> <span class="op">-</span><span class="fl">0.0546395854</span> <span class="op">-</span><span class="fl">0.0362406428</span> <span class="op">-</span><span class="fl">0.0042540216</span> <span class="op">-</span><span class="fl">0.0844669466</span> 
      YJL163C       YMR081C       YOR182C       YNL072W       YGL228W 
 <span class="fl">0.1633698773</span> <span class="op">-</span><span class="fl">0.2148284098</span> <span class="op">-</span><span class="fl">0.2597132559</span>  <span class="fl">0.1246731584</span>  <span class="fl">0.0192954760</span> 
      YDL218W       YPR184W       YHR037W       YER189W       YBL039C 
 <span class="fl">0.0752723458</span> <span class="op">-</span><span class="fl">0.0786889173</span>  <span class="fl">0.1668945669</span>  <span class="fl">0.1581546399</span>  <span class="fl">0.2235528391</span> 
      YMR295C       YKR105C       YFL067W       YOR235W       YCLX02C 
<span class="op">-</span><span class="fl">0.0654890777</span> <span class="op">-</span><span class="fl">0.0229559592</span>  <span class="fl">0.4094981264</span>  <span class="fl">0.0182779307</span> <span class="op">-</span><span class="fl">0.1755292354</span> 
      YHR092C       YLL022C       YJL118W       YDL102W       YLR270W 
 <span class="fl">0.3286916445</span> <span class="op">-</span><span class="fl">0.3351218696</span> <span class="op">-</span><span class="fl">0.1537702984</span>  <span class="fl">0.1801550890</span>  <span class="fl">0.0028156806</span> 
      YDL128W       YHR172W       YDR345C       YDR112W       YLR353W 
 <span class="fl">0.5251076083</span> <span class="op">-</span><span class="fl">0.1624860646</span>  <span class="fl">0.4098222863</span> <span class="op">-</span><span class="fl">0.0146686904</span>  <span class="fl">0.4575062062</span> 
      YML120C       YBR076W       YJL108C       YJL217W       YDR540C 
 <span class="fl">0.3085445399</span>  <span class="fl">0.1526587348</span> <span class="op">-</span><span class="fl">0.1626661949</span>  <span class="fl">0.0198431863</span>  <span class="fl">0.3624761479</span> 
      YLR355C       YNL274C       YPL280W       YPR194C       YMR069W 
 <span class="fl">0.3874400681</span> <span class="op">-</span><span class="fl">0.0115377428</span>  <span class="fl">0.0322258646</span>  <span class="fl">0.2890661764</span> <span class="op">-</span><span class="fl">0.2232485453</span> 
      YDL244W       YBR074W       YFL016C       YOR385W       YOL077C 
 <span class="fl">0.2104042212</span>  <span class="fl">0.3489285917</span>  <span class="fl">0.1224921408</span> <span class="op">-</span><span class="fl">0.1086568943</span> <span class="op">-</span><span class="fl">0.1686662376</span> 
      YPR045C       YLR462W       YAL035W       YNL044W       YOR096W 
 <span class="fl">0.0926815132</span>  <span class="fl">0.1309320155</span>  <span class="fl">0.1369497149</span>  <span class="fl">0.2149857566</span>  <span class="fl">0.1689311218</span> 
      YHR180W       YJL148W       YOL119C       YGR003W       YCR042C 
<span class="op">-</span><span class="fl">0.2770801725</span> <span class="op">-</span><span class="fl">0.2495145484</span>  <span class="fl">0.3723574161</span>  <span class="fl">0.1657071382</span>  <span class="fl">0.4567219046</span> 
      YJL135W       YAL044C       YNL028W       YFR054C       YLR273C 
<span class="op">-</span><span class="fl">0.0275517314</span>  <span class="fl">0.1595257918</span>  <span class="fl">0.0610316933</span> <span class="op">-</span><span class="fl">0.0054161479</span>  <span class="fl">0.2018959878</span> 
      YEL076C       YGR205W       YIL012W       YDL056W       YLL047W 
 <span class="fl">0.1807176952</span>  <span class="fl">0.2386590341</span>  <span class="fl">0.0095510031</span>  <span class="fl">0.2211287903</span> <span class="op">-</span><span class="fl">0.0908766519</span> 
      YJL211C       YDL147W       YOL041C       YOL113W       YBL009W 
 <span class="fl">0.1891826624</span>  <span class="fl">0.1871156140</span> <span class="op">-</span><span class="fl">0.2545846899</span>  <span class="fl">0.2596401285</span> <span class="op">-</span><span class="fl">0.0307749383</span> </code></pre></div>
<p>In the next statement we extract the names of the genes that have correlations with YAL022C greater than 0.6. First we test genes to see if they have a correlation with YAL022C greater than 0.6, which returns a vector of TRUE or FALSE values. This vector of Boolean values is than used to index into the rownames of the correlation matrix, pulling out the gene names where the statement was true.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pos.corr.YAL022C &lt;-<span class="st"> </span><span class="kw">rownames</span>(spellman.cor)[spellman.cor[<span class="st">&quot;YAL022C&quot;</span>,] <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.6</span>]
<span class="kw">length</span>(pos.corr.YAL022C)
[<span class="dv">1</span>] <span class="dv">15</span></code></pre></div>
<p>We then return to our wide data to show this set of genes that are strongly positively correlated with YAL022C, using faceted line graphs:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>pos.corr.YAL022C) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression, <span class="dt">color =</span> expt, <span class="dt">group =</span> gene)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_line</span>(<span class="dt">alpha =</span> <span class="fl">0.33</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>expt, <span class="dt">scales =</span> <span class="st">&quot;free_x&quot;</span>, <span class="dt">nrow =</span> <span class="dv">2</span>, <span class="dt">ncol =</span> <span class="dv">2</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&quot;none&quot;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-235-1.png" width="576" /></p>
<p>We can similarly filter for genes that have negative correlations with YAL022C.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">neg.corr.YAL022C &lt;-<span class="st"> </span><span class="kw">colnames</span>(spellman.cor)[spellman.cor[<span class="st">&quot;YAL022C&quot;</span>,] <span class="op">&lt;=</span><span class="st"> </span><span class="op">-</span><span class="fl">0.4</span>]</code></pre></div>
<p>As before we generate a faceted line plot showing these genes by experiment:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>neg.corr.YAL022C) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression, <span class="dt">color =</span> expt, <span class="dt">group =</span> gene)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>(<span class="dt">alpha =</span> <span class="fl">0.33</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>expt, <span class="dt">scales =</span> <span class="st">&quot;free_x&quot;</span>, <span class="dt">nrow =</span> <span class="dv">2</span>, <span class="dt">ncol =</span> <span class="dv">2</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&quot;none&quot;</span>)  </code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-237-1.png" width="576" /></p>
</div>
<div id="adding-new-columns-and-combining-filtered-data-frames" class="section level3">
<h3><span class="header-section-number">7.6.5</span> Adding new columns and combining filtered data frames</h3>
<p>Now let’s create a new data frame by: 1) filtering on our list of genes that have strong positive and negative correlations with YAL022C; and 2) creating a new variable, “corr.with.YAL022C”, which indicates the sign of the correlation. We’ll use this new variable to group genes when we create the plot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pos.corr.df &lt;-<span class="st"> </span>spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>pos.corr.YAL022C) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">corr.with.YAL022C =</span> <span class="st">&quot;positive&quot;</span>)

neg.corr.df &lt;-<span class="st"> </span>spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>neg.corr.YAL022C) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">corr.with.YAL022C =</span> <span class="st">&quot;negative&quot;</span>)

combined.pos.neg &lt;-<span class="st"> </span><span class="kw">rbind</span>(pos.corr.df, neg.corr.df)</code></pre></div>
<p>Finally, we plot the data, colored according to the correlation with YAL022C.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(combined.pos.neg, 
       <span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression,  <span class="dt">group =</span> gene,
           <span class="dt">color =</span> corr.with.YAL022C)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_line</span>(<span class="dt">alpha=</span><span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>expt, <span class="dt">nrow =</span><span class="dv">2</span>, <span class="dt">ncol =</span> <span class="dv">2</span>, <span class="dt">scales =</span> <span class="st">&quot;free_x&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="co"># changes legend title for discrete color legends</span>
<span class="st">  </span><span class="kw">scale_color_discrete</span>(<span class="dt">name =</span> <span class="st">&quot;Correlation with YAL022C&quot;</span>) </code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-239-1.png" width="768" /></p>
</div>
</div>
<div id="heat-maps" class="section level2">
<h2><span class="header-section-number">7.7</span> Heat maps</h2>
<p>Line plots can become visually busy very quickly, and hence aren’t very useful for more than a few tens of genes. An alternative is a “heat map” which depicts data in a grid or table like form, with values indicated by color. Heat maps are good for depicting large amounts of data and providing a coarse “10,000 foot view”. We’ll be looking at lots of heat maps when we get into clustering, but here’s a simple example based on the genes that were positively and negatively correlated with YAL022C.</p>
<div id="rcolorbrewer" class="section level3">
<h3><span class="header-section-number">7.7.1</span> RColorBrewer</h3>
<p>The <code>RColorBrewer</code> packages provides nice color schemes that are useful for creating heat maps. RColorBrewer defines a set of color palettes that have been optimized for color discrimination, to be color blind friendly, etc. Once you’ve installed the RColorBrewer package you can see the available color palettes as so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RColorBrewer)
<span class="co"># show representations of the palettes</span>
<span class="kw">par</span>(<span class="dt">cex =</span> <span class="fl">0.5</span>) <span class="co"># reduce size of text in the follow plot</span>
<span class="kw">display.brewer.all</span>()  </code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-240-1.png" width="576" /></p>
</div>
<div id="creating-a-heat-map-using-geom_tile" class="section level3">
<h3><span class="header-section-number">7.7.2</span> Creating a heat map using <code>geom_tile</code></h3>
<p>We’ll use the <code>geom_tile</code> geom to create the heat map. We’ll depict genes in rows, time points in columns, and expression values by the fill color of each tile. We’ll use the Red-to-Blue (“RdBu”) color scheme defined in RColorBrewer, however we’ll reverse the scheme so blues represent low expression and reds represent high expression.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># generate the color scheme to use</span>
color.scheme &lt;-<span class="st"> </span><span class="kw">rev</span>(<span class="kw">brewer.pal</span>(<span class="dv">8</span>,<span class="st">&quot;RdBu&quot;</span>))

<span class="co"># re-factor gene names so positive and negative genes are spatially distinct in plot</span>
combined.pos.neg<span class="op">$</span>gene &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw">factor</span>(combined.pos.neg<span class="op">$</span>gene, 
         <span class="dt">levels =</span> <span class="kw">c</span>(pos.corr.YAL022C, neg.corr.YAL022C))

combined.pos.neg <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(expt <span class="op">==</span><span class="st"> &quot;cdc28&quot;</span>) <span class="op">%&gt;%</span><span class="st">  </span><span class="co"># focus only on one condition</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> gene)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_tile</span>(<span class="kw">aes</span>(<span class="dt">fill =</span> expression)) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_fill_gradientn</span>(<span class="dt">colors=</span>color.scheme) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">xlab</span>(<span class="st">&quot;Time (mins)&quot;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-241-1.png" width="576" /></p>
<p>Can you visually see the breakpoint between the positively and negatively correlated sets of genes?</p>
</div>
<div id="a-fancy-figure" class="section level3">
<h3><span class="header-section-number">7.7.3</span> A “fancy” figure</h3>
<p>If we want to get really fancy, we can use cowplot’s <code>draw_plot</code> function that allows us to place plots at arbitrary locations and at arbitrary sizes onto the canvas. The coordinates of the canvas run from 0 to 1, and the point (0, 0) is in the lower left corner of the canvas. We’ll use <code>draw_plot</code> to draw a complex figure with a heatmap on the left, and two smaller line plots on the right.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
cdc28.filtered &lt;-<span class="st"> </span><span class="kw">filter</span>(combined.pos.neg, expt <span class="op">==</span><span class="st"> &quot;cdc28&quot;</span>)

pos.corr.lineplot &lt;-<span class="st">  </span>cdc28.filtered <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>pos.corr.YAL022C) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression, <span class="dt">group =</span> gene)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>(<span class="dt">alpha =</span> <span class="fl">0.33</span>, <span class="dt">color =</span> <span class="st">&#39;red&#39;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Expression&quot;</span>, 
      <span class="dt">title =</span> <span class="st">&quot;Genes Positively correlated</span><span class="ch">\n</span><span class="st">with YAL022C&quot;</span>)

neg.corr.lineplot &lt;-<span class="st">  </span>cdc28.filtered <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>neg.corr.YAL022C) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression, <span class="dt">group =</span> gene)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>(<span class="dt">alpha =</span> <span class="fl">0.33</span>, <span class="dt">color =</span> <span class="st">&#39;blue&#39;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Expression&quot;</span>, 
      <span class="dt">title =</span> <span class="st">&quot;Genes negatively correlated</span><span class="ch">\n</span><span class="st">with YAL022C&quot;</span>)


heat.map &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw">ggplot</span>(cdc28.filtered, <span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> gene)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_tile</span>(<span class="kw">aes</span>(<span class="dt">fill =</span> expression)) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_fill_gradientn</span>(<span class="dt">colors=</span>color.scheme) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Gene&quot;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&quot;none&quot;</span>)

fancy.plot &lt;-<span class="st"> </span><span class="kw">ggdraw</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">draw_plot</span>(heat.map, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dt">width =</span> <span class="fl">0.6</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">draw_plot</span>(neg.corr.lineplot, <span class="fl">0.6</span>, <span class="fl">0.5</span>, <span class="dt">width =</span> <span class="fl">0.4</span>, <span class="dt">height =</span> <span class="fl">0.5</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">draw_plot</span>(pos.corr.lineplot, <span class="fl">0.6</span>, <span class="dv">0</span>,   <span class="dt">width =</span> <span class="fl">0.4</span>, <span class="dt">height =</span> <span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">draw_plot_label</span>(<span class="kw">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;C&quot;</span>), <span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.6</span>, <span class="fl">0.6</span>), <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="fl">0.5</span>), <span class="dt">size =</span> <span class="dv">15</span>)

fancy.plot</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-242-1.png" width="768" /></p>
</div>
<div id="sorting-and-visualizing-genes-by-time-of-maximum-expression" class="section level3">
<h3><span class="header-section-number">7.7.4</span> Sorting and visualizing genes by time of maximum expression</h3>
<p>In a time series data one might be interested in examing genes with respect to some property of their dynamic expression. Here we’ll focus on the 1000 most variable genes, sort the data from the alpha factor synchronization experiment by time of maximum expression.</p>
<p>First let’s filter our tidy, long data frame to include only our top 1000 genes and the alpha factor measurements:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">top1k.alpha.long &lt;-
<span class="st">  </span>spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>top.genes.1k, expt <span class="op">==</span><span class="st"> &quot;alpha&quot;</span>)</code></pre></div>
<p>To find the time of maximum expression we’ll employ the function <code>which.max</code> (<code>which.min</code>), which finds the index of the maximum (minimum) element of a vector. For example to find the index of the maximum expression measurement for YAL022C we could do:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">top1k.alpha.long <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">==</span><span class="st"> &quot;YAL022C&quot;</span>) <span class="op">%$%</span><span class="st">  </span><span class="co"># note the exposition pipe operator!</span>
<span class="st">  </span><span class="kw">which.max</span>(expression)
[<span class="dv">1</span>] <span class="dv">18</span></code></pre></div>
<p>From the code above we find that the index of the observation at which YAL022C is maximal is 18. To get the corresponding time point we can do something like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">top1k.alpha.long <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">==</span><span class="st"> &quot;YAL022C&quot;</span>) <span class="op">%$%</span><span class="st">  </span><span class="co"># again note the exposition pipe operator!</span>
<span class="st">  </span>time[<span class="kw">which.max</span>(expression)]
[<span class="dv">1</span>] <span class="dv">119</span></code></pre></div>
<p>Thus YAL022C expression peaks at 119 minutes in the alpha factor experiment.</p>
<p>To find the index of maximal expression of all genes we can apply the <code>dplyr::group_by()</code> and <code>dplyr::summarize()</code> functions</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">peak.expression &lt;-<span class="st"> </span>
<span class="st">  </span>top1k.alpha.long <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">group_by</span>(gene) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">summarise</span>(<span class="dt">peak =</span> <span class="kw">which.max</span>(expression))

<span class="kw">dim</span>(peak.expression)
[<span class="dv">1</span>] <span class="dv">1000</span>    <span class="dv">2</span>
<span class="kw">head</span>(peak.expression, <span class="dt">n =</span> <span class="dv">10</span>)
<span class="co"># A tibble: 10 x 2</span>
   gene     peak
   <span class="op">&lt;</span>chr<span class="op">&gt;</span><span class="st">   </span><span class="er">&lt;</span>int<span class="op">&gt;</span>
<span class="st"> </span><span class="dv">1</span> YAL003W    <span class="dv">10</span>
 <span class="dv">2</span> YAL004W     <span class="dv">8</span>
 <span class="dv">3</span> YAL005C    <span class="dv">16</span>
 <span class="dv">4</span> YAL018C     <span class="dv">1</span>
 <span class="dv">5</span> YAL022C    <span class="dv">18</span>
 <span class="dv">6</span> YAL025C     <span class="dv">8</span>
 <span class="dv">7</span> YAL028W    <span class="dv">16</span>
 <span class="dv">8</span> YAL035W    <span class="dv">16</span>
 <span class="dv">9</span> YAL038W    <span class="dv">18</span>
<span class="dv">10</span> YAL044C    <span class="dv">17</span></code></pre></div>
<p>We can then generate a heatmap where we sort the rows (genes) of the heatmap by their time of peak expression. We introduce a new geom – <code>geom_raster</code> – which is like <code>geom_tile</code> but better suited for large data.</p>
<p>The explicit sorting of the data by peak expression is carried out in the call to <code>scale_y_discrete</code> where the limits (and order) of this axis are set with the <code>limits</code> argument (see <code>scale_y_discrete</code> and <code>discrete_scale</code> in the ggplot2 docs).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">top1k.alpha.long <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> gene)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_raster</span>(<span class="kw">aes</span>(<span class="dt">fill =</span> expression)) <span class="op">+</span><span class="st">  </span><span class="co"># </span>
<span class="st">  </span><span class="kw">scale_fill_gradientn</span>(<span class="dt">limits=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">2</span>),<span class="dt">colors=</span>color.scheme) <span class="op">+</span>
<span class="st">  </span><span class="kw">scale_y_discrete</span>(<span class="dt">limits=</span>peak.expression<span class="op">$</span>gene[<span class="kw">rev</span>(<span class="kw">order</span>(peak.expression<span class="op">$</span>peak))]) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Genes&quot;</span>, 
       <span class="dt">title =</span> <span class="st">&quot;1000 most variable genes&quot;</span>,
       <span class="dt">subtitle =</span> <span class="st">&quot;Sorted by time of peak expression&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="co"># the following line suppresses tick and labels on y-axi</span>
<span class="st">  </span><span class="kw">theme</span>(<span class="dt">axis.text.y =</span> <span class="kw">element_blank</span>(), <span class="dt">axis.ticks.y =</span> <span class="kw">element_blank</span>())</code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-247"></span>
<img src="bio723-book_files/figure-html/unnamed-chunk-247-1.png" alt="A Heatmap showing genes in the alpha-factor experiment, sorted by peak expression" width="576" />
<p class="caption">
Figure 7.1: A Heatmap showing genes in the alpha-factor experiment, sorted by peak expression
</p>
</div>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction-to-the-dplyr-package.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="vectors-and-functions.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
