<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 7 Data wrangling | Biology 723: Statistical Computing for Biologists</title>
  <meta name="description" content="Hands-on materials for Bio 723 course at Duke University.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 7 Data wrangling | Biology 723: Statistical Computing for Biologists" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 Data wrangling | Biology 723: Statistical Computing for Biologists" />
  
  <meta name="twitter:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

<meta name="author" content="Paul M. Magwene">


<meta name="date" content="2019-04-17">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="introduction-to-dplyr.html">
<link rel="next" href="functions-and-control-flow-statements.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script src="libs/htmlwidgets-1.3/htmlwidgets.js"></script>
<link href="libs/rglwidgetClass-2/rgl.css" rel="stylesheet" />
<script src="libs/rglwidgetClass-2/rglClass.src.js"></script>
<script src="libs/CanvasMatrix4-2016/CanvasMatrix.src.js"></script>
<script src="libs/rglWebGL-binding-0.99.16/rglWebGL.js"></script>
<link href="libs/crosstalk-1.0.0/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.0.0/js/crosstalk.min.js"></script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { background-color: #f8f8f8; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#accessing-older-versions-of-the-course-notes"><i class="fa fa-check"></i><b>1.1</b> Accessing older versions of the course notes</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#how-to-use-these-lecture-notes"><i class="fa fa-check"></i><b>1.2</b> How to use these lecture notes</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html"><i class="fa fa-check"></i><b>2</b> Getting Started with R</a><ul>
<li class="chapter" data-level="2.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-r"><i class="fa fa-check"></i><b>2.1</b> What is R?</a></li>
<li class="chapter" data-level="2.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-rstudio"><i class="fa fa-check"></i><b>2.2</b> What is RStudio?</a></li>
<li class="chapter" data-level="2.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#entering-commands-in-the-console"><i class="fa fa-check"></i><b>2.3</b> Entering commands in the console</a></li>
<li class="chapter" data-level="2.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#comments"><i class="fa fa-check"></i><b>2.4</b> Comments</a></li>
<li class="chapter" data-level="2.5" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#using-r-as-a-calculator"><i class="fa fa-check"></i><b>2.5</b> Using R as a Calculator</a><ul>
<li class="chapter" data-level="2.5.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#common-mathematical-functions"><i class="fa fa-check"></i><b>2.5.1</b> Common mathematical functions</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#variable-assignment"><i class="fa fa-check"></i><b>2.6</b> Variable assignment</a><ul>
<li class="chapter" data-level="2.6.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#valid-variable-names"><i class="fa fa-check"></i><b>2.6.1</b> Valid variable names</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#data-types"><i class="fa fa-check"></i><b>2.7</b> Data types</a><ul>
<li class="chapter" data-level="2.7.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#numeric-data-types"><i class="fa fa-check"></i><b>2.7.1</b> Numeric data types</a></li>
<li class="chapter" data-level="2.7.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#logical-values"><i class="fa fa-check"></i><b>2.7.2</b> Logical values</a></li>
<li class="chapter" data-level="2.7.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#character-strings"><i class="fa fa-check"></i><b>2.7.3</b> Character strings</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#packages"><i class="fa fa-check"></i><b>2.8</b> Packages</a><ul>
<li class="chapter" data-level="2.8.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-console"><i class="fa fa-check"></i><b>2.8.1</b> Installing packages from the console</a></li>
<li class="chapter" data-level="2.8.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#install-the-tidyverse-package"><i class="fa fa-check"></i><b>2.8.2</b> Install the tidyverse package</a></li>
<li class="chapter" data-level="2.8.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-rstudio-dialog"><i class="fa fa-check"></i><b>2.8.3</b> Installing packages from the RStudio dialog</a></li>
<li class="chapter" data-level="2.8.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#loading-packages-with-the-library-function"><i class="fa fa-check"></i><b>2.8.4</b> Loading packages with the <code>library()</code> function</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#the-r-help-system"><i class="fa fa-check"></i><b>2.9</b> The R Help System</a><ul>
<li class="chapter" data-level="2.9.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#getting-help-from-the-console"><i class="fa fa-check"></i><b>2.9.1</b> Getting help from the console</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html"><i class="fa fa-check"></i><b>3</b> R Markdown and R Notebooks</a><ul>
<li class="chapter" data-level="3.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#r-notebooks"><i class="fa fa-check"></i><b>3.1</b> R Notebooks</a></li>
<li class="chapter" data-level="3.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#creating-an-r-notebook"><i class="fa fa-check"></i><b>3.2</b> Creating an R Notebook</a></li>
<li class="chapter" data-level="3.3" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#the-default-r-notebook-template"><i class="fa fa-check"></i><b>3.3</b> The default R Notebook template</a></li>
<li class="chapter" data-level="3.4" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-and-non-code-blocks"><i class="fa fa-check"></i><b>3.4</b> Code and Non-code blocks</a><ul>
<li class="chapter" data-level="3.4.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#non-code-blocks"><i class="fa fa-check"></i><b>3.4.1</b> Non-code blocks</a></li>
<li class="chapter" data-level="3.4.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-blocks"><i class="fa fa-check"></i><b>3.4.2</b> Code blocks</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-a-code-chunk"><i class="fa fa-check"></i><b>3.5</b> Running a code chunk</a></li>
<li class="chapter" data-level="3.6" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-all-code-chunks-above"><i class="fa fa-check"></i><b>3.6</b> Running all code chunks above</a></li>
<li class="chapter" data-level="3.7" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#knitting-an-r-markdown-to-html"><i class="fa fa-check"></i><b>3.7</b> “Knitting an” R Markdown to HTML</a></li>
<li class="chapter" data-level="3.8" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#sharing-your-reproducible-r-notebook"><i class="fa fa-check"></i><b>3.8</b> Sharing your reproducible R Notebook</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-structures.html"><a href="data-structures.html"><i class="fa fa-check"></i><b>4</b> Data structures</a><ul>
<li class="chapter" data-level="4.1" data-path="data-structures.html"><a href="data-structures.html#vectors"><i class="fa fa-check"></i><b>4.1</b> Vectors</a><ul>
<li class="chapter" data-level="4.1.1" data-path="data-structures.html"><a href="data-structures.html#vector-arithmetic"><i class="fa fa-check"></i><b>4.1.1</b> Vector Arithmetic</a></li>
<li class="chapter" data-level="4.1.2" data-path="data-structures.html"><a href="data-structures.html#vector-recycling"><i class="fa fa-check"></i><b>4.1.2</b> Vector recycling</a></li>
<li class="chapter" data-level="4.1.3" data-path="data-structures.html"><a href="data-structures.html#simple-statistical-functions-for-numeric-vectors"><i class="fa fa-check"></i><b>4.1.3</b> Simple statistical functions for numeric vectors</a></li>
<li class="chapter" data-level="4.1.4" data-path="data-structures.html"><a href="data-structures.html#indexing-vectors"><i class="fa fa-check"></i><b>4.1.4</b> Indexing Vectors</a></li>
<li class="chapter" data-level="4.1.5" data-path="data-structures.html"><a href="data-structures.html#comparison-operators-applied-to-vectors"><i class="fa fa-check"></i><b>4.1.5</b> Comparison operators applied to vectors</a></li>
<li class="chapter" data-level="4.1.6" data-path="data-structures.html"><a href="data-structures.html#combining-indexing-and-comparison-of-vectors"><i class="fa fa-check"></i><b>4.1.6</b> Combining Indexing and Comparison of Vectors</a></li>
<li class="chapter" data-level="4.1.7" data-path="data-structures.html"><a href="data-structures.html#vector-manipulation"><i class="fa fa-check"></i><b>4.1.7</b> Vector manipulation</a></li>
<li class="chapter" data-level="4.1.8" data-path="data-structures.html"><a href="data-structures.html#vectors-from-regular-sequences"><i class="fa fa-check"></i><b>4.1.8</b> Vectors from regular sequences</a></li>
<li class="chapter" data-level="4.1.9" data-path="data-structures.html"><a href="data-structures.html#additional-functions-for-working-with-vectors"><i class="fa fa-check"></i><b>4.1.9</b> Additional functions for working with vectors</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="data-structures.html"><a href="data-structures.html#lists"><i class="fa fa-check"></i><b>4.2</b> Lists</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-structures.html"><a href="data-structures.html#length-and-type-of-lists"><i class="fa fa-check"></i><b>4.2.1</b> Length and type of lists</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-structures.html"><a href="data-structures.html#indexing-lists"><i class="fa fa-check"></i><b>4.2.2</b> Indexing lists</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-structures.html"><a href="data-structures.html#naming-list-elements"><i class="fa fa-check"></i><b>4.2.3</b> Naming list elements</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-structures.html"><a href="data-structures.html#the-operator"><i class="fa fa-check"></i><b>4.2.4</b> The <code>$</code> operator</a></li>
<li class="chapter" data-level="4.2.5" data-path="data-structures.html"><a href="data-structures.html#changing-and-adding-lists-items"><i class="fa fa-check"></i><b>4.2.5</b> Changing and adding lists items</a></li>
<li class="chapter" data-level="4.2.6" data-path="data-structures.html"><a href="data-structures.html#combining-lists"><i class="fa fa-check"></i><b>4.2.6</b> Combining lists</a></li>
<li class="chapter" data-level="4.2.7" data-path="data-structures.html"><a href="data-structures.html#converting-lists-to-vectors"><i class="fa fa-check"></i><b>4.2.7</b> Converting lists to vectors</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-structures.html"><a href="data-structures.html#data-frames"><i class="fa fa-check"></i><b>4.3</b> Data frames</a><ul>
<li class="chapter" data-level="4.3.1" data-path="data-structures.html"><a href="data-structures.html#creating-a-data-frame"><i class="fa fa-check"></i><b>4.3.1</b> Creating a data frame</a></li>
<li class="chapter" data-level="4.3.2" data-path="data-structures.html"><a href="data-structures.html#type-and-class-for-data-frames"><i class="fa fa-check"></i><b>4.3.2</b> Type and class for data frames</a></li>
<li class="chapter" data-level="4.3.3" data-path="data-structures.html"><a href="data-structures.html#length-and-dimension-for-data-frames"><i class="fa fa-check"></i><b>4.3.3</b> Length and dimension for data frames</a></li>
<li class="chapter" data-level="4.3.4" data-path="data-structures.html"><a href="data-structures.html#indexing-and-accessing-data-frames"><i class="fa fa-check"></i><b>4.3.4</b> Indexing and accessing data frames</a></li>
<li class="chapter" data-level="4.3.5" data-path="data-structures.html"><a href="data-structures.html#logical-indexing-of-data-frames"><i class="fa fa-check"></i><b>4.3.5</b> Logical indexing of data frames</a></li>
<li class="chapter" data-level="4.3.6" data-path="data-structures.html"><a href="data-structures.html#adding-columns-to-a-data-frame"><i class="fa fa-check"></i><b>4.3.6</b> Adding columns to a data frame</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html"><i class="fa fa-check"></i><b>5</b> Introduction to ggplot2</a><ul>
<li class="chapter" data-level="5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#loading-ggplot2"><i class="fa fa-check"></i><b>5.1</b> Loading ggplot2</a></li>
<li class="chapter" data-level="5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#example-data-set-andersons-iris-data"><i class="fa fa-check"></i><b>5.2</b> Example data set: Anderson’s Iris Data</a></li>
<li class="chapter" data-level="5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#template-for-single-layer-plots-in-ggplot2"><i class="fa fa-check"></i><b>5.3</b> Template for single layer plots in ggplot2</a></li>
<li class="chapter" data-level="5.4" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#an-aside-about-function-arguments"><i class="fa fa-check"></i><b>5.4</b> An aside about function arguments</a></li>
<li class="chapter" data-level="5.5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#strip-plots"><i class="fa fa-check"></i><b>5.5</b> Strip plots</a><ul>
<li class="chapter" data-level="5.5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#jittering-data"><i class="fa fa-check"></i><b>5.5.1</b> Jittering data</a></li>
<li class="chapter" data-level="5.5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-categorical-information"><i class="fa fa-check"></i><b>5.5.2</b> Adding categorical information</a></li>
<li class="chapter" data-level="5.5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#rotating-plot-coordinates"><i class="fa fa-check"></i><b>5.5.3</b> Rotating plot coordinates</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#histograms"><i class="fa fa-check"></i><b>5.6</b> Histograms</a><ul>
<li class="chapter" data-level="5.6.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#variations-on-histograms-when-considering-categorical-data"><i class="fa fa-check"></i><b>5.6.1</b> Variations on histograms when considering categorical data</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#faceting-to-depict-categorical-information"><i class="fa fa-check"></i><b>5.7</b> Faceting to depict categorical information</a></li>
<li class="chapter" data-level="5.8" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots"><i class="fa fa-check"></i><b>5.8</b> Density plots</a></li>
<li class="chapter" data-level="5.9" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-or-beanplot"><i class="fa fa-check"></i><b>5.9</b> Violin or Beanplot</a></li>
<li class="chapter" data-level="5.10" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplots"><i class="fa fa-check"></i><b>5.10</b> Boxplots</a></li>
<li class="chapter" data-level="5.11" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#building-complex-visualizations-with-layers"><i class="fa fa-check"></i><b>5.11</b> Building complex visualizations with layers</a></li>
<li class="chapter" data-level="5.12" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#useful-combination-plots"><i class="fa fa-check"></i><b>5.12</b> Useful combination plots</a><ul>
<li class="chapter" data-level="5.12.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplot-plus-strip-plot"><i class="fa fa-check"></i><b>5.12.1</b> Boxplot plus strip plot</a></li>
<li class="chapter" data-level="5.12.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#setting-shared-aesthetics"><i class="fa fa-check"></i><b>5.12.2</b> Setting shared aesthetics</a></li>
</ul></li>
<li class="chapter" data-level="5.13" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot-layers-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>5.13</b> ggplot layers can be assigned to variables</a><ul>
<li class="chapter" data-level="5.13.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-plot-plus-strip-plot"><i class="fa fa-check"></i><b>5.13.1</b> Violin plot plus strip plot</a></li>
</ul></li>
<li class="chapter" data-level="5.14" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-titles-and-tweaking-axis-labels"><i class="fa fa-check"></i><b>5.14</b> Adding titles and tweaking axis labels</a></li>
<li class="chapter" data-level="5.15" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot2-themes"><i class="fa fa-check"></i><b>5.15</b> ggplot2 themes</a><ul>
<li class="chapter" data-level="5.15.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#further-customization-with-ggplot2theme"><i class="fa fa-check"></i><b>5.15.1</b> Further customization with <code>ggplot2::theme</code></a></li>
</ul></li>
<li class="chapter" data-level="5.16" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#other-aspects-of-ggplots-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>5.16</b> Other aspects of ggplots can be assigned to variables</a></li>
<li class="chapter" data-level="5.17" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-plots"><i class="fa fa-check"></i><b>5.17</b> Bivariate plots</a><ul>
<li class="chapter" data-level="5.17.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#scatter-plots"><i class="fa fa-check"></i><b>5.17.1</b> Scatter plots</a></li>
<li class="chapter" data-level="5.17.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-a-trend-line-to-a-scatter-plot"><i class="fa fa-check"></i><b>5.17.2</b> Adding a trend line to a scatter plot</a></li>
</ul></li>
<li class="chapter" data-level="5.18" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-density-plots"><i class="fa fa-check"></i><b>5.18</b> Bivariate density plots</a></li>
<li class="chapter" data-level="5.19" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#combining-scatter-plots-and-density-plots-with-categorical-information"><i class="fa fa-check"></i><b>5.19</b> Combining Scatter Plots and Density Plots with Categorical Information</a></li>
<li class="chapter" data-level="5.20" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots-with-fill"><i class="fa fa-check"></i><b>5.20</b> Density plots with fill</a></li>
<li class="chapter" data-level="5.21" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#d-bin-and-hex-plots"><i class="fa fa-check"></i><b>5.21</b> 2D bin and hex plots</a></li>
<li class="chapter" data-level="5.22" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#the-cowplot-package"><i class="fa fa-check"></i><b>5.22</b> The <code>cowplot</code> package</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html"><i class="fa fa-check"></i><b>6</b> Introduction to <code>dplyr</code></a><ul>
<li class="chapter" data-level="6.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#libraries"><i class="fa fa-check"></i><b>6.1</b> Libraries</a></li>
<li class="chapter" data-level="6.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#reading-data-with-the-readr-package"><i class="fa fa-check"></i><b>6.2</b> Reading data with the <code>readr</code> package</a><ul>
<li class="chapter" data-level="6.2.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#reading-excel-files"><i class="fa fa-check"></i><b>6.2.1</b> Reading Excel files</a></li>
<li class="chapter" data-level="6.2.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#example-data-nc-births"><i class="fa fa-check"></i><b>6.2.2</b> Example data: NC Births</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#a-note-on-tibbles"><i class="fa fa-check"></i><b>6.3</b> A note on “tibbles”</a></li>
<li class="chapter" data-level="6.4" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#data-filtering-and-transformation-with-dplyr"><i class="fa fa-check"></i><b>6.4</b> Data filtering and transformation with <code>dplyr</code></a></li>
<li class="chapter" data-level="6.5" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#dplyrs-verbs"><i class="fa fa-check"></i><b>6.5</b> dplyr’s “verbs”</a><ul>
<li class="chapter" data-level="6.5.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#select"><i class="fa fa-check"></i><b>6.5.1</b> <code>select</code></a></li>
<li class="chapter" data-level="6.5.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#filter"><i class="fa fa-check"></i><b>6.5.2</b> <code>filter</code></a></li>
<li class="chapter" data-level="6.5.3" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#mutate"><i class="fa fa-check"></i><b>6.5.3</b> <code>mutate</code></a></li>
<li class="chapter" data-level="6.5.4" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#arrange"><i class="fa fa-check"></i><b>6.5.4</b> <code>arrange</code></a></li>
<li class="chapter" data-level="6.5.5" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#summarize"><i class="fa fa-check"></i><b>6.5.5</b> <code>summarize</code></a></li>
<li class="chapter" data-level="6.5.6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#group_by"><i class="fa fa-check"></i><b>6.5.6</b> <code>group_by</code></a></li>
<li class="chapter" data-level="6.5.7" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#combining-grouping-and-summarizing"><i class="fa fa-check"></i><b>6.5.7</b> Combining grouping and summarizing</a></li>
<li class="chapter" data-level="6.5.8" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#scoped-variants-of-mutate-and-summarize"><i class="fa fa-check"></i><b>6.5.8</b> Scoped variants of <code>mutate</code> and <code>summarize</code></a></li>
<li class="chapter" data-level="6.5.9" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#combining-summarize-with-grouping-aesthetics-in-ggplot2"><i class="fa fa-check"></i><b>6.5.9</b> Combining summarize with grouping aesthetics in <code>ggplot2</code></a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#pipes"><i class="fa fa-check"></i><b>6.6</b> Pipes</a><ul>
<li class="chapter" data-level="6.6.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#install-and-load-magrittr"><i class="fa fa-check"></i><b>6.6.1</b> Install and load <code>magrittr</code></a></li>
<li class="chapter" data-level="6.6.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-basic-pipe-operator"><i class="fa fa-check"></i><b>6.6.2</b> The basic pipe operator</a></li>
<li class="chapter" data-level="6.6.3" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#an-example-without-pipes"><i class="fa fa-check"></i><b>6.6.3</b> An example without pipes</a></li>
<li class="chapter" data-level="6.6.4" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-same-example-using-pipes"><i class="fa fa-check"></i><b>6.6.4</b> The same example using pipes</a></li>
<li class="chapter" data-level="6.6.5" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#assigning-the-output-of-a-statement-involving-pipes-to-a-variable"><i class="fa fa-check"></i><b>6.6.5</b> Assigning the output of a statement involving pipes to a variable</a></li>
<li class="chapter" data-level="6.6.6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#compound-assignment-pipe-operator"><i class="fa fa-check"></i><b>6.6.6</b> Compound assignment pipe operator</a></li>
<li class="chapter" data-level="6.6.7" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-dot-operator-with-pipes"><i class="fa fa-check"></i><b>6.6.7</b> The dot operator with pipes</a></li>
<li class="chapter" data-level="6.6.8" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-exposition-pipe-operator"><i class="fa fa-check"></i><b>6.6.8</b> The exposition pipe operator</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>7</b> Data wrangling</a><ul>
<li class="chapter" data-level="7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#libraries-1"><i class="fa fa-check"></i><b>7.1</b> Libraries</a></li>
<li class="chapter" data-level="7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#data"><i class="fa fa-check"></i><b>7.2</b> Data</a></li>
<li class="chapter" data-level="7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#renaming-data-frame-columms"><i class="fa fa-check"></i><b>7.3</b> Renaming data frame columms</a></li>
<li class="chapter" data-level="7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-unneeded-columns"><i class="fa fa-check"></i><b>7.4</b> Dropping unneeded columns</a><ul>
<li class="chapter" data-level="7.4.1" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-all-empty-columns"><i class="fa fa-check"></i><b>7.4.1</b> Finding all empty columns</a></li>
<li class="chapter" data-level="7.4.2" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-columns-by-matching-names"><i class="fa fa-check"></i><b>7.4.2</b> Dropping columns by matching names</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="data-wrangling.html"><a href="data-wrangling.html#merging-data-frames"><i class="fa fa-check"></i><b>7.5</b> Merging data frames</a></li>
<li class="chapter" data-level="7.6" data-path="data-wrangling.html"><a href="data-wrangling.html#reshaping-data-with-tidyr"><i class="fa fa-check"></i><b>7.6</b> Reshaping data with tidyr</a><ul>
<li class="chapter" data-level="7.6.1" data-path="data-wrangling.html"><a href="data-wrangling.html#wide-to-long-conversions-using-tidyrgather"><i class="fa fa-check"></i><b>7.6.1</b> Wide to long conversions using <code>tidyr::gather</code></a></li>
<li class="chapter" data-level="7.6.2" data-path="data-wrangling.html"><a href="data-wrangling.html#extracting-information-from-combined-variables-using-tidyrextract"><i class="fa fa-check"></i><b>7.6.2</b> Extracting information from combined variables using <code>tidyr::extract</code></a></li>
<li class="chapter" data-level="7.6.3" data-path="data-wrangling.html"><a href="data-wrangling.html#subsetting-rows"><i class="fa fa-check"></i><b>7.6.3</b> Subsetting rows</a></li>
<li class="chapter" data-level="7.6.4" data-path="data-wrangling.html"><a href="data-wrangling.html#splitting-columns"><i class="fa fa-check"></i><b>7.6.4</b> Splitting columns</a></li>
<li class="chapter" data-level="7.6.5" data-path="data-wrangling.html"><a href="data-wrangling.html#combining-data-frames"><i class="fa fa-check"></i><b>7.6.5</b> Combining data frames</a></li>
<li class="chapter" data-level="7.6.6" data-path="data-wrangling.html"><a href="data-wrangling.html#sorting-data-frame-rows"><i class="fa fa-check"></i><b>7.6.6</b> Sorting data frame rows</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="data-wrangling.html"><a href="data-wrangling.html#using-your-tidy-data"><i class="fa fa-check"></i><b>7.7</b> Using your tidy data</a><ul>
<li class="chapter" data-level="7.7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#visualizing-gene-expression-time-series"><i class="fa fa-check"></i><b>7.7.1</b> Visualizing gene expression time series</a></li>
<li class="chapter" data-level="7.7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-the-most-variable-genes"><i class="fa fa-check"></i><b>7.7.2</b> Finding the most variable genes</a></li>
<li class="chapter" data-level="7.7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#heat-maps"><i class="fa fa-check"></i><b>7.7.3</b> Heat maps</a></li>
</ul></li>
<li class="chapter" data-level="7.8" data-path="data-wrangling.html"><a href="data-wrangling.html#long-to-wide-conversion-using-tidyrspread"><i class="fa fa-check"></i><b>7.8</b> Long-to-wide conversion using <code>tidyr::spread</code></a></li>
<li class="chapter" data-level="7.9" data-path="data-wrangling.html"><a href="data-wrangling.html#exploring-bivariate-relationships-using-wide-data"><i class="fa fa-check"></i><b>7.9</b> Exploring bivariate relationships using “wide” data</a><ul>
<li class="chapter" data-level="7.9.1" data-path="data-wrangling.html"><a href="data-wrangling.html#large-scale-patterns-of-correlations"><i class="fa fa-check"></i><b>7.9.1</b> Large scale patterns of correlations</a></li>
<li class="chapter" data-level="7.9.2" data-path="data-wrangling.html"><a href="data-wrangling.html#adding-new-columns-and-combining-filtered-data-frames"><i class="fa fa-check"></i><b>7.9.2</b> Adding new columns and combining filtered data frames</a></li>
<li class="chapter" data-level="7.9.3" data-path="data-wrangling.html"><a href="data-wrangling.html#a-heat-mapped-sorted-by-correlations"><i class="fa fa-check"></i><b>7.9.3</b> A heat mapped sorted by correlations</a></li>
<li class="chapter" data-level="7.9.4" data-path="data-wrangling.html"><a href="data-wrangling.html#a-fancy-figure"><i class="fa fa-check"></i><b>7.9.4</b> A “fancy” figure</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html"><i class="fa fa-check"></i><b>8</b> Functions and control flow statements</a><ul>
<li class="chapter" data-level="8.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#writing-your-own-functions"><i class="fa fa-check"></i><b>8.1</b> Writing your own functions</a><ul>
<li class="chapter" data-level="8.1.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#function-arguments"><i class="fa fa-check"></i><b>8.1.1</b> Function arguments</a></li>
<li class="chapter" data-level="8.1.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#writing-functions-with-optional-arguments"><i class="fa fa-check"></i><b>8.1.2</b> Writing functions with optional arguments</a></li>
<li class="chapter" data-level="8.1.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#putting-r-functions-in-scripts"><i class="fa fa-check"></i><b>8.1.3</b> Putting R functions in Scripts</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#control-flow-statements"><i class="fa fa-check"></i><b>8.2</b> Control flow statements</a><ul>
<li class="chapter" data-level="8.2.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#if-and-if-else-statements"><i class="fa fa-check"></i><b>8.2.1</b> <code>if</code> and <code>if-else</code> statements</a></li>
<li class="chapter" data-level="8.2.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#for-loops"><i class="fa fa-check"></i><b>8.2.2</b> for loops</a></li>
<li class="chapter" data-level="8.2.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#break-statement"><i class="fa fa-check"></i><b>8.2.3</b> <code>break</code> statement</a></li>
<li class="chapter" data-level="8.2.4" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#repeat-loops"><i class="fa fa-check"></i><b>8.2.4</b> <code>repeat</code> loops</a></li>
<li class="chapter" data-level="8.2.5" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#next-statement"><i class="fa fa-check"></i><b>8.2.5</b> <code>next</code> statement</a></li>
<li class="chapter" data-level="8.2.6" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#while-statements"><i class="fa fa-check"></i><b>8.2.6</b> while statements</a></li>
<li class="chapter" data-level="8.2.7" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#ifelse"><i class="fa fa-check"></i><b>8.2.7</b> <code>ifelse</code></a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#map-and-related-tools"><i class="fa fa-check"></i><b>8.3</b> <code>map</code> and related tools</a><ul>
<li class="chapter" data-level="8.3.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#basic-map"><i class="fa fa-check"></i><b>8.3.1</b> basic <code>map</code></a></li>
<li class="chapter" data-level="8.3.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#map_if-and-map_at"><i class="fa fa-check"></i><b>8.3.2</b> <code>map_if</code> and <code>map_at</code></a></li>
<li class="chapter" data-level="8.3.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#mapping-in-parallel-using-map2"><i class="fa fa-check"></i><b>8.3.3</b> mapping in parallel using <code>map2</code></a></li>
<li class="chapter" data-level="8.3.4" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#map-variants-that-return-vectors"><i class="fa fa-check"></i><b>8.3.4</b> <code>map</code> variants that return vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="vector-algebra.html"><a href="vector-algebra.html"><i class="fa fa-check"></i><b>9</b> Vector algebra</a><ul>
<li class="chapter" data-level="9.1" data-path="vector-algebra.html"><a href="vector-algebra.html#libraries-2"><i class="fa fa-check"></i><b>9.1</b> Libraries</a></li>
<li class="chapter" data-level="9.2" data-path="vector-algebra.html"><a href="vector-algebra.html#vector-mathematics-in-r"><i class="fa fa-check"></i><b>9.2</b> Vector Mathematics in R</a></li>
<li class="chapter" data-level="9.3" data-path="vector-algebra.html"><a href="vector-algebra.html#simple-statistics-in-vector-form"><i class="fa fa-check"></i><b>9.3</b> Simple statistics in vector form</a><ul>
<li class="chapter" data-level="9.3.1" data-path="vector-algebra.html"><a href="vector-algebra.html#mean"><i class="fa fa-check"></i><b>9.3.1</b> Mean</a></li>
<li class="chapter" data-level="9.3.2" data-path="vector-algebra.html"><a href="vector-algebra.html#mean-centering"><i class="fa fa-check"></i><b>9.3.2</b> Mean centering</a></li>
<li class="chapter" data-level="9.3.3" data-path="vector-algebra.html"><a href="vector-algebra.html#variance-and-standard-deviation"><i class="fa fa-check"></i><b>9.3.3</b> Variance and standard deviation</a></li>
<li class="chapter" data-level="9.3.4" data-path="vector-algebra.html"><a href="vector-algebra.html#covariance-and-correlation"><i class="fa fa-check"></i><b>9.3.4</b> Covariance and correlation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="matrices-in-r.html"><a href="matrices-in-r.html"><i class="fa fa-check"></i><b>10</b> Matrices in R</a><ul>
<li class="chapter" data-level="10.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#creating-matrices-in-r"><i class="fa fa-check"></i><b>10.1</b> Creating matrices in R</a></li>
<li class="chapter" data-level="10.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-arithmetic-operations-in-r"><i class="fa fa-check"></i><b>10.2</b> Matrix arithmetic operations in R</a></li>
<li class="chapter" data-level="10.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#descriptive-statistics-as-matrix-functions"><i class="fa fa-check"></i><b>10.3</b> Descriptive statistics as matrix functions</a><ul>
<li class="chapter" data-level="10.3.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#mean-vector-and-matrix"><i class="fa fa-check"></i><b>10.3.1</b> Mean vector and matrix</a></li>
<li class="chapter" data-level="10.3.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#deviation-matrix"><i class="fa fa-check"></i><b>10.3.2</b> Deviation matrix</a></li>
<li class="chapter" data-level="10.3.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#covariance-matrix"><i class="fa fa-check"></i><b>10.3.3</b> Covariance matrix</a></li>
<li class="chapter" data-level="10.3.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#correlation-matrix"><i class="fa fa-check"></i><b>10.3.4</b> Correlation matrix</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-inverse"><i class="fa fa-check"></i><b>10.4</b> Matrix Inverse</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html"><i class="fa fa-check"></i><b>11</b> Fitting Regression Models in R</a><ul>
<li class="chapter" data-level="11.1" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#regression-terminology"><i class="fa fa-check"></i><b>11.1</b> Regression terminology</a></li>
<li class="chapter" data-level="11.2" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#the-optimality-criterion-for-least-squares-regression"><i class="fa fa-check"></i><b>11.2</b> The optimality criterion for least-squares regression</a></li>
<li class="chapter" data-level="11.3" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#vector-geometric-perspective"><i class="fa fa-check"></i><b>11.3</b> Vector geometric perspective</a></li>
<li class="chapter" data-level="11.4" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#solution-for-the-least-squares-criterion"><i class="fa fa-check"></i><b>11.4</b> Solution for the least-squares criterion</a></li>
<li class="chapter" data-level="11.5" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#libraries-3"><i class="fa fa-check"></i><b>11.5</b> Libraries</a></li>
<li class="chapter" data-level="11.6" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#illustrating-linear-regression-with-simulated-data"><i class="fa fa-check"></i><b>11.6</b> Illustrating linear regression with simulated data</a></li>
<li class="chapter" data-level="11.7" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#specifying-regression-models-in-r"><i class="fa fa-check"></i><b>11.7</b> Specifying Regression Models in R</a><ul>
<li class="chapter" data-level="11.7.1" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#fitted-values"><i class="fa fa-check"></i><b>11.7.1</b> Fitted values</a></li>
<li class="chapter" data-level="11.7.2" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#getting-the-model-coefficients"><i class="fa fa-check"></i><b>11.7.2</b> Getting the model coefficients</a></li>
</ul></li>
<li class="chapter" data-level="11.8" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#residuals"><i class="fa fa-check"></i><b>11.8</b> Residuals</a></li>
<li class="chapter" data-level="11.9" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#regression-as-sum-of-squares-decomposition"><i class="fa fa-check"></i><b>11.9</b> Regression as sum-of-squares decomposition</a></li>
<li class="chapter" data-level="11.10" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#variance-explained-by-a-regression-model"><i class="fa fa-check"></i><b>11.10</b> Variance “explained” by a regression model</a></li>
<li class="chapter" data-level="11.11" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#broom-a-library-for-converting-model-results-into-data-frames"><i class="fa fa-check"></i><b>11.11</b> Broom: a library for converting model results into data frames</a><ul>
<li class="chapter" data-level="11.11.1" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#broomtidy"><i class="fa fa-check"></i><b>11.11.1</b> <code>broom::tidy</code></a></li>
<li class="chapter" data-level="11.11.2" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#broomaugment"><i class="fa fa-check"></i><b>11.11.2</b> <code>broom::augment</code></a></li>
<li class="chapter" data-level="11.11.3" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#broomglance"><i class="fa fa-check"></i><b>11.11.3</b> <code>broom::glance</code></a></li>
</ul></li>
<li class="chapter" data-level="11.12" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#example-predicting-lion-age-based-on-nose-color"><i class="fa fa-check"></i><b>11.12</b> Example: Predicting lion age based on nose color</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="multiple-regression.html"><a href="multiple-regression.html"><i class="fa fa-check"></i><b>12</b> Multiple Regression</a><ul>
<li class="chapter" data-level="12.1" data-path="multiple-regression.html"><a href="multiple-regression.html#review-of-bivariate-regression"><i class="fa fa-check"></i><b>12.1</b> Review of bivariate regression</a></li>
<li class="chapter" data-level="12.2" data-path="multiple-regression.html"><a href="multiple-regression.html#multiple-regression-1"><i class="fa fa-check"></i><b>12.2</b> Multiple regression</a><ul>
<li class="chapter" data-level="12.2.1" data-path="multiple-regression.html"><a href="multiple-regression.html#variable-space-interpretation"><i class="fa fa-check"></i><b>12.2.1</b> Variable space interpretation</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="multiple-regression.html"><a href="multiple-regression.html#subject-space-representation"><i class="fa fa-check"></i><b>12.3</b> Subject space representation</a><ul>
<li class="chapter" data-level="12.3.1" data-path="multiple-regression.html"><a href="multiple-regression.html#coefficient-of-determination-for-multiple-regression"><i class="fa fa-check"></i><b>12.3.1</b> Coefficient of determination for multiple regression</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="multiple-regression.html"><a href="multiple-regression.html#interpretting-multiple-regression"><i class="fa fa-check"></i><b>12.4</b> Interpretting Multiple Regression</a></li>
<li class="chapter" data-level="12.5" data-path="multiple-regression.html"><a href="multiple-regression.html#new-libraries-to-install"><i class="fa fa-check"></i><b>12.5</b> New Libraries to install</a></li>
<li class="chapter" data-level="12.6" data-path="multiple-regression.html"><a href="multiple-regression.html#libraries-4"><i class="fa fa-check"></i><b>12.6</b> Libraries</a></li>
<li class="chapter" data-level="12.7" data-path="multiple-regression.html"><a href="multiple-regression.html#examplar-data"><i class="fa fa-check"></i><b>12.7</b> Examplar data</a></li>
<li class="chapter" data-level="12.8" data-path="multiple-regression.html"><a href="multiple-regression.html#data-exploration"><i class="fa fa-check"></i><b>12.8</b> Data exploration</a></li>
<li class="chapter" data-level="12.9" data-path="multiple-regression.html"><a href="multiple-regression.html#d-plots"><i class="fa fa-check"></i><b>12.9</b> 3D Plots</a><ul>
<li class="chapter" data-level="12.9.1" data-path="multiple-regression.html"><a href="multiple-regression.html#scatterplot3d"><i class="fa fa-check"></i><b>12.9.1</b> scatterplot3d</a></li>
<li class="chapter" data-level="12.9.2" data-path="multiple-regression.html"><a href="multiple-regression.html#rgl"><i class="fa fa-check"></i><b>12.9.2</b> rgl</a></li>
</ul></li>
<li class="chapter" data-level="12.10" data-path="multiple-regression.html"><a href="multiple-regression.html#fitting-the-regression-model"><i class="fa fa-check"></i><b>12.10</b> Fitting the regression model</a><ul>
<li class="chapter" data-level="12.10.1" data-path="multiple-regression.html"><a href="multiple-regression.html#visualizing-the-regression-model-in-scatterplot3d"><i class="fa fa-check"></i><b>12.10.1</b> Visualizing the regression model in scatterplot3d</a></li>
</ul></li>
<li class="chapter" data-level="12.11" data-path="multiple-regression.html"><a href="multiple-regression.html#interpretting-the-regression-model"><i class="fa fa-check"></i><b>12.11</b> Interpretting the regression model</a></li>
<li class="chapter" data-level="12.12" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-vector-geometry-of-the-regression-model"><i class="fa fa-check"></i><b>12.12</b> Exploring the Vector Geometry of the Regression Model</a></li>
<li class="chapter" data-level="12.13" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-residuals-from-the-model-fit"><i class="fa fa-check"></i><b>12.13</b> Exploring the Residuals from the Model Fit</a></li>
<li class="chapter" data-level="12.14" data-path="multiple-regression.html"><a href="multiple-regression.html#an-alternate-model"><i class="fa fa-check"></i><b>12.14</b> An alternate model</a></li>
<li class="chapter" data-level="12.15" data-path="multiple-regression.html"><a href="multiple-regression.html#fitting-a-curvilinear-model-using-lm"><i class="fa fa-check"></i><b>12.15</b> Fitting a curvilinear model using lm()</a></li>
<li class="chapter" data-level="12.16" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-impact-of-nearly-collinear-predictors-on-regression"><i class="fa fa-check"></i><b>12.16</b> Exploring the impact of nearly collinear predictors on regression</a></li>
<li class="chapter" data-level="12.17" data-path="multiple-regression.html"><a href="multiple-regression.html#loess-models"><i class="fa fa-check"></i><b>12.17</b> LOESS Models</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="logistic-regression.html"><a href="logistic-regression.html"><i class="fa fa-check"></i><b>13</b> Logistic regression</a><ul>
<li class="chapter" data-level="13.1" data-path="logistic-regression.html"><a href="logistic-regression.html#a-web-app-to-explore-the-logistic-regression-equation"><i class="fa fa-check"></i><b>13.1</b> A web app to explore the logistic regression equation</a></li>
<li class="chapter" data-level="13.2" data-path="logistic-regression.html"><a href="logistic-regression.html#titanic-data-set"><i class="fa fa-check"></i><b>13.2</b> Titanic data set</a></li>
<li class="chapter" data-level="13.3" data-path="logistic-regression.html"><a href="logistic-regression.html#subsetting-the-data"><i class="fa fa-check"></i><b>13.3</b> Subsetting the data</a></li>
<li class="chapter" data-level="13.4" data-path="logistic-regression.html"><a href="logistic-regression.html#visualizing-survival-as-a-function-of-age"><i class="fa fa-check"></i><b>13.4</b> Visualizing survival as a function of age</a></li>
<li class="chapter" data-level="13.5" data-path="logistic-regression.html"><a href="logistic-regression.html#fitting-the-logistic-regression-model"><i class="fa fa-check"></i><b>13.5</b> Fitting the logistic regression model</a></li>
<li class="chapter" data-level="13.6" data-path="logistic-regression.html"><a href="logistic-regression.html#visualizing-the-logistic-regression"><i class="fa fa-check"></i><b>13.6</b> Visualizing the logistic regression</a><ul>
<li class="chapter" data-level="13.6.1" data-path="logistic-regression.html"><a href="logistic-regression.html#quick-and-easy-visualization"><i class="fa fa-check"></i><b>13.6.1</b> Quick and easy visualization</a></li>
</ul></li>
<li class="chapter" data-level="13.7" data-path="logistic-regression.html"><a href="logistic-regression.html#impact-of-sex-and-passenger-class-on-the-models"><i class="fa fa-check"></i><b>13.7</b> Impact of sex and passenger class on the models</a></li>
<li class="chapter" data-level="13.8" data-path="logistic-regression.html"><a href="logistic-regression.html#fitting-multiple-models-based-on-groupings-use-dplyrdo"><i class="fa fa-check"></i><b>13.8</b> Fitting multiple models based on groupings use <code>dplyr::do</code></a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html"><i class="fa fa-check"></i><b>14</b> Principal Components Analysis</a><ul>
<li class="chapter" data-level="14.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#libraries-5"><i class="fa fa-check"></i><b>14.1</b> Libraries</a></li>
<li class="chapter" data-level="14.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#matrices-as-linear-transformations"><i class="fa fa-check"></i><b>14.2</b> Matrices as linear transformations</a></li>
<li class="chapter" data-level="14.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#eigenanalysis-in-r"><i class="fa fa-check"></i><b>14.3</b> Eigenanalysis in R</a></li>
<li class="chapter" data-level="14.4" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#principal-components-analysis-in-r"><i class="fa fa-check"></i><b>14.4</b> Principal Components Analysis in R</a><ul>
<li class="chapter" data-level="14.4.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#bioenv-dataset"><i class="fa fa-check"></i><b>14.4.1</b> Bioenv dataset</a></li>
<li class="chapter" data-level="14.4.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pca-of-the-bioenv-dataset"><i class="fa fa-check"></i><b>14.4.2</b> PCA of the Bioenv dataset</a></li>
<li class="chapter" data-level="14.4.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#calculating-factor-loadings"><i class="fa fa-check"></i><b>14.4.3</b> Calculating Factor Loadings</a></li>
</ul></li>
<li class="chapter" data-level="14.5" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#drawing-figures-to-represent-pca"><i class="fa fa-check"></i><b>14.5</b> Drawing Figures to Represent PCA</a><ul>
<li class="chapter" data-level="14.5.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pc-score-plots"><i class="fa fa-check"></i><b>14.5.1</b> PC Score Plots</a></li>
<li class="chapter" data-level="14.5.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#simultaneous-depiction-of-observations-and-variables-in-the-pc-space"><i class="fa fa-check"></i><b>14.5.2</b> Simultaneous Depiction of Observations and Variables in the PC Space</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="15" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html"><i class="fa fa-check"></i><b>15</b> Singular Value Decomposition</a><ul>
<li class="chapter" data-level="15.1" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#libraries-6"><i class="fa fa-check"></i><b>15.1</b> Libraries</a></li>
<li class="chapter" data-level="15.2" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#svd-in-r"><i class="fa fa-check"></i><b>15.2</b> SVD in R</a></li>
<li class="chapter" data-level="15.3" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#writing-our-own-pca-function-using-svd"><i class="fa fa-check"></i><b>15.3</b> Writing our own PCA function using SVD</a></li>
<li class="chapter" data-level="15.4" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#writing-a-biplot-function"><i class="fa fa-check"></i><b>15.4</b> Writing a biplot function</a></li>
<li class="chapter" data-level="15.5" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#data-compression-and-noise-filtering-using-svd"><i class="fa fa-check"></i><b>15.5</b> Data compression and noise filtering using SVD</a><ul>
<li class="chapter" data-level="15.5.1" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#example-image"><i class="fa fa-check"></i><b>15.5.1</b> Example image</a></li>
</ul></li>
<li class="chapter" data-level="15.6" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#image-compression-using-svd-in-r"><i class="fa fa-check"></i><b>15.6</b> Image compression using SVD in R</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html"><i class="fa fa-check"></i><b>16</b> Canonical Variates Analysis</a><ul>
<li class="chapter" data-level="16.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#libraries-7"><i class="fa fa-check"></i><b>16.1</b> Libraries</a></li>
<li class="chapter" data-level="16.2" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#discriminant-analysis-in-r"><i class="fa fa-check"></i><b>16.2</b> Discriminant Analysis in R</a><ul>
<li class="chapter" data-level="16.2.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#shorthand-formulae-in-r"><i class="fa fa-check"></i><b>16.2.1</b> Shorthand Formulae in R</a></li>
<li class="chapter" data-level="16.2.2" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#working-with-the-output-of-lda"><i class="fa fa-check"></i><b>16.2.2</b> Working with the output of <code>lda()</code></a></li>
</ul></li>
<li class="chapter" data-level="16.3" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#estimating-confidence-regions-for-group-means-in-cva"><i class="fa fa-check"></i><b>16.3</b> Estimating confidence regions for group means in CVA</a><ul>
<li class="chapter" data-level="16.3.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#drawing-the-cva-confidence-regions"><i class="fa fa-check"></i><b>16.3.1</b> Drawing the CVA confidence regions</a></li>
</ul></li>
<li class="chapter" data-level="16.4" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#calculating-the-within-and-between-group-covariance-matrices"><i class="fa fa-check"></i><b>16.4</b> Calculating the Within and Between Group Covariance Matrices</a><ul>
<li class="chapter" data-level="16.4.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#recapitulating-the-cva-analysis-of-lda"><i class="fa fa-check"></i><b>16.4.1</b> Recapitulating the CVA analysis of <code>lda()</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="17" data-path="clustering-in-r.html"><a href="clustering-in-r.html"><i class="fa fa-check"></i><b>17</b> Clustering in R</a><ul>
<li class="chapter" data-level="17.1" data-path="clustering-in-r.html"><a href="clustering-in-r.html#libraries-8"><i class="fa fa-check"></i><b>17.1</b> Libraries</a></li>
<li class="chapter" data-level="17.2" data-path="clustering-in-r.html"><a href="clustering-in-r.html#data-set"><i class="fa fa-check"></i><b>17.2</b> Data set</a></li>
<li class="chapter" data-level="17.3" data-path="clustering-in-r.html"><a href="clustering-in-r.html#hierarchical-clustering-in-r"><i class="fa fa-check"></i><b>17.3</b> Hierarchical Clustering in R</a></li>
<li class="chapter" data-level="17.4" data-path="clustering-in-r.html"><a href="clustering-in-r.html#manipulating-hierarchical-clusterings-with-dendextend"><i class="fa fa-check"></i><b>17.4</b> Manipulating hierarchical clusterings with dendextend</a></li>
<li class="chapter" data-level="17.5" data-path="clustering-in-r.html"><a href="clustering-in-r.html#plotting-dendrograms-in-dendextend"><i class="fa fa-check"></i><b>17.5</b> Plotting dendrograms in dendextend</a></li>
<li class="chapter" data-level="17.6" data-path="clustering-in-r.html"><a href="clustering-in-r.html#cutting-dendrograms"><i class="fa fa-check"></i><b>17.6</b> Cutting dendrograms</a></li>
<li class="chapter" data-level="17.7" data-path="clustering-in-r.html"><a href="clustering-in-r.html#looking-at-clusters"><i class="fa fa-check"></i><b>17.7</b> Looking at clusters</a></li>
<li class="chapter" data-level="17.8" data-path="clustering-in-r.html"><a href="clustering-in-r.html#generating-a-heat-map-from-a-cluster"><i class="fa fa-check"></i><b>17.8</b> Generating a heat map from a cluster</a></li>
<li class="chapter" data-level="17.9" data-path="clustering-in-r.html"><a href="clustering-in-r.html#working-with-sub-trees"><i class="fa fa-check"></i><b>17.9</b> Working with sub-trees</a></li>
<li class="chapter" data-level="17.10" data-path="clustering-in-r.html"><a href="clustering-in-r.html#setting-dendrogram-parameters-in-dendextend"><i class="fa fa-check"></i><b>17.10</b> Setting dendrogram parameters in dendextend</a></li>
<li class="chapter" data-level="17.11" data-path="clustering-in-r.html"><a href="clustering-in-r.html#combining-heatmaps-and-dendrograms"><i class="fa fa-check"></i><b>17.11</b> Combining heatmaps and dendrograms</a></li>
<li class="chapter" data-level="17.12" data-path="clustering-in-r.html"><a href="clustering-in-r.html#k-meansk-medoids-clustering-in-r"><i class="fa fa-check"></i><b>17.12</b> K-means/K-medoids Clustering in R</a><ul>
<li class="chapter" data-level="17.12.1" data-path="clustering-in-r.html"><a href="clustering-in-r.html#heat-map-from-k-medoids-cluster"><i class="fa fa-check"></i><b>17.12.1</b> Heat map from k-medoids cluster</a></li>
</ul></li>
<li class="chapter" data-level="17.13" data-path="clustering-in-r.html"><a href="clustering-in-r.html#combining-clusters-and-correlation-matrix-heatmaps"><i class="fa fa-check"></i><b>17.13</b> Combining clusters and correlation matrix heatmaps</a><ul>
<li class="chapter" data-level="17.13.1" data-path="clustering-in-r.html"><a href="clustering-in-r.html#hierarchical-clustering-visualized-on-correlation-matrix"><i class="fa fa-check"></i><b>17.13.1</b> Hierarchical clustering, visualized on correlation matrix</a></li>
<li class="chapter" data-level="17.13.2" data-path="clustering-in-r.html"><a href="clustering-in-r.html#k-medoids-clustering-visualized-on-the-correlation-matrix"><i class="fa fa-check"></i><b>17.13.2</b> K-medoids clustering visualized on the correlation matrix</a></li>
</ul></li>
<li class="chapter" data-level="17.14" data-path="clustering-in-r.html"><a href="clustering-in-r.html#depicting-the-data-within-clusters"><i class="fa fa-check"></i><b>17.14</b> Depicting the data within clusters</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html"><i class="fa fa-check"></i><b>18</b> Simulation I: Sampling distributions</a><ul>
<li class="chapter" data-level="18.1" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#sampling-distrubtions"><i class="fa fa-check"></i><b>18.1</b> Sampling distrubtions</a></li>
<li class="chapter" data-level="18.2" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#simulating-sampling-from-an-underlying-population"><i class="fa fa-check"></i><b>18.2</b> Simulating sampling from an underlying population</a></li>
<li class="chapter" data-level="18.3" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#seeding-the-pseudo-random-number-generator"><i class="fa fa-check"></i><b>18.3</b> Seeding the pseudo-random number generator</a></li>
<li class="chapter" data-level="18.4" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#libraries-9"><i class="fa fa-check"></i><b>18.4</b> Libraries</a><ul>
<li class="chapter" data-level="18.4.1" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#properties-of-the-underlying-population"><i class="fa fa-check"></i><b>18.4.1</b> Properties of the underlying population</a></li>
<li class="chapter" data-level="18.4.2" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#other-r-functions-related-to-the-normal-distribution"><i class="fa fa-check"></i><b>18.4.2</b> Other R functions related to the normal distribution</a></li>
</ul></li>
<li class="chapter" data-level="18.5" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#random-sampling-from-the-simulated-population"><i class="fa fa-check"></i><b>18.5</b> Random sampling from the simulated population</a><ul>
<li class="chapter" data-level="18.5.1" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#another-random-sample"><i class="fa fa-check"></i><b>18.5.1</b> Another random sample</a></li>
<li class="chapter" data-level="18.5.2" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#simulating-the-generation-of-many-random-samples"><i class="fa fa-check"></i><b>18.5.2</b> Simulating the generation of many random samples</a></li>
<li class="chapter" data-level="18.5.3" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#a-function-to-estimate-statistics-of-interest-in-a-random-sample"><i class="fa fa-check"></i><b>18.5.3</b> A function to estimate statistics of interest in a random sample</a></li>
<li class="chapter" data-level="18.5.4" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#generating-statistics-for-many-random-samples"><i class="fa fa-check"></i><b>18.5.4</b> Generating statistics for many random samples</a></li>
</ul></li>
<li class="chapter" data-level="18.6" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#simulated-sampling-distribution-of-the-mean"><i class="fa fa-check"></i><b>18.6</b> Simulated sampling distribution of the mean</a><ul>
<li class="chapter" data-level="18.6.1" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#differences-between-sampling-distributions-and-samplepopulation-distributions"><i class="fa fa-check"></i><b>18.6.1</b> Differences between sampling distributions and sample/population distributions</a></li>
<li class="chapter" data-level="18.6.2" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#using-sampling-distributions-to-understand-the-behavior-of-statistics-of-interest"><i class="fa fa-check"></i><b>18.6.2</b> Using sampling distributions to understand the behavior of statistics of interest</a></li>
<li class="chapter" data-level="18.6.3" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#sampling-distributions-for-different-sample-sizes"><i class="fa fa-check"></i><b>18.6.3</b> Sampling distributions for different sample sizes</a></li>
<li class="chapter" data-level="18.6.4" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#discussion-of-trends-for-sampling-distributions-of-different-sample-sizes"><i class="fa fa-check"></i><b>18.6.4</b> Discussion of trends for sampling distributions of different sample sizes</a></li>
</ul></li>
<li class="chapter" data-level="18.7" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#standard-error-of-the-mean"><i class="fa fa-check"></i><b>18.7</b> Standard Error of the Mean</a></li>
<li class="chapter" data-level="18.8" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#sampling-distribution-of-the-standard-deviation"><i class="fa fa-check"></i><b>18.8</b> Sampling Distribution of the Standard Deviation</a><ul>
<li class="chapter" data-level="18.8.1" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#standard-error-of-standard-deviations"><i class="fa fa-check"></i><b>18.8.1</b> Standard error of standard deviations</a></li>
</ul></li>
<li class="chapter" data-level="18.9" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#what-happens-to-the-sampling-distribution-of-the-mean-and-standard-deviation-when-our-sample-size-is-small"><i class="fa fa-check"></i><b>18.9</b> What happens to the sampling distribution of the mean and standard deviation when our sample size is small?</a><ul>
<li class="chapter" data-level="18.9.1" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#for-small-samples-sample-standard-deviations-systematically-underestimate-the-population-standard-deviation"><i class="fa fa-check"></i><b>18.9.1</b> For small samples, sample standard deviations systematically underestimate the population standard deviation</a></li>
<li class="chapter" data-level="18.9.2" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#underestimates-of-the-standard-deviation-for-small-n-lead-to-understimates-of-the-se-of-the-mean"><i class="fa fa-check"></i><b>18.9.2</b> Underestimates of the standard deviation for small <span class="math inline">\(n\)</span> lead to understimates of the SE of the mean</a></li>
<li class="chapter" data-level="18.9.3" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#the-t-distribution-is-the-appropriate-distribution-for-describing-the-sampling-distribution-of-the-mean-when-n-is-small"><i class="fa fa-check"></i><b>18.9.3</b> The t-distribution is the appropriate distribution for describing the sampling distribution of the mean when <span class="math inline">\(n\)</span> is small</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="19" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html"><i class="fa fa-check"></i><b>19</b> Simulating confidence intervals</a><ul>
<li class="chapter" data-level="19.1" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#confidence-intervals"><i class="fa fa-check"></i><b>19.1</b> Confidence intervals</a></li>
<li class="chapter" data-level="19.2" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#generic-formulation-for-confidence-intervals"><i class="fa fa-check"></i><b>19.2</b> Generic formulation for confidence intervals</a></li>
<li class="chapter" data-level="19.3" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#example-confidence-intervals-for-the-mean"><i class="fa fa-check"></i><b>19.3</b> Example: Confidence intervals for the mean</a><ul>
<li class="chapter" data-level="19.3.1" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#simulation-of-means"><i class="fa fa-check"></i><b>19.3.1</b> Simulation of means</a></li>
<li class="chapter" data-level="19.3.2" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#distance-between-sample-means-and-true-means"><i class="fa fa-check"></i><b>19.3.2</b> Distance between sample means and true means</a></li>
<li class="chapter" data-level="19.3.3" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#calculating-a-ci"><i class="fa fa-check"></i><b>19.3.3</b> Calculating a CI</a></li>
</ul></li>
<li class="chapter" data-level="19.4" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#a-problem-arises"><i class="fa fa-check"></i><b>19.4</b> A problem arises!</a></li>
<li class="chapter" data-level="19.5" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#confidence-intervals-under-sample-estimates-of-standard-errors"><i class="fa fa-check"></i><b>19.5</b> Confidence intervals under sample estimates of standard errors</a><ul>
<li class="chapter" data-level="19.5.1" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#the-spread-of-sample-estimates-of-the-mean-in-units-of-true-and-estimates-ses"><i class="fa fa-check"></i><b>19.5.1</b> The spread of sample estimates of the mean in units of true and estimates SEs</a></li>
<li class="chapter" data-level="19.5.2" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#quantifying-deviations-of-sample-estimates-of-the-mean"><i class="fa fa-check"></i><b>19.5.2</b> Quantifying deviations of sample estimates of the mean</a></li>
</ul></li>
<li class="chapter" data-level="19.6" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#t-distribution-for-confidence-intervals-of-the-mean"><i class="fa fa-check"></i><b>19.6</b> t-distribution for confidence intervals of the mean</a><ul>
<li class="chapter" data-level="19.6.1" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#t-distribution-vs-standard-normal-distribution"><i class="fa fa-check"></i><b>19.6.1</b> t-distribution vs standard normal distribution</a></li>
<li class="chapter" data-level="19.6.2" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#formula-for-confidence-intervals-of-the-mean-based-on-sample-estimates-of-the-se-and-the-t-distribution"><i class="fa fa-check"></i><b>19.6.2</b> Formula for confidence intervals of the mean based on sample estimates of the SE and the t-distribution</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Biology 723: Statistical Computing for Biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="data-wrangling" class="section level1">
<h1><span class="header-section-number">Chapter 7</span> Data wrangling</h1>
<p>In the real world you’ll often create a data set (or be given one) in a format that is less than ideal for analysis. This can happen for a number of reasons. For example, the data may have been recorded in a manner convenient for collection and visual inspection, but which does not work well for analysis and plotting. Or the data may be an amalgamation of multiple experiments, in which each of the experimenters used slightly different naming conventions. Or the data may have been produced by an instrument that produces output with a fixed format. Sometimes important experimental information is included in the column headers of a spreadsheet.</p>
<p>Whatever the case, we often find ourselves in the situation where we need to “wrangle” our data into a “tidy” format before we can proceed with visualization and analysis. The “R for Data Science” text discusses some desirable rules for “tidy” data in order to facilitate downstream analyses. These are:</p>
<ol style="list-style-type: decimal">
<li>Each variable must have its own column.</li>
<li>Each observation must have its own row.</li>
<li>Each value must have its own cell.</li>
</ol>
<p>In this lecture we’re going to walk through an extended example of wrangling some data into a “tidy” format.</p>
<div id="libraries-1" class="section level2">
<h2><span class="header-section-number">7.1</span> Libraries</h2>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(magrittr)
<span class="kw">library</span>(stringr)
<span class="kw">library</span>(tidyverse)
<span class="kw">library</span>(cowplot)</code></pre>
</div>
<div id="data" class="section level2">
<h2><span class="header-section-number">7.2</span> Data</h2>
<p>To illustrate a data wrangling pipeline, we’re going to use a gene expression microarray data set, based on the following paper:</p>
<ul>
<li>Spellman PT, et al. 1998. Comprehensive identification of cell cycle-regulated genes of the yeast Saccharomyces cerevisiae by microarray hybridization. Mol Biol Cell 9(12): 3273-97.</li>
</ul>
<p>In this paper, Spellman and colleagues tried to identify all the genes in the yeast genome (&gt;6000 genes) that exhibited oscillatory behaviors suggestive of cell cycle regulation. To do so, they combined gene expression measurements from six different types of cell cycle synchronization experiments.</p>
<p><em>Download the Spellman data to your filesystem</em> from <a href="https://github.com/Bio723-class/example-datasets/blob/master/spellman-combined.txt">this link</a> (right-click the “Download” button and save to your Downloads folder or similar).</p>
<p>I suggest that once you download the data, you open it in a spreadsheet program (e.g. Excel) or use the RStudio Data Viewer to get a sense of what the data looks like.</p>
<p>Let’s load it into R, using the <code>read_tsv()</code> function, using the appropriate file path.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># the filepath may differ on your computer</span>
spellman &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;~/Downloads/spellman-combined.txt&quot;</span>)
<span class="co">## Parsed with column specification:</span>
<span class="co">## cols(</span>
<span class="co">##   .default = col_double(),</span>
<span class="co">##   X1 = col_character(),</span>
<span class="co">##   clb = col_logical(),</span>
<span class="co">##   alpha = col_logical(),</span>
<span class="co">##   cdc15 = col_logical(),</span>
<span class="co">##   cdc28 = col_logical(),</span>
<span class="co">##   elu = col_logical()</span>
<span class="co">## )</span>
<span class="co">## See spec(...) for full column specifications.</span></code></pre>
<p>The initial dimenions of the data frame are:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(spellman)
<span class="co">## [1] 6178   83</span></code></pre>
<p>The six types of cell cycle synchronization experiments included in this data set are:</p>
<ol style="list-style-type: decimal">
<li>synchronization by alpha-factor = “alpha”</li>
<li>synchronization by cdc15 temperature sensitive mutants = “cdc15”</li>
<li>synchronization by cdc28 temperature sensitive mutants = “cdc28”</li>
<li>synchronization by elutration = “elu”</li>
<li>synchronization by cln3 mutatant strains = “cln3”</li>
<li>synchronization by clb2 mutant strains = “clb2”</li>
</ol>
</div>
<div id="renaming-data-frame-columms" class="section level2">
<h2><span class="header-section-number">7.3</span> Renaming data frame columms</h2>
<p>Notice that when we imported the data we got a warning message: <code>Missing column names filled in: 'X1' [1]</code>. In a data frame, every column must have a name. The first column of our data set did not have a name in the header, so <code>read_tsv</code> automatically gave it the name <code>X1</code>.</p>
<p>Our first task is to give the first column a more meaningful name. This column gives “systematic gene names” – a standardized naming scheme for genes in the yeast genome. We’ll use <code>dplyr::rename</code> to rename <code>X1</code> to <code>gene</code>. Note that <code>rename</code> can take multiple arguments if you need to rename multiple columns simultaneously.</p>
<pre class="sourceCode r"><code class="sourceCode r">spellman.clean &lt;-<span class="st"> </span>
<span class="st">  </span>spellman <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">rename</span>(<span class="dt">gene =</span> X1)</code></pre>
<p>Note the use of the compound assingment operator – <code>%&lt;&gt;%</code> – from the <code>magrittr</code> package, which we introduced in our last class session.</p>
</div>
<div id="dropping-unneeded-columns" class="section level2">
<h2><span class="header-section-number">7.4</span> Dropping unneeded columns</h2>
<p>Take a look at the Spellman data again in your spreadsheet program (or the RStudio data viewer). You’ll notice there are some blank columns. For example there is a column with the header “alpha” that has no entries. These are simply visual organizing elements that the creator of the spreadsheet added to separate the different experiments that are included in the data set.</p>
<p>We can use <code>dplyr::select()</code> to drop columns by prepending column names with the negative sign:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># drop the alpha column keeping all others</span>
spellman.clean <span class="op">%&lt;&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>alpha)  </code></pre>
<p>Note that usually <code>select()</code> keeps only the variables you specify. However if the first expression is negative, <code>select</code> will instead automatically keep all variables, dropping only those you specify.</p>
<div id="finding-all-empty-columns" class="section level3">
<h3><span class="header-section-number">7.4.1</span> Finding all empty columns</h3>
<p>In the example above, we looked at the data and saw that the “alpha” column was empty, and thus dropped it. This worked because there are only a modest number of columns in the data frame in it’s initial form. However, if our data frame contained thousands of columns, this “look and see” procedure would not be efficient. Can we come up with a general solution for removing empty columns from a data frame?</p>
<p>When you load a data frame from a spreadsheet, empty cells are given the value <code>NA</code>. In previous class sessions we were introduced to the function <code>is.na()</code> which tests each value in a vector or data frame for whether it’s NA or not. We can count NA values in a vector by summing the output of <code>is.na()</code>. Conversely we can count the number of “not NA” items by using the negation operator (<code>!</code>):</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># count number of NA values in the alpha0 column</span>
<span class="kw">sum</span>(<span class="kw">is.na</span>(spellman<span class="op">$</span>alpha0))
<span class="co">## [1] 165</span>

<span class="co"># count number of values that are NOT NA in alpha0</span>
<span class="kw">sum</span>(<span class="op">!</span><span class="kw">is.na</span>(spellman<span class="op">$</span>alpha0))
<span class="co">## [1] 6013</span></code></pre>
<p>This seems like it should get us close to a solution but <code>sum(is.na(..))</code> when applied to a data frame counts NAs across the entire data frame, not column-by-column.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># doesn&#39;t do what we hoped!</span>
<span class="kw">sum</span>(<span class="kw">is.na</span>(spellman))
<span class="co">## [1] 59017</span></code></pre>
<p>If we want sums of NAs by column, we instead use the <code>colSums()</code> function:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get number of NAs by column</span>
<span class="kw">colSums</span>(<span class="kw">is.na</span>(spellman))
<span class="co">##        X1    cln3-1    cln3-2       clb    clb2-2    clb2-1     alpha </span>
<span class="co">##         0       193       365      6178       454       142      6178 </span>
<span class="co">##    alpha0    alpha7   alpha14   alpha21   alpha28   alpha35   alpha42 </span>
<span class="co">##       165       525       191       312       267       207       123 </span>
<span class="co">##   alpha49   alpha56   alpha63   alpha70   alpha77   alpha84   alpha91 </span>
<span class="co">##       257       147       186       185       178       155       329 </span>
<span class="co">##   alpha98  alpha105  alpha112  alpha119     cdc15  cdc15_10  cdc15_30 </span>
<span class="co">##       209       174       222       251      6178       677       477 </span>
<span class="co">##  cdc15_50  cdc15_70  cdc15_80  cdc15_90 cdc15_100 cdc15_110 cdc15_120 </span>
<span class="co">##       501       608       573       562       606       570       611 </span>
<span class="co">## cdc15_130 cdc15_140 cdc15_150 cdc15_160 cdc15_170 cdc15_180 cdc15_190 </span>
<span class="co">##       495       574       811       583       571       803       613 </span>
<span class="co">## cdc15_200 cdc15_210 cdc15_220 cdc15_230 cdc15_240 cdc15_250 cdc15_270 </span>
<span class="co">##      1014       573       741       596       847       379       537 </span>
<span class="co">## cdc15_290     cdc28   cdc28_0  cdc28_10  cdc28_20  cdc28_30  cdc28_40 </span>
<span class="co">##       426      6178       122        72        67        55        66 </span>
<span class="co">##  cdc28_50  cdc28_60  cdc28_70  cdc28_80  cdc28_90 cdc28_100 cdc28_110 </span>
<span class="co">##        56        82        84        75       237       165       319 </span>
<span class="co">## cdc28_120 cdc28_130 cdc28_140 cdc28_150 cdc28_160       elu      elu0 </span>
<span class="co">##       312      1439      2159       521       543      6178       122 </span>
<span class="co">##     elu30     elu60     elu90    elu120    elu150    elu180    elu210 </span>
<span class="co">##       153       175       132       103       119       111       118 </span>
<span class="co">##    elu240    elu270    elu300    elu330    elu360    elu390 </span>
<span class="co">##       131       110       112       112       156       114</span></code></pre>
<p>Columns with <em>all</em> missing values can be more conveniently found by asking for those columns where the number of “not missing” values is zero:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get names of all columns for which all rows are NA</span>
<span class="co"># useing standard indexing</span>
<span class="kw">names</span>(spellman)[<span class="kw">colSums</span>(<span class="op">!</span><span class="kw">is.na</span>(spellman)) <span class="op">==</span><span class="st"> </span><span class="dv">0</span>]
<span class="co">## [1] &quot;clb&quot;   &quot;alpha&quot; &quot;cdc15&quot; &quot;cdc28&quot; &quot;elu&quot;</span></code></pre>
<p>We can combine the <code>colSums(!is.na())</code> idiom with the <code>dplyr::select_if</code> function to quickly remove all empty columns as so:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># this broke in dplyr 0.80 due to a bug (https://github.com/tidyverse/dplyr/issues/4213)</span>
<span class="co"># supposedly a fix is on the way. In the iterim, a work around is included</span>

<span class="co">#spellman.clean %&lt;&gt;%</span>
  <span class="co"># keep ONLY the non-empty columns</span>
  <span class="co">#select_if(colSums(!is.na(.)) &gt; 0)  </span>

not.all.na &lt;-<span class="st"> </span><span class="cf">function</span>(x) { 
  <span class="cf">if</span> (<span class="kw">sum</span>(<span class="op">!</span><span class="kw">is.na</span>(x)) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>){
    <span class="kw">return</span>(<span class="ot">TRUE</span>)
    }
  <span class="cf">else</span> {
    <span class="kw">return</span>(<span class="ot">FALSE</span>)
    }
}

spellman.clean <span class="op">%&lt;&gt;%</span>
<span class="st">  </span><span class="kw">select_if</span>(not.all.na)</code></pre>
</div>
<div id="dropping-columns-by-matching-names" class="section level3">
<h3><span class="header-section-number">7.4.2</span> Dropping columns by matching names</h3>
<p>Only two time points from the cln3 and clb2 experiments were reported in the original publication. Since complete time series are unavailable for these two experimental conditions we will drop them from further consideration.</p>
<p><code>select()</code> can be called be called with a number of “helper function” (<code>?select_helpers</code>). Here we’ll illustrate the <code>matches()</code> helper function which matches column names to a “regular expression”. Regular expressions (also referred to as “regex” or “regexp”) are a way of specifying patterns in strings. For the purposes of this document we’ll illustrate regexs by example; for a more detailed explanation of regular expressions see the the regex help(<code>?regex</code>) and the <a href="http://r4ds.had.co.nz/strings.html">Chapter on Strings in “R for Data Analysis”</a>:</p>
<p>Let’s see how to drop all the “cln3” and “clb2” columns from the data frame using <code>matches()</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r">spellman.clean <span class="op">%&lt;&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span><span class="kw">matches</span>(<span class="st">&quot;cln3&quot;</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span><span class="kw">matches</span>(<span class="st">&quot;clb2&quot;</span>))</code></pre>
<p>If we wanted we could have collapsed our two match statements into one as follows:</p>
<pre class="sourceCode r"><code class="sourceCode r">spellman.clean <span class="op">%&lt;&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span><span class="kw">matches</span>(<span class="st">&quot;cln3|clb2&quot;</span>))</code></pre>
<p>In this second example, the character “|” is specifing an OR match within the regular expression, so this regular expression matches column names that contain “cln3” OR “clb2”.</p>
</div>
</div>
<div id="merging-data-frames" class="section level2">
<h2><span class="header-section-number">7.5</span> Merging data frames</h2>
<p>Often you’ll find yourself in the situation where you want to combine information from multiple data sources. The usual requirement is that the data sources have one or more shared columns, that allow you to relate the entities or observations (rows) between the data sets. <code>dplyr</code> provides a variety of <code>join</code> functions to handle different data merging operators.</p>
<p>To illustrating merging or joining data sources, we’ll add information about each genes “common name” and a description of the gene functions to our Spellman data set. I’ve prepared a file with this info based on info I downloaded from the <a href="https://www.yeastgenome.org">Saccharomyces Genome Database</a>.</p>
<pre class="sourceCode r"><code class="sourceCode r">gene.info &lt;-<span class="st"> </span><span class="kw">read_csv</span>(<span class="st">&quot;https://github.com/bio304-class/bio304-course-notes/raw/master/datasets/yeast-ORF-info.csv&quot;</span>)
<span class="co">## Parsed with column specification:</span>
<span class="co">## cols(</span>
<span class="co">##   ftr.name = col_character(),</span>
<span class="co">##   std.name = col_character(),</span>
<span class="co">##   description = col_character()</span>
<span class="co">## )</span></code></pre>
<p>Having loaded the data, let’s get a quick overview of it’s structure:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(gene.info)
<span class="co">## [1] &quot;ftr.name&quot;    &quot;std.name&quot;    &quot;description&quot;</span>
<span class="kw">dim</span>(gene.info)
<span class="co">## [1] 6610    3</span>
<span class="kw">head</span>(gene.info)
<span class="co">## # A tibble: 6 x 3</span>
<span class="co">##   ftr.name  std.name description                                           </span>
<span class="co">##   &lt;chr&gt;     &lt;chr&gt;    &lt;chr&gt;                                                 </span>
<span class="co">## 1 YAL069W   &lt;NA&gt;     Dubious open reading frame; unlikely to encode a func…</span>
<span class="co">## 2 YAL068W-A &lt;NA&gt;     Dubious open reading frame; unlikely to encode a func…</span>
<span class="co">## 3 YAL068C   PAU8     Protein of unknown function; member of the seripauper…</span>
<span class="co">## 4 YAL067W-A &lt;NA&gt;     Putative protein of unknown function; identified by g…</span>
<span class="co">## 5 YAL067C   SEO1     Putative permease; member of the allantoate transport…</span>
<span class="co">## 6 YAL066W   &lt;NA&gt;     Dubious open reading frame; unlikely to encode a func…</span></code></pre>
<p>In <code>gene.info</code>, the <code>ftr.name</code> column corresponds to the <code>gene</code> column in our Spellman data set. The <code>std.name</code> column gives the “common” gene name (not every gene has a common name so there are lots of NAs). The <code>description</code> column gives a brief textual description of what the gene product does.</p>
<p>To combine <code>spellmean.clean</code> with <code>gene.info</code> we use the <code>left_join</code> function defined in dplyr. As noted in the description of the function, <code>left_join(x, y)</code> returns “all rows from x, and all columns from x and y. Rows in x with no match in y will have NA values in the new columns.” In addition, we have to specify the column to join by using the <code>by</code> argument to <code>left_join</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r">spellman.merged &lt;-
<span class="st">  </span><span class="kw">left_join</span>(spellman.clean, gene.info, <span class="dt">by =</span> <span class="kw">c</span>(<span class="st">&quot;gene&quot;</span> =<span class="st"> &quot;ftr.name&quot;</span>))</code></pre>
<p>By default, the joined columns are merged at the end of the data frame, so we’ll reorder variables to bring the <code>std.name</code> and <code>description</code> to the second and thirds columns, preserving the order of all the other colums.</p>
<pre class="sourceCode r"><code class="sourceCode r">spellman.merged <span class="op">%&lt;&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(gene, std.name, description, <span class="kw">everything</span>())

spellman.merged
<span class="co">## # A tibble: 6,178 x 76</span>
<span class="co">##    gene  std.name description alpha0 alpha7 alpha14 alpha21 alpha28 alpha35</span>
<span class="co">##    &lt;chr&gt; &lt;chr&gt;    &lt;chr&gt;        &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">##  1 YAL0… TFC3     Subunit of…  -0.15  -0.15   -0.21    0.17   -0.42  -0.44 </span>
<span class="co">##  2 YAL0… VPS8     Membrane-b…  -0.11   0.1     0.01    0.06    0.04  -0.26 </span>
<span class="co">##  3 YAL0… EFB1     Translatio…  -0.14  -0.71    0.1    -0.32   -0.4   -0.580</span>
<span class="co">##  4 YAL0… &lt;NA&gt;     Dubious op…  -0.02  -0.48   -0.11    0.12   -0.03   0.19 </span>
<span class="co">##  5 YAL0… SSA1     ATPase inv…  -0.05  -0.53   -0.47   -0.06    0.11  -0.07 </span>
<span class="co">##  6 YAL0… ERP2     Member of …  -0.6   -0.45   -0.13    0.35   -0.01   0.49 </span>
<span class="co">##  7 YAL0… FUN14    Integral m…  -0.28  -0.22   -0.06    0.22    0.25   0.13 </span>
<span class="co">##  8 YAL0… SPO7     Putative r…  -0.03  -0.27    0.17   -0.12   -0.27   0.06 </span>
<span class="co">##  9 YAL0… MDM10    Subunit of…  -0.05   0.13    0.13   -0.21   -0.45  -0.21 </span>
<span class="co">## 10 YAL0… SWC3     Protein of…  -0.31  -0.43   -0.3    -0.23   -0.13  -0.07 </span>
<span class="co">## # … with 6,168 more rows, and 67 more variables: alpha42 &lt;dbl&gt;,</span>
<span class="co">## #   alpha49 &lt;dbl&gt;, alpha56 &lt;dbl&gt;, alpha63 &lt;dbl&gt;, alpha70 &lt;dbl&gt;,</span>
<span class="co">## #   alpha77 &lt;dbl&gt;, alpha84 &lt;dbl&gt;, alpha91 &lt;dbl&gt;, alpha98 &lt;dbl&gt;,</span>
<span class="co">## #   alpha105 &lt;dbl&gt;, alpha112 &lt;dbl&gt;, alpha119 &lt;dbl&gt;, cdc15_10 &lt;dbl&gt;,</span>
<span class="co">## #   cdc15_30 &lt;dbl&gt;, cdc15_50 &lt;dbl&gt;, cdc15_70 &lt;dbl&gt;, cdc15_80 &lt;dbl&gt;,</span>
<span class="co">## #   cdc15_90 &lt;dbl&gt;, cdc15_100 &lt;dbl&gt;, cdc15_110 &lt;dbl&gt;, cdc15_120 &lt;dbl&gt;,</span>
<span class="co">## #   cdc15_130 &lt;dbl&gt;, cdc15_140 &lt;dbl&gt;, cdc15_150 &lt;dbl&gt;, cdc15_160 &lt;dbl&gt;,</span>
<span class="co">## #   cdc15_170 &lt;dbl&gt;, cdc15_180 &lt;dbl&gt;, cdc15_190 &lt;dbl&gt;, cdc15_200 &lt;dbl&gt;,</span>
<span class="co">## #   cdc15_210 &lt;dbl&gt;, cdc15_220 &lt;dbl&gt;, cdc15_230 &lt;dbl&gt;, cdc15_240 &lt;dbl&gt;,</span>
<span class="co">## #   cdc15_250 &lt;dbl&gt;, cdc15_270 &lt;dbl&gt;, cdc15_290 &lt;dbl&gt;, cdc28_0 &lt;dbl&gt;,</span>
<span class="co">## #   cdc28_10 &lt;dbl&gt;, cdc28_20 &lt;dbl&gt;, cdc28_30 &lt;dbl&gt;, cdc28_40 &lt;dbl&gt;,</span>
<span class="co">## #   cdc28_50 &lt;dbl&gt;, cdc28_60 &lt;dbl&gt;, cdc28_70 &lt;dbl&gt;, cdc28_80 &lt;dbl&gt;,</span>
<span class="co">## #   cdc28_90 &lt;dbl&gt;, cdc28_100 &lt;dbl&gt;, cdc28_110 &lt;dbl&gt;, cdc28_120 &lt;dbl&gt;,</span>
<span class="co">## #   cdc28_130 &lt;dbl&gt;, cdc28_140 &lt;dbl&gt;, cdc28_150 &lt;dbl&gt;, cdc28_160 &lt;dbl&gt;,</span>
<span class="co">## #   elu0 &lt;dbl&gt;, elu30 &lt;dbl&gt;, elu60 &lt;dbl&gt;, elu90 &lt;dbl&gt;, elu120 &lt;dbl&gt;,</span>
<span class="co">## #   elu150 &lt;dbl&gt;, elu180 &lt;dbl&gt;, elu210 &lt;dbl&gt;, elu240 &lt;dbl&gt;, elu270 &lt;dbl&gt;,</span>
<span class="co">## #   elu300 &lt;dbl&gt;, elu330 &lt;dbl&gt;, elu360 &lt;dbl&gt;, elu390 &lt;dbl&gt;</span></code></pre>
</div>
<div id="reshaping-data-with-tidyr" class="section level2">
<h2><span class="header-section-number">7.6</span> Reshaping data with tidyr</h2>
<p>The <code>tidyr</code> package provides functions for reshaping or tidying data frames. <code>tidyr</code> is yet another component of the tidyverse, and thus was loaded by the <code>library(tidyverse)</code>.</p>
<p>We’re going to look at two functions <code>tidyr::gather()</code> and <code>tidyr::extract()</code>, and how they can be combined with now familiar <code>dplyr</code> functions we’ve seen previously. The reading assignment for today’s class session covers a variety of other functions defined in <code>tidyr</code>.</p>
<p>The Spellman data, as I provided it to you, is in what we would call “wide” format. Each column (besides the <code>gene</code> column) corresponds to an experimental condition <em>and</em> time point. For example, “alpha0” is the alpha-factor experiment at time point 0 mins; “alpha7” is the alpha-factor experiment at time point 7 mins, etc. The cells within each column correspond to the expression of a corresponding gene (given by the first column which we renamed <code>gene</code>) in that particular experiment at that particular time point.</p>
<p>In every column (except “gene”), the cells represents the same abstract property of interest – the expression of a gene of interest in a particular experiment/time point. Our first task will be to rearrange our “wide” data frame that consists of many different columns representing gene expression into a “long” data frame with just a single column representing expression. We’ll also create a new column to keep track of which experiment and time point the measurement came from.</p>
<div id="wide-to-long-conversions-using-tidyrgather" class="section level3">
<h3><span class="header-section-number">7.6.1</span> Wide to long conversions using <code>tidyr::gather</code></h3>
<p><code>tidyr::gather()</code> takes multiple columns, and collapses them together into a smaller number of new columns. When using <code>gather()</code> you give the names of the <em>new</em> columns to create, as well as the names of any existing columns <code>gather()</code> should <em>not</em> collect together.</p>
<p>Here we want to collapse all 73 or the expression columns – “alpha0” to “elu390” – into two columns: 1) a column to represent the expt/time point of the measurement, and 2) a column to represent the corresponding expression value. The column we don’t want to touch are the <code>gene</code>, <code>std.name</code>, and <code>description</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># convert &quot;wide&quot; data to &quot;long&quot;</span>
spellman.long &lt;-
<span class="st">  </span>spellman.merged <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">gather</span>(expt.and.time, expression, <span class="op">-</span>gene, <span class="op">-</span>std.name, <span class="op">-</span>description)</code></pre>
<p>Take a moment to look at the data in the “long format”:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(spellman.long)
<span class="co">## # A tibble: 6 x 5</span>
<span class="co">##   gene   std.name description                      expt.and.time expression</span>
<span class="co">##   &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;                            &lt;chr&gt;              &lt;dbl&gt;</span>
<span class="co">## 1 YAL00… TFC3     Subunit of RNA polymerase III t… alpha0             -0.15</span>
<span class="co">## 2 YAL00… VPS8     Membrane-binding component of t… alpha0             -0.11</span>
<span class="co">## 3 YAL00… EFB1     Translation elongation factor 1… alpha0             -0.14</span>
<span class="co">## 4 YAL00… &lt;NA&gt;     Dubious open reading frame; unl… alpha0             -0.02</span>
<span class="co">## 5 YAL00… SSA1     ATPase involved in protein fold… alpha0             -0.05</span>
<span class="co">## 6 YAL00… ERP2     Member of the p24 family involv… alpha0             -0.6</span></code></pre>
<p>And compare the dimensions of the wide data to the new data:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(spellman.merged)  <span class="co"># for comparison</span>
<span class="co">## [1] 6178   76</span>
<span class="kw">dim</span>(spellman.long)
<span class="co">## [1] 450994      5</span></code></pre>
<p>As you see, we’ve gone from a data frame with 6178 rows and 76 columns (wide format), to a new data frame with 450994 rows and 5 columns (long format).</p>
</div>
<div id="extracting-information-from-combined-variables-using-tidyrextract" class="section level3">
<h3><span class="header-section-number">7.6.2</span> Extracting information from combined variables using <code>tidyr::extract</code></h3>
<p>The column <code>expt.and.time</code> violates one of our principles of tidy data: “Each variable must have its own column.”. This column conflates two different types of information – the experiment type and the time point of the measurement. Our next task is to split this information up into two new variables, which will help to facilitate downstream plotting and analysis.</p>
<p>One complicating factor is that the different experiments/time combinations have different naming conventions:</p>
<ul>
<li><p>The “alpha” and “elu” experiments are of the form “alpha0”, “alpha7”, “elu0”, “elu30”, etc. In this case, the first part of the string gives the experiment type (either alpha or elu) and the following digits give the time point.</p></li>
<li><p>In the “cdc15” and “cdc28” experiments the convention is slightly different; they are of the form “cdc15_0”, “cdc15_10”, “cdc28_0”, “cdc28_10”, etc. Here the part of the string before the underscore gives the experiment type, and the digits after the underscore give the time point.</p></li>
</ul>
<p>Because of the differences in naming conventions, we will find it easiest to break up <code>spellman.long</code> into a series of sub-data sets corresponding to each experiment type in order to extract out the experiment and time information. After processing each data subset separately, we will join the modified sub-data frames back together.</p>
</div>
<div id="subsetting-rows" class="section level3">
<h3><span class="header-section-number">7.6.3</span> Subsetting rows</h3>
<p>Let’s start by getting just the rows corresponding to the “alpha” experiment/times. Here we use <code>dplyr::filter</code> in combination with <code>stringr::str_detect</code> to get all those rows in which the <code>expt.and.time</code> variable contains the string “alpha”.</p>
<pre class="sourceCode r"><code class="sourceCode r">alpha.long &lt;-<span class="st"> </span>
<span class="st">  </span>spellman.long <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(<span class="kw">str_detect</span>(expt.and.time, <span class="st">&quot;alpha&quot;</span>))

<span class="co"># look at the new data frame</span>
<span class="kw">dim</span>(alpha.long)
<span class="co">## [1] 111204      5</span>
<span class="kw">head</span>(alpha.long, <span class="dt">n =</span> <span class="dv">10</span>)
<span class="co">## # A tibble: 10 x 5</span>
<span class="co">##    gene   std.name description                     expt.and.time expression</span>
<span class="co">##    &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;                           &lt;chr&gt;              &lt;dbl&gt;</span>
<span class="co">##  1 YAL00… TFC3     Subunit of RNA polymerase III … alpha0             -0.15</span>
<span class="co">##  2 YAL00… VPS8     Membrane-binding component of … alpha0             -0.11</span>
<span class="co">##  3 YAL00… EFB1     Translation elongation factor … alpha0             -0.14</span>
<span class="co">##  4 YAL00… &lt;NA&gt;     Dubious open reading frame; un… alpha0             -0.02</span>
<span class="co">##  5 YAL00… SSA1     ATPase involved in protein fol… alpha0             -0.05</span>
<span class="co">##  6 YAL00… ERP2     Member of the p24 family invol… alpha0             -0.6 </span>
<span class="co">##  7 YAL00… FUN14    Integral mitochondrial outer m… alpha0             -0.28</span>
<span class="co">##  8 YAL00… SPO7     Putative regulatory subunit of… alpha0             -0.03</span>
<span class="co">##  9 YAL01… MDM10    Subunit of both the ERMES and … alpha0             -0.05</span>
<span class="co">## 10 YAL01… SWC3     Protein of unknown function; c… alpha0             -0.31</span></code></pre>
</div>
<div id="splitting-columns" class="section level3">
<h3><span class="header-section-number">7.6.4</span> Splitting columns</h3>
<p>Having subsetted the data, we can now split <code>expt.and.time</code> into two new variables – <code>expt</code> and <code>time</code>. To do this we use <code>tidyr::extract</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r">alpha.long <span class="op">%&lt;&gt;%</span><span class="st"> </span>
<span class="st">  </span>tidyr<span class="op">::</span><span class="kw">extract</span>(expt.and.time,     <span class="co"># column we&#39;re extracting from</span>
                 <span class="kw">c</span>(<span class="st">&quot;expt&quot;</span>, <span class="st">&quot;time&quot;</span>), <span class="co"># new columns we&#39;re creating</span>
                 <span class="dt">regex=</span><span class="st">&quot;(alpha)([[:digit:]]+)&quot;</span>,  <span class="co"># regexp (see below)</span>
                 <span class="dt">convert=</span><span class="ot">TRUE</span>)      <span class="co"># automatically convert column types</span>

<span class="co"># </span><span class="al">NOTE</span><span class="co">: I&#39;m being explict about saying tidyr::extract because the</span>
<span class="co"># magrittr package defines a different extract function</span></code></pre>
<p>Let’s take a moment to look at the <code>regex</code> argument to <code>extract</code> – <code>regex="(alpha)([[:digit:]]+)"</code>. The regex is specified as a character string. Each part we want to match <em>and</em> extract is surround by parentheses. In this case we have two sets of parentheses corresponding to the two matches we want to make. The first part of the regex is <code>(alpha)</code>; here we’re looking to make an exact match to the string “alpha”. The second part of the regex reads <code>([[:digit:]]+)</code>. <code>[[:digit:]]</code> indicates we’re looking for a numeric digit. The <code>+</code> after <code>[[:digit:]]</code> indicates that we want to match <em>one or more</em> digits (i.e. to get a match we need to find at least one digit, but more than one digit should also be a match).</p>
<p>Let’s take a look at the new version of <code>alpha.long</code> following application of <code>extract</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(alpha.long, <span class="dt">n =</span> <span class="dv">10</span>)
<span class="co">## # A tibble: 10 x 6</span>
<span class="co">##    gene   std.name description                       expt   time expression</span>
<span class="co">##    &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;                             &lt;chr&gt; &lt;int&gt;      &lt;dbl&gt;</span>
<span class="co">##  1 YAL00… TFC3     Subunit of RNA polymerase III tr… alpha     0      -0.15</span>
<span class="co">##  2 YAL00… VPS8     Membrane-binding component of th… alpha     0      -0.11</span>
<span class="co">##  3 YAL00… EFB1     Translation elongation factor 1 … alpha     0      -0.14</span>
<span class="co">##  4 YAL00… &lt;NA&gt;     Dubious open reading frame; unli… alpha     0      -0.02</span>
<span class="co">##  5 YAL00… SSA1     ATPase involved in protein foldi… alpha     0      -0.05</span>
<span class="co">##  6 YAL00… ERP2     Member of the p24 family involve… alpha     0      -0.6 </span>
<span class="co">##  7 YAL00… FUN14    Integral mitochondrial outer mem… alpha     0      -0.28</span>
<span class="co">##  8 YAL00… SPO7     Putative regulatory subunit of N… alpha     0      -0.03</span>
<span class="co">##  9 YAL01… MDM10    Subunit of both the ERMES and th… alpha     0      -0.05</span>
<span class="co">## 10 YAL01… SWC3     Protein of unknown function; com… alpha     0      -0.31</span></code></pre>
<p>Notice our two new variables, both of which have appropriate types!</p>
<p>A data frame for the elutriation data can be created similarly:</p>
<pre class="sourceCode r"><code class="sourceCode r">elu.long &lt;-
<span class="st">  </span>spellman.long <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(<span class="kw">str_detect</span>(expt.and.time, <span class="st">&quot;elu&quot;</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span>tidyr<span class="op">::</span><span class="kw">extract</span>(expt.and.time,     <span class="co"># column we&#39;re extracting from</span>
                 <span class="kw">c</span>(<span class="st">&quot;expt&quot;</span>, <span class="st">&quot;time&quot;</span>), <span class="co"># new columns we&#39;re creating</span>
                 <span class="dt">regex=</span><span class="st">&quot;(elu)([[:digit:]]+)&quot;</span>,  <span class="co"># regexp (see below)</span>
                 <span class="dt">convert=</span><span class="ot">TRUE</span>)      <span class="co"># automatically convert column types</span></code></pre>
<div id="a-fancier-regex-for-the-cdc-experiments" class="section level4">
<h4><span class="header-section-number">7.6.4.1</span> A fancier regex for the cdc experiments</h4>
<p>Now let’s process the cdc experiments (cdc15 and cdc28). As before we extract the corresponding rows of the data frame using <code>filter</code> and <code>str_detect</code>. We then split <code>expt.and.time</code> using <code>tidyr::extract</code>. In this case we carry out the two steps in a single code block using pipes:</p>
<pre class="sourceCode r"><code class="sourceCode r">cdc.long &lt;-<span class="st"> </span>
<span class="st">  </span>spellman.long <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># both cdc15 and cdc28 contain &quot;cdc&quot; as a sub-string</span>
<span class="st">  </span><span class="kw">filter</span>(<span class="kw">str_detect</span>(expt.and.time, <span class="st">&quot;cdc&quot;</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span>tidyr<span class="op">::</span><span class="kw">extract</span>(expt.and.time, 
                 <span class="kw">c</span>(<span class="st">&quot;expt&quot;</span>, <span class="st">&quot;time&quot;</span>), 
                 <span class="dt">regex=</span><span class="st">&quot;(cdc15|cdc28)_([[:digit:]]+)&quot;</span>, <span class="co"># note the fancier regex</span>
                 <span class="dt">convert=</span><span class="ot">TRUE</span>)</code></pre>
<p>The regex – <code>"(cdc15|cdc28)_([[:digit:]]+)"</code> – is slightly fancier in this example. As before there are two parts we’re extracting: <code>(cdc15|cdc28)</code> and <code>([[:digit:]]+)</code>. The first parenthesized regexp is an “OR” – i.e. match “cdc15” <em>or</em> “cdc28”. The second parenthesized regexp is the same as we saw previously. Separating the two parenthesized regexps is an underscore (<code>_</code>). The underscore isn’t parenthesized because we only want to use it to make a match <em>not</em> to extract the corresponding match.</p>
</div>
</div>
<div id="combining-data-frames" class="section level3">
<h3><span class="header-section-number">7.6.5</span> Combining data frames</h3>
<p>If you have two or more data frames with identical columns, the rows of the data frames can be combined into a single data frame using <code>rbind</code> (defined in the <code>base</code> package). For example, to reassemble the <code>alpha.long</code>, <code>elu.long</code>, and <code>cdc.long</code> data frames into a single data frame we do:</p>
<pre class="sourceCode r"><code class="sourceCode r">spellman.final &lt;-<span class="st"> </span><span class="kw">rbind</span>(alpha.long, elu.long, cdc.long)

<span class="co"># check the dimensions of the new data frame</span>
<span class="kw">dim</span>(spellman.final)
<span class="co">## [1] 450994      6</span></code></pre>
</div>
<div id="sorting-data-frame-rows" class="section level3">
<h3><span class="header-section-number">7.6.6</span> Sorting data frame rows</h3>
<p>Currently the <code>spellman.final</code> data frame is sorted by time point and experiment.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(spellman.final, <span class="dt">n =</span> <span class="dv">10</span>)
<span class="co">## # A tibble: 10 x 6</span>
<span class="co">##    gene   std.name description                       expt   time expression</span>
<span class="co">##    &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;                             &lt;chr&gt; &lt;int&gt;      &lt;dbl&gt;</span>
<span class="co">##  1 YAL00… TFC3     Subunit of RNA polymerase III tr… alpha     0      -0.15</span>
<span class="co">##  2 YAL00… VPS8     Membrane-binding component of th… alpha     0      -0.11</span>
<span class="co">##  3 YAL00… EFB1     Translation elongation factor 1 … alpha     0      -0.14</span>
<span class="co">##  4 YAL00… &lt;NA&gt;     Dubious open reading frame; unli… alpha     0      -0.02</span>
<span class="co">##  5 YAL00… SSA1     ATPase involved in protein foldi… alpha     0      -0.05</span>
<span class="co">##  6 YAL00… ERP2     Member of the p24 family involve… alpha     0      -0.6 </span>
<span class="co">##  7 YAL00… FUN14    Integral mitochondrial outer mem… alpha     0      -0.28</span>
<span class="co">##  8 YAL00… SPO7     Putative regulatory subunit of N… alpha     0      -0.03</span>
<span class="co">##  9 YAL01… MDM10    Subunit of both the ERMES and th… alpha     0      -0.05</span>
<span class="co">## 10 YAL01… SWC3     Protein of unknown function; com… alpha     0      -0.31</span></code></pre>
<p>It might be useful instead to sort by gene and experiment. To do this we can use <code>dplyr::arrange</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r">spellman.final <span class="op">%&lt;&gt;%</span>
<span class="st">  </span><span class="kw">arrange</span>(gene, expt)

<span class="co"># look again at the rearranged data</span>
<span class="kw">head</span>(spellman.final, <span class="dt">n =</span> <span class="dv">10</span>)
<span class="co">## # A tibble: 10 x 6</span>
<span class="co">##    gene   std.name description                       expt   time expression</span>
<span class="co">##    &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;                             &lt;chr&gt; &lt;int&gt;      &lt;dbl&gt;</span>
<span class="co">##  1 YAL00… TFC3     Subunit of RNA polymerase III tr… alpha     0      -0.15</span>
<span class="co">##  2 YAL00… TFC3     Subunit of RNA polymerase III tr… alpha     7      -0.15</span>
<span class="co">##  3 YAL00… TFC3     Subunit of RNA polymerase III tr… alpha    14      -0.21</span>
<span class="co">##  4 YAL00… TFC3     Subunit of RNA polymerase III tr… alpha    21       0.17</span>
<span class="co">##  5 YAL00… TFC3     Subunit of RNA polymerase III tr… alpha    28      -0.42</span>
<span class="co">##  6 YAL00… TFC3     Subunit of RNA polymerase III tr… alpha    35      -0.44</span>
<span class="co">##  7 YAL00… TFC3     Subunit of RNA polymerase III tr… alpha    42      -0.15</span>
<span class="co">##  8 YAL00… TFC3     Subunit of RNA polymerase III tr… alpha    49       0.24</span>
<span class="co">##  9 YAL00… TFC3     Subunit of RNA polymerase III tr… alpha    56      -0.1 </span>
<span class="co">## 10 YAL00… TFC3     Subunit of RNA polymerase III tr… alpha    63      NA</span></code></pre>
</div>
</div>
<div id="using-your-tidy-data" class="section level2">
<h2><span class="header-section-number">7.7</span> Using your tidy data</h2>
<p>Whew – that was a fair amount of work to tidy our data! But having done so we can now carry out a wide variety of very powerful analyses.</p>
<div id="visualizing-gene-expression-time-series" class="section level3">
<h3><span class="header-section-number">7.7.1</span> Visualizing gene expression time series</h3>
<p>Let’s start by walking through a series of visualizations of gene expression time series. Each plot will show the expression of one or more genes, at different time points, in one or more experimental conditions. Our initial visualizations exploit the “long” versions of the tidy data.</p>
<p>First a single gene in a single experimental condition:</p>
<pre class="sourceCode r"><code class="sourceCode r">spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(expt <span class="op">==</span><span class="st"> &quot;alpha&quot;</span>, gene <span class="op">==</span><span class="st"> &quot;YAL022C&quot;</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Expression of YAL022C&quot;</span>)</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-234-1.png" width="576" /></p>
<p>We can easily modify the above code block to visualize the expression of multiple genes of interest:</p>
<pre class="sourceCode r"><code class="sourceCode r">genes.of.interest &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;YAL022C&quot;</span>, <span class="st">&quot;YAR018C&quot;</span>, <span class="st">&quot;YGR188C&quot;</span>)

spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>genes.of.interest, expt <span class="op">==</span><span class="st"> &quot;alpha&quot;</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression, <span class="dt">color =</span> gene)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Normalized expression&quot;</span>,
         <span class="dt">title =</span> <span class="st">&quot;Expression of multiple genes</span><span class="ch">\n</span><span class="st">following synchronization by alpha factor&quot;</span>)</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-235-1.png" width="576" /></p>
<p>By employing <code>facet_wrap()</code> we can visualize the relationship between this set of genes in each of the experiment types:</p>
<pre class="sourceCode r"><code class="sourceCode r">spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>genes.of.interest) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression, <span class="dt">color =</span> gene)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>expt) <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Normalized expression&quot;</span>,
         <span class="dt">title =</span> <span class="st">&quot;Expression of Multiple Genes</span><span class="ch">\n</span><span class="st">Across experiments&quot;</span>) </code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-236-1.png" width="576" /></p>
<p>The different experimental treatments were carried out for varying lengths of time due to the differences in their physiological effects. Plotting them all on the same time scale can obscure that patterns of oscillation we might be interested in, so let’s modify our code block so that plots that share the same y-axis, but have differently scaled x-axes.</p>
<pre class="sourceCode r"><code class="sourceCode r">spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>genes.of.interest) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression, <span class="dt">color =</span> gene)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>expt, <span class="dt">scales =</span> <span class="st">&quot;free_x&quot;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Normalized expression&quot;</span>,
         <span class="dt">title =</span> <span class="st">&quot;Expression of Multiple Genes</span><span class="ch">\n</span><span class="st">Across experiments&quot;</span>) </code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-237-1.png" width="576" /></p>
</div>
<div id="finding-the-most-variable-genes" class="section level3">
<h3><span class="header-section-number">7.7.2</span> Finding the most variable genes</h3>
<p>When dealing with vary large data sets, one ad hoc filtering criteria that is often employed is to focus on those variables that exhibit that greatest variation (variation is measure of the spread of data; we will give a precise definition in a later lecture). To do this, we first need to order our variables (genes) by their variation. Let’s see how we can accomplish this using our long data frame:</p>
<pre class="sourceCode r"><code class="sourceCode r">by.variance &lt;-
<span class="st">  </span>spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(gene) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">expression.var =</span> <span class="kw">var</span>(expression, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">arrange</span>(<span class="kw">desc</span>(expression.var))

<span class="kw">head</span>(by.variance)
<span class="co">## # A tibble: 6 x 2</span>
<span class="co">##   gene    expression.var</span>
<span class="co">##   &lt;chr&gt;            &lt;dbl&gt;</span>
<span class="co">## 1 YLR286C          2.157</span>
<span class="co">## 2 YNR067C          1.733</span>
<span class="co">## 3 YNL327W          1.653</span>
<span class="co">## 4 YGL028C          1.571</span>
<span class="co">## 5 YHL028W          1.521</span>
<span class="co">## 6 YKL164C          1.515</span></code></pre>
<p>The code above calculates the variance of each gene but ignores the fact that we have different experimental conditions. To take into account the experimental design of the data at hand, let’s calculate the average variance across the experimental conditions:</p>
<pre class="sourceCode r"><code class="sourceCode r">by.avg.variance &lt;-
<span class="st">  </span>spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(gene, expt) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">expression.var =</span> <span class="kw">var</span>(expression, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(gene) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">avg.expression.var =</span> <span class="kw">mean</span>(expression.var)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">arrange</span>(<span class="kw">desc</span>(avg.expression.var))

<span class="kw">head</span>(by.avg.variance)
<span class="co">## # A tibble: 6 x 2</span>
<span class="co">##   gene    avg.expression.var</span>
<span class="co">##   &lt;chr&gt;                &lt;dbl&gt;</span>
<span class="co">## 1 YFR014C              3.579</span>
<span class="co">## 2 YFR053C              2.377</span>
<span class="co">## 3 YBL032W              2.299</span>
<span class="co">## 4 YDR274C              2.173</span>
<span class="co">## 5 YLR286C              2.128</span>
<span class="co">## 6 YMR206W              1.937</span></code></pre>
<p>Based on the average experession variance across experimental conditions, let’s get the names of the 1000 most variable genes:</p>
<pre class="sourceCode r"><code class="sourceCode r">top.genes<span class="fl">.1</span>k &lt;-<span class="st"> </span>by.avg.variance[<span class="dv">1</span><span class="op">:</span><span class="dv">1000</span>,]<span class="op">$</span>gene

<span class="kw">head</span>(top.genes<span class="fl">.1</span>k)
<span class="co">## [1] &quot;YFR014C&quot; &quot;YFR053C&quot; &quot;YBL032W&quot; &quot;YDR274C&quot; &quot;YLR286C&quot; &quot;YMR206W&quot;</span></code></pre>
</div>
<div id="heat-maps" class="section level3">
<h3><span class="header-section-number">7.7.3</span> Heat maps</h3>
<p>In our prior visualizations we’ve used line plots to depict how gene expression changes over time. For example here are line plots for 15 genes in the data set, in the cdc28 experimental conditions:</p>
<pre class="sourceCode r"><code class="sourceCode r">genes.of.interest &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;YHR084W&quot;</span>, <span class="st">&quot;YBR083W&quot;</span>, <span class="st">&quot;YPL049C&quot;</span>, <span class="st">&quot;YDR480W&quot;</span>,
                       <span class="st">&quot;YGR040W&quot;</span>, <span class="st">&quot;YLR229C&quot;</span>, <span class="st">&quot;YDL159W&quot;</span>, <span class="st">&quot;YBL016W&quot;</span>,
                       <span class="st">&quot;YDR103W&quot;</span>, <span class="st">&quot;YJL157C&quot;</span>, <span class="st">&quot;YNL271C&quot;</span>, <span class="st">&quot;YDR461W&quot;</span>,
                       <span class="st">&quot;YHL007C&quot;</span>, <span class="st">&quot;YHR005C&quot;</span>, <span class="st">&quot;YJR086W&quot;</span>)

spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(expt <span class="op">==</span><span class="st"> &quot;cdc28&quot;</span>, gene <span class="op">%in%</span><span class="st"> </span>genes.of.interest) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression, <span class="dt">color=</span>gene)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (min)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Expression&quot;</span>)</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-241-1.png" width="576" /></p>
<p>Even with just 10 overlapping line plots, this figure is quite busy and it’s hard to make out the individual behavior of each gene.</p>
<p>An alternative approach to depicting such data is a “heat map” which depicts the same information in a grid like form, with the expression values indicated by color. Heat maps are good for depicting large amounts of data and providing a coarse “10,000 foot view”. We can create a heat map using <code>geom_tile</code> as follows:</p>
<pre class="sourceCode r"><code class="sourceCode r">spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(expt <span class="op">==</span><span class="st"> &quot;cdc28&quot;</span>, gene <span class="op">%in%</span><span class="st"> </span>genes.of.interest) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> gene)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_tile</span>(<span class="kw">aes</span>(<span class="dt">fill =</span> expression)) <span class="op">+</span>
<span class="st">    </span><span class="kw">xlab</span>(<span class="st">&quot;Time (mins)&quot;</span>)</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-242-1.png" width="576" /></p>
<p>This figure represents the same information as our line plot, but now there is row for each gene, and the expression of that gene at a given time point is represented by color (scale given on the right). Missing data is shown as gray boxes. Unfortunately, the default color scale used by ggplot is a very subtle gradient from light to dark blue. This make it hard to distinguish patterns of change. Let’s now see how we can improve that.</p>
<div id="better-color-schemes-with-rcolorbrewer" class="section level4">
<h4><span class="header-section-number">7.7.3.1</span> Better color schemes with RColorBrewer</h4>
<p>The <code>RColorBrewer</code> packages provides nice color schemes that are useful for creating heat maps. RColorBrewer defines a set of color palettes that have been optimized for color discrimination, many of which are color blind friendly, etc. Install the RColorBrewer package using the command line or the RStudio GUI.</p>
<p>Once you’ve installed the RColorBrewer package you can see the available color palettes as so:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RColorBrewer)

<span class="co"># show representations of the palettes</span>
<span class="kw">par</span>(<span class="dt">cex =</span> <span class="fl">0.5</span>) <span class="co"># reduce size of text in the following plot</span>
<span class="kw">display.brewer.all</span>()  </code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-243-1.png" width="576" /></p>
<p>We’ll use the Red-to-Blue (“RdBu”) color scheme defined in RColorBrewer, however we’ll reverse the scheme so blues represent low expression and reds represent high expression. We’ll divide the range of color values into 9 discrete bins.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># displays the RdBu color scheme divided into a palette of 9 colors</span>
<span class="kw">display.brewer.pal</span>(<span class="dv">9</span>, <span class="st">&quot;RdBu&quot;</span>)  </code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-244-1.png" width="576" /></p>
<pre class="sourceCode r"><code class="sourceCode r">
<span class="co"># assign the reversed (blue to red) RdBu palette</span>
color.scheme &lt;-<span class="st"> </span><span class="kw">rev</span>(<span class="kw">brewer.pal</span>(<span class="dv">9</span>,<span class="st">&quot;RdBu&quot;</span>))</code></pre>
<p>Now let’s regenerate the heat map we created previously with this new color scheme. To do this we specify a gradient color scale using the <code>scale_fill_gradientn()</code> function from ggplot. In addition to specifying the color scale, we also constrain the limits of the scale to insure it’s symmetric about zero.</p>
<pre class="sourceCode r"><code class="sourceCode r">spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(expt <span class="op">==</span><span class="st"> &quot;cdc28&quot;</span>, gene <span class="op">%in%</span><span class="st"> </span>genes.of.interest) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> gene)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_tile</span>(<span class="kw">aes</span>(<span class="dt">fill =</span> expression)) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_fill_gradientn</span>(<span class="dt">colors=</span>color.scheme,
                         <span class="dt">limits =</span> <span class="kw">c</span>(<span class="op">-</span><span class="fl">2.5</span>, <span class="fl">2.5</span>)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">xlab</span>(<span class="st">&quot;Time (mins)&quot;</span>)</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-245-1.png" width="576" /></p>
</div>
<div id="looking-for-patterns-using-sorted-data-and-heat-maps" class="section level4">
<h4><span class="header-section-number">7.7.3.2</span> Looking for patterns using sorted data and heat maps</h4>
<p>The real power of heat maps becomes apparent when you you rearrange the rows of the heat map to emphasize patterns of interest.</p>
<p>For example, let’s create a heat map in which we sort genes by the time of their maximal expression. This is one way to identify genes that reach their peak expression at similar times, which is one criteria one might use to identify genes acting in concert.</p>
<p>For simplicities sake we will restrict our attention to the cdc28 experiment, and only consider the 1000 most variables genes with no more than one missing observation in this experimental condition.</p>
<pre class="sourceCode r"><code class="sourceCode r">cdc28 &lt;-
<span class="st">  </span>spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(expt <span class="op">==</span><span class="st"> &quot;cdc28&quot;</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(gene) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(<span class="kw">sum</span>(<span class="kw">is.na</span>(expression)) <span class="op">&lt;=</span><span class="st"> </span><span class="dv">1</span>) <span class="op">%&gt;%</span>
<span class="st">  </span>ungroup   <span class="co"># removes grouping information from data frame</span>

top1k.genes &lt;-
<span class="st">  </span>cdc28 <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(gene) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">expression.var =</span> <span class="kw">var</span>(expression, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">arrange</span>(<span class="kw">desc</span>(expression.var)) <span class="op">%$%</span>
<span class="st">  </span>gene[<span class="dv">1</span><span class="op">:</span><span class="dv">1000</span>]
    
top1k.cdc28 &lt;-<span class="st"> </span>
<span class="st">  </span>cdc28 <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>top1k.genes)</code></pre>
<p>To find the time of maximum expression we’ll employ the function <code>which.max</code> (<code>which.min</code>), which finds the index of the maximum (minimum) element of a vector. For example to find the index of the maximum expression measurement for YAR018C we could do:</p>
<pre class="sourceCode r"><code class="sourceCode r">top1k.cdc28 <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">==</span><span class="st"> &quot;YAR018C&quot;</span>) <span class="op">%$%</span><span class="st">  </span><span class="co"># note the exposition pipe operator!</span>
<span class="st">  </span><span class="kw">which.max</span>(expression)
<span class="co">## [1] 8</span></code></pre>
<p>From the code above we find that the index of the observation at which YAR018C is maximal at 8. To get the corresponding time point we can do something like this:</p>
<pre class="sourceCode r"><code class="sourceCode r">top1k.cdc28 <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">==</span><span class="st"> &quot;YAR018C&quot;</span>) <span class="op">%$%</span><span class="st">  </span><span class="co"># again note the exposition pipe operator!</span>
<span class="st">  </span>time[<span class="kw">which.max</span>(expression)]
<span class="co">## [1] 70</span></code></pre>
<p>Thus YAR018C expression peaks at 70 minutes in the cdc28 experiment.</p>
<p>To find the index of maximal expression of all genes we can apply the <code>dplyr::group_by()</code> and <code>dplyr::summarize()</code> functions</p>
<pre class="sourceCode r"><code class="sourceCode r">peak.expression.cdc28 &lt;-<span class="st"> </span>
<span class="st">  </span>top1k.cdc28 <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">group_by</span>(gene) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">summarise</span>(<span class="dt">peak =</span> <span class="kw">which.max</span>(expression))

<span class="kw">head</span>(peak.expression.cdc28)
<span class="co">## # A tibble: 6 x 2</span>
<span class="co">##   gene       peak</span>
<span class="co">##   &lt;chr&gt;     &lt;int&gt;</span>
<span class="co">## 1 YAL003W      10</span>
<span class="co">## 2 YAL005C       2</span>
<span class="co">## 3 YAL022C      17</span>
<span class="co">## 4 YAL028W       5</span>
<span class="co">## 5 YAL035C-A    12</span>
<span class="co">## 6 YAL038W      15</span></code></pre>
<p>Let’s sort the order of genes by their peak expression:</p>
<pre class="sourceCode r"><code class="sourceCode r">peak.expression.cdc28 <span class="op">%&lt;&gt;%</span><span class="st"> </span><span class="kw">arrange</span>(peak)</code></pre>
<p>We can then generate a heatmap where we sort the rows (genes) of the heatmap by their time of peak expression. We introduce a new geom – <code>geom_raster</code> – which is like <code>geom_tile</code> but better suited for large data (hundreds to thousands of rows)</p>
<p>The explicit sorting of the data by peak expression is carried out in the call to <code>scale_y_discrete()</code> where the limits (and order) of this axis are set with the <code>limits</code> argument (see <code>scale_y_discrete</code> and <code>discrete_scale</code> in the ggplot2 docs).</p>
<pre class="sourceCode r"><code class="sourceCode r">
<span class="co"># we reverse the ordering because geom_raster (and geom_tile)</span>
<span class="co"># draw from the bottom row up, whereas we want to depict the</span>
<span class="co"># earliest peaking genes at the top of the figure</span>
gene.ordering &lt;-<span class="st"> </span><span class="kw">rev</span>(peak.expression.cdc28<span class="op">$</span>gene)

top1k.cdc28 <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> gene)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_raster</span>(<span class="kw">aes</span>(<span class="dt">fill =</span> expression)) <span class="op">+</span><span class="st">  </span><span class="co"># </span>
<span class="st">  </span><span class="kw">scale_fill_gradientn</span>(<span class="dt">limits=</span><span class="kw">c</span>(<span class="op">-</span><span class="fl">2.5</span>, <span class="fl">2.5</span>),
                       <span class="dt">colors=</span>color.scheme) <span class="op">+</span>
<span class="st">  </span><span class="kw">scale_y_discrete</span>(<span class="dt">limits=</span>gene.ordering) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Genes&quot;</span>, 
       <span class="dt">title =</span> <span class="st">&quot;1000 most variable genes&quot;</span>,
       <span class="dt">subtitle =</span> <span class="st">&quot;Sorted by time of peak expression&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="co"># the following line suppresses tick and labels on y-axis</span>
<span class="st">  </span><span class="kw">theme</span>(<span class="dt">axis.text.y =</span> <span class="kw">element_blank</span>(), <span class="dt">axis.ticks.y =</span> <span class="kw">element_blank</span>())</code></pre>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-251"></span>
<img src="bio723-book_files/figure-html/unnamed-chunk-251-1.png" alt="A Heatmap showing genes in the cdc28 experiment, sorted by peak expression" width="576" />
<p class="caption">
Figure 7.1: A Heatmap showing genes in the cdc28 experiment, sorted by peak expression
</p>
</div>
<p>The brightest red regions in each row of the heat map correspond to the times of peak expression, and the sorting of the rows helps to highlight those gene whose peak expression times are similar.</p>
</div>
</div>
</div>
<div id="long-to-wide-conversion-using-tidyrspread" class="section level2">
<h2><span class="header-section-number">7.8</span> Long-to-wide conversion using <code>tidyr::spread</code></h2>
<p>Our long data frame consists of four variables – <code>gene</code>, <code>expt</code>, <code>time</code>, and <code>expression</code>. This made it easy to create visualizations and summaries where time and expression were the primaries variables of interest, and gene and experiment type were categories we could condition on.</p>
<p>To facilitate analyses that emphasize comparison between genes, we want to create a new data frame in which each gene is itself treated as a variable of interest along with time, and experiment type remains a categorical variable. In this new data frame rather than just four columns in our data frame, we’ll have several thousand columns – one for each gene.</p>
<p>To accomplish this reshaping of data, we’ll use the function <code>tidyr::spread()</code>. <code>tidyr::spread()</code> is the inverse of <code>tidyr::gather()</code>. <code>gather()</code> took multiple columns and collapsed them together into a smaller number of new columns. The <code>tidyr</code> documentation calls this “collapsing into key-value pairs”. By contrast, <code>spread()</code> creates new columns by spreading “key-value pairs” (a column representing the “keys” and a column reprsenting the “values”) into multiple columns.</p>
<p>Here let’s use <code>spread()</code> to use the gene names (the “key”) and expression measures (the “values”) to create a new data frame where the genes are the primary variables (columns) of the data.</p>
<pre class="sourceCode r"><code class="sourceCode r">spellman.wide &lt;-
<span class="st">  </span>spellman.final <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>std.name, <span class="op">-</span>description) <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># drop unneeded columns</span>
<span class="st">  </span><span class="kw">spread</span>(gene, expression)</code></pre>
<p>Now let’s examine the dimensions of this wide version of the data:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(spellman.wide)
<span class="co">## [1]   73 6180</span></code></pre>
<p>And here’s a visual view of the first few rows and columns of the wide data:</p>
<pre class="sourceCode r"><code class="sourceCode r">spellman.wide[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">8</span>]
<span class="co">## # A tibble: 5 x 8</span>
<span class="co">##   expt   time YAL001C YAL002W YAL003W YAL004W YAL005C YAL007C</span>
<span class="co">##   &lt;chr&gt; &lt;int&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">## 1 alpha     0   -0.15   -0.11   -0.14   -0.02   -0.05   -0.6 </span>
<span class="co">## 2 alpha     7   -0.15    0.1    -0.71   -0.48   -0.53   -0.45</span>
<span class="co">## 3 alpha    14   -0.21    0.01    0.1    -0.11   -0.47   -0.13</span>
<span class="co">## 4 alpha    21    0.17    0.06   -0.32    0.12   -0.06    0.35</span>
<span class="co">## 5 alpha    28   -0.42    0.04   -0.4    -0.03    0.11   -0.01</span></code></pre>
<p>From this view we infer that the rows of the data set represent the various combination of experimental condition and time points, and the columns represents the 6178 genes in the data set plus the two columns for <code>expt</code> and <code>time</code>.</p>
</div>
<div id="exploring-bivariate-relationships-using-wide-data" class="section level2">
<h2><span class="header-section-number">7.9</span> Exploring bivariate relationships using “wide” data</h2>
<p>The “long” version of our data frame proved useful for exploring how gene expression changed over time. By contrast, our “wide” data frame is more convient for exploring how pairs of genes covary together. For example, we can generate bivariate scatter plots depicting the relationship between two genes of interest:</p>
<pre class="sourceCode r"><code class="sourceCode r">two.gene.plot &lt;-
<span class="st">  </span>spellman.wide <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">is.na</span>(YAR018C) <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(YAL022C)) <span class="op">%&gt;%</span><span class="st">  </span><span class="co"># remove NAs</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> YAR018C, <span class="dt">y =</span> YAL022C)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">theme</span>(<span class="dt">aspect.ratio =</span> <span class="dv">1</span>)

two.gene.plot</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-255-1.png" width="576" /></p>
<p>From the scatter plot we infer that the two genes are “positively correlated” with each other, meaning that high values of one tend to be associated with high values of the other (and the same for low values).</p>
<p>We can easily extend this visualization to facet the plot based on the experimental conditions:</p>
<pre class="sourceCode r"><code class="sourceCode r">two.gene.plot <span class="op">+</span><span class="st"> </span><span class="kw">facet_wrap</span>(<span class="op">~</span>expt, <span class="dt">nrow =</span> <span class="dv">2</span>, <span class="dt">ncol =</span> <span class="dv">2</span>) </code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-256-1.png" width="576" /></p>
<p>A statistic we use to measure the degree of association between pairs of continuous variables is called the “correlation coefficient”. Briefly, correlation is a measure of linear association between a pair of variables, and ranges from -1 to 1. A value near zero indicates the variables are uncorrelated (no linear association), while values approaching +1 indicate a strong positive association (the variables tend to get bigger or smaller together) while values near -1 indicate strong negative association (when one variable is larger, the other tends to be small).</p>
<p>Let’s calculate the correlation between YAR018C and YAL022C:</p>
<pre class="sourceCode r"><code class="sourceCode r">spellman.wide <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">is.na</span>(YAR018C) <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(YAL022C)) <span class="op">%&gt;%</span><span class="st">  </span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">cor =</span> <span class="kw">cor</span>(YAR018C, YAL022C))
<span class="co">## # A tibble: 1 x 1</span>
<span class="co">##      cor</span>
<span class="co">##    &lt;dbl&gt;</span>
<span class="co">## 1 0.6915</span></code></pre>
<p>The value of the correlation coefficient for YAR018C and YAL022C, ~0.69, indicates a fairly strong association between the two genes.</p>
<p>As we did for our visualization, we can also calculate the correlation coefficients for the two genes under each experimental condition:</p>
<pre class="sourceCode r"><code class="sourceCode r">spellman.wide <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">is.na</span>(YAR018C) <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(YAL022C)) <span class="op">%&gt;%</span><span class="st">  </span>
<span class="st">  </span><span class="kw">group_by</span>(expt) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">cor =</span> <span class="kw">cor</span>(YAR018C, YAL022C))
<span class="co">## # A tibble: 4 x 2</span>
<span class="co">##   expt     cor</span>
<span class="co">##   &lt;chr&gt;  &lt;dbl&gt;</span>
<span class="co">## 1 alpha 0.6341</span>
<span class="co">## 2 cdc15 0.5751</span>
<span class="co">## 3 cdc28 0.8474</span>
<span class="co">## 4 elu   0.7867</span></code></pre>
<p>This table suggests that the the strength of correlation between YAR018C and YAL022C may depend on the experimental conditions, with the highest correlations evident in the cdc28 and elu experiments.</p>
<div id="large-scale-patterns-of-correlations" class="section level3">
<h3><span class="header-section-number">7.9.1</span> Large scale patterns of correlations</h3>
<p>Now we’ll move from considering the correlation between two specific genes to looking at the correlation between many pairs of genes. As we did in the previous section, we’ll focus specifically onthe 1000 most variable genes in the cdc28 experiment.</p>
<p>First we filter our wide data set to only consider the cdc28 experiment and those genes in the top 1000 most variable genes in cdc28:</p>
<pre class="sourceCode r"><code class="sourceCode r">top1k.cdc28.wide &lt;-<span class="st"> </span>
<span class="st">  </span>top1k.cdc28 <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>std.name, <span class="op">-</span>description) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">spread</span>(gene, expression)</code></pre>
<p>With this restricted data set, we can then calculate the correlations between every pair of genes as follows:</p>
<pre class="sourceCode r"><code class="sourceCode r">cdc28.correlations &lt;-<span class="st"> </span>
<span class="st">  </span>top1k.cdc28.wide <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>expt, <span class="op">-</span>time)  <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># drop expt and time</span>
<span class="st">  </span><span class="kw">cor</span>(<span class="dt">use =</span> <span class="st">&quot;pairwise.complete.obs&quot;</span>)</code></pre>
<p>The argument <code>use = "pairwise.complete.obs"</code> tells the correlation function that for each pair of genes to use only the pariwse where there is a value for both genes (i.e. neither one can be NA).</p>
<p>Given <span class="math inline">\(n\)</span> genes, there are <span class="math inline">\(n \times n\)</span> pairs of correlations, as seen by the dimensions of the correlation matrix.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(cdc28.correlations)
<span class="co">## [1] 1000 1000</span></code></pre>
<p>To get the correlations with a gene of interest, we can index with the gene name on the rows of the correlation matrix. For example, to get the correlations between the gene YAR018C and the first 10 genes in the top 1000 set:</p>
<pre class="sourceCode r"><code class="sourceCode r">cdc28.correlations[<span class="st">&quot;YAR018C&quot;</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]  
<span class="co">##     YAL003W     YAL005C     YAL022C     YAL028W   YAL035C-A     YAL038W </span>
<span class="co">##  0.07100626 -0.53315493  0.84741624  0.33379901 -0.22316755 -0.03984599 </span>
<span class="co">##     YAL044C     YAL048C     YAL060W     YAL062W </span>
<span class="co">##  0.32253692  0.12220221  0.49445700 -0.60972118</span></code></pre>
<p>In the next statement we extract the names of the genes that have correlations with YAR018C greater than 0.6. First we test genes to see if they have a correlation with YAR018C greater than 0.6, which returns a vector of TRUE or FALSE values. This vector of Boolean values is than used to index into the rownames of the correlation matrix, pulling out the gene names where the statement was true.</p>
<pre class="sourceCode r"><code class="sourceCode r">pos.corr.YAR018C &lt;-<span class="st"> </span><span class="kw">rownames</span>(cdc28.correlations)[cdc28.correlations[<span class="st">&quot;YAR018C&quot;</span>,] <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.6</span>]
<span class="kw">length</span>(pos.corr.YAR018C)
<span class="co">## [1] 65</span></code></pre>
<p>We then return to our long data to show this set of genes that are strongly positively correlated with YAR018C.</p>
<pre class="sourceCode r"><code class="sourceCode r">top1k.cdc28 <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>pos.corr.YAR018C) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression, <span class="dt">group =</span> gene)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_line</span>(<span class="dt">alpha =</span> <span class="fl">0.33</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&quot;none&quot;</span>)</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-264-1.png" width="576" /></p>
<p>As is expected, genes with strong positive correlations with YAR018C. show similar temporal patterns with YAR018C.</p>
<p>We can similarly filter for genes that have negative correlations with YAR018C.</p>
<pre class="sourceCode r"><code class="sourceCode r">neg.corr.YAR018C &lt;-<span class="st"> </span><span class="kw">colnames</span>(cdc28.correlations)[cdc28.correlations[<span class="st">&quot;YAR018C&quot;</span>,] <span class="op">&lt;=</span><span class="st"> </span><span class="fl">-0.6</span>]</code></pre>
<p>As before we generate a line plot showing these genes:</p>
<pre class="sourceCode r"><code class="sourceCode r">top1k.cdc28 <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>neg.corr.YAR018C) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression, <span class="dt">group =</span> gene)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>(<span class="dt">alpha =</span> <span class="fl">0.33</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&quot;none&quot;</span>)  </code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-266-1.png" width="576" /></p>
</div>
<div id="adding-new-columns-and-combining-filtered-data-frames" class="section level3">
<h3><span class="header-section-number">7.9.2</span> Adding new columns and combining filtered data frames</h3>
<p>Now let’s create a new data frame by: 1) filtering on our list of genes that have strong positive and negative correlations with YAR018C; and 2) creating a new variable, “corr.with.YAR018C”, which indicates the sign of the correlation. We’ll use this new variable to group genes when we create the plot.</p>
<pre class="sourceCode r"><code class="sourceCode r">pos.corr.df &lt;-<span class="st"> </span>
<span class="st">  </span>top1k.cdc28 <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>pos.corr.YAR018C) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">corr.with.YAR018C =</span> <span class="st">&quot;positive&quot;</span>)

neg.corr.df &lt;-<span class="st"> </span>
<span class="st">  </span>top1k.cdc28 <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>neg.corr.YAR018C) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">corr.with.YAR018C =</span> <span class="st">&quot;negative&quot;</span>)

combined.pos.neg &lt;-<span class="st"> </span><span class="kw">rbind</span>(pos.corr.df, neg.corr.df)</code></pre>
<p>Finally, we plot the data, colored according to the correlation with YAR018C:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(combined.pos.neg, 
       <span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression,  <span class="dt">group =</span> gene,
           <span class="dt">color =</span> corr.with.YAR018C)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_line</span>(<span class="dt">alpha=</span><span class="fl">0.25</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression),
            <span class="dt">data =</span> <span class="kw">filter</span>(top1k.cdc28, gene <span class="op">==</span><span class="st"> &quot;YAR018C&quot;</span>),
            <span class="dt">color =</span> <span class="st">&quot;DarkRed&quot;</span>, <span class="dt">size =</span> <span class="dv">2</span>,<span class="dt">alpha=</span><span class="fl">0.5</span>) <span class="op">+</span>
<span class="st">  </span><span class="co"># changes legend title and values for color sclae</span>
<span class="st">  </span><span class="kw">scale_color_manual</span>(<span class="dt">name =</span> <span class="st">&quot;Correlation with YAR018C&quot;</span>,
                       <span class="dt">values =</span> <span class="kw">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;red&quot;</span>))  <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&quot;Genes strongly positively and negatively correlated with YAR018C&quot;</span>,
       <span class="dt">subtitle =</span> <span class="st">&quot;YAR018C shown in dark red&quot;</span>,
       <span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Expression&quot;</span>)</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-268-1.png" width="768" /></p>
</div>
<div id="a-heat-mapped-sorted-by-correlations" class="section level3">
<h3><span class="header-section-number">7.9.3</span> A heat mapped sorted by correlations</h3>
<p>In our previous heat map example figure, we sorted genes according to peak expression. Now let’s generate a heat map for the genes that are strongly correlated (both positive and negative) with YAR018C. We will sort the genes according to the sign of their correlation.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># re-factor gene names so positive and negative genes are spatially distinct in plot</span>
combined.pos.neg<span class="op">$</span>gene &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw">factor</span>(combined.pos.neg<span class="op">$</span>gene, 
         <span class="dt">levels =</span> <span class="kw">c</span>(pos.corr.YAR018C, neg.corr.YAR018C))

combined.pos.neg <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> gene)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_tile</span>(<span class="kw">aes</span>(<span class="dt">fill =</span> expression)) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_fill_gradientn</span>(<span class="dt">colors=</span>color.scheme) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">xlab</span>(<span class="st">&quot;Time (mins)&quot;</span>)</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-269-1.png" width="1152" /></p>
<p>The breakpoint between the positively and negatively correlated sets of genes is quite obvious in this figure.</p>
</div>
<div id="a-fancy-figure" class="section level3">
<h3><span class="header-section-number">7.9.4</span> A “fancy” figure</h3>
<p>Recall that we introduced the <code>cowplot</code> library in Chapter 6, as a way to combine different ggplot outputs into subfigures such as you might find in a published paper. Here we’ll make further use cowplot to combine our heat map and line plot visualizations of genes that covary with YAR018C.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(cowplot)</code></pre>
<p>cowplot’s <code>draw_plot()</code> function allows us to place plots at arbitrary locations and with arbitrary sizes onto the canvas. The coordinates of the canvas run from 0 to 1, and the point (0, 0) is in the lower left corner of the canvas. We’ll use <code>draw_plot</code> to draw a complex figure with a heatmap on the left, and two smaller line plots on the right.</p>
<pre class="sourceCode r"><code class="sourceCode r">pos.corr.lineplot &lt;-<span class="st">  </span>
<span class="st">  </span>combined.pos.neg <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>pos.corr.YAR018C) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression, <span class="dt">group =</span> gene)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>(<span class="dt">alpha =</span> <span class="fl">0.33</span>, <span class="dt">color =</span> <span class="st">&#39;red&#39;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Expression&quot;</span>, 
      <span class="dt">title =</span> <span class="st">&quot;Genes Positively correlated</span><span class="ch">\n</span><span class="st">with YAR018C&quot;</span>)

neg.corr.lineplot &lt;-<span class="st"> </span>
<span class="st">  </span>combined.pos.neg <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(gene <span class="op">%in%</span><span class="st"> </span>neg.corr.YAR018C) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> expression, <span class="dt">group =</span> gene)) <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_line</span>(<span class="dt">alpha =</span> <span class="fl">0.33</span>, <span class="dt">color =</span> <span class="st">&#39;blue&#39;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Expression&quot;</span>, 
      <span class="dt">title =</span> <span class="st">&quot;Genes negatively correlated</span><span class="ch">\n</span><span class="st">with YAR018C&quot;</span>)


heat.map &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw">ggplot</span>(combined.pos.neg, <span class="kw">aes</span>(<span class="dt">x =</span> time, <span class="dt">y =</span> gene)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_tile</span>(<span class="kw">aes</span>(<span class="dt">fill =</span> expression)) <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_fill_gradientn</span>(<span class="dt">colors=</span>color.scheme) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Time (mins)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Gene&quot;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&quot;none&quot;</span>)</code></pre>
<p>The coordinates of the canvas run from 0 to 1, and the point (0, 0) is in the lower left corner of the canvas. We’ll use <code>draw_plot</code> to draw a complex figure with a heatmap on the left, and two smaller line plots on the right.</p>
<p>I determined the coordinates below by experimentation to create a visually pleasing layout.</p>
<pre class="sourceCode r"><code class="sourceCode r">fancy.plot &lt;-<span class="st"> </span><span class="kw">ggdraw</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">draw_plot</span>(heat.map, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dt">width =</span> <span class="fl">0.6</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">draw_plot</span>(neg.corr.lineplot, <span class="fl">0.6</span>, <span class="fl">0.5</span>, <span class="dt">width =</span> <span class="fl">0.4</span>, <span class="dt">height =</span> <span class="fl">0.5</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">draw_plot</span>(pos.corr.lineplot, <span class="fl">0.6</span>, <span class="dv">0</span>,   <span class="dt">width =</span> <span class="fl">0.4</span>, <span class="dt">height =</span> <span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">draw_plot_label</span>(<span class="kw">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;C&quot;</span>), <span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.6</span>, <span class="fl">0.6</span>), <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="fl">0.5</span>), <span class="dt">size =</span> <span class="dv">15</span>)

fancy.plot</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-272-1.png" width="1536" /></p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction-to-dplyr.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="functions-and-control-flow-statements.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "section"
},
"toolbar": {
"position": "fixed"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
