<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 6 Introduction to dplyr | Biology 723: Statistical Computing for Biologists</title>
  <meta name="description" content="Hands-on materials for Bio 723 course at Duke University.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 6 Introduction to dplyr | Biology 723: Statistical Computing for Biologists" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 Introduction to dplyr | Biology 723: Statistical Computing for Biologists" />
  
  <meta name="twitter:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

<meta name="author" content="Paul M. Magwene">


<meta name="date" content="2019-04-10">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="introduction-to-ggplot2.html">
<link rel="next" href="data-wrangling.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script src="libs/htmlwidgets-1.3/htmlwidgets.js"></script>
<link href="libs/rglwidgetClass-2/rgl.css" rel="stylesheet" />
<script src="libs/rglwidgetClass-2/rglClass.src.js"></script>
<script src="libs/CanvasMatrix4-2016/CanvasMatrix.src.js"></script>
<script src="libs/rglWebGL-binding-0.99.16/rglWebGL.js"></script>
<link href="libs/crosstalk-1.0.0/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.0.0/js/crosstalk.min.js"></script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { background-color: #f8f8f8; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#how-to-use-these-lecture-notes"><i class="fa fa-check"></i><b>1.1</b> How to use these lecture notes</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html"><i class="fa fa-check"></i><b>2</b> Getting Started with R</a><ul>
<li class="chapter" data-level="2.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-r"><i class="fa fa-check"></i><b>2.1</b> What is R?</a></li>
<li class="chapter" data-level="2.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-rstudio"><i class="fa fa-check"></i><b>2.2</b> What is RStudio?</a></li>
<li class="chapter" data-level="2.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#entering-commands-in-the-console"><i class="fa fa-check"></i><b>2.3</b> Entering commands in the console</a></li>
<li class="chapter" data-level="2.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#comments"><i class="fa fa-check"></i><b>2.4</b> Comments</a></li>
<li class="chapter" data-level="2.5" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#using-r-as-a-calculator"><i class="fa fa-check"></i><b>2.5</b> Using R as a Calculator</a><ul>
<li class="chapter" data-level="2.5.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#common-mathematical-functions"><i class="fa fa-check"></i><b>2.5.1</b> Common mathematical functions</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#variable-assignment"><i class="fa fa-check"></i><b>2.6</b> Variable assignment</a><ul>
<li class="chapter" data-level="2.6.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#valid-variable-names"><i class="fa fa-check"></i><b>2.6.1</b> Valid variable names</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#data-types"><i class="fa fa-check"></i><b>2.7</b> Data types</a><ul>
<li class="chapter" data-level="2.7.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#numeric-data-types"><i class="fa fa-check"></i><b>2.7.1</b> Numeric data types</a></li>
<li class="chapter" data-level="2.7.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#logical-values"><i class="fa fa-check"></i><b>2.7.2</b> Logical values</a></li>
<li class="chapter" data-level="2.7.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#character-strings"><i class="fa fa-check"></i><b>2.7.3</b> Character strings</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#packages"><i class="fa fa-check"></i><b>2.8</b> Packages</a><ul>
<li class="chapter" data-level="2.8.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-console"><i class="fa fa-check"></i><b>2.8.1</b> Installing packages from the console</a></li>
<li class="chapter" data-level="2.8.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#install-the-tidyverse-package"><i class="fa fa-check"></i><b>2.8.2</b> Install the tidyverse package</a></li>
<li class="chapter" data-level="2.8.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-rstudio-dialog"><i class="fa fa-check"></i><b>2.8.3</b> Installing packages from the RStudio dialog</a></li>
<li class="chapter" data-level="2.8.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#loading-packages-with-the-library-function"><i class="fa fa-check"></i><b>2.8.4</b> Loading packages with the <code>library()</code> function</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#the-r-help-system"><i class="fa fa-check"></i><b>2.9</b> The R Help System</a><ul>
<li class="chapter" data-level="2.9.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#getting-help-from-the-console"><i class="fa fa-check"></i><b>2.9.1</b> Getting help from the console</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html"><i class="fa fa-check"></i><b>3</b> R Markdown and R Notebooks</a><ul>
<li class="chapter" data-level="3.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#r-notebooks"><i class="fa fa-check"></i><b>3.1</b> R Notebooks</a></li>
<li class="chapter" data-level="3.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#creating-an-r-notebook"><i class="fa fa-check"></i><b>3.2</b> Creating an R Notebook</a></li>
<li class="chapter" data-level="3.3" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#the-default-r-notebook-template"><i class="fa fa-check"></i><b>3.3</b> The default R Notebook template</a></li>
<li class="chapter" data-level="3.4" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-and-non-code-blocks"><i class="fa fa-check"></i><b>3.4</b> Code and Non-code blocks</a><ul>
<li class="chapter" data-level="3.4.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#non-code-blocks"><i class="fa fa-check"></i><b>3.4.1</b> Non-code blocks</a></li>
<li class="chapter" data-level="3.4.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-blocks"><i class="fa fa-check"></i><b>3.4.2</b> Code blocks</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-a-code-chunk"><i class="fa fa-check"></i><b>3.5</b> Running a code chunk</a></li>
<li class="chapter" data-level="3.6" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-all-code-chunks-above"><i class="fa fa-check"></i><b>3.6</b> Running all code chunks above</a></li>
<li class="chapter" data-level="3.7" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#knitting-an-r-markdown-to-html"><i class="fa fa-check"></i><b>3.7</b> “Knitting an” R Markdown to HTML</a></li>
<li class="chapter" data-level="3.8" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#sharing-your-reproducible-r-notebook"><i class="fa fa-check"></i><b>3.8</b> Sharing your reproducible R Notebook</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-structures.html"><a href="data-structures.html"><i class="fa fa-check"></i><b>4</b> Data structures</a><ul>
<li class="chapter" data-level="4.1" data-path="data-structures.html"><a href="data-structures.html#vectors"><i class="fa fa-check"></i><b>4.1</b> Vectors</a><ul>
<li class="chapter" data-level="4.1.1" data-path="data-structures.html"><a href="data-structures.html#vector-arithmetic"><i class="fa fa-check"></i><b>4.1.1</b> Vector Arithmetic</a></li>
<li class="chapter" data-level="4.1.2" data-path="data-structures.html"><a href="data-structures.html#vector-recycling"><i class="fa fa-check"></i><b>4.1.2</b> Vector recycling</a></li>
<li class="chapter" data-level="4.1.3" data-path="data-structures.html"><a href="data-structures.html#simple-statistical-functions-for-numeric-vectors"><i class="fa fa-check"></i><b>4.1.3</b> Simple statistical functions for numeric vectors</a></li>
<li class="chapter" data-level="4.1.4" data-path="data-structures.html"><a href="data-structures.html#indexing-vectors"><i class="fa fa-check"></i><b>4.1.4</b> Indexing Vectors</a></li>
<li class="chapter" data-level="4.1.5" data-path="data-structures.html"><a href="data-structures.html#comparison-operators-applied-to-vectors"><i class="fa fa-check"></i><b>4.1.5</b> Comparison operators applied to vectors</a></li>
<li class="chapter" data-level="4.1.6" data-path="data-structures.html"><a href="data-structures.html#combining-indexing-and-comparison-of-vectors"><i class="fa fa-check"></i><b>4.1.6</b> Combining Indexing and Comparison of Vectors</a></li>
<li class="chapter" data-level="4.1.7" data-path="data-structures.html"><a href="data-structures.html#vector-manipulation"><i class="fa fa-check"></i><b>4.1.7</b> Vector manipulation</a></li>
<li class="chapter" data-level="4.1.8" data-path="data-structures.html"><a href="data-structures.html#vectors-from-regular-sequences"><i class="fa fa-check"></i><b>4.1.8</b> Vectors from regular sequences</a></li>
<li class="chapter" data-level="4.1.9" data-path="data-structures.html"><a href="data-structures.html#additional-functions-for-working-with-vectors"><i class="fa fa-check"></i><b>4.1.9</b> Additional functions for working with vectors</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="data-structures.html"><a href="data-structures.html#lists"><i class="fa fa-check"></i><b>4.2</b> Lists</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-structures.html"><a href="data-structures.html#length-and-type-of-lists"><i class="fa fa-check"></i><b>4.2.1</b> Length and type of lists</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-structures.html"><a href="data-structures.html#indexing-lists"><i class="fa fa-check"></i><b>4.2.2</b> Indexing lists</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-structures.html"><a href="data-structures.html#naming-list-elements"><i class="fa fa-check"></i><b>4.2.3</b> Naming list elements</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-structures.html"><a href="data-structures.html#the-operator"><i class="fa fa-check"></i><b>4.2.4</b> The <code>$</code> operator</a></li>
<li class="chapter" data-level="4.2.5" data-path="data-structures.html"><a href="data-structures.html#changing-and-adding-lists-items"><i class="fa fa-check"></i><b>4.2.5</b> Changing and adding lists items</a></li>
<li class="chapter" data-level="4.2.6" data-path="data-structures.html"><a href="data-structures.html#combining-lists"><i class="fa fa-check"></i><b>4.2.6</b> Combining lists</a></li>
<li class="chapter" data-level="4.2.7" data-path="data-structures.html"><a href="data-structures.html#converting-lists-to-vectors"><i class="fa fa-check"></i><b>4.2.7</b> Converting lists to vectors</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-structures.html"><a href="data-structures.html#data-frames"><i class="fa fa-check"></i><b>4.3</b> Data frames</a><ul>
<li class="chapter" data-level="4.3.1" data-path="data-structures.html"><a href="data-structures.html#creating-a-data-frame"><i class="fa fa-check"></i><b>4.3.1</b> Creating a data frame</a></li>
<li class="chapter" data-level="4.3.2" data-path="data-structures.html"><a href="data-structures.html#type-and-class-for-data-frames"><i class="fa fa-check"></i><b>4.3.2</b> Type and class for data frames</a></li>
<li class="chapter" data-level="4.3.3" data-path="data-structures.html"><a href="data-structures.html#length-and-dimension-for-data-frames"><i class="fa fa-check"></i><b>4.3.3</b> Length and dimension for data frames</a></li>
<li class="chapter" data-level="4.3.4" data-path="data-structures.html"><a href="data-structures.html#indexing-and-accessing-data-frames"><i class="fa fa-check"></i><b>4.3.4</b> Indexing and accessing data frames</a></li>
<li class="chapter" data-level="4.3.5" data-path="data-structures.html"><a href="data-structures.html#logical-indexing-of-data-frames"><i class="fa fa-check"></i><b>4.3.5</b> Logical indexing of data frames</a></li>
<li class="chapter" data-level="4.3.6" data-path="data-structures.html"><a href="data-structures.html#adding-columns-to-a-data-frame"><i class="fa fa-check"></i><b>4.3.6</b> Adding columns to a data frame</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html"><i class="fa fa-check"></i><b>5</b> Introduction to ggplot2</a><ul>
<li class="chapter" data-level="5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#loading-ggplot2"><i class="fa fa-check"></i><b>5.1</b> Loading ggplot2</a></li>
<li class="chapter" data-level="5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#example-data-set-andersons-iris-data"><i class="fa fa-check"></i><b>5.2</b> Example data set: Anderson’s Iris Data</a></li>
<li class="chapter" data-level="5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#template-for-single-layer-plots-in-ggplot2"><i class="fa fa-check"></i><b>5.3</b> Template for single layer plots in ggplot2</a></li>
<li class="chapter" data-level="5.4" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#an-aside-about-function-arguments"><i class="fa fa-check"></i><b>5.4</b> An aside about function arguments</a></li>
<li class="chapter" data-level="5.5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#strip-plots"><i class="fa fa-check"></i><b>5.5</b> Strip plots</a><ul>
<li class="chapter" data-level="5.5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#jittering-data"><i class="fa fa-check"></i><b>5.5.1</b> Jittering data</a></li>
<li class="chapter" data-level="5.5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-categorical-information"><i class="fa fa-check"></i><b>5.5.2</b> Adding categorical information</a></li>
<li class="chapter" data-level="5.5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#rotating-plot-coordinates"><i class="fa fa-check"></i><b>5.5.3</b> Rotating plot coordinates</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#histograms"><i class="fa fa-check"></i><b>5.6</b> Histograms</a><ul>
<li class="chapter" data-level="5.6.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#variations-on-histograms-when-considering-categorical-data"><i class="fa fa-check"></i><b>5.6.1</b> Variations on histograms when considering categorical data</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#faceting-to-depict-categorical-information"><i class="fa fa-check"></i><b>5.7</b> Faceting to depict categorical information</a></li>
<li class="chapter" data-level="5.8" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots"><i class="fa fa-check"></i><b>5.8</b> Density plots</a></li>
<li class="chapter" data-level="5.9" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-or-beanplot"><i class="fa fa-check"></i><b>5.9</b> Violin or Beanplot</a></li>
<li class="chapter" data-level="5.10" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplots"><i class="fa fa-check"></i><b>5.10</b> Boxplots</a></li>
<li class="chapter" data-level="5.11" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#building-complex-visualizations-with-layers"><i class="fa fa-check"></i><b>5.11</b> Building complex visualizations with layers</a></li>
<li class="chapter" data-level="5.12" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#useful-combination-plots"><i class="fa fa-check"></i><b>5.12</b> Useful combination plots</a><ul>
<li class="chapter" data-level="5.12.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplot-plus-strip-plot"><i class="fa fa-check"></i><b>5.12.1</b> Boxplot plus strip plot</a></li>
<li class="chapter" data-level="5.12.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#setting-shared-aesthetics"><i class="fa fa-check"></i><b>5.12.2</b> Setting shared aesthetics</a></li>
</ul></li>
<li class="chapter" data-level="5.13" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot-layers-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>5.13</b> ggplot layers can be assigned to variables</a><ul>
<li class="chapter" data-level="5.13.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-plot-plus-strip-plot"><i class="fa fa-check"></i><b>5.13.1</b> Violin plot plus strip plot</a></li>
</ul></li>
<li class="chapter" data-level="5.14" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-titles-and-tweaking-axis-labels"><i class="fa fa-check"></i><b>5.14</b> Adding titles and tweaking axis labels</a></li>
<li class="chapter" data-level="5.15" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot2-themes"><i class="fa fa-check"></i><b>5.15</b> ggplot2 themes</a><ul>
<li class="chapter" data-level="5.15.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#further-customization-with-ggplot2theme"><i class="fa fa-check"></i><b>5.15.1</b> Further customization with <code>ggplot2::theme</code></a></li>
</ul></li>
<li class="chapter" data-level="5.16" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#other-aspects-of-ggplots-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>5.16</b> Other aspects of ggplots can be assigned to variables</a></li>
<li class="chapter" data-level="5.17" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-plots"><i class="fa fa-check"></i><b>5.17</b> Bivariate plots</a><ul>
<li class="chapter" data-level="5.17.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#scatter-plots"><i class="fa fa-check"></i><b>5.17.1</b> Scatter plots</a></li>
<li class="chapter" data-level="5.17.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-a-trend-line-to-a-scatter-plot"><i class="fa fa-check"></i><b>5.17.2</b> Adding a trend line to a scatter plot</a></li>
</ul></li>
<li class="chapter" data-level="5.18" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-density-plots"><i class="fa fa-check"></i><b>5.18</b> Bivariate density plots</a></li>
<li class="chapter" data-level="5.19" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#combining-scatter-plots-and-density-plots-with-categorical-information"><i class="fa fa-check"></i><b>5.19</b> Combining Scatter Plots and Density Plots with Categorical Information</a></li>
<li class="chapter" data-level="5.20" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots-with-fill"><i class="fa fa-check"></i><b>5.20</b> Density plots with fill</a></li>
<li class="chapter" data-level="5.21" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#d-bin-and-hex-plots"><i class="fa fa-check"></i><b>5.21</b> 2D bin and hex plots</a></li>
<li class="chapter" data-level="5.22" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#the-cowplot-package"><i class="fa fa-check"></i><b>5.22</b> The <code>cowplot</code> package</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html"><i class="fa fa-check"></i><b>6</b> Introduction to <code>dplyr</code></a><ul>
<li class="chapter" data-level="6.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#libraries"><i class="fa fa-check"></i><b>6.1</b> Libraries</a></li>
<li class="chapter" data-level="6.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#reading-data-with-the-readr-package"><i class="fa fa-check"></i><b>6.2</b> Reading data with the <code>readr</code> package</a><ul>
<li class="chapter" data-level="6.2.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#reading-excel-files"><i class="fa fa-check"></i><b>6.2.1</b> Reading Excel files</a></li>
<li class="chapter" data-level="6.2.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#example-data-nc-births"><i class="fa fa-check"></i><b>6.2.2</b> Example data: NC Births</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#a-note-on-tibbles"><i class="fa fa-check"></i><b>6.3</b> A note on “tibbles”</a></li>
<li class="chapter" data-level="6.4" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#data-filtering-and-transformation-with-dplyr"><i class="fa fa-check"></i><b>6.4</b> Data filtering and transformation with <code>dplyr</code></a></li>
<li class="chapter" data-level="6.5" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#dplyrs-verbs"><i class="fa fa-check"></i><b>6.5</b> dplyr’s “verbs”</a><ul>
<li class="chapter" data-level="6.5.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#select"><i class="fa fa-check"></i><b>6.5.1</b> <code>select</code></a></li>
<li class="chapter" data-level="6.5.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#filter"><i class="fa fa-check"></i><b>6.5.2</b> <code>filter</code></a></li>
<li class="chapter" data-level="6.5.3" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#mutate"><i class="fa fa-check"></i><b>6.5.3</b> <code>mutate</code></a></li>
<li class="chapter" data-level="6.5.4" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#arrange"><i class="fa fa-check"></i><b>6.5.4</b> <code>arrange</code></a></li>
<li class="chapter" data-level="6.5.5" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#summarize"><i class="fa fa-check"></i><b>6.5.5</b> <code>summarize</code></a></li>
<li class="chapter" data-level="6.5.6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#group_by"><i class="fa fa-check"></i><b>6.5.6</b> <code>group_by</code></a></li>
<li class="chapter" data-level="6.5.7" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#combining-grouping-and-summarizing"><i class="fa fa-check"></i><b>6.5.7</b> Combining grouping and summarizing</a></li>
<li class="chapter" data-level="6.5.8" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#scoped-variants-of-mutate-and-summarize"><i class="fa fa-check"></i><b>6.5.8</b> Scoped variants of <code>mutate</code> and <code>summarize</code></a></li>
<li class="chapter" data-level="6.5.9" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#combining-summarize-with-grouping-aesthetics-in-ggplot2"><i class="fa fa-check"></i><b>6.5.9</b> Combining summarize with grouping aesthetics in <code>ggplot2</code></a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#pipes"><i class="fa fa-check"></i><b>6.6</b> Pipes</a><ul>
<li class="chapter" data-level="6.6.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#install-and-load-magrittr"><i class="fa fa-check"></i><b>6.6.1</b> Install and load <code>magrittr</code></a></li>
<li class="chapter" data-level="6.6.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-basic-pipe-operator"><i class="fa fa-check"></i><b>6.6.2</b> The basic pipe operator</a></li>
<li class="chapter" data-level="6.6.3" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#an-example-without-pipes"><i class="fa fa-check"></i><b>6.6.3</b> An example without pipes</a></li>
<li class="chapter" data-level="6.6.4" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-same-example-using-pipes"><i class="fa fa-check"></i><b>6.6.4</b> The same example using pipes</a></li>
<li class="chapter" data-level="6.6.5" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#assigning-the-output-of-a-statement-involving-pipes-to-a-variable"><i class="fa fa-check"></i><b>6.6.5</b> Assigning the output of a statement involving pipes to a variable</a></li>
<li class="chapter" data-level="6.6.6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#compound-assignment-pipe-operator"><i class="fa fa-check"></i><b>6.6.6</b> Compound assignment pipe operator</a></li>
<li class="chapter" data-level="6.6.7" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-dot-operator-with-pipes"><i class="fa fa-check"></i><b>6.6.7</b> The dot operator with pipes</a></li>
<li class="chapter" data-level="6.6.8" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-exposition-pipe-operator"><i class="fa fa-check"></i><b>6.6.8</b> The exposition pipe operator</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>7</b> Data wrangling</a><ul>
<li class="chapter" data-level="7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#libraries-1"><i class="fa fa-check"></i><b>7.1</b> Libraries</a></li>
<li class="chapter" data-level="7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#data"><i class="fa fa-check"></i><b>7.2</b> Data</a></li>
<li class="chapter" data-level="7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#renaming-data-frame-columms"><i class="fa fa-check"></i><b>7.3</b> Renaming data frame columms</a></li>
<li class="chapter" data-level="7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-unneeded-columns"><i class="fa fa-check"></i><b>7.4</b> Dropping unneeded columns</a><ul>
<li class="chapter" data-level="7.4.1" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-all-empty-columns"><i class="fa fa-check"></i><b>7.4.1</b> Finding all empty columns</a></li>
<li class="chapter" data-level="7.4.2" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-columns-by-matching-names"><i class="fa fa-check"></i><b>7.4.2</b> Dropping columns by matching names</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="data-wrangling.html"><a href="data-wrangling.html#merging-data-frames"><i class="fa fa-check"></i><b>7.5</b> Merging data frames</a></li>
<li class="chapter" data-level="7.6" data-path="data-wrangling.html"><a href="data-wrangling.html#reshaping-data-with-tidyr"><i class="fa fa-check"></i><b>7.6</b> Reshaping data with tidyr</a><ul>
<li class="chapter" data-level="7.6.1" data-path="data-wrangling.html"><a href="data-wrangling.html#wide-to-long-conversions-using-tidyrgather"><i class="fa fa-check"></i><b>7.6.1</b> Wide to long conversions using <code>tidyr::gather</code></a></li>
<li class="chapter" data-level="7.6.2" data-path="data-wrangling.html"><a href="data-wrangling.html#extracting-information-from-combined-variables-using-tidyrextract"><i class="fa fa-check"></i><b>7.6.2</b> Extracting information from combined variables using <code>tidyr::extract</code></a></li>
<li class="chapter" data-level="7.6.3" data-path="data-wrangling.html"><a href="data-wrangling.html#subsetting-rows"><i class="fa fa-check"></i><b>7.6.3</b> Subsetting rows</a></li>
<li class="chapter" data-level="7.6.4" data-path="data-wrangling.html"><a href="data-wrangling.html#splitting-columns"><i class="fa fa-check"></i><b>7.6.4</b> Splitting columns</a></li>
<li class="chapter" data-level="7.6.5" data-path="data-wrangling.html"><a href="data-wrangling.html#combining-data-frames"><i class="fa fa-check"></i><b>7.6.5</b> Combining data frames</a></li>
<li class="chapter" data-level="7.6.6" data-path="data-wrangling.html"><a href="data-wrangling.html#sorting-data-frame-rows"><i class="fa fa-check"></i><b>7.6.6</b> Sorting data frame rows</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="data-wrangling.html"><a href="data-wrangling.html#using-your-tidy-data"><i class="fa fa-check"></i><b>7.7</b> Using your tidy data</a><ul>
<li class="chapter" data-level="7.7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#visualizing-gene-expression-time-series"><i class="fa fa-check"></i><b>7.7.1</b> Visualizing gene expression time series</a></li>
<li class="chapter" data-level="7.7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-the-most-variable-genes"><i class="fa fa-check"></i><b>7.7.2</b> Finding the most variable genes</a></li>
<li class="chapter" data-level="7.7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#heat-maps"><i class="fa fa-check"></i><b>7.7.3</b> Heat maps</a></li>
</ul></li>
<li class="chapter" data-level="7.8" data-path="data-wrangling.html"><a href="data-wrangling.html#long-to-wide-conversion-using-tidyrspread"><i class="fa fa-check"></i><b>7.8</b> Long-to-wide conversion using <code>tidyr::spread</code></a></li>
<li class="chapter" data-level="7.9" data-path="data-wrangling.html"><a href="data-wrangling.html#exploring-bivariate-relationships-using-wide-data"><i class="fa fa-check"></i><b>7.9</b> Exploring bivariate relationships using “wide” data</a><ul>
<li class="chapter" data-level="7.9.1" data-path="data-wrangling.html"><a href="data-wrangling.html#large-scale-patterns-of-correlations"><i class="fa fa-check"></i><b>7.9.1</b> Large scale patterns of correlations</a></li>
<li class="chapter" data-level="7.9.2" data-path="data-wrangling.html"><a href="data-wrangling.html#adding-new-columns-and-combining-filtered-data-frames"><i class="fa fa-check"></i><b>7.9.2</b> Adding new columns and combining filtered data frames</a></li>
<li class="chapter" data-level="7.9.3" data-path="data-wrangling.html"><a href="data-wrangling.html#a-heat-mapped-sorted-by-correlations"><i class="fa fa-check"></i><b>7.9.3</b> A heat mapped sorted by correlations</a></li>
<li class="chapter" data-level="7.9.4" data-path="data-wrangling.html"><a href="data-wrangling.html#a-fancy-figure"><i class="fa fa-check"></i><b>7.9.4</b> A “fancy” figure</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html"><i class="fa fa-check"></i><b>8</b> Functions and control flow statements</a><ul>
<li class="chapter" data-level="8.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#writing-your-own-functions"><i class="fa fa-check"></i><b>8.1</b> Writing your own functions</a><ul>
<li class="chapter" data-level="8.1.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#function-arguments"><i class="fa fa-check"></i><b>8.1.1</b> Function arguments</a></li>
<li class="chapter" data-level="8.1.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#writing-functions-with-optional-arguments"><i class="fa fa-check"></i><b>8.1.2</b> Writing functions with optional arguments</a></li>
<li class="chapter" data-level="8.1.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#putting-r-functions-in-scripts"><i class="fa fa-check"></i><b>8.1.3</b> Putting R functions in Scripts</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#control-flow-statements"><i class="fa fa-check"></i><b>8.2</b> Control flow statements</a><ul>
<li class="chapter" data-level="8.2.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#if-and-if-else-statements"><i class="fa fa-check"></i><b>8.2.1</b> <code>if</code> and <code>if-else</code> statements</a></li>
<li class="chapter" data-level="8.2.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#for-loops"><i class="fa fa-check"></i><b>8.2.2</b> for loops</a></li>
<li class="chapter" data-level="8.2.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#break-statement"><i class="fa fa-check"></i><b>8.2.3</b> <code>break</code> statement</a></li>
<li class="chapter" data-level="8.2.4" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#repeat-loops"><i class="fa fa-check"></i><b>8.2.4</b> <code>repeat</code> loops</a></li>
<li class="chapter" data-level="8.2.5" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#next-statement"><i class="fa fa-check"></i><b>8.2.5</b> <code>next</code> statement</a></li>
<li class="chapter" data-level="8.2.6" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#while-statements"><i class="fa fa-check"></i><b>8.2.6</b> while statements</a></li>
<li class="chapter" data-level="8.2.7" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#ifelse"><i class="fa fa-check"></i><b>8.2.7</b> <code>ifelse</code></a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#map-and-related-tools"><i class="fa fa-check"></i><b>8.3</b> <code>map</code> and related tools</a><ul>
<li class="chapter" data-level="8.3.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#basic-map"><i class="fa fa-check"></i><b>8.3.1</b> basic <code>map</code></a></li>
<li class="chapter" data-level="8.3.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#map_if-and-map_at"><i class="fa fa-check"></i><b>8.3.2</b> <code>map_if</code> and <code>map_at</code></a></li>
<li class="chapter" data-level="8.3.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#mapping-in-parallel-using-map2"><i class="fa fa-check"></i><b>8.3.3</b> mapping in parallel using <code>map2</code></a></li>
<li class="chapter" data-level="8.3.4" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#map-variants-that-return-vectors"><i class="fa fa-check"></i><b>8.3.4</b> <code>map</code> variants that return vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="vector-algebra.html"><a href="vector-algebra.html"><i class="fa fa-check"></i><b>9</b> Vector algebra</a><ul>
<li class="chapter" data-level="9.1" data-path="vector-algebra.html"><a href="vector-algebra.html#libraries-2"><i class="fa fa-check"></i><b>9.1</b> Libraries</a></li>
<li class="chapter" data-level="9.2" data-path="vector-algebra.html"><a href="vector-algebra.html#vector-mathematics-in-r"><i class="fa fa-check"></i><b>9.2</b> Vector Mathematics in R</a></li>
<li class="chapter" data-level="9.3" data-path="vector-algebra.html"><a href="vector-algebra.html#simple-statistics-in-vector-form"><i class="fa fa-check"></i><b>9.3</b> Simple statistics in vector form</a><ul>
<li class="chapter" data-level="9.3.1" data-path="vector-algebra.html"><a href="vector-algebra.html#mean"><i class="fa fa-check"></i><b>9.3.1</b> Mean</a></li>
<li class="chapter" data-level="9.3.2" data-path="vector-algebra.html"><a href="vector-algebra.html#mean-centering"><i class="fa fa-check"></i><b>9.3.2</b> Mean centering</a></li>
<li class="chapter" data-level="9.3.3" data-path="vector-algebra.html"><a href="vector-algebra.html#variance-and-standard-deviation"><i class="fa fa-check"></i><b>9.3.3</b> Variance and standard deviation</a></li>
<li class="chapter" data-level="9.3.4" data-path="vector-algebra.html"><a href="vector-algebra.html#covariance-and-correlation"><i class="fa fa-check"></i><b>9.3.4</b> Covariance and correlation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="matrices-in-r.html"><a href="matrices-in-r.html"><i class="fa fa-check"></i><b>10</b> Matrices in R</a><ul>
<li class="chapter" data-level="10.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#creating-matrices-in-r"><i class="fa fa-check"></i><b>10.1</b> Creating matrices in R</a></li>
<li class="chapter" data-level="10.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-arithmetic-operations-in-r"><i class="fa fa-check"></i><b>10.2</b> Matrix arithmetic operations in R</a></li>
<li class="chapter" data-level="10.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#descriptive-statistics-as-matrix-functions"><i class="fa fa-check"></i><b>10.3</b> Descriptive statistics as matrix functions</a><ul>
<li class="chapter" data-level="10.3.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#mean-vector-and-matrix"><i class="fa fa-check"></i><b>10.3.1</b> Mean vector and matrix</a></li>
<li class="chapter" data-level="10.3.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#deviation-matrix"><i class="fa fa-check"></i><b>10.3.2</b> Deviation matrix</a></li>
<li class="chapter" data-level="10.3.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#covariance-matrix"><i class="fa fa-check"></i><b>10.3.3</b> Covariance matrix</a></li>
<li class="chapter" data-level="10.3.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#correlation-matrix"><i class="fa fa-check"></i><b>10.3.4</b> Correlation matrix</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-inverse"><i class="fa fa-check"></i><b>10.4</b> Matrix Inverse</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html"><i class="fa fa-check"></i><b>11</b> Fitting Regression Models in R</a><ul>
<li class="chapter" data-level="11.1" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#regression-terminology"><i class="fa fa-check"></i><b>11.1</b> Regression terminology</a></li>
<li class="chapter" data-level="11.2" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#the-optimality-criterion-for-least-squares-regression"><i class="fa fa-check"></i><b>11.2</b> The optimality criterion for least-squares regression</a></li>
<li class="chapter" data-level="11.3" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#vector-geometric-perspective"><i class="fa fa-check"></i><b>11.3</b> Vector geometric perspective</a></li>
<li class="chapter" data-level="11.4" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#solution-for-the-least-squares-criterion"><i class="fa fa-check"></i><b>11.4</b> Solution for the least-squares criterion</a></li>
<li class="chapter" data-level="11.5" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#libraries-3"><i class="fa fa-check"></i><b>11.5</b> Libraries</a></li>
<li class="chapter" data-level="11.6" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#illustrating-linear-regression-with-simulated-data"><i class="fa fa-check"></i><b>11.6</b> Illustrating linear regression with simulated data</a></li>
<li class="chapter" data-level="11.7" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#specifying-regression-models-in-r"><i class="fa fa-check"></i><b>11.7</b> Specifying Regression Models in R</a><ul>
<li class="chapter" data-level="11.7.1" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#fitted-values"><i class="fa fa-check"></i><b>11.7.1</b> Fitted values</a></li>
<li class="chapter" data-level="11.7.2" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#getting-the-model-coefficients"><i class="fa fa-check"></i><b>11.7.2</b> Getting the model coefficients</a></li>
</ul></li>
<li class="chapter" data-level="11.8" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#residuals"><i class="fa fa-check"></i><b>11.8</b> Residuals</a></li>
<li class="chapter" data-level="11.9" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#regression-as-sum-of-squares-decomposition"><i class="fa fa-check"></i><b>11.9</b> Regression as sum-of-squares decomposition</a></li>
<li class="chapter" data-level="11.10" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#variance-explained-by-a-regression-model"><i class="fa fa-check"></i><b>11.10</b> Variance “explained” by a regression model</a></li>
<li class="chapter" data-level="11.11" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#broom-a-library-for-converting-model-results-into-data-frames"><i class="fa fa-check"></i><b>11.11</b> Broom: a library for converting model results into data frames</a><ul>
<li class="chapter" data-level="11.11.1" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#broomtidy"><i class="fa fa-check"></i><b>11.11.1</b> <code>broom::tidy</code></a></li>
<li class="chapter" data-level="11.11.2" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#broomaugment"><i class="fa fa-check"></i><b>11.11.2</b> <code>broom::augment</code></a></li>
<li class="chapter" data-level="11.11.3" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#broomglance"><i class="fa fa-check"></i><b>11.11.3</b> <code>broom::glance</code></a></li>
</ul></li>
<li class="chapter" data-level="11.12" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#example-predicting-lion-age-based-on-nose-color"><i class="fa fa-check"></i><b>11.12</b> Example: Predicting lion age based on nose color</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="multiple-regression.html"><a href="multiple-regression.html"><i class="fa fa-check"></i><b>12</b> Multiple Regression</a><ul>
<li class="chapter" data-level="12.1" data-path="multiple-regression.html"><a href="multiple-regression.html#review-of-bivariate-regression"><i class="fa fa-check"></i><b>12.1</b> Review of bivariate regression</a></li>
<li class="chapter" data-level="12.2" data-path="multiple-regression.html"><a href="multiple-regression.html#multiple-regression-1"><i class="fa fa-check"></i><b>12.2</b> Multiple regression</a><ul>
<li class="chapter" data-level="12.2.1" data-path="multiple-regression.html"><a href="multiple-regression.html#variable-space-interpretation"><i class="fa fa-check"></i><b>12.2.1</b> Variable space interpretation</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="multiple-regression.html"><a href="multiple-regression.html#subject-space-representation"><i class="fa fa-check"></i><b>12.3</b> Subject space representation</a><ul>
<li class="chapter" data-level="12.3.1" data-path="multiple-regression.html"><a href="multiple-regression.html#coefficient-of-determination-for-multiple-regression"><i class="fa fa-check"></i><b>12.3.1</b> Coefficient of determination for multiple regression</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="multiple-regression.html"><a href="multiple-regression.html#interpretting-multiple-regression"><i class="fa fa-check"></i><b>12.4</b> Interpretting Multiple Regression</a></li>
<li class="chapter" data-level="12.5" data-path="multiple-regression.html"><a href="multiple-regression.html#new-libraries-to-install"><i class="fa fa-check"></i><b>12.5</b> New Libraries to install</a></li>
<li class="chapter" data-level="12.6" data-path="multiple-regression.html"><a href="multiple-regression.html#libraries-4"><i class="fa fa-check"></i><b>12.6</b> Libraries</a></li>
<li class="chapter" data-level="12.7" data-path="multiple-regression.html"><a href="multiple-regression.html#examplar-data"><i class="fa fa-check"></i><b>12.7</b> Examplar data</a></li>
<li class="chapter" data-level="12.8" data-path="multiple-regression.html"><a href="multiple-regression.html#data-exploration"><i class="fa fa-check"></i><b>12.8</b> Data exploration</a></li>
<li class="chapter" data-level="12.9" data-path="multiple-regression.html"><a href="multiple-regression.html#d-plots"><i class="fa fa-check"></i><b>12.9</b> 3D Plots</a><ul>
<li class="chapter" data-level="12.9.1" data-path="multiple-regression.html"><a href="multiple-regression.html#scatterplot3d"><i class="fa fa-check"></i><b>12.9.1</b> scatterplot3d</a></li>
<li class="chapter" data-level="12.9.2" data-path="multiple-regression.html"><a href="multiple-regression.html#rgl"><i class="fa fa-check"></i><b>12.9.2</b> rgl</a></li>
</ul></li>
<li class="chapter" data-level="12.10" data-path="multiple-regression.html"><a href="multiple-regression.html#fitting-the-regression-model"><i class="fa fa-check"></i><b>12.10</b> Fitting the regression model</a><ul>
<li class="chapter" data-level="12.10.1" data-path="multiple-regression.html"><a href="multiple-regression.html#visualizing-the-regression-model-in-scatterplot3d"><i class="fa fa-check"></i><b>12.10.1</b> Visualizing the regression model in scatterplot3d</a></li>
</ul></li>
<li class="chapter" data-level="12.11" data-path="multiple-regression.html"><a href="multiple-regression.html#interpretting-the-regression-model"><i class="fa fa-check"></i><b>12.11</b> Interpretting the regression model</a></li>
<li class="chapter" data-level="12.12" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-vector-geometry-of-the-regression-model"><i class="fa fa-check"></i><b>12.12</b> Exploring the Vector Geometry of the Regression Model</a></li>
<li class="chapter" data-level="12.13" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-residuals-from-the-model-fit"><i class="fa fa-check"></i><b>12.13</b> Exploring the Residuals from the Model Fit</a></li>
<li class="chapter" data-level="12.14" data-path="multiple-regression.html"><a href="multiple-regression.html#an-alternate-model"><i class="fa fa-check"></i><b>12.14</b> An alternate model</a></li>
<li class="chapter" data-level="12.15" data-path="multiple-regression.html"><a href="multiple-regression.html#fitting-a-curvilinear-model-using-lm"><i class="fa fa-check"></i><b>12.15</b> Fitting a curvilinear model using lm()</a></li>
<li class="chapter" data-level="12.16" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-impact-of-nearly-collinear-predictors-on-regression"><i class="fa fa-check"></i><b>12.16</b> Exploring the impact of nearly collinear predictors on regression</a></li>
<li class="chapter" data-level="12.17" data-path="multiple-regression.html"><a href="multiple-regression.html#loess-models"><i class="fa fa-check"></i><b>12.17</b> LOESS Models</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="logistic-regression.html"><a href="logistic-regression.html"><i class="fa fa-check"></i><b>13</b> Logistic regression</a><ul>
<li class="chapter" data-level="13.1" data-path="logistic-regression.html"><a href="logistic-regression.html#a-web-app-to-explore-the-logistic-regression-equation"><i class="fa fa-check"></i><b>13.1</b> A web app to explore the logistic regression equation</a></li>
<li class="chapter" data-level="13.2" data-path="logistic-regression.html"><a href="logistic-regression.html#titanic-data-set"><i class="fa fa-check"></i><b>13.2</b> Titanic data set</a></li>
<li class="chapter" data-level="13.3" data-path="logistic-regression.html"><a href="logistic-regression.html#subsetting-the-data"><i class="fa fa-check"></i><b>13.3</b> Subsetting the data</a></li>
<li class="chapter" data-level="13.4" data-path="logistic-regression.html"><a href="logistic-regression.html#visualizing-survival-as-a-function-of-age"><i class="fa fa-check"></i><b>13.4</b> Visualizing survival as a function of age</a></li>
<li class="chapter" data-level="13.5" data-path="logistic-regression.html"><a href="logistic-regression.html#fitting-the-logistic-regression-model"><i class="fa fa-check"></i><b>13.5</b> Fitting the logistic regression model</a></li>
<li class="chapter" data-level="13.6" data-path="logistic-regression.html"><a href="logistic-regression.html#visualizing-the-logistic-regression"><i class="fa fa-check"></i><b>13.6</b> Visualizing the logistic regression</a><ul>
<li class="chapter" data-level="13.6.1" data-path="logistic-regression.html"><a href="logistic-regression.html#quick-and-easy-visualization"><i class="fa fa-check"></i><b>13.6.1</b> Quick and easy visualization</a></li>
</ul></li>
<li class="chapter" data-level="13.7" data-path="logistic-regression.html"><a href="logistic-regression.html#impact-of-sex-and-passenger-class-on-the-models"><i class="fa fa-check"></i><b>13.7</b> Impact of sex and passenger class on the models</a></li>
<li class="chapter" data-level="13.8" data-path="logistic-regression.html"><a href="logistic-regression.html#fitting-multiple-models-based-on-groupings-use-dplyrdo"><i class="fa fa-check"></i><b>13.8</b> Fitting multiple models based on groupings use <code>dplyr::do</code></a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html"><i class="fa fa-check"></i><b>14</b> Principal Components Analysis</a><ul>
<li class="chapter" data-level="14.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#libraries-5"><i class="fa fa-check"></i><b>14.1</b> Libraries</a></li>
<li class="chapter" data-level="14.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#matrices-as-linear-transformations"><i class="fa fa-check"></i><b>14.2</b> Matrices as linear transformations</a></li>
<li class="chapter" data-level="14.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#eigenanalysis-in-r"><i class="fa fa-check"></i><b>14.3</b> Eigenanalysis in R</a></li>
<li class="chapter" data-level="14.4" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#principal-components-analysis-in-r"><i class="fa fa-check"></i><b>14.4</b> Principal Components Analysis in R</a><ul>
<li class="chapter" data-level="14.4.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#bioenv-dataset"><i class="fa fa-check"></i><b>14.4.1</b> Bioenv dataset</a></li>
<li class="chapter" data-level="14.4.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pca-of-the-bioenv-dataset"><i class="fa fa-check"></i><b>14.4.2</b> PCA of the Bioenv dataset</a></li>
<li class="chapter" data-level="14.4.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#calculating-factor-loadings"><i class="fa fa-check"></i><b>14.4.3</b> Calculating Factor Loadings</a></li>
</ul></li>
<li class="chapter" data-level="14.5" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#drawing-figures-to-represent-pca"><i class="fa fa-check"></i><b>14.5</b> Drawing Figures to Represent PCA</a><ul>
<li class="chapter" data-level="14.5.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pc-score-plots"><i class="fa fa-check"></i><b>14.5.1</b> PC Score Plots</a></li>
<li class="chapter" data-level="14.5.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#simultaneous-depiction-of-observations-and-variables-in-the-pc-space"><i class="fa fa-check"></i><b>14.5.2</b> Simultaneous Depiction of Observations and Variables in the PC Space</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="15" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html"><i class="fa fa-check"></i><b>15</b> Singular Value Decomposition</a><ul>
<li class="chapter" data-level="15.1" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#libraries-6"><i class="fa fa-check"></i><b>15.1</b> Libraries</a></li>
<li class="chapter" data-level="15.2" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#svd-in-r"><i class="fa fa-check"></i><b>15.2</b> SVD in R</a></li>
<li class="chapter" data-level="15.3" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#writing-our-own-pca-function-using-svd"><i class="fa fa-check"></i><b>15.3</b> Writing our own PCA function using SVD</a></li>
<li class="chapter" data-level="15.4" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#writing-a-biplot-function"><i class="fa fa-check"></i><b>15.4</b> Writing a biplot function</a></li>
<li class="chapter" data-level="15.5" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#data-compression-and-noise-filtering-using-svd"><i class="fa fa-check"></i><b>15.5</b> Data compression and noise filtering using SVD</a><ul>
<li class="chapter" data-level="15.5.1" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#example-image"><i class="fa fa-check"></i><b>15.5.1</b> Example image</a></li>
</ul></li>
<li class="chapter" data-level="15.6" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#image-compression-using-svd-in-r"><i class="fa fa-check"></i><b>15.6</b> Image compression using SVD in R</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html"><i class="fa fa-check"></i><b>16</b> Canonical Variates Analysis</a><ul>
<li class="chapter" data-level="16.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#libraries-7"><i class="fa fa-check"></i><b>16.1</b> Libraries</a></li>
<li class="chapter" data-level="16.2" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#discriminant-analysis-in-r"><i class="fa fa-check"></i><b>16.2</b> Discriminant Analysis in R</a><ul>
<li class="chapter" data-level="16.2.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#shorthand-formulae-in-r"><i class="fa fa-check"></i><b>16.2.1</b> Shorthand Formulae in R</a></li>
<li class="chapter" data-level="16.2.2" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#working-with-the-output-of-lda"><i class="fa fa-check"></i><b>16.2.2</b> Working with the output of <code>lda()</code></a></li>
</ul></li>
<li class="chapter" data-level="16.3" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#estimating-confidence-regions-for-group-means-in-cva"><i class="fa fa-check"></i><b>16.3</b> Estimating confidence regions for group means in CVA</a><ul>
<li class="chapter" data-level="16.3.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#drawing-the-cva-confidence-regions"><i class="fa fa-check"></i><b>16.3.1</b> Drawing the CVA confidence regions</a></li>
</ul></li>
<li class="chapter" data-level="16.4" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#calculating-the-within-and-between-group-covariance-matrices"><i class="fa fa-check"></i><b>16.4</b> Calculating the Within and Between Group Covariance Matrices</a><ul>
<li class="chapter" data-level="16.4.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#recapitulating-the-cva-analysis-of-lda"><i class="fa fa-check"></i><b>16.4.1</b> Recapitulating the CVA analysis of <code>lda()</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="17" data-path="clustering-in-r.html"><a href="clustering-in-r.html"><i class="fa fa-check"></i><b>17</b> Clustering in R</a><ul>
<li class="chapter" data-level="17.1" data-path="clustering-in-r.html"><a href="clustering-in-r.html#libraries-8"><i class="fa fa-check"></i><b>17.1</b> Libraries</a></li>
<li class="chapter" data-level="17.2" data-path="clustering-in-r.html"><a href="clustering-in-r.html#data-set"><i class="fa fa-check"></i><b>17.2</b> Data set</a></li>
<li class="chapter" data-level="17.3" data-path="clustering-in-r.html"><a href="clustering-in-r.html#hierarchical-clustering-in-r"><i class="fa fa-check"></i><b>17.3</b> Hierarchical Clustering in R</a></li>
<li class="chapter" data-level="17.4" data-path="clustering-in-r.html"><a href="clustering-in-r.html#manipulating-hierarchical-clusterings-with-dendextend"><i class="fa fa-check"></i><b>17.4</b> Manipulating hierarchical clusterings with dendextend</a></li>
<li class="chapter" data-level="17.5" data-path="clustering-in-r.html"><a href="clustering-in-r.html#plotting-dendrograms-in-dendextend"><i class="fa fa-check"></i><b>17.5</b> Plotting dendrograms in dendextend</a></li>
<li class="chapter" data-level="17.6" data-path="clustering-in-r.html"><a href="clustering-in-r.html#cutting-dendrograms"><i class="fa fa-check"></i><b>17.6</b> Cutting dendrograms</a></li>
<li class="chapter" data-level="17.7" data-path="clustering-in-r.html"><a href="clustering-in-r.html#looking-at-clusters"><i class="fa fa-check"></i><b>17.7</b> Looking at clusters</a></li>
<li class="chapter" data-level="17.8" data-path="clustering-in-r.html"><a href="clustering-in-r.html#generating-a-heat-map-from-a-cluster"><i class="fa fa-check"></i><b>17.8</b> Generating a heat map from a cluster</a></li>
<li class="chapter" data-level="17.9" data-path="clustering-in-r.html"><a href="clustering-in-r.html#working-with-sub-trees"><i class="fa fa-check"></i><b>17.9</b> Working with sub-trees</a></li>
<li class="chapter" data-level="17.10" data-path="clustering-in-r.html"><a href="clustering-in-r.html#setting-dendrogram-parameters-in-dendextend"><i class="fa fa-check"></i><b>17.10</b> Setting dendrogram parameters in dendextend</a></li>
<li class="chapter" data-level="17.11" data-path="clustering-in-r.html"><a href="clustering-in-r.html#combining-heatmaps-and-dendrograms"><i class="fa fa-check"></i><b>17.11</b> Combining heatmaps and dendrograms</a></li>
<li class="chapter" data-level="17.12" data-path="clustering-in-r.html"><a href="clustering-in-r.html#k-meansk-medoids-clustering-in-r"><i class="fa fa-check"></i><b>17.12</b> K-means/K-medoids Clustering in R</a><ul>
<li class="chapter" data-level="17.12.1" data-path="clustering-in-r.html"><a href="clustering-in-r.html#heat-map-from-k-medoids-cluster"><i class="fa fa-check"></i><b>17.12.1</b> Heat map from k-medoids cluster</a></li>
</ul></li>
<li class="chapter" data-level="17.13" data-path="clustering-in-r.html"><a href="clustering-in-r.html#combining-clusters-and-correlation-matrix-heatmaps"><i class="fa fa-check"></i><b>17.13</b> Combining clusters and correlation matrix heatmaps</a><ul>
<li class="chapter" data-level="17.13.1" data-path="clustering-in-r.html"><a href="clustering-in-r.html#hierarchical-clustering-visualized-on-correlation-matrix"><i class="fa fa-check"></i><b>17.13.1</b> Hierarchical clustering, visualized on correlation matrix</a></li>
<li class="chapter" data-level="17.13.2" data-path="clustering-in-r.html"><a href="clustering-in-r.html#k-medoids-clustering-visualized-on-the-correlation-matrix"><i class="fa fa-check"></i><b>17.13.2</b> K-medoids clustering visualized on the correlation matrix</a></li>
</ul></li>
<li class="chapter" data-level="17.14" data-path="clustering-in-r.html"><a href="clustering-in-r.html#depicting-the-data-within-clusters"><i class="fa fa-check"></i><b>17.14</b> Depicting the data within clusters</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html"><i class="fa fa-check"></i><b>18</b> Simulation I: Sampling distributions</a><ul>
<li class="chapter" data-level="18.1" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#sampling-distrubtions"><i class="fa fa-check"></i><b>18.1</b> Sampling distrubtions</a></li>
<li class="chapter" data-level="18.2" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#simulating-sampling-from-an-underlying-population"><i class="fa fa-check"></i><b>18.2</b> Simulating sampling from an underlying population</a></li>
<li class="chapter" data-level="18.3" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#seeding-the-pseudo-random-number-generator"><i class="fa fa-check"></i><b>18.3</b> Seeding the pseudo-random number generator</a></li>
<li class="chapter" data-level="18.4" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#libraries-9"><i class="fa fa-check"></i><b>18.4</b> Libraries</a><ul>
<li class="chapter" data-level="18.4.1" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#properties-of-the-underlying-population"><i class="fa fa-check"></i><b>18.4.1</b> Properties of the underlying population</a></li>
<li class="chapter" data-level="18.4.2" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#other-r-functions-related-to-the-normal-distribution"><i class="fa fa-check"></i><b>18.4.2</b> Other R functions related to the normal distribution</a></li>
</ul></li>
<li class="chapter" data-level="18.5" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#random-sampling-from-the-simulated-population"><i class="fa fa-check"></i><b>18.5</b> Random sampling from the simulated population</a><ul>
<li class="chapter" data-level="18.5.1" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#another-random-sample"><i class="fa fa-check"></i><b>18.5.1</b> Another random sample</a></li>
<li class="chapter" data-level="18.5.2" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#simulating-the-generation-of-many-random-samples"><i class="fa fa-check"></i><b>18.5.2</b> Simulating the generation of many random samples</a></li>
<li class="chapter" data-level="18.5.3" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#a-function-to-estimate-statistics-of-interest-in-a-random-sample"><i class="fa fa-check"></i><b>18.5.3</b> A function to estimate statistics of interest in a random sample</a></li>
<li class="chapter" data-level="18.5.4" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#generating-statistics-for-many-random-samples"><i class="fa fa-check"></i><b>18.5.4</b> Generating statistics for many random samples</a></li>
</ul></li>
<li class="chapter" data-level="18.6" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#simulated-sampling-distribution-of-the-mean"><i class="fa fa-check"></i><b>18.6</b> Simulated sampling distribution of the mean</a><ul>
<li class="chapter" data-level="18.6.1" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#differences-between-sampling-distributions-and-samplepopulation-distributions"><i class="fa fa-check"></i><b>18.6.1</b> Differences between sampling distributions and sample/population distributions</a></li>
<li class="chapter" data-level="18.6.2" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#using-sampling-distributions-to-understand-the-behavior-of-statistics-of-interest"><i class="fa fa-check"></i><b>18.6.2</b> Using sampling distributions to understand the behavior of statistics of interest</a></li>
<li class="chapter" data-level="18.6.3" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#sampling-distributions-for-different-sample-sizes"><i class="fa fa-check"></i><b>18.6.3</b> Sampling distributions for different sample sizes</a></li>
<li class="chapter" data-level="18.6.4" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#discussion-of-trends-for-sampling-distributions-of-different-sample-sizes"><i class="fa fa-check"></i><b>18.6.4</b> Discussion of trends for sampling distributions of different sample sizes</a></li>
</ul></li>
<li class="chapter" data-level="18.7" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#standard-error-of-the-mean"><i class="fa fa-check"></i><b>18.7</b> Standard Error of the Mean</a></li>
<li class="chapter" data-level="18.8" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#sampling-distribution-of-the-standard-deviation"><i class="fa fa-check"></i><b>18.8</b> Sampling Distribution of the Standard Deviation</a><ul>
<li class="chapter" data-level="18.8.1" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#standard-error-of-standard-deviations"><i class="fa fa-check"></i><b>18.8.1</b> Standard error of standard deviations</a></li>
</ul></li>
<li class="chapter" data-level="18.9" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#what-happens-to-the-sampling-distribution-of-the-mean-and-standard-deviation-when-our-sample-size-is-small"><i class="fa fa-check"></i><b>18.9</b> What happens to the sampling distribution of the mean and standard deviation when our sample size is small?</a><ul>
<li class="chapter" data-level="18.9.1" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#for-small-samples-sample-standard-deviations-systematically-underestimate-the-population-standard-deviation"><i class="fa fa-check"></i><b>18.9.1</b> For small samples, sample standard deviations systematically underestimate the population standard deviation</a></li>
<li class="chapter" data-level="18.9.2" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#underestimates-of-the-standard-deviation-for-small-n-lead-to-understimates-of-the-se-of-the-mean"><i class="fa fa-check"></i><b>18.9.2</b> Underestimates of the standard deviation for small <span class="math inline">\(n\)</span> lead to understimates of the SE of the mean</a></li>
<li class="chapter" data-level="18.9.3" data-path="simulation-i-sampling-distributions.html"><a href="simulation-i-sampling-distributions.html#the-t-distribution-is-the-appropriate-distribution-for-describing-the-sampling-distribution-of-the-mean-when-n-is-small"><i class="fa fa-check"></i><b>18.9.3</b> The t-distribution is the appropriate distribution for describing the sampling distribution of the mean when <span class="math inline">\(n\)</span> is small</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Biology 723: Statistical Computing for Biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="introduction-to-dplyr" class="section level1">
<h1><span class="header-section-number">Chapter 6</span> Introduction to <code>dplyr</code></h1>
<p>In today’s class we introduce a new package, <code>dplyr</code>, which, along with ggplot2 will be used in almost every class session. We will also introduce the <code>readr</code> package, for reading tabular data.</p>
<div id="libraries" class="section level2">
<h2><span class="header-section-number">6.1</span> Libraries</h2>
<p>Both <code>readr</code> and <code>dplyr</code> are members of the tidyverse, so a single invocation of <code>library()</code> makes the functions defined in these two packages available for our use:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)  </code></pre>
</div>
<div id="reading-data-with-the-readr-package" class="section level2">
<h2><span class="header-section-number">6.2</span> Reading data with the <code>readr</code> package</h2>
<p>The <a href="http://readr.tidyverse.org"><code>readr</code></a> package defines a number of functions for reading data tables from common file formats like Comma-Separated-Value (CSV) and Tab-Separated-Value (TSV) files.</p>
<p>The two most frequently used <code>readr</code> functions we’ll use in this class are <code>read_csv()</code> and <code>read_tsv()</code> for reading CSV and TSV files respectively. There are some variants of these basic function, which you can read about by invoking the help system (<code>?read_csv</code>).</p>
<div id="reading-excel-files" class="section level3">
<h3><span class="header-section-number">6.2.1</span> Reading Excel files</h3>
<p>The tidyverse also includes a package called <a href="http://readxl.tidyverse.org"><code>readxl</code></a> which can be used to read Excel spreadsheets (recent versions with <code>.xls</code> and <code>.xlsx</code> extensions). Excel files are somewhat more complicated to deal with because they can include separate “sheets”. We won’t use <code>readxl</code> in this class, but documentation and examples of how <code>readxl</code> is used can be found at the page linked above.</p>
</div>
<div id="example-data-nc-births" class="section level3">
<h3><span class="header-section-number">6.2.2</span> Example data: NC Births</h3>
<p>For today’s hands on session we’ll use a data set that contains information on 150 cases of mothers and their newborns in North Carolina in 2004. This data set is available at the following URL:</p>
<ul>
<li><a href="https://github.com/Bio723-class/example-datasets/raw/master/nc-births.txt" class="uri">https://github.com/Bio723-class/example-datasets/raw/master/nc-births.txt</a></li>
</ul>
<p>The births data is a TSV file, so we’ll use the <code>read_tsv()</code> function to read it:</p>
<pre class="sourceCode r"><code class="sourceCode r">births &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;https://github.com/Bio723-class/example-datasets/raw/master/nc-births.txt&quot;</span>)
<span class="co">## Parsed with column specification:</span>
<span class="co">## cols(</span>
<span class="co">##   fAge = col_double(),</span>
<span class="co">##   mAge = col_double(),</span>
<span class="co">##   weeks = col_double(),</span>
<span class="co">##   premature = col_character(),</span>
<span class="co">##   visits = col_double(),</span>
<span class="co">##   gained = col_double(),</span>
<span class="co">##   weight = col_double(),</span>
<span class="co">##   sexBaby = col_character(),</span>
<span class="co">##   smoke = col_character()</span>
<span class="co">## )</span></code></pre>
<p>Notice that when you used <code>read_tsv()</code> the function printed information about how it “parsed” the data (i.e. the types it assigned to each of the columns).</p>
<p>The variables in the data set are:</p>
<ul>
<li>father’s age (<code>fAge</code>),</li>
<li>mother’s age (<code>mAge</code>),<br />
</li>
<li>weeks of gestation (<code>weeks</code>)</li>
<li>whether the birth was premature or full term (<code>premature</code>)</li>
<li>number of OB/GYN visits (<code>visits</code>)</li>
<li>mother’s weight gained in pounds (<code>gained</code>)</li>
<li>babies birth weight (<code>weight</code>)</li>
<li>sex of the baby (<code>sexBaby</code>)</li>
<li>whether the mother was a smoker (<code>smoke</code>).</li>
</ul>
<p>Notice too that we read the TSV file directly from a remote location via a URL. If instead, you wanted to load a local file on your computer you would specify the “path” – i.e. the location on your hard drive where you stored the file. For example, here is how I would load the same file if it was stored in the Downloads directory on my Mac laptop:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load the data from a local file</span>
births &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;/Users/pmagwene/Downloads/nc-births.txt&quot;</span>)</code></pre>
</div>
</div>
<div id="a-note-on-tibbles" class="section level2">
<h2><span class="header-section-number">6.3</span> A note on “tibbles”</h2>
<p>You may have noticed that most of the functions defined in tidyverse related packages return not data frames, but rather something called a “tibble”. You can think about tibbles as light-weight data frames. In fact if you ask about the “class” of a tibble you’ll see that it includes <code>data.frame</code> as one of it’s classes as well as <code>tbl</code> and <code>tbl_df</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(births)
<span class="co">## [1] &quot;spec_tbl_df&quot; &quot;tbl_df&quot;      &quot;tbl&quot;         &quot;data.frame&quot;</span></code></pre>
<p>There are some minor differences between data frame and tibbles. For example, tibbles print differently in the console and don’t automatically change variable names and types in the same way that standard data frames do. Usually tibbles can be used wherever a standard data frame is expected, but you may occasionally find a function that only works with a standard data frame. It’s easy to convert a tibble to a standard data frame using the <code>as.data.frame</code> function:</p>
<pre class="sourceCode r"><code class="sourceCode r">births.std.df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(births)</code></pre>
<p>For more details about tibbles, see the <a href="http://r4ds.had.co.nz/tibbles.html">Tibbles chapter in R for Data Analysis</a>.</p>
</div>
<div id="data-filtering-and-transformation-with-dplyr" class="section level2">
<h2><span class="header-section-number">6.4</span> Data filtering and transformation with <code>dplyr</code></h2>
<p><code>dplyr</code> is powerful tool for data filter and transformation. In the same way that <code>ggplot2</code> attempts to provide a “grammar of graphics”, <code>dplyr</code> aims to provide a “grammar of data manipulation”. In today’s material we will see how <code>dplyr</code> complements and simplifies standard data frame indexing and subsetting operations. However, <code>dplyr</code> is focused only on data frames and doesn’t completely replace the basic subsetting operations, and so being adept with both <code>dplyr</code> and the indexing approaches we’ve seen previously is important. If you’re curious about the name “dplyr”, the package’s originator Hadley Wickham says it’s supposed to invoke the idea of pliers for data frames (<a href="https://github.com/tidyverse/dplyr/issues/1857">Github: Meaning of dplyrs name</a>)</p>
</div>
<div id="dplyrs-verbs" class="section level2">
<h2><span class="header-section-number">6.5</span> dplyr’s “verbs”</h2>
<p>The primary functions in the <code>dplyr</code> package can be thought of as a set of “verbs”, each verb corresponding to a common data manipulation task. Some of the most frequently used verbs/functions in <code>dplyr</code> include:</p>
<ul>
<li><code>select</code> – select columns</li>
<li><code>filter</code> – filter rows</li>
<li><code>mutate</code> – create new columns</li>
<li><code>arrange</code>– reorder rows</li>
<li><code>summarize</code> – summarize values</li>
<li><code>group_by</code> – split data frame on some grouping variable. Can be powerfully combined with <code>summarize</code></li>
</ul>
<p>All of these functions return new data frames rather than modifying the existing data frame (though some of the functions support in place modification of data frames via optional arguments).We illustrate these below by example using the NC births data.</p>
<div id="select" class="section level3">
<h3><span class="header-section-number">6.5.1</span> <code>select</code></h3>
<p>The <code>select</code> function subsets the columns (variables) of a data frame. For example, to select just the weeks and weight columns from the births data set we could do:</p>
<pre class="sourceCode r"><code class="sourceCode r">wks.weight &lt;-<span class="st"> </span><span class="kw">select</span>(births, weeks, weight)
<span class="kw">dim</span>(wks.weight)  <span class="co"># dim should be 50 x 2</span>
<span class="co">## [1] 150   2</span>
<span class="kw">head</span>(wks.weight)
<span class="co">## # A tibble: 6 x 2</span>
<span class="co">##   weeks weight</span>
<span class="co">##   &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">## 1    39   6.88</span>
<span class="co">## 2    39   7.69</span>
<span class="co">## 3    40   8.88</span>
<span class="co">## 4    40   9   </span>
<span class="co">## 5    40   7.94</span>
<span class="co">## 6    40   8.25</span></code></pre>
<p>The equivalent using standard indexing would be:</p>
<pre class="sourceCode r"><code class="sourceCode r">wks.wt.alt &lt;-<span class="st"> </span>births[<span class="kw">c</span>(<span class="st">&quot;weeks&quot;</span>, <span class="st">&quot;weight&quot;</span>)]
<span class="kw">dim</span>(wks.wt.alt)
<span class="co">## [1] 150   2</span>
<span class="kw">head</span>(wks.wt.alt)
<span class="co">## # A tibble: 6 x 2</span>
<span class="co">##   weeks weight</span>
<span class="co">##   &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">## 1    39   6.88</span>
<span class="co">## 2    39   7.69</span>
<span class="co">## 3    40   8.88</span>
<span class="co">## 4    40   9   </span>
<span class="co">## 5    40   7.94</span>
<span class="co">## 6    40   8.25</span></code></pre>
<p><strong>Notes</strong>:<br />
* The first argument to all of the <code>dplyr</code> functions is the data frame you’re operating on</p>
<ul>
<li>When using functions defined in <code>dplyr</code> and <code>ggplot2</code> variable names are (usually) not quoted or used with the <code>$</code> operator. This is a design feature of these libraries and makes it easier to carry out interactive analyes because it saves a fair amount of typing.</li>
</ul>
</div>
<div id="filter" class="section level3">
<h3><span class="header-section-number">6.5.2</span> <code>filter</code></h3>
<p>The <code>filter</code> function returns those rows of the data set that meet the given logical criterion.</p>
<p>For example, to get all the premature babies in the data set we could use filter as so:</p>
<pre class="sourceCode r"><code class="sourceCode r">premies &lt;-<span class="st"> </span><span class="kw">filter</span>(births, premature <span class="op">==</span><span class="st"> &quot;premie&quot;</span>)
<span class="kw">dim</span>(premies)
<span class="co">## [1] 21  9</span></code></pre>
<p>The equivalent using standard indexing would be:</p>
<pre class="sourceCode r"><code class="sourceCode r">premies.alt &lt;-<span class="st"> </span>births[births<span class="op">$</span>premature <span class="op">==</span><span class="st"> &quot;premie&quot;</span>,] </code></pre>
<p>The <code>filter</code> function will work with more than one logical argument, and these are joined together using Boolean AND logic (i.e. intersection). For example, to find those babies that were premature <em>and</em> whose mothers were smokers we could do:</p>
<pre class="sourceCode r"><code class="sourceCode r">smoking.premies &lt;-<span class="st"> </span><span class="kw">filter</span>(births, premature <span class="op">==</span><span class="st"> &quot;premie&quot;</span>, smoke <span class="op">==</span><span class="st"> &quot;smoker&quot;</span>)</code></pre>
<p>The equivalent call using standard indexing is:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># don&#39;t forget the trailing comma to indicate rows!</span>
smoking.premies.alt &lt;-<span class="st"> </span>births[(births<span class="op">$</span>premature <span class="op">==</span><span class="st"> &quot;premie&quot;</span>) <span class="op">&amp;</span><span class="st"> </span>(births<span class="op">$</span>smoke <span class="op">==</span><span class="st"> &quot;smoker&quot;</span>),]</code></pre>
<p><code>filter</code> also accepts logical statements chained together using the standard Boolean operators. For example, to find babies who were premature <em>or</em> whose moms were older than 35 you could use the OR operator <code>|</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r">premies.or.oldmom &lt;-<span class="st"> </span><span class="kw">filter</span>(births, premature <span class="op">==</span><span class="st"> &quot;premie&quot;</span> <span class="op">|</span><span class="st"> </span>fAge <span class="op">&gt;</span><span class="st"> </span><span class="dv">35</span>)</code></pre>
</div>
<div id="mutate" class="section level3">
<h3><span class="header-section-number">6.5.3</span> <code>mutate</code></h3>
<p>The <code>mutate</code> function creates a new data frame that is the same as input data frame but with additional variables (columns) as specified by the function arguments. In the example below, I create two new variables, <code>weight.in.kg</code> and a <code>mom.smoked</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># to make code more readable it&#39;s sometime useful to spread out</span>
<span class="co"># function arguments over multiple lines like I&#39;ve done here</span>
births.plus &lt;-<span class="st"> </span><span class="kw">mutate</span>(births, 
                      <span class="dt">weight.in.kg =</span> weight <span class="op">/</span><span class="st"> </span><span class="fl">2.2</span>,
                      <span class="dt">mom.smoked =</span> (smoke <span class="op">==</span><span class="st"> &quot;smoker&quot;</span>))

<span class="kw">head</span>(births.plus)
<span class="co">## # A tibble: 6 x 11</span>
<span class="co">##    fAge  mAge weeks premature visits gained weight sexBaby smoke</span>
<span class="co">##   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;</span>
<span class="co">## 1    31    30    39 full term     13      1   6.88 male    smok…</span>
<span class="co">## 2    34    36    39 full term      5     35   7.69 male    nons…</span>
<span class="co">## 3    36    35    40 full term     12     29   8.88 male    nons…</span>
<span class="co">## 4    41    40    40 full term     13     30   9    female  nons…</span>
<span class="co">## 5    42    37    40 full term     NA     10   7.94 male    nons…</span>
<span class="co">## 6    37    28    40 full term     12     35   8.25 male    smok…</span>
<span class="co">## # … with 2 more variables: weight.in.kg &lt;dbl&gt;, mom.smoked &lt;lgl&gt;</span></code></pre>
<p>The equivalent using standard indexing would be to create a new data frame from <code>births</code>, appending the new variables to the end as so:</p>
<pre class="sourceCode r"><code class="sourceCode r">births.plus.alt &lt;-<span class="st"> </span><span class="kw">data.frame</span>(births,
                              <span class="dt">weight.in.kg =</span> births<span class="op">$</span>weight <span class="op">/</span><span class="st"> </span><span class="fl">2.2</span>,
                              <span class="dt">mom.smoked =</span> (births<span class="op">$</span>smoke <span class="op">==</span><span class="st"> &quot;smoker&quot;</span>))</code></pre>
</div>
<div id="arrange" class="section level3">
<h3><span class="header-section-number">6.5.4</span> <code>arrange</code></h3>
<p>Arrange creates a new data frame where the rows are sorted according to their values for one or more variables. For example, to sort by mothers age we could do:</p>
<pre class="sourceCode r"><code class="sourceCode r">young.moms.first &lt;-<span class="st"> </span><span class="kw">arrange</span>(births, mAge)
<span class="kw">head</span>(young.moms.first)
<span class="co">## # A tibble: 6 x 9</span>
<span class="co">##    fAge  mAge weeks premature visits gained weight sexBaby smoke    </span>
<span class="co">##   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;    </span>
<span class="co">## 1    18    15    37 full term     12     76   8.44 male    nonsmoker</span>
<span class="co">## 2    NA    16    40 full term      4     12   6    female  nonsmoker</span>
<span class="co">## 3    21    16    38 full term     15     75   7.56 female  smoker   </span>
<span class="co">## 4    26    17    38 full term     11     30   9.5  female  nonsmoker</span>
<span class="co">## 5    17    17    29 premie         4     10   2.63 female  nonsmoker</span>
<span class="co">## 6    20    17    40 full term     17     38   7.19 male    nonsmoker</span></code></pre>
<p>The equivalent to <code>arrange</code> using standard indexing would be to use the information returned by the <code>order</code> function:</p>
<pre class="sourceCode r"><code class="sourceCode r">young.moms.first.alt &lt;-<span class="st"> </span>births[<span class="kw">order</span>(births<span class="op">$</span>mAge),]
<span class="kw">head</span>(young.moms.first.alt)
<span class="co">## # A tibble: 6 x 9</span>
<span class="co">##    fAge  mAge weeks premature visits gained weight sexBaby smoke    </span>
<span class="co">##   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;    </span>
<span class="co">## 1    18    15    37 full term     12     76   8.44 male    nonsmoker</span>
<span class="co">## 2    NA    16    40 full term      4     12   6    female  nonsmoker</span>
<span class="co">## 3    21    16    38 full term     15     75   7.56 female  smoker   </span>
<span class="co">## 4    26    17    38 full term     11     30   9.5  female  nonsmoker</span>
<span class="co">## 5    17    17    29 premie         4     10   2.63 female  nonsmoker</span>
<span class="co">## 6    20    17    40 full term     17     38   7.19 male    nonsmoker</span></code></pre>
<p>When using <code>arrange</code>, multiple sorting variables can be specified:</p>
<pre class="sourceCode r"><code class="sourceCode r">sorted.by.moms.and.dads &lt;-<span class="st"> </span><span class="kw">arrange</span>(births, mAge, fAge)
<span class="kw">head</span>(sorted.by.moms.and.dads)
<span class="co">## # A tibble: 6 x 9</span>
<span class="co">##    fAge  mAge weeks premature visits gained weight sexBaby smoke    </span>
<span class="co">##   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;    </span>
<span class="co">## 1    18    15    37 full term     12     76   8.44 male    nonsmoker</span>
<span class="co">## 2    21    16    38 full term     15     75   7.56 female  smoker   </span>
<span class="co">## 3    NA    16    40 full term      4     12   6    female  nonsmoker</span>
<span class="co">## 4    17    17    29 premie         4     10   2.63 female  nonsmoker</span>
<span class="co">## 5    20    17    40 full term     17     38   7.19 male    nonsmoker</span>
<span class="co">## 6    26    17    38 full term     11     30   9.5  female  nonsmoker</span></code></pre>
<p>If you want to sort in descending order, you can combing <code>arrange</code> with the <code>desc</code> (=descend) function, also defined in <code>dplyr</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r">old.moms.first &lt;-<span class="st"> </span><span class="kw">arrange</span>(births, <span class="kw">desc</span>(mAge))
<span class="kw">head</span>(old.moms.first)
<span class="co">## # A tibble: 6 x 9</span>
<span class="co">##    fAge  mAge weeks premature visits gained weight sexBaby smoke    </span>
<span class="co">##   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;    </span>
<span class="co">## 1    NA    41    33 premie        13      0   5.69 female  nonsmoker</span>
<span class="co">## 2    41    40    40 full term     13     30   9    female  nonsmoker</span>
<span class="co">## 3    33    40    36 premie        13     23   7.81 female  nonsmoker</span>
<span class="co">## 4    40    40    38 full term     13     38   7.31 male    nonsmoker</span>
<span class="co">## 5    46    39    38 full term     10     35   6.75 male    smoker   </span>
<span class="co">## 6    NA    38    32 premie        10     16   2.19 female  smoker</span></code></pre>
</div>
<div id="summarize" class="section level3">
<h3><span class="header-section-number">6.5.5</span> <code>summarize</code></h3>
<p><code>summarize</code> applies a function of interest to one or more variables in a data frame, reducing a vector of values to a single value and returning the results in a data frame. This is most often used to calculate statistics like means, medians, count, etc. As we’ll see below, this is powerful when combined with the <code>group_by</code> function.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summarize</span>(births, 
          <span class="dt">mean.wt =</span> <span class="kw">mean</span>(weight), 
          <span class="dt">median.wks =</span> <span class="kw">median</span>(weeks))
<span class="co">## # A tibble: 1 x 2</span>
<span class="co">##   mean.wt median.wks</span>
<span class="co">##     &lt;dbl&gt;      &lt;dbl&gt;</span>
<span class="co">## 1   7.046         39</span></code></pre>
<p>You’ll need to be diligent if your data has missing values (NAs). For example, by default the <code>mean</code> function returns NA if any of the input values are NA:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summarize</span>(births, 
          <span class="dt">mean.gained =</span> <span class="kw">mean</span>(gained))
<span class="co">## # A tibble: 1 x 1</span>
<span class="co">##   mean.gained</span>
<span class="co">##         &lt;dbl&gt;</span>
<span class="co">## 1          NA</span></code></pre>
<p>However, if you read the <code>mean</code> docs (<code>?mean</code>) you’ll see that there is an <code>na.rm</code> argument that indicates whether NA values should be removed before computing the mean. This is what we want so we instead call summarize as follows:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summarize</span>(births, 
          <span class="dt">mean.gained =</span> <span class="kw">mean</span>(gained, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>))
<span class="co">## # A tibble: 1 x 1</span>
<span class="co">##   mean.gained</span>
<span class="co">##         &lt;dbl&gt;</span>
<span class="co">## 1       32.45</span></code></pre>
</div>
<div id="group_by" class="section level3">
<h3><span class="header-section-number">6.5.6</span> <code>group_by</code></h3>
<p>The <code>group_by</code> function implicitly adds grouping information to a data frame.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># group the births by whether mom smoked or not</span>
by_smoking &lt;-<span class="st"> </span><span class="kw">group_by</span>(births, smoke)</code></pre>
<p>The object returned by <code>group_by</code> is a “grouped data frame”:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(by_smoking)
<span class="co">## [1] &quot;grouped_df&quot; &quot;tbl_df&quot;     &quot;tbl&quot;        &quot;data.frame&quot;</span></code></pre>
<p>Some functions, like <code>count()</code> and <code>summarize()</code> (see below) know how to use the grouping information. For example, to count the number of births conditional on mother smoking status we could do:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">count</span>(by_smoking)
<span class="co">## # A tibble: 2 x 2</span>
<span class="co">##   smoke         n</span>
<span class="co">##   &lt;chr&gt;     &lt;int&gt;</span>
<span class="co">## 1 nonsmoker   100</span>
<span class="co">## 2 smoker       50</span></code></pre>
<p><code>group_by</code> also works with multiple grouping variables, with each added grouping variable specified as an additional argument:</p>
<pre class="sourceCode r"><code class="sourceCode r">by_smoking.and.mAge &lt;-<span class="st"> </span><span class="kw">group_by</span>(births, smoke, mAge <span class="op">&gt;</span><span class="st"> </span><span class="dv">35</span>)</code></pre>
</div>
<div id="combining-grouping-and-summarizing" class="section level3">
<h3><span class="header-section-number">6.5.7</span> Combining grouping and summarizing</h3>
<p>Grouped data frames can be combined with the <code>summarize</code> function we saw above. For example, if we wanted to calculate mean birth weight, broken down by whether the baby’s mother smoked or not we could call <code>summarize</code> with our <code>by_smoking</code> grouped data frame:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summarize</span>(by_smoking, <span class="dt">mean.wt =</span> <span class="kw">mean</span>(weight))
<span class="co">## # A tibble: 2 x 2</span>
<span class="co">##   smoke     mean.wt</span>
<span class="co">##   &lt;chr&gt;       &lt;dbl&gt;</span>
<span class="co">## 1 nonsmoker   7.180</span>
<span class="co">## 2 smoker      6.779</span></code></pre>
<p>Similarly to get the mean birth weight of children conditioned on mothers smoking status and age:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summarize</span>(by_smoking.and.mAge, <span class="kw">mean</span>(weight))
<span class="co">## # A tibble: 4 x 3</span>
<span class="co">##   smoke     `mAge &gt; 35` `mean(weight)`</span>
<span class="co">##   &lt;chr&gt;     &lt;lgl&gt;                &lt;dbl&gt;</span>
<span class="co">## 1 nonsmoker FALSE                7.171</span>
<span class="co">## 2 nonsmoker TRUE                 7.258</span>
<span class="co">## 3 smoker    FALSE                6.832</span>
<span class="co">## 4 smoker    TRUE                 6.302</span></code></pre>
</div>
<div id="scoped-variants-of-mutate-and-summarize" class="section level3">
<h3><span class="header-section-number">6.5.8</span> Scoped variants of <code>mutate</code> and <code>summarize</code></h3>
<p>Both the <code>mutate()</code> and <code>summarize()</code> functions provide “scoped” alternatives, that allow us to apply the operation on a selection of variables. These variants are often used in combination with grouping. We’ll look at the <code>summarize</code> versions – <code>summarize_all()</code>, <code>summarize_at()</code>, and <code>summarize_if()</code>. See the documentation (<code>?mutate_all</code>) for descriptions of the <code>mutate</code> versions.</p>
<div id="summarize_all" class="section level4">
<h4><span class="header-section-number">6.5.8.1</span> <code>summarize_all()</code></h4>
<p><code>summarize_all()</code> applies a one or more functions to all columns in a data frame. Here we illustrate a simple version of this with the iris data:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># group by species</span>
by_species &lt;-<span class="st"> </span><span class="kw">group_by</span>(iris, Species)
<span class="co"># calculate the mean of every variable, grouped by species</span>
<span class="kw">summarize_all</span>(by_species, mean)  
<span class="co">## # A tibble: 3 x 5</span>
<span class="co">##   Species    Sepal.Length Sepal.Width Petal.Length Petal.Width</span>
<span class="co">##   &lt;fct&gt;             &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;</span>
<span class="co">## 1 setosa            5.006       3.428        1.462       0.246</span>
<span class="co">## 2 versicolor        5.936       2.77         4.26        1.326</span>
<span class="co">## 3 virginica         6.588       2.974        5.552       2.026</span></code></pre>
<p>Note that if we try and apply <code>summarize_all()</code> in the same way to the grouped data frame <code>by_smoking</code> we’ll get a bunch of warning messages:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summarize_all</span>(by_smoking, mean)
<span class="co">## # A tibble: 2 x 9</span>
<span class="co">##   smoke      fAge  mAge weeks premature visits gained weight sexBaby</span>
<span class="co">##   &lt;chr&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">## 1 nonsmoker    NA  26.9 38.55        NA   NA       NA  7.180      NA</span>
<span class="co">## 2 smoker       NA  26   38.54        NA   10.8     NA  6.779      NA</span></code></pre>
<p>Here’s an example of one of these warnings:</p>
<pre><code>Warning messages:
1: In mean.default(premature) :
  argument is not numeric or logical: returning NA</code></pre>
<p>This message is telling us that we can’t apply the <code>mean()</code> function to the data frame column <code>premature</code> because this is not a numerical or logical vector. Despite this and the other similar warnings, <code>summarize_all()</code> does return a result, but the means for any non-numeric values are replaced with NAs, as shown below:</p>
<pre><code># A tibble: 2 x 9
  smoke      fAge  mAge weeks premature visits gained weight sexBaby
  &lt;chr&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;
1 nonsmoker    NA  26.9  38.6        NA   NA       NA   7.18      NA
2 smoker       NA  26.0  38.5        NA   10.8     NA   6.78      NA</code></pre>
<p>If you examine the output above, you’ll see that there are several variables that are numeric, however we still got NAs when we calculated the grouped means. This is because those variables contain NA values. The <code>mean</code> function has an optional argument, <code>na.rm</code>, which tells the function to remove any missing data before calculating the mean. Thus we can modify our call to <code>summarize_all</code> as follows:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># calculate mean of all variables, grouped by smoking status</span>
<span class="kw">summarize_all</span>(by_smoking, mean, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)
<span class="co">## # A tibble: 2 x 9</span>
<span class="co">##   smoke      fAge  mAge weeks premature visits gained weight sexBaby</span>
<span class="co">##   &lt;chr&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">## 1 nonsmoker 29.81  26.9 38.55        NA  11.86  32.55  7.180      NA</span>
<span class="co">## 2 smoker    29.71  26   38.54        NA  10.8   32.27  6.779      NA</span></code></pre>
<p>Note that the non-numeric data columns still lead to NA values.</p>
</div>
<div id="summarize_if" class="section level4">
<h4><span class="header-section-number">6.5.8.2</span> <code>summarize_if()</code></h4>
<p><code>summarize_if()</code> is similar to <code>summarize_all()</code>, except it only applies the function of interest to those variables that match a particular predicate (i.e. are TRUE for a particular TRUE/FALSE test).</p>
<p>Here we use <code>summarize_if()</code> to apply the <code>mean()</code> function to only those variables (columns) that are numeric.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># calculate mean of all numeric variables, grouped by smoking status</span>
<span class="kw">summarize_if</span>(by_smoking, is.numeric, mean, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)
<span class="co">## # A tibble: 2 x 7</span>
<span class="co">##   smoke      fAge  mAge weeks visits gained weight</span>
<span class="co">##   &lt;chr&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">## 1 nonsmoker 29.81  26.9 38.55  11.86  32.55  7.180</span>
<span class="co">## 2 smoker    29.71  26   38.54  10.8   32.27  6.779</span></code></pre>
</div>
<div id="summarize_at" class="section level4">
<h4><span class="header-section-number">6.5.8.3</span> <code>summarize_at()</code></h4>
<p><code>summarize_at()</code> allows us to apply functions of interest only to specific variables.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># calculate mean of gained and weight variables, grouped by smoking status</span>
<span class="kw">summarize_at</span>(by_smoking, <span class="kw">c</span>(<span class="st">&quot;gained&quot;</span>, <span class="st">&quot;weight&quot;</span>), mean, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)
<span class="co">## # A tibble: 2 x 3</span>
<span class="co">##   smoke     gained weight</span>
<span class="co">##   &lt;chr&gt;      &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">## 1 nonsmoker  32.55  7.180</span>
<span class="co">## 2 smoker     32.27  6.779</span></code></pre>
<p>All three of the scoped summarize functions can also be used to apply multiple functions, by wrapping the function names in a call to <code>dplyr::funs()</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># calculate mean and std deviation of </span>
<span class="co"># gained and weight variables, grouped by smoking status</span>
<span class="kw">summarize_at</span>(by_smoking, <span class="kw">c</span>(<span class="st">&quot;gained&quot;</span>, <span class="st">&quot;weight&quot;</span>), <span class="kw">funs</span>(mean, sd), <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)
<span class="co">## # A tibble: 2 x 5</span>
<span class="co">##   smoke     gained_mean weight_mean gained_sd weight_sd</span>
<span class="co">##   &lt;chr&gt;           &lt;dbl&gt;       &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span>
<span class="co">## 1 nonsmoker       32.55       7.180     15.23     1.434</span>
<span class="co">## 2 smoker          32.27       6.779     16.65     1.597</span></code></pre>
<p><code>summarize_at()</code> accepts as the the argument for variables a character vector of column names, a numeric vector of column positions, or a list of columns generated by the <code>dplyr::vars()</code> function, which can be be used as so:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># reformatted to promote readability of arguments</span>
<span class="kw">summarize_at</span>(by_smoking, 
             <span class="kw">vars</span>(gained, weight),
             <span class="kw">funs</span>(mean, sd),
             <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)
<span class="co">## # A tibble: 2 x 5</span>
<span class="co">##   smoke     gained_mean weight_mean gained_sd weight_sd</span>
<span class="co">##   &lt;chr&gt;           &lt;dbl&gt;       &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span>
<span class="co">## 1 nonsmoker       32.55       7.180     15.23     1.434</span>
<span class="co">## 2 smoker          32.27       6.779     16.65     1.597</span></code></pre>
</div>
</div>
<div id="combining-summarize-with-grouping-aesthetics-in-ggplot2" class="section level3">
<h3><span class="header-section-number">6.5.9</span> Combining summarize with grouping aesthetics in <code>ggplot2</code></h3>
<p>We’ve already seen an instance of grouping (conditioning) when we used aesthetics like color or fill to distinguish subgroups in different types of statistical graphics. Below is an example where we integrate information from a <code>group_by</code>/<code>summarize</code> operation into a plot:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># calculate mean weights, conditioned on smoking status</span>
wt.by.smoking &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw">summarize</span>(by_smoking, <span class="dt">mean_weight =</span> <span class="kw">mean</span>(weight, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>))

<span class="co"># create density plot for all the data</span>
<span class="co"># and then use geom_vline to draw vertical lines at the means for</span>
<span class="co"># each group</span>
<span class="kw">ggplot</span>(births) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_density</span>(<span class="kw">aes</span>(<span class="dt">x =</span> weight, <span class="dt">color =</span> smoke)) <span class="op">+</span><span class="st"> </span><span class="co"># data drawn from births</span>
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">data =</span> wt.by.smoking,  <span class="co"># note use of different data frame!</span>
             <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">xintercept =</span> mean_weight, <span class="dt">color =</span> smoke),
             <span class="dt">linetype =</span> <span class="st">&#39;dashed&#39;</span>)</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-194-1.png" width="576" /></p>
</div>
</div>
<div id="pipes" class="section level2">
<h2><span class="header-section-number">6.6</span> Pipes</h2>
<p><code>dplyr</code> includes a very useful operator available called a pipe available to us. Pipes are powerful because they allow us to chain together sets of operations in a very intuitive fashion while minimizing nested function calls. We can think of pipes as taking the output of one function and feeding it as the <em>first argument</em> to another function call, where we’ve already specified the subsequent arguments.</p>
<p>Pipes are actually defined in another packaged called <code>magrittr</code>. We’ll look at the basic pipe operator and then look at a few additional “special” pipes that <code>magrittr</code> provides.</p>
<div id="install-and-load-magrittr" class="section level3">
<h3><span class="header-section-number">6.6.1</span> Install and load <code>magrittr</code></h3>
<p>In magrittr in not already installed, install it via the command line or the RStudio GUI. Having done so, you will need to load magrittr via the <code>library()</code> function:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(magrittr)</code></pre>
</div>
<div id="the-basic-pipe-operator" class="section level3">
<h3><span class="header-section-number">6.6.2</span> The basic pipe operator</h3>
<p>The pipe operator is designated by <code>%&gt;%</code>. Using pipes, the expression <code>x %&gt;% f()</code> is equivalent to <code>f(x)</code> and the expression <code>x %&gt;% f(y)</code> is equivalent to <code>f(x,y)</code>. The documentation on pipes (see <code>?magrittr</code>) uses the notation <code>lhs %&gt;% rhs</code> where <code>lhs</code> and <code>rhs</code> are short for “left-hand side” and “right-hand side” respectively. I’ll use this same notation in some of the explanations that follow.</p>
<pre class="sourceCode r"><code class="sourceCode r">births <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">head</span>()   <span class="co"># same as head(births)</span>
<span class="co">## # A tibble: 6 x 9</span>
<span class="co">##    fAge  mAge weeks premature visits gained weight sexBaby smoke    </span>
<span class="co">##   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;    </span>
<span class="co">## 1    31    30    39 full term     13      1   6.88 male    smoker   </span>
<span class="co">## 2    34    36    39 full term      5     35   7.69 male    nonsmoker</span>
<span class="co">## 3    36    35    40 full term     12     29   8.88 male    nonsmoker</span>
<span class="co">## 4    41    40    40 full term     13     30   9    female  nonsmoker</span>
<span class="co">## 5    42    37    40 full term     NA     10   7.94 male    nonsmoker</span>
<span class="co">## 6    37    28    40 full term     12     35   8.25 male    smoker</span>
births <span class="op">%&gt;%</span><span class="st"> </span>head     <span class="co"># you can even leave the parentheses out</span>
<span class="co">## # A tibble: 6 x 9</span>
<span class="co">##    fAge  mAge weeks premature visits gained weight sexBaby smoke    </span>
<span class="co">##   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;    </span>
<span class="co">## 1    31    30    39 full term     13      1   6.88 male    smoker   </span>
<span class="co">## 2    34    36    39 full term      5     35   7.69 male    nonsmoker</span>
<span class="co">## 3    36    35    40 full term     12     29   8.88 male    nonsmoker</span>
<span class="co">## 4    41    40    40 full term     13     30   9    female  nonsmoker</span>
<span class="co">## 5    42    37    40 full term     NA     10   7.94 male    nonsmoker</span>
<span class="co">## 6    37    28    40 full term     12     35   8.25 male    smoker</span>
births <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">head</span>(<span class="dv">10</span>) <span class="co"># same as head(births, 10)</span>
<span class="co">## # A tibble: 10 x 9</span>
<span class="co">##     fAge  mAge weeks premature visits gained weight sexBaby smoke    </span>
<span class="co">##    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;    </span>
<span class="co">##  1    31    30    39 full term     13      1   6.88 male    smoker   </span>
<span class="co">##  2    34    36    39 full term      5     35   7.69 male    nonsmoker</span>
<span class="co">##  3    36    35    40 full term     12     29   8.88 male    nonsmoker</span>
<span class="co">##  4    41    40    40 full term     13     30   9    female  nonsmoker</span>
<span class="co">##  5    42    37    40 full term     NA     10   7.94 male    nonsmoker</span>
<span class="co">##  6    37    28    40 full term     12     35   8.25 male    smoker   </span>
<span class="co">##  7    35    35    28 premie         6     29   1.63 female  nonsmoker</span>
<span class="co">##  8    28    21    35 premie         9     15   5.5  female  smoker   </span>
<span class="co">##  9    22    20    32 premie         5     40   2.69 male    smoker   </span>
<span class="co">## 10    36    25    40 full term     13     34   8.75 female  nonsmoker</span></code></pre>
<p>Multiple pipes can be chained together, such that <code>x %&gt;% f() %&gt;% g() %&gt;% h()</code> is equivalent to <code>h(g(f(x)))</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># equivalent to: head(arrange(births, weight), 10)</span>
births <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">arrange</span>(weight) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">head</span>(<span class="dv">10</span>) 
<span class="co">## # A tibble: 10 x 9</span>
<span class="co">##     fAge  mAge weeks premature visits gained weight sexBaby smoke    </span>
<span class="co">##    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;    </span>
<span class="co">##  1    35    35    28 premie         6     29   1.63 female  nonsmoker</span>
<span class="co">##  2    NA    18    33 premie         7     40   1.69 male    smoker   </span>
<span class="co">##  3    NA    38    32 premie        10     16   2.19 female  smoker   </span>
<span class="co">##  4    17    17    29 premie         4     10   2.63 female  nonsmoker</span>
<span class="co">##  5    22    20    32 premie         5     40   2.69 male    smoker   </span>
<span class="co">##  6    38    37    26 premie         5     25   3.63 male    nonsmoker</span>
<span class="co">##  7    25    22    34 premie        10     20   3.75 male    nonsmoker</span>
<span class="co">##  8    NA    24    38 full term     16     50   3.75 female  nonsmoker</span>
<span class="co">##  9    30    25    35 premie        15     40   4.5  male    smoker   </span>
<span class="co">## 10    19    20    34 premie        13      6   4.5  male    nonsmoker</span></code></pre>
<p>When there are multiple piping operations, I like to arrange the statements vertically to help emphasize the flow of processing and to facilitate debugging and/or modification. I would usually rearrange the above code block as follows:</p>
<pre class="sourceCode r"><code class="sourceCode r">births <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">arrange</span>(weight) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">head</span>(<span class="dv">10</span>)
<span class="co">## # A tibble: 10 x 9</span>
<span class="co">##     fAge  mAge weeks premature visits gained weight sexBaby smoke    </span>
<span class="co">##    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;    </span>
<span class="co">##  1    35    35    28 premie         6     29   1.63 female  nonsmoker</span>
<span class="co">##  2    NA    18    33 premie         7     40   1.69 male    smoker   </span>
<span class="co">##  3    NA    38    32 premie        10     16   2.19 female  smoker   </span>
<span class="co">##  4    17    17    29 premie         4     10   2.63 female  nonsmoker</span>
<span class="co">##  5    22    20    32 premie         5     40   2.69 male    smoker   </span>
<span class="co">##  6    38    37    26 premie         5     25   3.63 male    nonsmoker</span>
<span class="co">##  7    25    22    34 premie        10     20   3.75 male    nonsmoker</span>
<span class="co">##  8    NA    24    38 full term     16     50   3.75 female  nonsmoker</span>
<span class="co">##  9    30    25    35 premie        15     40   4.5  male    smoker   </span>
<span class="co">## 10    19    20    34 premie        13      6   4.5  male    nonsmoker</span></code></pre>
</div>
<div id="an-example-without-pipes" class="section level3">
<h3><span class="header-section-number">6.6.3</span> An example without pipes</h3>
<p>To illustrate how pipes help us, first let’s look at an example set of analysis steps without using pipes. Let’s say we wanted to explore the relationship between father’s age and baby’s birth weight. We’ll start this process of exploration by generating a bivariate scatter plot. Being good scientists we want to express our data in SI units, so we’ll need to converts pounds to kilograms. You’ll also recall that a number of the cases have missing data on father’s age, so we’ll want to remove those before we plot them. Here’s how we might accomplish these steps:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># add a new column for weight in kg</span>
births.kg &lt;-<span class="st"> </span><span class="kw">mutate</span>(births, <span class="dt">weight.kg =</span> weight <span class="op">/</span><span class="st"> </span><span class="fl">2.2</span>)

<span class="co"># filter out the NA fathers</span>
filtered.births &lt;-<span class="st"> </span><span class="kw">filter</span>(births.kg, <span class="op">!</span><span class="kw">is.na</span>(fAge))

<span class="co"># create our plot</span>
<span class="kw">ggplot</span>(filtered.births, <span class="kw">aes</span>(<span class="dt">x =</span> fAge, <span class="dt">y =</span> weight.kg)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Father&#39;s Age (years)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Birth Weight (kg)&quot;</span>)</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-199-1.png" width="576" /></p>
<p>Notice that we created two “temporary” data frames along the way – <code>births.kg</code> and <code>filtered.births</code>. These probably aren’t of particular interest to us, but we needed to generate them to build the plot we wanted. If you were particularly masochistic you could avoid these temporary data frames by using nested functions call like this:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># You SHOULD NOT write nested code like this.</span>
<span class="co"># Code like this is hard to debug and understand!</span>
<span class="kw">ggplot</span>(<span class="kw">filter</span>(<span class="kw">mutate</span>(births, <span class="dt">weight.kg =</span> weight <span class="op">/</span><span class="st"> </span><span class="fl">2.2</span>), <span class="op">!</span><span class="kw">is.na</span>(fAge)), 
       <span class="kw">aes</span>(<span class="dt">x =</span> fAge, <span class="dt">y =</span> weight.kg)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Father&#39;s Age (years)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Birth Weight (kg)&quot;</span>)</code></pre>
</div>
<div id="the-same-example-using-pipes" class="section level3">
<h3><span class="header-section-number">6.6.4</span> The same example using pipes</h3>
<p>The pipe operator makes the output of one statement (<code>lhs</code>) as the first input of a following function (<code>rhs</code>). This simplifies the above example to:</p>
<pre class="sourceCode r"><code class="sourceCode r">births <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">weight.kg =</span> weight <span class="op">/</span><span class="st"> </span><span class="fl">2.2</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">is.na</span>(fAge)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> fAge, <span class="dt">y =</span> weight.kg)) <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Father&#39;s Age (years)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Birth Weight (kg)&quot;</span>)</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-201-1.png" width="576" /></p>
<p>In the example above, we feed the data frame into the <code>mutate</code> function. <code>mutate</code> expects a data frame as a first argument, and subsequent arguments specify the new variables to be created. <code>births %&gt;% mutate(weight.kg = weight / 2.2)</code> is thus equivalent to <code>mutate(births, weight.kg = weight / 2.2))</code>. We then pipe the output to <code>filter</code>, removing NA fathers, and then pipe that output as the input to ggplot.</p>
<p>As mentioned previously, it’s good coding style to write each discrete step as its own line when using piping. This make it easier to understand what the steps of the analysis are as well as facilitating changes to the code (commenting out lines, adding lines, etc)</p>
</div>
<div id="assigning-the-output-of-a-statement-involving-pipes-to-a-variable" class="section level3">
<h3><span class="header-section-number">6.6.5</span> Assigning the output of a statement involving pipes to a variable</h3>
<p>It’s important to recognize that pipes are simply a convenient way to chain together a series of expression. Just like any other compound expression, the output of a series of pipe statements can be assigned to a variable, like so:</p>
<pre class="sourceCode r"><code class="sourceCode r">stats.old.moms &lt;-
<span class="st">  </span>births <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(mAge <span class="op">&gt;</span><span class="st"> </span><span class="dv">35</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">median.gestation =</span> <span class="kw">median</span>(weeks), 
            <span class="dt">mean.weight =</span> <span class="kw">mean</span>(weight))

stats.old.moms
<span class="co">## # A tibble: 1 x 2</span>
<span class="co">##   median.gestation mean.weight</span>
<span class="co">##              &lt;dbl&gt;       &lt;dbl&gt;</span>
<span class="co">## 1               38       6.939</span></code></pre>
<p>Note that our summary table, <code>stats.old.moms</code>, is itself a data frame.</p>
</div>
<div id="compound-assignment-pipe-operator" class="section level3">
<h3><span class="header-section-number">6.6.6</span> Compound assignment pipe operator</h3>
<p>A fairly common operation when working interactively in R is to update an existing data frame. <code>magrittr</code> defines another pipe operator – <code>%&lt;&gt;%</code> – called the “compound assignment” pipe operator, to facilitate this. The compound assignment pipe operator has the basic usage <code>lhs %&lt;&gt;% rhs</code>. This operator evaluates the function on the <code>rhs</code> using the <code>lhs</code> as the first argument, and <em>then</em> updates the <code>lhs</code> with the resulting value. This is simply shorthand for writing <code>lhs &lt;- lhs %&gt;% rhs</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r">stats.old.moms <span class="op">%&lt;&gt;%</span><span class="st">  </span><span class="co"># note compound pipe operator!</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">mean.weight.kg =</span> mean.weight <span class="op">/</span><span class="st"> </span><span class="fl">2.2</span>)</code></pre>
</div>
<div id="the-dot-operator-with-pipes" class="section level3">
<h3><span class="header-section-number">6.6.7</span> The dot operator with pipes</h3>
<p>When working with pipes, sometimes you’ll want to use the <code>lhs</code> in multiple places on the <code>rhs</code>, or as something other than the first argument to the <code>rhs</code>. <code>magrittr</code> provides for this situation by using the dot (<code>.</code>) operator as a placeholder. Using the dot operator, the expression <code>y %&gt;% f(x, .)</code> is equivalent to <code>f(x,y)</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">c</span>(<span class="st">&quot;dog&quot;</span>, <span class="st">&quot;cakes&quot;</span>, <span class="st">&quot;sauce&quot;</span>, <span class="st">&quot;house&quot;</span>) <span class="op">%&gt;%</span><span class="st">  </span><span class="co"># create a vector</span>
<span class="st">  </span><span class="kw">sample</span>(<span class="dv">1</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># pick a random single element of that vector</span>
<span class="st">  </span><span class="kw">str_c</span>(<span class="st">&quot;hot&quot;</span>, .)  <span class="co"># string concatenate the pick with the word &quot;hot&quot;</span>
<span class="co">## [1] &quot;hotcakes&quot;</span></code></pre>
</div>
<div id="the-exposition-pipe-operator" class="section level3">
<h3><span class="header-section-number">6.6.8</span> The exposition pipe operator</h3>
<p><code>magrittr</code> defines another operator called the “exposition pipe operator”, designed <code>%$%</code>. This operator exposes the names in the <code>lhs</code> to the expression on the <code>rhs</code>.</p>
<p>Here is an example of using the exposition pipe operator to simply return the vector of weights:</p>
<pre class="sourceCode r"><code class="sourceCode r">births <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(premature <span class="op">==</span><span class="st"> &quot;premie&quot;</span>) <span class="op">%$%</span><span class="st">  </span><span class="co"># note the different pipe operator!</span>
<span class="st">  </span>weight
<span class="co">##  [1] 1.63 5.50 2.69 6.50 7.81 4.75 3.75 2.19 6.81 4.69 6.75 4.50 5.94 4.50</span>
<span class="co">## [15] 5.06 5.69 1.69 6.31 2.63 5.88 3.63</span></code></pre>
<p>If we wanted to calculate the minimum and maximum weight of premature babies in the data set we could do the following (though I’d usually prefer <code>summarize()</code> unless I needed the results in the form of a vector):</p>
<pre class="sourceCode r"><code class="sourceCode r">births <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(mAge <span class="op">&gt;</span><span class="st"> </span><span class="dv">35</span>) <span class="op">%$%</span><span class="st">  </span><span class="co"># note the different pipe operator!</span>
<span class="st">  </span><span class="kw">c</span>(<span class="kw">min</span>(weight), <span class="kw">max</span>(weight)) 
<span class="co">## [1]  2.19 10.13</span></code></pre>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction-to-ggplot2.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="data-wrangling.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "section"
},
"toolbar": {
"position": "fixed"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
