<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 12 Linear Regression Models II | Biology 723: Statistical Computing for Biologists</title>
  <meta name="description" content="Hands-on materials for Bio 723 course at Duke University." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 12 Linear Regression Models II | Biology 723: Statistical Computing for Biologists" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 12 Linear Regression Models II | Biology 723: Statistical Computing for Biologists" />
  
  <meta name="twitter:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

<meta name="author" content="Paul M. Magwene" />


<meta name="date" content="2021-04-07" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="linear-regression-models-i.html"/>
<link rel="next" href="principal-components-analysis.html"/>
<script src="libs/header-attrs-2.6/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#accessing-older-versions-of-the-course-notes"><i class="fa fa-check"></i><b>1.1</b> Accessing older versions of the course notes</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#how-to-use-these-lecture-notes"><i class="fa fa-check"></i><b>1.2</b> How to use these lecture notes</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html"><i class="fa fa-check"></i><b>2</b> Getting Started with R</a>
<ul>
<li class="chapter" data-level="2.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-r"><i class="fa fa-check"></i><b>2.1</b> What is R?</a></li>
<li class="chapter" data-level="2.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-rstudio"><i class="fa fa-check"></i><b>2.2</b> What is RStudio?</a></li>
<li class="chapter" data-level="2.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#entering-commands-in-the-console"><i class="fa fa-check"></i><b>2.3</b> Entering commands in the console</a></li>
<li class="chapter" data-level="2.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#comments"><i class="fa fa-check"></i><b>2.4</b> Comments</a></li>
<li class="chapter" data-level="2.5" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#using-r-as-a-calculator"><i class="fa fa-check"></i><b>2.5</b> Using R as a Calculator</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#common-mathematical-functions"><i class="fa fa-check"></i><b>2.5.1</b> Common mathematical functions</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#variable-assignment"><i class="fa fa-check"></i><b>2.6</b> Variable assignment</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#valid-variable-names"><i class="fa fa-check"></i><b>2.6.1</b> Valid variable names</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#data-types"><i class="fa fa-check"></i><b>2.7</b> Data types</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#numeric-data-types"><i class="fa fa-check"></i><b>2.7.1</b> Numeric data types</a></li>
<li class="chapter" data-level="2.7.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#logical-values"><i class="fa fa-check"></i><b>2.7.2</b> Logical values</a></li>
<li class="chapter" data-level="2.7.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#character-strings"><i class="fa fa-check"></i><b>2.7.3</b> Character strings</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#packages"><i class="fa fa-check"></i><b>2.8</b> Packages</a>
<ul>
<li class="chapter" data-level="2.8.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-console"><i class="fa fa-check"></i><b>2.8.1</b> Installing packages from the console</a></li>
<li class="chapter" data-level="2.8.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#install-the-tidyverse-package"><i class="fa fa-check"></i><b>2.8.2</b> Install the tidyverse package</a></li>
<li class="chapter" data-level="2.8.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-rstudio-dialog"><i class="fa fa-check"></i><b>2.8.3</b> Installing packages from the RStudio dialog</a></li>
<li class="chapter" data-level="2.8.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#loading-packages-with-the-library-function"><i class="fa fa-check"></i><b>2.8.4</b> Loading packages with the <code>library()</code> function</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#the-r-help-system"><i class="fa fa-check"></i><b>2.9</b> The R Help System</a>
<ul>
<li class="chapter" data-level="2.9.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#getting-help-from-the-console"><i class="fa fa-check"></i><b>2.9.1</b> Getting help from the console</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html"><i class="fa fa-check"></i><b>3</b> R Markdown and R Notebooks</a>
<ul>
<li class="chapter" data-level="3.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#r-notebooks"><i class="fa fa-check"></i><b>3.1</b> R Notebooks</a></li>
<li class="chapter" data-level="3.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#creating-an-r-notebook"><i class="fa fa-check"></i><b>3.2</b> Creating an R Notebook</a></li>
<li class="chapter" data-level="3.3" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#the-default-r-notebook-template"><i class="fa fa-check"></i><b>3.3</b> The default R Notebook template</a></li>
<li class="chapter" data-level="3.4" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-and-non-code-blocks"><i class="fa fa-check"></i><b>3.4</b> Code and Non-code blocks</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#non-code-blocks"><i class="fa fa-check"></i><b>3.4.1</b> Non-code blocks</a></li>
<li class="chapter" data-level="3.4.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-blocks"><i class="fa fa-check"></i><b>3.4.2</b> Code blocks</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-a-code-chunk"><i class="fa fa-check"></i><b>3.5</b> Running a code chunk</a></li>
<li class="chapter" data-level="3.6" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-all-code-chunks-above"><i class="fa fa-check"></i><b>3.6</b> Running all code chunks above</a></li>
<li class="chapter" data-level="3.7" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#knitting-r-markdown-to-html"><i class="fa fa-check"></i><b>3.7</b> “Knitting” R Markdown to HTML</a></li>
<li class="chapter" data-level="3.8" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#sharing-your-reproducible-r-notebook"><i class="fa fa-check"></i><b>3.8</b> Sharing your reproducible R Notebook</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-structures.html"><a href="data-structures.html"><i class="fa fa-check"></i><b>4</b> Data structures</a>
<ul>
<li class="chapter" data-level="4.1" data-path="data-structures.html"><a href="data-structures.html#vectors"><i class="fa fa-check"></i><b>4.1</b> Vectors</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="data-structures.html"><a href="data-structures.html#vector-arithmetic"><i class="fa fa-check"></i><b>4.1.1</b> Vector Arithmetic</a></li>
<li class="chapter" data-level="4.1.2" data-path="data-structures.html"><a href="data-structures.html#vector-recycling"><i class="fa fa-check"></i><b>4.1.2</b> Vector recycling</a></li>
<li class="chapter" data-level="4.1.3" data-path="data-structures.html"><a href="data-structures.html#simple-statistical-functions-for-numeric-vectors"><i class="fa fa-check"></i><b>4.1.3</b> Simple statistical functions for numeric vectors</a></li>
<li class="chapter" data-level="4.1.4" data-path="data-structures.html"><a href="data-structures.html#indexing-vectors"><i class="fa fa-check"></i><b>4.1.4</b> Indexing Vectors</a></li>
<li class="chapter" data-level="4.1.5" data-path="data-structures.html"><a href="data-structures.html#comparison-operators-applied-to-vectors"><i class="fa fa-check"></i><b>4.1.5</b> Comparison operators applied to vectors</a></li>
<li class="chapter" data-level="4.1.6" data-path="data-structures.html"><a href="data-structures.html#combining-indexing-and-comparison-of-vectors"><i class="fa fa-check"></i><b>4.1.6</b> Combining Indexing and Comparison of Vectors</a></li>
<li class="chapter" data-level="4.1.7" data-path="data-structures.html"><a href="data-structures.html#vector-manipulation"><i class="fa fa-check"></i><b>4.1.7</b> Vector manipulation</a></li>
<li class="chapter" data-level="4.1.8" data-path="data-structures.html"><a href="data-structures.html#vectors-from-regular-sequences"><i class="fa fa-check"></i><b>4.1.8</b> Vectors from regular sequences</a></li>
<li class="chapter" data-level="4.1.9" data-path="data-structures.html"><a href="data-structures.html#additional-functions-for-working-with-vectors"><i class="fa fa-check"></i><b>4.1.9</b> Additional functions for working with vectors</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="data-structures.html"><a href="data-structures.html#lists"><i class="fa fa-check"></i><b>4.2</b> Lists</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="data-structures.html"><a href="data-structures.html#length-and-type-of-lists"><i class="fa fa-check"></i><b>4.2.1</b> Length and type of lists</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-structures.html"><a href="data-structures.html#indexing-lists"><i class="fa fa-check"></i><b>4.2.2</b> Indexing lists</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-structures.html"><a href="data-structures.html#naming-list-elements"><i class="fa fa-check"></i><b>4.2.3</b> Naming list elements</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-structures.html"><a href="data-structures.html#the-operator"><i class="fa fa-check"></i><b>4.2.4</b> The <code>$</code> operator</a></li>
<li class="chapter" data-level="4.2.5" data-path="data-structures.html"><a href="data-structures.html#changing-and-adding-lists-items"><i class="fa fa-check"></i><b>4.2.5</b> Changing and adding lists items</a></li>
<li class="chapter" data-level="4.2.6" data-path="data-structures.html"><a href="data-structures.html#combining-lists"><i class="fa fa-check"></i><b>4.2.6</b> Combining lists</a></li>
<li class="chapter" data-level="4.2.7" data-path="data-structures.html"><a href="data-structures.html#converting-lists-to-vectors"><i class="fa fa-check"></i><b>4.2.7</b> Converting lists to vectors</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-structures.html"><a href="data-structures.html#data-frames"><i class="fa fa-check"></i><b>4.3</b> Data frames</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="data-structures.html"><a href="data-structures.html#creating-a-data-frame"><i class="fa fa-check"></i><b>4.3.1</b> Creating a data frame</a></li>
<li class="chapter" data-level="4.3.2" data-path="data-structures.html"><a href="data-structures.html#type-and-class-for-data-frames"><i class="fa fa-check"></i><b>4.3.2</b> Type and class for data frames</a></li>
<li class="chapter" data-level="4.3.3" data-path="data-structures.html"><a href="data-structures.html#length-and-dimension-for-data-frames"><i class="fa fa-check"></i><b>4.3.3</b> Length and dimension for data frames</a></li>
<li class="chapter" data-level="4.3.4" data-path="data-structures.html"><a href="data-structures.html#indexing-and-accessing-data-frames"><i class="fa fa-check"></i><b>4.3.4</b> Indexing and accessing data frames</a></li>
<li class="chapter" data-level="4.3.5" data-path="data-structures.html"><a href="data-structures.html#logical-indexing-of-data-frames"><i class="fa fa-check"></i><b>4.3.5</b> Logical indexing of data frames</a></li>
<li class="chapter" data-level="4.3.6" data-path="data-structures.html"><a href="data-structures.html#adding-columns-to-a-data-frame"><i class="fa fa-check"></i><b>4.3.6</b> Adding columns to a data frame</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html"><i class="fa fa-check"></i><b>5</b> Functions and control flow statements</a>
<ul>
<li class="chapter" data-level="5.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#writing-your-own-functions"><i class="fa fa-check"></i><b>5.1</b> Writing your own functions</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#function-arguments"><i class="fa fa-check"></i><b>5.1.1</b> Function arguments</a></li>
<li class="chapter" data-level="5.1.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#writing-functions-with-optional-arguments"><i class="fa fa-check"></i><b>5.1.2</b> Writing functions with optional arguments</a></li>
<li class="chapter" data-level="5.1.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#putting-r-functions-in-scripts"><i class="fa fa-check"></i><b>5.1.3</b> Putting R functions in Scripts</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#control-flow-statements"><i class="fa fa-check"></i><b>5.2</b> Control flow statements</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#if-and-if-else-statements"><i class="fa fa-check"></i><b>5.2.1</b> <code>if</code> and <code>if-else</code> statements</a></li>
<li class="chapter" data-level="5.2.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#for-loops"><i class="fa fa-check"></i><b>5.2.2</b> for loops</a></li>
<li class="chapter" data-level="5.2.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#break-statement"><i class="fa fa-check"></i><b>5.2.3</b> <code>break</code> statement</a></li>
<li class="chapter" data-level="5.2.4" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#repeat-loops"><i class="fa fa-check"></i><b>5.2.4</b> <code>repeat</code> loops</a></li>
<li class="chapter" data-level="5.2.5" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#next-statement"><i class="fa fa-check"></i><b>5.2.5</b> <code>next</code> statement</a></li>
<li class="chapter" data-level="5.2.6" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#while-statements"><i class="fa fa-check"></i><b>5.2.6</b> while statements</a></li>
<li class="chapter" data-level="5.2.7" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#ifelse"><i class="fa fa-check"></i><b>5.2.7</b> <code>ifelse</code></a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#map-and-related-tools"><i class="fa fa-check"></i><b>5.3</b> <code>map</code> and related tools</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#basic-map"><i class="fa fa-check"></i><b>5.3.1</b> basic <code>map</code></a></li>
<li class="chapter" data-level="5.3.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#map_if-and-map_at"><i class="fa fa-check"></i><b>5.3.2</b> <code>map_if</code> and <code>map_at</code></a></li>
<li class="chapter" data-level="5.3.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#mapping-in-parallel-using-map2"><i class="fa fa-check"></i><b>5.3.3</b> mapping in parallel using <code>map2</code></a></li>
<li class="chapter" data-level="5.3.4" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#map-variants-that-return-vectors"><i class="fa fa-check"></i><b>5.3.4</b> <code>map</code> variants that return vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html"><i class="fa fa-check"></i><b>6</b> Introduction to ggplot2</a>
<ul>
<li class="chapter" data-level="6.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#loading-ggplot2"><i class="fa fa-check"></i><b>6.1</b> Loading ggplot2</a></li>
<li class="chapter" data-level="6.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#example-data-set-andersons-iris-data"><i class="fa fa-check"></i><b>6.2</b> Example data set: Anderson’s Iris Data</a></li>
<li class="chapter" data-level="6.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#template-for-single-layer-plots-in-ggplot2"><i class="fa fa-check"></i><b>6.3</b> Template for single layer plots in ggplot2</a></li>
<li class="chapter" data-level="6.4" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#an-aside-about-function-arguments"><i class="fa fa-check"></i><b>6.4</b> An aside about function arguments</a></li>
<li class="chapter" data-level="6.5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#strip-plots"><i class="fa fa-check"></i><b>6.5</b> Strip plots</a>
<ul>
<li class="chapter" data-level="6.5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#jittering-data"><i class="fa fa-check"></i><b>6.5.1</b> Jittering data</a></li>
<li class="chapter" data-level="6.5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-categorical-information"><i class="fa fa-check"></i><b>6.5.2</b> Adding categorical information</a></li>
<li class="chapter" data-level="6.5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#rotating-plot-coordinates"><i class="fa fa-check"></i><b>6.5.3</b> Rotating plot coordinates</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#histograms"><i class="fa fa-check"></i><b>6.6</b> Histograms</a>
<ul>
<li class="chapter" data-level="6.6.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#variations-on-histograms-when-considering-categorical-data"><i class="fa fa-check"></i><b>6.6.1</b> Variations on histograms when considering categorical data</a></li>
</ul></li>
<li class="chapter" data-level="6.7" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#faceting-to-depict-categorical-information"><i class="fa fa-check"></i><b>6.7</b> Faceting to depict categorical information</a></li>
<li class="chapter" data-level="6.8" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots"><i class="fa fa-check"></i><b>6.8</b> Density plots</a></li>
<li class="chapter" data-level="6.9" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-or-beanplot"><i class="fa fa-check"></i><b>6.9</b> Violin or Beanplot</a></li>
<li class="chapter" data-level="6.10" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplots"><i class="fa fa-check"></i><b>6.10</b> Boxplots</a></li>
<li class="chapter" data-level="6.11" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#building-complex-visualizations-with-layers"><i class="fa fa-check"></i><b>6.11</b> Building complex visualizations with layers</a></li>
<li class="chapter" data-level="6.12" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#useful-combination-plots"><i class="fa fa-check"></i><b>6.12</b> Useful combination plots</a>
<ul>
<li class="chapter" data-level="6.12.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplot-plus-strip-plot"><i class="fa fa-check"></i><b>6.12.1</b> Boxplot plus strip plot</a></li>
<li class="chapter" data-level="6.12.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#setting-shared-aesthetics"><i class="fa fa-check"></i><b>6.12.2</b> Setting shared aesthetics</a></li>
</ul></li>
<li class="chapter" data-level="6.13" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot-layers-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>6.13</b> ggplot layers can be assigned to variables</a>
<ul>
<li class="chapter" data-level="6.13.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-plot-plus-strip-plot"><i class="fa fa-check"></i><b>6.13.1</b> Violin plot plus strip plot</a></li>
</ul></li>
<li class="chapter" data-level="6.14" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-titles-and-tweaking-axis-labels"><i class="fa fa-check"></i><b>6.14</b> Adding titles and tweaking axis labels</a></li>
<li class="chapter" data-level="6.15" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot2-themes"><i class="fa fa-check"></i><b>6.15</b> ggplot2 themes</a>
<ul>
<li class="chapter" data-level="6.15.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#further-customization-with-ggplot2theme"><i class="fa fa-check"></i><b>6.15.1</b> Further customization with <code>ggplot2::theme</code></a></li>
</ul></li>
<li class="chapter" data-level="6.16" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#other-aspects-of-ggplots-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>6.16</b> Other aspects of ggplots can be assigned to variables</a></li>
<li class="chapter" data-level="6.17" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-plots"><i class="fa fa-check"></i><b>6.17</b> Bivariate plots</a>
<ul>
<li class="chapter" data-level="6.17.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#scatter-plots"><i class="fa fa-check"></i><b>6.17.1</b> Scatter plots</a></li>
<li class="chapter" data-level="6.17.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-a-trend-line-to-a-scatter-plot"><i class="fa fa-check"></i><b>6.17.2</b> Adding a trend line to a scatter plot</a></li>
</ul></li>
<li class="chapter" data-level="6.18" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-density-plots"><i class="fa fa-check"></i><b>6.18</b> Bivariate density plots</a></li>
<li class="chapter" data-level="6.19" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#combining-scatter-plots-and-density-plots-with-categorical-information"><i class="fa fa-check"></i><b>6.19</b> Combining Scatter Plots and Density Plots with Categorical Information</a></li>
<li class="chapter" data-level="6.20" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots-with-fill"><i class="fa fa-check"></i><b>6.20</b> Density plots with fill</a></li>
<li class="chapter" data-level="6.21" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#d-bin-and-hex-plots"><i class="fa fa-check"></i><b>6.21</b> 2D bin and hex plots</a></li>
<li class="chapter" data-level="6.22" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#the-cowplot-package"><i class="fa fa-check"></i><b>6.22</b> The <code>cowplot</code> package</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html"><i class="fa fa-check"></i><b>7</b> Introduction to <code>dplyr</code></a>
<ul>
<li class="chapter" data-level="7.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#libraries"><i class="fa fa-check"></i><b>7.1</b> Libraries</a></li>
<li class="chapter" data-level="7.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#reading-data-with-the-readr-package"><i class="fa fa-check"></i><b>7.2</b> Reading data with the <code>readr</code> package</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#reading-excel-files"><i class="fa fa-check"></i><b>7.2.1</b> Reading Excel files</a></li>
<li class="chapter" data-level="7.2.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#example-data-nc-births"><i class="fa fa-check"></i><b>7.2.2</b> Example data: NC Births</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#a-note-on-tibbles"><i class="fa fa-check"></i><b>7.3</b> A note on “tibbles”</a></li>
<li class="chapter" data-level="7.4" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#data-filtering-and-transformation-with-dplyr"><i class="fa fa-check"></i><b>7.4</b> Data filtering and transformation with <code>dplyr</code></a></li>
<li class="chapter" data-level="7.5" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#dplyrs-verbs"><i class="fa fa-check"></i><b>7.5</b> dplyr’s “verbs”</a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#select"><i class="fa fa-check"></i><b>7.5.1</b> <code>select</code></a></li>
<li class="chapter" data-level="7.5.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#filter"><i class="fa fa-check"></i><b>7.5.2</b> <code>filter</code></a></li>
<li class="chapter" data-level="7.5.3" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#mutate"><i class="fa fa-check"></i><b>7.5.3</b> <code>mutate</code></a></li>
<li class="chapter" data-level="7.5.4" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#arrange"><i class="fa fa-check"></i><b>7.5.4</b> <code>arrange</code></a></li>
<li class="chapter" data-level="7.5.5" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#summarize"><i class="fa fa-check"></i><b>7.5.5</b> <code>summarize</code></a></li>
<li class="chapter" data-level="7.5.6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#group_by"><i class="fa fa-check"></i><b>7.5.6</b> <code>group_by</code></a></li>
<li class="chapter" data-level="7.5.7" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#combining-grouping-and-summarizing"><i class="fa fa-check"></i><b>7.5.7</b> Combining grouping and summarizing</a></li>
<li class="chapter" data-level="7.5.8" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#scoped-variants-of-mutate-and-summarize"><i class="fa fa-check"></i><b>7.5.8</b> Scoped variants of <code>mutate</code> and <code>summarize</code></a></li>
<li class="chapter" data-level="7.5.9" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#combining-summarize-with-grouping-aesthetics-in-ggplot2"><i class="fa fa-check"></i><b>7.5.9</b> Combining summarize with grouping aesthetics in <code>ggplot2</code></a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#pipes"><i class="fa fa-check"></i><b>7.6</b> Pipes</a>
<ul>
<li class="chapter" data-level="7.6.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#install-and-load-magrittr"><i class="fa fa-check"></i><b>7.6.1</b> Install and load <code>magrittr</code></a></li>
<li class="chapter" data-level="7.6.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-basic-pipe-operator"><i class="fa fa-check"></i><b>7.6.2</b> The basic pipe operator</a></li>
<li class="chapter" data-level="7.6.3" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#an-example-without-pipes"><i class="fa fa-check"></i><b>7.6.3</b> An example without pipes</a></li>
<li class="chapter" data-level="7.6.4" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-same-example-using-pipes"><i class="fa fa-check"></i><b>7.6.4</b> The same example using pipes</a></li>
<li class="chapter" data-level="7.6.5" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#assigning-the-output-of-a-statement-involving-pipes-to-a-variable"><i class="fa fa-check"></i><b>7.6.5</b> Assigning the output of a statement involving pipes to a variable</a></li>
<li class="chapter" data-level="7.6.6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#compound-assignment-pipe-operator"><i class="fa fa-check"></i><b>7.6.6</b> Compound assignment pipe operator</a></li>
<li class="chapter" data-level="7.6.7" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-dot-operator-with-pipes"><i class="fa fa-check"></i><b>7.6.7</b> The dot operator with pipes</a></li>
<li class="chapter" data-level="7.6.8" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-exposition-pipe-operator"><i class="fa fa-check"></i><b>7.6.8</b> The exposition pipe operator</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>8</b> Data wrangling</a>
<ul>
<li class="chapter" data-level="8.1" data-path="data-wrangling.html"><a href="data-wrangling.html#libraries-1"><i class="fa fa-check"></i><b>8.1</b> Libraries</a></li>
<li class="chapter" data-level="8.2" data-path="data-wrangling.html"><a href="data-wrangling.html#data"><i class="fa fa-check"></i><b>8.2</b> Data</a></li>
<li class="chapter" data-level="8.3" data-path="data-wrangling.html"><a href="data-wrangling.html#renaming-data-frame-columms"><i class="fa fa-check"></i><b>8.3</b> Renaming data frame columms</a></li>
<li class="chapter" data-level="8.4" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-unneeded-columns"><i class="fa fa-check"></i><b>8.4</b> Dropping unneeded columns</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-all-empty-columns"><i class="fa fa-check"></i><b>8.4.1</b> Finding all empty columns</a></li>
<li class="chapter" data-level="8.4.2" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-columns-by-matching-names"><i class="fa fa-check"></i><b>8.4.2</b> Dropping columns by matching names</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="data-wrangling.html"><a href="data-wrangling.html#merging-data-frames"><i class="fa fa-check"></i><b>8.5</b> Merging data frames</a></li>
<li class="chapter" data-level="8.6" data-path="data-wrangling.html"><a href="data-wrangling.html#reshaping-data-with-tidyr"><i class="fa fa-check"></i><b>8.6</b> Reshaping data with tidyr</a>
<ul>
<li class="chapter" data-level="8.6.1" data-path="data-wrangling.html"><a href="data-wrangling.html#wide-to-long-conversions-using-tidyrgather"><i class="fa fa-check"></i><b>8.6.1</b> Wide to long conversions using <code>tidyr::gather</code></a></li>
<li class="chapter" data-level="8.6.2" data-path="data-wrangling.html"><a href="data-wrangling.html#extracting-information-from-combined-variables-using-tidyrextract"><i class="fa fa-check"></i><b>8.6.2</b> Extracting information from combined variables using <code>tidyr::extract</code></a></li>
<li class="chapter" data-level="8.6.3" data-path="data-wrangling.html"><a href="data-wrangling.html#subsetting-rows"><i class="fa fa-check"></i><b>8.6.3</b> Subsetting rows</a></li>
<li class="chapter" data-level="8.6.4" data-path="data-wrangling.html"><a href="data-wrangling.html#splitting-columns"><i class="fa fa-check"></i><b>8.6.4</b> Splitting columns</a></li>
<li class="chapter" data-level="8.6.5" data-path="data-wrangling.html"><a href="data-wrangling.html#combining-data-frames"><i class="fa fa-check"></i><b>8.6.5</b> Combining data frames</a></li>
<li class="chapter" data-level="8.6.6" data-path="data-wrangling.html"><a href="data-wrangling.html#sorting-data-frame-rows"><i class="fa fa-check"></i><b>8.6.6</b> Sorting data frame rows</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="data-wrangling.html"><a href="data-wrangling.html#using-your-tidy-data"><i class="fa fa-check"></i><b>8.7</b> Using your tidy data</a>
<ul>
<li class="chapter" data-level="8.7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#visualizing-gene-expression-time-series"><i class="fa fa-check"></i><b>8.7.1</b> Visualizing gene expression time series</a></li>
<li class="chapter" data-level="8.7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-the-most-variable-genes"><i class="fa fa-check"></i><b>8.7.2</b> Finding the most variable genes</a></li>
<li class="chapter" data-level="8.7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#heat-maps"><i class="fa fa-check"></i><b>8.7.3</b> Heat maps</a></li>
</ul></li>
<li class="chapter" data-level="8.8" data-path="data-wrangling.html"><a href="data-wrangling.html#long-to-wide-conversion-using-tidyrspread"><i class="fa fa-check"></i><b>8.8</b> Long-to-wide conversion using <code>tidyr::spread</code></a></li>
<li class="chapter" data-level="8.9" data-path="data-wrangling.html"><a href="data-wrangling.html#exploring-bivariate-relationships-using-wide-data"><i class="fa fa-check"></i><b>8.9</b> Exploring bivariate relationships using “wide” data</a>
<ul>
<li class="chapter" data-level="8.9.1" data-path="data-wrangling.html"><a href="data-wrangling.html#large-scale-patterns-of-correlations"><i class="fa fa-check"></i><b>8.9.1</b> Large scale patterns of correlations</a></li>
<li class="chapter" data-level="8.9.2" data-path="data-wrangling.html"><a href="data-wrangling.html#adding-new-columns-and-combining-filtered-data-frames"><i class="fa fa-check"></i><b>8.9.2</b> Adding new columns and combining filtered data frames</a></li>
<li class="chapter" data-level="8.9.3" data-path="data-wrangling.html"><a href="data-wrangling.html#a-heat-mapped-sorted-by-correlations"><i class="fa fa-check"></i><b>8.9.3</b> A heat mapped sorted by correlations</a></li>
<li class="chapter" data-level="8.9.4" data-path="data-wrangling.html"><a href="data-wrangling.html#a-fancy-figure"><i class="fa fa-check"></i><b>8.9.4</b> A “fancy” figure</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="vector-algebra.html"><a href="vector-algebra.html"><i class="fa fa-check"></i><b>9</b> Vector algebra</a>
<ul>
<li class="chapter" data-level="9.1" data-path="vector-algebra.html"><a href="vector-algebra.html#libraries-2"><i class="fa fa-check"></i><b>9.1</b> Libraries</a></li>
<li class="chapter" data-level="9.2" data-path="vector-algebra.html"><a href="vector-algebra.html#vector-mathematics-in-r"><i class="fa fa-check"></i><b>9.2</b> Vector Mathematics in R</a></li>
<li class="chapter" data-level="9.3" data-path="vector-algebra.html"><a href="vector-algebra.html#simple-statistics-in-vector-form"><i class="fa fa-check"></i><b>9.3</b> Simple statistics in vector form</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="vector-algebra.html"><a href="vector-algebra.html#mean"><i class="fa fa-check"></i><b>9.3.1</b> Mean</a></li>
<li class="chapter" data-level="9.3.2" data-path="vector-algebra.html"><a href="vector-algebra.html#mean-centering"><i class="fa fa-check"></i><b>9.3.2</b> Mean centering</a></li>
<li class="chapter" data-level="9.3.3" data-path="vector-algebra.html"><a href="vector-algebra.html#variance-and-standard-deviation"><i class="fa fa-check"></i><b>9.3.3</b> Variance and standard deviation</a></li>
<li class="chapter" data-level="9.3.4" data-path="vector-algebra.html"><a href="vector-algebra.html#covariance-and-correlation"><i class="fa fa-check"></i><b>9.3.4</b> Covariance and correlation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="matrices-in-r.html"><a href="matrices-in-r.html"><i class="fa fa-check"></i><b>10</b> Matrices in R</a>
<ul>
<li class="chapter" data-level="10.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#creating-matrices-in-r"><i class="fa fa-check"></i><b>10.1</b> Creating matrices in R</a></li>
<li class="chapter" data-level="10.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-arithmetic-operations-in-r"><i class="fa fa-check"></i><b>10.2</b> Matrix arithmetic operations in R</a></li>
<li class="chapter" data-level="10.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#descriptive-statistics-as-matrix-functions"><i class="fa fa-check"></i><b>10.3</b> Descriptive statistics as matrix functions</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#mean-vector-and-matrix"><i class="fa fa-check"></i><b>10.3.1</b> Mean vector and matrix</a></li>
<li class="chapter" data-level="10.3.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#deviation-matrix"><i class="fa fa-check"></i><b>10.3.2</b> Deviation matrix</a></li>
<li class="chapter" data-level="10.3.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#covariance-matrix"><i class="fa fa-check"></i><b>10.3.3</b> Covariance matrix</a></li>
<li class="chapter" data-level="10.3.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#correlation-matrix"><i class="fa fa-check"></i><b>10.3.4</b> Correlation matrix</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-inverse"><i class="fa fa-check"></i><b>10.4</b> Matrix Inverse</a></li>
<li class="chapter" data-level="10.5" data-path="matrices-in-r.html"><a href="matrices-in-r.html#solving-sets-of-simultaneous-equations"><i class="fa fa-check"></i><b>10.5</b> Solving sets of simultaneous equations</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html"><i class="fa fa-check"></i><b>11</b> Linear Regression Models I</a>
<ul>
<li class="chapter" data-level="11.1" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#linear-functions"><i class="fa fa-check"></i><b>11.1</b> Linear functions</a></li>
<li class="chapter" data-level="11.2" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#linear-regression"><i class="fa fa-check"></i><b>11.2</b> Linear regression</a></li>
<li class="chapter" data-level="11.3" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#the-optimality-criterion-for-least-squares-regression"><i class="fa fa-check"></i><b>11.3</b> The optimality criterion for least-squares regression</a></li>
<li class="chapter" data-level="11.4" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#geometry-of-linear-regression"><i class="fa fa-check"></i><b>11.4</b> Geometry of linear regression</a></li>
<li class="chapter" data-level="11.5" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#solution-for-the-least-squares-criterion"><i class="fa fa-check"></i><b>11.5</b> Solution for the least-squares criterion</a>
<ul>
<li class="chapter" data-level="11.5.1" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#bivariate-regression-estimating-coefficients"><i class="fa fa-check"></i><b>11.5.1</b> Bivariate regression, estimating coefficients</a></li>
<li class="chapter" data-level="11.5.2" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#multiple-regression-estimating-coefficients"><i class="fa fa-check"></i><b>11.5.2</b> Multiple regression, estimating coefficients</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#residuals"><i class="fa fa-check"></i><b>11.6</b> Residuals</a></li>
<li class="chapter" data-level="11.7" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#regression-as-sum-of-squares-decomposition"><i class="fa fa-check"></i><b>11.7</b> Regression as sum-of-squares decomposition</a></li>
<li class="chapter" data-level="11.8" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#variance-explained-by-a-regression-model"><i class="fa fa-check"></i><b>11.8</b> Variance “explained” by a regression model</a></li>
<li class="chapter" data-level="11.9" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#goodness-of-fit"><i class="fa fa-check"></i><b>11.9</b> Goodness of fit</a></li>
<li class="chapter" data-level="11.10" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#interpretting-regression"><i class="fa fa-check"></i><b>11.10</b> Interpretting Regression</a></li>
<li class="chapter" data-level="11.11" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#illustrating-linear-regression-with-simulated-data"><i class="fa fa-check"></i><b>11.11</b> Illustrating linear regression with simulated data</a></li>
<li class="chapter" data-level="11.12" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#calculating-the-regression-model-using-vector-arithmetic"><i class="fa fa-check"></i><b>11.12</b> Calculating the regression model using vector arithmetic</a>
<ul>
<li class="chapter" data-level="11.12.1" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#calculating-the-predicted-values-and-residuals"><i class="fa fa-check"></i><b>11.12.1</b> Calculating the predicted values and residuals</a></li>
</ul></li>
<li class="chapter" data-level="11.13" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#plotting-various-aspects-of-a-regression"><i class="fa fa-check"></i><b>11.13</b> Plotting various aspects of a regression</a>
<ul>
<li class="chapter" data-level="11.13.1" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#residual-plots"><i class="fa fa-check"></i><b>11.13.1</b> Residual plots</a></li>
</ul></li>
<li class="chapter" data-level="11.14" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#calculating-the-coefficient-of-determination"><i class="fa fa-check"></i><b>11.14</b> Calculating the coefficient of determination</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html"><i class="fa fa-check"></i><b>12</b> Linear Regression Models II</a>
<ul>
<li class="chapter" data-level="12.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#new-libraries-to-install"><i class="fa fa-check"></i><b>12.1</b> New Libraries to install</a></li>
<li class="chapter" data-level="12.2" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#standard-libraries"><i class="fa fa-check"></i><b>12.2</b> Standard libraries</a></li>
<li class="chapter" data-level="12.3" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#specifying-regression-models-in-r"><i class="fa fa-check"></i><b>12.3</b> Specifying Regression Models in R</a>
<ul>
<li class="chapter" data-level="12.3.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#interpretting-summary-output-from-lm"><i class="fa fa-check"></i><b>12.3.1</b> Interpretting summary output from lm()</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#quick-bivariate-regression-plots-in-ggplot"><i class="fa fa-check"></i><b>12.4</b> Quick bivariate regression plots in ggplot</a></li>
<li class="chapter" data-level="12.5" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#more-about-the-data-structure-returned-by-lm"><i class="fa fa-check"></i><b>12.5</b> More about the data structure returned by <code>lm()</code></a>
<ul>
<li class="chapter" data-level="12.5.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#fitted-values"><i class="fa fa-check"></i><b>12.5.1</b> Fitted values</a></li>
<li class="chapter" data-level="12.5.2" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#getting-the-model-coefficients"><i class="fa fa-check"></i><b>12.5.2</b> Getting the model coefficients</a></li>
</ul></li>
<li class="chapter" data-level="12.6" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#broom-a-library-for-converting-model-results-into-data-frames"><i class="fa fa-check"></i><b>12.6</b> Broom: a library for converting model results into data frames</a>
<ul>
<li class="chapter" data-level="12.6.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#broomtidy"><i class="fa fa-check"></i><b>12.6.1</b> <code>broom::tidy</code></a></li>
<li class="chapter" data-level="12.6.2" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#broomaugment"><i class="fa fa-check"></i><b>12.6.2</b> <code>broom::augment</code></a></li>
<li class="chapter" data-level="12.6.3" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#broomglance"><i class="fa fa-check"></i><b>12.6.3</b> <code>broom::glance</code></a></li>
</ul></li>
<li class="chapter" data-level="12.7" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#qq-plots"><i class="fa fa-check"></i><b>12.7</b> qq-plots</a></li>
<li class="chapter" data-level="12.8" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#multiple-regression"><i class="fa fa-check"></i><b>12.8</b> Multiple regression</a>
<ul>
<li class="chapter" data-level="12.8.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#exploration-of-the-trees-data-set"><i class="fa fa-check"></i><b>12.8.1</b> Exploration of the trees data set</a></li>
</ul></li>
<li class="chapter" data-level="12.9" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#d-plots"><i class="fa fa-check"></i><b>12.9</b> 3D Plots</a>
<ul>
<li class="chapter" data-level="12.9.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#scatterplot3d"><i class="fa fa-check"></i><b>12.9.1</b> scatterplot3d</a></li>
</ul></li>
<li class="chapter" data-level="12.10" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#fitting-the-regression-model"><i class="fa fa-check"></i><b>12.10</b> Fitting the regression model</a>
<ul>
<li class="chapter" data-level="12.10.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#visualizing-the-regression-model-in-scatterplot3d"><i class="fa fa-check"></i><b>12.10.1</b> Visualizing the regression model in scatterplot3d</a></li>
</ul></li>
<li class="chapter" data-level="12.11" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#interpretting-the-regression-model"><i class="fa fa-check"></i><b>12.11</b> Interpretting the regression model</a></li>
<li class="chapter" data-level="12.12" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#exploring-the-vector-geometry-of-the-regression-model"><i class="fa fa-check"></i><b>12.12</b> Exploring the Vector Geometry of the Regression Model</a></li>
<li class="chapter" data-level="12.13" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#exploring-the-residuals-from-the-model-fit"><i class="fa fa-check"></i><b>12.13</b> Exploring the Residuals from the Model Fit</a></li>
<li class="chapter" data-level="12.14" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#an-alternate-model"><i class="fa fa-check"></i><b>12.14</b> An alternate model</a></li>
<li class="chapter" data-level="12.15" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#exploring-the-impact-of-nearly-collinear-predictors-on-regression"><i class="fa fa-check"></i><b>12.15</b> Exploring the impact of nearly collinear predictors on regression</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html"><i class="fa fa-check"></i><b>13</b> Principal Components Analysis</a>
<ul>
<li class="chapter" data-level="13.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#libraries-3"><i class="fa fa-check"></i><b>13.1</b> Libraries</a></li>
<li class="chapter" data-level="13.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#matrices-as-linear-transformations"><i class="fa fa-check"></i><b>13.2</b> Matrices as linear transformations</a></li>
<li class="chapter" data-level="13.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#eigenanalysis-in-r"><i class="fa fa-check"></i><b>13.3</b> Eigenanalysis in R</a></li>
<li class="chapter" data-level="13.4" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#principal-components-analysis-in-r"><i class="fa fa-check"></i><b>13.4</b> Principal Components Analysis in R</a>
<ul>
<li class="chapter" data-level="13.4.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#bioenv-dataset"><i class="fa fa-check"></i><b>13.4.1</b> Bioenv dataset</a></li>
<li class="chapter" data-level="13.4.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pca-of-the-bioenv-dataset"><i class="fa fa-check"></i><b>13.4.2</b> PCA of the Bioenv dataset</a></li>
<li class="chapter" data-level="13.4.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#calculating-factor-loadings"><i class="fa fa-check"></i><b>13.4.3</b> Calculating Factor Loadings</a></li>
</ul></li>
<li class="chapter" data-level="13.5" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#drawing-figures-to-represent-pca"><i class="fa fa-check"></i><b>13.5</b> Drawing Figures to Represent PCA</a>
<ul>
<li class="chapter" data-level="13.5.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pc-score-plots"><i class="fa fa-check"></i><b>13.5.1</b> PC Score Plots</a></li>
<li class="chapter" data-level="13.5.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#simultaneous-depiction-of-observations-and-variables-in-the-pc-space"><i class="fa fa-check"></i><b>13.5.2</b> Simultaneous Depiction of Observations and Variables in the PC Space</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="14" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html"><i class="fa fa-check"></i><b>14</b> Canonical Variates Analysis</a>
<ul>
<li class="chapter" data-level="14.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#libraries-4"><i class="fa fa-check"></i><b>14.1</b> Libraries</a></li>
<li class="chapter" data-level="14.2" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#discriminant-analysis-in-r"><i class="fa fa-check"></i><b>14.2</b> Discriminant Analysis in R</a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#shorthand-formulae-in-r"><i class="fa fa-check"></i><b>14.2.1</b> Shorthand Formulae in R</a></li>
<li class="chapter" data-level="14.2.2" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#working-with-the-output-of-lda"><i class="fa fa-check"></i><b>14.2.2</b> Working with the output of <code>lda()</code></a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#estimating-confidence-regions-for-group-means-in-cva"><i class="fa fa-check"></i><b>14.3</b> Estimating confidence regions for group means in CVA</a>
<ul>
<li class="chapter" data-level="14.3.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#drawing-the-cva-confidence-regions"><i class="fa fa-check"></i><b>14.3.1</b> Drawing the CVA confidence regions</a></li>
</ul></li>
<li class="chapter" data-level="14.4" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#calculating-the-within-and-between-group-covariance-matrices"><i class="fa fa-check"></i><b>14.4</b> Calculating the Within and Between Group Covariance Matrices</a>
<ul>
<li class="chapter" data-level="14.4.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#recapitulating-the-cva-analysis-of-lda"><i class="fa fa-check"></i><b>14.4.1</b> Recapitulating the CVA analysis of <code>lda()</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="15" data-path="clustering-in-r.html"><a href="clustering-in-r.html"><i class="fa fa-check"></i><b>15</b> Clustering in R</a>
<ul>
<li class="chapter" data-level="15.1" data-path="clustering-in-r.html"><a href="clustering-in-r.html#libraries-5"><i class="fa fa-check"></i><b>15.1</b> Libraries</a></li>
<li class="chapter" data-level="15.2" data-path="clustering-in-r.html"><a href="clustering-in-r.html#data-set"><i class="fa fa-check"></i><b>15.2</b> Data set</a></li>
<li class="chapter" data-level="15.3" data-path="clustering-in-r.html"><a href="clustering-in-r.html#hierarchical-clustering-in-r"><i class="fa fa-check"></i><b>15.3</b> Hierarchical Clustering in R</a></li>
<li class="chapter" data-level="15.4" data-path="clustering-in-r.html"><a href="clustering-in-r.html#manipulating-hierarchical-clusterings-with-dendextend"><i class="fa fa-check"></i><b>15.4</b> Manipulating hierarchical clusterings with dendextend</a></li>
<li class="chapter" data-level="15.5" data-path="clustering-in-r.html"><a href="clustering-in-r.html#plotting-dendrograms-in-dendextend"><i class="fa fa-check"></i><b>15.5</b> Plotting dendrograms in dendextend</a></li>
<li class="chapter" data-level="15.6" data-path="clustering-in-r.html"><a href="clustering-in-r.html#cutting-dendrograms"><i class="fa fa-check"></i><b>15.6</b> Cutting dendrograms</a></li>
<li class="chapter" data-level="15.7" data-path="clustering-in-r.html"><a href="clustering-in-r.html#looking-at-clusters"><i class="fa fa-check"></i><b>15.7</b> Looking at clusters</a></li>
<li class="chapter" data-level="15.8" data-path="clustering-in-r.html"><a href="clustering-in-r.html#generating-a-heat-map-from-a-cluster"><i class="fa fa-check"></i><b>15.8</b> Generating a heat map from a cluster</a></li>
<li class="chapter" data-level="15.9" data-path="clustering-in-r.html"><a href="clustering-in-r.html#working-with-sub-trees"><i class="fa fa-check"></i><b>15.9</b> Working with sub-trees</a></li>
<li class="chapter" data-level="15.10" data-path="clustering-in-r.html"><a href="clustering-in-r.html#setting-dendrogram-parameters-in-dendextend"><i class="fa fa-check"></i><b>15.10</b> Setting dendrogram parameters in dendextend</a></li>
<li class="chapter" data-level="15.11" data-path="clustering-in-r.html"><a href="clustering-in-r.html#combining-heatmaps-and-dendrograms"><i class="fa fa-check"></i><b>15.11</b> Combining heatmaps and dendrograms</a></li>
<li class="chapter" data-level="15.12" data-path="clustering-in-r.html"><a href="clustering-in-r.html#k-meansk-medoids-clustering-in-r"><i class="fa fa-check"></i><b>15.12</b> K-means/K-medoids Clustering in R</a>
<ul>
<li class="chapter" data-level="15.12.1" data-path="clustering-in-r.html"><a href="clustering-in-r.html#heat-map-from-k-medoids-cluster"><i class="fa fa-check"></i><b>15.12.1</b> Heat map from k-medoids cluster</a></li>
</ul></li>
<li class="chapter" data-level="15.13" data-path="clustering-in-r.html"><a href="clustering-in-r.html#combining-clusters-and-correlation-matrix-heatmaps"><i class="fa fa-check"></i><b>15.13</b> Combining clusters and correlation matrix heatmaps</a>
<ul>
<li class="chapter" data-level="15.13.1" data-path="clustering-in-r.html"><a href="clustering-in-r.html#hierarchical-clustering-visualized-on-correlation-matrix"><i class="fa fa-check"></i><b>15.13.1</b> Hierarchical clustering, visualized on correlation matrix</a></li>
<li class="chapter" data-level="15.13.2" data-path="clustering-in-r.html"><a href="clustering-in-r.html#k-medoids-clustering-visualized-on-the-correlation-matrix"><i class="fa fa-check"></i><b>15.13.2</b> K-medoids clustering visualized on the correlation matrix</a></li>
</ul></li>
<li class="chapter" data-level="15.14" data-path="clustering-in-r.html"><a href="clustering-in-r.html#depicting-the-data-within-clusters"><i class="fa fa-check"></i><b>15.14</b> Depicting the data within clusters</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Biology 723: Statistical Computing for Biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="linear-regression-models-ii" class="section level1" number="12">
<h1><span class="header-section-number">Chapter 12</span> Linear Regression Models II</h1>
<p>Last week we reviewed the mathematical basis of linear regression, and we saw how to fit bivariate and regression models using vector operations. This week we’ll look at R’s built-in tools for fitting regression models and we’ll look at a couple of options for producing 3D plots.</p>
<div id="new-libraries-to-install" class="section level2" number="12.1">
<h2><span class="header-section-number">12.1</span> New Libraries to install</h2>
<p>We’ll be using several new packages for this class session. Install the following packages via one of the standard install mechanisms:</p>
<ul>
<li><code>broom</code></li>
<li><code>scatterplot3d</code></li>
</ul>
</div>
<div id="standard-libraries" class="section level2" number="12.2">
<h2><span class="header-section-number">12.2</span> Standard libraries</h2>
<div class="sourceCode" id="cb361"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb361-1"><a href="linear-regression-models-ii.html#cb361-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
</div>
<div id="specifying-regression-models-in-r" class="section level2" number="12.3">
<h2><span class="header-section-number">12.3</span> Specifying Regression Models in R</h2>
<p>As one would expect, R has a built-in function for fitting linear regression models. The function <code>lm()</code> can be used to fit bivariate and multiple regression models, as well asanalysis of variance, analysis of covariance, and other linear models.</p>
<p>We’ll start by illustrating bivariate regression with the <a href="https://github.com/Bio723-class/example-datasets/raw/master/ABD-lion-noses.csv">lion nose pigmentation data set</a> introduced in the assignment for the previous class session. Let’s load the lion data and remind ourselves of the basic structure:</p>
<div class="sourceCode" id="cb362"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb362-1"><a href="linear-regression-models-ii.html#cb362-1" aria-hidden="true" tabindex="-1"></a>lions <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;https://github.com/Bio723-class/example-datasets/raw/master/ABD-lion-noses.csv&quot;</span>)</span>
<span id="cb362-2"><a href="linear-regression-models-ii.html#cb362-2" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb362-3"><a href="linear-regression-models-ii.html#cb362-3" aria-hidden="true" tabindex="-1"></a><span class="do">## ── Column specification ────────────────────────────────────────────────────────</span></span>
<span id="cb362-4"><a href="linear-regression-models-ii.html#cb362-4" aria-hidden="true" tabindex="-1"></a><span class="do">## cols(</span></span>
<span id="cb362-5"><a href="linear-regression-models-ii.html#cb362-5" aria-hidden="true" tabindex="-1"></a><span class="do">##   proportionBlack = col_double(),</span></span>
<span id="cb362-6"><a href="linear-regression-models-ii.html#cb362-6" aria-hidden="true" tabindex="-1"></a><span class="do">##   ageInYears = col_double()</span></span>
<span id="cb362-7"><a href="linear-regression-models-ii.html#cb362-7" aria-hidden="true" tabindex="-1"></a><span class="do">## )</span></span>
<span id="cb362-8"><a href="linear-regression-models-ii.html#cb362-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(lions)</span>
<span id="cb362-9"><a href="linear-regression-models-ii.html#cb362-9" aria-hidden="true" tabindex="-1"></a><span class="do">## # A tibble: 6 x 2</span></span>
<span id="cb362-10"><a href="linear-regression-models-ii.html#cb362-10" aria-hidden="true" tabindex="-1"></a><span class="do">##   proportionBlack ageInYears</span></span>
<span id="cb362-11"><a href="linear-regression-models-ii.html#cb362-11" aria-hidden="true" tabindex="-1"></a><span class="do">##             &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span id="cb362-12"><a href="linear-regression-models-ii.html#cb362-12" aria-hidden="true" tabindex="-1"></a><span class="do">## 1            0.21        1.1</span></span>
<span id="cb362-13"><a href="linear-regression-models-ii.html#cb362-13" aria-hidden="true" tabindex="-1"></a><span class="do">## 2            0.14        1.5</span></span>
<span id="cb362-14"><a href="linear-regression-models-ii.html#cb362-14" aria-hidden="true" tabindex="-1"></a><span class="do">## 3            0.11        1.9</span></span>
<span id="cb362-15"><a href="linear-regression-models-ii.html#cb362-15" aria-hidden="true" tabindex="-1"></a><span class="do">## 4            0.13        2.2</span></span>
<span id="cb362-16"><a href="linear-regression-models-ii.html#cb362-16" aria-hidden="true" tabindex="-1"></a><span class="do">## 5            0.12        2.6</span></span>
<span id="cb362-17"><a href="linear-regression-models-ii.html#cb362-17" aria-hidden="true" tabindex="-1"></a><span class="do">## 6            0.13        3.2</span></span></code></pre></div>
<p>Recall that our predictor (explanatory) variable is <code>proportionBlack</code> and our outcome (response) variable is <code>ageInYears</code>. We use the <code>lm()</code> function to fit the regression of <code>proportionBlack</code> on <code>ageInYears</code> as so:</p>
<div class="sourceCode" id="cb363"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb363-1"><a href="linear-regression-models-ii.html#cb363-1" aria-hidden="true" tabindex="-1"></a>fit.lions <span class="ot">&lt;-</span> <span class="fu">lm</span>(ageInYears <span class="sc">~</span> proportionBlack, lions)</span></code></pre></div>
<p>The first argument to <code>lm</code> is an R “formula,” the second argument is a data frame.Formulas are R’s way of specifying models, though they find other uses as well (e.g. we saw the formula syntax when we introduced the <code>facet_wrap</code> and <code>facet_grid</code> functions from ggplot). The general form of a formula in R is <code>response variable ~ explanatory variables</code>. In the code example above, we have only a single explanatory variable, and thus our response variable is <code>ageInYears</code> and our explanatory variable is <code>proportionBlack</code>.</p>
<p>The <code>lm</code> function returns a list with a number of different components. The ones of most interest to us are <code>fitted.values</code>, <code>coefficients</code>, <code>residuals</code>, and (see the <code>lm</code> documentation for full details.)</p>
<div class="sourceCode" id="cb364"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb364-1"><a href="linear-regression-models-ii.html#cb364-1" aria-hidden="true" tabindex="-1"></a>fit.lions</span>
<span id="cb364-2"><a href="linear-regression-models-ii.html#cb364-2" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb364-3"><a href="linear-regression-models-ii.html#cb364-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Call:</span></span>
<span id="cb364-4"><a href="linear-regression-models-ii.html#cb364-4" aria-hidden="true" tabindex="-1"></a><span class="do">## lm(formula = ageInYears ~ proportionBlack, data = lions)</span></span>
<span id="cb364-5"><a href="linear-regression-models-ii.html#cb364-5" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb364-6"><a href="linear-regression-models-ii.html#cb364-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Coefficients:</span></span>
<span id="cb364-7"><a href="linear-regression-models-ii.html#cb364-7" aria-hidden="true" tabindex="-1"></a><span class="do">##     (Intercept)  proportionBlack  </span></span>
<span id="cb364-8"><a href="linear-regression-models-ii.html#cb364-8" aria-hidden="true" tabindex="-1"></a><span class="do">##           0.879           10.647</span></span></code></pre></div>
<div id="interpretting-summary-output-from-lm" class="section level3" number="12.3.1">
<h3><span class="header-section-number">12.3.1</span> Interpretting summary output from lm()</h3>
<p>Calling <code>summary</code> on a fit model provides more detailed output:</p>
<div class="sourceCode" id="cb365"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb365-1"><a href="linear-regression-models-ii.html#cb365-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(fit.lions)</span>
<span id="cb365-2"><a href="linear-regression-models-ii.html#cb365-2" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb365-3"><a href="linear-regression-models-ii.html#cb365-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Call:</span></span>
<span id="cb365-4"><a href="linear-regression-models-ii.html#cb365-4" aria-hidden="true" tabindex="-1"></a><span class="do">## lm(formula = ageInYears ~ proportionBlack, data = lions)</span></span>
<span id="cb365-5"><a href="linear-regression-models-ii.html#cb365-5" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb365-6"><a href="linear-regression-models-ii.html#cb365-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Residuals:</span></span>
<span id="cb365-7"><a href="linear-regression-models-ii.html#cb365-7" aria-hidden="true" tabindex="-1"></a><span class="do">##     Min      1Q  Median      3Q     Max </span></span>
<span id="cb365-8"><a href="linear-regression-models-ii.html#cb365-8" aria-hidden="true" tabindex="-1"></a><span class="do">## -2.5449 -1.1117 -0.5285  0.9635  4.3421 </span></span>
<span id="cb365-9"><a href="linear-regression-models-ii.html#cb365-9" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb365-10"><a href="linear-regression-models-ii.html#cb365-10" aria-hidden="true" tabindex="-1"></a><span class="do">## Coefficients:</span></span>
<span id="cb365-11"><a href="linear-regression-models-ii.html#cb365-11" aria-hidden="true" tabindex="-1"></a><span class="do">##                 Estimate Std. Error t value Pr(&gt;|t|)    </span></span>
<span id="cb365-12"><a href="linear-regression-models-ii.html#cb365-12" aria-hidden="true" tabindex="-1"></a><span class="do">## (Intercept)       0.8790     0.5688   1.545    0.133    </span></span>
<span id="cb365-13"><a href="linear-regression-models-ii.html#cb365-13" aria-hidden="true" tabindex="-1"></a><span class="do">## proportionBlack  10.6471     1.5095   7.053 7.68e-08 ***</span></span>
<span id="cb365-14"><a href="linear-regression-models-ii.html#cb365-14" aria-hidden="true" tabindex="-1"></a><span class="do">## ---</span></span>
<span id="cb365-15"><a href="linear-regression-models-ii.html#cb365-15" aria-hidden="true" tabindex="-1"></a><span class="do">## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</span></span>
<span id="cb365-16"><a href="linear-regression-models-ii.html#cb365-16" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb365-17"><a href="linear-regression-models-ii.html#cb365-17" aria-hidden="true" tabindex="-1"></a><span class="do">## Residual standard error: 1.669 on 30 degrees of freedom</span></span>
<span id="cb365-18"><a href="linear-regression-models-ii.html#cb365-18" aria-hidden="true" tabindex="-1"></a><span class="do">## Multiple R-squared:  0.6238, Adjusted R-squared:  0.6113 </span></span>
<span id="cb365-19"><a href="linear-regression-models-ii.html#cb365-19" aria-hidden="true" tabindex="-1"></a><span class="do">## F-statistic: 49.75 on 1 and 30 DF,  p-value: 7.677e-08</span></span></code></pre></div>
<p>The <code>summary()</code> function provides textual output reporting a number of features of the model fit. We’ll focus on the section labeled coefficients. For the model we fit above this looks like:</p>
<pre><code>Coefficients:
                 Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)       0.8790     0.5688   1.545    0.133    
proportionBlack  10.6471     1.5095   7.053 7.68e-08 ***
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
<p>The first column of this printed table gives the values of the estimated coefficients. The first coefficient, labeled “(Intercept)” is the intercept of the regression line (<span class="math inline">\(a\)</span> in the formulate <span class="math inline">\(\widehat{Y} = a + bX\)</span>) – here the estimated value is 0.879. Since our model includes only one predictor variable (proportionBlack) we have only one additional coefficient (b = 10.647). Thus the mathematical equation describing our fit model is $ = 0.88 + 10.65() $</p>
<p>The second column (<code>Std. Error</code>) in the coefficients table gives the standard error of the estimated coefficients (based on assumption of multivariate normality of the data).</p>
<p>The third column (<code>t value</code>) gives a calculated t-value for the null hypothesis that the corresponding coefficient is zero, and the fourth column (<code>Pr(&gt;|t|)</code>) is the probability of observing a t-value that large under the null hypothesis that the corresponding has a value of zero (i.e. a p-value). For the intercept, we see that the p-value is 0.133; thus we don’t have strong statistical evidence to reject the null hypothesis that this coefficient is zero. However, for the slope the p-value is &lt; 1e-8. This is strong evidence on which to reject the null hypothesis of a zero coefficient (a zero slope would imply that)</p>
</div>
</div>
<div id="quick-bivariate-regression-plots-in-ggplot" class="section level2" number="12.4">
<h2><span class="header-section-number">12.4</span> Quick bivariate regression plots in ggplot</h2>
<p>Since linear model fitting is a fairly common task, the ggplot library includes a geometric mapping, <code>geom_smooth()</code>, that will fit a linear model (as well as other models, as we’ll see in future lection) for us and generate the corresponding regression plot. We can create this by specifying <code>method="lm"</code> as the argument to <code>geom_smooth()</code>:</p>
<div class="sourceCode" id="cb367"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb367-1"><a href="linear-regression-models-ii.html#cb367-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(lions, <span class="fu">aes</span>(<span class="at">x =</span> proportionBlack, <span class="at">y =</span> ageInYears)) <span class="sc">+</span> </span>
<span id="cb367-2"><a href="linear-regression-models-ii.html#cb367-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.75</span>) <span class="sc">+</span> </span>
<span id="cb367-3"><a href="linear-regression-models-ii.html#cb367-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method=</span><span class="st">&quot;lm&quot;</span>, <span class="at">color =</span> <span class="st">&#39;red&#39;</span>, <span class="at">fullrange=</span><span class="cn">TRUE</span>) <span class="sc">+</span> </span>
<span id="cb367-4"><a href="linear-regression-models-ii.html#cb367-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="dv">0</span>,<span class="dv">1</span>)</span>
<span id="cb367-5"><a href="linear-regression-models-ii.html#cb367-5" aria-hidden="true" tabindex="-1"></a><span class="do">## `geom_smooth()` using formula &#39;y ~ x&#39;</span></span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-363-1.png" width="576" /></p>
<p>By default, <code>geom_smooth</code> draws 95% confidence intervals for the regression model (the shaded gray area around the regression line). These confidence intervals reflect the uncertainty in the estimates of the slope and intercept, as reflected by the standard error of the coefficients discussed above. Note that confidence intervals for a linear regression model are wider far away from the mean values of <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span>.</p>
</div>
<div id="more-about-the-data-structure-returned-by-lm" class="section level2" number="12.5">
<h2><span class="header-section-number">12.5</span> More about the data structure returned by <code>lm()</code></h2>
<p>The data structure returned by <code>lm()</code> is a list-like object with multiple fields:</p>
<div class="sourceCode" id="cb368"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb368-1"><a href="linear-regression-models-ii.html#cb368-1" aria-hidden="true" tabindex="-1"></a><span class="fu">typeof</span>(fit.lions)</span>
<span id="cb368-2"><a href="linear-regression-models-ii.html#cb368-2" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] &quot;list&quot;</span></span></code></pre></div>
<p>We can access the names of all the fields associated with the fit object:</p>
<div class="sourceCode" id="cb369"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb369-1"><a href="linear-regression-models-ii.html#cb369-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(fit.lions)</span>
<span id="cb369-2"><a href="linear-regression-models-ii.html#cb369-2" aria-hidden="true" tabindex="-1"></a><span class="do">##  [1] &quot;coefficients&quot;  &quot;residuals&quot;     &quot;effects&quot;       &quot;rank&quot;         </span></span>
<span id="cb369-3"><a href="linear-regression-models-ii.html#cb369-3" aria-hidden="true" tabindex="-1"></a><span class="do">##  [5] &quot;fitted.values&quot; &quot;assign&quot;        &quot;qr&quot;            &quot;df.residual&quot;  </span></span>
<span id="cb369-4"><a href="linear-regression-models-ii.html#cb369-4" aria-hidden="true" tabindex="-1"></a><span class="do">##  [9] &quot;xlevels&quot;       &quot;call&quot;          &quot;terms&quot;         &quot;model&quot;</span></span></code></pre></div>
<div id="fitted-values" class="section level3" number="12.5.1">
<h3><span class="header-section-number">12.5.1</span> Fitted values</h3>
<p>The component <code>fitted.values</code> gives the predicted values of <span class="math inline">\(Y\)</span> (<span class="math inline">\(\hat{Y}\)</span> in the equations above) for each observed value of <span class="math inline">\(X\)</span>. We can plot these predicted values of <span class="math inline">\(Y\)</span>, as shown below. Notice how the predicted values all fall on a line (the regression line itself!)</p>
<div class="sourceCode" id="cb370"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb370-1"><a href="linear-regression-models-ii.html#cb370-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(lions, <span class="fu">aes</span>(<span class="at">x =</span> proportionBlack, <span class="at">y =</span> ageInYears)) <span class="sc">+</span> </span>
<span id="cb370-2"><a href="linear-regression-models-ii.html#cb370-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha=</span><span class="fl">0.7</span>) <span class="sc">+</span>                           <span class="co"># observed data</span></span>
<span id="cb370-3"><a href="linear-regression-models-ii.html#cb370-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> proportionBlack, <span class="at">y =</span> fit.lions<span class="sc">$</span>fitted.values),  <span class="co"># predicted data</span></span>
<span id="cb370-4"><a href="linear-regression-models-ii.html#cb370-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">color=</span><span class="st">&#39;red&#39;</span>, <span class="at">alpha=</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb370-5"><a href="linear-regression-models-ii.html#cb370-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">xend =</span> proportionBlack, <span class="at">yend =</span> fit.lions<span class="sc">$</span>fitted.values),</span>
<span id="cb370-6"><a href="linear-regression-models-ii.html#cb370-6" aria-hidden="true" tabindex="-1"></a>               <span class="at">color=</span><span class="st">&#39;red&#39;</span>, <span class="at">linetype=</span><span class="st">&#39;dashed&#39;</span>, <span class="at">alpha=</span><span class="fl">0.25</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-366"></span>
<img src="bio723-book_files/figure-html/unnamed-chunk-366-1.png" alt="Observed (black) and predicted (red) values in a linear regression of Y on X.  Dashed lines indicate the residuals from the regression." width="576" />
<p class="caption">
Figure 12.1: Observed (black) and predicted (red) values in a linear regression of Y on X. Dashed lines indicate the residuals from the regression.
</p>
</div>
</div>
<div id="getting-the-model-coefficients" class="section level3" number="12.5.2">
<h3><span class="header-section-number">12.5.2</span> Getting the model coefficients</h3>
<p>The <code>coefficients</code> components gives the value of the model parameters, namely the intercept and slope. These are the same values reported by the <code>summary()</code> function described above:</p>
<div class="sourceCode" id="cb371"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb371-1"><a href="linear-regression-models-ii.html#cb371-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> fit.lions<span class="sc">$</span>coefficients</span>
<span id="cb371-2"><a href="linear-regression-models-ii.html#cb371-2" aria-hidden="true" tabindex="-1"></a><span class="do">##     (Intercept) proportionBlack </span></span>
<span id="cb371-3"><a href="linear-regression-models-ii.html#cb371-3" aria-hidden="true" tabindex="-1"></a><span class="do">##       0.8790062      10.6471194</span></span></code></pre></div>
<p>Typically these coefficients are given with more precision than we’d want to report. For the purpose of printing, can create character strings with rounded coefficients using the <code>sprintf()</code> function so:</p>
<p>As shown above, the estimated slope is 10.65 and the estimated intercept is 0.88. The model estimated by our linear regression is thus <span class="math inline">\(\widehat{\text{age}} = 0.88 + 10.65\text{proportion black}\)</span>.</p>
</div>
</div>
<div id="broom-a-library-for-converting-model-results-into-data-frames" class="section level2" number="12.6">
<h2><span class="header-section-number">12.6</span> Broom: a library for converting model results into data frames</h2>
<p>The data structure we got back when we used the <code>lm</code> function to carry out linear regression, carries lots of useful information it isn’t a particularly “tidy” way to access the data. The R package Broom converts "statistical analysis objects from R into tidy data frames, so that they can more easily be combined, reshaped and otherwise processed with tools like ‘dplyr,’ ‘tidyr’ and ‘ggplot2.’ The discussion of Broom below is drawn from the <a href="https://cran.r-project.org/web/packages/broom/vignettes/broom.html">Introduction to Broom</a></p>
<p><strong>If you haven’t already done so, install the <code>broom</code> package before proceeding.</strong></p>
<div class="sourceCode" id="cb372"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb372-1"><a href="linear-regression-models-ii.html#cb372-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(broom)</span></code></pre></div>
<p>There are three <code>broom</code> functions that are particularly useful for our purposes. They are:</p>
<ol style="list-style-type: decimal">
<li><code>tidy</code> – constructs a data frame that summarizes the model’s statistical findings.</li>
<li><code>augment</code> – add columns to the original data that was modeled. This includes predictions, residuals, and cluster assignments.</li>
<li><code>glance</code> – construct a concise one-row summary of the model.</li>
</ol>
<div id="broomtidy" class="section level3" number="12.6.1">
<h3><span class="header-section-number">12.6.1</span> <code>broom::tidy</code></h3>
<p><code>tidy</code> applied to a regression model object returns a table giving the estimated coefficients and other information about the uncertainty of those estimates and corresponding p-values. This mirrors the table of coefficients that apply <code>summary()</code> provided to us, but instead of just printing this information we get it in the form of a data frame that we can do further computations on.</p>
<div class="sourceCode" id="cb373"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb373-1"><a href="linear-regression-models-ii.html#cb373-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tidy</span>(fit.lions)</span>
<span id="cb373-2"><a href="linear-regression-models-ii.html#cb373-2" aria-hidden="true" tabindex="-1"></a><span class="do">## # A tibble: 2 x 5</span></span>
<span id="cb373-3"><a href="linear-regression-models-ii.html#cb373-3" aria-hidden="true" tabindex="-1"></a><span class="do">##   term            estimate std.error statistic      p.value</span></span>
<span id="cb373-4"><a href="linear-regression-models-ii.html#cb373-4" aria-hidden="true" tabindex="-1"></a><span class="do">##   &lt;chr&gt;              &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;        &lt;dbl&gt;</span></span>
<span id="cb373-5"><a href="linear-regression-models-ii.html#cb373-5" aria-hidden="true" tabindex="-1"></a><span class="do">## 1 (Intercept)        0.879     0.569      1.55 0.133       </span></span>
<span id="cb373-6"><a href="linear-regression-models-ii.html#cb373-6" aria-hidden="true" tabindex="-1"></a><span class="do">## 2 proportionBlack   10.6       1.51       7.05 0.0000000768</span></span></code></pre></div>
</div>
<div id="broomaugment" class="section level3" number="12.6.2">
<h3><span class="header-section-number">12.6.2</span> <code>broom::augment</code></h3>
<p><code>augment</code> creates a data frame that combines the original data with related information from the model fit.</p>
<div class="sourceCode" id="cb374"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb374-1"><a href="linear-regression-models-ii.html#cb374-1" aria-hidden="true" tabindex="-1"></a>lions.augmented <span class="ot">&lt;-</span> <span class="fu">augment</span>(fit.lions, lions)</span>
<span id="cb374-2"><a href="linear-regression-models-ii.html#cb374-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(lions.augmented)</span>
<span id="cb374-3"><a href="linear-regression-models-ii.html#cb374-3" aria-hidden="true" tabindex="-1"></a><span class="do">## # A tibble: 6 x 8</span></span>
<span id="cb374-4"><a href="linear-regression-models-ii.html#cb374-4" aria-hidden="true" tabindex="-1"></a><span class="do">##   proportionBlack ageInYears .fitted  .resid .std.resid   .hat .sigma   .cooksd</span></span>
<span id="cb374-5"><a href="linear-regression-models-ii.html#cb374-5" aria-hidden="true" tabindex="-1"></a><span class="do">##             &lt;dbl&gt;      &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;     &lt;dbl&gt;</span></span>
<span id="cb374-6"><a href="linear-regression-models-ii.html#cb374-6" aria-hidden="true" tabindex="-1"></a><span class="do">## 1            0.21        1.1    3.11 -2.01      -1.23   0.0415   1.65 0.0330   </span></span>
<span id="cb374-7"><a href="linear-regression-models-ii.html#cb374-7" aria-hidden="true" tabindex="-1"></a><span class="do">## 2            0.14        1.5    2.37 -0.870     -0.537  0.0584   1.69 0.00894  </span></span>
<span id="cb374-8"><a href="linear-regression-models-ii.html#cb374-8" aria-hidden="true" tabindex="-1"></a><span class="do">## 3            0.11        1.9    2.05 -0.150     -0.0932 0.0681   1.70 0.000318 </span></span>
<span id="cb374-9"><a href="linear-regression-models-ii.html#cb374-9" aria-hidden="true" tabindex="-1"></a><span class="do">## 4            0.13        2.2    2.26 -0.0631    -0.0391 0.0615   1.70 0.0000499</span></span>
<span id="cb374-10"><a href="linear-regression-models-ii.html#cb374-10" aria-hidden="true" tabindex="-1"></a><span class="do">## 5            0.12        2.6    2.16  0.443      0.275  0.0647   1.70 0.00261  </span></span>
<span id="cb374-11"><a href="linear-regression-models-ii.html#cb374-11" aria-hidden="true" tabindex="-1"></a><span class="do">## 6            0.13        3.2    2.26  0.937      0.580  0.0615   1.69 0.0110</span></span></code></pre></div>
<p>Now, in addition to the <code>proportionBlack</code> and <code>ageInYears</code> variables of the original data, we have columns like <code>.fitted</code> (value of Y predicted by the model for the corresponding value of X), <code>.resid</code> (difference between the actual Y and the predicted value), and a variety of other information for evalulating model uncertainty.</p>
<p>One thing we can do with this “augmented” data frame is to use it to better visualize and explore the model. For example, if we wanted to generate a figure highlighting the deviations from the model using vertical lines emanating from the regression line, we could do something like this:</p>
<div class="sourceCode" id="cb375"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb375-1"><a href="linear-regression-models-ii.html#cb375-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(lions.augmented, <span class="fu">aes</span>(proportionBlack, ageInYears)) <span class="sc">+</span> </span>
<span id="cb375-2"><a href="linear-regression-models-ii.html#cb375-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">geom_smooth</span>(<span class="at">method=</span><span class="st">&quot;lm&quot;</span>, <span class="at">color=</span><span class="st">&quot;red&quot;</span>,<span class="at">se=</span><span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb375-3"><a href="linear-regression-models-ii.html#cb375-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">xend =</span> proportionBlack, <span class="at">yend =</span> .fitted), <span class="at">linetype=</span><span class="st">&quot;dashed&quot;</span>)</span>
<span id="cb375-4"><a href="linear-regression-models-ii.html#cb375-4" aria-hidden="true" tabindex="-1"></a><span class="do">## `geom_smooth()` using formula &#39;y ~ x&#39;</span></span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-372-1.png" width="576" /></p>
<p>An another example, we can quickly create a residual plot using the augmented data frame as so:</p>
<div class="sourceCode" id="cb376"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb376-1"><a href="linear-regression-models-ii.html#cb376-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(lions.augmented, <span class="fu">aes</span>(proportionBlack, .resid)) <span class="sc">+</span> </span>
<span id="cb376-2"><a href="linear-regression-models-ii.html#cb376-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb376-3"><a href="linear-regression-models-ii.html#cb376-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">color =</span> <span class="st">&quot;red&quot;</span>, <span class="at">linetype=</span><span class="st">&#39;dashed&#39;</span>) <span class="sc">+</span> </span>
<span id="cb376-4"><a href="linear-regression-models-ii.html#cb376-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">&quot;Residuals&quot;</span>, <span class="at">title =</span> <span class="st">&quot;Residual plot the lions regression model.&quot;</span>)</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-373-1.png" width="576" /></p>
</div>
<div id="broomglance" class="section level3" number="12.6.3">
<h3><span class="header-section-number">12.6.3</span> <code>broom::glance</code></h3>
<p><code>glance()</code> provides summary information about the goodness of fit of the model. Most relevant for our current discussion is the column giving the coefficient of determination (<code>r.squared</code>):</p>
<div class="sourceCode" id="cb377"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb377-1"><a href="linear-regression-models-ii.html#cb377-1" aria-hidden="true" tabindex="-1"></a><span class="fu">glance</span>(fit.lions)</span>
<span id="cb377-2"><a href="linear-regression-models-ii.html#cb377-2" aria-hidden="true" tabindex="-1"></a><span class="do">## # A tibble: 1 x 12</span></span>
<span id="cb377-3"><a href="linear-regression-models-ii.html#cb377-3" aria-hidden="true" tabindex="-1"></a><span class="do">##   r.squared adj.r.squared sigma statistic p.value    df logLik   AIC   BIC</span></span>
<span id="cb377-4"><a href="linear-regression-models-ii.html#cb377-4" aria-hidden="true" tabindex="-1"></a><span class="do">##       &lt;dbl&gt;         &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span id="cb377-5"><a href="linear-regression-models-ii.html#cb377-5" aria-hidden="true" tabindex="-1"></a><span class="do">## 1     0.624         0.611  1.67      49.8 7.68e-8     1  -60.8  128.  132.</span></span>
<span id="cb377-6"><a href="linear-regression-models-ii.html#cb377-6" aria-hidden="true" tabindex="-1"></a><span class="do">## # … with 3 more variables: deviance &lt;dbl&gt;, df.residual &lt;int&gt;, nobs &lt;int&gt;</span></span></code></pre></div>
</div>
</div>
<div id="qq-plots" class="section level2" number="12.7">
<h2><span class="header-section-number">12.7</span> qq-plots</h2>
<p>From our residuals plot of the lions data set, there may be some indication of greater variance of residuals for larger values of the predictor variable.</p>
<p>Let’s check how normal the residuals look using a diagnostic plot called a QQ-plot (quantile-quantile plot). A qq-plot is a graphical method for comparing distributions by plotting the respective quantiles against each other. Typically we plot sample quantiles against theoretical quantiles; for example to compare the sample quantiles to the theoretical expectation of normality. In the example below we construct the QQ-plot using “standardized residuals” from the regression which are just z-scores for the residuals.</p>
<div class="sourceCode" id="cb378"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb378-1"><a href="linear-regression-models-ii.html#cb378-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(lions.augmented, <span class="fu">aes</span>(<span class="at">sample =</span> .std.resid)) <span class="sc">+</span> </span>
<span id="cb378-2"><a href="linear-regression-models-ii.html#cb378-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_qq</span>() <span class="sc">+</span></span>
<span id="cb378-3"><a href="linear-regression-models-ii.html#cb378-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_qq_line</span>(<span class="at">color=</span><span class="st">&quot;firebrick&quot;</span>)</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-375-1.png" width="576" /></p>
<p>Based on the QQ-plot, the residuals seem to diverge somewhat from a normal distirbution, as there’s noticeable curvature in the QQ-plot. When we test for the normality of the residuals using Shapiro-Wilk’s test for normality, we fail to reject the null hypothesis of normality at a significance threshold of <span class="math inline">\(\alpha=0.05\)</span>:</p>
<div class="sourceCode" id="cb379"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb379-1"><a href="linear-regression-models-ii.html#cb379-1" aria-hidden="true" tabindex="-1"></a><span class="fu">shapiro.test</span>(lions.augmented<span class="sc">$</span>.resid)</span>
<span id="cb379-2"><a href="linear-regression-models-ii.html#cb379-2" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb379-3"><a href="linear-regression-models-ii.html#cb379-3" aria-hidden="true" tabindex="-1"></a><span class="do">##  Shapiro-Wilk normality test</span></span>
<span id="cb379-4"><a href="linear-regression-models-ii.html#cb379-4" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb379-5"><a href="linear-regression-models-ii.html#cb379-5" aria-hidden="true" tabindex="-1"></a><span class="do">## data:  lions.augmented$.resid</span></span>
<span id="cb379-6"><a href="linear-regression-models-ii.html#cb379-6" aria-hidden="true" tabindex="-1"></a><span class="do">## W = 0.93879, p-value = 0.0692</span></span></code></pre></div>
<p>Even though we failed to reject the null hypothesis of normality for the residuals, but the P-value is very close to significance, suggesting some caution in applying the linear model.</p>
</div>
<div id="multiple-regression" class="section level2" number="12.8">
<h2><span class="header-section-number">12.8</span> Multiple regression</h2>
<p>To illustrate multiple regression in R we’ll use a built in dataset called <code>trees</code>. <code>trees</code> consists of measurements of the girth, height, and volume of 31 black cherry trees (<code>?trees</code> for more info).</p>
<p>Let’s assume we’re lumberjacks, but our permit only allows us to harvest a fixed number of trees. We get paid by the total volume of wood we harvest, so we’re interested in predicting a tree’s volume (hard to measure directly) as a function of its girth and height (relatively easy to measure), so we can pick the best trees to harvest. We’ll therefore calculate a multiple regression of volume on height and width.</p>
<div id="exploration-of-the-trees-data-set" class="section level3" number="12.8.1">
<h3><span class="header-section-number">12.8.1</span> Exploration of the trees data set</h3>
<p>We’ll start with some summary tables and diagnostic plots to familiarize ourselves with the data:</p>
<div class="sourceCode" id="cb380"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb380-1"><a href="linear-regression-models-ii.html#cb380-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(trees)</span>
<span id="cb380-2"><a href="linear-regression-models-ii.html#cb380-2" aria-hidden="true" tabindex="-1"></a><span class="do">## # A tibble: 6 x 3</span></span>
<span id="cb380-3"><a href="linear-regression-models-ii.html#cb380-3" aria-hidden="true" tabindex="-1"></a><span class="do">##   Girth Height Volume</span></span>
<span id="cb380-4"><a href="linear-regression-models-ii.html#cb380-4" aria-hidden="true" tabindex="-1"></a><span class="do">##   &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;</span></span>
<span id="cb380-5"><a href="linear-regression-models-ii.html#cb380-5" aria-hidden="true" tabindex="-1"></a><span class="do">## 1   8.3     70   10.3</span></span>
<span id="cb380-6"><a href="linear-regression-models-ii.html#cb380-6" aria-hidden="true" tabindex="-1"></a><span class="do">## 2   8.6     65   10.3</span></span>
<span id="cb380-7"><a href="linear-regression-models-ii.html#cb380-7" aria-hidden="true" tabindex="-1"></a><span class="do">## 3   8.8     63   10.2</span></span>
<span id="cb380-8"><a href="linear-regression-models-ii.html#cb380-8" aria-hidden="true" tabindex="-1"></a><span class="do">## 4  10.5     72   16.4</span></span>
<span id="cb380-9"><a href="linear-regression-models-ii.html#cb380-9" aria-hidden="true" tabindex="-1"></a><span class="do">## 5  10.7     81   18.8</span></span>
<span id="cb380-10"><a href="linear-regression-models-ii.html#cb380-10" aria-hidden="true" tabindex="-1"></a><span class="do">## 6  10.8     83   19.7</span></span></code></pre></div>
<p>We’ll use the <code>GGally::ggpairs()</code> function introduced in problem set 01 to create a scatterplot matrix depicting the pairwise relationships between all the variables</p>
<div class="sourceCode" id="cb381"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb381-1"><a href="linear-regression-models-ii.html#cb381-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GGally)</span>
<span id="cb381-2"><a href="linear-regression-models-ii.html#cb381-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Registered S3 method overwritten by &#39;GGally&#39;:</span></span>
<span id="cb381-3"><a href="linear-regression-models-ii.html#cb381-3" aria-hidden="true" tabindex="-1"></a><span class="do">##   method from   </span></span>
<span id="cb381-4"><a href="linear-regression-models-ii.html#cb381-4" aria-hidden="true" tabindex="-1"></a><span class="do">##   +.gg   ggplot2</span></span>
<span id="cb381-5"><a href="linear-regression-models-ii.html#cb381-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggpairs</span>(trees)</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-378-1.png" width="576" /></p>
<p>As one might expect for morphological measurements related to size, the scatterplot matrix shows that all the variables are positively correlated, and girth and volume have a particularly strong correlation.</p>
</div>
</div>
<div id="d-plots" class="section level2" number="12.9">
<h2><span class="header-section-number">12.9</span> 3D Plots</h2>
<p>ggplot has no built in facilities for 3D scatter plots so we’ll use two new packages, <a href="https://cran.r-project.org/package=scatterplot3d"><code>scatterplot3D</code></a> and <a href="https://cran.r-project.org/web/packages/rgl/vignettes/rgl.html"><code>rgl</code></a>, to generate 3D visualizations.</p>
<div id="scatterplot3d" class="section level3" number="12.9.1">
<h3><span class="header-section-number">12.9.1</span> scatterplot3d</h3>
<div class="sourceCode" id="cb382"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb382-1"><a href="linear-regression-models-ii.html#cb382-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scatterplot3d)  <span class="co"># install this package first if needed</span></span>
<span id="cb382-2"><a href="linear-regression-models-ii.html#cb382-2" aria-hidden="true" tabindex="-1"></a><span class="fu">scatterplot3d</span>(trees,</span>
<span id="cb382-3"><a href="linear-regression-models-ii.html#cb382-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">main =</span> <span class="st">&#39;Tree Volume as</span><span class="sc">\n</span><span class="st">a function of Girth and Height&#39;</span>)</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-379-1.png" width="576" /></p>
<p>The argument <code>pch</code> sets the type of plotting character to use in the plot (for a graphical key of the available plotting characters see <a href="https://www.statmethods.net/advgraphs/parameters.html">this link</a>) and <code>color</code> sets plotting character colors. We can change the angle of the 3D plot using the <code>angle</code> argument:</p>
<div class="sourceCode" id="cb383"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb383-1"><a href="linear-regression-models-ii.html#cb383-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scatterplot3d</span>(trees, </span>
<span id="cb383-2"><a href="linear-regression-models-ii.html#cb383-2" aria-hidden="true" tabindex="-1"></a>              <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">color=</span><span class="st">&quot;steelblue&quot;</span>,</span>
<span id="cb383-3"><a href="linear-regression-models-ii.html#cb383-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">angle=</span><span class="dv">75</span>,</span>
<span id="cb383-4"><a href="linear-regression-models-ii.html#cb383-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">main =</span> <span class="st">&#39;Tree Volume as</span><span class="sc">\n</span><span class="st">a function of Girth and Height&#39;</span>)</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-380-1.png" width="576" /></p>
<p>We can add vertical lines to the plot using the <code>type</code> argument and remove the box around the plot:</p>
<div class="sourceCode" id="cb384"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb384-1"><a href="linear-regression-models-ii.html#cb384-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scatterplot3d</span>(trees, </span>
<span id="cb384-2"><a href="linear-regression-models-ii.html#cb384-2" aria-hidden="true" tabindex="-1"></a>              <span class="at">pch =</span> <span class="dv">16</span>, </span>
<span id="cb384-3"><a href="linear-regression-models-ii.html#cb384-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">color=</span><span class="st">&quot;steelblue&quot;</span>,</span>
<span id="cb384-4"><a href="linear-regression-models-ii.html#cb384-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">angle=</span><span class="dv">75</span>,</span>
<span id="cb384-5"><a href="linear-regression-models-ii.html#cb384-5" aria-hidden="true" tabindex="-1"></a>              <span class="at">box =</span> <span class="cn">FALSE</span>,</span>
<span id="cb384-6"><a href="linear-regression-models-ii.html#cb384-6" aria-hidden="true" tabindex="-1"></a>              <span class="at">type =</span> <span class="st">&quot;h&quot;</span>,</span>
<span id="cb384-7"><a href="linear-regression-models-ii.html#cb384-7" aria-hidden="true" tabindex="-1"></a>              <span class="at">main =</span> <span class="st">&#39;Tree Volume as</span><span class="sc">\n</span><span class="st">a function of Girth and Height&#39;</span>)</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-381-1.png" width="576" /></p>
<p>For more examples of how you can modify plots generated with the <code>scatterplot3d</code> package see <a href="http://www.sthda.com/english/wiki/scatterplot3d-3d-graphics-r-software-and-data-visualization#change-the-global-appearance-of-the-graph">this web page</a>).</p>
<!-- ### rgl -->
<!-- The package `rgl` is another package that we can use for 3D visualization. `rgl` is powerful because it lets us create interactive plots we can rotate and zoom in/out on.  -->
<!-- You can then create an interactive 3D plot as so: -->
<!-- ```{r} -->
<!-- library(rgl) -->
<!-- # create 3D scatter, using spheres to draw points -->
<!-- plot3d(trees$Girth, trees$Height, trees$Volume, -->
<!--       xlab = "Girth", ylab = "Height", zlab = "Volume", -->
<!--        type = "s",  -->
<!--        size = 1.5, -->
<!--        col = "red") -->
<!-- rglwidget() # only need to include this line if using in a markdown document -->
<!-- ``` -->
</div>
</div>
<div id="fitting-the-regression-model" class="section level2" number="12.10">
<h2><span class="header-section-number">12.10</span> Fitting the regression model</h2>
<p>From the 3D scatter plot it looks like we ought to be able to find a plane through the data that fits the scatter fairly well.</p>
<p>Let’s use the <code>lm()</code> function to calculate the multiple regression and <code>summary()</code> to get the details of the model:</p>
<div class="sourceCode" id="cb385"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb385-1"><a href="linear-regression-models-ii.html#cb385-1" aria-hidden="true" tabindex="-1"></a>fit.trees <span class="ot">&lt;-</span> <span class="fu">lm</span>(Volume <span class="sc">~</span> Girth <span class="sc">+</span> Height, <span class="at">data=</span>trees)</span>
<span id="cb385-2"><a href="linear-regression-models-ii.html#cb385-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb385-3"><a href="linear-regression-models-ii.html#cb385-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(fit.trees)</span>
<span id="cb385-4"><a href="linear-regression-models-ii.html#cb385-4" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb385-5"><a href="linear-regression-models-ii.html#cb385-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Call:</span></span>
<span id="cb385-6"><a href="linear-regression-models-ii.html#cb385-6" aria-hidden="true" tabindex="-1"></a><span class="do">## lm(formula = Volume ~ Girth + Height, data = trees)</span></span>
<span id="cb385-7"><a href="linear-regression-models-ii.html#cb385-7" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb385-8"><a href="linear-regression-models-ii.html#cb385-8" aria-hidden="true" tabindex="-1"></a><span class="do">## Residuals:</span></span>
<span id="cb385-9"><a href="linear-regression-models-ii.html#cb385-9" aria-hidden="true" tabindex="-1"></a><span class="do">##     Min      1Q  Median      3Q     Max </span></span>
<span id="cb385-10"><a href="linear-regression-models-ii.html#cb385-10" aria-hidden="true" tabindex="-1"></a><span class="do">## -6.4065 -2.6493 -0.2876  2.2003  8.4847 </span></span>
<span id="cb385-11"><a href="linear-regression-models-ii.html#cb385-11" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb385-12"><a href="linear-regression-models-ii.html#cb385-12" aria-hidden="true" tabindex="-1"></a><span class="do">## Coefficients:</span></span>
<span id="cb385-13"><a href="linear-regression-models-ii.html#cb385-13" aria-hidden="true" tabindex="-1"></a><span class="do">##             Estimate Std. Error t value Pr(&gt;|t|)    </span></span>
<span id="cb385-14"><a href="linear-regression-models-ii.html#cb385-14" aria-hidden="true" tabindex="-1"></a><span class="do">## (Intercept) -57.9877     8.6382  -6.713 2.75e-07 ***</span></span>
<span id="cb385-15"><a href="linear-regression-models-ii.html#cb385-15" aria-hidden="true" tabindex="-1"></a><span class="do">## Girth         4.7082     0.2643  17.816  &lt; 2e-16 ***</span></span>
<span id="cb385-16"><a href="linear-regression-models-ii.html#cb385-16" aria-hidden="true" tabindex="-1"></a><span class="do">## Height        0.3393     0.1302   2.607   0.0145 *  </span></span>
<span id="cb385-17"><a href="linear-regression-models-ii.html#cb385-17" aria-hidden="true" tabindex="-1"></a><span class="do">## ---</span></span>
<span id="cb385-18"><a href="linear-regression-models-ii.html#cb385-18" aria-hidden="true" tabindex="-1"></a><span class="do">## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</span></span>
<span id="cb385-19"><a href="linear-regression-models-ii.html#cb385-19" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb385-20"><a href="linear-regression-models-ii.html#cb385-20" aria-hidden="true" tabindex="-1"></a><span class="do">## Residual standard error: 3.882 on 28 degrees of freedom</span></span>
<span id="cb385-21"><a href="linear-regression-models-ii.html#cb385-21" aria-hidden="true" tabindex="-1"></a><span class="do">## Multiple R-squared:  0.948,  Adjusted R-squared:  0.9442 </span></span>
<span id="cb385-22"><a href="linear-regression-models-ii.html#cb385-22" aria-hidden="true" tabindex="-1"></a><span class="do">## F-statistic:   255 on 2 and 28 DF,  p-value: &lt; 2.2e-16</span></span></code></pre></div>
<div id="visualizing-the-regression-model-in-scatterplot3d" class="section level3" number="12.10.1">
<h3><span class="header-section-number">12.10.1</span> Visualizing the regression model in scatterplot3d</h3>
<p>To visualize the multiple regression, let’s use the <code>scatterplot3d</code> package to draw the 3D scatter of plots and the plane that corresponds to the regression model:</p>
<div class="sourceCode" id="cb386"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb386-1"><a href="linear-regression-models-ii.html#cb386-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">scatterplot3d</span>(trees, <span class="at">angle=</span><span class="dv">55</span>,<span class="at">type=</span><span class="st">&#39;h&#39;</span>,</span>
<span id="cb386-2"><a href="linear-regression-models-ii.html#cb386-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">pch =</span> <span class="dv">16</span>, <span class="at">color =</span> <span class="st">&quot;steelblue&quot;</span>,</span>
<span id="cb386-3"><a href="linear-regression-models-ii.html#cb386-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">main =</span> <span class="st">&#39;Tree Volume as</span><span class="sc">\n</span><span class="st">a function of Girth and Height&#39;</span>)</span>
<span id="cb386-4"><a href="linear-regression-models-ii.html#cb386-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb386-5"><a href="linear-regression-models-ii.html#cb386-5" aria-hidden="true" tabindex="-1"></a><span class="co"># add a plane representing the fit of the model</span></span>
<span id="cb386-6"><a href="linear-regression-models-ii.html#cb386-6" aria-hidden="true" tabindex="-1"></a>p<span class="sc">$</span><span class="fu">plane3d</span>(fit.trees, <span class="at">col=</span><span class="st">&#39;orangered&#39;</span>)</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-383-1.png" width="576" /></p>
<!-- If instead of `scatterplot3d`, we wanted to use `rgl` to depict the model fit we can use the `rgl.planes` function as shown below. -->
<!-- ```{r, eval=FALSE} -->
<!-- coefs <- coef(fit.trees) -->
<!-- b1 <- coefs["Girth"] -->
<!-- b2 <- coefs["Height"] -->
<!-- c <- -1 -->
<!-- a <- coefs["(Intercept)"] -->
<!-- plot3d(trees$Girth, trees$Height, trees$Volume, -->
<!--       xlab = "Girth", ylab = "Height", zlab = "Volume",  -->
<!--        type = "s",  -->
<!--        size = 1.5, -->
<!--        col = "red") -->
<!-- rgl.planes(b1, b2, c, a, alpha = 0.8, color = "gray") -->
<!-- rglwidget() -->
<!-- ``` -->
<p>From the figures it looks like the regression model fits pretty well, as we anticipated from the pairwise relationships.</p>
</div>
</div>
<div id="interpretting-the-regression-model" class="section level2" number="12.11">
<h2><span class="header-section-number">12.11</span> Interpretting the regression model</h2>
<p>The regression equation is: <span class="math inline">\(\hat{y}\)</span> = + <span class="math inline">\(x_1\)</span> +<span class="math inline">\(x_2\)</span>, where <span class="math inline">\(y\)</span> is Volume, and <span class="math inline">\(x_1\)</span> and <span class="math inline">\(x_2\)</span> are Girth and Height respectively. Since they’re on different scales the coefficients for Girth and Height aren’t directly comparable. Both coefficients are significant at the <span class="math inline">\(p&lt;0.05\)</span> level, but note that Girth is the much stronger predictor. In fact the addition of height explains only a minor additional fraction of variation in tree volume, so from the lumberjack’s perspective the additional trouble of measuring height probably isn’t worth it.</p>
</div>
<div id="exploring-the-vector-geometry-of-the-regression-model" class="section level2" number="12.12">
<h2><span class="header-section-number">12.12</span> Exploring the Vector Geometry of the Regression Model</h2>
<p>Recall the <code>broom:tidy</code> produces a tabular summary of the coefficients of the model and their associated statistics:</p>
<div class="sourceCode" id="cb387"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb387-1"><a href="linear-regression-models-ii.html#cb387-1" aria-hidden="true" tabindex="-1"></a>broom<span class="sc">::</span><span class="fu">tidy</span>(fit.trees)</span>
<span id="cb387-2"><a href="linear-regression-models-ii.html#cb387-2" aria-hidden="true" tabindex="-1"></a><span class="do">## # A tibble: 3 x 5</span></span>
<span id="cb387-3"><a href="linear-regression-models-ii.html#cb387-3" aria-hidden="true" tabindex="-1"></a><span class="do">##   term        estimate std.error statistic  p.value</span></span>
<span id="cb387-4"><a href="linear-regression-models-ii.html#cb387-4" aria-hidden="true" tabindex="-1"></a><span class="do">##   &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;</span></span>
<span id="cb387-5"><a href="linear-regression-models-ii.html#cb387-5" aria-hidden="true" tabindex="-1"></a><span class="do">## 1 (Intercept)  -58.0       8.64      -6.71 2.75e- 7</span></span>
<span id="cb387-6"><a href="linear-regression-models-ii.html#cb387-6" aria-hidden="true" tabindex="-1"></a><span class="do">## 2 Girth          4.71      0.264     17.8  8.22e-17</span></span>
<span id="cb387-7"><a href="linear-regression-models-ii.html#cb387-7" aria-hidden="true" tabindex="-1"></a><span class="do">## 3 Height         0.339     0.130      2.61 1.45e- 2</span></span></code></pre></div>
<p><code>broom:glance</code> provides information about the fit of the model:</p>
<div class="sourceCode" id="cb388"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb388-1"><a href="linear-regression-models-ii.html#cb388-1" aria-hidden="true" tabindex="-1"></a>broom<span class="sc">::</span><span class="fu">glance</span>(fit.trees)</span>
<span id="cb388-2"><a href="linear-regression-models-ii.html#cb388-2" aria-hidden="true" tabindex="-1"></a><span class="do">## # A tibble: 1 x 12</span></span>
<span id="cb388-3"><a href="linear-regression-models-ii.html#cb388-3" aria-hidden="true" tabindex="-1"></a><span class="do">##   r.squared adj.r.squared sigma statistic  p.value    df logLik   AIC   BIC</span></span>
<span id="cb388-4"><a href="linear-regression-models-ii.html#cb388-4" aria-hidden="true" tabindex="-1"></a><span class="do">##       &lt;dbl&gt;         &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span id="cb388-5"><a href="linear-regression-models-ii.html#cb388-5" aria-hidden="true" tabindex="-1"></a><span class="do">## 1     0.948         0.944  3.88      255. 1.07e-18     2  -84.5  177.  183.</span></span>
<span id="cb388-6"><a href="linear-regression-models-ii.html#cb388-6" aria-hidden="true" tabindex="-1"></a><span class="do">## # … with 3 more variables: deviance &lt;dbl&gt;, df.residual &lt;int&gt;, nobs &lt;int&gt;</span></span></code></pre></div>
<p>Let’s use our knowledge of vector geometry to further explore the relationship between the predicted Volume and the predictor variables. By definition the vector representing the predicted values lies in the subspace (in this case a plane) defined by Height and Girth, so let’s do some simple calculations to understand their length and angular relationships:</p>
<div class="sourceCode" id="cb389"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb389-1"><a href="linear-regression-models-ii.html#cb389-1" aria-hidden="true" tabindex="-1"></a><span class="co"># proportional to length of vectors</span></span>
<span id="cb389-2"><a href="linear-regression-models-ii.html#cb389-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sd</span>(fit.trees<span class="sc">$</span>fitted.values)</span>
<span id="cb389-3"><a href="linear-regression-models-ii.html#cb389-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 16.00434</span></span>
<span id="cb389-4"><a href="linear-regression-models-ii.html#cb389-4" aria-hidden="true" tabindex="-1"></a><span class="fu">sd</span>(trees<span class="sc">$</span>Height)</span>
<span id="cb389-5"><a href="linear-regression-models-ii.html#cb389-5" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 6.371813</span></span>
<span id="cb389-6"><a href="linear-regression-models-ii.html#cb389-6" aria-hidden="true" tabindex="-1"></a><span class="fu">sd</span>(trees<span class="sc">$</span>Girth)</span>
<span id="cb389-7"><a href="linear-regression-models-ii.html#cb389-7" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 3.138139</span></span>
<span id="cb389-8"><a href="linear-regression-models-ii.html#cb389-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb389-9"><a href="linear-regression-models-ii.html#cb389-9" aria-hidden="true" tabindex="-1"></a><span class="co"># cosines of angles btw vectors</span></span>
<span id="cb389-10"><a href="linear-regression-models-ii.html#cb389-10" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(trees<span class="sc">$</span>Height, trees<span class="sc">$</span>Girth)</span>
<span id="cb389-11"><a href="linear-regression-models-ii.html#cb389-11" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 0.5192801</span></span>
<span id="cb389-12"><a href="linear-regression-models-ii.html#cb389-12" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(trees<span class="sc">$</span>Girth, fit.trees<span class="sc">$</span>fitted.values)</span>
<span id="cb389-13"><a href="linear-regression-models-ii.html#cb389-13" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 0.9933158</span></span>
<span id="cb389-14"><a href="linear-regression-models-ii.html#cb389-14" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(trees<span class="sc">$</span>Height, fit.trees<span class="sc">$</span>fitted.values)</span>
<span id="cb389-15"><a href="linear-regression-models-ii.html#cb389-15" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 0.6144545</span></span>
<span id="cb389-16"><a href="linear-regression-models-ii.html#cb389-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb389-17"><a href="linear-regression-models-ii.html#cb389-17" aria-hidden="true" tabindex="-1"></a><span class="co"># angles btw vectors in degrees</span></span>
<span id="cb389-18"><a href="linear-regression-models-ii.html#cb389-18" aria-hidden="true" tabindex="-1"></a><span class="fu">acos</span>(<span class="fu">cor</span>(trees<span class="sc">$</span>Girth, trees<span class="sc">$</span>Height)) <span class="sc">*</span> (<span class="dv">180</span><span class="sc">/</span>pi)</span>
<span id="cb389-19"><a href="linear-regression-models-ii.html#cb389-19" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 58.71603</span></span>
<span id="cb389-20"><a href="linear-regression-models-ii.html#cb389-20" aria-hidden="true" tabindex="-1"></a><span class="fu">acos</span>(<span class="fu">cor</span>(trees<span class="sc">$</span>Girth, fit.trees<span class="sc">$</span>fitted.values)) <span class="sc">*</span> (<span class="dv">180</span><span class="sc">/</span>pi)</span>
<span id="cb389-21"><a href="linear-regression-models-ii.html#cb389-21" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 6.628322</span></span>
<span id="cb389-22"><a href="linear-regression-models-ii.html#cb389-22" aria-hidden="true" tabindex="-1"></a><span class="fu">acos</span>(<span class="fu">cor</span>(trees<span class="sc">$</span>Height, fit.trees<span class="sc">$</span>fitted.values)) <span class="sc">*</span> (<span class="dv">180</span><span class="sc">/</span>pi)</span>
<span id="cb389-23"><a href="linear-regression-models-ii.html#cb389-23" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 52.08771</span></span></code></pre></div>
<p>Notice that <span class="math inline">\(\text{Girth}\)</span> is very highly correlated with the <span class="math inline">\(\widehat{\text{Volume}}\)</span> and hence the angle between these two vectors is very small (about 6.6 degrees). By contrast, <span class="math inline">\(\text{Height}\)</span> is only moderately correlated with <span class="math inline">\(\widehat{\text{Volume}}\)</span> and the angle between them is significantly larger (about 52 degrees).</p>
</div>
<div id="exploring-the-residuals-from-the-model-fit" class="section level2" number="12.13">
<h2><span class="header-section-number">12.13</span> Exploring the Residuals from the Model Fit</h2>
<p>Now let’s look at the residuals from the regression. The residuals represent the `unexplained’ variance:</p>
<div class="sourceCode" id="cb390"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb390-1"><a href="linear-regression-models-ii.html#cb390-1" aria-hidden="true" tabindex="-1"></a>trees.augmented <span class="ot">&lt;-</span> <span class="fu">augment</span>(fit.trees, trees)</span>
<span id="cb390-2"><a href="linear-regression-models-ii.html#cb390-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb390-3"><a href="linear-regression-models-ii.html#cb390-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(trees.augmented, <span class="fu">aes</span>(<span class="at">x =</span> Girth, <span class="at">y =</span> .resid)) <span class="sc">+</span> </span>
<span id="cb390-4"><a href="linear-regression-models-ii.html#cb390-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb390-5"><a href="linear-regression-models-ii.html#cb390-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">color=</span><span class="st">&#39;red&#39;</span>, <span class="at">linetype=</span><span class="st">&#39;dashed&#39;</span>) <span class="sc">+</span></span>
<span id="cb390-6"><a href="linear-regression-models-ii.html#cb390-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Girth&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Residuals&quot;</span>)</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-387-1.png" width="576" /></p>
<p>Ideally the residuals should be evenly scattered around zero, with no trends as we go from high to low values of the dependent variable. As you can see, the residuals are somewhat u-shaped or j-shaped suggesting that there may be a non-linear aspect of the relationship that our model isn’t capturing.</p>
</div>
<div id="an-alternate-model" class="section level2" number="12.14">
<h2><span class="header-section-number">12.14</span> An alternate model</h2>
<p>Let’s think about the relationships we’re actually modeling for a few minutes. For the sake of simplicity let’s consider the trunk of a tree to be a cylinder. How do the dimensions of this cylinder relate to its volume? You can look up the formula for the volume of a cylinder, but the key thing you’ll want to note is that volume of the cylinder should be proportional to a characteristic length of the cylinder cubed (<span class="math inline">\(V \propto \mathrm{L}^3\)</span>). This suggests that if we want to fit a linear model we should relate Girth and Height to <span class="math inline">\(\sqrt[3]{\mathrm{Volume}}\)</span>:</p>
<div class="sourceCode" id="cb391"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb391-1"><a href="linear-regression-models-ii.html#cb391-1" aria-hidden="true" tabindex="-1"></a>trees.cuberoot <span class="ot">&lt;-</span> <span class="fu">mutate</span>(trees, <span class="at">cuberoot.Volume =</span> Volume<span class="sc">^</span><span class="fl">0.33</span>)</span>
<span id="cb391-2"><a href="linear-regression-models-ii.html#cb391-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb391-3"><a href="linear-regression-models-ii.html#cb391-3" aria-hidden="true" tabindex="-1"></a>fit.trees.cuberoot <span class="ot">&lt;-</span> <span class="fu">lm</span>(cuberoot.Volume <span class="sc">~</span> Girth <span class="sc">+</span> Height, <span class="at">data =</span> trees.cuberoot)</span>
<span id="cb391-4"><a href="linear-regression-models-ii.html#cb391-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb391-5"><a href="linear-regression-models-ii.html#cb391-5" aria-hidden="true" tabindex="-1"></a>broom<span class="sc">::</span><span class="fu">glance</span>(fit.trees) <span class="co"># summary of fit of original model</span></span>
<span id="cb391-6"><a href="linear-regression-models-ii.html#cb391-6" aria-hidden="true" tabindex="-1"></a><span class="do">## # A tibble: 1 x 12</span></span>
<span id="cb391-7"><a href="linear-regression-models-ii.html#cb391-7" aria-hidden="true" tabindex="-1"></a><span class="do">##   r.squared adj.r.squared sigma statistic  p.value    df logLik   AIC   BIC</span></span>
<span id="cb391-8"><a href="linear-regression-models-ii.html#cb391-8" aria-hidden="true" tabindex="-1"></a><span class="do">##       &lt;dbl&gt;         &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span id="cb391-9"><a href="linear-regression-models-ii.html#cb391-9" aria-hidden="true" tabindex="-1"></a><span class="do">## 1     0.948         0.944  3.88      255. 1.07e-18     2  -84.5  177.  183.</span></span>
<span id="cb391-10"><a href="linear-regression-models-ii.html#cb391-10" aria-hidden="true" tabindex="-1"></a><span class="do">## # … with 3 more variables: deviance &lt;dbl&gt;, df.residual &lt;int&gt;, nobs &lt;int&gt;</span></span>
<span id="cb391-11"><a href="linear-regression-models-ii.html#cb391-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb391-12"><a href="linear-regression-models-ii.html#cb391-12" aria-hidden="true" tabindex="-1"></a>broom<span class="sc">::</span><span class="fu">glance</span>(fit.trees.cuberoot)  <span class="co"># summary of fit of alternate model</span></span>
<span id="cb391-13"><a href="linear-regression-models-ii.html#cb391-13" aria-hidden="true" tabindex="-1"></a><span class="do">## # A tibble: 1 x 12</span></span>
<span id="cb391-14"><a href="linear-regression-models-ii.html#cb391-14" aria-hidden="true" tabindex="-1"></a><span class="do">##   r.squared adj.r.squared  sigma statistic  p.value    df logLik   AIC   BIC</span></span>
<span id="cb391-15"><a href="linear-regression-models-ii.html#cb391-15" aria-hidden="true" tabindex="-1"></a><span class="do">##       &lt;dbl&gt;         &lt;dbl&gt;  &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span id="cb391-16"><a href="linear-regression-models-ii.html#cb391-16" aria-hidden="true" tabindex="-1"></a><span class="do">## 1     0.978         0.976 0.0811      612. 7.77e-24     2   35.5 -62.9 -57.2</span></span>
<span id="cb391-17"><a href="linear-regression-models-ii.html#cb391-17" aria-hidden="true" tabindex="-1"></a><span class="do">## # … with 3 more variables: deviance &lt;dbl&gt;, df.residual &lt;int&gt;, nobs &lt;int&gt;</span></span></code></pre></div>
<p>Comparing the summary tables, we see indeed that using the cube root of Volume improves the fit of our model some. Let’s examine the residuals of this alternate model.</p>
<div class="sourceCode" id="cb392"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb392-1"><a href="linear-regression-models-ii.html#cb392-1" aria-hidden="true" tabindex="-1"></a>trees.cuberoot <span class="ot">&lt;-</span> broom<span class="sc">::</span><span class="fu">augment</span>(fit.trees.cuberoot, trees.cuberoot)</span>
<span id="cb392-2"><a href="linear-regression-models-ii.html#cb392-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb392-3"><a href="linear-regression-models-ii.html#cb392-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(trees.cuberoot, <span class="fu">aes</span>(<span class="at">x =</span> cuberoot.Volume, <span class="at">y =</span> .resid)) <span class="sc">+</span> </span>
<span id="cb392-4"><a href="linear-regression-models-ii.html#cb392-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb392-5"><a href="linear-regression-models-ii.html#cb392-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">color=</span><span class="st">&#39;red&#39;</span>, <span class="at">linetype=</span><span class="st">&#39;dashed&#39;</span>) <span class="sc">+</span></span>
<span id="cb392-6"><a href="linear-regression-models-ii.html#cb392-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Girth&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Residuals&quot;</span>)</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-389-1.png" width="576" /></p>
<p>As we can see the transformation we applied to the data did seem to make our residuals more uniform across the range of observations.</p>
</div>
<div id="exploring-the-impact-of-nearly-collinear-predictors-on-regression" class="section level2" number="12.15">
<h2><span class="header-section-number">12.15</span> Exploring the impact of nearly collinear predictors on regression</h2>
<p>In lecture we discussed the problems that can arise in regression when your predictor variables are nearly collinear. In this section we’ll illustrate some of these issues.</p>
<p>Consider again the <code>trees</code> data set. Recall that two of the variables – Girth and Volume – are highly correlated and thus nearly collinear.</p>
<div class="sourceCode" id="cb393"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb393-1"><a href="linear-regression-models-ii.html#cb393-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(trees)</span>
<span id="cb393-2"><a href="linear-regression-models-ii.html#cb393-2" aria-hidden="true" tabindex="-1"></a><span class="do">##            Girth    Height    Volume</span></span>
<span id="cb393-3"><a href="linear-regression-models-ii.html#cb393-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Girth  1.0000000 0.5192801 0.9671194</span></span>
<span id="cb393-4"><a href="linear-regression-models-ii.html#cb393-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Height 0.5192801 1.0000000 0.5982497</span></span>
<span id="cb393-5"><a href="linear-regression-models-ii.html#cb393-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Volume 0.9671194 0.5982497 1.0000000</span></span></code></pre></div>
<p>Let’s explore what happens when we treat Height as the dependent variable, and Girth and Volume as the predictor variables.</p>
<div class="sourceCode" id="cb394"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb394-1"><a href="linear-regression-models-ii.html#cb394-1" aria-hidden="true" tabindex="-1"></a>fit.Height <span class="ot">&lt;-</span> <span class="fu">lm</span>(Height <span class="sc">~</span> Girth <span class="sc">+</span> Volume, <span class="at">data =</span> trees)</span>
<span id="cb394-2"><a href="linear-regression-models-ii.html#cb394-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb394-3"><a href="linear-regression-models-ii.html#cb394-3" aria-hidden="true" tabindex="-1"></a>broom<span class="sc">::</span><span class="fu">glance</span>(fit.Height)</span>
<span id="cb394-4"><a href="linear-regression-models-ii.html#cb394-4" aria-hidden="true" tabindex="-1"></a><span class="do">## # A tibble: 1 x 12</span></span>
<span id="cb394-5"><a href="linear-regression-models-ii.html#cb394-5" aria-hidden="true" tabindex="-1"></a><span class="do">##   r.squared adj.r.squared sigma statistic p.value    df logLik   AIC   BIC</span></span>
<span id="cb394-6"><a href="linear-regression-models-ii.html#cb394-6" aria-hidden="true" tabindex="-1"></a><span class="do">##       &lt;dbl&gt;         &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span id="cb394-7"><a href="linear-regression-models-ii.html#cb394-7" aria-hidden="true" tabindex="-1"></a><span class="do">## 1     0.412         0.370  5.06      9.82 5.87e-4     2  -92.6  193.  199.</span></span>
<span id="cb394-8"><a href="linear-regression-models-ii.html#cb394-8" aria-hidden="true" tabindex="-1"></a><span class="do">## # … with 3 more variables: deviance &lt;dbl&gt;, df.residual &lt;int&gt;, nobs &lt;int&gt;</span></span></code></pre></div>
<p>We can, of course, fit the linear model despite the near collinearity, and we find that the model does have some predictive power, with <span class="math inline">\(R^2 = 0.41\)</span>, and with Volume being the more significant predictor.</p>
<p>Now, let’s created a slightly different version of the trees data set by add some noise to the three variables. Our goal here is to simulate a data set we might have created had we measured a slightly different set of trees during our sampling. We’ll use the <code>jitter()</code> function to add uniform noise to the data set.</p>
<div class="sourceCode" id="cb395"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb395-1"><a href="linear-regression-models-ii.html#cb395-1" aria-hidden="true" tabindex="-1"></a>jitter.Girth <span class="ot">&lt;-</span> <span class="fu">jitter</span>(trees<span class="sc">$</span>Girth, <span class="at">amount=</span> <span class="fl">0.5</span> <span class="sc">*</span> <span class="fu">sd</span>(trees<span class="sc">$</span>Girth))</span>
<span id="cb395-2"><a href="linear-regression-models-ii.html#cb395-2" aria-hidden="true" tabindex="-1"></a>jitter.Height <span class="ot">&lt;-</span> <span class="fu">jitter</span>(trees<span class="sc">$</span>Height, <span class="at">amount=</span> <span class="fl">0.5</span> <span class="sc">*</span> <span class="fu">sd</span>(trees<span class="sc">$</span>Height))</span>
<span id="cb395-3"><a href="linear-regression-models-ii.html#cb395-3" aria-hidden="true" tabindex="-1"></a>jitter.Volume <span class="ot">&lt;-</span> <span class="fu">jitter</span>(trees<span class="sc">$</span>Volume, <span class="at">amount=</span> <span class="fl">0.5</span> <span class="sc">*</span> <span class="fu">sd</span>(trees<span class="sc">$</span>Volume))</span>
<span id="cb395-4"><a href="linear-regression-models-ii.html#cb395-4" aria-hidden="true" tabindex="-1"></a>jitter.trees <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Girth =</span> jitter.Girth, </span>
<span id="cb395-5"><a href="linear-regression-models-ii.html#cb395-5" aria-hidden="true" tabindex="-1"></a>                        <span class="at">Height =</span> jitter.Height, </span>
<span id="cb395-6"><a href="linear-regression-models-ii.html#cb395-6" aria-hidden="true" tabindex="-1"></a>                        <span class="at">Volume =</span> jitter.Volume)</span></code></pre></div>
<p>Here we added uniform noise proportional to the one-quarter the standard deviation of each variable. Let’s take a moment to convince ourselves that our new data set, <code>jitter.trees</code>, is not too different from the <code>trees</code> data set from which it was derived.</p>
<div class="sourceCode" id="cb396"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb396-1"><a href="linear-regression-models-ii.html#cb396-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">20190227</span>)</span>
<span id="cb396-2"><a href="linear-regression-models-ii.html#cb396-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb396-3"><a href="linear-regression-models-ii.html#cb396-3" aria-hidden="true" tabindex="-1"></a><span class="co"># compare this to broom::tidy(trees)</span></span>
<span id="cb396-4"><a href="linear-regression-models-ii.html#cb396-4" aria-hidden="true" tabindex="-1"></a>broom<span class="sc">::</span><span class="fu">tidy</span>(jitter.trees)</span>
<span id="cb396-5"><a href="linear-regression-models-ii.html#cb396-5" aria-hidden="true" tabindex="-1"></a><span class="do">## # A tibble: 3 x 13</span></span>
<span id="cb396-6"><a href="linear-regression-models-ii.html#cb396-6" aria-hidden="true" tabindex="-1"></a><span class="do">##   column     n  mean    sd median trimmed   mad   min   max range   skew</span></span>
<span id="cb396-7"><a href="linear-regression-models-ii.html#cb396-7" aria-hidden="true" tabindex="-1"></a><span class="do">##   &lt;chr&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;</span></span>
<span id="cb396-8"><a href="linear-regression-models-ii.html#cb396-8" aria-hidden="true" tabindex="-1"></a><span class="do">## 1 Girth     31  13.2  3.36   12.6    13.0  2.48  8.78  20.7  11.9  0.563</span></span>
<span id="cb396-9"><a href="linear-regression-models-ii.html#cb396-9" aria-hidden="true" tabindex="-1"></a><span class="do">## 2 Height    31  76.1  6.29   77.2    76.5  3.84 63.2   85.6  22.4 -0.445</span></span>
<span id="cb396-10"><a href="linear-regression-models-ii.html#cb396-10" aria-hidden="true" tabindex="-1"></a><span class="do">## 3 Volume    31  30.2 15.9    27.3    28.9  9.82  7.29  73.8  66.5  0.849</span></span>
<span id="cb396-11"><a href="linear-regression-models-ii.html#cb396-11" aria-hidden="true" tabindex="-1"></a><span class="do">## # … with 2 more variables: kurtosis &lt;dbl&gt;, se &lt;dbl&gt;</span></span>
<span id="cb396-12"><a href="linear-regression-models-ii.html#cb396-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb396-13"><a href="linear-regression-models-ii.html#cb396-13" aria-hidden="true" tabindex="-1"></a><span class="co"># correlations among jittered variables are</span></span>
<span id="cb396-14"><a href="linear-regression-models-ii.html#cb396-14" aria-hidden="true" tabindex="-1"></a><span class="co"># similar to those of the original variables</span></span>
<span id="cb396-15"><a href="linear-regression-models-ii.html#cb396-15" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(jitter.trees)</span>
<span id="cb396-16"><a href="linear-regression-models-ii.html#cb396-16" aria-hidden="true" tabindex="-1"></a><span class="do">##            Girth    Height    Volume</span></span>
<span id="cb396-17"><a href="linear-regression-models-ii.html#cb396-17" aria-hidden="true" tabindex="-1"></a><span class="do">## Girth  1.0000000 0.4571782 0.8781239</span></span>
<span id="cb396-18"><a href="linear-regression-models-ii.html#cb396-18" aria-hidden="true" tabindex="-1"></a><span class="do">## Height 0.4571782 1.0000000 0.4509685</span></span>
<span id="cb396-19"><a href="linear-regression-models-ii.html#cb396-19" aria-hidden="true" tabindex="-1"></a><span class="do">## Volume 0.8781239 0.4509685 1.0000000</span></span>
<span id="cb396-20"><a href="linear-regression-models-ii.html#cb396-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb396-21"><a href="linear-regression-models-ii.html#cb396-21" aria-hidden="true" tabindex="-1"></a><span class="do">## jittered variables are highly correlatd with original variables</span></span>
<span id="cb396-22"><a href="linear-regression-models-ii.html#cb396-22" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(trees<span class="sc">$</span>Height, jitter.trees<span class="sc">$</span>Height)</span>
<span id="cb396-23"><a href="linear-regression-models-ii.html#cb396-23" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 0.9506738</span></span>
<span id="cb396-24"><a href="linear-regression-models-ii.html#cb396-24" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(trees<span class="sc">$</span>Girth, jitter.trees<span class="sc">$</span>Girth)</span>
<span id="cb396-25"><a href="linear-regression-models-ii.html#cb396-25" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 0.9573967</span></span>
<span id="cb396-26"><a href="linear-regression-models-ii.html#cb396-26" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(trees<span class="sc">$</span>Volume, jitter.trees<span class="sc">$</span>Volume)</span>
<span id="cb396-27"><a href="linear-regression-models-ii.html#cb396-27" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 0.9478109</span></span></code></pre></div>
<p>Now that we’ve convinced ourselves that our jittered data set is a decent approximation to our original data set, let’s re-calculate the linear regression, and compare the coefficients of the jittered model to the original model:</p>
<div class="sourceCode" id="cb397"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb397-1"><a href="linear-regression-models-ii.html#cb397-1" aria-hidden="true" tabindex="-1"></a>fit.Height.jitter <span class="ot">&lt;-</span> <span class="fu">lm</span>(Height <span class="sc">~</span> Girth <span class="sc">+</span> Volume, <span class="at">data =</span> jitter.trees)</span>
<span id="cb397-2"><a href="linear-regression-models-ii.html#cb397-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb397-3"><a href="linear-regression-models-ii.html#cb397-3" aria-hidden="true" tabindex="-1"></a>broom<span class="sc">::</span><span class="fu">tidy</span>(fit.Height)</span>
<span id="cb397-4"><a href="linear-regression-models-ii.html#cb397-4" aria-hidden="true" tabindex="-1"></a><span class="do">## # A tibble: 3 x 5</span></span>
<span id="cb397-5"><a href="linear-regression-models-ii.html#cb397-5" aria-hidden="true" tabindex="-1"></a><span class="do">##   term        estimate std.error statistic  p.value</span></span>
<span id="cb397-6"><a href="linear-regression-models-ii.html#cb397-6" aria-hidden="true" tabindex="-1"></a><span class="do">##   &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;</span></span>
<span id="cb397-7"><a href="linear-regression-models-ii.html#cb397-7" aria-hidden="true" tabindex="-1"></a><span class="do">## 1 (Intercept)   83.3       9.09       9.17 6.33e-10</span></span>
<span id="cb397-8"><a href="linear-regression-models-ii.html#cb397-8" aria-hidden="true" tabindex="-1"></a><span class="do">## 2 Girth         -1.86      1.16      -1.61 1.19e- 1</span></span>
<span id="cb397-9"><a href="linear-regression-models-ii.html#cb397-9" aria-hidden="true" tabindex="-1"></a><span class="do">## 3 Volume         0.576     0.221      2.61 1.45e- 2</span></span>
<span id="cb397-10"><a href="linear-regression-models-ii.html#cb397-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb397-11"><a href="linear-regression-models-ii.html#cb397-11" aria-hidden="true" tabindex="-1"></a>broom<span class="sc">::</span><span class="fu">tidy</span>(fit.Height.jitter)</span>
<span id="cb397-12"><a href="linear-regression-models-ii.html#cb397-12" aria-hidden="true" tabindex="-1"></a><span class="do">## # A tibble: 3 x 5</span></span>
<span id="cb397-13"><a href="linear-regression-models-ii.html#cb397-13" aria-hidden="true" tabindex="-1"></a><span class="do">##   term        estimate std.error statistic  p.value</span></span>
<span id="cb397-14"><a href="linear-regression-models-ii.html#cb397-14" aria-hidden="true" tabindex="-1"></a><span class="do">##   &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;</span></span>
<span id="cb397-15"><a href="linear-regression-models-ii.html#cb397-15" aria-hidden="true" tabindex="-1"></a><span class="do">## 1 (Intercept)  66.9        5.45     12.3   8.63e-13</span></span>
<span id="cb397-16"><a href="linear-regression-models-ii.html#cb397-16" aria-hidden="true" tabindex="-1"></a><span class="do">## 2 Girth         0.501      0.655     0.766 4.50e- 1</span></span>
<span id="cb397-17"><a href="linear-regression-models-ii.html#cb397-17" aria-hidden="true" tabindex="-1"></a><span class="do">## 3 Volume        0.0856     0.138     0.620 5.40e- 1</span></span></code></pre></div>
<p>We see that the coefficients of the linear model have changed substantially between the original data and the jittered data. Our model is unstable to relatively modest changes to the data!</p>
<p>Let’s draw some plots to illustrate how different the models fit to the original and jittered data are:</p>
<div class="sourceCode" id="cb398"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb398-1"><a href="linear-regression-models-ii.html#cb398-1" aria-hidden="true" tabindex="-1"></a><span class="co"># draw 3d scatter plots with small points so as not to obscure regression planes</span></span>
<span id="cb398-2"><a href="linear-regression-models-ii.html#cb398-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">scatterplot3d</span>(<span class="at">x=</span>trees<span class="sc">$</span>Girth, <span class="at">y=</span>trees<span class="sc">$</span>Volume, <span class="at">z=</span>trees<span class="sc">$</span>Height, </span>
<span id="cb398-3"><a href="linear-regression-models-ii.html#cb398-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">angle=</span><span class="dv">15</span>, <span class="at">type=</span><span class="st">&#39;p&#39;</span>, <span class="at">pch=</span><span class="st">&#39;.&#39;</span>)</span>
<span id="cb398-4"><a href="linear-regression-models-ii.html#cb398-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb398-5"><a href="linear-regression-models-ii.html#cb398-5" aria-hidden="true" tabindex="-1"></a><span class="co"># original model</span></span>
<span id="cb398-6"><a href="linear-regression-models-ii.html#cb398-6" aria-hidden="true" tabindex="-1"></a>p<span class="sc">$</span><span class="fu">plane3d</span>(fit.Height, <span class="at">col=</span><span class="st">&#39;orangered&#39;</span>)</span>
<span id="cb398-7"><a href="linear-regression-models-ii.html#cb398-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb398-8"><a href="linear-regression-models-ii.html#cb398-8" aria-hidden="true" tabindex="-1"></a><span class="co"># jittered model</span></span>
<span id="cb398-9"><a href="linear-regression-models-ii.html#cb398-9" aria-hidden="true" tabindex="-1"></a>p<span class="sc">$</span><span class="fu">plane3d</span>(fit.Height.jitter, <span class="at">col=</span><span class="st">&#39;blue&#39;</span>)</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-395-1.png" width="576" /></p>
<p>Let’s do the same comparison for the multiple regression of Volume on Height and Girth. In this case the predictor variables are  nearly collinear.</p>
<div class="sourceCode" id="cb399"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb399-1"><a href="linear-regression-models-ii.html#cb399-1" aria-hidden="true" tabindex="-1"></a>fit.Volume <span class="ot">&lt;-</span> <span class="fu">lm</span>(Volume <span class="sc">~</span> Girth <span class="sc">+</span> Height, <span class="at">data =</span> trees)</span>
<span id="cb399-2"><a href="linear-regression-models-ii.html#cb399-2" aria-hidden="true" tabindex="-1"></a>fit.Volume.jitter <span class="ot">&lt;-</span> <span class="fu">lm</span>(Volume <span class="sc">~</span> Girth <span class="sc">+</span> Height, <span class="at">data =</span> jitter.trees)</span>
<span id="cb399-3"><a href="linear-regression-models-ii.html#cb399-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb399-4"><a href="linear-regression-models-ii.html#cb399-4" aria-hidden="true" tabindex="-1"></a><span class="fu">coefficients</span>(fit.Volume)</span>
<span id="cb399-5"><a href="linear-regression-models-ii.html#cb399-5" aria-hidden="true" tabindex="-1"></a><span class="do">## (Intercept)       Girth      Height </span></span>
<span id="cb399-6"><a href="linear-regression-models-ii.html#cb399-6" aria-hidden="true" tabindex="-1"></a><span class="do">## -57.9876589   4.7081605   0.3392512</span></span>
<span id="cb399-7"><a href="linear-regression-models-ii.html#cb399-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb399-8"><a href="linear-regression-models-ii.html#cb399-8" aria-hidden="true" tabindex="-1"></a><span class="fu">coefficients</span>(fit.Volume.jitter)</span>
<span id="cb399-9"><a href="linear-regression-models-ii.html#cb399-9" aria-hidden="true" tabindex="-1"></a><span class="do">## (Intercept)       Girth      Height </span></span>
<span id="cb399-10"><a href="linear-regression-models-ii.html#cb399-10" aria-hidden="true" tabindex="-1"></a><span class="do">##  -34.893773    4.025177    0.158098</span></span></code></pre></div>
<p>For this model, we see that the coefficients have changed only a small amount. The underlying data, <code>jitter.trees</code>, is the same in both cases, but now our model is stable because the predictor variables are only modestly correlated with each other.</p>
<p>Let’s generate another plot to illustrate the similarity of the models fit to the original and jittered data when Girth and Height are used to predict Volume.</p>
<div class="sourceCode" id="cb400"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb400-1"><a href="linear-regression-models-ii.html#cb400-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">scatterplot3d</span>(<span class="at">x=</span>trees<span class="sc">$</span>Girth, <span class="at">y=</span>trees<span class="sc">$</span>Height, <span class="at">z=</span>trees<span class="sc">$</span>Volume, </span>
<span id="cb400-2"><a href="linear-regression-models-ii.html#cb400-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">angle=</span><span class="dv">55</span>, <span class="at">type=</span><span class="st">&#39;p&#39;</span>, <span class="at">pch=</span><span class="st">&#39;.&#39;</span>)</span>
<span id="cb400-3"><a href="linear-regression-models-ii.html#cb400-3" aria-hidden="true" tabindex="-1"></a>p<span class="sc">$</span><span class="fu">plane3d</span>(fit.Volume, <span class="at">col=</span><span class="st">&#39;orangered&#39;</span>)</span>
<span id="cb400-4"><a href="linear-regression-models-ii.html#cb400-4" aria-hidden="true" tabindex="-1"></a>p<span class="sc">$</span><span class="fu">plane3d</span>(fit.Volume.jitter, <span class="at">col=</span><span class="st">&#39;blue&#39;</span>)</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-397-1.png" width="576" /></p>
<p>Finally, let’s do some vector calculations to quantify how the angular deviation between the fit data and the predictor variables changes between the original and jittered data set for the two different multiple regressions:</p>
<div class="sourceCode" id="cb401"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb401-1"><a href="linear-regression-models-ii.html#cb401-1" aria-hidden="true" tabindex="-1"></a><span class="co"># write a quickie fxn to express angle between vectors in degrees</span></span>
<span id="cb401-2"><a href="linear-regression-models-ii.html#cb401-2" aria-hidden="true" tabindex="-1"></a>vec.angle <span class="ot">&lt;-</span> <span class="cf">function</span>(x,y) { <span class="fu">acos</span>(<span class="fu">cor</span>(x,y)) <span class="sc">*</span> (<span class="dv">180</span><span class="sc">/</span>pi)}</span>
<span id="cb401-3"><a href="linear-regression-models-ii.html#cb401-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb401-4"><a href="linear-regression-models-ii.html#cb401-4" aria-hidden="true" tabindex="-1"></a><span class="co"># vector angles for fit of Height ~ Girth + Volume (orig)</span></span>
<span id="cb401-5"><a href="linear-regression-models-ii.html#cb401-5" aria-hidden="true" tabindex="-1"></a><span class="fu">vec.angle</span>(fit.Height<span class="sc">$</span>fitted.values, trees<span class="sc">$</span>Girth)</span>
<span id="cb401-6"><a href="linear-regression-models-ii.html#cb401-6" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 36.02644</span></span>
<span id="cb401-7"><a href="linear-regression-models-ii.html#cb401-7" aria-hidden="true" tabindex="-1"></a><span class="fu">vec.angle</span>(fit.Height<span class="sc">$</span>fitted.values, trees<span class="sc">$</span>Volume)</span>
<span id="cb401-8"><a href="linear-regression-models-ii.html#cb401-8" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 21.29297</span></span>
<span id="cb401-9"><a href="linear-regression-models-ii.html#cb401-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb401-10"><a href="linear-regression-models-ii.html#cb401-10" aria-hidden="true" tabindex="-1"></a><span class="co"># vector angles for fit of Height ~ Girth + Volume (jittered)</span></span>
<span id="cb401-11"><a href="linear-regression-models-ii.html#cb401-11" aria-hidden="true" tabindex="-1"></a><span class="fu">vec.angle</span>(fit.Height.jitter<span class="sc">$</span>fitted.values, jitter.trees<span class="sc">$</span>Girth)</span>
<span id="cb401-12"><a href="linear-regression-models-ii.html#cb401-12" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 12.75399</span></span>
<span id="cb401-13"><a href="linear-regression-models-ii.html#cb401-13" aria-hidden="true" tabindex="-1"></a><span class="fu">vec.angle</span>(fit.Height.jitter<span class="sc">$</span>fitted.values, jitter.trees<span class="sc">$</span>Volume)</span>
<span id="cb401-14"><a href="linear-regression-models-ii.html#cb401-14" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 15.82913</span></span></code></pre></div>
<p>Now the same comparison for the non-collinear model <code>Volume ~ Girth + Height</code>.</p>
<div class="sourceCode" id="cb402"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb402-1"><a href="linear-regression-models-ii.html#cb402-1" aria-hidden="true" tabindex="-1"></a><span class="co"># vector angles for fit of Volume ~ Girth + Height (orig)</span></span>
<span id="cb402-2"><a href="linear-regression-models-ii.html#cb402-2" aria-hidden="true" tabindex="-1"></a><span class="fu">vec.angle</span>(fit.Volume<span class="sc">$</span>fitted.values, trees<span class="sc">$</span>Girth)</span>
<span id="cb402-3"><a href="linear-regression-models-ii.html#cb402-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 6.628322</span></span>
<span id="cb402-4"><a href="linear-regression-models-ii.html#cb402-4" aria-hidden="true" tabindex="-1"></a><span class="fu">vec.angle</span>(fit.Volume<span class="sc">$</span>fitted.values, trees<span class="sc">$</span>Height)</span>
<span id="cb402-5"><a href="linear-regression-models-ii.html#cb402-5" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 52.08771</span></span>
<span id="cb402-6"><a href="linear-regression-models-ii.html#cb402-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb402-7"><a href="linear-regression-models-ii.html#cb402-7" aria-hidden="true" tabindex="-1"></a><span class="co"># vector angles for fit of Volume ~ Girth + Height (jittered)</span></span>
<span id="cb402-8"><a href="linear-regression-models-ii.html#cb402-8" aria-hidden="true" tabindex="-1"></a><span class="fu">vec.angle</span>(fit.Volume.jitter<span class="sc">$</span>fitted.values, jitter.trees<span class="sc">$</span>Girth)</span>
<span id="cb402-9"><a href="linear-regression-models-ii.html#cb402-9" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 3.627345</span></span>
<span id="cb402-10"><a href="linear-regression-models-ii.html#cb402-10" aria-hidden="true" tabindex="-1"></a><span class="fu">vec.angle</span>(fit.Volume.jitter<span class="sc">$</span>fitted.values, jitter.trees<span class="sc">$</span>Height)</span>
<span id="cb402-11"><a href="linear-regression-models-ii.html#cb402-11" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 59.16748</span></span></code></pre></div>
<p>As these calculation illustrate, the change in the regression plane in the jittered date is much smaller when the dependent variable are not nearly colinear.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="linear-regression-models-i.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="principal-components-analysis.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "section"
},
"toolbar": {
"position": "fixed"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
