<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 13 More regression models | Biology 723: Statistical Computing for Biologists</title>
  <meta name="description" content="Hands-on materials for Bio 723 course at Duke University." />
  <meta name="generator" content="bookdown 0.17.2 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 13 More regression models | Biology 723: Statistical Computing for Biologists" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 13 More regression models | Biology 723: Statistical Computing for Biologists" />
  
  <meta name="twitter:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

<meta name="author" content="Paul M. Magwene" />


<meta name="date" content="2020-04-14" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="linear-regression-models-ii.html"/>
<link rel="next" href="principal-components-analysis.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/htmlwidgets-1.5.1/htmlwidgets.js"></script>
<link href="libs/rglwidgetClass-2/rgl.css" rel="stylesheet" />
<script src="libs/rglwidgetClass-2/rglClass.src.js"></script>
<script src="libs/CanvasMatrix4-2016/CanvasMatrix.src.js"></script>
<script src="libs/rglWebGL-binding-0.100.30/rglWebGL.js"></script>
<link href="libs/crosstalk-1.0.0/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.0.0/js/crosstalk.min.js"></script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { background-color: #f8f8f8; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#accessing-older-versions-of-the-course-notes"><i class="fa fa-check"></i><b>1.1</b> Accessing older versions of the course notes</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#how-to-use-these-lecture-notes"><i class="fa fa-check"></i><b>1.2</b> How to use these lecture notes</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html"><i class="fa fa-check"></i><b>2</b> Getting Started with R</a><ul>
<li class="chapter" data-level="2.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-r"><i class="fa fa-check"></i><b>2.1</b> What is R?</a></li>
<li class="chapter" data-level="2.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-rstudio"><i class="fa fa-check"></i><b>2.2</b> What is RStudio?</a></li>
<li class="chapter" data-level="2.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#entering-commands-in-the-console"><i class="fa fa-check"></i><b>2.3</b> Entering commands in the console</a></li>
<li class="chapter" data-level="2.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#comments"><i class="fa fa-check"></i><b>2.4</b> Comments</a></li>
<li class="chapter" data-level="2.5" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#using-r-as-a-calculator"><i class="fa fa-check"></i><b>2.5</b> Using R as a Calculator</a><ul>
<li class="chapter" data-level="2.5.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#common-mathematical-functions"><i class="fa fa-check"></i><b>2.5.1</b> Common mathematical functions</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#variable-assignment"><i class="fa fa-check"></i><b>2.6</b> Variable assignment</a><ul>
<li class="chapter" data-level="2.6.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#valid-variable-names"><i class="fa fa-check"></i><b>2.6.1</b> Valid variable names</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#data-types"><i class="fa fa-check"></i><b>2.7</b> Data types</a><ul>
<li class="chapter" data-level="2.7.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#numeric-data-types"><i class="fa fa-check"></i><b>2.7.1</b> Numeric data types</a></li>
<li class="chapter" data-level="2.7.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#logical-values"><i class="fa fa-check"></i><b>2.7.2</b> Logical values</a></li>
<li class="chapter" data-level="2.7.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#character-strings"><i class="fa fa-check"></i><b>2.7.3</b> Character strings</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#packages"><i class="fa fa-check"></i><b>2.8</b> Packages</a><ul>
<li class="chapter" data-level="2.8.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-console"><i class="fa fa-check"></i><b>2.8.1</b> Installing packages from the console</a></li>
<li class="chapter" data-level="2.8.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#install-the-tidyverse-package"><i class="fa fa-check"></i><b>2.8.2</b> Install the tidyverse package</a></li>
<li class="chapter" data-level="2.8.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-rstudio-dialog"><i class="fa fa-check"></i><b>2.8.3</b> Installing packages from the RStudio dialog</a></li>
<li class="chapter" data-level="2.8.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#loading-packages-with-the-library-function"><i class="fa fa-check"></i><b>2.8.4</b> Loading packages with the <code>library()</code> function</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#the-r-help-system"><i class="fa fa-check"></i><b>2.9</b> The R Help System</a><ul>
<li class="chapter" data-level="2.9.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#getting-help-from-the-console"><i class="fa fa-check"></i><b>2.9.1</b> Getting help from the console</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html"><i class="fa fa-check"></i><b>3</b> R Markdown and R Notebooks</a><ul>
<li class="chapter" data-level="3.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#r-notebooks"><i class="fa fa-check"></i><b>3.1</b> R Notebooks</a></li>
<li class="chapter" data-level="3.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#creating-an-r-notebook"><i class="fa fa-check"></i><b>3.2</b> Creating an R Notebook</a></li>
<li class="chapter" data-level="3.3" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#the-default-r-notebook-template"><i class="fa fa-check"></i><b>3.3</b> The default R Notebook template</a></li>
<li class="chapter" data-level="3.4" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-and-non-code-blocks"><i class="fa fa-check"></i><b>3.4</b> Code and Non-code blocks</a><ul>
<li class="chapter" data-level="3.4.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#non-code-blocks"><i class="fa fa-check"></i><b>3.4.1</b> Non-code blocks</a></li>
<li class="chapter" data-level="3.4.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-blocks"><i class="fa fa-check"></i><b>3.4.2</b> Code blocks</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-a-code-chunk"><i class="fa fa-check"></i><b>3.5</b> Running a code chunk</a></li>
<li class="chapter" data-level="3.6" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-all-code-chunks-above"><i class="fa fa-check"></i><b>3.6</b> Running all code chunks above</a></li>
<li class="chapter" data-level="3.7" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#knitting-an-r-markdown-to-html"><i class="fa fa-check"></i><b>3.7</b> “Knitting an” R Markdown to HTML</a></li>
<li class="chapter" data-level="3.8" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#sharing-your-reproducible-r-notebook"><i class="fa fa-check"></i><b>3.8</b> Sharing your reproducible R Notebook</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-structures.html"><a href="data-structures.html"><i class="fa fa-check"></i><b>4</b> Data structures</a><ul>
<li class="chapter" data-level="4.1" data-path="data-structures.html"><a href="data-structures.html#vectors"><i class="fa fa-check"></i><b>4.1</b> Vectors</a><ul>
<li class="chapter" data-level="4.1.1" data-path="data-structures.html"><a href="data-structures.html#vector-arithmetic"><i class="fa fa-check"></i><b>4.1.1</b> Vector Arithmetic</a></li>
<li class="chapter" data-level="4.1.2" data-path="data-structures.html"><a href="data-structures.html#vector-recycling"><i class="fa fa-check"></i><b>4.1.2</b> Vector recycling</a></li>
<li class="chapter" data-level="4.1.3" data-path="data-structures.html"><a href="data-structures.html#simple-statistical-functions-for-numeric-vectors"><i class="fa fa-check"></i><b>4.1.3</b> Simple statistical functions for numeric vectors</a></li>
<li class="chapter" data-level="4.1.4" data-path="data-structures.html"><a href="data-structures.html#indexing-vectors"><i class="fa fa-check"></i><b>4.1.4</b> Indexing Vectors</a></li>
<li class="chapter" data-level="4.1.5" data-path="data-structures.html"><a href="data-structures.html#comparison-operators-applied-to-vectors"><i class="fa fa-check"></i><b>4.1.5</b> Comparison operators applied to vectors</a></li>
<li class="chapter" data-level="4.1.6" data-path="data-structures.html"><a href="data-structures.html#combining-indexing-and-comparison-of-vectors"><i class="fa fa-check"></i><b>4.1.6</b> Combining Indexing and Comparison of Vectors</a></li>
<li class="chapter" data-level="4.1.7" data-path="data-structures.html"><a href="data-structures.html#vector-manipulation"><i class="fa fa-check"></i><b>4.1.7</b> Vector manipulation</a></li>
<li class="chapter" data-level="4.1.8" data-path="data-structures.html"><a href="data-structures.html#vectors-from-regular-sequences"><i class="fa fa-check"></i><b>4.1.8</b> Vectors from regular sequences</a></li>
<li class="chapter" data-level="4.1.9" data-path="data-structures.html"><a href="data-structures.html#additional-functions-for-working-with-vectors"><i class="fa fa-check"></i><b>4.1.9</b> Additional functions for working with vectors</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="data-structures.html"><a href="data-structures.html#lists"><i class="fa fa-check"></i><b>4.2</b> Lists</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-structures.html"><a href="data-structures.html#length-and-type-of-lists"><i class="fa fa-check"></i><b>4.2.1</b> Length and type of lists</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-structures.html"><a href="data-structures.html#indexing-lists"><i class="fa fa-check"></i><b>4.2.2</b> Indexing lists</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-structures.html"><a href="data-structures.html#naming-list-elements"><i class="fa fa-check"></i><b>4.2.3</b> Naming list elements</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-structures.html"><a href="data-structures.html#the-operator"><i class="fa fa-check"></i><b>4.2.4</b> The <code>$</code> operator</a></li>
<li class="chapter" data-level="4.2.5" data-path="data-structures.html"><a href="data-structures.html#changing-and-adding-lists-items"><i class="fa fa-check"></i><b>4.2.5</b> Changing and adding lists items</a></li>
<li class="chapter" data-level="4.2.6" data-path="data-structures.html"><a href="data-structures.html#combining-lists"><i class="fa fa-check"></i><b>4.2.6</b> Combining lists</a></li>
<li class="chapter" data-level="4.2.7" data-path="data-structures.html"><a href="data-structures.html#converting-lists-to-vectors"><i class="fa fa-check"></i><b>4.2.7</b> Converting lists to vectors</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-structures.html"><a href="data-structures.html#data-frames"><i class="fa fa-check"></i><b>4.3</b> Data frames</a><ul>
<li class="chapter" data-level="4.3.1" data-path="data-structures.html"><a href="data-structures.html#creating-a-data-frame"><i class="fa fa-check"></i><b>4.3.1</b> Creating a data frame</a></li>
<li class="chapter" data-level="4.3.2" data-path="data-structures.html"><a href="data-structures.html#type-and-class-for-data-frames"><i class="fa fa-check"></i><b>4.3.2</b> Type and class for data frames</a></li>
<li class="chapter" data-level="4.3.3" data-path="data-structures.html"><a href="data-structures.html#length-and-dimension-for-data-frames"><i class="fa fa-check"></i><b>4.3.3</b> Length and dimension for data frames</a></li>
<li class="chapter" data-level="4.3.4" data-path="data-structures.html"><a href="data-structures.html#indexing-and-accessing-data-frames"><i class="fa fa-check"></i><b>4.3.4</b> Indexing and accessing data frames</a></li>
<li class="chapter" data-level="4.3.5" data-path="data-structures.html"><a href="data-structures.html#logical-indexing-of-data-frames"><i class="fa fa-check"></i><b>4.3.5</b> Logical indexing of data frames</a></li>
<li class="chapter" data-level="4.3.6" data-path="data-structures.html"><a href="data-structures.html#adding-columns-to-a-data-frame"><i class="fa fa-check"></i><b>4.3.6</b> Adding columns to a data frame</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html"><i class="fa fa-check"></i><b>5</b> Functions and control flow statements</a><ul>
<li class="chapter" data-level="5.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#writing-your-own-functions"><i class="fa fa-check"></i><b>5.1</b> Writing your own functions</a><ul>
<li class="chapter" data-level="5.1.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#function-arguments"><i class="fa fa-check"></i><b>5.1.1</b> Function arguments</a></li>
<li class="chapter" data-level="5.1.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#writing-functions-with-optional-arguments"><i class="fa fa-check"></i><b>5.1.2</b> Writing functions with optional arguments</a></li>
<li class="chapter" data-level="5.1.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#putting-r-functions-in-scripts"><i class="fa fa-check"></i><b>5.1.3</b> Putting R functions in Scripts</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#control-flow-statements"><i class="fa fa-check"></i><b>5.2</b> Control flow statements</a><ul>
<li class="chapter" data-level="5.2.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#if-and-if-else-statements"><i class="fa fa-check"></i><b>5.2.1</b> <code>if</code> and <code>if-else</code> statements</a></li>
<li class="chapter" data-level="5.2.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#for-loops"><i class="fa fa-check"></i><b>5.2.2</b> for loops</a></li>
<li class="chapter" data-level="5.2.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#break-statement"><i class="fa fa-check"></i><b>5.2.3</b> <code>break</code> statement</a></li>
<li class="chapter" data-level="5.2.4" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#repeat-loops"><i class="fa fa-check"></i><b>5.2.4</b> <code>repeat</code> loops</a></li>
<li class="chapter" data-level="5.2.5" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#next-statement"><i class="fa fa-check"></i><b>5.2.5</b> <code>next</code> statement</a></li>
<li class="chapter" data-level="5.2.6" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#while-statements"><i class="fa fa-check"></i><b>5.2.6</b> while statements</a></li>
<li class="chapter" data-level="5.2.7" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#ifelse"><i class="fa fa-check"></i><b>5.2.7</b> <code>ifelse</code></a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#map-and-related-tools"><i class="fa fa-check"></i><b>5.3</b> <code>map</code> and related tools</a><ul>
<li class="chapter" data-level="5.3.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#basic-map"><i class="fa fa-check"></i><b>5.3.1</b> basic <code>map</code></a></li>
<li class="chapter" data-level="5.3.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#map_if-and-map_at"><i class="fa fa-check"></i><b>5.3.2</b> <code>map_if</code> and <code>map_at</code></a></li>
<li class="chapter" data-level="5.3.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#mapping-in-parallel-using-map2"><i class="fa fa-check"></i><b>5.3.3</b> mapping in parallel using <code>map2</code></a></li>
<li class="chapter" data-level="5.3.4" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#map-variants-that-return-vectors"><i class="fa fa-check"></i><b>5.3.4</b> <code>map</code> variants that return vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html"><i class="fa fa-check"></i><b>6</b> Introduction to ggplot2</a><ul>
<li class="chapter" data-level="6.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#loading-ggplot2"><i class="fa fa-check"></i><b>6.1</b> Loading ggplot2</a></li>
<li class="chapter" data-level="6.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#example-data-set-andersons-iris-data"><i class="fa fa-check"></i><b>6.2</b> Example data set: Anderson’s Iris Data</a></li>
<li class="chapter" data-level="6.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#template-for-single-layer-plots-in-ggplot2"><i class="fa fa-check"></i><b>6.3</b> Template for single layer plots in ggplot2</a></li>
<li class="chapter" data-level="6.4" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#an-aside-about-function-arguments"><i class="fa fa-check"></i><b>6.4</b> An aside about function arguments</a></li>
<li class="chapter" data-level="6.5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#strip-plots"><i class="fa fa-check"></i><b>6.5</b> Strip plots</a><ul>
<li class="chapter" data-level="6.5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#jittering-data"><i class="fa fa-check"></i><b>6.5.1</b> Jittering data</a></li>
<li class="chapter" data-level="6.5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-categorical-information"><i class="fa fa-check"></i><b>6.5.2</b> Adding categorical information</a></li>
<li class="chapter" data-level="6.5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#rotating-plot-coordinates"><i class="fa fa-check"></i><b>6.5.3</b> Rotating plot coordinates</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#histograms"><i class="fa fa-check"></i><b>6.6</b> Histograms</a><ul>
<li class="chapter" data-level="6.6.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#variations-on-histograms-when-considering-categorical-data"><i class="fa fa-check"></i><b>6.6.1</b> Variations on histograms when considering categorical data</a></li>
</ul></li>
<li class="chapter" data-level="6.7" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#faceting-to-depict-categorical-information"><i class="fa fa-check"></i><b>6.7</b> Faceting to depict categorical information</a></li>
<li class="chapter" data-level="6.8" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots"><i class="fa fa-check"></i><b>6.8</b> Density plots</a></li>
<li class="chapter" data-level="6.9" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-or-beanplot"><i class="fa fa-check"></i><b>6.9</b> Violin or Beanplot</a></li>
<li class="chapter" data-level="6.10" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplots"><i class="fa fa-check"></i><b>6.10</b> Boxplots</a></li>
<li class="chapter" data-level="6.11" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#building-complex-visualizations-with-layers"><i class="fa fa-check"></i><b>6.11</b> Building complex visualizations with layers</a></li>
<li class="chapter" data-level="6.12" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#useful-combination-plots"><i class="fa fa-check"></i><b>6.12</b> Useful combination plots</a><ul>
<li class="chapter" data-level="6.12.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplot-plus-strip-plot"><i class="fa fa-check"></i><b>6.12.1</b> Boxplot plus strip plot</a></li>
<li class="chapter" data-level="6.12.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#setting-shared-aesthetics"><i class="fa fa-check"></i><b>6.12.2</b> Setting shared aesthetics</a></li>
</ul></li>
<li class="chapter" data-level="6.13" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot-layers-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>6.13</b> ggplot layers can be assigned to variables</a><ul>
<li class="chapter" data-level="6.13.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-plot-plus-strip-plot"><i class="fa fa-check"></i><b>6.13.1</b> Violin plot plus strip plot</a></li>
</ul></li>
<li class="chapter" data-level="6.14" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-titles-and-tweaking-axis-labels"><i class="fa fa-check"></i><b>6.14</b> Adding titles and tweaking axis labels</a></li>
<li class="chapter" data-level="6.15" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot2-themes"><i class="fa fa-check"></i><b>6.15</b> ggplot2 themes</a><ul>
<li class="chapter" data-level="6.15.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#further-customization-with-ggplot2theme"><i class="fa fa-check"></i><b>6.15.1</b> Further customization with <code>ggplot2::theme</code></a></li>
</ul></li>
<li class="chapter" data-level="6.16" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#other-aspects-of-ggplots-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>6.16</b> Other aspects of ggplots can be assigned to variables</a></li>
<li class="chapter" data-level="6.17" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-plots"><i class="fa fa-check"></i><b>6.17</b> Bivariate plots</a><ul>
<li class="chapter" data-level="6.17.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#scatter-plots"><i class="fa fa-check"></i><b>6.17.1</b> Scatter plots</a></li>
<li class="chapter" data-level="6.17.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-a-trend-line-to-a-scatter-plot"><i class="fa fa-check"></i><b>6.17.2</b> Adding a trend line to a scatter plot</a></li>
</ul></li>
<li class="chapter" data-level="6.18" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-density-plots"><i class="fa fa-check"></i><b>6.18</b> Bivariate density plots</a></li>
<li class="chapter" data-level="6.19" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#combining-scatter-plots-and-density-plots-with-categorical-information"><i class="fa fa-check"></i><b>6.19</b> Combining Scatter Plots and Density Plots with Categorical Information</a></li>
<li class="chapter" data-level="6.20" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots-with-fill"><i class="fa fa-check"></i><b>6.20</b> Density plots with fill</a></li>
<li class="chapter" data-level="6.21" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#d-bin-and-hex-plots"><i class="fa fa-check"></i><b>6.21</b> 2D bin and hex plots</a></li>
<li class="chapter" data-level="6.22" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#the-cowplot-package"><i class="fa fa-check"></i><b>6.22</b> The <code>cowplot</code> package</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html"><i class="fa fa-check"></i><b>7</b> Introduction to <code>dplyr</code></a><ul>
<li class="chapter" data-level="7.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#libraries"><i class="fa fa-check"></i><b>7.1</b> Libraries</a></li>
<li class="chapter" data-level="7.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#reading-data-with-the-readr-package"><i class="fa fa-check"></i><b>7.2</b> Reading data with the <code>readr</code> package</a><ul>
<li class="chapter" data-level="7.2.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#reading-excel-files"><i class="fa fa-check"></i><b>7.2.1</b> Reading Excel files</a></li>
<li class="chapter" data-level="7.2.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#example-data-nc-births"><i class="fa fa-check"></i><b>7.2.2</b> Example data: NC Births</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#a-note-on-tibbles"><i class="fa fa-check"></i><b>7.3</b> A note on “tibbles”</a></li>
<li class="chapter" data-level="7.4" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#data-filtering-and-transformation-with-dplyr"><i class="fa fa-check"></i><b>7.4</b> Data filtering and transformation with <code>dplyr</code></a></li>
<li class="chapter" data-level="7.5" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#dplyrs-verbs"><i class="fa fa-check"></i><b>7.5</b> dplyr’s “verbs”</a><ul>
<li class="chapter" data-level="7.5.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#select"><i class="fa fa-check"></i><b>7.5.1</b> <code>select</code></a></li>
<li class="chapter" data-level="7.5.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#filter"><i class="fa fa-check"></i><b>7.5.2</b> <code>filter</code></a></li>
<li class="chapter" data-level="7.5.3" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#mutate"><i class="fa fa-check"></i><b>7.5.3</b> <code>mutate</code></a></li>
<li class="chapter" data-level="7.5.4" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#arrange"><i class="fa fa-check"></i><b>7.5.4</b> <code>arrange</code></a></li>
<li class="chapter" data-level="7.5.5" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#summarize"><i class="fa fa-check"></i><b>7.5.5</b> <code>summarize</code></a></li>
<li class="chapter" data-level="7.5.6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#group_by"><i class="fa fa-check"></i><b>7.5.6</b> <code>group_by</code></a></li>
<li class="chapter" data-level="7.5.7" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#combining-grouping-and-summarizing"><i class="fa fa-check"></i><b>7.5.7</b> Combining grouping and summarizing</a></li>
<li class="chapter" data-level="7.5.8" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#scoped-variants-of-mutate-and-summarize"><i class="fa fa-check"></i><b>7.5.8</b> Scoped variants of <code>mutate</code> and <code>summarize</code></a></li>
<li class="chapter" data-level="7.5.9" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#combining-summarize-with-grouping-aesthetics-in-ggplot2"><i class="fa fa-check"></i><b>7.5.9</b> Combining summarize with grouping aesthetics in <code>ggplot2</code></a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#pipes"><i class="fa fa-check"></i><b>7.6</b> Pipes</a><ul>
<li class="chapter" data-level="7.6.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#install-and-load-magrittr"><i class="fa fa-check"></i><b>7.6.1</b> Install and load <code>magrittr</code></a></li>
<li class="chapter" data-level="7.6.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-basic-pipe-operator"><i class="fa fa-check"></i><b>7.6.2</b> The basic pipe operator</a></li>
<li class="chapter" data-level="7.6.3" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#an-example-without-pipes"><i class="fa fa-check"></i><b>7.6.3</b> An example without pipes</a></li>
<li class="chapter" data-level="7.6.4" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-same-example-using-pipes"><i class="fa fa-check"></i><b>7.6.4</b> The same example using pipes</a></li>
<li class="chapter" data-level="7.6.5" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#assigning-the-output-of-a-statement-involving-pipes-to-a-variable"><i class="fa fa-check"></i><b>7.6.5</b> Assigning the output of a statement involving pipes to a variable</a></li>
<li class="chapter" data-level="7.6.6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#compound-assignment-pipe-operator"><i class="fa fa-check"></i><b>7.6.6</b> Compound assignment pipe operator</a></li>
<li class="chapter" data-level="7.6.7" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-dot-operator-with-pipes"><i class="fa fa-check"></i><b>7.6.7</b> The dot operator with pipes</a></li>
<li class="chapter" data-level="7.6.8" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-exposition-pipe-operator"><i class="fa fa-check"></i><b>7.6.8</b> The exposition pipe operator</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>8</b> Data wrangling</a><ul>
<li class="chapter" data-level="8.1" data-path="data-wrangling.html"><a href="data-wrangling.html#libraries-1"><i class="fa fa-check"></i><b>8.1</b> Libraries</a></li>
<li class="chapter" data-level="8.2" data-path="data-wrangling.html"><a href="data-wrangling.html#data"><i class="fa fa-check"></i><b>8.2</b> Data</a></li>
<li class="chapter" data-level="8.3" data-path="data-wrangling.html"><a href="data-wrangling.html#renaming-data-frame-columms"><i class="fa fa-check"></i><b>8.3</b> Renaming data frame columms</a></li>
<li class="chapter" data-level="8.4" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-unneeded-columns"><i class="fa fa-check"></i><b>8.4</b> Dropping unneeded columns</a><ul>
<li class="chapter" data-level="8.4.1" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-all-empty-columns"><i class="fa fa-check"></i><b>8.4.1</b> Finding all empty columns</a></li>
<li class="chapter" data-level="8.4.2" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-columns-by-matching-names"><i class="fa fa-check"></i><b>8.4.2</b> Dropping columns by matching names</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="data-wrangling.html"><a href="data-wrangling.html#merging-data-frames"><i class="fa fa-check"></i><b>8.5</b> Merging data frames</a></li>
<li class="chapter" data-level="8.6" data-path="data-wrangling.html"><a href="data-wrangling.html#reshaping-data-with-tidyr"><i class="fa fa-check"></i><b>8.6</b> Reshaping data with tidyr</a><ul>
<li class="chapter" data-level="8.6.1" data-path="data-wrangling.html"><a href="data-wrangling.html#wide-to-long-conversions-using-tidyrgather"><i class="fa fa-check"></i><b>8.6.1</b> Wide to long conversions using <code>tidyr::gather</code></a></li>
<li class="chapter" data-level="8.6.2" data-path="data-wrangling.html"><a href="data-wrangling.html#extracting-information-from-combined-variables-using-tidyrextract"><i class="fa fa-check"></i><b>8.6.2</b> Extracting information from combined variables using <code>tidyr::extract</code></a></li>
<li class="chapter" data-level="8.6.3" data-path="data-wrangling.html"><a href="data-wrangling.html#subsetting-rows"><i class="fa fa-check"></i><b>8.6.3</b> Subsetting rows</a></li>
<li class="chapter" data-level="8.6.4" data-path="data-wrangling.html"><a href="data-wrangling.html#splitting-columns"><i class="fa fa-check"></i><b>8.6.4</b> Splitting columns</a></li>
<li class="chapter" data-level="8.6.5" data-path="data-wrangling.html"><a href="data-wrangling.html#combining-data-frames"><i class="fa fa-check"></i><b>8.6.5</b> Combining data frames</a></li>
<li class="chapter" data-level="8.6.6" data-path="data-wrangling.html"><a href="data-wrangling.html#sorting-data-frame-rows"><i class="fa fa-check"></i><b>8.6.6</b> Sorting data frame rows</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="data-wrangling.html"><a href="data-wrangling.html#using-your-tidy-data"><i class="fa fa-check"></i><b>8.7</b> Using your tidy data</a><ul>
<li class="chapter" data-level="8.7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#visualizing-gene-expression-time-series"><i class="fa fa-check"></i><b>8.7.1</b> Visualizing gene expression time series</a></li>
<li class="chapter" data-level="8.7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-the-most-variable-genes"><i class="fa fa-check"></i><b>8.7.2</b> Finding the most variable genes</a></li>
<li class="chapter" data-level="8.7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#heat-maps"><i class="fa fa-check"></i><b>8.7.3</b> Heat maps</a></li>
</ul></li>
<li class="chapter" data-level="8.8" data-path="data-wrangling.html"><a href="data-wrangling.html#long-to-wide-conversion-using-tidyrspread"><i class="fa fa-check"></i><b>8.8</b> Long-to-wide conversion using <code>tidyr::spread</code></a></li>
<li class="chapter" data-level="8.9" data-path="data-wrangling.html"><a href="data-wrangling.html#exploring-bivariate-relationships-using-wide-data"><i class="fa fa-check"></i><b>8.9</b> Exploring bivariate relationships using “wide” data</a><ul>
<li class="chapter" data-level="8.9.1" data-path="data-wrangling.html"><a href="data-wrangling.html#large-scale-patterns-of-correlations"><i class="fa fa-check"></i><b>8.9.1</b> Large scale patterns of correlations</a></li>
<li class="chapter" data-level="8.9.2" data-path="data-wrangling.html"><a href="data-wrangling.html#adding-new-columns-and-combining-filtered-data-frames"><i class="fa fa-check"></i><b>8.9.2</b> Adding new columns and combining filtered data frames</a></li>
<li class="chapter" data-level="8.9.3" data-path="data-wrangling.html"><a href="data-wrangling.html#a-heat-mapped-sorted-by-correlations"><i class="fa fa-check"></i><b>8.9.3</b> A heat mapped sorted by correlations</a></li>
<li class="chapter" data-level="8.9.4" data-path="data-wrangling.html"><a href="data-wrangling.html#a-fancy-figure"><i class="fa fa-check"></i><b>8.9.4</b> A “fancy” figure</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="vector-algebra.html"><a href="vector-algebra.html"><i class="fa fa-check"></i><b>9</b> Vector algebra</a><ul>
<li class="chapter" data-level="9.1" data-path="vector-algebra.html"><a href="vector-algebra.html#libraries-2"><i class="fa fa-check"></i><b>9.1</b> Libraries</a></li>
<li class="chapter" data-level="9.2" data-path="vector-algebra.html"><a href="vector-algebra.html#vector-mathematics-in-r"><i class="fa fa-check"></i><b>9.2</b> Vector Mathematics in R</a></li>
<li class="chapter" data-level="9.3" data-path="vector-algebra.html"><a href="vector-algebra.html#simple-statistics-in-vector-form"><i class="fa fa-check"></i><b>9.3</b> Simple statistics in vector form</a><ul>
<li class="chapter" data-level="9.3.1" data-path="vector-algebra.html"><a href="vector-algebra.html#mean"><i class="fa fa-check"></i><b>9.3.1</b> Mean</a></li>
<li class="chapter" data-level="9.3.2" data-path="vector-algebra.html"><a href="vector-algebra.html#mean-centering"><i class="fa fa-check"></i><b>9.3.2</b> Mean centering</a></li>
<li class="chapter" data-level="9.3.3" data-path="vector-algebra.html"><a href="vector-algebra.html#variance-and-standard-deviation"><i class="fa fa-check"></i><b>9.3.3</b> Variance and standard deviation</a></li>
<li class="chapter" data-level="9.3.4" data-path="vector-algebra.html"><a href="vector-algebra.html#covariance-and-correlation"><i class="fa fa-check"></i><b>9.3.4</b> Covariance and correlation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="matrices-in-r.html"><a href="matrices-in-r.html"><i class="fa fa-check"></i><b>10</b> Matrices in R</a><ul>
<li class="chapter" data-level="10.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#creating-matrices-in-r"><i class="fa fa-check"></i><b>10.1</b> Creating matrices in R</a></li>
<li class="chapter" data-level="10.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-arithmetic-operations-in-r"><i class="fa fa-check"></i><b>10.2</b> Matrix arithmetic operations in R</a></li>
<li class="chapter" data-level="10.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#descriptive-statistics-as-matrix-functions"><i class="fa fa-check"></i><b>10.3</b> Descriptive statistics as matrix functions</a><ul>
<li class="chapter" data-level="10.3.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#mean-vector-and-matrix"><i class="fa fa-check"></i><b>10.3.1</b> Mean vector and matrix</a></li>
<li class="chapter" data-level="10.3.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#deviation-matrix"><i class="fa fa-check"></i><b>10.3.2</b> Deviation matrix</a></li>
<li class="chapter" data-level="10.3.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#covariance-matrix"><i class="fa fa-check"></i><b>10.3.3</b> Covariance matrix</a></li>
<li class="chapter" data-level="10.3.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#correlation-matrix"><i class="fa fa-check"></i><b>10.3.4</b> Correlation matrix</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-inverse"><i class="fa fa-check"></i><b>10.4</b> Matrix Inverse</a></li>
<li class="chapter" data-level="10.5" data-path="matrices-in-r.html"><a href="matrices-in-r.html#solving-sets-of-simultaneous-equations"><i class="fa fa-check"></i><b>10.5</b> Solving sets of simultaneous equations</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html"><i class="fa fa-check"></i><b>11</b> Linear Regression Models I</a><ul>
<li class="chapter" data-level="11.1" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#linear-functions"><i class="fa fa-check"></i><b>11.1</b> Linear functions</a></li>
<li class="chapter" data-level="11.2" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#linear-regression"><i class="fa fa-check"></i><b>11.2</b> Linear regression</a></li>
<li class="chapter" data-level="11.3" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#the-optimality-criterion-for-least-squares-regression"><i class="fa fa-check"></i><b>11.3</b> The optimality criterion for least-squares regression</a></li>
<li class="chapter" data-level="11.4" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#geometry-of-linear-regression"><i class="fa fa-check"></i><b>11.4</b> Geometry of linear regression</a></li>
<li class="chapter" data-level="11.5" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#solution-for-the-least-squares-criterion"><i class="fa fa-check"></i><b>11.5</b> Solution for the least-squares criterion</a><ul>
<li class="chapter" data-level="11.5.1" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#bivariate-regression-estimating-coefficients"><i class="fa fa-check"></i><b>11.5.1</b> Bivariate regression, estimating coefficients</a></li>
<li class="chapter" data-level="11.5.2" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#multiple-regression-estimating-coefficients"><i class="fa fa-check"></i><b>11.5.2</b> Multiple regression, estimating coefficients</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#residuals"><i class="fa fa-check"></i><b>11.6</b> Residuals</a></li>
<li class="chapter" data-level="11.7" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#regression-as-sum-of-squares-decomposition"><i class="fa fa-check"></i><b>11.7</b> Regression as sum-of-squares decomposition</a></li>
<li class="chapter" data-level="11.8" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#variance-explained-by-a-regression-model"><i class="fa fa-check"></i><b>11.8</b> Variance “explained” by a regression model</a></li>
<li class="chapter" data-level="11.9" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#goodness-of-fit"><i class="fa fa-check"></i><b>11.9</b> Goodness of fit</a></li>
<li class="chapter" data-level="11.10" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#interpretting-regression"><i class="fa fa-check"></i><b>11.10</b> Interpretting Regression</a></li>
<li class="chapter" data-level="11.11" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#illustrating-linear-regression-with-simulated-data"><i class="fa fa-check"></i><b>11.11</b> Illustrating linear regression with simulated data</a></li>
<li class="chapter" data-level="11.12" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#calculating-the-regression-model-using-vector-arithmetic"><i class="fa fa-check"></i><b>11.12</b> Calculating the regression model using vector arithmetic</a><ul>
<li class="chapter" data-level="11.12.1" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#calculating-the-predicted-values-and-residuals"><i class="fa fa-check"></i><b>11.12.1</b> Calculating the predicted values and residuals</a></li>
</ul></li>
<li class="chapter" data-level="11.13" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#plotting-various-aspects-of-a-regression"><i class="fa fa-check"></i><b>11.13</b> Plotting various aspects of a regression</a><ul>
<li class="chapter" data-level="11.13.1" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#residual-plots"><i class="fa fa-check"></i><b>11.13.1</b> Residual plots</a></li>
</ul></li>
<li class="chapter" data-level="11.14" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#calculating-the-coefficient-of-determination"><i class="fa fa-check"></i><b>11.14</b> Calculating the coefficient of determination</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html"><i class="fa fa-check"></i><b>12</b> Linear Regression Models II</a><ul>
<li class="chapter" data-level="12.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#new-libraries-to-install"><i class="fa fa-check"></i><b>12.1</b> New Libraries to install</a></li>
<li class="chapter" data-level="12.2" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#standard-libraries"><i class="fa fa-check"></i><b>12.2</b> Standard libraries</a></li>
<li class="chapter" data-level="12.3" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#specifying-regression-models-in-r"><i class="fa fa-check"></i><b>12.3</b> Specifying Regression Models in R</a><ul>
<li class="chapter" data-level="12.3.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#interpretting-summary-output-from-lm"><i class="fa fa-check"></i><b>12.3.1</b> Interpretting summary output from lm()</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#quick-bivariate-regression-plots-in-ggplot"><i class="fa fa-check"></i><b>12.4</b> Quick bivariate regression plots in ggplot</a></li>
<li class="chapter" data-level="12.5" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#more-about-the-data-structure-returned-by-lm"><i class="fa fa-check"></i><b>12.5</b> More about the data structure returned by <code>lm()</code></a><ul>
<li class="chapter" data-level="12.5.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#fitted-values"><i class="fa fa-check"></i><b>12.5.1</b> Fitted values</a></li>
<li class="chapter" data-level="12.5.2" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#getting-the-model-coefficients"><i class="fa fa-check"></i><b>12.5.2</b> Getting the model coefficients</a></li>
</ul></li>
<li class="chapter" data-level="12.6" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#broom-a-library-for-converting-model-results-into-data-frames"><i class="fa fa-check"></i><b>12.6</b> Broom: a library for converting model results into data frames</a><ul>
<li class="chapter" data-level="12.6.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#broomtidy"><i class="fa fa-check"></i><b>12.6.1</b> <code>broom::tidy</code></a></li>
<li class="chapter" data-level="12.6.2" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#broomaugment"><i class="fa fa-check"></i><b>12.6.2</b> <code>broom::augment</code></a></li>
<li class="chapter" data-level="12.6.3" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#broomglance"><i class="fa fa-check"></i><b>12.6.3</b> <code>broom::glance</code></a></li>
</ul></li>
<li class="chapter" data-level="12.7" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#qq-plots"><i class="fa fa-check"></i><b>12.7</b> qq-plots</a></li>
<li class="chapter" data-level="12.8" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#multiple-regression"><i class="fa fa-check"></i><b>12.8</b> Multiple regression</a><ul>
<li class="chapter" data-level="12.8.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#exploration-of-the-trees-data-set"><i class="fa fa-check"></i><b>12.8.1</b> Exploration of the trees data set</a></li>
</ul></li>
<li class="chapter" data-level="12.9" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#d-plots"><i class="fa fa-check"></i><b>12.9</b> 3D Plots</a><ul>
<li class="chapter" data-level="12.9.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#scatterplot3d"><i class="fa fa-check"></i><b>12.9.1</b> scatterplot3d</a></li>
<li class="chapter" data-level="12.9.2" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#rgl"><i class="fa fa-check"></i><b>12.9.2</b> rgl</a></li>
</ul></li>
<li class="chapter" data-level="12.10" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#fitting-the-regression-model"><i class="fa fa-check"></i><b>12.10</b> Fitting the regression model</a><ul>
<li class="chapter" data-level="12.10.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#visualizing-the-regression-model-in-scatterplot3d"><i class="fa fa-check"></i><b>12.10.1</b> Visualizing the regression model in scatterplot3d</a></li>
</ul></li>
<li class="chapter" data-level="12.11" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#interpretting-the-regression-model"><i class="fa fa-check"></i><b>12.11</b> Interpretting the regression model</a></li>
<li class="chapter" data-level="12.12" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#exploring-the-vector-geometry-of-the-regression-model"><i class="fa fa-check"></i><b>12.12</b> Exploring the Vector Geometry of the Regression Model</a></li>
<li class="chapter" data-level="12.13" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#exploring-the-residuals-from-the-model-fit"><i class="fa fa-check"></i><b>12.13</b> Exploring the Residuals from the Model Fit</a></li>
<li class="chapter" data-level="12.14" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#an-alternate-model"><i class="fa fa-check"></i><b>12.14</b> An alternate model</a></li>
<li class="chapter" data-level="12.15" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#exploring-the-impact-of-nearly-collinear-predictors-on-regression"><i class="fa fa-check"></i><b>12.15</b> Exploring the impact of nearly collinear predictors on regression</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="more-regression-models.html"><a href="more-regression-models.html"><i class="fa fa-check"></i><b>13</b> More regression models</a><ul>
<li class="chapter" data-level="13.1" data-path="more-regression-models.html"><a href="more-regression-models.html#curvilinear-regression"><i class="fa fa-check"></i><b>13.1</b> Curvilinear regression</a></li>
<li class="chapter" data-level="13.2" data-path="more-regression-models.html"><a href="more-regression-models.html#loess-regression"><i class="fa fa-check"></i><b>13.2</b> LOESS regression</a></li>
<li class="chapter" data-level="13.3" data-path="more-regression-models.html"><a href="more-regression-models.html#logistic-regression"><i class="fa fa-check"></i><b>13.3</b> Logistic regression</a><ul>
<li class="chapter" data-level="13.3.1" data-path="more-regression-models.html"><a href="more-regression-models.html#a-web-app-to-explore-the-logistic-regression-equation"><i class="fa fa-check"></i><b>13.3.1</b> A web app to explore the logistic regression equation</a></li>
<li class="chapter" data-level="13.3.2" data-path="more-regression-models.html"><a href="more-regression-models.html#titanic-data-set"><i class="fa fa-check"></i><b>13.3.2</b> Titanic data set</a></li>
<li class="chapter" data-level="13.3.3" data-path="more-regression-models.html"><a href="more-regression-models.html#subsetting-the-data"><i class="fa fa-check"></i><b>13.3.3</b> Subsetting the data</a></li>
<li class="chapter" data-level="13.3.4" data-path="more-regression-models.html"><a href="more-regression-models.html#visualizing-survival-as-a-function-of-age"><i class="fa fa-check"></i><b>13.3.4</b> Visualizing survival as a function of age</a></li>
<li class="chapter" data-level="13.3.5" data-path="more-regression-models.html"><a href="more-regression-models.html#fitting-the-logistic-regression-model"><i class="fa fa-check"></i><b>13.3.5</b> Fitting the logistic regression model</a></li>
<li class="chapter" data-level="13.3.6" data-path="more-regression-models.html"><a href="more-regression-models.html#visualizing-the-logistic-regression"><i class="fa fa-check"></i><b>13.3.6</b> Visualizing the logistic regression</a></li>
<li class="chapter" data-level="13.3.7" data-path="more-regression-models.html"><a href="more-regression-models.html#quick-and-easy-visualization"><i class="fa fa-check"></i><b>13.3.7</b> Quick and easy visualization</a></li>
<li class="chapter" data-level="13.3.8" data-path="more-regression-models.html"><a href="more-regression-models.html#impact-of-sex-and-passenger-class-on-the-models"><i class="fa fa-check"></i><b>13.3.8</b> Impact of sex and passenger class on the models</a></li>
<li class="chapter" data-level="13.3.9" data-path="more-regression-models.html"><a href="more-regression-models.html#fitting-multiple-models-based-on-groupings-use-dplyrdo"><i class="fa fa-check"></i><b>13.3.9</b> Fitting multiple models based on groupings use <code>dplyr::do</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="14" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html"><i class="fa fa-check"></i><b>14</b> Principal Components Analysis</a><ul>
<li class="chapter" data-level="14.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#libraries-3"><i class="fa fa-check"></i><b>14.1</b> Libraries</a></li>
<li class="chapter" data-level="14.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#matrices-as-linear-transformations"><i class="fa fa-check"></i><b>14.2</b> Matrices as linear transformations</a></li>
<li class="chapter" data-level="14.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#eigenanalysis-in-r"><i class="fa fa-check"></i><b>14.3</b> Eigenanalysis in R</a></li>
<li class="chapter" data-level="14.4" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#principal-components-analysis-in-r"><i class="fa fa-check"></i><b>14.4</b> Principal Components Analysis in R</a><ul>
<li class="chapter" data-level="14.4.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#bioenv-dataset"><i class="fa fa-check"></i><b>14.4.1</b> Bioenv dataset</a></li>
<li class="chapter" data-level="14.4.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pca-of-the-bioenv-dataset"><i class="fa fa-check"></i><b>14.4.2</b> PCA of the Bioenv dataset</a></li>
<li class="chapter" data-level="14.4.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#calculating-factor-loadings"><i class="fa fa-check"></i><b>14.4.3</b> Calculating Factor Loadings</a></li>
</ul></li>
<li class="chapter" data-level="14.5" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#drawing-figures-to-represent-pca"><i class="fa fa-check"></i><b>14.5</b> Drawing Figures to Represent PCA</a><ul>
<li class="chapter" data-level="14.5.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pc-score-plots"><i class="fa fa-check"></i><b>14.5.1</b> PC Score Plots</a></li>
<li class="chapter" data-level="14.5.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#simultaneous-depiction-of-observations-and-variables-in-the-pc-space"><i class="fa fa-check"></i><b>14.5.2</b> Simultaneous Depiction of Observations and Variables in the PC Space</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="15" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html"><i class="fa fa-check"></i><b>15</b> Canonical Variates Analysis</a><ul>
<li class="chapter" data-level="15.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#libraries-4"><i class="fa fa-check"></i><b>15.1</b> Libraries</a></li>
<li class="chapter" data-level="15.2" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#discriminant-analysis-in-r"><i class="fa fa-check"></i><b>15.2</b> Discriminant Analysis in R</a><ul>
<li class="chapter" data-level="15.2.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#shorthand-formulae-in-r"><i class="fa fa-check"></i><b>15.2.1</b> Shorthand Formulae in R</a></li>
<li class="chapter" data-level="15.2.2" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#working-with-the-output-of-lda"><i class="fa fa-check"></i><b>15.2.2</b> Working with the output of <code>lda()</code></a></li>
</ul></li>
<li class="chapter" data-level="15.3" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#estimating-confidence-regions-for-group-means-in-cva"><i class="fa fa-check"></i><b>15.3</b> Estimating confidence regions for group means in CVA</a><ul>
<li class="chapter" data-level="15.3.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#drawing-the-cva-confidence-regions"><i class="fa fa-check"></i><b>15.3.1</b> Drawing the CVA confidence regions</a></li>
</ul></li>
<li class="chapter" data-level="15.4" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#calculating-the-within-and-between-group-covariance-matrices"><i class="fa fa-check"></i><b>15.4</b> Calculating the Within and Between Group Covariance Matrices</a><ul>
<li class="chapter" data-level="15.4.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#recapitulating-the-cva-analysis-of-lda"><i class="fa fa-check"></i><b>15.4.1</b> Recapitulating the CVA analysis of <code>lda()</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="16" data-path="clustering-in-r.html"><a href="clustering-in-r.html"><i class="fa fa-check"></i><b>16</b> Clustering in R</a><ul>
<li class="chapter" data-level="16.1" data-path="clustering-in-r.html"><a href="clustering-in-r.html#libraries-5"><i class="fa fa-check"></i><b>16.1</b> Libraries</a></li>
<li class="chapter" data-level="16.2" data-path="clustering-in-r.html"><a href="clustering-in-r.html#data-set"><i class="fa fa-check"></i><b>16.2</b> Data set</a></li>
<li class="chapter" data-level="16.3" data-path="clustering-in-r.html"><a href="clustering-in-r.html#hierarchical-clustering-in-r"><i class="fa fa-check"></i><b>16.3</b> Hierarchical Clustering in R</a></li>
<li class="chapter" data-level="16.4" data-path="clustering-in-r.html"><a href="clustering-in-r.html#manipulating-hierarchical-clusterings-with-dendextend"><i class="fa fa-check"></i><b>16.4</b> Manipulating hierarchical clusterings with dendextend</a></li>
<li class="chapter" data-level="16.5" data-path="clustering-in-r.html"><a href="clustering-in-r.html#plotting-dendrograms-in-dendextend"><i class="fa fa-check"></i><b>16.5</b> Plotting dendrograms in dendextend</a></li>
<li class="chapter" data-level="16.6" data-path="clustering-in-r.html"><a href="clustering-in-r.html#cutting-dendrograms"><i class="fa fa-check"></i><b>16.6</b> Cutting dendrograms</a></li>
<li class="chapter" data-level="16.7" data-path="clustering-in-r.html"><a href="clustering-in-r.html#looking-at-clusters"><i class="fa fa-check"></i><b>16.7</b> Looking at clusters</a></li>
<li class="chapter" data-level="16.8" data-path="clustering-in-r.html"><a href="clustering-in-r.html#generating-a-heat-map-from-a-cluster"><i class="fa fa-check"></i><b>16.8</b> Generating a heat map from a cluster</a></li>
<li class="chapter" data-level="16.9" data-path="clustering-in-r.html"><a href="clustering-in-r.html#working-with-sub-trees"><i class="fa fa-check"></i><b>16.9</b> Working with sub-trees</a></li>
<li class="chapter" data-level="16.10" data-path="clustering-in-r.html"><a href="clustering-in-r.html#setting-dendrogram-parameters-in-dendextend"><i class="fa fa-check"></i><b>16.10</b> Setting dendrogram parameters in dendextend</a></li>
<li class="chapter" data-level="16.11" data-path="clustering-in-r.html"><a href="clustering-in-r.html#combining-heatmaps-and-dendrograms"><i class="fa fa-check"></i><b>16.11</b> Combining heatmaps and dendrograms</a></li>
<li class="chapter" data-level="16.12" data-path="clustering-in-r.html"><a href="clustering-in-r.html#k-meansk-medoids-clustering-in-r"><i class="fa fa-check"></i><b>16.12</b> K-means/K-medoids Clustering in R</a><ul>
<li class="chapter" data-level="16.12.1" data-path="clustering-in-r.html"><a href="clustering-in-r.html#heat-map-from-k-medoids-cluster"><i class="fa fa-check"></i><b>16.12.1</b> Heat map from k-medoids cluster</a></li>
</ul></li>
<li class="chapter" data-level="16.13" data-path="clustering-in-r.html"><a href="clustering-in-r.html#combining-clusters-and-correlation-matrix-heatmaps"><i class="fa fa-check"></i><b>16.13</b> Combining clusters and correlation matrix heatmaps</a><ul>
<li class="chapter" data-level="16.13.1" data-path="clustering-in-r.html"><a href="clustering-in-r.html#hierarchical-clustering-visualized-on-correlation-matrix"><i class="fa fa-check"></i><b>16.13.1</b> Hierarchical clustering, visualized on correlation matrix</a></li>
<li class="chapter" data-level="16.13.2" data-path="clustering-in-r.html"><a href="clustering-in-r.html#k-medoids-clustering-visualized-on-the-correlation-matrix"><i class="fa fa-check"></i><b>16.13.2</b> K-medoids clustering visualized on the correlation matrix</a></li>
</ul></li>
<li class="chapter" data-level="16.14" data-path="clustering-in-r.html"><a href="clustering-in-r.html#depicting-the-data-within-clusters"><i class="fa fa-check"></i><b>16.14</b> Depicting the data within clusters</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Biology 723: Statistical Computing for Biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="more-regression-models" class="section level1">
<h1><span class="header-section-number">Chapter 13</span> More regression models</h1>
<div id="curvilinear-regression" class="section level2">
<h2><span class="header-section-number">13.1</span> Curvilinear regression</h2>
<p>In the previous chapter we used the built-in <code>trees</code> data set, which consists of measurements of the girth, height, and volume of 31 black cherry trees, to illustrate the fitting of multiple regression models.</p>
<p>After exploring the data set and examining residuals from the model fit we realized that:</p>
<ol style="list-style-type: lower-alpha">
<li><p>the residual from the linear model suggested a straight line relationship might not be appropriate so we transformed the volume data in order to fit a straight line relationship between <span class="math inline">\(\sqrt[3]{V}\)</span> and Girth and Height.</p></li>
<li><p>A model based on Girth as the only predictor variable is almost as good as one involving both Girth and Height.</p></li>
</ol>
<p>Here we illustrate how to fit a cubic regression to the model <code>Volume ~ Girth</code>. Remember that a cubic regression is still a form of linear regression, as it’s in the coefficients.</p>
<div class="sourceCode" id="cb405"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb405-1" title="1">fit.curvilinear &lt;-<span class="st"> </span><span class="kw">lm</span>(Volume <span class="op">~</span><span class="st"> </span><span class="kw">I</span>(Girth<span class="op">^</span><span class="dv">3</span>), <span class="dt">data=</span>trees)</a>
<a class="sourceLine" id="cb405-2" title="2"></a>
<a class="sourceLine" id="cb405-3" title="3">broom<span class="op">::</span><span class="kw">tidy</span>(fit.curvilinear)</a>
<a class="sourceLine" id="cb405-4" title="4"><span class="co">## # A tibble: 2 x 5</span></a>
<a class="sourceLine" id="cb405-5" title="5"><span class="co">##   term        estimate std.error statistic  p.value</span></a>
<a class="sourceLine" id="cb405-6" title="6"><span class="co">##   &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb405-7" title="7"><span class="co">## 1 (Intercept)  8.04     1.04          7.71 1.66e- 8</span></a>
<a class="sourceLine" id="cb405-8" title="8"><span class="co">## 2 I(Girth^3)   0.00814  0.000312     26.1  1.09e-21</span></a>
<a class="sourceLine" id="cb405-9" title="9"></a>
<a class="sourceLine" id="cb405-10" title="10">broom<span class="op">::</span><span class="kw">glance</span>(fit.curvilinear)</a>
<a class="sourceLine" id="cb405-11" title="11"><span class="co">## # A tibble: 1 x 11</span></a>
<a class="sourceLine" id="cb405-12" title="12"><span class="co">##   r.squared adj.r.squared sigma statistic  p.value    df logLik   AIC   BIC</span></a>
<a class="sourceLine" id="cb405-13" title="13"><span class="co">##       &lt;dbl&gt;         &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;int&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb405-14" title="14"><span class="co">## 1     0.959         0.958  3.38      681. 1.09e-21     2  -80.7  167.  172.</span></a>
<a class="sourceLine" id="cb405-15" title="15"><span class="co">## # … with 2 more variables: deviance &lt;dbl&gt;, df.residual &lt;int&gt;</span></a></code></pre></div>
<p>Here’s how we can visualize the corresponding curvilinear regression using ggplot:</p>
<div class="sourceCode" id="cb406"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb406-1" title="1"><span class="kw">ggplot</span>(trees, <span class="kw">aes</span>(<span class="dt">x =</span> Girth, <span class="dt">y =</span> Volume)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb406-2" title="2"><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb406-3" title="3"><span class="st">  </span><span class="kw">geom_smooth</span>(<span class="dt">method =</span> <span class="st">&quot;lm&quot;</span>, </a>
<a class="sourceLine" id="cb406-4" title="4">              <span class="dt">formula =</span> y <span class="op">~</span><span class="st"> </span><span class="kw">I</span>(x<span class="op">^</span><span class="dv">3</span>),</a>
<a class="sourceLine" id="cb406-5" title="5">              <span class="dt">se =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-403-1.png" width="576" /></p>
<p>The <code>I()</code> function used above requires a little explanation. Normally, the R formula syntax (see <code>?formula</code>) treats the carat symbol, <code>^</code>, as short-hand for factor crossing to the specified degree. For example, the formula <code>(a+b+c)^2</code> would be interpretted as the model with main effects and all second order interaction terms, i.e. <code>a + b + c + a:b + a:c + b:c</code> where the colons indicate interactions. The <code>I()</code> function `protects’ the object in it’s argument; in this case telling the regression function to treat this as Girth raised to the third power as opposed to trying to construct interaction terms for Girth.</p>
</div>
<div id="loess-regression" class="section level2">
<h2><span class="header-section-number">13.2</span> LOESS regression</h2>
<p>LOESS (aka LOWESS; ‘Locally weighted scatterplot smoothing’) is a
modeling technique that fits a curve (or surface) to a set of data using
a large number of local linear regressions. Local weighted regressions are fit
at numerous regions across the data range, using a weighting function
that drops off as you move away from the center of the fitting region
(hence the "local aspect). LOESS combines the simplicity of least
squares fitting with the flexibility of non-linear techniques and
doesn’t require the user to specify a functional form ahead of time in
order to fit the model. It does however require relatively dense
sampling in order to produce robust fits.</p>
<p>Formally, at each point <span class="math inline">\(x_i\)</span> we estimate the regression coefficients
<span class="math inline">\(\hat{\beta}_j(x)\)</span> as the values that minimize:
<span class="math display">\[
\sum_{k=1}^n w_k(x_i)(y_k - \beta_0 - \beta_1 x_k - \ldots - \beta_d x_k^2)^2
\]</span>
where <span class="math inline">\(d\)</span> is the degree of the polynomial (usually 1 or 2) and <span class="math inline">\(w_k\)</span>
is a weight function. The most common choice of weighting function is
called the “tri-cube” function as is defined as:</p>
<p><span class="math display">\[\begin{align*}
w_k(x_i) &amp;= (1-|x_i|^3)^3, \mbox{for}\ |x_i| \lt 1 \\
     &amp;= 0, \mbox{for}\ |x_i| \geq 1
\end{align*}\]</span>
where <span class="math inline">\(|x_i|\)</span> is the normalized distance (as determined by the span parameter of the LOESS model) of the observation <span class="math inline">\(x_i\)</span> from the focal observation <span class="math inline">\(x_k\)</span>.</p>
<p>The primary parameter that a user must decide on when using LOESS is the
size of the neighborhood function to apply (i.e. over what distance
should the weight function drop to zero). This is referred to as the
“span” in the R documentation, or as the parameter <span class="math inline">\(\alpha\)</span> in many of
the papers that discuss LOESS. The appropriate span can be determined by
experimentation or, more rigorously by cross-validation.</p>
<p>We’ll illustrate fitting a Loess model using data on Barack Obama’s approval ratings over the period from 2008 to 2001 (<a href="https://github.com/Bio723-class/example-datasets/raw/master/obama-polls-2008-2011.txt"><code>obama-polls.txt</code></a>).</p>
<div class="sourceCode" id="cb407"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb407-1" title="1">polls &lt;-<span class="st"> </span><span class="kw">read_delim</span>(<span class="st">&#39;https://github.com/Bio723-class/example-datasets/raw/master/obama-polls-2008-2011.txt&#39;</span>,</a>
<a class="sourceLine" id="cb407-2" title="2">                    <span class="dt">delim=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">trim_ws=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb407-3" title="3"><span class="co"># note that we needed to use &quot;trim_ws&quot; above because there were </span></a>
<a class="sourceLine" id="cb407-4" title="4"><span class="co"># some lurking spaces in the fields of that tab delimited data file</span></a>
<a class="sourceLine" id="cb407-5" title="5"></a>
<a class="sourceLine" id="cb407-6" title="6"><span class="kw">head</span>(polls)</a>
<a class="sourceLine" id="cb407-7" title="7"><span class="co">## # A tibble: 6 x 6</span></a>
<a class="sourceLine" id="cb407-8" title="8"><span class="co">##   Pollster         Dates      `N/Pop` Approve Disapprove Undecided</span></a>
<a class="sourceLine" id="cb407-9" title="9"><span class="co">##   &lt;chr&gt;            &lt;chr&gt;      &lt;chr&gt;     &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;    </span></a>
<a class="sourceLine" id="cb407-10" title="10"><span class="co">## 1 Rasmussen        9/17-19/11 1500 LV      46         52 -        </span></a>
<a class="sourceLine" id="cb407-11" title="11"><span class="co">## 2 Rasmussen        9/14-16/11 1500 LV      45         55 -        </span></a>
<a class="sourceLine" id="cb407-12" title="12"><span class="co">## 3 Gallup           9/13-15/11 1500 A       39         52 -        </span></a>
<a class="sourceLine" id="cb407-13" title="13"><span class="co">## 4 CBS/Times        9/10-15/11 1452 A       43         50 7        </span></a>
<a class="sourceLine" id="cb407-14" title="14"><span class="co">## 5 Marist/McClatchy 9/13-14/11 825 RV       39         52 9        </span></a>
<a class="sourceLine" id="cb407-15" title="15"><span class="co">## 6 Rasmussen        9/11-13/11 1500 LV      45         54 -</span></a></code></pre></div>
<p>Notice that the <code>Dates</code> column is not very tidy. Each “date” is actually a range of dates of the form <code>Month/DayStart-DayEnd/Year</code> (e.g. “9/1/09” is September 01, 2009). Even nastier, some dates are in the form <code>Month/Day/Year</code> (only a single day) or <code>MonthStart/DayStart-MonthEnd/DayEnd/Year</code> (e.g. “2/26-3/1/11” is February 26,2011 to March 01, 2011) . Whoever formatted the data in this fashion must really hate tidy data! To deal with this nightmare we’re going to use the <code>tidyr::extract()</code> function to employ regular expressions (regex) to parse this complicated data field into it’s constituent parts. For more details on regular expression see the <a href="https://github.com/rstudio/cheatsheets/raw/master/regex.pdf">R Regular Expession Cheat Sheet</a> and <a href="https://r4ds.had.co.nz/strings.html#matching-patterns-with-regular-expressions">R for Data Science</a>.</p>
<div class="sourceCode" id="cb408"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb408-1" title="1">polls &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb408-2" title="2"><span class="st">  </span>polls <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb408-3" title="3"><span class="st">  </span></a>
<a class="sourceLine" id="cb408-4" title="4"><span class="st">  </span><span class="co"># first separate left most and right most fields as month and year respectively</span></a>
<a class="sourceLine" id="cb408-5" title="5"><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">extract</span>(<span class="st">&quot;Dates&quot;</span>, <span class="kw">c</span>(<span class="st">&quot;month&quot;</span>, <span class="st">&quot;day.range&quot;</span>, <span class="st">&quot;year&quot;</span>), <span class="dt">regex=</span><span class="st">&quot;(</span><span class="ch">\\</span><span class="st">d+)/(.+)/(</span><span class="ch">\\</span><span class="st">d+$)&quot;</span>, <span class="dt">convert =</span> <span class="ot">TRUE</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb408-6" title="6"><span class="st">  </span></a>
<a class="sourceLine" id="cb408-7" title="7"><span class="st">  </span><span class="co"># now deal with the complicated middle field. For simplicities sake we&#39;re just</span></a>
<a class="sourceLine" id="cb408-8" title="8"><span class="st">  </span><span class="co"># going to focus on extracting the start day</span></a>
<a class="sourceLine" id="cb408-9" title="9"><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">extract</span>(<span class="st">&quot;day.range&quot;</span>, <span class="kw">c</span>(<span class="st">&quot;day.start&quot;</span>, <span class="st">&quot;day.other&quot;</span>), <span class="dt">regex =</span> <span class="st">&quot;(</span><span class="ch">\\</span><span class="st">d+)(.+)&quot;</span>, <span class="dt">convert =</span> <span class="ot">TRUE</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb408-10" title="10"><span class="st">  </span></a>
<a class="sourceLine" id="cb408-11" title="11"><span class="st">  </span><span class="co"># finally convert YY to 20YY</span></a>
<a class="sourceLine" id="cb408-12" title="12"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">year =</span> <span class="dv">2000</span> <span class="op">+</span><span class="st"> </span>year) </a>
<a class="sourceLine" id="cb408-13" title="13"></a>
<a class="sourceLine" id="cb408-14" title="14"><span class="kw">head</span>(polls)</a>
<a class="sourceLine" id="cb408-15" title="15"><span class="co">## # A tibble: 6 x 9</span></a>
<a class="sourceLine" id="cb408-16" title="16"><span class="co">##   Pollster  month day.start day.other  year `N/Pop` Approve Disapprove Undecided</span></a>
<a class="sourceLine" id="cb408-17" title="17"><span class="co">##   &lt;chr&gt;     &lt;int&gt;     &lt;int&gt; &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt;     &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;    </span></a>
<a class="sourceLine" id="cb408-18" title="18"><span class="co">## 1 Rasmussen     9        17 -19        2011 1500 LV      46         52 -        </span></a>
<a class="sourceLine" id="cb408-19" title="19"><span class="co">## 2 Rasmussen     9        14 -16        2011 1500 LV      45         55 -        </span></a>
<a class="sourceLine" id="cb408-20" title="20"><span class="co">## 3 Gallup        9        13 -15        2011 1500 A       39         52 -        </span></a>
<a class="sourceLine" id="cb408-21" title="21"><span class="co">## 4 CBS/Times     9        10 -15        2011 1452 A       43         50 7        </span></a>
<a class="sourceLine" id="cb408-22" title="22"><span class="co">## 5 Marist/M…     9        13 -14        2011 825 RV       39         52 9        </span></a>
<a class="sourceLine" id="cb408-23" title="23"><span class="co">## 6 Rasmussen     9        11 -13        2011 1500 LV      45         54 -</span></a></code></pre></div>
<p>For the next steps we’ll need the <code>lubridate</code> library (install if needed):</p>
<div class="sourceCode" id="cb409"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb409-1" title="1"><span class="kw">library</span>(lubridate)</a>
<a class="sourceLine" id="cb409-2" title="2"></a>
<a class="sourceLine" id="cb409-3" title="3">polls &lt;-</a>
<a class="sourceLine" id="cb409-4" title="4"><span class="st">  </span>polls <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb409-5" title="5"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">date =</span> <span class="kw">make_date</span>(<span class="dt">year =</span> year, <span class="dt">month=</span>month, <span class="dt">day =</span> day.start))</a>
<a class="sourceLine" id="cb409-6" title="6"></a>
<a class="sourceLine" id="cb409-7" title="7"><span class="kw">head</span>(polls)</a>
<a class="sourceLine" id="cb409-8" title="8"><span class="co">## # A tibble: 6 x 10</span></a>
<a class="sourceLine" id="cb409-9" title="9"><span class="co">##   Pollster month day.start day.other  year `N/Pop` Approve Disapprove Undecided</span></a>
<a class="sourceLine" id="cb409-10" title="10"><span class="co">##   &lt;chr&gt;    &lt;int&gt;     &lt;int&gt; &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt;     &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;    </span></a>
<a class="sourceLine" id="cb409-11" title="11"><span class="co">## 1 Rasmuss…     9        17 -19        2011 1500 LV      46         52 -        </span></a>
<a class="sourceLine" id="cb409-12" title="12"><span class="co">## 2 Rasmuss…     9        14 -16        2011 1500 LV      45         55 -        </span></a>
<a class="sourceLine" id="cb409-13" title="13"><span class="co">## 3 Gallup       9        13 -15        2011 1500 A       39         52 -        </span></a>
<a class="sourceLine" id="cb409-14" title="14"><span class="co">## 4 CBS/Tim…     9        10 -15        2011 1452 A       43         50 7        </span></a>
<a class="sourceLine" id="cb409-15" title="15"><span class="co">## 5 Marist/…     9        13 -14        2011 825 RV       39         52 9        </span></a>
<a class="sourceLine" id="cb409-16" title="16"><span class="co">## 6 Rasmuss…     9        11 -13        2011 1500 LV      45         54 -        </span></a>
<a class="sourceLine" id="cb409-17" title="17"><span class="co">## # … with 1 more variable: date &lt;date&gt;</span></a></code></pre></div>
<div class="sourceCode" id="cb410"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb410-1" title="1">polls.plot &lt;-</a>
<a class="sourceLine" id="cb410-2" title="2"><span class="st">  </span>polls <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb410-3" title="3"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> date, <span class="dt">y =</span> Approve)) <span class="op">+</span></a>
<a class="sourceLine" id="cb410-4" title="4"><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">alpha=</span><span class="fl">0.5</span>, <span class="dt">pch=</span><span class="dv">1</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb410-5" title="5"><span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Date&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Approval Rating&quot;</span>,</a>
<a class="sourceLine" id="cb410-6" title="6">       <span class="dt">title =</span> <span class="st">&quot;Barack Obama&#39;s Approval Ratings, 2008-2011&quot;</span>)</a>
<a class="sourceLine" id="cb410-7" title="7"></a>
<a class="sourceLine" id="cb410-8" title="8">polls.plot</a></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-407-1.png" width="576" /></p>
<p>We can fit the LOESS as so, and get back the predicted values using the <code>predict()</code> function:</p>
<div class="sourceCode" id="cb411"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb411-1" title="1">loess.approval &lt;-<span class="st"> </span><span class="kw">loess</span>(Approve <span class="op">~</span><span class="st"> </span><span class="kw">as.numeric</span>(date), <span class="dt">data =</span> polls)</a>
<a class="sourceLine" id="cb411-2" title="2">loess.predicted.values &lt;-<span class="st"> </span><span class="kw">predict</span>(loess.approval)</a>
<a class="sourceLine" id="cb411-3" title="3"><span class="kw">head</span>(loess.predicted.values)</a>
<a class="sourceLine" id="cb411-4" title="4"><span class="co">## [1] 44.55653 44.59349 44.60572 44.64216 44.60572 44.63006</span></a></code></pre></div>
<p>Usually we’ll want to visualize the LOESS regression, which we can conveniently do with <code>ggplot::geom_smooth</code> without having to explicitly calculate the LOESS:</p>
<div class="sourceCode" id="cb412"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb412-1" title="1">polls.plot <span class="op">+</span></a>
<a class="sourceLine" id="cb412-2" title="2"><span class="st">  </span><span class="kw">geom_smooth</span>(<span class="dt">color=</span><span class="st">&#39;red&#39;</span>, <span class="dt">method=</span><span class="st">&quot;loess&quot;</span>, <span class="dt">se=</span><span class="ot">FALSE</span>)</a></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-409-1.png" width="576" /></p>
<p>Here’s the same data fit with a smaller <code>span</code> (the paramater that controls the “local neighborhood” size in LOESS):</p>
<div class="sourceCode" id="cb413"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb413-1" title="1">polls.plot <span class="op">+</span></a>
<a class="sourceLine" id="cb413-2" title="2"><span class="st">  </span><span class="kw">geom_smooth</span>(<span class="dt">color=</span><span class="st">&#39;red&#39;</span>, <span class="dt">method=</span><span class="st">&quot;loess&quot;</span>, <span class="dt">se=</span><span class="ot">FALSE</span>, <span class="dt">span=</span><span class="fl">0.1</span>)</a></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-410-1.png" width="576" /></p>
<p>The high density of the polling justifies the smaller span, and the additional deviations apparent when the LOESS is fit with the smaller span likely reflect real world changes in approval, induced by a variety of political and other news events.</p>
<p>For example, we can zoom in on 2011:</p>
<div class="sourceCode" id="cb414"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb414-1" title="1">polls.plot <span class="op">+</span></a>
<a class="sourceLine" id="cb414-2" title="2"><span class="st">  </span><span class="kw">geom_smooth</span>(<span class="dt">color=</span><span class="st">&#39;red&#39;</span>, <span class="dt">method=</span><span class="st">&quot;loess&quot;</span>, <span class="dt">se=</span><span class="ot">FALSE</span>, <span class="dt">span=</span><span class="fl">0.1</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb414-3" title="3"><span class="st">  </span><span class="kw">coord_cartesian</span>(<span class="dt">xlim=</span><span class="kw">c</span>(<span class="kw">ymd</span>(<span class="dv">20110101</span>), <span class="kw">ymd</span>(<span class="dv">20110901</span>)), <span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">35</span>,<span class="dv">65</span>)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb414-4" title="4"><span class="st">  </span><span class="kw">scale_x_date</span>(<span class="dt">date_breaks=</span><span class="st">&quot;1 month&quot;</span>, <span class="dt">date_label=</span><span class="st">&quot;%B&quot;</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb414-5" title="5"><span class="st">  </span><span class="kw">labs</span>(<span class="dt">title=</span><span class="st">&quot;Barack Obama&#39;s Approval Ratings, Jan - Sep 2011&quot;</span>)</a></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-411-1.png" width="576" /></p>
<p>Increased approval ratings in January coincide with the approval of a tax deal and a speech to the nation following the shooting of congresswoman Gabbie Giffords in Tuscson, AZ (<a href="https://www.cnbc.com/id/41139968" class="uri">https://www.cnbc.com/id/41139968</a>). The spike apparent in early May coincides with the death of Osama Bin Laden. You might take a look at major policitcal events in other years to see if you can identify drivers behind other approval rating shifts.</p>
</div>
<div id="logistic-regression" class="section level2">
<h2><span class="header-section-number">13.3</span> Logistic regression</h2>
<p>Logistic regression is used when the dependent variable is discrete (often binary). The explanatory variables may be either continuous or discrete.</p>
<p>Examples:</p>
<ul>
<li>Whether a gene is turned off (=0) or on (=1) as a function of levels of various proteins</li>
<li>Whether an individual is healthy (=0) or diseased (=1) as a function of various risk factors.</li>
<li>Whether an individual died (=0) or survived (=1) some selective event as a function of behavior, morphology, etc.</li>
</ul>
<p>We model the binary response variable, <span class="math inline">\(Y\)</span>, as a function of the predictor variables, <span class="math inline">\(X_1\)</span>, <span class="math inline">\(X_2\)</span>, etc as :</p>
<p><span class="math display">\[
P(Y = 1|X_1,\ldots,X_p) = f(\beta_0 + \beta_1 X_1 + \beta_2 X_2 + \cdots + \beta_p X_p)
\]</span></p>
<p>So we’re modeling the <em>probability of the state of Y as a function of a linear combination of the predictor variables</em>.</p>
<p>For logistic regression, <span class="math inline">\(f\)</span> is the logistic function:
<span class="math display">\[
f(z) = \frac{e^z}{1+e^z} = \frac{1}{1 + e^{-z}}
\]</span></p>
<p>Therefore, the bivariate logistic regression is given by:
<span class="math display">\[
P(Y = 1|X) = \frac{1}{1 + e^{-(\beta_0 + \beta_1 X)}}
\]</span></p>
<p>Note that <span class="math inline">\(\beta_0\)</span> here is akin to the intercept in our standard linear regression.</p>
<div id="a-web-app-to-explore-the-logistic-regression-equation" class="section level3">
<h3><span class="header-section-number">13.3.1</span> A web app to explore the logistic regression equation</h3>
<p>To help you develop an intuition for the logistic regression equation, I’ve developed <a href="https://magwenelab.shinyapps.io/exploring_logistic_regression/">a small web app</a>, that allows you to explore how the shape of the regression curve responds to changes in the regression coefficients <span class="math inline">\(\beta_0\)</span> and <span class="math inline">\(\beta_1\)</span>. Open the app in another browser window and play with the sliders that control the coeffients <span class="math inline">\(B_0\)</span> and <span class="math inline">\(B_1\)</span>. In the assignment associated with today’s class you’ll be asked to answer some specific questions based on this app.</p>
</div>
<div id="titanic-data-set" class="section level3">
<h3><span class="header-section-number">13.3.2</span> Titanic data set</h3>
<p><a href="http://bit.ly/bio304-titanic-data"><code>titanic.csv</code></a> contains information about passengers on the Titanic. Variables in this data set include information such as sex, age, passenger class (1st, 2nd, 3rd), and whether or not they survived the sinking of the ship (0 = died, 1 = survived).</p>
<div class="sourceCode" id="cb415"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb415-1" title="1"><span class="kw">library</span>(tidyverse)</a>
<a class="sourceLine" id="cb415-2" title="2"><span class="kw">library</span>(broom)</a>
<a class="sourceLine" id="cb415-3" title="3"><span class="kw">library</span>(cowplot)</a>
<a class="sourceLine" id="cb415-4" title="4"><span class="kw">library</span>(ggthemes)</a></code></pre></div>
<div class="sourceCode" id="cb416"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb416-1" title="1">titanic &lt;-<span class="st"> </span><span class="kw">read_csv</span>(<span class="st">&quot;http://bit.ly/bio304-titanic-data&quot;</span>)</a>
<a class="sourceLine" id="cb416-2" title="2"><span class="kw">names</span>(titanic)</a>
<a class="sourceLine" id="cb416-3" title="3"><span class="co">##  [1] &quot;pclass&quot;    &quot;survived&quot;  &quot;name&quot;      &quot;sex&quot;       &quot;age&quot;       &quot;sibsp&quot;    </span></a>
<a class="sourceLine" id="cb416-4" title="4"><span class="co">##  [7] &quot;parch&quot;     &quot;ticket&quot;    &quot;fare&quot;      &quot;cabin&quot;     &quot;embarked&quot;  &quot;boat&quot;     </span></a>
<a class="sourceLine" id="cb416-5" title="5"><span class="co">## [13] &quot;body&quot;      &quot;home.dest&quot;</span></a></code></pre></div>
</div>
<div id="subsetting-the-data" class="section level3">
<h3><span class="header-section-number">13.3.3</span> Subsetting the data</h3>
<p>We’ve all heard the phrase, “Women and children first”, so we might expect that the probability that a passenger survived the sinking of the Titanic is related to their sex and/or age. Let’s create separate data subsets for male and female passengers.</p>
<div class="sourceCode" id="cb417"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb417-1" title="1">male &lt;-<span class="st"> </span><span class="kw">filter</span>(titanic, sex <span class="op">==</span><span class="st"> &quot;male&quot;</span>)</a>
<a class="sourceLine" id="cb417-2" title="2">female &lt;-<span class="st"> </span><span class="kw">filter</span>(titanic, sex <span class="op">==</span><span class="st"> &quot;female&quot;</span>)</a></code></pre></div>
</div>
<div id="visualizing-survival-as-a-function-of-age" class="section level3">
<h3><span class="header-section-number">13.3.4</span> Visualizing survival as a function of age</h3>
<p>Let’s create visualizations of survival as a function of age for the male and female passengers.</p>
<div class="sourceCode" id="cb418"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb418-1" title="1">fcolor =<span class="st"> &quot;lightcoral&quot;</span></a>
<a class="sourceLine" id="cb418-2" title="2">mcolor =<span class="st"> &quot;lightsteelblue&quot;</span></a>
<a class="sourceLine" id="cb418-3" title="3"></a>
<a class="sourceLine" id="cb418-4" title="4">female.plot &lt;-<span class="st"> </span><span class="kw">ggplot</span>(female, <span class="kw">aes</span>(<span class="dt">x =</span> age, <span class="dt">y =</span> survived)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb418-5" title="5"><span class="st">  </span><span class="kw">geom_jitter</span>(<span class="dt">width =</span> <span class="dv">0</span>, <span class="dt">height =</span> <span class="fl">0.05</span>, <span class="dt">color =</span> fcolor) <span class="op">+</span></a>
<a class="sourceLine" id="cb418-6" title="6"><span class="st">  </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&quot;Female Passengers&quot;</span>)</a>
<a class="sourceLine" id="cb418-7" title="7"></a>
<a class="sourceLine" id="cb418-8" title="8">male.plot &lt;-<span class="st"> </span><span class="kw">ggplot</span>(male, <span class="kw">aes</span>(<span class="dt">x =</span> age, <span class="dt">y =</span> survived)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb418-9" title="9"><span class="st">  </span><span class="kw">geom_jitter</span>(<span class="dt">width =</span> <span class="dv">0</span>, <span class="dt">height =</span> <span class="fl">0.05</span>, <span class="dt">color =</span> mcolor) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb418-10" title="10"><span class="st">  </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&quot;Male Passengers&quot;</span>)</a>
<a class="sourceLine" id="cb418-11" title="11"></a>
<a class="sourceLine" id="cb418-12" title="12"><span class="kw">plot_grid</span>(female.plot, male.plot)</a></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-415-1.png" width="864" /></p>
<p>The jittered points with Y-axis value around one are passengers who survived, the point jittered around zero are those who died.</p>
</div>
<div id="fitting-the-logistic-regression-model" class="section level3">
<h3><span class="header-section-number">13.3.5</span> Fitting the logistic regression model</h3>
<p>The function <code>glm</code> (generalized linear model) can be used to fit the logistic regression model (as well as other models). Setting the argument <code>family = binomial</code> gives us logistic regression. Note that when fitting the model the dependent variable needs to be numeric, so if the data is provided as Boolean (logical) TRUE/FALSE values, they should be converted to integers using <code>as.numeric()</code>.</p>
<p>First we fit the regression for the famale passengers.</p>
<div class="sourceCode" id="cb419"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb419-1" title="1">fit.female &lt;-<span class="st"> </span><span class="kw">glm</span>(survived <span class="op">~</span><span class="st"> </span>age, <span class="dt">family =</span> binomial, female)</a>
<a class="sourceLine" id="cb419-2" title="2"><span class="kw">tidy</span>(fit.female)</a>
<a class="sourceLine" id="cb419-3" title="3"><span class="co">## # A tibble: 2 x 5</span></a>
<a class="sourceLine" id="cb419-4" title="4"><span class="co">##   term        estimate std.error statistic p.value</span></a>
<a class="sourceLine" id="cb419-5" title="5"><span class="co">##   &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb419-6" title="6"><span class="co">## 1 (Intercept)   0.493    0.254        1.94 0.0523 </span></a>
<a class="sourceLine" id="cb419-7" title="7"><span class="co">## 2 age           0.0225   0.00854      2.64 0.00834</span></a></code></pre></div>
<p>The column “estimate” gives the coefficients of the model. The “intercept”" estimate corresponds to <span class="math inline">\(B_0\)</span> in the logistic regression equation, the “age” estimate corresponds to the coefficient <span class="math inline">\(B_1\)</span> in the equation.</p>
<p>Now we repeat the same step for the male passengers.</p>
<div class="sourceCode" id="cb420"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb420-1" title="1">fit.male &lt;-<span class="st"> </span><span class="kw">glm</span>(survived <span class="op">~</span><span class="st"> </span>age, <span class="dt">family =</span> binomial, male)</a>
<a class="sourceLine" id="cb420-2" title="2"><span class="kw">tidy</span>(fit.male)</a>
<a class="sourceLine" id="cb420-3" title="3"><span class="co">## # A tibble: 2 x 5</span></a>
<a class="sourceLine" id="cb420-4" title="4"><span class="co">##   term        estimate std.error statistic p.value</span></a>
<a class="sourceLine" id="cb420-5" title="5"><span class="co">##   &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb420-6" title="6"><span class="co">## 1 (Intercept)  -0.661    0.225       -2.94 0.00329</span></a>
<a class="sourceLine" id="cb420-7" title="7"><span class="co">## 2 age          -0.0238   0.00728     -3.27 0.00109</span></a></code></pre></div>
<p>Notice that the female coefficients are both positive, while the male coefficients are negative. We’ll visualize what this means in terms of the model below.</p>
</div>
<div id="visualizing-the-logistic-regression" class="section level3">
<h3><span class="header-section-number">13.3.6</span> Visualizing the logistic regression</h3>
<p>To visualize the logistic regression fit, we first use the <code>predict</code> function to generate the model predictions about probability of survival as a function of age.</p>
<div class="sourceCode" id="cb421"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb421-1" title="1">ages &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dv">0</span>, <span class="dv">75</span>, <span class="dv">1</span>) <span class="co"># predict survival for ages 0 to 75</span></a>
<a class="sourceLine" id="cb421-2" title="2"></a>
<a class="sourceLine" id="cb421-3" title="3">predicted.female &lt;-<span class="st"> </span><span class="kw">predict</span>(fit.female, </a>
<a class="sourceLine" id="cb421-4" title="4">                            <span class="dt">newdata =</span> <span class="kw">data.frame</span>(<span class="dt">age =</span> ages),</a>
<a class="sourceLine" id="cb421-5" title="5">                            <span class="dt">type =</span> <span class="st">&quot;response&quot;</span>)</a>
<a class="sourceLine" id="cb421-6" title="6"></a>
<a class="sourceLine" id="cb421-7" title="7">predicted.male &lt;-<span class="st"> </span><span class="kw">predict</span>(fit.male,</a>
<a class="sourceLine" id="cb421-8" title="8">                          <span class="dt">newdata =</span> <span class="kw">data.frame</span>(<span class="dt">age =</span> ages),</a>
<a class="sourceLine" id="cb421-9" title="9">                          <span class="dt">type =</span> <span class="st">&quot;response&quot;</span>)</a>
<a class="sourceLine" id="cb421-10" title="10">                            </a></code></pre></div>
<p>Having generated the predicted probabilities of survival we can then add these prediction lines to our previous plot using <code>geom_line</code>.</p>
<div class="sourceCode" id="cb422"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb422-1" title="1">female.logistic.plot &lt;-<span class="st"> </span>female.plot <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb422-2" title="2"><span class="st">  </span><span class="kw">geom_line</span>(<span class="dt">data =</span> <span class="kw">data.frame</span>(<span class="dt">age =</span> ages, <span class="dt">survived =</span> predicted.female),</a>
<a class="sourceLine" id="cb422-3" title="3">            <span class="dt">color =</span> fcolor, <span class="dt">size =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb422-4" title="4"></a>
<a class="sourceLine" id="cb422-5" title="5">male.logistic.plot &lt;-<span class="st"> </span>male.plot <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb422-6" title="6"><span class="st">  </span><span class="kw">geom_line</span>(<span class="dt">data =</span> <span class="kw">data.frame</span>(<span class="dt">age =</span> ages, <span class="dt">survived =</span> predicted.male),</a>
<a class="sourceLine" id="cb422-7" title="7">            <span class="dt">color =</span> mcolor, <span class="dt">size =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb422-8" title="8"></a>
<a class="sourceLine" id="cb422-9" title="9"><span class="kw">plot_grid</span>(female.logistic.plot, male.logistic.plot)</a></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-419-1.png" width="864" /></p>
<p>We see that for the female passengers, the logistic regression predicts that the probability of survival <em>increases</em> with passenger age. In contrast, the model fit to the male passengers suggests that the probability of survival decreases with passenger age. For the male passengers, the data is consistent with “children first”; for female passengers this model doesn’t seem to hold. However, there are other factors to consider as we’ll see below.</p>
</div>
<div id="quick-and-easy-visualization" class="section level3">
<h3><span class="header-section-number">13.3.7</span> Quick and easy visualization</h3>
<p>Here’s an alternative “quick and easy” way to generate the plot above using the awesome power of ggplot. The downside of this approach is we don’t generate the detailed information on the model, which is something you’d certainly want to have in any real analysis.</p>
<div class="sourceCode" id="cb423"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb423-1" title="1"><span class="kw">ggplot</span>(titanic, <span class="kw">aes</span>(<span class="dt">x=</span>age, <span class="dt">y=</span>survived, <span class="dt">color=</span>sex)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb423-2" title="2"><span class="st">  </span><span class="kw">geom_jitter</span>(<span class="dt">width =</span> <span class="dv">0</span>, <span class="dt">height =</span> <span class="fl">0.05</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb423-3" title="3"><span class="st">  </span><span class="kw">geom_smooth</span>(<span class="dt">method=</span><span class="st">&quot;glm&quot;</span>,  <span class="dt">method.args =</span> <span class="kw">list</span>(<span class="dt">family=</span><span class="st">&quot;binomial&quot;</span>))  <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb423-4" title="4"><span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Age&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;P(Survival)&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb423-5" title="5"><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>sex) <span class="op">+</span></a>
<a class="sourceLine" id="cb423-6" title="6"><span class="st">  </span><span class="kw">scale_color_manual</span>(<span class="dt">values =</span> <span class="kw">c</span>(fcolor, mcolor))</a></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-420-1.png" width="864" /></p>
</div>
<div id="impact-of-sex-and-passenger-class-on-the-models" class="section level3">
<h3><span class="header-section-number">13.3.8</span> Impact of sex and passenger class on the models</h3>
<p>In our previous analysis we considered the relationship between survival and age, conditioned (facted) on passenger sex. In a complex data set like this one, it is often useful to condition on multiple variables simultaneously. Lets extend our visualization to look at the regression faceted on both class and sex, using <code>facet_grid</code>:</p>
<div class="sourceCode" id="cb424"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb424-1" title="1"><span class="kw">ggplot</span>(titanic, <span class="kw">aes</span>(<span class="dt">x=</span>age, <span class="dt">y=</span>survived, <span class="dt">color=</span>sex)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb424-2" title="2"><span class="st">  </span><span class="kw">geom_jitter</span>(<span class="dt">width =</span> <span class="dv">0</span>, <span class="dt">height =</span> <span class="fl">0.05</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb424-3" title="3"><span class="st">  </span><span class="kw">geom_smooth</span>(<span class="dt">method=</span><span class="st">&quot;glm&quot;</span>,  <span class="dt">method.args =</span> <span class="kw">list</span>(<span class="dt">family=</span><span class="st">&quot;binomial&quot;</span>))  <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb424-4" title="4"><span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Age&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;P(Survival)&quot;</span>) <span class="op">+</span><span class="st">  </span></a>
<a class="sourceLine" id="cb424-5" title="5"><span class="st">  </span><span class="kw">facet_grid</span>(pclass <span class="op">~</span><span class="st"> </span>sex) <span class="op">+</span></a>
<a class="sourceLine" id="cb424-6" title="6"><span class="st">  </span><span class="kw">scale_color_manual</span>(<span class="dt">values =</span> <span class="kw">c</span>(fcolor, mcolor)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb424-7" title="7"><span class="st">  </span><span class="kw">theme_few</span>()</a></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-421-1.png" width="864" /></p>
<p>Having conditioned on both sex and ticket class, our figure now reveals a much more complex relationship between age and survival. Almost all first class female passengers survived, regardless of age. For second calss female passengers, the logistic regression suggests a very modest decrease in survival with increasing age. The negative relationship between age and survival is stronger still for third class females. Male passengers on the other hand show a negative relationship between sex and survival, regardless of class, but the models suggest that there are still class specific differences in this relationship.</p>
</div>
<div id="fitting-multiple-models-based-on-groupings-use-dplyrdo" class="section level3">
<h3><span class="header-section-number">13.3.9</span> Fitting multiple models based on groupings use <code>dplyr::do</code></h3>
<p>In the figure above we used <code>ggplot</code> and <code>facet_grid</code> to visualize logistic regression of survival on age, conditioned on both sex and class. What if we wanted to calculate the terms of the logistic regressions for each combination of these two categorical variables? There are three passenger classes and two sexes, meaning we’d have to create six data subsets and fit the model six times if we used the same approach we used previously. Luckily, <code>dplyr</code> provides a powerful function called <code>do()</code> that allows us to carry out arbitrary computations on grouped data.</p>
<p>There are two ways to use <code>do()</code>. The first way is to give the expressions you evaluate in <code>do()</code> a name, in which case <code>do()</code> will store the results in a column. The second way to use <code>do()</code> is for the expression to return a data frame.</p>
<p>In this first example, the model fits are stored in the <code>fits</code> column. When using <code>do()</code> you can refer to the groupings using a period (<code>.</code>):</p>
<div class="sourceCode" id="cb425"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb425-1" title="1">grouped.models &lt;-</a>
<a class="sourceLine" id="cb425-2" title="2"><span class="st">  </span>titanic <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb425-3" title="3"><span class="st">  </span><span class="kw">group_by</span>(sex, pclass) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb425-4" title="4"><span class="st">  </span><span class="kw">do</span>(<span class="dt">fits =</span> <span class="kw">glm</span>(survived <span class="op">~</span><span class="st"> </span>age, <span class="dt">family =</span> binomial, <span class="dt">data =</span> .))</a>
<a class="sourceLine" id="cb425-5" title="5"></a>
<a class="sourceLine" id="cb425-6" title="6">grouped.models</a>
<a class="sourceLine" id="cb425-7" title="7"><span class="co">## # A tibble: 6 x 3</span></a>
<a class="sourceLine" id="cb425-8" title="8"><span class="co">##   sex    pclass fits  </span></a>
<a class="sourceLine" id="cb425-9" title="9"><span class="co">##   &lt;chr&gt;   &lt;dbl&gt; &lt;list&gt;</span></a>
<a class="sourceLine" id="cb425-10" title="10"><span class="co">## 1 female      1 &lt;glm&gt; </span></a>
<a class="sourceLine" id="cb425-11" title="11"><span class="co">## 2 female      2 &lt;glm&gt; </span></a>
<a class="sourceLine" id="cb425-12" title="12"><span class="co">## 3 female      3 &lt;glm&gt; </span></a>
<a class="sourceLine" id="cb425-13" title="13"><span class="co">## 4 male        1 &lt;glm&gt; </span></a>
<a class="sourceLine" id="cb425-14" title="14"><span class="co">## 5 male        2 &lt;glm&gt; </span></a>
<a class="sourceLine" id="cb425-15" title="15"><span class="co">## 6 male        3 &lt;glm&gt;</span></a></code></pre></div>
<p>Notice that the “fits” column doesn’t explicitly print out the details of the model. The object returned by <code>glm()</code> can’t be simply represented as text string (it’s a list), so we seea place holder string that tells us that there is data here represented a glm object. However, we can access the the columns with the fits just like any other variable:</p>
<div class="sourceCode" id="cb426"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb426-1" title="1"><span class="co"># get the summary of the second logistic regression (Female, 2nd Class) </span></a>
<a class="sourceLine" id="cb426-2" title="2"><span class="kw">tidy</span>(grouped.models<span class="op">$</span>fits[[<span class="dv">2</span>]])</a>
<a class="sourceLine" id="cb426-3" title="3"><span class="co">## # A tibble: 2 x 5</span></a>
<a class="sourceLine" id="cb426-4" title="4"><span class="co">##   term        estimate std.error statistic  p.value</span></a>
<a class="sourceLine" id="cb426-5" title="5"><span class="co">##   &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb426-6" title="6"><span class="co">## 1 (Intercept)   3.49      0.904       3.86 0.000112</span></a>
<a class="sourceLine" id="cb426-7" title="7"><span class="co">## 2 age          -0.0450    0.0255     -1.77 0.0773</span></a></code></pre></div>
<p>Now we illustrate the second approach to using <code>do()</code>. When no name is provided, <code>do()</code> expects its expression to return a dataframe. Here we use the <code>broom::tidy()</code> function to get the key results of each fit model into a data frame:</p>
<div class="sourceCode" id="cb427"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb427-1" title="1">titanic <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb427-2" title="2"><span class="st">  </span><span class="kw">group_by</span>(sex, pclass) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb427-3" title="3"><span class="st">  </span><span class="kw">do</span>(<span class="kw">tidy</span>(<span class="kw">glm</span>(survived <span class="op">~</span><span class="st"> </span>age, <span class="dt">family =</span> binomial, <span class="dt">data =</span> .)))</a>
<a class="sourceLine" id="cb427-4" title="4"><span class="co">## # A tibble: 12 x 7</span></a>
<a class="sourceLine" id="cb427-5" title="5"><span class="co">##    sex    pclass term        estimate std.error statistic    p.value</span></a>
<a class="sourceLine" id="cb427-6" title="6"><span class="co">##    &lt;chr&gt;   &lt;dbl&gt; &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;      &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb427-7" title="7"><span class="co">##  1 female      1 (Intercept)  2.90       1.24       2.34  0.0193    </span></a>
<a class="sourceLine" id="cb427-8" title="8"><span class="co">##  2 female      1 age          0.00960    0.0326     0.294 0.769     </span></a>
<a class="sourceLine" id="cb427-9" title="9"><span class="co">##  3 female      2 (Intercept)  3.49       0.904      3.86  0.000112  </span></a>
<a class="sourceLine" id="cb427-10" title="10"><span class="co">##  4 female      2 age         -0.0450     0.0255    -1.77  0.0773    </span></a>
<a class="sourceLine" id="cb427-11" title="11"><span class="co">##  5 female      3 (Intercept)  0.289      0.341      0.846 0.398     </span></a>
<a class="sourceLine" id="cb427-12" title="12"><span class="co">##  6 female      3 age         -0.0178     0.0136    -1.31  0.190     </span></a>
<a class="sourceLine" id="cb427-13" title="13"><span class="co">##  7 male        1 (Intercept)  0.880      0.527      1.67  0.0951    </span></a>
<a class="sourceLine" id="cb427-14" title="14"><span class="co">##  8 male        1 age         -0.0374     0.0128    -2.93  0.00335   </span></a>
<a class="sourceLine" id="cb427-15" title="15"><span class="co">##  9 male        2 (Intercept)  1.04       0.599      1.74  0.0818    </span></a>
<a class="sourceLine" id="cb427-16" title="16"><span class="co">## 10 male        2 age         -0.112      0.0249    -4.50  0.00000672</span></a>
<a class="sourceLine" id="cb427-17" title="17"><span class="co">## 11 male        3 (Intercept) -0.761      0.342     -2.23  0.0259    </span></a>
<a class="sourceLine" id="cb427-18" title="18"><span class="co">## 12 male        3 age         -0.0339     0.0134    -2.53  0.0113</span></a></code></pre></div>
<p>Using this approach we get a nice data frame showing the logistic regression coefficients, and associated statistics (standard error, P-values, etc) for the regression of survival on age, for each combination of sex and class.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="linear-regression-models-ii.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="principal-components-analysis.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "section"
},
"toolbar": {
"position": "fixed"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
