<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Biology 723: Statistical Computing for Biologists</title>
  <meta name="description" content="Hands-on materials for Bio 723 course at Duke University.">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Biology 723: Statistical Computing for Biologists" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Biology 723: Statistical Computing for Biologists" />
  
  <meta name="twitter:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

<meta name="author" content="Paul M. Magwene">


<meta name="date" content="2018-03-21">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="matrices-in-r.html">
<link rel="next" href="principal-components-analysis.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#how-to-use-these-lecture-notes"><i class="fa fa-check"></i><b>1.1</b> How to use these lecture notes</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html"><i class="fa fa-check"></i><b>2</b> Getting Started with R</a><ul>
<li class="chapter" data-level="2.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-r"><i class="fa fa-check"></i><b>2.1</b> What is R?</a></li>
<li class="chapter" data-level="2.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-rstudio"><i class="fa fa-check"></i><b>2.2</b> What is RStudio?</a></li>
<li class="chapter" data-level="2.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#entering-commands-in-the-console"><i class="fa fa-check"></i><b>2.3</b> Entering commands in the console</a></li>
<li class="chapter" data-level="2.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#comments"><i class="fa fa-check"></i><b>2.4</b> Comments</a></li>
<li class="chapter" data-level="2.5" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#using-r-as-a-calculator"><i class="fa fa-check"></i><b>2.5</b> Using R as a Calculator</a><ul>
<li class="chapter" data-level="2.5.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#common-mathematical-functions"><i class="fa fa-check"></i><b>2.5.1</b> Common mathematical functions</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#numeric-data-types-in-r"><i class="fa fa-check"></i><b>2.6</b> Numeric data types in R</a></li>
<li class="chapter" data-level="2.7" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#variable-assignment-in-r"><i class="fa fa-check"></i><b>2.7</b> Variable assignment in R</a><ul>
<li class="chapter" data-level="2.7.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#valid-variable-names"><i class="fa fa-check"></i><b>2.7.1</b> Valid variable names</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#working-with-vectors-in-r"><i class="fa fa-check"></i><b>2.8</b> Working with Vectors in R</a><ul>
<li class="chapter" data-level="2.8.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#vector-arithmetic"><i class="fa fa-check"></i><b>2.8.1</b> Vector Arithmetic</a></li>
<li class="chapter" data-level="2.8.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#vector-recycling"><i class="fa fa-check"></i><b>2.8.2</b> Vector recycling</a></li>
<li class="chapter" data-level="2.8.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#simple-statistical-functions-for-numeric-vectors"><i class="fa fa-check"></i><b>2.8.3</b> Simple statistical functions for numeric vectors</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#the-r-help-system"><i class="fa fa-check"></i><b>2.9</b> The R Help System</a><ul>
<li class="chapter" data-level="2.9.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#getting-help-from-the-console"><i class="fa fa-check"></i><b>2.9.1</b> Getting help from the console</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#packages"><i class="fa fa-check"></i><b>2.10</b> Packages</a><ul>
<li class="chapter" data-level="2.10.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-console"><i class="fa fa-check"></i><b>2.10.1</b> Installing packages from the console</a></li>
<li class="chapter" data-level="2.10.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#install-the-tidyverse-package"><i class="fa fa-check"></i><b>2.10.2</b> Install the tidyverse package</a></li>
<li class="chapter" data-level="2.10.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-rstudio-dialog"><i class="fa fa-check"></i><b>2.10.3</b> Installing packages from the RStudio dialog</a></li>
<li class="chapter" data-level="2.10.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#install-the-stringr-package"><i class="fa fa-check"></i><b>2.10.4</b> Install the stringr package</a></li>
<li class="chapter" data-level="2.10.5" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#loading-packages-with-the-library-function"><i class="fa fa-check"></i><b>2.10.5</b> Loading packages with the <code>library()</code> function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html"><i class="fa fa-check"></i><b>3</b> R Basics: Data types and data structures</a><ul>
<li class="chapter" data-level="3.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-boolean-values"><i class="fa fa-check"></i><b>3.1</b> Logical (Boolean) values</a><ul>
<li class="chapter" data-level="3.1.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#comparison-operators-applied-to-numeric-vectors"><i class="fa fa-check"></i><b>3.1.1</b> Comparison operators applied to numeric vectors</a></li>
<li class="chapter" data-level="3.1.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-vectors"><i class="fa fa-check"></i><b>3.1.2</b> Logical vectors</a></li>
<li class="chapter" data-level="3.1.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-operators"><i class="fa fa-check"></i><b>3.1.3</b> Logical operators</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#character-strings"><i class="fa fa-check"></i><b>3.2</b> Character strings</a><ul>
<li class="chapter" data-level="3.2.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#the-stringr-package"><i class="fa fa-check"></i><b>3.2.1</b> The <code>stringr</code> package</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#vectors"><i class="fa fa-check"></i><b>3.3</b> Vectors</a><ul>
<li class="chapter" data-level="3.3.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#indexing-vectors"><i class="fa fa-check"></i><b>3.3.1</b> Indexing Vectors</a></li>
<li class="chapter" data-level="3.3.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#combining-indexing-and-comparison-of-vectors"><i class="fa fa-check"></i><b>3.3.2</b> Combining Indexing and Comparison of Vectors</a></li>
<li class="chapter" data-level="3.3.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#vector-manipulation"><i class="fa fa-check"></i><b>3.3.3</b> Vector manipulation</a></li>
<li class="chapter" data-level="3.3.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#vectors-from-regular-sequences"><i class="fa fa-check"></i><b>3.3.4</b> Vectors from regular sequences</a></li>
<li class="chapter" data-level="3.3.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#additional-functions-for-working-with-vectors"><i class="fa fa-check"></i><b>3.3.5</b> Additional functions for working with vectors</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#lists"><i class="fa fa-check"></i><b>3.4</b> Lists</a><ul>
<li class="chapter" data-level="3.4.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#length-and-type-of-lists"><i class="fa fa-check"></i><b>3.4.1</b> Length and type of lists</a></li>
<li class="chapter" data-level="3.4.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#indexing-lists"><i class="fa fa-check"></i><b>3.4.2</b> Indexing lists</a></li>
<li class="chapter" data-level="3.4.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#naming-list-elements"><i class="fa fa-check"></i><b>3.4.3</b> Naming list elements</a></li>
<li class="chapter" data-level="3.4.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#the-operator"><i class="fa fa-check"></i><b>3.4.4</b> The <code>$</code> operator</a></li>
<li class="chapter" data-level="3.4.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#changing-and-adding-lists-items"><i class="fa fa-check"></i><b>3.4.5</b> Changing and adding lists items</a></li>
<li class="chapter" data-level="3.4.6" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#combining-lists"><i class="fa fa-check"></i><b>3.4.6</b> Combining lists</a></li>
<li class="chapter" data-level="3.4.7" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#converting-lists-to-vectors"><i class="fa fa-check"></i><b>3.4.7</b> Converting lists to vectors</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#data-frames"><i class="fa fa-check"></i><b>3.5</b> Data frames</a><ul>
<li class="chapter" data-level="3.5.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#creating-a-data-frame"><i class="fa fa-check"></i><b>3.5.1</b> Creating a data frame</a></li>
<li class="chapter" data-level="3.5.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#type-and-class-for-data-frames"><i class="fa fa-check"></i><b>3.5.2</b> Type and class for data frames</a></li>
<li class="chapter" data-level="3.5.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#length-and-dimension-for-data-frames"><i class="fa fa-check"></i><b>3.5.3</b> Length and dimension for data frames</a></li>
<li class="chapter" data-level="3.5.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#indexing-and-accessing-data-frames"><i class="fa fa-check"></i><b>3.5.4</b> Indexing and accessing data frames</a></li>
<li class="chapter" data-level="3.5.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-indexing-of-data-frames"><i class="fa fa-check"></i><b>3.5.5</b> Logical indexing of data frames</a></li>
<li class="chapter" data-level="3.5.6" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#adding-columns-to-a-data-frame"><i class="fa fa-check"></i><b>3.5.6</b> Adding columns to a data frame</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html"><i class="fa fa-check"></i><b>4</b> R Markdown and R Notebooks</a><ul>
<li class="chapter" data-level="4.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#r-notebooks"><i class="fa fa-check"></i><b>4.1</b> R Notebooks</a></li>
<li class="chapter" data-level="4.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#creating-an-r-notebook"><i class="fa fa-check"></i><b>4.2</b> Creating an R Notebook</a></li>
<li class="chapter" data-level="4.3" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#the-default-r-notebook-template"><i class="fa fa-check"></i><b>4.3</b> The default R Notebook template</a></li>
<li class="chapter" data-level="4.4" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-and-non-code-blocks"><i class="fa fa-check"></i><b>4.4</b> Code and Non-code blocks</a><ul>
<li class="chapter" data-level="4.4.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#non-code-blocks"><i class="fa fa-check"></i><b>4.4.1</b> Non-code blocks</a></li>
<li class="chapter" data-level="4.4.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-blocks"><i class="fa fa-check"></i><b>4.4.2</b> Code blocks</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-a-code-chunk"><i class="fa fa-check"></i><b>4.5</b> Running a code chunk</a></li>
<li class="chapter" data-level="4.6" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-all-code-chunks-above"><i class="fa fa-check"></i><b>4.6</b> Running all code chunks above</a></li>
<li class="chapter" data-level="4.7" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#knitting-an-r-markdown-to-html"><i class="fa fa-check"></i><b>4.7</b> “Knitting an” R Markdown to HTML</a></li>
<li class="chapter" data-level="4.8" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#sharing-your-reproducible-r-notebook"><i class="fa fa-check"></i><b>4.8</b> Sharing your reproducible R Notebook</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html"><i class="fa fa-check"></i><b>5</b> Introduction to ggplot2</a><ul>
<li class="chapter" data-level="5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#loading-ggplot2"><i class="fa fa-check"></i><b>5.1</b> Loading ggplot2</a></li>
<li class="chapter" data-level="5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#example-data-set-andersons-iris-data"><i class="fa fa-check"></i><b>5.2</b> Example data set: Anderson’s Iris Data</a></li>
<li class="chapter" data-level="5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#template-for-single-layer-plots-in-ggplot2"><i class="fa fa-check"></i><b>5.3</b> Template for single layer plots in ggplot2</a></li>
<li class="chapter" data-level="5.4" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#an-aside-about-function-arguments"><i class="fa fa-check"></i><b>5.4</b> An aside about function arguments</a></li>
<li class="chapter" data-level="5.5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#strip-plots"><i class="fa fa-check"></i><b>5.5</b> Strip plots</a><ul>
<li class="chapter" data-level="5.5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#jittering-data"><i class="fa fa-check"></i><b>5.5.1</b> Jittering data</a></li>
<li class="chapter" data-level="5.5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-categorical-information"><i class="fa fa-check"></i><b>5.5.2</b> Adding categorical information</a></li>
<li class="chapter" data-level="5.5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#rotating-plot-coordinates"><i class="fa fa-check"></i><b>5.5.3</b> Rotating plot coordinates</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#histograms"><i class="fa fa-check"></i><b>5.6</b> Histograms</a><ul>
<li class="chapter" data-level="5.6.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#variations-on-histograms-when-considering-categorical-data"><i class="fa fa-check"></i><b>5.6.1</b> Variations on histograms when considering categorical data</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#faceting-to-depict-categorical-information"><i class="fa fa-check"></i><b>5.7</b> Faceting to depict categorical information</a></li>
<li class="chapter" data-level="5.8" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots"><i class="fa fa-check"></i><b>5.8</b> Density plots</a></li>
<li class="chapter" data-level="5.9" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-or-beanplot"><i class="fa fa-check"></i><b>5.9</b> Violin or Beanplot</a></li>
<li class="chapter" data-level="5.10" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplots"><i class="fa fa-check"></i><b>5.10</b> Boxplots</a></li>
<li class="chapter" data-level="5.11" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#building-complex-visualizations-with-layers"><i class="fa fa-check"></i><b>5.11</b> Building complex visualizations with layers</a></li>
<li class="chapter" data-level="5.12" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#useful-combination-plots"><i class="fa fa-check"></i><b>5.12</b> Useful combination plots</a><ul>
<li class="chapter" data-level="5.12.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplot-plus-strip-plot"><i class="fa fa-check"></i><b>5.12.1</b> Boxplot plus strip plot</a></li>
<li class="chapter" data-level="5.12.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#setting-shared-aesthetics"><i class="fa fa-check"></i><b>5.12.2</b> Setting shared aesthetics</a></li>
</ul></li>
<li class="chapter" data-level="5.13" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot-layers-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>5.13</b> ggplot layers can be assigned to variables</a><ul>
<li class="chapter" data-level="5.13.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-plot-plus-strip-plot"><i class="fa fa-check"></i><b>5.13.1</b> Violin plot plus strip plot</a></li>
</ul></li>
<li class="chapter" data-level="5.14" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-titles-and-tweaking-axis-labels"><i class="fa fa-check"></i><b>5.14</b> Adding titles and tweaking axis labels</a></li>
<li class="chapter" data-level="5.15" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot2-themes"><i class="fa fa-check"></i><b>5.15</b> ggplot2 themes</a><ul>
<li class="chapter" data-level="5.15.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#further-customization-with-ggplot2theme"><i class="fa fa-check"></i><b>5.15.1</b> Further customization with <code>ggplot2::theme</code></a></li>
</ul></li>
<li class="chapter" data-level="5.16" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#other-aspects-of-ggplots-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>5.16</b> Other aspects of ggplots can be assigned to variables</a></li>
<li class="chapter" data-level="5.17" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-plots"><i class="fa fa-check"></i><b>5.17</b> Bivariate plots</a><ul>
<li class="chapter" data-level="5.17.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#scatter-plots"><i class="fa fa-check"></i><b>5.17.1</b> Scatter plots</a></li>
</ul></li>
<li class="chapter" data-level="5.18" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-density-plots"><i class="fa fa-check"></i><b>5.18</b> Bivariate density plots</a></li>
<li class="chapter" data-level="5.19" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#combining-scatter-plots-and-density-plots-with-categorical-information"><i class="fa fa-check"></i><b>5.19</b> Combining Scatter Plots and Density Plots with Categorical Information</a></li>
<li class="chapter" data-level="5.20" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots-with-fill"><i class="fa fa-check"></i><b>5.20</b> Density plots with fill</a></li>
<li class="chapter" data-level="5.21" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#d-bin-and-hex-plots"><i class="fa fa-check"></i><b>5.21</b> 2D bin and hex plots</a></li>
<li class="chapter" data-level="5.22" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#the-cowplot-package"><i class="fa fa-check"></i><b>5.22</b> The <code>cowplot</code> package</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html"><i class="fa fa-check"></i><b>6</b> Introduction to the <code>dplyr</code> package</a><ul>
<li class="chapter" data-level="6.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#libraries"><i class="fa fa-check"></i><b>6.1</b> Libraries</a></li>
<li class="chapter" data-level="6.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#reading-data-with-the-readr-package"><i class="fa fa-check"></i><b>6.2</b> Reading data with the <code>readr</code> package</a><ul>
<li class="chapter" data-level="6.2.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#example-data-nc-births"><i class="fa fa-check"></i><b>6.2.1</b> Example data: NC Births</a></li>
<li class="chapter" data-level="6.2.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#reading-excel-files"><i class="fa fa-check"></i><b>6.2.2</b> Reading Excel files</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#a-note-on-tibbles"><i class="fa fa-check"></i><b>6.3</b> A note on “tibbles”</a></li>
<li class="chapter" data-level="6.4" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#data-filtering-and-transformation-with-dplyr"><i class="fa fa-check"></i><b>6.4</b> Data filtering and transformation with <code>dplyr</code></a></li>
<li class="chapter" data-level="6.5" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#dplyrs-verbs"><i class="fa fa-check"></i><b>6.5</b> dplyr’s “verbs”</a><ul>
<li class="chapter" data-level="6.5.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#select"><i class="fa fa-check"></i><b>6.5.1</b> <code>select</code></a></li>
<li class="chapter" data-level="6.5.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#filter"><i class="fa fa-check"></i><b>6.5.2</b> <code>filter</code></a></li>
<li class="chapter" data-level="6.5.3" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#mutate"><i class="fa fa-check"></i><b>6.5.3</b> <code>mutate</code></a></li>
<li class="chapter" data-level="6.5.4" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#arrange"><i class="fa fa-check"></i><b>6.5.4</b> <code>arrange</code></a></li>
<li class="chapter" data-level="6.5.5" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#summarize"><i class="fa fa-check"></i><b>6.5.5</b> <code>summarize</code></a></li>
<li class="chapter" data-level="6.5.6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#group_by"><i class="fa fa-check"></i><b>6.5.6</b> <code>group_by</code></a></li>
<li class="chapter" data-level="6.5.7" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#combining-grouping-and-summarizing"><i class="fa fa-check"></i><b>6.5.7</b> Combining grouping and summarizing</a></li>
<li class="chapter" data-level="6.5.8" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#scoped-variants-of-mutate-and-summarize"><i class="fa fa-check"></i><b>6.5.8</b> Scoped variants of <code>mutate</code> and <code>summarize</code></a></li>
<li class="chapter" data-level="6.5.9" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#combining-summarize-with-grouping-aesthetics-in-ggplot2"><i class="fa fa-check"></i><b>6.5.9</b> Combining summarize with grouping aesthetics in <code>ggplot2</code></a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#pipes"><i class="fa fa-check"></i><b>6.6</b> Pipes</a><ul>
<li class="chapter" data-level="6.6.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#install-and-load-magrittr"><i class="fa fa-check"></i><b>6.6.1</b> Install and load <code>magrittr</code></a></li>
<li class="chapter" data-level="6.6.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-basic-pipe-operator"><i class="fa fa-check"></i><b>6.6.2</b> The basic pipe operator</a></li>
<li class="chapter" data-level="6.6.3" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#an-example-without-pipes"><i class="fa fa-check"></i><b>6.6.3</b> An example without pipes</a></li>
<li class="chapter" data-level="6.6.4" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-same-example-using-pipes"><i class="fa fa-check"></i><b>6.6.4</b> The same example using pipes</a></li>
<li class="chapter" data-level="6.6.5" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#assigning-the-output-of-a-statement-involving-pipes-to-a-variable"><i class="fa fa-check"></i><b>6.6.5</b> Assigning the output of a statement involving pipes to a variable</a></li>
<li class="chapter" data-level="6.6.6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#compound-assignment-pipe-operator"><i class="fa fa-check"></i><b>6.6.6</b> Compound assignment pipe operator</a></li>
<li class="chapter" data-level="6.6.7" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-dot-operator-with-pipes"><i class="fa fa-check"></i><b>6.6.7</b> The dot operator with pipes</a></li>
<li class="chapter" data-level="6.6.8" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-exposition-pipe-operator"><i class="fa fa-check"></i><b>6.6.8</b> The exposition pipe operator</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>7</b> Data wrangling</a><ul>
<li class="chapter" data-level="7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#libraries-1"><i class="fa fa-check"></i><b>7.1</b> Libraries</a></li>
<li class="chapter" data-level="7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#data"><i class="fa fa-check"></i><b>7.2</b> Data</a></li>
<li class="chapter" data-level="7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#renaming-data-frame-columms"><i class="fa fa-check"></i><b>7.3</b> Renaming data frame columms</a></li>
<li class="chapter" data-level="7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-unneeded-columns"><i class="fa fa-check"></i><b>7.4</b> Dropping unneeded columns</a><ul>
<li class="chapter" data-level="7.4.1" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-all-empty-columns"><i class="fa fa-check"></i><b>7.4.1</b> Finding all empty columns</a></li>
<li class="chapter" data-level="7.4.2" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-columns-by-matching-names"><i class="fa fa-check"></i><b>7.4.2</b> Dropping columns by matching names</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="data-wrangling.html"><a href="data-wrangling.html#reshaping-data-with-tidyr"><i class="fa fa-check"></i><b>7.5</b> Reshaping data with tidyr</a><ul>
<li class="chapter" data-level="7.5.1" data-path="data-wrangling.html"><a href="data-wrangling.html#wide-to-long-conversions-using-tidyrgather"><i class="fa fa-check"></i><b>7.5.1</b> Wide to long conversions using <code>tidyr::gather</code></a></li>
<li class="chapter" data-level="7.5.2" data-path="data-wrangling.html"><a href="data-wrangling.html#extracting-information-from-combined-variables-using-tidyrextract"><i class="fa fa-check"></i><b>7.5.2</b> Extracting information from combined variables using <code>tidyr::extract</code></a></li>
<li class="chapter" data-level="7.5.3" data-path="data-wrangling.html"><a href="data-wrangling.html#subsetting-rows"><i class="fa fa-check"></i><b>7.5.3</b> Subsetting rows</a></li>
<li class="chapter" data-level="7.5.4" data-path="data-wrangling.html"><a href="data-wrangling.html#splitting-columns"><i class="fa fa-check"></i><b>7.5.4</b> Splitting columns</a></li>
<li class="chapter" data-level="7.5.5" data-path="data-wrangling.html"><a href="data-wrangling.html#combining-data-frame-rows"><i class="fa fa-check"></i><b>7.5.5</b> Combining data frame rows</a></li>
<li class="chapter" data-level="7.5.6" data-path="data-wrangling.html"><a href="data-wrangling.html#sorting-data-frame-rows"><i class="fa fa-check"></i><b>7.5.6</b> Sorting data frame rows</a></li>
<li class="chapter" data-level="7.5.7" data-path="data-wrangling.html"><a href="data-wrangling.html#collapsing-columns-using-tidyrunite"><i class="fa fa-check"></i><b>7.5.7</b> Collapsing columns using <code>tidyr::unite</code></a></li>
<li class="chapter" data-level="7.5.8" data-path="data-wrangling.html"><a href="data-wrangling.html#splitting-columns-using-tidyrseparate"><i class="fa fa-check"></i><b>7.5.8</b> Splitting columns using <code>tidyr::separate</code></a></li>
<li class="chapter" data-level="7.5.9" data-path="data-wrangling.html"><a href="data-wrangling.html#long-to-wide-conversion-using-tidyrspread"><i class="fa fa-check"></i><b>7.5.9</b> Long-to-wide conversion using <code>tidyr::spread</code></a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="data-wrangling.html"><a href="data-wrangling.html#using-your-tidy-data"><i class="fa fa-check"></i><b>7.6</b> Using your tidy data</a><ul>
<li class="chapter" data-level="7.6.1" data-path="data-wrangling.html"><a href="data-wrangling.html#visualizing-gene-expression-time-series"><i class="fa fa-check"></i><b>7.6.1</b> Visualizing gene expression time series</a></li>
<li class="chapter" data-level="7.6.2" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-the-most-variable-genes"><i class="fa fa-check"></i><b>7.6.2</b> Finding the most variable genes</a></li>
<li class="chapter" data-level="7.6.3" data-path="data-wrangling.html"><a href="data-wrangling.html#exploring-bivariate-relationships-using-wide-data"><i class="fa fa-check"></i><b>7.6.3</b> Exploring bivariate relationships using “wide” data</a></li>
<li class="chapter" data-level="7.6.4" data-path="data-wrangling.html"><a href="data-wrangling.html#calculating-correlations-between-genes"><i class="fa fa-check"></i><b>7.6.4</b> Calculating correlations between genes</a></li>
<li class="chapter" data-level="7.6.5" data-path="data-wrangling.html"><a href="data-wrangling.html#adding-new-columns-and-combining-filtered-data-frames"><i class="fa fa-check"></i><b>7.6.5</b> Adding new columns and combining filtered data frames</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="data-wrangling.html"><a href="data-wrangling.html#heat-maps"><i class="fa fa-check"></i><b>7.7</b> Heat maps</a><ul>
<li class="chapter" data-level="7.7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#rcolorbrewer"><i class="fa fa-check"></i><b>7.7.1</b> RColorBrewer</a></li>
<li class="chapter" data-level="7.7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#creating-a-heat-map-using-geom_tile"><i class="fa fa-check"></i><b>7.7.2</b> Creating a heat map using <code>geom_tile</code></a></li>
<li class="chapter" data-level="7.7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#a-fancy-figure"><i class="fa fa-check"></i><b>7.7.3</b> A “fancy” figure</a></li>
<li class="chapter" data-level="7.7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#sorting-and-visualizing-genes-by-time-of-maximum-expression"><i class="fa fa-check"></i><b>7.7.4</b> Sorting and visualizing genes by time of maximum expression</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html"><i class="fa fa-check"></i><b>8</b> Vectors and Functions</a><ul>
<li class="chapter" data-level="8.1" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#libraries-2"><i class="fa fa-check"></i><b>8.1</b> Libraries</a></li>
<li class="chapter" data-level="8.2" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#vector-mathematics-in-r"><i class="fa fa-check"></i><b>8.2</b> Vector Mathematics in R</a></li>
<li class="chapter" data-level="8.3" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#writing-your-own-functions-in-r"><i class="fa fa-check"></i><b>8.3</b> Writing Your Own Functions in R</a></li>
<li class="chapter" data-level="8.4" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#function-arguments"><i class="fa fa-check"></i><b>8.4</b> Function arguments</a></li>
<li class="chapter" data-level="8.5" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#writing-functions-with-optional-arguments"><i class="fa fa-check"></i><b>8.5</b> Writing functions with optional arguments</a></li>
<li class="chapter" data-level="8.6" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#putting-r-functions-in-scripts"><i class="fa fa-check"></i><b>8.6</b> Putting R functions in Scripts</a></li>
<li class="chapter" data-level="8.7" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#simple-statistics-in-vector-form"><i class="fa fa-check"></i><b>8.7</b> Simple statistics in vector form</a><ul>
<li class="chapter" data-level="8.7.1" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#mean"><i class="fa fa-check"></i><b>8.7.1</b> Mean</a></li>
<li class="chapter" data-level="8.7.2" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#variance-and-standard-deviation"><i class="fa fa-check"></i><b>8.7.2</b> Variance and standard deviation</a></li>
<li class="chapter" data-level="8.7.3" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#covariance-and-correlation"><i class="fa fa-check"></i><b>8.7.3</b> Covariance and correlation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html"><i class="fa fa-check"></i><b>9</b> Bivariate Linear Regression</a><ul>
<li class="chapter" data-level="9.1" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#regression-terminology"><i class="fa fa-check"></i><b>9.1</b> Regression terminology</a></li>
<li class="chapter" data-level="9.2" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#libraries-3"><i class="fa fa-check"></i><b>9.2</b> Libraries</a></li>
<li class="chapter" data-level="9.3" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#illustrating-linear-regression-with-simulated-data"><i class="fa fa-check"></i><b>9.3</b> Illustrating linear regression with simulated data</a></li>
<li class="chapter" data-level="9.4" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#specifying-regression-models-in-r"><i class="fa fa-check"></i><b>9.4</b> Specifying Regression Models in R</a><ul>
<li class="chapter" data-level="9.4.1" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#fitted-values"><i class="fa fa-check"></i><b>9.4.1</b> Fitted values</a></li>
<li class="chapter" data-level="9.4.2" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#getting-the-model-coefficients"><i class="fa fa-check"></i><b>9.4.2</b> Getting the model coefficients</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#residuals"><i class="fa fa-check"></i><b>9.5</b> Residuals</a></li>
<li class="chapter" data-level="9.6" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#regression-as-sum-of-squares-decomposition"><i class="fa fa-check"></i><b>9.6</b> Regression as sum-of-squares decomposition</a></li>
<li class="chapter" data-level="9.7" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#variance-explained-by-a-regression-model"><i class="fa fa-check"></i><b>9.7</b> Variance “explained” by a regression model</a></li>
<li class="chapter" data-level="9.8" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broom-a-library-for-converting-model-results-into-data-frames"><i class="fa fa-check"></i><b>9.8</b> Broom: a library for converting model results into data frames</a><ul>
<li class="chapter" data-level="9.8.1" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broomtidy"><i class="fa fa-check"></i><b>9.8.1</b> <code>broom::tidy</code></a></li>
<li class="chapter" data-level="9.8.2" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broomaugment"><i class="fa fa-check"></i><b>9.8.2</b> <code>broom::augment</code></a></li>
<li class="chapter" data-level="9.8.3" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broomglance"><i class="fa fa-check"></i><b>9.8.3</b> <code>broom::glance</code></a></li>
</ul></li>
<li class="chapter" data-level="9.9" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#bivariate-linear-regression-on-a-real-world-data-set"><i class="fa fa-check"></i><b>9.9</b> Bivariate linear regression on a real world data set</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="matrices-in-r.html"><a href="matrices-in-r.html"><i class="fa fa-check"></i><b>10</b> Matrices in R</a><ul>
<li class="chapter" data-level="10.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#creating-matrices-in-r"><i class="fa fa-check"></i><b>10.1</b> Creating matrices in R</a></li>
<li class="chapter" data-level="10.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-arithmetic-operations-in-r"><i class="fa fa-check"></i><b>10.2</b> Matrix arithmetic operations in R</a></li>
<li class="chapter" data-level="10.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#descriptive-statistics-as-matrix-functions"><i class="fa fa-check"></i><b>10.3</b> Descriptive statistics as matrix functions</a><ul>
<li class="chapter" data-level="10.3.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#mean-vector-and-matrix"><i class="fa fa-check"></i><b>10.3.1</b> Mean vector and matrix</a></li>
<li class="chapter" data-level="10.3.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#deviation-matrix"><i class="fa fa-check"></i><b>10.3.2</b> Deviation matrix</a></li>
<li class="chapter" data-level="10.3.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#covariance-matrix"><i class="fa fa-check"></i><b>10.3.3</b> Covariance matrix</a></li>
<li class="chapter" data-level="10.3.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#correlation-matrix"><i class="fa fa-check"></i><b>10.3.4</b> Correlation matrix</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-inverse"><i class="fa fa-check"></i><b>10.4</b> Matrix Inverse</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="multiple-regression.html"><a href="multiple-regression.html"><i class="fa fa-check"></i><b>11</b> Multiple Regression</a><ul>
<li class="chapter" data-level="11.1" data-path="multiple-regression.html"><a href="multiple-regression.html#new-libraries-to-install"><i class="fa fa-check"></i><b>11.1</b> New Libraries to install</a></li>
<li class="chapter" data-level="11.2" data-path="multiple-regression.html"><a href="multiple-regression.html#libraries-4"><i class="fa fa-check"></i><b>11.2</b> Libraries</a></li>
<li class="chapter" data-level="11.3" data-path="multiple-regression.html"><a href="multiple-regression.html#examplar-data"><i class="fa fa-check"></i><b>11.3</b> Examplar data</a></li>
<li class="chapter" data-level="11.4" data-path="multiple-regression.html"><a href="multiple-regression.html#data-exploration"><i class="fa fa-check"></i><b>11.4</b> Data exploration</a></li>
<li class="chapter" data-level="11.5" data-path="multiple-regression.html"><a href="multiple-regression.html#d-plots"><i class="fa fa-check"></i><b>11.5</b> 3D Plots</a><ul>
<li class="chapter" data-level="11.5.1" data-path="multiple-regression.html"><a href="multiple-regression.html#scatterplot3d"><i class="fa fa-check"></i><b>11.5.1</b> scatterplot3d</a></li>
<li class="chapter" data-level="11.5.2" data-path="multiple-regression.html"><a href="multiple-regression.html#rgl"><i class="fa fa-check"></i><b>11.5.2</b> rgl</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="multiple-regression.html"><a href="multiple-regression.html#fitting-the-regression-model"><i class="fa fa-check"></i><b>11.6</b> Fitting the regression model</a><ul>
<li class="chapter" data-level="11.6.1" data-path="multiple-regression.html"><a href="multiple-regression.html#visualizing-the-regression-model-in-scatterplot3d"><i class="fa fa-check"></i><b>11.6.1</b> Visualizing the regression model in scatterplot3d</a></li>
</ul></li>
<li class="chapter" data-level="11.7" data-path="multiple-regression.html"><a href="multiple-regression.html#interpretting-the-regression-model"><i class="fa fa-check"></i><b>11.7</b> Interpretting the regression model</a></li>
<li class="chapter" data-level="11.8" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-vector-geometry-of-the-regression-model"><i class="fa fa-check"></i><b>11.8</b> Exploring the Vector Geometry of the Regression Model</a></li>
<li class="chapter" data-level="11.9" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-residuals-from-the-model-fit"><i class="fa fa-check"></i><b>11.9</b> Exploring the Residuals from the Model Fit</a></li>
<li class="chapter" data-level="11.10" data-path="multiple-regression.html"><a href="multiple-regression.html#an-alternate-model"><i class="fa fa-check"></i><b>11.10</b> An alternate model</a></li>
<li class="chapter" data-level="11.11" data-path="multiple-regression.html"><a href="multiple-regression.html#fitting-a-curvilinear-model-using-lm"><i class="fa fa-check"></i><b>11.11</b> Fitting a curvilinear model using lm()</a></li>
<li class="chapter" data-level="11.12" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-impact-of-nearly-collinear-predictors-on-regression"><i class="fa fa-check"></i><b>11.12</b> Exploring the impact of nearly collinear predictors on regression</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html"><i class="fa fa-check"></i><b>12</b> Principal Components Analysis</a><ul>
<li class="chapter" data-level="12.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#libraries-5"><i class="fa fa-check"></i><b>12.1</b> Libraries</a></li>
<li class="chapter" data-level="12.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#matrices-as-linear-transformations"><i class="fa fa-check"></i><b>12.2</b> Matrices as linear transformations</a></li>
<li class="chapter" data-level="12.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#eigenanalysis-in-r"><i class="fa fa-check"></i><b>12.3</b> Eigenanalysis in R</a></li>
<li class="chapter" data-level="12.4" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#principal-components-analysis-in-r"><i class="fa fa-check"></i><b>12.4</b> Principal Components Analysis in R</a><ul>
<li class="chapter" data-level="12.4.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#bioenv-dataset"><i class="fa fa-check"></i><b>12.4.1</b> Bioenv dataset</a></li>
<li class="chapter" data-level="12.4.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pca-of-the-bioenv-dataset"><i class="fa fa-check"></i><b>12.4.2</b> PCA of the Bioenv dataset</a></li>
<li class="chapter" data-level="12.4.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#calculating-factor-loadings"><i class="fa fa-check"></i><b>12.4.3</b> Calculating Factor Loadings</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#drawing-figures-to-represent-pca"><i class="fa fa-check"></i><b>12.5</b> Drawing Figures to Represent PCA</a><ul>
<li class="chapter" data-level="12.5.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pc-score-plots"><i class="fa fa-check"></i><b>12.5.1</b> PC Score Plots</a></li>
<li class="chapter" data-level="12.5.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#simultaneous-depiction-of-observations-and-variables-in-the-pc-space"><i class="fa fa-check"></i><b>12.5.2</b> Simultaneous Depiction of Observations and Variables in the PC Space</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html"><i class="fa fa-check"></i><b>13</b> Simulating Sampling Distributions and Confidence Intervals</a><ul>
<li class="chapter" data-level="13.1" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#libraries-6"><i class="fa fa-check"></i><b>13.1</b> Libraries</a></li>
<li class="chapter" data-level="13.2" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#data-set-simulated-male-heights"><i class="fa fa-check"></i><b>13.2</b> Data set: Simulated male heights</a><ul>
<li class="chapter" data-level="13.2.1" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#seeding-the-pseudo-random-number-generator"><i class="fa fa-check"></i><b>13.2.1</b> Seeding the pseudo-random number generator</a></li>
<li class="chapter" data-level="13.2.2" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#generating-the-simulated-population"><i class="fa fa-check"></i><b>13.2.2</b> Generating the simulated population</a></li>
<li class="chapter" data-level="13.2.3" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#properties-of-the-simulated-population"><i class="fa fa-check"></i><b>13.2.3</b> Properties of the simulated population</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#random-sampling-from-the-simulated-population"><i class="fa fa-check"></i><b>13.3</b> Random sampling from the simulated population</a><ul>
<li class="chapter" data-level="13.3.1" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#another-random-sample"><i class="fa fa-check"></i><b>13.3.1</b> Another random sample</a></li>
<li class="chapter" data-level="13.3.2" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#simulating-the-generation-of-many-random-samples"><i class="fa fa-check"></i><b>13.3.2</b> Simulating the generation of many random samples</a></li>
<li class="chapter" data-level="13.3.3" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#a-function-to-estimate-statistics-of-interest-in-a-random-sample"><i class="fa fa-check"></i><b>13.3.3</b> A function to estimate statistics of interest in a random sample</a></li>
<li class="chapter" data-level="13.3.4" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#generating-statistics-for-many-random-samples"><i class="fa fa-check"></i><b>13.3.4</b> Generating statistics for many random samples</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#simulated-sampling-distribution-of-the-mean"><i class="fa fa-check"></i><b>13.4</b> Simulated sampling distribution of the mean</a><ul>
<li class="chapter" data-level="13.4.1" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#sampling-distributions-for-different-sample-sizes"><i class="fa fa-check"></i><b>13.4.1</b> Sampling distributions for different sample sizes</a></li>
<li class="chapter" data-level="13.4.2" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#discussion-of-trends-for-sampling-distributions-of-different-sizes"><i class="fa fa-check"></i><b>13.4.2</b> Discussion of trends for sampling distributions of different sizes</a></li>
</ul></li>
<li class="chapter" data-level="13.5" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#standard-error-of-the-mean"><i class="fa fa-check"></i><b>13.5</b> Standard Error of the Mean</a></li>
<li class="chapter" data-level="13.6" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#interpretting-trends-for-the-sampling-distribution-of-the-mean"><i class="fa fa-check"></i><b>13.6</b> Interpretting Trends for the Sampling Distribution of the Mean</a></li>
<li class="chapter" data-level="13.7" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#sampling-distribution-of-the-standard-deviation"><i class="fa fa-check"></i><b>13.7</b> Sampling Distribution of the Standard Deviation</a></li>
<li class="chapter" data-level="13.8" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#confidence-intervals"><i class="fa fa-check"></i><b>13.8</b> Confidence Intervals</a></li>
<li class="chapter" data-level="13.9" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#defining-confidence-intervals"><i class="fa fa-check"></i><b>13.9</b> Defining confidence intervals</a></li>
<li class="chapter" data-level="13.10" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#simulating-confidence-intervals-for-the-mean"><i class="fa fa-check"></i><b>13.10</b> Simulating confidence intervals for the mean</a></li>
<li class="chapter" data-level="13.11" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#sample-estimate-of-the-standard-error-of-the-mean"><i class="fa fa-check"></i><b>13.11</b> Sample estimate of the standard error of the mean</a></li>
<li class="chapter" data-level="13.12" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#ci-of-mean-simulation"><i class="fa fa-check"></i><b>13.12</b> CI of mean simulation</a></li>
<li class="chapter" data-level="13.13" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#generating-a-table-of-cis-and-corresponding-margins-of-error"><i class="fa fa-check"></i><b>13.13</b> Generating a table of CIs and corresponding margins of error</a></li>
<li class="chapter" data-level="13.14" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#conditions-for-sampling-distribution-of-the-mean-to-be-nearly-normal"><i class="fa fa-check"></i><b>13.14</b> Conditions for sampling distribution of the mean to be nearly normal</a></li>
<li class="chapter" data-level="13.15" data-path="simulating-sampling-distributions-and-confidence-intervals.html"><a href="simulating-sampling-distributions-and-confidence-intervals.html#interpreting-our-simulation"><i class="fa fa-check"></i><b>13.15</b> Interpreting our simulation</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Biology 723: Statistical Computing for Biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="multiple-regression" class="section level1">
<h1><span class="header-section-number">Chapter 11</span> Multiple Regression</h1>
<div id="new-libraries-to-install" class="section level2">
<h2><span class="header-section-number">11.1</span> New Libraries to install</h2>
<p>We’ll be using several new packages for this class session. Install the following packages via one of the standard install mechanisms:</p>
<ul>
<li><code>scatterplot3d</code></li>
<li><code>rgl</code> – NOTE: On OS X, <code>rgl</code> requires you to install a program called XQuartz. XQuartz can be downloaded from the <a href="https://www.xquartz.org">XQuartz Home Page</a>. If you’re on a Mac, install XQuartz before installing <code>rgl</code>. You may have to reboot your computer after installing XQuartz.</li>
</ul>
</div>
<div id="libraries-4" class="section level2">
<h2><span class="header-section-number">11.2</span> Libraries</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)
<span class="kw">library</span>(broom)  <span class="co"># for working w/lm output</span></code></pre></div>
</div>
<div id="examplar-data" class="section level2">
<h2><span class="header-section-number">11.3</span> Examplar data</h2>
<p>To illustrate multiple regression in R we’ll use a built in dataset called <code>trees</code>. <code>trees</code> consists of measurements of the girth, height, and volume of 31 black cherry trees (<code>?trees</code> for more info).</p>
<p>Let’s assume we’re lumberjacks, but our permit only allows us to harvest a fixed number of trees. We get paid by the total volume of wood we harvest, so we’re interested in predicting a tree’s volume (hard to measure directly) as a function of its girth and height (relatively easy to measure), so we can pick the best trees to harvest. We’ll therefore calculate a multiple regression of volume on height and width.</p>
</div>
<div id="data-exploration" class="section level2">
<h2><span class="header-section-number">11.4</span> Data exploration</h2>
<p>We’ll start with some summary tables and diagnostic plots to familiarize ourselves with the data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(trees)
[<span class="dv">1</span>] <span class="st">&quot;Girth&quot;</span>  <span class="st">&quot;Height&quot;</span> <span class="st">&quot;Volume&quot;</span>

<span class="kw">dim</span>(trees)
[<span class="dv">1</span>] <span class="dv">31</span>  <span class="dv">3</span>

<span class="kw">summary</span>(trees)
     Girth          Height       Volume    
 Min.   <span class="op">:</span><span class="st"> </span><span class="fl">8.3</span>   Min.   <span class="op">:</span><span class="dv">63</span>   Min.   <span class="op">:</span><span class="fl">10.2</span>  
 1st Qu.<span class="op">:</span><span class="fl">11.1</span>   1st Qu.<span class="op">:</span><span class="dv">72</span>   1st Qu.<span class="op">:</span><span class="fl">19.4</span>  
 Median <span class="op">:</span><span class="fl">12.9</span>   Median <span class="op">:</span><span class="dv">76</span>   Median <span class="op">:</span><span class="fl">24.2</span>  
 Mean   <span class="op">:</span><span class="fl">13.2</span>   Mean   <span class="op">:</span><span class="dv">76</span>   Mean   <span class="op">:</span><span class="fl">30.2</span>  
 3rd Qu.<span class="op">:</span><span class="fl">15.2</span>   3rd Qu.<span class="op">:</span><span class="dv">80</span>   3rd Qu.<span class="op">:</span><span class="fl">37.3</span>  
 Max.   <span class="op">:</span><span class="fl">20.6</span>   Max.   <span class="op">:</span><span class="dv">87</span>   Max.   <span class="op">:</span><span class="fl">77.0</span>  </code></pre></div>
<p>We’ll use the <code>GGally::ggpairs()</code> function introduced in problem set 01 to create a scatterplot matrix depicting the pairwise relationships between all the variables</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GGally)
<span class="kw">ggpairs</span>(trees)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-316-1.png" width="576" /></p>
<p>As one might expect for morphological measurements related to size, the scatterplot matrix shows that all the variables are positively correlated, and girth and volume have a particularly strong correlation.</p>
</div>
<div id="d-plots" class="section level2">
<h2><span class="header-section-number">11.5</span> 3D Plots</h2>
<p>ggplot has no built in facilities for 3D scatter plots so we’ll use two new packages, <a href="https://cran.r-project.org/package=scatterplot3d"><code>scatterplot3D</code></a> and <a href="https://cran.r-project.org/web/packages/rgl/vignettes/rgl.html"><code>rgl</code></a>, to generate 3D visualizations.</p>
<div id="scatterplot3d" class="section level3">
<h3><span class="header-section-number">11.5.1</span> scatterplot3d</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scatterplot3d)  <span class="co"># install this package first if needed</span>
<span class="kw">scatterplot3d</span>(trees,
              <span class="dt">main =</span> <span class="st">&#39;Tree Volume as</span><span class="ch">\n</span><span class="st">a function of Girth and Height&#39;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-317-1.png" width="576" /></p>
<p>The argument <code>pch</code> sets the type of plotting character to use in the plot (for a graphical key of the available plotting characters see <a href="https://www.statmethods.net/advgraphs/parameters.html">this link</a>) and <code>color</code> sets plotting character colors. We can change the angle of the 3D plot using the <code>angle</code> argument:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">scatterplot3d</span>(trees, 
              <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">color=</span><span class="st">&quot;steelblue&quot;</span>,
              <span class="dt">angle=</span><span class="dv">75</span>,
              <span class="dt">main =</span> <span class="st">&#39;Tree Volume as</span><span class="ch">\n</span><span class="st">a function of Girth and Height&#39;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-318-1.png" width="576" /></p>
<p>We can add vertical lines to the plot using the <code>type</code> argument and remove the box around the plot:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">scatterplot3d</span>(trees, 
              <span class="dt">pch =</span> <span class="dv">16</span>, 
              <span class="dt">color=</span><span class="st">&quot;steelblue&quot;</span>,
              <span class="dt">angle=</span><span class="dv">75</span>,
              <span class="dt">box =</span> <span class="ot">FALSE</span>,
              <span class="dt">type =</span> <span class="st">&quot;h&quot;</span>,
              <span class="dt">main =</span> <span class="st">&#39;Tree Volume as</span><span class="ch">\n</span><span class="st">a function of Girth and Height&#39;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-319-1.png" width="576" /></p>
<p>For more examples of how you can modify plots generated with the <code>scatterplot3d</code> package see <a href="http://www.sthda.com/english/wiki/scatterplot3d-3d-graphics-r-software-and-data-visualization#change-the-global-appearance-of-the-graph">this web page</a>).</p>
</div>
<div id="rgl" class="section level3">
<h3><span class="header-section-number">11.5.2</span> rgl</h3>
<p>The package <code>rgl</code> is another package that we can use for 3D visualization. <code>rgl</code> is powerful because it lets us create interactive plots we can rotate and zoom in/out on.</p>
<p>To make rgl plots work in HTML document, make your codeblock start with this header:</p>
<pre><code>{r, results = &quot;asis&quot;}</code></pre>
<p>You can then create an interactive 3D plot as so:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(rgl)

<span class="co"># create 3D scatter, using spheres to draw points</span>
<span class="kw">plot3d</span>(trees<span class="op">$</span>Girth, trees<span class="op">$</span>Height, trees<span class="op">$</span>Volume,
      <span class="dt">xlab =</span> <span class="st">&quot;Girth&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Height&quot;</span>, <span class="dt">zlab =</span> <span class="st">&quot;Volume&quot;</span>,
       <span class="dt">type =</span> <span class="st">&quot;s&quot;</span>, 
       <span class="dt">size =</span> <span class="fl">1.5</span>,
       <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>)

<span class="kw">rglwidget</span>() <span class="co"># only need to include this line if using in a markdown document</span></code></pre></div>
</div>
</div>
<div id="fitting-the-regression-model" class="section level2">
<h2><span class="header-section-number">11.6</span> Fitting the regression model</h2>
<p>From the 3D scatter plot it looks like we ought to be able to find a plane through the data that fits the scatter fairly well.</p>
<p>Let’s use the <code>lm()</code> function to calculate the multiple regression and <code>summary()</code> to get the details of the model:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit.trees &lt;-<span class="st"> </span><span class="kw">lm</span>(Volume <span class="op">~</span><span class="st"> </span>Girth <span class="op">+</span><span class="st"> </span>Height, <span class="dt">data=</span>trees)

<span class="kw">summary</span>(fit.trees)

Call<span class="op">:</span>
<span class="kw">lm</span>(<span class="dt">formula =</span> Volume <span class="op">~</span><span class="st"> </span>Girth <span class="op">+</span><span class="st"> </span>Height, <span class="dt">data =</span> trees)

Residuals<span class="op">:</span>
<span class="st">   </span>Min     1Q Median     3Q    Max 
<span class="op">-</span><span class="fl">6.406</span> <span class="op">-</span><span class="fl">2.649</span> <span class="op">-</span><span class="fl">0.288</span>  <span class="fl">2.200</span>  <span class="fl">8.485</span> 

Coefficients<span class="op">:</span>
<span class="st">            </span>Estimate Std. Error t value <span class="kw">Pr</span>(<span class="op">&gt;</span><span class="er">|</span>t<span class="op">|</span>)    
(Intercept)  <span class="op">-</span><span class="fl">57.988</span>      <span class="fl">8.638</span>   <span class="op">-</span><span class="fl">6.71</span>  <span class="fl">2.7e-07</span> <span class="op">**</span><span class="er">*</span>
Girth          <span class="fl">4.708</span>      <span class="fl">0.264</span>   <span class="fl">17.82</span>  <span class="op">&lt;</span><span class="st"> </span><span class="fl">2e-16</span> <span class="op">**</span><span class="er">*</span>
Height         <span class="fl">0.339</span>      <span class="fl">0.130</span>    <span class="fl">2.61</span>    <span class="fl">0.014</span> <span class="op">*</span><span class="st">  </span>
<span class="op">---</span>
Signif. codes<span class="op">:</span><span class="st">  </span><span class="dv">0</span> <span class="st">&#39;***&#39;</span> <span class="fl">0.001</span> <span class="st">&#39;**&#39;</span> <span class="fl">0.01</span> <span class="st">&#39;*&#39;</span> <span class="fl">0.05</span> <span class="st">&#39;.&#39;</span> <span class="fl">0.1</span> <span class="st">&#39; &#39;</span> <span class="dv">1</span>

Residual standard error<span class="op">:</span><span class="st"> </span><span class="fl">3.88</span> on <span class="dv">28</span> degrees of freedom
Multiple R<span class="op">-</span>squared<span class="op">:</span><span class="st">  </span><span class="fl">0.948</span>, Adjusted R<span class="op">-</span>squared<span class="op">:</span><span class="st">  </span><span class="fl">0.944</span> 
F<span class="op">-</span>statistic<span class="op">:</span><span class="st">  </span><span class="dv">255</span> on <span class="dv">2</span> and <span class="dv">28</span> DF,  p<span class="op">-</span>value<span class="op">:</span><span class="st"> </span><span class="er">&lt;</span><span class="fl">2e-16</span></code></pre></div>
<div id="visualizing-the-regression-model-in-scatterplot3d" class="section level3">
<h3><span class="header-section-number">11.6.1</span> Visualizing the regression model in scatterplot3d</h3>
<p>To visualize the multiple regression, let’s use the <code>scatterplot3d</code> package to draw the 3D scatter of plots and the plane that corresponds to the regression model:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">scatterplot3d</span>(trees, <span class="dt">angle=</span><span class="dv">55</span>,<span class="dt">type=</span><span class="st">&#39;h&#39;</span>,
                   <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">color =</span> <span class="st">&quot;steelblue&quot;</span>,
                   <span class="dt">main =</span> <span class="st">&#39;Tree Volume as</span><span class="ch">\n</span><span class="st">a function of Girth and Height&#39;</span>)

<span class="co"># add a plane representing the fit of the model</span>
p<span class="op">$</span><span class="kw">plane3d</span>(fit.trees, <span class="dt">col=</span><span class="st">&#39;orangered&#39;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-322-1.png" width="576" /></p>
<p>If instead of <code>scatterplot3d</code>, we wanted to use <code>rgl</code> to depict the model fit we can use the <code>rgl.planes</code> function as shown below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">coefs &lt;-<span class="st"> </span><span class="kw">coef</span>(fit.trees)
b1 &lt;-<span class="st"> </span>coefs[<span class="st">&quot;Girth&quot;</span>]
b2 &lt;-<span class="st"> </span>coefs[<span class="st">&quot;Height&quot;</span>]
c &lt;-<span class="st"> </span><span class="op">-</span><span class="dv">1</span>
a &lt;-<span class="st"> </span>coefs[<span class="st">&quot;(Intercept)&quot;</span>]
<span class="kw">plot3d</span>(trees<span class="op">$</span>Girth, trees<span class="op">$</span>Height, trees<span class="op">$</span>Volume,
      <span class="dt">xlab =</span> <span class="st">&quot;Girth&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;Height&quot;</span>, <span class="dt">zlab =</span> <span class="st">&quot;Volume&quot;</span>, 
       <span class="dt">type =</span> <span class="st">&quot;s&quot;</span>, 
       <span class="dt">size =</span> <span class="fl">1.5</span>,
       <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>)
<span class="kw">rgl.planes</span>(b1, b2, c, a, <span class="dt">alpha =</span> <span class="fl">0.9</span>, <span class="dt">color =</span> <span class="st">&quot;gray&quot;</span>)
<span class="kw">rglwidget</span>()</code></pre></div>
<p>From the figures it looks like the regression model fits pretty well, as we anticipated from the pairwise relationships.</p>
</div>
</div>
<div id="interpretting-the-regression-model" class="section level2">
<h2><span class="header-section-number">11.7</span> Interpretting the regression model</h2>
<p>The regression equation is: <span class="math inline">\(\hat{y}\)</span> = + <span class="math inline">\(x_1\)</span> +<span class="math inline">\(x_2\)</span>, where <span class="math inline">\(y\)</span> is Volume, and <span class="math inline">\(x_1\)</span> and <span class="math inline">\(x_2\)</span> are Girth and Height respectively. Since they’re on different scales the coefficients for Girth and Height aren’t directly comparable. Both coefficients are significant at the <span class="math inline">\(p&lt;0.05\)</span> level, but note that Girth is the much stronger predictor. In fact the addition of height explains only a minor additional fraction of variation in tree volume, so from the lumberjack’s perspective the additional trouble of measuring height probably isn’t worth it.</p>
</div>
<div id="exploring-the-vector-geometry-of-the-regression-model" class="section level2">
<h2><span class="header-section-number">11.8</span> Exploring the Vector Geometry of the Regression Model</h2>
<p>The object returned by the <code>lm()</code> function hold lots of useful information:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(fit.trees)
 [<span class="dv">1</span>] <span class="st">&quot;coefficients&quot;</span>  <span class="st">&quot;residuals&quot;</span>     <span class="st">&quot;effects&quot;</span>       <span class="st">&quot;rank&quot;</span>         
 [<span class="dv">5</span>] <span class="st">&quot;fitted.values&quot;</span> <span class="st">&quot;assign&quot;</span>        <span class="st">&quot;qr&quot;</span>            <span class="st">&quot;df.residual&quot;</span>  
 [<span class="dv">9</span>] <span class="st">&quot;xlevels&quot;</span>       <span class="st">&quot;call&quot;</span>          <span class="st">&quot;terms&quot;</span>         <span class="st">&quot;model&quot;</span>        </code></pre></div>
<p>The <code>fitted.values</code> correspond to the predicted values of the outcome variable (<span class="math inline">\(\hat{y}\)</span>). Alternate we can get this information in useful table form using functions from the <code>broom</code> library:</p>
<p>Recall the <code>broom:tidy</code> produces a tabular summary of the coefficients of the model and their associated statistics:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">broom<span class="op">::</span><span class="kw">tidy</span>(fit.trees)
         term estimate std.error statistic   p.value
<span class="dv">1</span> (Intercept) <span class="op">-</span><span class="fl">57.9877</span>    <span class="fl">8.6382</span>    <span class="op">-</span><span class="fl">6.713</span> <span class="fl">2.750e-07</span>
<span class="dv">2</span>       Girth   <span class="fl">4.7082</span>    <span class="fl">0.2643</span>    <span class="fl">17.816</span> <span class="fl">8.223e-17</span>
<span class="dv">3</span>      Height   <span class="fl">0.3393</span>    <span class="fl">0.1302</span>     <span class="fl">2.607</span> <span class="fl">1.449e-02</span></code></pre></div>
<p><code>broom:glance</code> provides information about the fit of the model:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">broom<span class="op">::</span><span class="kw">glance</span>(fit.trees)
  r.squared adj.r.squared sigma statistic   p.value df logLik   AIC   BIC
<span class="dv">1</span>     <span class="fl">0.948</span>        <span class="fl">0.9442</span> <span class="fl">3.882</span>       <span class="dv">255</span> <span class="fl">1.071e-18</span>  <span class="dv">3</span> <span class="op">-</span><span class="fl">84.45</span> <span class="fl">176.9</span> <span class="fl">182.6</span>
  deviance df.residual
<span class="dv">1</span>    <span class="fl">421.9</span>          <span class="dv">28</span></code></pre></div>
<p>Let’s use our knowledge of vector geometry to further explore the relationship between the predicted Volume and the predictor variables. By definition the vector representing the predicted values lies in the subspace (in this case a plane) defined by Height and Girth, so let’s do some simple calculations to understand their length and angular relationships:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># proportional to length of vectors</span>
<span class="kw">sd</span>(fit.trees<span class="op">$</span>fitted.values)
[<span class="dv">1</span>] <span class="dv">16</span>
<span class="kw">sd</span>(trees<span class="op">$</span>Height)
[<span class="dv">1</span>] <span class="fl">6.372</span>
<span class="kw">sd</span>(trees<span class="op">$</span>Girth)
[<span class="dv">1</span>] <span class="fl">3.138</span>

<span class="co"># cosines of angles btw vectors</span>
<span class="kw">cor</span>(trees<span class="op">$</span>Height, trees<span class="op">$</span>Girth)
[<span class="dv">1</span>] <span class="fl">0.5193</span>
<span class="kw">cor</span>(trees<span class="op">$</span>Girth, fit.trees<span class="op">$</span>fitted.values)
[<span class="dv">1</span>] <span class="fl">0.9933</span>
<span class="kw">cor</span>(trees<span class="op">$</span>Height, fit.trees<span class="op">$</span>fitted.values)
[<span class="dv">1</span>] <span class="fl">0.6145</span>

<span class="co"># angles btw vectors in degrees</span>
<span class="kw">acos</span>(<span class="kw">cor</span>(trees<span class="op">$</span>Girth, trees<span class="op">$</span>Height)) <span class="op">*</span><span class="st"> </span>(<span class="dv">180</span><span class="op">/</span>pi)
[<span class="dv">1</span>] <span class="fl">58.72</span>
<span class="kw">acos</span>(<span class="kw">cor</span>(trees<span class="op">$</span>Girth, fit.trees<span class="op">$</span>fitted.values)) <span class="op">*</span><span class="st"> </span>(<span class="dv">180</span><span class="op">/</span>pi)
[<span class="dv">1</span>] <span class="fl">6.628</span>
<span class="kw">acos</span>(<span class="kw">cor</span>(trees<span class="op">$</span>Height, fit.trees<span class="op">$</span>fitted.values)) <span class="op">*</span><span class="st"> </span>(<span class="dv">180</span><span class="op">/</span>pi)
[<span class="dv">1</span>] <span class="fl">52.09</span></code></pre></div>
</div>
<div id="exploring-the-residuals-from-the-model-fit" class="section level2">
<h2><span class="header-section-number">11.9</span> Exploring the Residuals from the Model Fit</h2>
<p>Now let’s look at the residuals from the regression. The residuals represent the `unexplained’ variance:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">trees.augmented &lt;-<span class="st"> </span><span class="kw">augment</span>(fit.trees, trees)

<span class="kw">ggplot</span>(trees.augmented, <span class="kw">aes</span>(<span class="dt">x =</span> Girth, <span class="dt">y =</span> .resid)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="dv">0</span>, <span class="dt">color=</span><span class="st">&#39;red&#39;</span>, <span class="dt">linetype=</span><span class="st">&#39;dashed&#39;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Girth&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Residuals&quot;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-328-1.png" width="576" /></p>
<p>Ideally the residuals should be evenly scattered around zero, with no trends as we go from high to low values of the dependent variable. As you can see, the residuals are somewhat u-shaped or j-shaped suggesting that there may be a non-linear aspect of the relationship that our model isn’t capturing.</p>
</div>
<div id="an-alternate-model" class="section level2">
<h2><span class="header-section-number">11.10</span> An alternate model</h2>
<p>Let’s think about the relationships we’re actually modeling for a few minutes. For the sake of simplicity let’s consider the trunk of a tree to be a cylinder. How do the dimensions of this cylinder relate to its volume? You can look up the formula for the volume of a cylinder, but the key thing you’ll want to note is that volume of the cylinder should be proportional to a characteristic length of the cylinder cubed (<span class="math inline">\(V \propto \mathrm{L}^3\)</span>). This suggests that if we want to fit a linear model we should relate Girth and Height to <span class="math inline">\(\sqrt[3]{\mathrm{Volume}}\)</span>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">trees.cuberoot &lt;-<span class="st"> </span><span class="kw">mutate</span>(trees, <span class="dt">cuberoot.Volume =</span> Volume<span class="op">^</span><span class="fl">0.33</span>)

fit.trees.cuberoot &lt;-<span class="st"> </span><span class="kw">lm</span>(cuberoot.Volume <span class="op">~</span><span class="st"> </span>Girth <span class="op">+</span><span class="st"> </span>Height, <span class="dt">data =</span> trees.cuberoot)

broom<span class="op">::</span><span class="kw">glance</span>(fit.trees) <span class="co"># summary of fit of original model</span>
  r.squared adj.r.squared sigma statistic   p.value df logLik   AIC   BIC
<span class="dv">1</span>     <span class="fl">0.948</span>        <span class="fl">0.9442</span> <span class="fl">3.882</span>       <span class="dv">255</span> <span class="fl">1.071e-18</span>  <span class="dv">3</span> <span class="op">-</span><span class="fl">84.45</span> <span class="fl">176.9</span> <span class="fl">182.6</span>
  deviance df.residual
<span class="dv">1</span>    <span class="fl">421.9</span>          <span class="dv">28</span>

broom<span class="op">::</span><span class="kw">glance</span>(fit.trees.cuberoot)  <span class="co"># summary of fit of alternate model</span>
  r.squared adj.r.squared   sigma statistic   p.value df logLik    AIC
<span class="dv">1</span>    <span class="fl">0.9776</span>        <span class="fl">0.9761</span> <span class="fl">0.08108</span>     <span class="fl">612.4</span> <span class="fl">7.768e-24</span>  <span class="dv">3</span>  <span class="fl">35.47</span> <span class="op">-</span><span class="fl">62.94</span>
     BIC deviance df.residual
<span class="dv">1</span> <span class="op">-</span><span class="fl">57.21</span>   <span class="fl">0.1841</span>          <span class="dv">28</span></code></pre></div>
<p>Comparing the summary tables, we see indeed that using the cube root of Volume improves the fit of our model some. Let’s examine the residuals of this alternate model.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">trees.cuberoot &lt;-<span class="st"> </span>broom<span class="op">::</span><span class="kw">augment</span>(fit.trees.cuberoot, trees.cuberoot)

<span class="kw">ggplot</span>(trees.cuberoot, <span class="kw">aes</span>(<span class="dt">x =</span> cuberoot.Volume, <span class="dt">y =</span> .resid)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="dv">0</span>, <span class="dt">color=</span><span class="st">&#39;red&#39;</span>, <span class="dt">linetype=</span><span class="st">&#39;dashed&#39;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Girth&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Residuals&quot;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-330-1.png" width="576" /></p>
<p>As we can see the transformation we applied to the data did seem to make our residuals more uniform across the range of observations.</p>
</div>
<div id="fitting-a-curvilinear-model-using-lm" class="section level2">
<h2><span class="header-section-number">11.11</span> Fitting a curvilinear model using lm()</h2>
<p>Above we transformed the volume data in order to fit a straight line relationship between <span class="math inline">\(\sqrt[3]{V}\)</span> and Girth and Hieght. However, we could just as easily have applied a cubic regression to the original variables (remember this is still linear in the coefficients). Since Height didn’t add much to additional information, we’ll simplify the model to consider only Girth.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit.curvilinear &lt;-<span class="st"> </span><span class="kw">lm</span>(Volume <span class="op">~</span><span class="st"> </span><span class="kw">I</span>(Girth<span class="op">^</span><span class="dv">3</span>), <span class="dt">data=</span>trees)

broom<span class="op">::</span><span class="kw">tidy</span>(fit.curvilinear)
         term estimate std.error statistic   p.value
<span class="dv">1</span> (Intercept) <span class="fl">8.042696</span> <span class="fl">1.0426698</span>     <span class="fl">7.714</span> <span class="fl">1.663e-08</span>
<span class="dv">2</span>  <span class="kw">I</span>(Girth<span class="op">^</span><span class="dv">3</span>) <span class="fl">0.008137</span> <span class="fl">0.0003118</span>    <span class="fl">26.098</span> <span class="fl">1.087e-21</span>

broom<span class="op">::</span><span class="kw">glance</span>(fit.curvilinear)
  r.squared adj.r.squared sigma statistic   p.value df logLik   AIC   BIC
<span class="dv">1</span>    <span class="fl">0.9592</span>        <span class="fl">0.9578</span> <span class="fl">3.379</span>     <span class="fl">681.1</span> <span class="fl">1.087e-21</span>  <span class="dv">2</span>  <span class="op">-</span><span class="fl">80.7</span> <span class="fl">167.4</span> <span class="fl">171.7</span>
  deviance df.residual
<span class="dv">1</span>      <span class="dv">331</span>          <span class="dv">29</span></code></pre></div>
<p>Here’s how we can visualize the corresponding curvilinear regression using ggplot:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(trees, <span class="kw">aes</span>(<span class="dt">x =</span> Girth, <span class="dt">y =</span> Volume)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_smooth</span>(<span class="dt">method =</span> <span class="st">&quot;lm&quot;</span>, 
              <span class="dt">formula =</span> y <span class="op">~</span><span class="st"> </span><span class="kw">I</span>(x<span class="op">^</span><span class="dv">3</span>),
              <span class="dt">se =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-332-1.png" width="576" /></p>
<p>The <code>I()</code> function used above requires a little explanation. Normally, the R formula syntax (see <code>?formula</code>) treats the carat symbol, <code>^</code>, as short-hand for factor crossing to the specified degree. For example, the formula <code>(a+b+c)^2</code> would be interpretted as the model with main effects and all second order interaction terms, i.e. <code>a + b + c + a:b + a:c + b:c</code> where the colons indicate interactions. The <code>I()</code> function `protects’ the object in it’s argument; in this case telling the regression function to treat this as Girth raised to the third power as opposed to trying to construct interaction terms for Girth.</p>
</div>
<div id="exploring-the-impact-of-nearly-collinear-predictors-on-regression" class="section level2">
<h2><span class="header-section-number">11.12</span> Exploring the impact of nearly collinear predictors on regression</h2>
<p>In lecture we discussed the problems that can arise in regression when your predictor variables are nearly collinear. In this section we’ll illustrate some of these issues.</p>
<p>Consider again the <code>trees</code> data set. Recall that two of the variables – Girth and Volume – are highly correlated and thus nearly collinear.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">cor</span>(trees)
        Girth Height Volume
Girth  <span class="fl">1.0000</span> <span class="fl">0.5193</span> <span class="fl">0.9671</span>
Height <span class="fl">0.5193</span> <span class="fl">1.0000</span> <span class="fl">0.5982</span>
Volume <span class="fl">0.9671</span> <span class="fl">0.5982</span> <span class="fl">1.0000</span></code></pre></div>
<p>Let’s explore what happens when we treat Height as the dependent variable, and Girth and Volume as the predictor variables.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit.Height &lt;-<span class="st"> </span><span class="kw">lm</span>(Height <span class="op">~</span><span class="st"> </span>Girth <span class="op">+</span><span class="st"> </span>Volume, <span class="dt">data =</span> trees)

broom<span class="op">::</span><span class="kw">glance</span>(fit.Height)
  r.squared adj.r.squared sigma statistic   p.value df logLik   AIC BIC
<span class="dv">1</span>    <span class="fl">0.4123</span>        <span class="fl">0.3703</span> <span class="fl">5.056</span>      <span class="fl">9.82</span> <span class="fl">0.0005868</span>  <span class="dv">3</span> <span class="op">-</span><span class="fl">92.65</span> <span class="fl">193.3</span> <span class="dv">199</span>
  deviance df.residual
<span class="dv">1</span>    <span class="fl">715.9</span>          <span class="dv">28</span></code></pre></div>
<p>We can, of course, fit the linear model despite the near collinearity, and we find that the model does have some predictive power, with <span class="math inline">\(R^2 = 0.41\)</span>, and with Volume being the more significant predictor.</p>
<p>Now, let’s created a slightly different version of the trees data set by add some noise to the three variables. Our goal here is to simulate a data set we might have created had we measured a slightly different set of trees during our sampling. We’ll use the <code>jitter()</code> function to add uniform noise to the data set.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">jitter.Girth &lt;-<span class="st"> </span><span class="kw">jitter</span>(trees<span class="op">$</span>Girth, <span class="dt">amount=</span> <span class="fl">0.5</span> <span class="op">*</span><span class="st"> </span><span class="kw">sd</span>(trees<span class="op">$</span>Girth))
jitter.Height &lt;-<span class="st"> </span><span class="kw">jitter</span>(trees<span class="op">$</span>Height, <span class="dt">amount=</span> <span class="fl">0.5</span> <span class="op">*</span><span class="st"> </span><span class="kw">sd</span>(trees<span class="op">$</span>Height))
jitter.Volume &lt;-<span class="st"> </span><span class="kw">jitter</span>(trees<span class="op">$</span>Volume, <span class="dt">amount=</span> <span class="fl">0.5</span> <span class="op">*</span><span class="st"> </span><span class="kw">sd</span>(trees<span class="op">$</span>Volume))
jitter.trees &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">Girth =</span> jitter.Girth, 
                        <span class="dt">Height =</span> jitter.Height, 
                        <span class="dt">Volume =</span> jitter.Volume)</code></pre></div>
<p>Here we added uniform noise proportional to the one-quarter the standard deviation of each variable. Let’s take a moment to convince ourselves that our new data set, <code>jitter.trees</code>, is not too different from the <code>trees</code> data set from which it was derived.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">20180228</span>)

<span class="co"># compare this to broom::tidy(trees)</span>
broom<span class="op">::</span><span class="kw">tidy</span>(jitter.trees)
  column  n  mean     sd median trimmed   mad   min   max range     skew
<span class="dv">1</span>  Girth <span class="dv">31</span> <span class="fl">13.13</span>  <span class="fl">3.255</span>  <span class="fl">12.68</span>   <span class="fl">12.93</span> <span class="fl">3.142</span>  <span class="fl">7.90</span> <span class="fl">21.94</span> <span class="fl">14.04</span>  <span class="fl">0.65776</span>
<span class="dv">2</span> Height <span class="dv">31</span> <span class="fl">75.62</span>  <span class="fl">6.833</span>  <span class="fl">75.20</span>   <span class="fl">75.73</span> <span class="fl">6.743</span> <span class="fl">61.89</span> <span class="fl">88.13</span> <span class="fl">26.23</span> <span class="op">-</span><span class="fl">0.08102</span>
<span class="dv">3</span> Volume <span class="dv">31</span> <span class="fl">30.38</span> <span class="fl">17.022</span>  <span class="fl">21.65</span>   <span class="fl">28.56</span> <span class="fl">8.937</span> <span class="fl">10.06</span> <span class="fl">71.95</span> <span class="fl">61.90</span>  <span class="fl">0.84178</span>
  kurtosis     se
<span class="dv">1</span> <span class="op">-</span><span class="fl">0.08395</span> <span class="fl">0.5845</span>
<span class="dv">2</span> <span class="op">-</span><span class="fl">0.88116</span> <span class="fl">1.2273</span>
<span class="dv">3</span> <span class="op">-</span><span class="fl">0.64818</span> <span class="fl">3.0573</span>

<span class="co"># correlations among jittered variables are</span>
<span class="co"># similar to those of the original variables</span>
<span class="kw">cor</span>(jitter.trees)
        Girth Height Volume
Girth  <span class="fl">1.0000</span> <span class="fl">0.5020</span> <span class="fl">0.8609</span>
Height <span class="fl">0.5020</span> <span class="fl">1.0000</span> <span class="fl">0.4846</span>
Volume <span class="fl">0.8609</span> <span class="fl">0.4846</span> <span class="fl">1.0000</span>

## jittered variables are highly correlatd with original variables
<span class="kw">cor</span>(trees<span class="op">$</span>Height, jitter.trees<span class="op">$</span>Height)
[<span class="dv">1</span>] <span class="fl">0.9618</span>
<span class="kw">cor</span>(trees<span class="op">$</span>Girth, jitter.trees<span class="op">$</span>Girth)
[<span class="dv">1</span>] <span class="fl">0.9595</span>
<span class="kw">cor</span>(trees<span class="op">$</span>Volume, jitter.trees<span class="op">$</span>Volume)
[<span class="dv">1</span>] <span class="fl">0.9519</span></code></pre></div>
<p>Now that we’ve convinced ourselves that our jittered data set is a decent approximation to our original data set, let’s re-calculate the linear regression, and compare the coefficients of the jittered model to the original model:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit.Height.jitter &lt;-<span class="st"> </span><span class="kw">lm</span>(Height <span class="op">~</span><span class="st"> </span>Girth <span class="op">+</span><span class="st"> </span>Volume, <span class="dt">data =</span> jitter.trees)

broom<span class="op">::</span><span class="kw">tidy</span>(fit.Height)
         term estimate std.error statistic   p.value
<span class="dv">1</span> (Intercept)  <span class="fl">83.2958</span>    <span class="fl">9.0866</span>     <span class="fl">9.167</span> <span class="fl">6.333e-10</span>
<span class="dv">2</span>       Girth  <span class="op">-</span><span class="fl">1.8615</span>    <span class="fl">1.1567</span>    <span class="op">-</span><span class="fl">1.609</span> <span class="fl">1.188e-01</span>
<span class="dv">3</span>      Volume   <span class="fl">0.5756</span>    <span class="fl">0.2208</span>     <span class="fl">2.607</span> <span class="fl">1.449e-02</span>

broom<span class="op">::</span><span class="kw">tidy</span>(fit.Height.jitter)
         term estimate std.error statistic   p.value
<span class="dv">1</span> (Intercept) <span class="fl">64.11041</span>    <span class="fl">5.8941</span>   <span class="fl">10.8770</span> <span class="fl">1.458e-11</span>
<span class="dv">2</span>       Girth  <span class="fl">0.68831</span>    <span class="fl">0.6697</span>    <span class="fl">1.0277</span> <span class="fl">3.129e-01</span>
<span class="dv">3</span>      Volume  <span class="fl">0.08122</span>    <span class="fl">0.1281</span>    <span class="fl">0.6343</span> <span class="fl">5.310e-01</span></code></pre></div>
<p>We see that the coefficients of the linear model have changed substantially between the original data and the jittered data. Our model is unstable to relatively modest changes to the data!</p>
<p>Let’s draw some plots to illustrate how different the models fit to the original and jittered data are:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># draw 3d scatter plots with small points so as not to obscure regression planes</span>
p &lt;-<span class="st"> </span><span class="kw">scatterplot3d</span>(<span class="dt">x=</span>trees<span class="op">$</span>Girth, <span class="dt">y=</span>trees<span class="op">$</span>Volume, <span class="dt">z=</span>trees<span class="op">$</span>Height, 
                      <span class="dt">angle=</span><span class="dv">15</span>, <span class="dt">type=</span><span class="st">&#39;p&#39;</span>, <span class="dt">pch=</span><span class="st">&#39;.&#39;</span>)

<span class="co"># original model</span>
p<span class="op">$</span><span class="kw">plane3d</span>(fit.Height, <span class="dt">col=</span><span class="st">&#39;orangered&#39;</span>)

<span class="co"># jittered model</span>
p<span class="op">$</span><span class="kw">plane3d</span>(fit.Height.jitter, <span class="dt">col=</span><span class="st">&#39;blue&#39;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-338-1.png" width="576" /></p>
<p>Let’s do the same comparison for the multiple regression of Volume on Height and Girth. In this case the predictor variables are  nearly collinear.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit.Volume &lt;-<span class="st"> </span><span class="kw">lm</span>(Volume <span class="op">~</span><span class="st"> </span>Girth <span class="op">+</span><span class="st"> </span>Height, <span class="dt">data =</span> trees)
fit.Volume.jitter &lt;-<span class="st"> </span><span class="kw">lm</span>(Volume <span class="op">~</span><span class="st"> </span>Girth <span class="op">+</span><span class="st"> </span>Height, <span class="dt">data =</span> jitter.trees)

<span class="kw">coefficients</span>(fit.Volume)
(Intercept)       Girth      Height 
   <span class="op">-</span><span class="fl">57.9877</span>      <span class="fl">4.7082</span>      <span class="fl">0.3393</span> 

<span class="kw">coefficients</span>(fit.Volume.jitter)
(Intercept)       Girth      Height 
   <span class="op">-</span><span class="fl">39.5087</span>      <span class="fl">4.3190</span>      <span class="fl">0.1744</span> </code></pre></div>
<p>For this model, we see that the coefficients have changed only a small amount. The underlying data, <code>jitter.trees</code>, is the same in both cases, but now our model is stable because the predictor variables are only modestly correlated with each other.</p>
<p>Let’s generate another plot to illustrate the similarity of the models fit to the original and jittered data when Girth and Height are used to predict Volume.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p &lt;-<span class="st"> </span><span class="kw">scatterplot3d</span>(<span class="dt">x=</span>trees<span class="op">$</span>Girth, <span class="dt">y=</span>trees<span class="op">$</span>Height, <span class="dt">z=</span>trees<span class="op">$</span>Volume, 
                     <span class="dt">angle=</span><span class="dv">55</span>, <span class="dt">type=</span><span class="st">&#39;p&#39;</span>, <span class="dt">pch=</span><span class="st">&#39;.&#39;</span>)
p<span class="op">$</span><span class="kw">plane3d</span>(fit.Volume, <span class="dt">col=</span><span class="st">&#39;orangered&#39;</span>)
p<span class="op">$</span><span class="kw">plane3d</span>(fit.Volume.jitter, <span class="dt">col=</span><span class="st">&#39;blue&#39;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-340-1.png" width="576" /></p>
<p>Finally, let’s do some vector calculations to quantify how the angular deviation between the fit data and the predictor variables changes between the original and jittered data set for the two different multiple regressions:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># write a quickie fxn to express angle between vectors in degrees</span>
vec.angle &lt;-<span class="st"> </span><span class="cf">function</span>(x,y) { <span class="kw">acos</span>(<span class="kw">cor</span>(x,y)) <span class="op">*</span><span class="st"> </span>(<span class="dv">180</span><span class="op">/</span>pi)}

<span class="co"># vector angles for fit of Height ~ Girth + Volume (orig)</span>
<span class="kw">vec.angle</span>(fit.Height<span class="op">$</span>fitted.values, trees<span class="op">$</span>Girth)
[<span class="dv">1</span>] <span class="fl">36.03</span>
<span class="kw">vec.angle</span>(fit.Height<span class="op">$</span>fitted.values, trees<span class="op">$</span>Volume)
[<span class="dv">1</span>] <span class="fl">21.29</span>

<span class="co"># vector angles for fit of Height ~ Girth + Volume (jittered)</span>
<span class="kw">vec.angle</span>(fit.Height.jitter<span class="op">$</span>fitted.values, jitter.trees<span class="op">$</span>Girth)
[<span class="dv">1</span>] <span class="fl">11.59</span>
<span class="kw">vec.angle</span>(fit.Height.jitter<span class="op">$</span>fitted.values, jitter.trees<span class="op">$</span>Volume)
[<span class="dv">1</span>] <span class="fl">18.99</span></code></pre></div>
<p>Now the same comparison for the non-collinear model <code>Volume ~ Girth + Height</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># vector angles for fit of Volume ~ Girth + Height (orig)</span>
<span class="kw">vec.angle</span>(fit.Volume<span class="op">$</span>fitted.values, trees<span class="op">$</span>Girth)
[<span class="dv">1</span>] <span class="fl">6.628</span>
<span class="kw">vec.angle</span>(fit.Volume<span class="op">$</span>fitted.values, trees<span class="op">$</span>Height)
[<span class="dv">1</span>] <span class="fl">52.09</span>

<span class="co"># vector angles for fit of Volume ~ Girth + Height (jittered)</span>
<span class="kw">vec.angle</span>(fit.Volume.jitter<span class="op">$</span>fitted.values, jitter.trees<span class="op">$</span>Girth)
[<span class="dv">1</span>] <span class="fl">4.023</span>
<span class="kw">vec.angle</span>(fit.Volume.jitter<span class="op">$</span>fitted.values, jitter.trees<span class="op">$</span>Height)
[<span class="dv">1</span>] <span class="fl">55.84</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="matrices-in-r.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="principal-components-analysis.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
