<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 14 Principal Components Analysis | Biology 723: Statistical Computing for Biologists</title>
  <meta name="description" content="Hands-on materials for Bio 723 course at Duke University.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 14 Principal Components Analysis | Biology 723: Statistical Computing for Biologists" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 14 Principal Components Analysis | Biology 723: Statistical Computing for Biologists" />
  
  <meta name="twitter:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

<meta name="author" content="Paul M. Magwene">


<meta name="date" content="2019-03-20">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="logistic-regression.html">
<link rel="next" href="singular-value-decomposition.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script src="libs/htmlwidgets-1.3/htmlwidgets.js"></script>
<link href="libs/rglwidgetClass-2/rgl.css" rel="stylesheet" />
<script src="libs/rglwidgetClass-2/rglClass.src.js"></script>
<script src="libs/CanvasMatrix4-2016/CanvasMatrix.src.js"></script>
<script src="libs/rglWebGL-binding-0.99.16/rglWebGL.js"></script>
<link href="libs/crosstalk-1.0.0/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.0.0/js/crosstalk.min.js"></script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { background-color: #f8f8f8; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#how-to-use-these-lecture-notes"><i class="fa fa-check"></i><b>1.1</b> How to use these lecture notes</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html"><i class="fa fa-check"></i><b>2</b> Getting Started with R</a><ul>
<li class="chapter" data-level="2.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-r"><i class="fa fa-check"></i><b>2.1</b> What is R?</a></li>
<li class="chapter" data-level="2.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-rstudio"><i class="fa fa-check"></i><b>2.2</b> What is RStudio?</a></li>
<li class="chapter" data-level="2.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#entering-commands-in-the-console"><i class="fa fa-check"></i><b>2.3</b> Entering commands in the console</a></li>
<li class="chapter" data-level="2.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#comments"><i class="fa fa-check"></i><b>2.4</b> Comments</a></li>
<li class="chapter" data-level="2.5" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#using-r-as-a-calculator"><i class="fa fa-check"></i><b>2.5</b> Using R as a Calculator</a><ul>
<li class="chapter" data-level="2.5.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#common-mathematical-functions"><i class="fa fa-check"></i><b>2.5.1</b> Common mathematical functions</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#variable-assignment"><i class="fa fa-check"></i><b>2.6</b> Variable assignment</a><ul>
<li class="chapter" data-level="2.6.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#valid-variable-names"><i class="fa fa-check"></i><b>2.6.1</b> Valid variable names</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#data-types"><i class="fa fa-check"></i><b>2.7</b> Data types</a><ul>
<li class="chapter" data-level="2.7.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#numeric-data-types"><i class="fa fa-check"></i><b>2.7.1</b> Numeric data types</a></li>
<li class="chapter" data-level="2.7.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#logical-values"><i class="fa fa-check"></i><b>2.7.2</b> Logical values</a></li>
<li class="chapter" data-level="2.7.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#character-strings"><i class="fa fa-check"></i><b>2.7.3</b> Character strings</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#packages"><i class="fa fa-check"></i><b>2.8</b> Packages</a><ul>
<li class="chapter" data-level="2.8.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-console"><i class="fa fa-check"></i><b>2.8.1</b> Installing packages from the console</a></li>
<li class="chapter" data-level="2.8.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#install-the-tidyverse-package"><i class="fa fa-check"></i><b>2.8.2</b> Install the tidyverse package</a></li>
<li class="chapter" data-level="2.8.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-rstudio-dialog"><i class="fa fa-check"></i><b>2.8.3</b> Installing packages from the RStudio dialog</a></li>
<li class="chapter" data-level="2.8.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#loading-packages-with-the-library-function"><i class="fa fa-check"></i><b>2.8.4</b> Loading packages with the <code>library()</code> function</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#the-r-help-system"><i class="fa fa-check"></i><b>2.9</b> The R Help System</a><ul>
<li class="chapter" data-level="2.9.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#getting-help-from-the-console"><i class="fa fa-check"></i><b>2.9.1</b> Getting help from the console</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html"><i class="fa fa-check"></i><b>3</b> R Markdown and R Notebooks</a><ul>
<li class="chapter" data-level="3.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#r-notebooks"><i class="fa fa-check"></i><b>3.1</b> R Notebooks</a></li>
<li class="chapter" data-level="3.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#creating-an-r-notebook"><i class="fa fa-check"></i><b>3.2</b> Creating an R Notebook</a></li>
<li class="chapter" data-level="3.3" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#the-default-r-notebook-template"><i class="fa fa-check"></i><b>3.3</b> The default R Notebook template</a></li>
<li class="chapter" data-level="3.4" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-and-non-code-blocks"><i class="fa fa-check"></i><b>3.4</b> Code and Non-code blocks</a><ul>
<li class="chapter" data-level="3.4.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#non-code-blocks"><i class="fa fa-check"></i><b>3.4.1</b> Non-code blocks</a></li>
<li class="chapter" data-level="3.4.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-blocks"><i class="fa fa-check"></i><b>3.4.2</b> Code blocks</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-a-code-chunk"><i class="fa fa-check"></i><b>3.5</b> Running a code chunk</a></li>
<li class="chapter" data-level="3.6" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-all-code-chunks-above"><i class="fa fa-check"></i><b>3.6</b> Running all code chunks above</a></li>
<li class="chapter" data-level="3.7" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#knitting-an-r-markdown-to-html"><i class="fa fa-check"></i><b>3.7</b> “Knitting an” R Markdown to HTML</a></li>
<li class="chapter" data-level="3.8" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#sharing-your-reproducible-r-notebook"><i class="fa fa-check"></i><b>3.8</b> Sharing your reproducible R Notebook</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-structures.html"><a href="data-structures.html"><i class="fa fa-check"></i><b>4</b> Data structures</a><ul>
<li class="chapter" data-level="4.1" data-path="data-structures.html"><a href="data-structures.html#vectors"><i class="fa fa-check"></i><b>4.1</b> Vectors</a><ul>
<li class="chapter" data-level="4.1.1" data-path="data-structures.html"><a href="data-structures.html#vector-arithmetic"><i class="fa fa-check"></i><b>4.1.1</b> Vector Arithmetic</a></li>
<li class="chapter" data-level="4.1.2" data-path="data-structures.html"><a href="data-structures.html#vector-recycling"><i class="fa fa-check"></i><b>4.1.2</b> Vector recycling</a></li>
<li class="chapter" data-level="4.1.3" data-path="data-structures.html"><a href="data-structures.html#simple-statistical-functions-for-numeric-vectors"><i class="fa fa-check"></i><b>4.1.3</b> Simple statistical functions for numeric vectors</a></li>
<li class="chapter" data-level="4.1.4" data-path="data-structures.html"><a href="data-structures.html#indexing-vectors"><i class="fa fa-check"></i><b>4.1.4</b> Indexing Vectors</a></li>
<li class="chapter" data-level="4.1.5" data-path="data-structures.html"><a href="data-structures.html#comparison-operators-applied-to-vectors"><i class="fa fa-check"></i><b>4.1.5</b> Comparison operators applied to vectors</a></li>
<li class="chapter" data-level="4.1.6" data-path="data-structures.html"><a href="data-structures.html#combining-indexing-and-comparison-of-vectors"><i class="fa fa-check"></i><b>4.1.6</b> Combining Indexing and Comparison of Vectors</a></li>
<li class="chapter" data-level="4.1.7" data-path="data-structures.html"><a href="data-structures.html#vector-manipulation"><i class="fa fa-check"></i><b>4.1.7</b> Vector manipulation</a></li>
<li class="chapter" data-level="4.1.8" data-path="data-structures.html"><a href="data-structures.html#vectors-from-regular-sequences"><i class="fa fa-check"></i><b>4.1.8</b> Vectors from regular sequences</a></li>
<li class="chapter" data-level="4.1.9" data-path="data-structures.html"><a href="data-structures.html#additional-functions-for-working-with-vectors"><i class="fa fa-check"></i><b>4.1.9</b> Additional functions for working with vectors</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="data-structures.html"><a href="data-structures.html#lists"><i class="fa fa-check"></i><b>4.2</b> Lists</a><ul>
<li class="chapter" data-level="4.2.1" data-path="data-structures.html"><a href="data-structures.html#length-and-type-of-lists"><i class="fa fa-check"></i><b>4.2.1</b> Length and type of lists</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-structures.html"><a href="data-structures.html#indexing-lists"><i class="fa fa-check"></i><b>4.2.2</b> Indexing lists</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-structures.html"><a href="data-structures.html#naming-list-elements"><i class="fa fa-check"></i><b>4.2.3</b> Naming list elements</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-structures.html"><a href="data-structures.html#the-operator"><i class="fa fa-check"></i><b>4.2.4</b> The <code>$</code> operator</a></li>
<li class="chapter" data-level="4.2.5" data-path="data-structures.html"><a href="data-structures.html#changing-and-adding-lists-items"><i class="fa fa-check"></i><b>4.2.5</b> Changing and adding lists items</a></li>
<li class="chapter" data-level="4.2.6" data-path="data-structures.html"><a href="data-structures.html#combining-lists"><i class="fa fa-check"></i><b>4.2.6</b> Combining lists</a></li>
<li class="chapter" data-level="4.2.7" data-path="data-structures.html"><a href="data-structures.html#converting-lists-to-vectors"><i class="fa fa-check"></i><b>4.2.7</b> Converting lists to vectors</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-structures.html"><a href="data-structures.html#data-frames"><i class="fa fa-check"></i><b>4.3</b> Data frames</a><ul>
<li class="chapter" data-level="4.3.1" data-path="data-structures.html"><a href="data-structures.html#creating-a-data-frame"><i class="fa fa-check"></i><b>4.3.1</b> Creating a data frame</a></li>
<li class="chapter" data-level="4.3.2" data-path="data-structures.html"><a href="data-structures.html#type-and-class-for-data-frames"><i class="fa fa-check"></i><b>4.3.2</b> Type and class for data frames</a></li>
<li class="chapter" data-level="4.3.3" data-path="data-structures.html"><a href="data-structures.html#length-and-dimension-for-data-frames"><i class="fa fa-check"></i><b>4.3.3</b> Length and dimension for data frames</a></li>
<li class="chapter" data-level="4.3.4" data-path="data-structures.html"><a href="data-structures.html#indexing-and-accessing-data-frames"><i class="fa fa-check"></i><b>4.3.4</b> Indexing and accessing data frames</a></li>
<li class="chapter" data-level="4.3.5" data-path="data-structures.html"><a href="data-structures.html#logical-indexing-of-data-frames"><i class="fa fa-check"></i><b>4.3.5</b> Logical indexing of data frames</a></li>
<li class="chapter" data-level="4.3.6" data-path="data-structures.html"><a href="data-structures.html#adding-columns-to-a-data-frame"><i class="fa fa-check"></i><b>4.3.6</b> Adding columns to a data frame</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html"><i class="fa fa-check"></i><b>5</b> Introduction to ggplot2</a><ul>
<li class="chapter" data-level="5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#loading-ggplot2"><i class="fa fa-check"></i><b>5.1</b> Loading ggplot2</a></li>
<li class="chapter" data-level="5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#example-data-set-andersons-iris-data"><i class="fa fa-check"></i><b>5.2</b> Example data set: Anderson’s Iris Data</a></li>
<li class="chapter" data-level="5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#template-for-single-layer-plots-in-ggplot2"><i class="fa fa-check"></i><b>5.3</b> Template for single layer plots in ggplot2</a></li>
<li class="chapter" data-level="5.4" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#an-aside-about-function-arguments"><i class="fa fa-check"></i><b>5.4</b> An aside about function arguments</a></li>
<li class="chapter" data-level="5.5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#strip-plots"><i class="fa fa-check"></i><b>5.5</b> Strip plots</a><ul>
<li class="chapter" data-level="5.5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#jittering-data"><i class="fa fa-check"></i><b>5.5.1</b> Jittering data</a></li>
<li class="chapter" data-level="5.5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-categorical-information"><i class="fa fa-check"></i><b>5.5.2</b> Adding categorical information</a></li>
<li class="chapter" data-level="5.5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#rotating-plot-coordinates"><i class="fa fa-check"></i><b>5.5.3</b> Rotating plot coordinates</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#histograms"><i class="fa fa-check"></i><b>5.6</b> Histograms</a><ul>
<li class="chapter" data-level="5.6.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#variations-on-histograms-when-considering-categorical-data"><i class="fa fa-check"></i><b>5.6.1</b> Variations on histograms when considering categorical data</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#faceting-to-depict-categorical-information"><i class="fa fa-check"></i><b>5.7</b> Faceting to depict categorical information</a></li>
<li class="chapter" data-level="5.8" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots"><i class="fa fa-check"></i><b>5.8</b> Density plots</a></li>
<li class="chapter" data-level="5.9" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-or-beanplot"><i class="fa fa-check"></i><b>5.9</b> Violin or Beanplot</a></li>
<li class="chapter" data-level="5.10" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplots"><i class="fa fa-check"></i><b>5.10</b> Boxplots</a></li>
<li class="chapter" data-level="5.11" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#building-complex-visualizations-with-layers"><i class="fa fa-check"></i><b>5.11</b> Building complex visualizations with layers</a></li>
<li class="chapter" data-level="5.12" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#useful-combination-plots"><i class="fa fa-check"></i><b>5.12</b> Useful combination plots</a><ul>
<li class="chapter" data-level="5.12.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplot-plus-strip-plot"><i class="fa fa-check"></i><b>5.12.1</b> Boxplot plus strip plot</a></li>
<li class="chapter" data-level="5.12.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#setting-shared-aesthetics"><i class="fa fa-check"></i><b>5.12.2</b> Setting shared aesthetics</a></li>
</ul></li>
<li class="chapter" data-level="5.13" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot-layers-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>5.13</b> ggplot layers can be assigned to variables</a><ul>
<li class="chapter" data-level="5.13.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-plot-plus-strip-plot"><i class="fa fa-check"></i><b>5.13.1</b> Violin plot plus strip plot</a></li>
</ul></li>
<li class="chapter" data-level="5.14" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-titles-and-tweaking-axis-labels"><i class="fa fa-check"></i><b>5.14</b> Adding titles and tweaking axis labels</a></li>
<li class="chapter" data-level="5.15" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot2-themes"><i class="fa fa-check"></i><b>5.15</b> ggplot2 themes</a><ul>
<li class="chapter" data-level="5.15.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#further-customization-with-ggplot2theme"><i class="fa fa-check"></i><b>5.15.1</b> Further customization with <code>ggplot2::theme</code></a></li>
</ul></li>
<li class="chapter" data-level="5.16" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#other-aspects-of-ggplots-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>5.16</b> Other aspects of ggplots can be assigned to variables</a></li>
<li class="chapter" data-level="5.17" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-plots"><i class="fa fa-check"></i><b>5.17</b> Bivariate plots</a><ul>
<li class="chapter" data-level="5.17.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#scatter-plots"><i class="fa fa-check"></i><b>5.17.1</b> Scatter plots</a></li>
<li class="chapter" data-level="5.17.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-a-trend-line-to-a-scatter-plot"><i class="fa fa-check"></i><b>5.17.2</b> Adding a trend line to a scatter plot</a></li>
</ul></li>
<li class="chapter" data-level="5.18" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-density-plots"><i class="fa fa-check"></i><b>5.18</b> Bivariate density plots</a></li>
<li class="chapter" data-level="5.19" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#combining-scatter-plots-and-density-plots-with-categorical-information"><i class="fa fa-check"></i><b>5.19</b> Combining Scatter Plots and Density Plots with Categorical Information</a></li>
<li class="chapter" data-level="5.20" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots-with-fill"><i class="fa fa-check"></i><b>5.20</b> Density plots with fill</a></li>
<li class="chapter" data-level="5.21" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#d-bin-and-hex-plots"><i class="fa fa-check"></i><b>5.21</b> 2D bin and hex plots</a></li>
<li class="chapter" data-level="5.22" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#the-cowplot-package"><i class="fa fa-check"></i><b>5.22</b> The <code>cowplot</code> package</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html"><i class="fa fa-check"></i><b>6</b> Introduction to <code>dplyr</code></a><ul>
<li class="chapter" data-level="6.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#libraries"><i class="fa fa-check"></i><b>6.1</b> Libraries</a></li>
<li class="chapter" data-level="6.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#reading-data-with-the-readr-package"><i class="fa fa-check"></i><b>6.2</b> Reading data with the <code>readr</code> package</a><ul>
<li class="chapter" data-level="6.2.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#reading-excel-files"><i class="fa fa-check"></i><b>6.2.1</b> Reading Excel files</a></li>
<li class="chapter" data-level="6.2.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#example-data-nc-births"><i class="fa fa-check"></i><b>6.2.2</b> Example data: NC Births</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#a-note-on-tibbles"><i class="fa fa-check"></i><b>6.3</b> A note on “tibbles”</a></li>
<li class="chapter" data-level="6.4" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#data-filtering-and-transformation-with-dplyr"><i class="fa fa-check"></i><b>6.4</b> Data filtering and transformation with <code>dplyr</code></a></li>
<li class="chapter" data-level="6.5" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#dplyrs-verbs"><i class="fa fa-check"></i><b>6.5</b> dplyr’s “verbs”</a><ul>
<li class="chapter" data-level="6.5.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#select"><i class="fa fa-check"></i><b>6.5.1</b> <code>select</code></a></li>
<li class="chapter" data-level="6.5.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#filter"><i class="fa fa-check"></i><b>6.5.2</b> <code>filter</code></a></li>
<li class="chapter" data-level="6.5.3" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#mutate"><i class="fa fa-check"></i><b>6.5.3</b> <code>mutate</code></a></li>
<li class="chapter" data-level="6.5.4" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#arrange"><i class="fa fa-check"></i><b>6.5.4</b> <code>arrange</code></a></li>
<li class="chapter" data-level="6.5.5" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#summarize"><i class="fa fa-check"></i><b>6.5.5</b> <code>summarize</code></a></li>
<li class="chapter" data-level="6.5.6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#group_by"><i class="fa fa-check"></i><b>6.5.6</b> <code>group_by</code></a></li>
<li class="chapter" data-level="6.5.7" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#combining-grouping-and-summarizing"><i class="fa fa-check"></i><b>6.5.7</b> Combining grouping and summarizing</a></li>
<li class="chapter" data-level="6.5.8" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#scoped-variants-of-mutate-and-summarize"><i class="fa fa-check"></i><b>6.5.8</b> Scoped variants of <code>mutate</code> and <code>summarize</code></a></li>
<li class="chapter" data-level="6.5.9" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#combining-summarize-with-grouping-aesthetics-in-ggplot2"><i class="fa fa-check"></i><b>6.5.9</b> Combining summarize with grouping aesthetics in <code>ggplot2</code></a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#pipes"><i class="fa fa-check"></i><b>6.6</b> Pipes</a><ul>
<li class="chapter" data-level="6.6.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#install-and-load-magrittr"><i class="fa fa-check"></i><b>6.6.1</b> Install and load <code>magrittr</code></a></li>
<li class="chapter" data-level="6.6.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-basic-pipe-operator"><i class="fa fa-check"></i><b>6.6.2</b> The basic pipe operator</a></li>
<li class="chapter" data-level="6.6.3" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#an-example-without-pipes"><i class="fa fa-check"></i><b>6.6.3</b> An example without pipes</a></li>
<li class="chapter" data-level="6.6.4" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-same-example-using-pipes"><i class="fa fa-check"></i><b>6.6.4</b> The same example using pipes</a></li>
<li class="chapter" data-level="6.6.5" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#assigning-the-output-of-a-statement-involving-pipes-to-a-variable"><i class="fa fa-check"></i><b>6.6.5</b> Assigning the output of a statement involving pipes to a variable</a></li>
<li class="chapter" data-level="6.6.6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#compound-assignment-pipe-operator"><i class="fa fa-check"></i><b>6.6.6</b> Compound assignment pipe operator</a></li>
<li class="chapter" data-level="6.6.7" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-dot-operator-with-pipes"><i class="fa fa-check"></i><b>6.6.7</b> The dot operator with pipes</a></li>
<li class="chapter" data-level="6.6.8" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-exposition-pipe-operator"><i class="fa fa-check"></i><b>6.6.8</b> The exposition pipe operator</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>7</b> Data wrangling</a><ul>
<li class="chapter" data-level="7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#libraries-1"><i class="fa fa-check"></i><b>7.1</b> Libraries</a></li>
<li class="chapter" data-level="7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#data"><i class="fa fa-check"></i><b>7.2</b> Data</a></li>
<li class="chapter" data-level="7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#renaming-data-frame-columms"><i class="fa fa-check"></i><b>7.3</b> Renaming data frame columms</a></li>
<li class="chapter" data-level="7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-unneeded-columns"><i class="fa fa-check"></i><b>7.4</b> Dropping unneeded columns</a><ul>
<li class="chapter" data-level="7.4.1" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-all-empty-columns"><i class="fa fa-check"></i><b>7.4.1</b> Finding all empty columns</a></li>
<li class="chapter" data-level="7.4.2" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-columns-by-matching-names"><i class="fa fa-check"></i><b>7.4.2</b> Dropping columns by matching names</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="data-wrangling.html"><a href="data-wrangling.html#merging-data-frames"><i class="fa fa-check"></i><b>7.5</b> Merging data frames</a></li>
<li class="chapter" data-level="7.6" data-path="data-wrangling.html"><a href="data-wrangling.html#reshaping-data-with-tidyr"><i class="fa fa-check"></i><b>7.6</b> Reshaping data with tidyr</a><ul>
<li class="chapter" data-level="7.6.1" data-path="data-wrangling.html"><a href="data-wrangling.html#wide-to-long-conversions-using-tidyrgather"><i class="fa fa-check"></i><b>7.6.1</b> Wide to long conversions using <code>tidyr::gather</code></a></li>
<li class="chapter" data-level="7.6.2" data-path="data-wrangling.html"><a href="data-wrangling.html#extracting-information-from-combined-variables-using-tidyrextract"><i class="fa fa-check"></i><b>7.6.2</b> Extracting information from combined variables using <code>tidyr::extract</code></a></li>
<li class="chapter" data-level="7.6.3" data-path="data-wrangling.html"><a href="data-wrangling.html#subsetting-rows"><i class="fa fa-check"></i><b>7.6.3</b> Subsetting rows</a></li>
<li class="chapter" data-level="7.6.4" data-path="data-wrangling.html"><a href="data-wrangling.html#splitting-columns"><i class="fa fa-check"></i><b>7.6.4</b> Splitting columns</a></li>
<li class="chapter" data-level="7.6.5" data-path="data-wrangling.html"><a href="data-wrangling.html#combining-data-frames"><i class="fa fa-check"></i><b>7.6.5</b> Combining data frames</a></li>
<li class="chapter" data-level="7.6.6" data-path="data-wrangling.html"><a href="data-wrangling.html#sorting-data-frame-rows"><i class="fa fa-check"></i><b>7.6.6</b> Sorting data frame rows</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="data-wrangling.html"><a href="data-wrangling.html#using-your-tidy-data"><i class="fa fa-check"></i><b>7.7</b> Using your tidy data</a><ul>
<li class="chapter" data-level="7.7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#visualizing-gene-expression-time-series"><i class="fa fa-check"></i><b>7.7.1</b> Visualizing gene expression time series</a></li>
<li class="chapter" data-level="7.7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-the-most-variable-genes"><i class="fa fa-check"></i><b>7.7.2</b> Finding the most variable genes</a></li>
<li class="chapter" data-level="7.7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#heat-maps"><i class="fa fa-check"></i><b>7.7.3</b> Heat maps</a></li>
</ul></li>
<li class="chapter" data-level="7.8" data-path="data-wrangling.html"><a href="data-wrangling.html#long-to-wide-conversion-using-tidyrspread"><i class="fa fa-check"></i><b>7.8</b> Long-to-wide conversion using <code>tidyr::spread</code></a></li>
<li class="chapter" data-level="7.9" data-path="data-wrangling.html"><a href="data-wrangling.html#exploring-bivariate-relationships-using-wide-data"><i class="fa fa-check"></i><b>7.9</b> Exploring bivariate relationships using “wide” data</a><ul>
<li class="chapter" data-level="7.9.1" data-path="data-wrangling.html"><a href="data-wrangling.html#large-scale-patterns-of-correlations"><i class="fa fa-check"></i><b>7.9.1</b> Large scale patterns of correlations</a></li>
<li class="chapter" data-level="7.9.2" data-path="data-wrangling.html"><a href="data-wrangling.html#adding-new-columns-and-combining-filtered-data-frames"><i class="fa fa-check"></i><b>7.9.2</b> Adding new columns and combining filtered data frames</a></li>
<li class="chapter" data-level="7.9.3" data-path="data-wrangling.html"><a href="data-wrangling.html#a-heat-mapped-sorted-by-correlations"><i class="fa fa-check"></i><b>7.9.3</b> A heat mapped sorted by correlations</a></li>
<li class="chapter" data-level="7.9.4" data-path="data-wrangling.html"><a href="data-wrangling.html#a-fancy-figure"><i class="fa fa-check"></i><b>7.9.4</b> A “fancy” figure</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html"><i class="fa fa-check"></i><b>8</b> Functions and control flow statements</a><ul>
<li class="chapter" data-level="8.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#writing-your-own-functions"><i class="fa fa-check"></i><b>8.1</b> Writing your own functions</a><ul>
<li class="chapter" data-level="8.1.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#function-arguments"><i class="fa fa-check"></i><b>8.1.1</b> Function arguments</a></li>
<li class="chapter" data-level="8.1.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#writing-functions-with-optional-arguments"><i class="fa fa-check"></i><b>8.1.2</b> Writing functions with optional arguments</a></li>
<li class="chapter" data-level="8.1.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#putting-r-functions-in-scripts"><i class="fa fa-check"></i><b>8.1.3</b> Putting R functions in Scripts</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#control-flow-statements"><i class="fa fa-check"></i><b>8.2</b> Control flow statements</a><ul>
<li class="chapter" data-level="8.2.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#if-and-if-else-statements"><i class="fa fa-check"></i><b>8.2.1</b> <code>if</code> and <code>if-else</code> statements</a></li>
<li class="chapter" data-level="8.2.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#for-loops"><i class="fa fa-check"></i><b>8.2.2</b> for loops</a></li>
<li class="chapter" data-level="8.2.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#break-statement"><i class="fa fa-check"></i><b>8.2.3</b> <code>break</code> statement</a></li>
<li class="chapter" data-level="8.2.4" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#repeat-loops"><i class="fa fa-check"></i><b>8.2.4</b> <code>repeat</code> loops</a></li>
<li class="chapter" data-level="8.2.5" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#next-statement"><i class="fa fa-check"></i><b>8.2.5</b> <code>next</code> statement</a></li>
<li class="chapter" data-level="8.2.6" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#while-statements"><i class="fa fa-check"></i><b>8.2.6</b> while statements</a></li>
<li class="chapter" data-level="8.2.7" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#ifelse"><i class="fa fa-check"></i><b>8.2.7</b> <code>ifelse</code></a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#map-and-related-tools"><i class="fa fa-check"></i><b>8.3</b> <code>map</code> and related tools</a><ul>
<li class="chapter" data-level="8.3.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#basic-map"><i class="fa fa-check"></i><b>8.3.1</b> basic <code>map</code></a></li>
<li class="chapter" data-level="8.3.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#map_if-and-map_at"><i class="fa fa-check"></i><b>8.3.2</b> <code>map_if</code> and <code>map_at</code></a></li>
<li class="chapter" data-level="8.3.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#mapping-in-parallel-using-map2"><i class="fa fa-check"></i><b>8.3.3</b> mapping in parallel using <code>map2</code></a></li>
<li class="chapter" data-level="8.3.4" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#map-variants-that-return-vectors"><i class="fa fa-check"></i><b>8.3.4</b> <code>map</code> variants that return vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="vector-algebra.html"><a href="vector-algebra.html"><i class="fa fa-check"></i><b>9</b> Vector algebra</a><ul>
<li class="chapter" data-level="9.1" data-path="vector-algebra.html"><a href="vector-algebra.html#libraries-2"><i class="fa fa-check"></i><b>9.1</b> Libraries</a></li>
<li class="chapter" data-level="9.2" data-path="vector-algebra.html"><a href="vector-algebra.html#vector-mathematics-in-r"><i class="fa fa-check"></i><b>9.2</b> Vector Mathematics in R</a></li>
<li class="chapter" data-level="9.3" data-path="vector-algebra.html"><a href="vector-algebra.html#simple-statistics-in-vector-form"><i class="fa fa-check"></i><b>9.3</b> Simple statistics in vector form</a><ul>
<li class="chapter" data-level="9.3.1" data-path="vector-algebra.html"><a href="vector-algebra.html#mean"><i class="fa fa-check"></i><b>9.3.1</b> Mean</a></li>
<li class="chapter" data-level="9.3.2" data-path="vector-algebra.html"><a href="vector-algebra.html#mean-centering"><i class="fa fa-check"></i><b>9.3.2</b> Mean centering</a></li>
<li class="chapter" data-level="9.3.3" data-path="vector-algebra.html"><a href="vector-algebra.html#variance-and-standard-deviation"><i class="fa fa-check"></i><b>9.3.3</b> Variance and standard deviation</a></li>
<li class="chapter" data-level="9.3.4" data-path="vector-algebra.html"><a href="vector-algebra.html#covariance-and-correlation"><i class="fa fa-check"></i><b>9.3.4</b> Covariance and correlation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="matrices-in-r.html"><a href="matrices-in-r.html"><i class="fa fa-check"></i><b>10</b> Matrices in R</a><ul>
<li class="chapter" data-level="10.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#creating-matrices-in-r"><i class="fa fa-check"></i><b>10.1</b> Creating matrices in R</a></li>
<li class="chapter" data-level="10.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-arithmetic-operations-in-r"><i class="fa fa-check"></i><b>10.2</b> Matrix arithmetic operations in R</a></li>
<li class="chapter" data-level="10.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#descriptive-statistics-as-matrix-functions"><i class="fa fa-check"></i><b>10.3</b> Descriptive statistics as matrix functions</a><ul>
<li class="chapter" data-level="10.3.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#mean-vector-and-matrix"><i class="fa fa-check"></i><b>10.3.1</b> Mean vector and matrix</a></li>
<li class="chapter" data-level="10.3.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#deviation-matrix"><i class="fa fa-check"></i><b>10.3.2</b> Deviation matrix</a></li>
<li class="chapter" data-level="10.3.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#covariance-matrix"><i class="fa fa-check"></i><b>10.3.3</b> Covariance matrix</a></li>
<li class="chapter" data-level="10.3.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#correlation-matrix"><i class="fa fa-check"></i><b>10.3.4</b> Correlation matrix</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-inverse"><i class="fa fa-check"></i><b>10.4</b> Matrix Inverse</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html"><i class="fa fa-check"></i><b>11</b> Fitting Regression Models in R</a><ul>
<li class="chapter" data-level="11.1" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#regression-terminology"><i class="fa fa-check"></i><b>11.1</b> Regression terminology</a></li>
<li class="chapter" data-level="11.2" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#the-optimality-criterion-for-least-squares-regression"><i class="fa fa-check"></i><b>11.2</b> The optimality criterion for least-squares regression</a></li>
<li class="chapter" data-level="11.3" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#vector-geometric-perspective"><i class="fa fa-check"></i><b>11.3</b> Vector geometric perspective</a></li>
<li class="chapter" data-level="11.4" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#solution-for-the-least-squares-criterion"><i class="fa fa-check"></i><b>11.4</b> Solution for the least-squares criterion</a></li>
<li class="chapter" data-level="11.5" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#libraries-3"><i class="fa fa-check"></i><b>11.5</b> Libraries</a></li>
<li class="chapter" data-level="11.6" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#illustrating-linear-regression-with-simulated-data"><i class="fa fa-check"></i><b>11.6</b> Illustrating linear regression with simulated data</a></li>
<li class="chapter" data-level="11.7" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#specifying-regression-models-in-r"><i class="fa fa-check"></i><b>11.7</b> Specifying Regression Models in R</a><ul>
<li class="chapter" data-level="11.7.1" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#fitted-values"><i class="fa fa-check"></i><b>11.7.1</b> Fitted values</a></li>
<li class="chapter" data-level="11.7.2" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#getting-the-model-coefficients"><i class="fa fa-check"></i><b>11.7.2</b> Getting the model coefficients</a></li>
</ul></li>
<li class="chapter" data-level="11.8" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#residuals"><i class="fa fa-check"></i><b>11.8</b> Residuals</a></li>
<li class="chapter" data-level="11.9" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#regression-as-sum-of-squares-decomposition"><i class="fa fa-check"></i><b>11.9</b> Regression as sum-of-squares decomposition</a></li>
<li class="chapter" data-level="11.10" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#variance-explained-by-a-regression-model"><i class="fa fa-check"></i><b>11.10</b> Variance “explained” by a regression model</a></li>
<li class="chapter" data-level="11.11" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#broom-a-library-for-converting-model-results-into-data-frames"><i class="fa fa-check"></i><b>11.11</b> Broom: a library for converting model results into data frames</a><ul>
<li class="chapter" data-level="11.11.1" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#broomtidy"><i class="fa fa-check"></i><b>11.11.1</b> <code>broom::tidy</code></a></li>
<li class="chapter" data-level="11.11.2" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#broomaugment"><i class="fa fa-check"></i><b>11.11.2</b> <code>broom::augment</code></a></li>
<li class="chapter" data-level="11.11.3" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#broomglance"><i class="fa fa-check"></i><b>11.11.3</b> <code>broom::glance</code></a></li>
</ul></li>
<li class="chapter" data-level="11.12" data-path="fitting-regression-models-in-r.html"><a href="fitting-regression-models-in-r.html#example-predicting-lion-age-based-on-nose-color"><i class="fa fa-check"></i><b>11.12</b> Example: Predicting lion age based on nose color</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="multiple-regression.html"><a href="multiple-regression.html"><i class="fa fa-check"></i><b>12</b> Multiple Regression</a><ul>
<li class="chapter" data-level="12.1" data-path="multiple-regression.html"><a href="multiple-regression.html#review-of-bivariate-regression"><i class="fa fa-check"></i><b>12.1</b> Review of bivariate regression</a></li>
<li class="chapter" data-level="12.2" data-path="multiple-regression.html"><a href="multiple-regression.html#multiple-regression-1"><i class="fa fa-check"></i><b>12.2</b> Multiple regression</a><ul>
<li class="chapter" data-level="12.2.1" data-path="multiple-regression.html"><a href="multiple-regression.html#variable-space-interpretation"><i class="fa fa-check"></i><b>12.2.1</b> Variable space interpretation</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="multiple-regression.html"><a href="multiple-regression.html#subject-space-representation"><i class="fa fa-check"></i><b>12.3</b> Subject space representation</a><ul>
<li class="chapter" data-level="12.3.1" data-path="multiple-regression.html"><a href="multiple-regression.html#coefficient-of-determination-for-multiple-regression"><i class="fa fa-check"></i><b>12.3.1</b> Coefficient of determination for multiple regression</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="multiple-regression.html"><a href="multiple-regression.html#interpretting-multiple-regression"><i class="fa fa-check"></i><b>12.4</b> Interpretting Multiple Regression</a></li>
<li class="chapter" data-level="12.5" data-path="multiple-regression.html"><a href="multiple-regression.html#new-libraries-to-install"><i class="fa fa-check"></i><b>12.5</b> New Libraries to install</a></li>
<li class="chapter" data-level="12.6" data-path="multiple-regression.html"><a href="multiple-regression.html#libraries-4"><i class="fa fa-check"></i><b>12.6</b> Libraries</a></li>
<li class="chapter" data-level="12.7" data-path="multiple-regression.html"><a href="multiple-regression.html#examplar-data"><i class="fa fa-check"></i><b>12.7</b> Examplar data</a></li>
<li class="chapter" data-level="12.8" data-path="multiple-regression.html"><a href="multiple-regression.html#data-exploration"><i class="fa fa-check"></i><b>12.8</b> Data exploration</a></li>
<li class="chapter" data-level="12.9" data-path="multiple-regression.html"><a href="multiple-regression.html#d-plots"><i class="fa fa-check"></i><b>12.9</b> 3D Plots</a><ul>
<li class="chapter" data-level="12.9.1" data-path="multiple-regression.html"><a href="multiple-regression.html#scatterplot3d"><i class="fa fa-check"></i><b>12.9.1</b> scatterplot3d</a></li>
<li class="chapter" data-level="12.9.2" data-path="multiple-regression.html"><a href="multiple-regression.html#rgl"><i class="fa fa-check"></i><b>12.9.2</b> rgl</a></li>
</ul></li>
<li class="chapter" data-level="12.10" data-path="multiple-regression.html"><a href="multiple-regression.html#fitting-the-regression-model"><i class="fa fa-check"></i><b>12.10</b> Fitting the regression model</a><ul>
<li class="chapter" data-level="12.10.1" data-path="multiple-regression.html"><a href="multiple-regression.html#visualizing-the-regression-model-in-scatterplot3d"><i class="fa fa-check"></i><b>12.10.1</b> Visualizing the regression model in scatterplot3d</a></li>
</ul></li>
<li class="chapter" data-level="12.11" data-path="multiple-regression.html"><a href="multiple-regression.html#interpretting-the-regression-model"><i class="fa fa-check"></i><b>12.11</b> Interpretting the regression model</a></li>
<li class="chapter" data-level="12.12" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-vector-geometry-of-the-regression-model"><i class="fa fa-check"></i><b>12.12</b> Exploring the Vector Geometry of the Regression Model</a></li>
<li class="chapter" data-level="12.13" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-residuals-from-the-model-fit"><i class="fa fa-check"></i><b>12.13</b> Exploring the Residuals from the Model Fit</a></li>
<li class="chapter" data-level="12.14" data-path="multiple-regression.html"><a href="multiple-regression.html#an-alternate-model"><i class="fa fa-check"></i><b>12.14</b> An alternate model</a></li>
<li class="chapter" data-level="12.15" data-path="multiple-regression.html"><a href="multiple-regression.html#fitting-a-curvilinear-model-using-lm"><i class="fa fa-check"></i><b>12.15</b> Fitting a curvilinear model using lm()</a></li>
<li class="chapter" data-level="12.16" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-impact-of-nearly-collinear-predictors-on-regression"><i class="fa fa-check"></i><b>12.16</b> Exploring the impact of nearly collinear predictors on regression</a></li>
<li class="chapter" data-level="12.17" data-path="multiple-regression.html"><a href="multiple-regression.html#loess-models"><i class="fa fa-check"></i><b>12.17</b> LOESS Models</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="logistic-regression.html"><a href="logistic-regression.html"><i class="fa fa-check"></i><b>13</b> Logistic regression</a><ul>
<li class="chapter" data-level="13.1" data-path="logistic-regression.html"><a href="logistic-regression.html#a-web-app-to-explore-the-logistic-regression-equation"><i class="fa fa-check"></i><b>13.1</b> A web app to explore the logistic regression equation</a></li>
<li class="chapter" data-level="13.2" data-path="logistic-regression.html"><a href="logistic-regression.html#titanic-data-set"><i class="fa fa-check"></i><b>13.2</b> Titanic data set</a></li>
<li class="chapter" data-level="13.3" data-path="logistic-regression.html"><a href="logistic-regression.html#subsetting-the-data"><i class="fa fa-check"></i><b>13.3</b> Subsetting the data</a></li>
<li class="chapter" data-level="13.4" data-path="logistic-regression.html"><a href="logistic-regression.html#visualizing-survival-as-a-function-of-age"><i class="fa fa-check"></i><b>13.4</b> Visualizing survival as a function of age</a></li>
<li class="chapter" data-level="13.5" data-path="logistic-regression.html"><a href="logistic-regression.html#fitting-the-logistic-regression-model"><i class="fa fa-check"></i><b>13.5</b> Fitting the logistic regression model</a></li>
<li class="chapter" data-level="13.6" data-path="logistic-regression.html"><a href="logistic-regression.html#visualizing-the-logistic-regression"><i class="fa fa-check"></i><b>13.6</b> Visualizing the logistic regression</a><ul>
<li class="chapter" data-level="13.6.1" data-path="logistic-regression.html"><a href="logistic-regression.html#quick-and-easy-visualization"><i class="fa fa-check"></i><b>13.6.1</b> Quick and easy visualization</a></li>
</ul></li>
<li class="chapter" data-level="13.7" data-path="logistic-regression.html"><a href="logistic-regression.html#impact-of-sex-and-passenger-class-on-the-models"><i class="fa fa-check"></i><b>13.7</b> Impact of sex and passenger class on the models</a></li>
<li class="chapter" data-level="13.8" data-path="logistic-regression.html"><a href="logistic-regression.html#fitting-multiple-models-based-on-groupings-use-dplyrdo"><i class="fa fa-check"></i><b>13.8</b> Fitting multiple models based on groupings use <code>dplyr::do</code></a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html"><i class="fa fa-check"></i><b>14</b> Principal Components Analysis</a><ul>
<li class="chapter" data-level="14.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#libraries-5"><i class="fa fa-check"></i><b>14.1</b> Libraries</a></li>
<li class="chapter" data-level="14.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#matrices-as-linear-transformations"><i class="fa fa-check"></i><b>14.2</b> Matrices as linear transformations</a></li>
<li class="chapter" data-level="14.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#eigenanalysis-in-r"><i class="fa fa-check"></i><b>14.3</b> Eigenanalysis in R</a></li>
<li class="chapter" data-level="14.4" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#principal-components-analysis-in-r"><i class="fa fa-check"></i><b>14.4</b> Principal Components Analysis in R</a><ul>
<li class="chapter" data-level="14.4.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#bioenv-dataset"><i class="fa fa-check"></i><b>14.4.1</b> Bioenv dataset</a></li>
<li class="chapter" data-level="14.4.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pca-of-the-bioenv-dataset"><i class="fa fa-check"></i><b>14.4.2</b> PCA of the Bioenv dataset</a></li>
<li class="chapter" data-level="14.4.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#calculating-factor-loadings"><i class="fa fa-check"></i><b>14.4.3</b> Calculating Factor Loadings</a></li>
</ul></li>
<li class="chapter" data-level="14.5" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#drawing-figures-to-represent-pca"><i class="fa fa-check"></i><b>14.5</b> Drawing Figures to Represent PCA</a><ul>
<li class="chapter" data-level="14.5.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pc-score-plots"><i class="fa fa-check"></i><b>14.5.1</b> PC Score Plots</a></li>
<li class="chapter" data-level="14.5.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#simultaneous-depiction-of-observations-and-variables-in-the-pc-space"><i class="fa fa-check"></i><b>14.5.2</b> Simultaneous Depiction of Observations and Variables in the PC Space</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="15" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html"><i class="fa fa-check"></i><b>15</b> Singular Value Decomposition</a><ul>
<li class="chapter" data-level="15.1" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#libraries-6"><i class="fa fa-check"></i><b>15.1</b> Libraries</a></li>
<li class="chapter" data-level="15.2" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#svd-in-r"><i class="fa fa-check"></i><b>15.2</b> SVD in R</a></li>
<li class="chapter" data-level="15.3" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#writing-our-own-pca-function-using-svd"><i class="fa fa-check"></i><b>15.3</b> Writing our own PCA function using SVD</a></li>
<li class="chapter" data-level="15.4" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#writing-a-biplot-function"><i class="fa fa-check"></i><b>15.4</b> Writing a biplot function</a></li>
<li class="chapter" data-level="15.5" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#data-compression-and-noise-filtering-using-svd"><i class="fa fa-check"></i><b>15.5</b> Data compression and noise filtering using SVD</a><ul>
<li class="chapter" data-level="15.5.1" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#example-image"><i class="fa fa-check"></i><b>15.5.1</b> Example image</a></li>
</ul></li>
<li class="chapter" data-level="15.6" data-path="singular-value-decomposition.html"><a href="singular-value-decomposition.html#image-compression-using-svd-in-r"><i class="fa fa-check"></i><b>15.6</b> Image compression using SVD in R</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Biology 723: Statistical Computing for Biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="principal-components-analysis" class="section level1">
<h1><span class="header-section-number">Chapter 14</span> Principal Components Analysis</h1>
<div id="libraries-5" class="section level2">
<h2><span class="header-section-number">14.1</span> Libraries</h2>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)
<span class="kw">library</span>(broom)
<span class="kw">library</span>(GGally)
<span class="kw">library</span>(cowplot)</code></pre>
</div>
<div id="matrices-as-linear-transformations" class="section level2">
<h2><span class="header-section-number">14.2</span> Matrices as linear transformations</h2>
<p>In lecture we introduced the notion that pre-multiplying a vector, <span class="math inline">\(x\)</span>, by a matrix <span class="math inline">\(\mathbf{A}\)</span> represents a linear transformation of <span class="math inline">\(x\)</span>. Let’s explore that visually for 2D vectors.</p>
<p>For illustration let’s generate a 2D vector representing the coordinates of points on the sine function.</p>
<pre class="sourceCode r"><code class="sourceCode r">sin.xy &lt;-<span class="st"> </span><span class="kw">data_frame</span>(<span class="dt">x =</span> <span class="kw">seq</span>(<span class="dv">0</span>,<span class="dv">2</span><span class="op">*</span>pi, <span class="dt">length.out =</span> <span class="dv">50</span>) <span class="op">-</span><span class="st"> </span>pi,
                     <span class="dt">y =</span> <span class="kw">sin</span>(x))

<span class="kw">ggplot</span>(sin.xy, <span class="kw">aes</span>(<span class="dt">x =</span> x, <span class="dt">y=</span> y)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>()</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-419-1.png" width="576" /></p>
<p>Let’s start with some of the transformations we discussed in lecture. First we look at reflection about the x-axis:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># this matrix represents reflection about the x-axis</span>
A &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">1</span>,  <span class="dv">0</span>,
              <span class="dv">0</span>, <span class="dv">-1</span>), <span class="dt">byrow=</span><span class="ot">TRUE</span>, <span class="dt">nrow=</span><span class="dv">2</span>)

<span class="co"># apply this transformation to each of the points in our vector `sin.xy`</span>
Ax &lt;-<span class="st"> </span>A <span class="op">%*%</span><span class="st"> </span><span class="kw">t</span>(sin.xy)

<span class="co"># creae a data frame from this transformation</span>
Ax.df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">t</span>(Ax)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">rename</span>(<span class="dt">x =</span> V1, <span class="dt">y =</span> V2)

<span class="co"># plot the transformed points</span>
<span class="kw">ggplot</span>(Ax.df, <span class="kw">aes</span>(<span class="dt">x =</span> x, <span class="dt">y =</span> y)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>()</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-420-1.png" width="576" /></p>
<p>If instead we apply a matrix that represents a shear in the x-axis we get:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># shear parallel to the x-axis</span>
A &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">1</span>, <span class="fl">1.5</span>, 
              <span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">byrow=</span><span class="ot">TRUE</span>, <span class="dt">nrow=</span><span class="dv">2</span>)

Ax &lt;-<span class="st"> </span>A <span class="op">%*%</span><span class="st"> </span><span class="kw">t</span>(sin.xy)

Ax.df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">t</span>(Ax)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">rename</span>(<span class="dt">x =</span> V1, <span class="dt">y =</span> V2)
<span class="kw">ggplot</span>(Ax.df, <span class="kw">aes</span>(<span class="dt">x =</span> x, <span class="dt">y=</span> y)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>()</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-421-1.png" width="576" /></p>
<p>And finally rotation:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="co"># rotation by pi/2 radians (90 degrees)</span>
A &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="kw">cos</span>(pi<span class="op">/</span><span class="dv">2</span>), <span class="op">-</span><span class="kw">sin</span>(pi<span class="op">/</span><span class="dv">2</span>), 
              <span class="kw">sin</span>(pi<span class="op">/</span><span class="dv">2</span>), <span class="kw">cos</span>(pi<span class="op">/</span><span class="dv">2</span>)), <span class="dt">byrow=</span><span class="ot">TRUE</span>, <span class="dt">nrow=</span><span class="dv">2</span>)

Ax &lt;-<span class="st"> </span>A <span class="op">%*%</span><span class="st"> </span><span class="kw">t</span>(sin.xy)

Ax.df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">t</span>(Ax)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">rename</span>(<span class="dt">x =</span> V1, <span class="dt">y =</span> V2)
<span class="kw">ggplot</span>(Ax.df, <span class="kw">aes</span>(<span class="dt">x =</span> x, <span class="dt">y=</span> y)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>()</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-422-1.png" width="576" /></p>
</div>
<div id="eigenanalysis-in-r" class="section level2">
<h2><span class="header-section-number">14.3</span> Eigenanalysis in R</h2>
<p>As we discussed in lecture, the eigenvectors of a square matrix, <span class="math inline">\(\mathbf{A}\)</span>, point in the directions that are unchanged by the transformation specified by <span class="math inline">\(\mathbf{A}\)</span>.</p>
<p>Let’s start with yet another transformation matrix, <span class="math inline">\(\mathbf{A}\)</span>, the effects of which are illustrated below:</p>
<pre class="sourceCode r"><code class="sourceCode r">A &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="fl">1.0</span>, <span class="fl">1.5</span>,
              <span class="dv">0</span>,   <span class="fl">2.0</span>), <span class="dt">byrow=</span><span class="ot">TRUE</span>, <span class="dt">nrow=</span><span class="dv">2</span>)

Ax &lt;-<span class="st"> </span>A <span class="op">%*%</span><span class="st"> </span><span class="kw">t</span>(sin.xy)
Ax.df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">t</span>(Ax)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">rename</span>(<span class="dt">x =</span> V1, <span class="dt">y =</span> V2)

<span class="kw">ggplot</span>(<span class="dt">mapping=</span><span class="kw">aes</span>(<span class="dt">x =</span> x, <span class="dt">y =</span> y)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">data =</span> sin.xy, <span class="dt">shape=</span><span class="dv">16</span>, <span class="dt">alpha=</span><span class="fl">0.5</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">data =</span> Ax.df, <span class="dt">shape=</span><span class="dv">17</span>, <span class="dt">alpha=</span><span class="fl">0.5</span>, <span class="dt">color=</span><span class="st">&#39;red&#39;</span>) </code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-423-1.png" width="576" /></p>
<p>This transformation is a combination of stretching (dilation) in the Y-axis, and shear parallel to the x-axis.</p>
<p>The <code>eigen()</code> function computes the eigenvalues and eigenvectors of a square matrix:</p>
<pre class="sourceCode r"><code class="sourceCode r">eigen.A &lt;-<span class="st"> </span><span class="kw">eigen</span>(A)
eigen.A
<span class="co">## eigen() decomposition</span>
<span class="co">## $values</span>
<span class="co">## [1] 2 1</span>
<span class="co">## </span>
<span class="co">## $vectors</span>
<span class="co">##           [,1] [,2]</span>
<span class="co">## [1,] 0.8320503    1</span>
<span class="co">## [2,] 0.5547002    0</span></code></pre>
<p>The following relationships relate <span class="math inline">\(\mathbf{A}\)</span> to it’s eigenvectors and eigenvalues:</p>
<p><span class="math display">\[\mathbf{V}^{-1}\mathbf{A}\mathbf{V}  =  \mathbf{D} \]</span></p>
<p><span class="math display">\[\mathbf{A}  =  \mathbf{V}\mathbf{D}\mathbf{V}^{-1}\]</span></p>
<p>where <span class="math inline">\(\mathbf{V}\)</span> is a matrix where the columns represent the eigenvectors, and <span class="math inline">\(\mathbf{D}\)</span> is a diagonal matrix of eigenvalues.</p>
<p>Let’s confirm that those relationships hold for our example:</p>
<pre class="sourceCode r"><code class="sourceCode r">V &lt;-<span class="st"> </span>eigen.A<span class="op">$</span>vectors
D &lt;-<span class="st"> </span><span class="kw">diag</span>(eigen.A<span class="op">$</span>values) <span class="co"># diagonal matrix of eigenvalues</span>
Vinv &lt;-<span class="st"> </span><span class="kw">solve</span>(V)

V <span class="op">%*%</span><span class="st"> </span>D <span class="op">%*%</span><span class="st"> </span>Vinv  <span class="co"># reconstruct our original matrix (see lecture slides)</span>
<span class="co">##      [,1] [,2]</span>
<span class="co">## [1,]    1  1.5</span>
<span class="co">## [2,]    0  2.0</span></code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">all.equal</span>(Vinv <span class="op">%*%</span><span class="st"> </span>A <span class="op">%*%</span><span class="st"> </span>V, D) <span class="co"># test &#39;near equality&#39;</span>
<span class="co">## [1] TRUE</span></code></pre>
<pre class="sourceCode r"><code class="sourceCode r">V[,<span class="dv">1</span>] <span class="op">%*%</span><span class="st"> </span>V[,<span class="dv">2</span>] <span class="co"># note that the eigenvectors are NOT orthogonal. Why?</span>
<span class="co">##           [,1]</span>
<span class="co">## [1,] 0.8320503</span></code></pre>
<p>Since we’re dealing with a 2D transformation matrix, we can easily calculate the slopes of the eigenvectors:</p>
<pre class="sourceCode r"><code class="sourceCode r">eigenvec.A.slope1 =<span class="st"> </span>eigen.A<span class="op">$</span>vectors[<span class="dv">2</span>,<span class="dv">1</span>]<span class="op">/</span>eigen.A<span class="op">$</span>vectors[<span class="dv">1</span>,<span class="dv">1</span>]
eigenvec.A.slope2 =<span class="st"> </span>eigen.A<span class="op">$</span>vectors[<span class="dv">2</span>,<span class="dv">2</span>]<span class="op">/</span>eigen.A<span class="op">$</span>vectors[<span class="dv">1</span>,<span class="dv">2</span>]</code></pre>
<p>Now we plot the eigenvectors to show the directions that are unaffected by the transformation represented by the matrix <span class="math inline">\(\mathbf{A}\)</span>:</p>
<pre class="sourceCode r"><code class="sourceCode r">
<span class="co"># Note that the slopes can be infinite (inf), indicating a vertical eigenvector (parallel to y-axis)</span>
<span class="co"># this function is to draw the eigenvectors correctly regardless of the slope</span>
geom_ab_or_vline &lt;-<span class="st"> </span><span class="cf">function</span>(slope, intercept, ...) {
  <span class="cf">if</span> (<span class="kw">is.infinite</span>(slope)) {
    <span class="kw">return</span>(<span class="kw">geom_vline</span>(<span class="dt">yintercept =</span> intercept, ...))
  } <span class="cf">else</span> {
    <span class="kw">return</span>(<span class="kw">geom_abline</span>(<span class="dt">slope=</span>slope, <span class="dt">intercept=</span>intercept, ...))
  }
}


<span class="kw">ggplot</span>(<span class="dt">mapping=</span><span class="kw">aes</span>(<span class="dt">x =</span> x, <span class="dt">y =</span> y)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">data =</span> sin.xy, <span class="dt">alpha=</span><span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">data =</span> Ax.df, <span class="dt">color=</span><span class="st">&#39;red&#39;</span>, <span class="dt">alpha=</span><span class="fl">0.5</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_ab_or_vline</span>(<span class="dt">slope =</span> eigenvec.A.slope1,<span class="dt">intercept=</span><span class="dv">0</span>,
              <span class="dt">color=</span><span class="st">&#39;red&#39;</span>,<span class="dt">linetype=</span><span class="st">&#39;dashed&#39;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_ab_or_vline</span>(<span class="dt">slope =</span> eigenvec.A.slope2,<span class="dt">intercept=</span><span class="dv">0</span>,
              <span class="dt">color=</span><span class="st">&#39;red&#39;</span>,<span class="dt">linetype=</span><span class="st">&#39;dashed&#39;</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">coord_fixed</span>()</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-429-1.png" width="576" /></p>
<p>Referring back to the eigenvectors, we see that it’s the second eigenvector that is represented by the horizontal line:</p>
<pre class="sourceCode r"><code class="sourceCode r">eigen.A<span class="op">$</span>vectors
<span class="co">##           [,1] [,2]</span>
<span class="co">## [1,] 0.8320503    1</span>
<span class="co">## [2,] 0.5547002    0</span></code></pre>
<p>The implies that lines pointing in the horizontal are unchanged in the direction by the transformation represented by A. To drive home this point, here’s another configuration of points and the corresponding points under the same transformation:</p>
<pre class="sourceCode r"><code class="sourceCode r">horiz.x &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>,<span class="dt">by=</span><span class="fl">0.1</span>)
horiz.y &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">0</span>, <span class="kw">length</span>(horiz.x))

horiz.x2 &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="op">-</span><span class="fl">0.5</span>,<span class="fl">0.5</span>,<span class="dt">by=</span><span class="fl">0.1</span>)
horiz.y2 &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="fl">0.5</span>, <span class="kw">length</span>(horiz.x2))

horiz.x3 &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="op">-</span><span class="fl">0.5</span>,<span class="fl">0.5</span>,<span class="dt">by=</span><span class="fl">0.1</span>)
horiz.y3 &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="op">-</span><span class="fl">0.5</span>, <span class="kw">length</span>(horiz.x3))

vert.y &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>,<span class="dt">by=</span><span class="fl">0.1</span>)
vert.x &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">0</span>, <span class="kw">length</span>(vert.y))

x &lt;-<span class="st"> </span><span class="kw">c</span>(horiz.x, horiz.x2, horiz.x3, vert.x)
y &lt;-<span class="st"> </span><span class="kw">c</span>(horiz.y, horiz.y2, horiz.y3, vert.y)
cross.df &lt;-<span class="st"> </span><span class="kw">data_frame</span>(<span class="dt">x =</span> x, <span class="dt">y =</span> y)

Ax &lt;-<span class="st"> </span>A <span class="op">%*%</span><span class="st"> </span><span class="kw">t</span>(cross.df)
Ax.df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">t</span>(Ax)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">rename</span>(<span class="dt">x =</span> V1, <span class="dt">y =</span> V2)


plot.A &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw">ggplot</span>(cross.df, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x=</span>x,<span class="dt">y=</span>y)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span>
<span class="st">  </span><span class="kw">lims</span>(<span class="dt">x=</span><span class="kw">c</span>(<span class="op">-</span><span class="fl">1.5</span>,<span class="fl">1.5</span>), <span class="dt">y=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">2</span>)) <span class="op">+</span>
<span class="st">  </span><span class="kw">coord_fixed</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;X&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Y&quot;</span>, <span class="dt">title=</span><span class="st">&quot;Points prior to transformation&quot;</span>)
  
plot.B &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw">ggplot</span>(Ax.df, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x=</span>x,<span class="dt">y=</span>y)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">color=</span><span class="st">&#39;red&#39;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">lims</span>(<span class="dt">x=</span><span class="kw">c</span>(<span class="op">-</span><span class="fl">1.5</span>,<span class="fl">1.5</span>), <span class="dt">y=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>,<span class="dv">2</span>)) <span class="op">+</span>
<span class="st">  </span><span class="kw">coord_fixed</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;X&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Y&quot;</span>, <span class="dt">title=</span><span class="st">&quot;Points after transformation&quot;</span>)
  
<span class="kw">plot_grid</span>(plot.A, plot.B)</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-431-1.png" width="576" /></p>
</div>
<div id="principal-components-analysis-in-r" class="section level2">
<h2><span class="header-section-number">14.4</span> Principal Components Analysis in R</h2>
<p>There are two functions in R for carrying out PCA - <code>princomp()</code> and
<code>prcomp()</code>. The <code>princomp()</code> function uses the
<code>eigen()</code> function to carry out the analysis on the covariance
matrix or correlation matrix, while  carries out an
equivalent analysis, starting from a data matrix, using a technique called
singular value decomposition (SVD). The SVD routine has greater numerical
accuracy, so the <code>prcomp()</code> function should generally be preferred. The
<code>princomp()</code> function is useful when you don’t have access to
the original data, but you do have a covariance or correlation matrix (a
frequent situation when re-analyzing data from the literature). We’ll concentrate on using the <code>prcomp()</code> function.</p>
<div id="bioenv-dataset" class="section level3">
<h3><span class="header-section-number">14.4.1</span> Bioenv dataset</h3>
<p>To demonstrate PCA we’ll use a dataset called `bioenv.txt’ (see class wiki), obtained from a book called “Biplots in Practice” (M. Greenacre, 2010). Here is Greenacre’s description of the dataset:</p>
<blockquote>
<p>The context is in marine biology and the data consist of two sets of variables observed at the same locations on the sea-bed: the first is a set of biological variables, the counts of five groups of species, and the second is a set of four environmental variables. The data set, called “bioenv”, is shown in Exhibit 2.1. The species groups are abbreviated as “a” to “e”. The environmental variables are “pollution”, a composite index of pollution combining measurements of heavy metal concentrations and hydrocarbons; depth, the depth in metres of the sea-bed where the sample was taken; “temperature”, the temperature of the water at the sampling point; and “sediment”, a classification of the substrate of the sample into one of three sediment categories.</p>
</blockquote>
<p>We’ll start by reading the <code>bioenv.txt</code> data set from the Github repository:</p>
<pre class="sourceCode r"><code class="sourceCode r">bioenv &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&#39;https://github.com/Bio723-class/example-datasets/raw/master/bioenv.txt&#39;</span>) 
<span class="co">## Parsed with column specification:</span>
<span class="co">## cols(</span>
<span class="co">##   X1 = col_character(),</span>
<span class="co">##   a = col_double(),</span>
<span class="co">##   b = col_double(),</span>
<span class="co">##   c = col_double(),</span>
<span class="co">##   d = col_double(),</span>
<span class="co">##   e = col_double(),</span>
<span class="co">##   Pollution = col_double(),</span>
<span class="co">##   Depth = col_double(),</span>
<span class="co">##   Temperature = col_double(),</span>
<span class="co">##   Sediment = col_character()</span>
<span class="co">## )</span>
<span class="kw">names</span>(bioenv)
<span class="co">##  [1] &quot;X1&quot;          &quot;a&quot;           &quot;b&quot;           &quot;c&quot;           &quot;d&quot;          </span>
<span class="co">##  [6] &quot;e&quot;           &quot;Pollution&quot;   &quot;Depth&quot;       &quot;Temperature&quot; &quot;Sediment&quot;</span></code></pre>
<p>Notice that the first column got assigned the generic name <code>X1</code>. This is because there is a missing column header in the <code>bioenv.txt</code> file. This first column corresponds to the sampling sites. Before we move on let’s give this column a more meaningful name:</p>
<pre class="sourceCode r"><code class="sourceCode r">bioenv &lt;-
<span class="st">  </span>bioenv <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">rename</span>(<span class="dt">Site =</span> X1)

<span class="kw">names</span>(bioenv)
<span class="co">##  [1] &quot;Site&quot;        &quot;a&quot;           &quot;b&quot;           &quot;c&quot;           &quot;d&quot;          </span>
<span class="co">##  [6] &quot;e&quot;           &quot;Pollution&quot;   &quot;Depth&quot;       &quot;Temperature&quot; &quot;Sediment&quot;</span></code></pre>
<p>The columns labeled <code>a</code> to <code>e</code> contain the counts of the five species at each site, while the remaining columns give additional information about the physical properties of each sampling site. For our purposes today we’ll confine our attention to the abundance data.</p>
<pre class="sourceCode r"><code class="sourceCode r">abundance &lt;-<span class="st"> </span>
<span class="st">  </span>bioenv <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(Site, a, b, c, d, e)

<span class="kw">head</span>(abundance)
<span class="co">## # A tibble: 6 x 6</span>
<span class="co">##   Site      a     b     c     d     e</span>
<span class="co">##   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">## 1 s1        0     2     9    14     2</span>
<span class="co">## 2 s2       26     4    13    11     0</span>
<span class="co">## 3 s3        0    10     9     8     0</span>
<span class="co">## 4 s4        0     0    15     3     0</span>
<span class="co">## 5 s5       13     5     3    10     7</span>
<span class="co">## 6 s6       31    21    13    16     5</span></code></pre>
<p>The data is currently in a “wide” format. For the purposes of plotting it will be more convenient to generate a “long” version of the data using functions from the tidyr library (see the Data Wrangling chapter).</p>
<pre class="sourceCode r"><code class="sourceCode r">long.abundance &lt;-
<span class="st">  </span>abundance <span class="op">%&gt;%</span>
<span class="st">  </span>tidyr<span class="op">::</span><span class="kw">gather</span>(Species, Count, <span class="op">-</span>Site)

<span class="kw">head</span>(long.abundance)
<span class="co">## # A tibble: 6 x 3</span>
<span class="co">##   Site  Species Count</span>
<span class="co">##   &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;</span>
<span class="co">## 1 s1    a           0</span>
<span class="co">## 2 s2    a          26</span>
<span class="co">## 3 s3    a           0</span>
<span class="co">## 4 s4    a           0</span>
<span class="co">## 5 s5    a          13</span>
<span class="co">## 6 s6    a          31</span></code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(long.abundance, <span class="kw">aes</span>(<span class="dt">x =</span> Species, <span class="dt">y =</span> Count)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_boxplot</span>() <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Species&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Count&quot;</span>, <span class="dt">title=</span><span class="st">&quot;Distribution of</span><span class="ch">\n</span><span class="st">Species Counts per Site&quot;</span>)</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-436-1.png" width="576" /></p>
<p>From the boxplot it looks like the counts for species `e’ are smaller on average, and less variable. The mean and variance functions confirm that.</p>
<pre class="sourceCode r"><code class="sourceCode r">long.abundance <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(Species) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="kw">mean</span>(Count), <span class="kw">var</span>(Count))
<span class="co">## # A tibble: 5 x 3</span>
<span class="co">##   Species `mean(Count)` `var(Count)`</span>
<span class="co">##   &lt;chr&gt;           &lt;dbl&gt;        &lt;dbl&gt;</span>
<span class="co">## 1 a              13.47        157.6 </span>
<span class="co">## 2 b               8.733        83.44</span>
<span class="co">## 3 c               8.4          73.63</span>
<span class="co">## 4 d              10.9          44.44</span>
<span class="co">## 5 e               2.967        15.69</span></code></pre>
<p>A correlation matrix suggests weak to moderate associations between the variables:</p>
<pre class="sourceCode r"><code class="sourceCode r">abundance.only &lt;-
<span class="st">  </span>abundance <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>Site) <span class="co"># drop the Site column</span>
  
<span class="kw">cor</span>(abundance.only)
<span class="co">##            a           b           c            d            e</span>
<span class="co">## a  1.0000000  0.67339954 -0.23992888  0.358192050  0.273522301</span>
<span class="co">## b  0.6733995  1.00000000 -0.08041947  0.501834036  0.036914702</span>
<span class="co">## c -0.2399289 -0.08041947  1.00000000  0.081504483 -0.343540453</span>
<span class="co">## d  0.3581921  0.50183404  0.08150448  1.000000000 -0.004048517</span>
<span class="co">## e  0.2735223  0.03691470 -0.34354045 -0.004048517  1.000000000</span></code></pre>
<p>However a scatterplot matrix generated by the <code>GGally::ggapirs()</code> function suggests that many of the relationships have a strong non-linear element.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggpairs</span>(abundance.only)</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-439-1.png" width="576" /></p>
</div>
<div id="pca-of-the-bioenv-dataset" class="section level3">
<h3><span class="header-section-number">14.4.2</span> PCA of the Bioenv dataset</h3>
<p>Linearity is not a requirement for PCA, as it’s simply a rigid rotation of the original data. So we’ll continue with our analysis after taking a moment to read the help on the <code>prcomp()</code> function that is used to carry-out PCA in R.</p>
<pre class="sourceCode r"><code class="sourceCode r">abundance.pca &lt;-<span class="st"> </span><span class="kw">prcomp</span>(abundance.only, <span class="dt">center=</span><span class="ot">TRUE</span>, <span class="dt">retx=</span><span class="ot">TRUE</span>)
    <span class="co"># center=TRUE mean centers the data</span>
    <span class="co"># retx=TRUE returns the PC scores</span>
    <span class="co"># if you want to do PCA on the correlation matrix set scale.=TRUE</span>
    <span class="co">#    -- notice the period after scale!</span>

<span class="kw">summary</span>(abundance.pca)
<span class="co">## Importance of components:</span>
<span class="co">##                            PC1    PC2    PC3     PC4     PC5</span>
<span class="co">## Standard deviation     14.8653 8.8149 6.2193 5.03477 3.48231</span>
<span class="co">## Proportion of Variance  0.5895 0.2073 0.1032 0.06763 0.03235</span>
<span class="co">## Cumulative Proportion   0.5895 0.7968 0.9000 0.96765 1.00000</span></code></pre>
<p>We see that approximately 59% of the variance in the data is capture by the first PC, and approximately 90% by the first three PCs.</p>
<p>Let’s compare the values return by PCA to what we would get if we carried out eigenanalysis of the covariance matrix that corresponds to our data.</p>
<p>First the list object return by <code>prcomp()</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r">abundance.pca
<span class="co">## Standard deviations (1, .., p=5):</span>
<span class="co">## [1] 14.865306  8.814912  6.219250  5.034774  3.482308</span>
<span class="co">## </span>
<span class="co">## Rotation (n x k) = (5 x 5):</span>
<span class="co">##           PC1         PC2         PC3         PC4         PC5</span>
<span class="co">## a  0.81064462  0.07052882 -0.53108427  0.18442140 -0.14771336</span>
<span class="co">## b  0.51264394 -0.27799671  0.47711910 -0.63418946  0.17342177</span>
<span class="co">## c -0.16235135 -0.88665551 -0.40897655 -0.01149647  0.14173943</span>
<span class="co">## d  0.22207108 -0.31665237  0.56250980  0.72941223 -0.04422938</span>
<span class="co">## e  0.06616623  0.17696554 -0.08141111  0.17781482  0.96231977</span></code></pre>
<p>And now the corresponding values returned by eigenanaysis of the covariance matrix generated from the abundance data:</p>
<pre class="sourceCode r"><code class="sourceCode r">eig.abundance &lt;-<span class="st"> </span><span class="kw">eigen</span>(<span class="kw">cov</span>(abundance.only))
eig.abundance<span class="op">$</span>vectors  <span class="co"># compare to rotation matrix of PCA</span>
<span class="co">##             [,1]        [,2]        [,3]        [,4]        [,5]</span>
<span class="co">## [1,]  0.81064462 -0.07052882  0.53108427  0.18442140 -0.14771336</span>
<span class="co">## [2,]  0.51264394  0.27799671 -0.47711910 -0.63418946  0.17342177</span>
<span class="co">## [3,] -0.16235135  0.88665551  0.40897655 -0.01149647  0.14173943</span>
<span class="co">## [4,]  0.22207108  0.31665237 -0.56250980  0.72941223 -0.04422938</span>
<span class="co">## [5,]  0.06616623 -0.17696554  0.08141111  0.17781482  0.96231977</span>
<span class="kw">sqrt</span>(eig.abundance<span class="op">$</span>values) <span class="co"># compare to sdev of PCA</span>
<span class="co">## [1] 14.865306  8.814912  6.219250  5.034774  3.482308</span></code></pre>
<p>Notice that the <code>rotation</code> object returned by the <code>prcomp()</code> represents the scaled eigenvectors (scaled to have length 1). The standard deviations of the PCA are the square roots of the eigenvalues of the covariance matrix.</p>
</div>
<div id="calculating-factor-loadings" class="section level3">
<h3><span class="header-section-number">14.4.3</span> Calculating Factor Loadings</h3>
<p>Let’s calculate the “factor loadings” associated with the PCs:</p>
<pre class="sourceCode r"><code class="sourceCode r">V &lt;-<span class="st"> </span>abundance.pca<span class="op">$</span>rotation <span class="co"># eigenvectors</span>
L &lt;-<span class="st"> </span><span class="kw">diag</span>(abundance.pca<span class="op">$</span>sdev) <span class="co"># diag mtx w/sqrts of eigenvalues on diag.</span>

abundance.loadings &lt;-<span class="st"> </span>V <span class="op">%*%</span><span class="st"> </span>L
abundance.loadings
<span class="co">##         [,1]       [,2]       [,3]        [,4]       [,5]</span>
<span class="co">## a 12.0504801  0.6217053 -3.3029460  0.92852016 -0.5143835</span>
<span class="co">## b  7.6206090 -2.4505164  2.9673232 -3.19300085  0.6039081</span>
<span class="co">## c -2.4134024 -7.8157898 -2.5435276 -0.05788214  0.4935804</span>
<span class="co">## d  3.3011545 -2.7912626  3.4983893  3.67242602 -0.1540203</span>
<span class="co">## e  0.9835813  1.5599356 -0.5063161  0.89525751  3.3510942</span></code></pre>
<p>The magnitude of the factor loadings is what you want to focus on. For example, species <code>a</code> and <code>b</code> contribute most to the first PC, while species <code>c</code> has the largest influence on PC2.</p>
<p>You can think of the factor loadings, as defined above, as the components (i.e lengths of the projected vectors) of the original variables with respect to the PC basis vectors. Since vector length is proportional to the standard deviation of the variables they represent, you can think of the loadings as giving the standard deviation of the original variables with respect the PC axes. This implies that the loadings squared sum to the total variance in the original data, as illustrated below.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sum</span>(abundance.loadings<span class="op">**</span><span class="dv">2</span>)
<span class="co">## [1] 374.8345</span>

abundance.only <span class="op">%&gt;%</span><span class="st"> </span>purrr<span class="op">::</span><span class="kw">map_dbl</span>(var) <span class="op">%&gt;%</span><span class="st"> </span>sum
<span class="co">## [1] 374.8345</span></code></pre>
</div>
</div>
<div id="drawing-figures-to-represent-pca" class="section level2">
<h2><span class="header-section-number">14.5</span> Drawing Figures to Represent PCA</h2>
<div id="pc-score-plots" class="section level3">
<h3><span class="header-section-number">14.5.1</span> PC Score Plots</h3>
<p>The simplest PCA figure is to depict the PC scores, i.e. the projection of the observations into the space defined by the PC axes. Let’s make a figure with three subplots, depicting PC1 vs PC2, PC1 vs PC3, and PC2 vs. PC3.</p>
<pre class="sourceCode r"><code class="sourceCode r">pca.scores.df &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(abundance.pca<span class="op">$</span>x)

coord.system &lt;-<span class="st"> </span><span class="kw">coord_fixed</span>(<span class="dt">ratio=</span><span class="dv">1</span>, <span class="dt">xlim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">30</span>,<span class="dv">30</span>),<span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">30</span>,<span class="dv">30</span>))

pc1v2 &lt;-
<span class="st">  </span>pca.scores.df <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> PC1, <span class="dt">y=</span> PC2)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span>coord.system

pc1v3 &lt;-
<span class="st">  </span>pca.scores.df <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> PC1, <span class="dt">y=</span> PC3)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span>coord.system

pc2v3 &lt;-
<span class="st">  </span>pca.scores.df <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> PC2, <span class="dt">y=</span> PC3)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span>coord.system

cowplot<span class="op">::</span><span class="kw">plot_grid</span>(pc1v2, pc1v3, pc2v3, <span class="dt">align=</span><span class="st">&quot;hv&quot;</span>)</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-445-1.png" width="576" /></p>
<p>When plotting PC scores, it is <strong>very important</strong> to keep the aspect ratio with a value of 1, so that the distance between points in the plot is an accurate representation of the distance in the PC space. Note too that I used the <code>xlim</code> and <code>ylim</code> arguments to keep the axis limits the same in all plots; comparable scaling of axes is important when comparing plots. Also note the use of the <code>align="hv"</code> argument to <code>plot_grid()</code> to keep my plots the same size when combining them into a single figure.</p>
</div>
<div id="simultaneous-depiction-of-observations-and-variables-in-the-pc-space" class="section level3">
<h3><span class="header-section-number">14.5.2</span> Simultaneous Depiction of Observations and Variables in the PC Space</h3>
<p>Let’s return to our simple PC score plot. As we discussed above, the loadings are components of the original variables in the space of the PCs. This implies we can depict those loadings in the same PC basis that we use to depict the scores.</p>
<p>First let’s create a data frame with the loadings from the first two PCs as well as a column representation the variable names:</p>
<pre class="sourceCode r"><code class="sourceCode r">loadings<span class="fl">.1</span>and2 &lt;-
<span class="st">  </span><span class="kw">data.frame</span>(abundance.loadings[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">rename</span>(<span class="dt">PC1.loading =</span> X1, <span class="dt">PC2.loading =</span> X2) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">variable =</span> <span class="kw">row.names</span>(abundance.loadings))

loadings<span class="fl">.1</span>and2
<span class="co">## # A tibble: 5 x 3</span>
<span class="co">##   PC1.loading PC2.loading variable</span>
<span class="co">##         &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;   </span>
<span class="co">## 1     12.05        0.6217 a       </span>
<span class="co">## 2      7.621      -2.451  b       </span>
<span class="co">## 3     -2.413      -7.816  c       </span>
<span class="co">## 4      3.301      -2.791  d       </span>
<span class="co">## 5      0.9836      1.560  e</span></code></pre>
<p>With this data frame in hand, we can now draw a set of vector to represent the original variables projected into the subspace defined by PCs 1 and 2:</p>
<pre class="sourceCode r"><code class="sourceCode r">pc1v2.biplot &lt;-
<span class="st">  </span>pc1v2 <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_segment</span>(<span class="dt">data=</span>loadings<span class="fl">.1</span>and2,
               <span class="kw">aes</span>(<span class="dt">x =</span> <span class="dv">0</span>, <span class="dt">y =</span> <span class="dv">0</span>, <span class="dt">xend =</span> PC1.loading, <span class="dt">yend =</span> PC2.loading),
               <span class="dt">color=</span><span class="st">&#39;red&#39;</span>, 
               <span class="dt">arrow =</span> <span class="kw">arrow</span>(<span class="dt">angle=</span><span class="dv">15</span>, <span class="dt">length=</span><span class="kw">unit</span>(<span class="fl">0.1</span>,<span class="st">&quot;inches&quot;</span>))) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_text</span>(<span class="dt">data=</span>loadings<span class="fl">.1</span>and2,
            <span class="kw">aes</span>(<span class="dt">x =</span> PC1.loading, <span class="dt">y =</span> PC2.loading, <span class="dt">label=</span>variable),
            <span class="dt">color=</span><span class="st">&#39;red&#39;</span>, <span class="dt">nudge_x =</span> <span class="dv">1</span>, <span class="dt">nudge_y =</span> <span class="dv">1</span>)


pc1v2.biplot</code></pre>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-447-1.png" width="768" /></p>
<p>The figure above is called a “biplot”", as it simultaneously depicts both the observations and variables in the same space. From this biplot we can immediately see that variable <code>a</code> is highly correlated with PC1, but only weakly associated with PC2. Conversely, variable <code>c</code> is strongly correlated with PC2 but only weakly so with PC1. We can also approximate the correlations among the variables themselves – for example <code>b</code> and <code>d</code> are fairly strongly correlated, but weakly correlated with <code>c</code>. Keep in mind however that with respect to the relationships among the variables, this visualization is a 2D projection of a 5D space so the geometry is approximate.</p>
<p>The biplot is a generally useful tool for multivariate analyses and there are a number of different ways to define biplots. We’ll study biplots more formally in a few weeks after we’ve covered singular value decomposition.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="logistic-regression.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="singular-value-decomposition.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "section"
},
"toolbar": {
"position": "fixed"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
