<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 13 Principal Components Analysis | Biology 723: Statistical Computing for Biologists</title>
  <meta name="description" content="Hands-on materials for Bio 723 course at Duke University." />
  <meta name="generator" content="bookdown 0.32 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 13 Principal Components Analysis | Biology 723: Statistical Computing for Biologists" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 13 Principal Components Analysis | Biology 723: Statistical Computing for Biologists" />
  
  <meta name="twitter:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

<meta name="author" content="Paul M. Magwene" />


<meta name="date" content="2023-03-08" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="linear-regression-models-ii.html"/>
<link rel="next" href="canonical-variates-analysis.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#accessing-older-versions-of-the-course-notes"><i class="fa fa-check"></i><b>1.1</b> Accessing older versions of the course notes</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#how-to-use-these-lecture-notes"><i class="fa fa-check"></i><b>1.2</b> How to use these lecture notes</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html"><i class="fa fa-check"></i><b>2</b> Getting Started with R</a>
<ul>
<li class="chapter" data-level="2.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-r"><i class="fa fa-check"></i><b>2.1</b> What is R?</a></li>
<li class="chapter" data-level="2.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-rstudio"><i class="fa fa-check"></i><b>2.2</b> What is RStudio?</a></li>
<li class="chapter" data-level="2.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#entering-commands-in-the-console"><i class="fa fa-check"></i><b>2.3</b> Entering commands in the console</a></li>
<li class="chapter" data-level="2.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#comments"><i class="fa fa-check"></i><b>2.4</b> Comments</a></li>
<li class="chapter" data-level="2.5" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#using-r-as-a-calculator"><i class="fa fa-check"></i><b>2.5</b> Using R as a Calculator</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#common-mathematical-functions"><i class="fa fa-check"></i><b>2.5.1</b> Common mathematical functions</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#variable-assignment"><i class="fa fa-check"></i><b>2.6</b> Variable assignment</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#valid-variable-names"><i class="fa fa-check"></i><b>2.6.1</b> Valid variable names</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#data-types"><i class="fa fa-check"></i><b>2.7</b> Data types</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#numeric-data-types"><i class="fa fa-check"></i><b>2.7.1</b> Numeric data types</a></li>
<li class="chapter" data-level="2.7.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#logical-values"><i class="fa fa-check"></i><b>2.7.2</b> Logical values</a></li>
<li class="chapter" data-level="2.7.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#character-strings"><i class="fa fa-check"></i><b>2.7.3</b> Character strings</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#packages"><i class="fa fa-check"></i><b>2.8</b> Packages</a>
<ul>
<li class="chapter" data-level="2.8.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-console"><i class="fa fa-check"></i><b>2.8.1</b> Installing packages from the console</a></li>
<li class="chapter" data-level="2.8.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#install-the-tidyverse-package"><i class="fa fa-check"></i><b>2.8.2</b> Install the tidyverse package</a></li>
<li class="chapter" data-level="2.8.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-rstudio-dialog"><i class="fa fa-check"></i><b>2.8.3</b> Installing packages from the RStudio dialog</a></li>
<li class="chapter" data-level="2.8.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#loading-packages-with-the-library-function"><i class="fa fa-check"></i><b>2.8.4</b> Loading packages with the <code>library()</code> function</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#the-r-help-system"><i class="fa fa-check"></i><b>2.9</b> The R Help System</a>
<ul>
<li class="chapter" data-level="2.9.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#getting-help-from-the-console"><i class="fa fa-check"></i><b>2.9.1</b> Getting help from the console</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html"><i class="fa fa-check"></i><b>3</b> R Markdown and R Notebooks</a>
<ul>
<li class="chapter" data-level="3.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#r-notebooks"><i class="fa fa-check"></i><b>3.1</b> R Notebooks</a></li>
<li class="chapter" data-level="3.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#creating-an-r-notebook"><i class="fa fa-check"></i><b>3.2</b> Creating an R Notebook</a></li>
<li class="chapter" data-level="3.3" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#the-default-r-notebook-template"><i class="fa fa-check"></i><b>3.3</b> The default R Notebook template</a></li>
<li class="chapter" data-level="3.4" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-and-non-code-blocks"><i class="fa fa-check"></i><b>3.4</b> Code and Non-code blocks</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#non-code-blocks"><i class="fa fa-check"></i><b>3.4.1</b> Non-code blocks</a></li>
<li class="chapter" data-level="3.4.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-blocks"><i class="fa fa-check"></i><b>3.4.2</b> Code blocks</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-a-code-chunk"><i class="fa fa-check"></i><b>3.5</b> Running a code chunk</a></li>
<li class="chapter" data-level="3.6" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-all-code-chunks-above"><i class="fa fa-check"></i><b>3.6</b> Running all code chunks above</a></li>
<li class="chapter" data-level="3.7" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#knitting-r-markdown-to-html"><i class="fa fa-check"></i><b>3.7</b> “Knitting” R Markdown to HTML</a></li>
<li class="chapter" data-level="3.8" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#sharing-your-reproducible-r-notebook"><i class="fa fa-check"></i><b>3.8</b> Sharing your reproducible R Notebook</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-structures.html"><a href="data-structures.html"><i class="fa fa-check"></i><b>4</b> Data structures</a>
<ul>
<li class="chapter" data-level="4.1" data-path="data-structures.html"><a href="data-structures.html#vectors"><i class="fa fa-check"></i><b>4.1</b> Vectors</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="data-structures.html"><a href="data-structures.html#vector-arithmetic"><i class="fa fa-check"></i><b>4.1.1</b> Vector Arithmetic</a></li>
<li class="chapter" data-level="4.1.2" data-path="data-structures.html"><a href="data-structures.html#vector-recycling"><i class="fa fa-check"></i><b>4.1.2</b> Vector recycling</a></li>
<li class="chapter" data-level="4.1.3" data-path="data-structures.html"><a href="data-structures.html#simple-statistical-functions-for-numeric-vectors"><i class="fa fa-check"></i><b>4.1.3</b> Simple statistical functions for numeric vectors</a></li>
<li class="chapter" data-level="4.1.4" data-path="data-structures.html"><a href="data-structures.html#indexing-vectors"><i class="fa fa-check"></i><b>4.1.4</b> Indexing Vectors</a></li>
<li class="chapter" data-level="4.1.5" data-path="data-structures.html"><a href="data-structures.html#comparison-operators-applied-to-vectors"><i class="fa fa-check"></i><b>4.1.5</b> Comparison operators applied to vectors</a></li>
<li class="chapter" data-level="4.1.6" data-path="data-structures.html"><a href="data-structures.html#combining-indexing-and-comparison-of-vectors"><i class="fa fa-check"></i><b>4.1.6</b> Combining Indexing and Comparison of Vectors</a></li>
<li class="chapter" data-level="4.1.7" data-path="data-structures.html"><a href="data-structures.html#vector-manipulation"><i class="fa fa-check"></i><b>4.1.7</b> Vector manipulation</a></li>
<li class="chapter" data-level="4.1.8" data-path="data-structures.html"><a href="data-structures.html#vectors-from-regular-sequences"><i class="fa fa-check"></i><b>4.1.8</b> Vectors from regular sequences</a></li>
<li class="chapter" data-level="4.1.9" data-path="data-structures.html"><a href="data-structures.html#additional-functions-for-working-with-vectors"><i class="fa fa-check"></i><b>4.1.9</b> Additional functions for working with vectors</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="data-structures.html"><a href="data-structures.html#lists"><i class="fa fa-check"></i><b>4.2</b> Lists</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="data-structures.html"><a href="data-structures.html#length-and-type-of-lists"><i class="fa fa-check"></i><b>4.2.1</b> Length and type of lists</a></li>
<li class="chapter" data-level="4.2.2" data-path="data-structures.html"><a href="data-structures.html#indexing-lists"><i class="fa fa-check"></i><b>4.2.2</b> Indexing lists</a></li>
<li class="chapter" data-level="4.2.3" data-path="data-structures.html"><a href="data-structures.html#naming-list-elements"><i class="fa fa-check"></i><b>4.2.3</b> Naming list elements</a></li>
<li class="chapter" data-level="4.2.4" data-path="data-structures.html"><a href="data-structures.html#the-operator"><i class="fa fa-check"></i><b>4.2.4</b> The <code>$</code> operator</a></li>
<li class="chapter" data-level="4.2.5" data-path="data-structures.html"><a href="data-structures.html#changing-and-adding-lists-items"><i class="fa fa-check"></i><b>4.2.5</b> Changing and adding lists items</a></li>
<li class="chapter" data-level="4.2.6" data-path="data-structures.html"><a href="data-structures.html#combining-lists"><i class="fa fa-check"></i><b>4.2.6</b> Combining lists</a></li>
<li class="chapter" data-level="4.2.7" data-path="data-structures.html"><a href="data-structures.html#converting-lists-to-vectors"><i class="fa fa-check"></i><b>4.2.7</b> Converting lists to vectors</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="data-structures.html"><a href="data-structures.html#data-frames"><i class="fa fa-check"></i><b>4.3</b> Data frames</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="data-structures.html"><a href="data-structures.html#creating-a-data-frame"><i class="fa fa-check"></i><b>4.3.1</b> Creating a data frame</a></li>
<li class="chapter" data-level="4.3.2" data-path="data-structures.html"><a href="data-structures.html#type-and-class-for-data-frames"><i class="fa fa-check"></i><b>4.3.2</b> Type and class for data frames</a></li>
<li class="chapter" data-level="4.3.3" data-path="data-structures.html"><a href="data-structures.html#length-and-dimension-for-data-frames"><i class="fa fa-check"></i><b>4.3.3</b> Length and dimension for data frames</a></li>
<li class="chapter" data-level="4.3.4" data-path="data-structures.html"><a href="data-structures.html#indexing-and-accessing-data-frames"><i class="fa fa-check"></i><b>4.3.4</b> Indexing and accessing data frames</a></li>
<li class="chapter" data-level="4.3.5" data-path="data-structures.html"><a href="data-structures.html#logical-indexing-of-data-frames"><i class="fa fa-check"></i><b>4.3.5</b> Logical indexing of data frames</a></li>
<li class="chapter" data-level="4.3.6" data-path="data-structures.html"><a href="data-structures.html#adding-columns-to-a-data-frame"><i class="fa fa-check"></i><b>4.3.6</b> Adding columns to a data frame</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html"><i class="fa fa-check"></i><b>5</b> Functions and control flow statements</a>
<ul>
<li class="chapter" data-level="5.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#writing-your-own-functions"><i class="fa fa-check"></i><b>5.1</b> Writing your own functions</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#function-arguments"><i class="fa fa-check"></i><b>5.1.1</b> Function arguments</a></li>
<li class="chapter" data-level="5.1.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#writing-functions-with-optional-arguments"><i class="fa fa-check"></i><b>5.1.2</b> Writing functions with optional arguments</a></li>
<li class="chapter" data-level="5.1.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#putting-r-functions-in-scripts"><i class="fa fa-check"></i><b>5.1.3</b> Putting R functions in Scripts</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#control-flow-statements"><i class="fa fa-check"></i><b>5.2</b> Control flow statements</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#if-and-if-else-statements"><i class="fa fa-check"></i><b>5.2.1</b> <code>if</code> and <code>if-else</code> statements</a></li>
<li class="chapter" data-level="5.2.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#for-loops"><i class="fa fa-check"></i><b>5.2.2</b> for loops</a></li>
<li class="chapter" data-level="5.2.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#break-statement"><i class="fa fa-check"></i><b>5.2.3</b> <code>break</code> statement</a></li>
<li class="chapter" data-level="5.2.4" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#repeat-loops"><i class="fa fa-check"></i><b>5.2.4</b> <code>repeat</code> loops</a></li>
<li class="chapter" data-level="5.2.5" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#next-statement"><i class="fa fa-check"></i><b>5.2.5</b> <code>next</code> statement</a></li>
<li class="chapter" data-level="5.2.6" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#while-statements"><i class="fa fa-check"></i><b>5.2.6</b> while statements</a></li>
<li class="chapter" data-level="5.2.7" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#ifelse"><i class="fa fa-check"></i><b>5.2.7</b> <code>ifelse</code></a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#map-and-related-tools"><i class="fa fa-check"></i><b>5.3</b> <code>map</code> and related tools</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#basic-map"><i class="fa fa-check"></i><b>5.3.1</b> basic <code>map</code></a></li>
<li class="chapter" data-level="5.3.2" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#map_if-and-map_at"><i class="fa fa-check"></i><b>5.3.2</b> <code>map_if</code> and <code>map_at</code></a></li>
<li class="chapter" data-level="5.3.3" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#mapping-in-parallel-using-map2"><i class="fa fa-check"></i><b>5.3.3</b> mapping in parallel using <code>map2</code></a></li>
<li class="chapter" data-level="5.3.4" data-path="functions-and-control-flow-statements.html"><a href="functions-and-control-flow-statements.html#map-variants-that-return-vectors"><i class="fa fa-check"></i><b>5.3.4</b> <code>map</code> variants that return vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html"><i class="fa fa-check"></i><b>6</b> Introduction to ggplot2</a>
<ul>
<li class="chapter" data-level="6.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#loading-ggplot2"><i class="fa fa-check"></i><b>6.1</b> Loading ggplot2</a></li>
<li class="chapter" data-level="6.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#example-data-set-andersons-iris-data"><i class="fa fa-check"></i><b>6.2</b> Example data set: Anderson’s Iris Data</a></li>
<li class="chapter" data-level="6.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#template-for-single-layer-plots-in-ggplot2"><i class="fa fa-check"></i><b>6.3</b> Template for single layer plots in ggplot2</a></li>
<li class="chapter" data-level="6.4" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#an-aside-about-function-arguments"><i class="fa fa-check"></i><b>6.4</b> An aside about function arguments</a></li>
<li class="chapter" data-level="6.5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#strip-plots"><i class="fa fa-check"></i><b>6.5</b> Strip plots</a>
<ul>
<li class="chapter" data-level="6.5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#jittering-data"><i class="fa fa-check"></i><b>6.5.1</b> Jittering data</a></li>
<li class="chapter" data-level="6.5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-categorical-information"><i class="fa fa-check"></i><b>6.5.2</b> Adding categorical information</a></li>
<li class="chapter" data-level="6.5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#rotating-plot-coordinates"><i class="fa fa-check"></i><b>6.5.3</b> Rotating plot coordinates</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#histograms"><i class="fa fa-check"></i><b>6.6</b> Histograms</a>
<ul>
<li class="chapter" data-level="6.6.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#variations-on-histograms-when-considering-categorical-data"><i class="fa fa-check"></i><b>6.6.1</b> Variations on histograms when considering categorical data</a></li>
</ul></li>
<li class="chapter" data-level="6.7" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#faceting-to-depict-categorical-information"><i class="fa fa-check"></i><b>6.7</b> Faceting to depict categorical information</a></li>
<li class="chapter" data-level="6.8" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots"><i class="fa fa-check"></i><b>6.8</b> Density plots</a></li>
<li class="chapter" data-level="6.9" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-or-beanplot"><i class="fa fa-check"></i><b>6.9</b> Violin or Beanplot</a></li>
<li class="chapter" data-level="6.10" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplots"><i class="fa fa-check"></i><b>6.10</b> Boxplots</a></li>
<li class="chapter" data-level="6.11" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#building-complex-visualizations-with-layers"><i class="fa fa-check"></i><b>6.11</b> Building complex visualizations with layers</a></li>
<li class="chapter" data-level="6.12" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#useful-combination-plots"><i class="fa fa-check"></i><b>6.12</b> Useful combination plots</a>
<ul>
<li class="chapter" data-level="6.12.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplot-plus-strip-plot"><i class="fa fa-check"></i><b>6.12.1</b> Boxplot plus strip plot</a></li>
<li class="chapter" data-level="6.12.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#setting-shared-aesthetics"><i class="fa fa-check"></i><b>6.12.2</b> Setting shared aesthetics</a></li>
</ul></li>
<li class="chapter" data-level="6.13" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot-layers-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>6.13</b> ggplot layers can be assigned to variables</a>
<ul>
<li class="chapter" data-level="6.13.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-plot-plus-strip-plot"><i class="fa fa-check"></i><b>6.13.1</b> Violin plot plus strip plot</a></li>
</ul></li>
<li class="chapter" data-level="6.14" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-titles-and-tweaking-axis-labels"><i class="fa fa-check"></i><b>6.14</b> Adding titles and tweaking axis labels</a></li>
<li class="chapter" data-level="6.15" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot2-themes"><i class="fa fa-check"></i><b>6.15</b> ggplot2 themes</a>
<ul>
<li class="chapter" data-level="6.15.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#further-customization-with-ggplot2theme"><i class="fa fa-check"></i><b>6.15.1</b> Further customization with <code>ggplot2::theme</code></a></li>
</ul></li>
<li class="chapter" data-level="6.16" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#other-aspects-of-ggplots-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>6.16</b> Other aspects of ggplots can be assigned to variables</a></li>
<li class="chapter" data-level="6.17" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-plots"><i class="fa fa-check"></i><b>6.17</b> Bivariate plots</a>
<ul>
<li class="chapter" data-level="6.17.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#scatter-plots"><i class="fa fa-check"></i><b>6.17.1</b> Scatter plots</a></li>
<li class="chapter" data-level="6.17.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-a-trend-line-to-a-scatter-plot"><i class="fa fa-check"></i><b>6.17.2</b> Adding a trend line to a scatter plot</a></li>
</ul></li>
<li class="chapter" data-level="6.18" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-density-plots"><i class="fa fa-check"></i><b>6.18</b> Bivariate density plots</a></li>
<li class="chapter" data-level="6.19" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#combining-scatter-plots-and-density-plots-with-categorical-information"><i class="fa fa-check"></i><b>6.19</b> Combining Scatter Plots and Density Plots with Categorical Information</a></li>
<li class="chapter" data-level="6.20" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots-with-fill"><i class="fa fa-check"></i><b>6.20</b> Density plots with fill</a></li>
<li class="chapter" data-level="6.21" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#d-bin-and-hex-plots"><i class="fa fa-check"></i><b>6.21</b> 2D bin and hex plots</a></li>
<li class="chapter" data-level="6.22" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#the-cowplot-package"><i class="fa fa-check"></i><b>6.22</b> The <code>cowplot</code> package</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html"><i class="fa fa-check"></i><b>7</b> Introduction to <code>dplyr</code></a>
<ul>
<li class="chapter" data-level="7.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#libraries"><i class="fa fa-check"></i><b>7.1</b> Libraries</a></li>
<li class="chapter" data-level="7.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#reading-data-with-the-readr-package"><i class="fa fa-check"></i><b>7.2</b> Reading data with the <code>readr</code> package</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#reading-excel-files"><i class="fa fa-check"></i><b>7.2.1</b> Reading Excel files</a></li>
<li class="chapter" data-level="7.2.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#example-data-nc-births"><i class="fa fa-check"></i><b>7.2.2</b> Example data: NC Births</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#a-note-on-tibbles"><i class="fa fa-check"></i><b>7.3</b> A note on “tibbles”</a></li>
<li class="chapter" data-level="7.4" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#data-filtering-and-transformation-with-dplyr"><i class="fa fa-check"></i><b>7.4</b> Data filtering and transformation with <code>dplyr</code></a></li>
<li class="chapter" data-level="7.5" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#dplyrs-verbs"><i class="fa fa-check"></i><b>7.5</b> dplyr’s “verbs”</a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#select"><i class="fa fa-check"></i><b>7.5.1</b> <code>select</code></a></li>
<li class="chapter" data-level="7.5.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#filter"><i class="fa fa-check"></i><b>7.5.2</b> <code>filter</code></a></li>
<li class="chapter" data-level="7.5.3" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#mutate"><i class="fa fa-check"></i><b>7.5.3</b> <code>mutate</code></a></li>
<li class="chapter" data-level="7.5.4" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#arrange"><i class="fa fa-check"></i><b>7.5.4</b> <code>arrange</code></a></li>
<li class="chapter" data-level="7.5.5" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#summarize"><i class="fa fa-check"></i><b>7.5.5</b> <code>summarize</code></a></li>
<li class="chapter" data-level="7.5.6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#group_by"><i class="fa fa-check"></i><b>7.5.6</b> <code>group_by</code></a></li>
<li class="chapter" data-level="7.5.7" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#combining-grouping-and-summarizing"><i class="fa fa-check"></i><b>7.5.7</b> Combining grouping and summarizing</a></li>
<li class="chapter" data-level="7.5.8" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#scoped-variants-of-mutate-and-summarize"><i class="fa fa-check"></i><b>7.5.8</b> Scoped variants of <code>mutate</code> and <code>summarize</code></a></li>
<li class="chapter" data-level="7.5.9" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#combining-summarize-with-grouping-aesthetics-in-ggplot2"><i class="fa fa-check"></i><b>7.5.9</b> Combining summarize with grouping aesthetics in <code>ggplot2</code></a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#pipes"><i class="fa fa-check"></i><b>7.6</b> Pipes</a>
<ul>
<li class="chapter" data-level="7.6.1" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#install-and-load-magrittr"><i class="fa fa-check"></i><b>7.6.1</b> Install and load <code>magrittr</code></a></li>
<li class="chapter" data-level="7.6.2" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-basic-pipe-operator"><i class="fa fa-check"></i><b>7.6.2</b> The basic pipe operator</a></li>
<li class="chapter" data-level="7.6.3" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#an-example-without-pipes"><i class="fa fa-check"></i><b>7.6.3</b> An example without pipes</a></li>
<li class="chapter" data-level="7.6.4" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-same-example-using-pipes"><i class="fa fa-check"></i><b>7.6.4</b> The same example using pipes</a></li>
<li class="chapter" data-level="7.6.5" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#assigning-the-output-of-a-statement-involving-pipes-to-a-variable"><i class="fa fa-check"></i><b>7.6.5</b> Assigning the output of a statement involving pipes to a variable</a></li>
<li class="chapter" data-level="7.6.6" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#compound-assignment-pipe-operator"><i class="fa fa-check"></i><b>7.6.6</b> Compound assignment pipe operator</a></li>
<li class="chapter" data-level="7.6.7" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-dot-operator-with-pipes"><i class="fa fa-check"></i><b>7.6.7</b> The dot operator with pipes</a></li>
<li class="chapter" data-level="7.6.8" data-path="introduction-to-dplyr.html"><a href="introduction-to-dplyr.html#the-exposition-pipe-operator"><i class="fa fa-check"></i><b>7.6.8</b> The exposition pipe operator</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>8</b> Data wrangling</a>
<ul>
<li class="chapter" data-level="8.1" data-path="data-wrangling.html"><a href="data-wrangling.html#libraries-1"><i class="fa fa-check"></i><b>8.1</b> Libraries</a></li>
<li class="chapter" data-level="8.2" data-path="data-wrangling.html"><a href="data-wrangling.html#data"><i class="fa fa-check"></i><b>8.2</b> Data</a></li>
<li class="chapter" data-level="8.3" data-path="data-wrangling.html"><a href="data-wrangling.html#loading-the-data"><i class="fa fa-check"></i><b>8.3</b> Loading the data</a></li>
<li class="chapter" data-level="8.4" data-path="data-wrangling.html"><a href="data-wrangling.html#check-your-data"><i class="fa fa-check"></i><b>8.4</b> Check your data!</a></li>
<li class="chapter" data-level="8.5" data-path="data-wrangling.html"><a href="data-wrangling.html#renaming-data-frame-columms"><i class="fa fa-check"></i><b>8.5</b> Renaming data frame columms</a></li>
<li class="chapter" data-level="8.6" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-unneeded-columns"><i class="fa fa-check"></i><b>8.6</b> Dropping unneeded columns</a>
<ul>
<li class="chapter" data-level="8.6.1" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-all-empty-columns"><i class="fa fa-check"></i><b>8.6.1</b> Finding all empty columns</a></li>
<li class="chapter" data-level="8.6.2" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-columns-by-matching-names"><i class="fa fa-check"></i><b>8.6.2</b> Dropping columns by matching names</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="data-wrangling.html"><a href="data-wrangling.html#merging-data-frames"><i class="fa fa-check"></i><b>8.7</b> Merging data frames</a></li>
<li class="chapter" data-level="8.8" data-path="data-wrangling.html"><a href="data-wrangling.html#reshaping-data-with-tidyr"><i class="fa fa-check"></i><b>8.8</b> Reshaping data with tidyr</a>
<ul>
<li class="chapter" data-level="8.8.1" data-path="data-wrangling.html"><a href="data-wrangling.html#wide-to-long-conversions-using-pivot_longer"><i class="fa fa-check"></i><b>8.8.1</b> Wide to long conversions using <code>pivot_longer()</code></a></li>
<li class="chapter" data-level="8.8.2" data-path="data-wrangling.html"><a href="data-wrangling.html#extracting-information-from-combined-variables-using-extract"><i class="fa fa-check"></i><b>8.8.2</b> Extracting information from combined variables using <code>extract()</code></a></li>
<li class="chapter" data-level="8.8.3" data-path="data-wrangling.html"><a href="data-wrangling.html#subsetting-rows"><i class="fa fa-check"></i><b>8.8.3</b> Subsetting rows</a></li>
<li class="chapter" data-level="8.8.4" data-path="data-wrangling.html"><a href="data-wrangling.html#splitting-columns"><i class="fa fa-check"></i><b>8.8.4</b> Splitting columns</a></li>
<li class="chapter" data-level="8.8.5" data-path="data-wrangling.html"><a href="data-wrangling.html#combining-data-frames"><i class="fa fa-check"></i><b>8.8.5</b> Combining data frames</a></li>
<li class="chapter" data-level="8.8.6" data-path="data-wrangling.html"><a href="data-wrangling.html#sorting-data-frame-rows"><i class="fa fa-check"></i><b>8.8.6</b> Sorting data frame rows</a></li>
</ul></li>
<li class="chapter" data-level="8.9" data-path="data-wrangling.html"><a href="data-wrangling.html#using-your-tidy-data"><i class="fa fa-check"></i><b>8.9</b> Using your tidy data</a>
<ul>
<li class="chapter" data-level="8.9.1" data-path="data-wrangling.html"><a href="data-wrangling.html#visualizing-gene-expression-time-series"><i class="fa fa-check"></i><b>8.9.1</b> Visualizing gene expression time series</a></li>
<li class="chapter" data-level="8.9.2" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-the-most-variable-genes"><i class="fa fa-check"></i><b>8.9.2</b> Finding the most variable genes</a></li>
<li class="chapter" data-level="8.9.3" data-path="data-wrangling.html"><a href="data-wrangling.html#heat-maps"><i class="fa fa-check"></i><b>8.9.3</b> Heat maps</a></li>
</ul></li>
<li class="chapter" data-level="8.10" data-path="data-wrangling.html"><a href="data-wrangling.html#long-to-wide-conversion-using-pivot_wider"><i class="fa fa-check"></i><b>8.10</b> Long-to-wide conversion using <code>pivot_wider()</code></a></li>
<li class="chapter" data-level="8.11" data-path="data-wrangling.html"><a href="data-wrangling.html#exploring-bivariate-relationships-using-wide-data"><i class="fa fa-check"></i><b>8.11</b> Exploring bivariate relationships using “wide” data</a>
<ul>
<li class="chapter" data-level="8.11.1" data-path="data-wrangling.html"><a href="data-wrangling.html#large-scale-patterns-of-correlations"><i class="fa fa-check"></i><b>8.11.1</b> Large scale patterns of correlations</a></li>
<li class="chapter" data-level="8.11.2" data-path="data-wrangling.html"><a href="data-wrangling.html#adding-new-columns-and-combining-filtered-data-frames"><i class="fa fa-check"></i><b>8.11.2</b> Adding new columns and combining filtered data frames</a></li>
<li class="chapter" data-level="8.11.3" data-path="data-wrangling.html"><a href="data-wrangling.html#a-heat-mapped-sorted-by-correlations"><i class="fa fa-check"></i><b>8.11.3</b> A heat mapped sorted by correlations</a></li>
<li class="chapter" data-level="8.11.4" data-path="data-wrangling.html"><a href="data-wrangling.html#a-fancy-figure"><i class="fa fa-check"></i><b>8.11.4</b> A “fancy” figure</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="vector-algebra.html"><a href="vector-algebra.html"><i class="fa fa-check"></i><b>9</b> Vector algebra</a>
<ul>
<li class="chapter" data-level="9.1" data-path="vector-algebra.html"><a href="vector-algebra.html#libraries-2"><i class="fa fa-check"></i><b>9.1</b> Libraries</a></li>
<li class="chapter" data-level="9.2" data-path="vector-algebra.html"><a href="vector-algebra.html#vector-mathematics-in-r"><i class="fa fa-check"></i><b>9.2</b> Vector Mathematics in R</a></li>
<li class="chapter" data-level="9.3" data-path="vector-algebra.html"><a href="vector-algebra.html#simple-statistics-in-vector-form"><i class="fa fa-check"></i><b>9.3</b> Simple statistics in vector form</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="vector-algebra.html"><a href="vector-algebra.html#mean"><i class="fa fa-check"></i><b>9.3.1</b> Mean</a></li>
<li class="chapter" data-level="9.3.2" data-path="vector-algebra.html"><a href="vector-algebra.html#mean-centering"><i class="fa fa-check"></i><b>9.3.2</b> Mean centering</a></li>
<li class="chapter" data-level="9.3.3" data-path="vector-algebra.html"><a href="vector-algebra.html#variance-and-standard-deviation"><i class="fa fa-check"></i><b>9.3.3</b> Variance and standard deviation</a></li>
<li class="chapter" data-level="9.3.4" data-path="vector-algebra.html"><a href="vector-algebra.html#covariance-and-correlation"><i class="fa fa-check"></i><b>9.3.4</b> Covariance and correlation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="matrices-in-r.html"><a href="matrices-in-r.html"><i class="fa fa-check"></i><b>10</b> Matrices in R</a>
<ul>
<li class="chapter" data-level="10.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#creating-matrices-in-r"><i class="fa fa-check"></i><b>10.1</b> Creating matrices in R</a></li>
<li class="chapter" data-level="10.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-arithmetic-operations-in-r"><i class="fa fa-check"></i><b>10.2</b> Matrix arithmetic operations in R</a></li>
<li class="chapter" data-level="10.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#descriptive-statistics-as-matrix-functions"><i class="fa fa-check"></i><b>10.3</b> Descriptive statistics as matrix functions</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#mean-vector-and-matrix"><i class="fa fa-check"></i><b>10.3.1</b> Mean vector and matrix</a></li>
<li class="chapter" data-level="10.3.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#deviation-matrix"><i class="fa fa-check"></i><b>10.3.2</b> Deviation matrix</a></li>
<li class="chapter" data-level="10.3.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#covariance-matrix"><i class="fa fa-check"></i><b>10.3.3</b> Covariance matrix</a></li>
<li class="chapter" data-level="10.3.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#correlation-matrix"><i class="fa fa-check"></i><b>10.3.4</b> Correlation matrix</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-inverse"><i class="fa fa-check"></i><b>10.4</b> Matrix Inverse</a></li>
<li class="chapter" data-level="10.5" data-path="matrices-in-r.html"><a href="matrices-in-r.html#solving-sets-of-simultaneous-equations"><i class="fa fa-check"></i><b>10.5</b> Solving sets of simultaneous equations</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html"><i class="fa fa-check"></i><b>11</b> Linear Regression Models I</a>
<ul>
<li class="chapter" data-level="11.1" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#linear-functions"><i class="fa fa-check"></i><b>11.1</b> Linear functions</a></li>
<li class="chapter" data-level="11.2" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#linear-regression"><i class="fa fa-check"></i><b>11.2</b> Linear regression</a></li>
<li class="chapter" data-level="11.3" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#the-optimality-criterion-for-least-squares-regression"><i class="fa fa-check"></i><b>11.3</b> The optimality criterion for least-squares regression</a></li>
<li class="chapter" data-level="11.4" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#geometry-of-linear-regression"><i class="fa fa-check"></i><b>11.4</b> Geometry of linear regression</a></li>
<li class="chapter" data-level="11.5" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#solution-for-the-least-squares-criterion"><i class="fa fa-check"></i><b>11.5</b> Solution for the least-squares criterion</a>
<ul>
<li class="chapter" data-level="11.5.1" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#bivariate-regression-estimating-coefficients"><i class="fa fa-check"></i><b>11.5.1</b> Bivariate regression, estimating coefficients</a></li>
<li class="chapter" data-level="11.5.2" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#multiple-regression-estimating-coefficients"><i class="fa fa-check"></i><b>11.5.2</b> Multiple regression, estimating coefficients</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#residuals"><i class="fa fa-check"></i><b>11.6</b> Residuals</a></li>
<li class="chapter" data-level="11.7" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#regression-as-sum-of-squares-decomposition"><i class="fa fa-check"></i><b>11.7</b> Regression as sum-of-squares decomposition</a></li>
<li class="chapter" data-level="11.8" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#variance-explained-by-a-regression-model"><i class="fa fa-check"></i><b>11.8</b> Variance “explained” by a regression model</a></li>
<li class="chapter" data-level="11.9" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#goodness-of-fit"><i class="fa fa-check"></i><b>11.9</b> Goodness of fit</a></li>
<li class="chapter" data-level="11.10" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#interpretting-regression"><i class="fa fa-check"></i><b>11.10</b> Interpretting Regression</a></li>
<li class="chapter" data-level="11.11" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#illustrating-linear-regression-with-simulated-data"><i class="fa fa-check"></i><b>11.11</b> Illustrating linear regression with simulated data</a></li>
<li class="chapter" data-level="11.12" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#calculating-the-regression-model-using-vector-arithmetic"><i class="fa fa-check"></i><b>11.12</b> Calculating the regression model using vector arithmetic</a>
<ul>
<li class="chapter" data-level="11.12.1" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#calculating-the-predicted-values-and-residuals"><i class="fa fa-check"></i><b>11.12.1</b> Calculating the predicted values and residuals</a></li>
</ul></li>
<li class="chapter" data-level="11.13" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#plotting-various-aspects-of-a-regression"><i class="fa fa-check"></i><b>11.13</b> Plotting various aspects of a regression</a>
<ul>
<li class="chapter" data-level="11.13.1" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#residual-plots"><i class="fa fa-check"></i><b>11.13.1</b> Residual plots</a></li>
</ul></li>
<li class="chapter" data-level="11.14" data-path="linear-regression-models-i.html"><a href="linear-regression-models-i.html#calculating-the-coefficient-of-determination"><i class="fa fa-check"></i><b>11.14</b> Calculating the coefficient of determination</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html"><i class="fa fa-check"></i><b>12</b> Linear Regression Models II</a>
<ul>
<li class="chapter" data-level="12.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#new-libraries-to-install"><i class="fa fa-check"></i><b>12.1</b> New Libraries to install</a></li>
<li class="chapter" data-level="12.2" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#standard-libraries"><i class="fa fa-check"></i><b>12.2</b> Standard libraries</a></li>
<li class="chapter" data-level="12.3" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#specifying-regression-models-in-r"><i class="fa fa-check"></i><b>12.3</b> Specifying Regression Models in R</a>
<ul>
<li class="chapter" data-level="12.3.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#example-data"><i class="fa fa-check"></i><b>12.3.1</b> Example data</a></li>
<li class="chapter" data-level="12.3.2" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#fitting-the-model-using-lm"><i class="fa fa-check"></i><b>12.3.2</b> Fitting the model using <code>lm()</code></a></li>
<li class="chapter" data-level="12.3.3" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#interpretting-summary-output-from-lm"><i class="fa fa-check"></i><b>12.3.3</b> Interpretting summary output from lm()</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#quick-bivariate-regression-plots-in-ggplot"><i class="fa fa-check"></i><b>12.4</b> Quick bivariate regression plots in ggplot</a></li>
<li class="chapter" data-level="12.5" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#more-about-the-data-structure-returned-by-lm"><i class="fa fa-check"></i><b>12.5</b> More about the data structure returned by <code>lm()</code></a>
<ul>
<li class="chapter" data-level="12.5.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#fitted-values"><i class="fa fa-check"></i><b>12.5.1</b> Fitted values</a></li>
<li class="chapter" data-level="12.5.2" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#getting-the-model-coefficients"><i class="fa fa-check"></i><b>12.5.2</b> Getting the model coefficients</a></li>
</ul></li>
<li class="chapter" data-level="12.6" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#broom-a-library-for-converting-model-results-into-data-frames"><i class="fa fa-check"></i><b>12.6</b> Broom: a library for converting model results into data frames</a>
<ul>
<li class="chapter" data-level="12.6.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#broomtidy"><i class="fa fa-check"></i><b>12.6.1</b> <code>broom::tidy</code></a></li>
<li class="chapter" data-level="12.6.2" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#broomaugment"><i class="fa fa-check"></i><b>12.6.2</b> <code>broom::augment</code></a></li>
<li class="chapter" data-level="12.6.3" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#broomglance"><i class="fa fa-check"></i><b>12.6.3</b> <code>broom::glance</code></a></li>
</ul></li>
<li class="chapter" data-level="12.7" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#qq-plots"><i class="fa fa-check"></i><b>12.7</b> qq-plots</a></li>
<li class="chapter" data-level="12.8" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#multiple-regression"><i class="fa fa-check"></i><b>12.8</b> Multiple regression</a>
<ul>
<li class="chapter" data-level="12.8.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#exploration-of-the-trees-data-set"><i class="fa fa-check"></i><b>12.8.1</b> Exploration of the trees data set</a></li>
</ul></li>
<li class="chapter" data-level="12.9" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#d-plots"><i class="fa fa-check"></i><b>12.9</b> 3D Plots</a>
<ul>
<li class="chapter" data-level="12.9.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#scatterplot3d"><i class="fa fa-check"></i><b>12.9.1</b> scatterplot3d</a></li>
</ul></li>
<li class="chapter" data-level="12.10" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#fitting-the-regression-model"><i class="fa fa-check"></i><b>12.10</b> Fitting the regression model</a>
<ul>
<li class="chapter" data-level="12.10.1" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#visualizing-the-regression-model-in-scatterplot3d"><i class="fa fa-check"></i><b>12.10.1</b> Visualizing the regression model in scatterplot3d</a></li>
</ul></li>
<li class="chapter" data-level="12.11" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#interpretting-the-regression-model"><i class="fa fa-check"></i><b>12.11</b> Interpretting the regression model</a></li>
<li class="chapter" data-level="12.12" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#exploring-the-vector-geometry-of-the-regression-model"><i class="fa fa-check"></i><b>12.12</b> Exploring the Vector Geometry of the Regression Model</a></li>
<li class="chapter" data-level="12.13" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#exploring-the-residuals-from-the-model-fit"><i class="fa fa-check"></i><b>12.13</b> Exploring the Residuals from the Model Fit</a></li>
<li class="chapter" data-level="12.14" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#an-alternate-model"><i class="fa fa-check"></i><b>12.14</b> An alternate model</a></li>
<li class="chapter" data-level="12.15" data-path="linear-regression-models-ii.html"><a href="linear-regression-models-ii.html#exploring-the-impact-of-nearly-collinear-predictors-on-regression"><i class="fa fa-check"></i><b>12.15</b> Exploring the impact of nearly collinear predictors on regression</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html"><i class="fa fa-check"></i><b>13</b> Principal Components Analysis</a>
<ul>
<li class="chapter" data-level="13.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#libraries-3"><i class="fa fa-check"></i><b>13.1</b> Libraries</a></li>
<li class="chapter" data-level="13.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#matrices-as-linear-transformations"><i class="fa fa-check"></i><b>13.2</b> Matrices as linear transformations</a></li>
<li class="chapter" data-level="13.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#eigenanalysis-in-r"><i class="fa fa-check"></i><b>13.3</b> Eigenanalysis in R</a></li>
<li class="chapter" data-level="13.4" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#principal-components-analysis-in-r"><i class="fa fa-check"></i><b>13.4</b> Principal Components Analysis in R</a>
<ul>
<li class="chapter" data-level="13.4.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#bioenv-dataset"><i class="fa fa-check"></i><b>13.4.1</b> Bioenv dataset</a></li>
<li class="chapter" data-level="13.4.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pca-of-the-bioenv-dataset"><i class="fa fa-check"></i><b>13.4.2</b> PCA of the Bioenv dataset</a></li>
<li class="chapter" data-level="13.4.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#calculating-factor-loadings"><i class="fa fa-check"></i><b>13.4.3</b> Calculating Factor Loadings</a></li>
</ul></li>
<li class="chapter" data-level="13.5" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#drawing-figures-to-represent-pca"><i class="fa fa-check"></i><b>13.5</b> Drawing Figures to Represent PCA</a>
<ul>
<li class="chapter" data-level="13.5.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pc-score-plots"><i class="fa fa-check"></i><b>13.5.1</b> PC Score Plots</a></li>
<li class="chapter" data-level="13.5.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#simultaneous-depiction-of-observations-and-variables-in-the-pc-space"><i class="fa fa-check"></i><b>13.5.2</b> Simultaneous Depiction of Observations and Variables in the PC Space</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="14" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html"><i class="fa fa-check"></i><b>14</b> Canonical Variates Analysis</a>
<ul>
<li class="chapter" data-level="14.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#libraries-4"><i class="fa fa-check"></i><b>14.1</b> Libraries</a></li>
<li class="chapter" data-level="14.2" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#discriminant-analysis-in-r"><i class="fa fa-check"></i><b>14.2</b> Discriminant Analysis in R</a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#shorthand-formulae-in-r"><i class="fa fa-check"></i><b>14.2.1</b> Shorthand Formulae in R</a></li>
<li class="chapter" data-level="14.2.2" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#working-with-the-output-of-lda"><i class="fa fa-check"></i><b>14.2.2</b> Working with the output of <code>lda()</code></a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#estimating-confidence-regions-for-group-means-in-cva"><i class="fa fa-check"></i><b>14.3</b> Estimating confidence regions for group means in CVA</a>
<ul>
<li class="chapter" data-level="14.3.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#drawing-the-cva-confidence-regions"><i class="fa fa-check"></i><b>14.3.1</b> Drawing the CVA confidence regions</a></li>
</ul></li>
<li class="chapter" data-level="14.4" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#calculating-the-within-and-between-group-covariance-matrices"><i class="fa fa-check"></i><b>14.4</b> Calculating the Within and Between Group Covariance Matrices</a>
<ul>
<li class="chapter" data-level="14.4.1" data-path="canonical-variates-analysis.html"><a href="canonical-variates-analysis.html#recapitulating-the-cva-analysis-of-lda"><i class="fa fa-check"></i><b>14.4.1</b> Recapitulating the CVA analysis of <code>lda()</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="15" data-path="clustering-in-r.html"><a href="clustering-in-r.html"><i class="fa fa-check"></i><b>15</b> Clustering in R</a>
<ul>
<li class="chapter" data-level="15.1" data-path="clustering-in-r.html"><a href="clustering-in-r.html#libraries-5"><i class="fa fa-check"></i><b>15.1</b> Libraries</a></li>
<li class="chapter" data-level="15.2" data-path="clustering-in-r.html"><a href="clustering-in-r.html#data-set"><i class="fa fa-check"></i><b>15.2</b> Data set</a></li>
<li class="chapter" data-level="15.3" data-path="clustering-in-r.html"><a href="clustering-in-r.html#hierarchical-clustering-in-r"><i class="fa fa-check"></i><b>15.3</b> Hierarchical Clustering in R</a></li>
<li class="chapter" data-level="15.4" data-path="clustering-in-r.html"><a href="clustering-in-r.html#manipulating-hierarchical-clusterings-with-dendextend"><i class="fa fa-check"></i><b>15.4</b> Manipulating hierarchical clusterings with dendextend</a></li>
<li class="chapter" data-level="15.5" data-path="clustering-in-r.html"><a href="clustering-in-r.html#plotting-dendrograms-in-dendextend"><i class="fa fa-check"></i><b>15.5</b> Plotting dendrograms in dendextend</a></li>
<li class="chapter" data-level="15.6" data-path="clustering-in-r.html"><a href="clustering-in-r.html#cutting-dendrograms"><i class="fa fa-check"></i><b>15.6</b> Cutting dendrograms</a></li>
<li class="chapter" data-level="15.7" data-path="clustering-in-r.html"><a href="clustering-in-r.html#looking-at-clusters"><i class="fa fa-check"></i><b>15.7</b> Looking at clusters</a></li>
<li class="chapter" data-level="15.8" data-path="clustering-in-r.html"><a href="clustering-in-r.html#generating-a-heat-map-from-a-cluster"><i class="fa fa-check"></i><b>15.8</b> Generating a heat map from a cluster</a></li>
<li class="chapter" data-level="15.9" data-path="clustering-in-r.html"><a href="clustering-in-r.html#working-with-sub-trees"><i class="fa fa-check"></i><b>15.9</b> Working with sub-trees</a></li>
<li class="chapter" data-level="15.10" data-path="clustering-in-r.html"><a href="clustering-in-r.html#setting-dendrogram-parameters-in-dendextend"><i class="fa fa-check"></i><b>15.10</b> Setting dendrogram parameters in dendextend</a></li>
<li class="chapter" data-level="15.11" data-path="clustering-in-r.html"><a href="clustering-in-r.html#combining-heatmaps-and-dendrograms"><i class="fa fa-check"></i><b>15.11</b> Combining heatmaps and dendrograms</a></li>
<li class="chapter" data-level="15.12" data-path="clustering-in-r.html"><a href="clustering-in-r.html#k-meansk-medoids-clustering-in-r"><i class="fa fa-check"></i><b>15.12</b> K-means/K-medoids Clustering in R</a>
<ul>
<li class="chapter" data-level="15.12.1" data-path="clustering-in-r.html"><a href="clustering-in-r.html#heat-map-from-k-medoids-cluster"><i class="fa fa-check"></i><b>15.12.1</b> Heat map from k-medoids cluster</a></li>
</ul></li>
<li class="chapter" data-level="15.13" data-path="clustering-in-r.html"><a href="clustering-in-r.html#combining-clusters-and-correlation-matrix-heatmaps"><i class="fa fa-check"></i><b>15.13</b> Combining clusters and correlation matrix heatmaps</a>
<ul>
<li class="chapter" data-level="15.13.1" data-path="clustering-in-r.html"><a href="clustering-in-r.html#hierarchical-clustering-visualized-on-correlation-matrix"><i class="fa fa-check"></i><b>15.13.1</b> Hierarchical clustering, visualized on correlation matrix</a></li>
<li class="chapter" data-level="15.13.2" data-path="clustering-in-r.html"><a href="clustering-in-r.html#k-medoids-clustering-visualized-on-the-correlation-matrix"><i class="fa fa-check"></i><b>15.13.2</b> K-medoids clustering visualized on the correlation matrix</a></li>
</ul></li>
<li class="chapter" data-level="15.14" data-path="clustering-in-r.html"><a href="clustering-in-r.html#depicting-the-data-within-clusters"><i class="fa fa-check"></i><b>15.14</b> Depicting the data within clusters</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="non-linear-regression-models.html"><a href="non-linear-regression-models.html"><i class="fa fa-check"></i><b>16</b> Non-linear regression models</a>
<ul>
<li class="chapter" data-level="16.1" data-path="non-linear-regression-models.html"><a href="non-linear-regression-models.html#loess-regression"><i class="fa fa-check"></i><b>16.1</b> LOESS regression</a></li>
<li class="chapter" data-level="16.2" data-path="non-linear-regression-models.html"><a href="non-linear-regression-models.html#logistic-regression"><i class="fa fa-check"></i><b>16.2</b> Logistic regression</a>
<ul>
<li class="chapter" data-level="16.2.1" data-path="non-linear-regression-models.html"><a href="non-linear-regression-models.html#a-web-app-to-explore-the-logistic-regression-equation"><i class="fa fa-check"></i><b>16.2.1</b> A web app to explore the logistic regression equation</a></li>
<li class="chapter" data-level="16.2.2" data-path="non-linear-regression-models.html"><a href="non-linear-regression-models.html#titanic-data-set"><i class="fa fa-check"></i><b>16.2.2</b> Titanic data set</a></li>
<li class="chapter" data-level="16.2.3" data-path="non-linear-regression-models.html"><a href="non-linear-regression-models.html#subsetting-the-data"><i class="fa fa-check"></i><b>16.2.3</b> Subsetting the data</a></li>
<li class="chapter" data-level="16.2.4" data-path="non-linear-regression-models.html"><a href="non-linear-regression-models.html#visualizing-survival-as-a-function-of-age"><i class="fa fa-check"></i><b>16.2.4</b> Visualizing survival as a function of age</a></li>
<li class="chapter" data-level="16.2.5" data-path="non-linear-regression-models.html"><a href="non-linear-regression-models.html#fitting-the-logistic-regression-model"><i class="fa fa-check"></i><b>16.2.5</b> Fitting the logistic regression model</a></li>
<li class="chapter" data-level="16.2.6" data-path="non-linear-regression-models.html"><a href="non-linear-regression-models.html#visualizing-the-logistic-regression"><i class="fa fa-check"></i><b>16.2.6</b> Visualizing the logistic regression</a></li>
<li class="chapter" data-level="16.2.7" data-path="non-linear-regression-models.html"><a href="non-linear-regression-models.html#quick-and-easy-visualization"><i class="fa fa-check"></i><b>16.2.7</b> Quick and easy visualization</a></li>
<li class="chapter" data-level="16.2.8" data-path="non-linear-regression-models.html"><a href="non-linear-regression-models.html#impact-of-sex-and-passenger-class-on-the-models"><i class="fa fa-check"></i><b>16.2.8</b> Impact of sex and passenger class on the models</a></li>
<li class="chapter" data-level="16.2.9" data-path="non-linear-regression-models.html"><a href="non-linear-regression-models.html#fitting-multiple-models-based-on-groupings-use-dplyrdo"><i class="fa fa-check"></i><b>16.2.9</b> Fitting multiple models based on groupings use <code>dplyr::do</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="17" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html"><i class="fa fa-check"></i><b>17</b> Randomization, Jackknife, and Bootstrap</a>
<ul>
<li class="chapter" data-level="17.1" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#libraries-6"><i class="fa fa-check"></i><b>17.1</b> Libraries</a></li>
<li class="chapter" data-level="17.2" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#randomization"><i class="fa fa-check"></i><b>17.2</b> Randomization</a></li>
<li class="chapter" data-level="17.3" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#using-randomization-to-test-for-a-difference-in-means"><i class="fa fa-check"></i><b>17.3</b> Using randomization to test for a difference in means</a></li>
<li class="chapter" data-level="17.4" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#using-randomization-to-test-for-equality-of-variances"><i class="fa fa-check"></i><b>17.4</b> Using randomization to test for equality of variances</a></li>
<li class="chapter" data-level="17.5" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#jackknifing-in-r"><i class="fa fa-check"></i><b>17.5</b> Jackknifing in R</a></li>
<li class="chapter" data-level="17.6" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#bootstrapping-in-r"><i class="fa fa-check"></i><b>17.6</b> Bootstrapping in R</a>
<ul>
<li class="chapter" data-level="17.6.1" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#a-more-sophisticated-application-of-the-bootstrap"><i class="fa fa-check"></i><b>17.6.1</b> A more sophisticated application of the bootstrap</a></li>
<li class="chapter" data-level="17.6.2" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#about-logistic-regression"><i class="fa fa-check"></i><b>17.6.2</b> About logistic regression</a></li>
<li class="chapter" data-level="17.6.3" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#bumpus-house-sparrow-data-set"><i class="fa fa-check"></i><b>17.6.3</b> Bumpus house sparrow data set</a></li>
<li class="chapter" data-level="17.6.4" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#predicting-survival-as-a-function-of-body-weight"><i class="fa fa-check"></i><b>17.6.4</b> Predicting survival as a function of body weight</a></li>
<li class="chapter" data-level="17.6.5" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#fitting-the-logistic-regression"><i class="fa fa-check"></i><b>17.6.5</b> Fitting the logistic regression</a></li>
<li class="chapter" data-level="17.6.6" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#defining-an-appropriate-function-for-boot"><i class="fa fa-check"></i><b>17.6.6</b> Defining an appropriate function for <code>boot</code></a></li>
<li class="chapter" data-level="17.6.7" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#calculating-bootstrap-confidence-intervals"><i class="fa fa-check"></i><b>17.6.7</b> Calculating bootstrap confidence intervals</a></li>
<li class="chapter" data-level="17.6.8" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#visualizing-bootstrap-confidence-intervals"><i class="fa fa-check"></i><b>17.6.8</b> Visualizing bootstrap confidence intervals</a></li>
</ul></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html"><i class="fa fa-check"></i><b>A</b> Simulating sampling distributions</a>
<ul>
<li class="chapter" data-level="A.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#sampling-distributions"><i class="fa fa-check"></i><b>A.1</b> Sampling distributions</a></li>
<li class="chapter" data-level="A.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#simulating-sampling-from-an-underlying-population"><i class="fa fa-check"></i><b>A.2</b> Simulating sampling from an underlying population</a></li>
<li class="chapter" data-level="A.3" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#seeding-the-pseudo-random-number-generator"><i class="fa fa-check"></i><b>A.3</b> Seeding the pseudo-random number generator</a></li>
<li class="chapter" data-level="A.4" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#libraries-7"><i class="fa fa-check"></i><b>A.4</b> Libraries</a>
<ul>
<li class="chapter" data-level="A.4.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#properties-of-the-underlying-population"><i class="fa fa-check"></i><b>A.4.1</b> Properties of the underlying population</a></li>
<li class="chapter" data-level="A.4.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#other-r-functions-related-to-the-normal-distribution"><i class="fa fa-check"></i><b>A.4.2</b> Other R functions related to the normal distribution</a></li>
</ul></li>
<li class="chapter" data-level="A.5" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#random-sampling-from-the-simulated-population"><i class="fa fa-check"></i><b>A.5</b> Random sampling from the simulated population</a>
<ul>
<li class="chapter" data-level="A.5.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#another-random-sample"><i class="fa fa-check"></i><b>A.5.1</b> Another random sample</a></li>
<li class="chapter" data-level="A.5.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#simulating-the-generation-of-many-random-samples"><i class="fa fa-check"></i><b>A.5.2</b> Simulating the generation of many random samples</a></li>
<li class="chapter" data-level="A.5.3" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#a-function-to-estimate-statistics-of-interest-in-a-random-sample"><i class="fa fa-check"></i><b>A.5.3</b> A function to estimate statistics of interest in a random sample</a></li>
<li class="chapter" data-level="A.5.4" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#generating-statistics-for-many-random-samples"><i class="fa fa-check"></i><b>A.5.4</b> Generating statistics for many random samples</a></li>
</ul></li>
<li class="chapter" data-level="A.6" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#simulated-sampling-distribution-of-the-mean"><i class="fa fa-check"></i><b>A.6</b> Simulated sampling distribution of the mean</a>
<ul>
<li class="chapter" data-level="A.6.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#differences-between-sampling-distributions-and-samplepopulation-distributions"><i class="fa fa-check"></i><b>A.6.1</b> Differences between sampling distributions and sample/population distributions</a></li>
<li class="chapter" data-level="A.6.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#using-sampling-distributions-to-understand-the-behavior-of-statistics-of-interest"><i class="fa fa-check"></i><b>A.6.2</b> Using sampling distributions to understand the behavior of statistics of interest</a></li>
<li class="chapter" data-level="A.6.3" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#sampling-distributions-for-different-sample-sizes"><i class="fa fa-check"></i><b>A.6.3</b> Sampling distributions for different sample sizes</a></li>
<li class="chapter" data-level="A.6.4" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#discussion-of-trends-for-sampling-distributions-of-different-sample-sizes"><i class="fa fa-check"></i><b>A.6.4</b> Discussion of trends for sampling distributions of different sample sizes</a></li>
</ul></li>
<li class="chapter" data-level="A.7" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#standard-error-of-the-mean"><i class="fa fa-check"></i><b>A.7</b> Standard Error of the Mean</a></li>
<li class="chapter" data-level="A.8" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#sampling-distribution-of-the-standard-deviation"><i class="fa fa-check"></i><b>A.8</b> Sampling Distribution of the Standard Deviation</a>
<ul>
<li class="chapter" data-level="A.8.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#standard-error-of-standard-deviations"><i class="fa fa-check"></i><b>A.8.1</b> Standard error of standard deviations</a></li>
</ul></li>
<li class="chapter" data-level="A.9" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#what-happens-to-the-sampling-distribution-of-the-mean-and-standard-deviation-when-our-sample-size-is-small"><i class="fa fa-check"></i><b>A.9</b> What happens to the sampling distribution of the mean and standard deviation when our sample size is small?</a>
<ul>
<li class="chapter" data-level="A.9.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#for-small-samples-sample-standard-deviations-systematically-underestimate-the-population-standard-deviation"><i class="fa fa-check"></i><b>A.9.1</b> For small samples, sample standard deviations systematically underestimate the population standard deviation</a></li>
<li class="chapter" data-level="A.9.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#underestimates-of-the-standard-deviation-for-small-n-lead-to-understimates-of-the-se-of-the-mean"><i class="fa fa-check"></i><b>A.9.2</b> Underestimates of the standard deviation for small <span class="math inline">\(n\)</span> lead to understimates of the SE of the mean</a></li>
<li class="chapter" data-level="A.9.3" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#the-t-distribution-is-the-appropriate-distribution-for-describing-the-sampling-distribution-of-the-mean-when-n-is-small"><i class="fa fa-check"></i><b>A.9.3</b> The t-distribution is the appropriate distribution for describing the sampling distribution of the mean when <span class="math inline">\(n\)</span> is small</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="B" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html"><i class="fa fa-check"></i><b>B</b> Simulating confidence intervals</a>
<ul>
<li class="chapter" data-level="B.1" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#confidence-intervals"><i class="fa fa-check"></i><b>B.1</b> Confidence intervals</a></li>
<li class="chapter" data-level="B.2" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#generic-formulation-for-confidence-intervals"><i class="fa fa-check"></i><b>B.2</b> Generic formulation for confidence intervals</a></li>
<li class="chapter" data-level="B.3" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#example-confidence-intervals-for-the-mean"><i class="fa fa-check"></i><b>B.3</b> Example: Confidence intervals for the mean</a>
<ul>
<li class="chapter" data-level="B.3.1" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#simulation-of-means"><i class="fa fa-check"></i><b>B.3.1</b> Simulation of means</a></li>
<li class="chapter" data-level="B.3.2" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#distance-between-sample-means-and-true-means"><i class="fa fa-check"></i><b>B.3.2</b> Distance between sample means and true means</a></li>
<li class="chapter" data-level="B.3.3" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#calculating-a-ci"><i class="fa fa-check"></i><b>B.3.3</b> Calculating a CI</a></li>
</ul></li>
<li class="chapter" data-level="B.4" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#a-problem-arises"><i class="fa fa-check"></i><b>B.4</b> A problem arises!</a></li>
<li class="chapter" data-level="B.5" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#confidence-intervals-under-sample-estimates-of-standard-errors"><i class="fa fa-check"></i><b>B.5</b> Confidence intervals under sample estimates of standard errors</a>
<ul>
<li class="chapter" data-level="B.5.1" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#the-spread-of-sample-estimates-of-the-mean-in-units-of-true-and-estimates-ses"><i class="fa fa-check"></i><b>B.5.1</b> The spread of sample estimates of the mean in units of true and estimates SEs</a></li>
<li class="chapter" data-level="B.5.2" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#quantifying-deviations-of-sample-estimates-of-the-mean"><i class="fa fa-check"></i><b>B.5.2</b> Quantifying deviations of sample estimates of the mean</a></li>
</ul></li>
<li class="chapter" data-level="B.6" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#t-distribution-for-confidence-intervals-of-the-mean"><i class="fa fa-check"></i><b>B.6</b> t-distribution for confidence intervals of the mean</a>
<ul>
<li class="chapter" data-level="B.6.1" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#t-distribution-vs-standard-normal-distribution"><i class="fa fa-check"></i><b>B.6.1</b> t-distribution vs standard normal distribution</a></li>
<li class="chapter" data-level="B.6.2" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#formula-for-confidence-intervals-of-the-mean-based-on-sample-estimates-of-the-se-and-the-t-distribution"><i class="fa fa-check"></i><b>B.6.2</b> Formula for confidence intervals of the mean based on sample estimates of the SE and the t-distribution</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Biology 723: Statistical Computing for Biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="principal-components-analysis" class="section level1 hasAnchor" number="13">
<h1><span class="header-section-number">Chapter 13</span> Principal Components Analysis<a href="principal-components-analysis.html#principal-components-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="libraries-3" class="section level2 hasAnchor" number="13.1">
<h2><span class="header-section-number">13.1</span> Libraries<a href="principal-components-analysis.html#libraries-3" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb408"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb408-1"><a href="principal-components-analysis.html#cb408-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb408-2"><a href="principal-components-analysis.html#cb408-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(broom)</span>
<span id="cb408-3"><a href="principal-components-analysis.html#cb408-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GGally)</span>
<span id="cb408-4"><a href="principal-components-analysis.html#cb408-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span></code></pre></div>
</div>
<div id="matrices-as-linear-transformations" class="section level2 hasAnchor" number="13.2">
<h2><span class="header-section-number">13.2</span> Matrices as linear transformations<a href="principal-components-analysis.html#matrices-as-linear-transformations" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In lecture we introduced the notion that pre-multiplying a vector, <span class="math inline">\(x\)</span>, by a matrix <span class="math inline">\(\mathbf{A}\)</span> represents a linear transformation of <span class="math inline">\(x\)</span>. Let’s explore that visually for 2D vectors.</p>
<p>For illustration let’s generate a 2D vector representing the coordinates of points on the sine function.</p>
<div class="sourceCode" id="cb409"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb409-1"><a href="principal-components-analysis.html#cb409-1" aria-hidden="true" tabindex="-1"></a>sin.xy <span class="ot">&lt;-</span> <span class="fu">data_frame</span>(<span class="at">x =</span> <span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">2</span><span class="sc">*</span>pi, <span class="at">length.out =</span> <span class="dv">50</span>) <span class="sc">-</span> pi,</span>
<span id="cb409-2"><a href="principal-components-analysis.html#cb409-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">y =</span> <span class="fu">sin</span>(x))</span>
<span id="cb409-3"><a href="principal-components-analysis.html#cb409-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb409-4"><a href="principal-components-analysis.html#cb409-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(sin.xy, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y=</span> y)) <span class="sc">+</span> <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-406-1.png" width="576" /></p>
<p>Let’s start with some of the transformations we discussed in lecture. First we look at reflection about the x-axis:</p>
<div class="sourceCode" id="cb410"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb410-1"><a href="principal-components-analysis.html#cb410-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this matrix represents reflection about the x-axis</span></span>
<span id="cb410-2"><a href="principal-components-analysis.html#cb410-2" aria-hidden="true" tabindex="-1"></a>A <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">1</span>,  <span class="dv">0</span>,</span>
<span id="cb410-3"><a href="principal-components-analysis.html#cb410-3" aria-hidden="true" tabindex="-1"></a>              <span class="dv">0</span>, <span class="sc">-</span><span class="dv">1</span>), <span class="at">byrow=</span><span class="cn">TRUE</span>, <span class="at">nrow=</span><span class="dv">2</span>)</span>
<span id="cb410-4"><a href="principal-components-analysis.html#cb410-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb410-5"><a href="principal-components-analysis.html#cb410-5" aria-hidden="true" tabindex="-1"></a><span class="co"># apply this transformation to each of the points in our vector `sin.xy`</span></span>
<span id="cb410-6"><a href="principal-components-analysis.html#cb410-6" aria-hidden="true" tabindex="-1"></a>Ax <span class="ot">&lt;-</span> A <span class="sc">%*%</span> <span class="fu">t</span>(sin.xy)</span>
<span id="cb410-7"><a href="principal-components-analysis.html#cb410-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb410-8"><a href="principal-components-analysis.html#cb410-8" aria-hidden="true" tabindex="-1"></a><span class="co"># creae a data frame from this transformation</span></span>
<span id="cb410-9"><a href="principal-components-analysis.html#cb410-9" aria-hidden="true" tabindex="-1"></a>Ax.df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">t</span>(Ax)) <span class="sc">%&gt;%</span> <span class="fu">rename</span>(<span class="at">x =</span> V1, <span class="at">y =</span> V2)</span>
<span id="cb410-10"><a href="principal-components-analysis.html#cb410-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb410-11"><a href="principal-components-analysis.html#cb410-11" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the transformed points</span></span>
<span id="cb410-12"><a href="principal-components-analysis.html#cb410-12" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(Ax.df, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y)) <span class="sc">+</span> <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-407-1.png" width="576" /></p>
<p>If instead we apply a matrix that represents a shear in the x-axis we get:</p>
<div class="sourceCode" id="cb411"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb411-1"><a href="principal-components-analysis.html#cb411-1" aria-hidden="true" tabindex="-1"></a><span class="co"># shear parallel to the x-axis</span></span>
<span id="cb411-2"><a href="principal-components-analysis.html#cb411-2" aria-hidden="true" tabindex="-1"></a>A <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">1</span>, <span class="fl">1.5</span>, </span>
<span id="cb411-3"><a href="principal-components-analysis.html#cb411-3" aria-hidden="true" tabindex="-1"></a>              <span class="dv">0</span>, <span class="dv">1</span>), <span class="at">byrow=</span><span class="cn">TRUE</span>, <span class="at">nrow=</span><span class="dv">2</span>)</span>
<span id="cb411-4"><a href="principal-components-analysis.html#cb411-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb411-5"><a href="principal-components-analysis.html#cb411-5" aria-hidden="true" tabindex="-1"></a>Ax <span class="ot">&lt;-</span> A <span class="sc">%*%</span> <span class="fu">t</span>(sin.xy)</span>
<span id="cb411-6"><a href="principal-components-analysis.html#cb411-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb411-7"><a href="principal-components-analysis.html#cb411-7" aria-hidden="true" tabindex="-1"></a>Ax.df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">t</span>(Ax)) <span class="sc">%&gt;%</span> <span class="fu">rename</span>(<span class="at">x =</span> V1, <span class="at">y =</span> V2)</span>
<span id="cb411-8"><a href="principal-components-analysis.html#cb411-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(Ax.df, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y=</span> y)) <span class="sc">+</span> <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-408-1.png" width="576" /></p>
<p>And finally rotation:</p>
<div class="sourceCode" id="cb412"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb412-1"><a href="principal-components-analysis.html#cb412-1" aria-hidden="true" tabindex="-1"></a><span class="co"># rotation by pi/2 radians (90 degrees)</span></span>
<span id="cb412-2"><a href="principal-components-analysis.html#cb412-2" aria-hidden="true" tabindex="-1"></a>A <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fu">cos</span>(pi<span class="sc">/</span><span class="dv">2</span>), <span class="sc">-</span><span class="fu">sin</span>(pi<span class="sc">/</span><span class="dv">2</span>), </span>
<span id="cb412-3"><a href="principal-components-analysis.html#cb412-3" aria-hidden="true" tabindex="-1"></a>              <span class="fu">sin</span>(pi<span class="sc">/</span><span class="dv">2</span>), <span class="fu">cos</span>(pi<span class="sc">/</span><span class="dv">2</span>)), <span class="at">byrow=</span><span class="cn">TRUE</span>, <span class="at">nrow=</span><span class="dv">2</span>)</span>
<span id="cb412-4"><a href="principal-components-analysis.html#cb412-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-5"><a href="principal-components-analysis.html#cb412-5" aria-hidden="true" tabindex="-1"></a>Ax <span class="ot">&lt;-</span> A <span class="sc">%*%</span> <span class="fu">t</span>(sin.xy)</span>
<span id="cb412-6"><a href="principal-components-analysis.html#cb412-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb412-7"><a href="principal-components-analysis.html#cb412-7" aria-hidden="true" tabindex="-1"></a>Ax.df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">t</span>(Ax)) <span class="sc">%&gt;%</span> <span class="fu">rename</span>(<span class="at">x =</span> V1, <span class="at">y =</span> V2)</span>
<span id="cb412-8"><a href="principal-components-analysis.html#cb412-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(Ax.df, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y=</span> y)) <span class="sc">+</span> <span class="fu">geom_point</span>()</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-409-1.png" width="576" /></p>
</div>
<div id="eigenanalysis-in-r" class="section level2 hasAnchor" number="13.3">
<h2><span class="header-section-number">13.3</span> Eigenanalysis in R<a href="principal-components-analysis.html#eigenanalysis-in-r" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>As we discussed in lecture, the eigenvectors of a square matrix, <span class="math inline">\(\mathbf{A}\)</span>, point in the directions that are unchanged by the transformation specified by <span class="math inline">\(\mathbf{A}\)</span>.</p>
<p>Let’s start with yet another transformation matrix, <span class="math inline">\(\mathbf{A}\)</span>, the effects of which are illustrated below:</p>
<div class="sourceCode" id="cb413"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb413-1"><a href="principal-components-analysis.html#cb413-1" aria-hidden="true" tabindex="-1"></a>A <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fl">1.0</span>, <span class="fl">1.5</span>,</span>
<span id="cb413-2"><a href="principal-components-analysis.html#cb413-2" aria-hidden="true" tabindex="-1"></a>              <span class="dv">0</span>,   <span class="fl">2.0</span>), <span class="at">byrow=</span><span class="cn">TRUE</span>, <span class="at">nrow=</span><span class="dv">2</span>)</span>
<span id="cb413-3"><a href="principal-components-analysis.html#cb413-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb413-4"><a href="principal-components-analysis.html#cb413-4" aria-hidden="true" tabindex="-1"></a>Ax <span class="ot">&lt;-</span> A <span class="sc">%*%</span> <span class="fu">t</span>(sin.xy)</span>
<span id="cb413-5"><a href="principal-components-analysis.html#cb413-5" aria-hidden="true" tabindex="-1"></a>Ax.df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">t</span>(Ax)) <span class="sc">%&gt;%</span> <span class="fu">rename</span>(<span class="at">x =</span> V1, <span class="at">y =</span> V2)</span>
<span id="cb413-6"><a href="principal-components-analysis.html#cb413-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb413-7"><a href="principal-components-analysis.html#cb413-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y)) <span class="sc">+</span> </span>
<span id="cb413-8"><a href="principal-components-analysis.html#cb413-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> sin.xy, <span class="at">shape=</span><span class="dv">16</span>, <span class="at">alpha=</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb413-9"><a href="principal-components-analysis.html#cb413-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> Ax.df, <span class="at">shape=</span><span class="dv">17</span>, <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">color=</span><span class="st">&#39;red&#39;</span>) </span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-410-1.png" width="576" /></p>
<p>This transformation is a combination of stretching (dilation) in the Y-axis, and shear parallel to the x-axis.</p>
<p>The <code>eigen()</code> function computes the eigenvalues and eigenvectors of a square matrix:</p>
<div class="sourceCode" id="cb414"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb414-1"><a href="principal-components-analysis.html#cb414-1" aria-hidden="true" tabindex="-1"></a>eigen.A <span class="ot">&lt;-</span> <span class="fu">eigen</span>(A)</span>
<span id="cb414-2"><a href="principal-components-analysis.html#cb414-2" aria-hidden="true" tabindex="-1"></a>eigen.A</span>
<span id="cb414-3"><a href="principal-components-analysis.html#cb414-3" aria-hidden="true" tabindex="-1"></a><span class="do">## eigen() decomposition</span></span>
<span id="cb414-4"><a href="principal-components-analysis.html#cb414-4" aria-hidden="true" tabindex="-1"></a><span class="do">## $values</span></span>
<span id="cb414-5"><a href="principal-components-analysis.html#cb414-5" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 2 1</span></span>
<span id="cb414-6"><a href="principal-components-analysis.html#cb414-6" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb414-7"><a href="principal-components-analysis.html#cb414-7" aria-hidden="true" tabindex="-1"></a><span class="do">## $vectors</span></span>
<span id="cb414-8"><a href="principal-components-analysis.html#cb414-8" aria-hidden="true" tabindex="-1"></a><span class="do">##           [,1] [,2]</span></span>
<span id="cb414-9"><a href="principal-components-analysis.html#cb414-9" aria-hidden="true" tabindex="-1"></a><span class="do">## [1,] 0.8320503    1</span></span>
<span id="cb414-10"><a href="principal-components-analysis.html#cb414-10" aria-hidden="true" tabindex="-1"></a><span class="do">## [2,] 0.5547002    0</span></span></code></pre></div>
<p>The following relationships relate <span class="math inline">\(\mathbf{A}\)</span> to it’s eigenvectors and eigenvalues:</p>
<p><span class="math display">\[\mathbf{V}^{-1}\mathbf{A}\mathbf{V}  =  \mathbf{D} \]</span></p>
<p><span class="math display">\[\mathbf{A}  =  \mathbf{V}\mathbf{D}\mathbf{V}^{-1}\]</span></p>
<p>where <span class="math inline">\(\mathbf{V}\)</span> is a matrix where the columns represent the eigenvectors, and <span class="math inline">\(\mathbf{D}\)</span> is a diagonal matrix of eigenvalues.</p>
<p>Let’s confirm that those relationships hold for our example:</p>
<div class="sourceCode" id="cb415"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb415-1"><a href="principal-components-analysis.html#cb415-1" aria-hidden="true" tabindex="-1"></a>V <span class="ot">&lt;-</span> eigen.A<span class="sc">$</span>vectors</span>
<span id="cb415-2"><a href="principal-components-analysis.html#cb415-2" aria-hidden="true" tabindex="-1"></a>D <span class="ot">&lt;-</span> <span class="fu">diag</span>(eigen.A<span class="sc">$</span>values) <span class="co"># diagonal matrix of eigenvalues</span></span>
<span id="cb415-3"><a href="principal-components-analysis.html#cb415-3" aria-hidden="true" tabindex="-1"></a>Vinv <span class="ot">&lt;-</span> <span class="fu">solve</span>(V)</span>
<span id="cb415-4"><a href="principal-components-analysis.html#cb415-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb415-5"><a href="principal-components-analysis.html#cb415-5" aria-hidden="true" tabindex="-1"></a>V <span class="sc">%*%</span> D <span class="sc">%*%</span> Vinv  <span class="co"># reconstruct our original matrix (see lecture slides)</span></span>
<span id="cb415-6"><a href="principal-components-analysis.html#cb415-6" aria-hidden="true" tabindex="-1"></a><span class="do">##      [,1] [,2]</span></span>
<span id="cb415-7"><a href="principal-components-analysis.html#cb415-7" aria-hidden="true" tabindex="-1"></a><span class="do">## [1,]    1  1.5</span></span>
<span id="cb415-8"><a href="principal-components-analysis.html#cb415-8" aria-hidden="true" tabindex="-1"></a><span class="do">## [2,]    0  2.0</span></span></code></pre></div>
<div class="sourceCode" id="cb416"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb416-1"><a href="principal-components-analysis.html#cb416-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(Vinv <span class="sc">%*%</span> A <span class="sc">%*%</span> V, D) <span class="co"># test &#39;near equality&#39;</span></span>
<span id="cb416-2"><a href="principal-components-analysis.html#cb416-2" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] TRUE</span></span></code></pre></div>
<div class="sourceCode" id="cb417"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb417-1"><a href="principal-components-analysis.html#cb417-1" aria-hidden="true" tabindex="-1"></a>V[,<span class="dv">1</span>] <span class="sc">%*%</span> V[,<span class="dv">2</span>] <span class="co"># note that the eigenvectors are NOT orthogonal. Why?</span></span>
<span id="cb417-2"><a href="principal-components-analysis.html#cb417-2" aria-hidden="true" tabindex="-1"></a><span class="do">##           [,1]</span></span>
<span id="cb417-3"><a href="principal-components-analysis.html#cb417-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1,] 0.8320503</span></span></code></pre></div>
<p>Since we’re dealing with a 2D transformation matrix, we can easily calculate the slopes of the eigenvectors:</p>
<div class="sourceCode" id="cb418"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb418-1"><a href="principal-components-analysis.html#cb418-1" aria-hidden="true" tabindex="-1"></a>eigenvec.A.slope1 <span class="ot">=</span> eigen.A<span class="sc">$</span>vectors[<span class="dv">2</span>,<span class="dv">1</span>]<span class="sc">/</span>eigen.A<span class="sc">$</span>vectors[<span class="dv">1</span>,<span class="dv">1</span>]</span>
<span id="cb418-2"><a href="principal-components-analysis.html#cb418-2" aria-hidden="true" tabindex="-1"></a>eigenvec.A.slope2 <span class="ot">=</span> eigen.A<span class="sc">$</span>vectors[<span class="dv">2</span>,<span class="dv">2</span>]<span class="sc">/</span>eigen.A<span class="sc">$</span>vectors[<span class="dv">1</span>,<span class="dv">2</span>]</span></code></pre></div>
<p>Now we plot the eigenvectors to show the directions that are unaffected by the transformation represented by the matrix <span class="math inline">\(\mathbf{A}\)</span>:</p>
<div class="sourceCode" id="cb419"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb419-1"><a href="principal-components-analysis.html#cb419-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb419-2"><a href="principal-components-analysis.html#cb419-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Note that the slopes can be infinite (inf), indicating a vertical eigenvector (parallel to y-axis)</span></span>
<span id="cb419-3"><a href="principal-components-analysis.html#cb419-3" aria-hidden="true" tabindex="-1"></a><span class="co"># this function is to draw the eigenvectors correctly regardless of the slope</span></span>
<span id="cb419-4"><a href="principal-components-analysis.html#cb419-4" aria-hidden="true" tabindex="-1"></a>geom_ab_or_vline <span class="ot">&lt;-</span> <span class="cf">function</span>(slope, intercept, ...) {</span>
<span id="cb419-5"><a href="principal-components-analysis.html#cb419-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">is.infinite</span>(slope)) {</span>
<span id="cb419-6"><a href="principal-components-analysis.html#cb419-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">geom_vline</span>(<span class="at">yintercept =</span> intercept, ...))</span>
<span id="cb419-7"><a href="principal-components-analysis.html#cb419-7" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb419-8"><a href="principal-components-analysis.html#cb419-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">geom_abline</span>(<span class="at">slope=</span>slope, <span class="at">intercept=</span>intercept, ...))</span>
<span id="cb419-9"><a href="principal-components-analysis.html#cb419-9" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb419-10"><a href="principal-components-analysis.html#cb419-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb419-11"><a href="principal-components-analysis.html#cb419-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb419-12"><a href="principal-components-analysis.html#cb419-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb419-13"><a href="principal-components-analysis.html#cb419-13" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y)) <span class="sc">+</span></span>
<span id="cb419-14"><a href="principal-components-analysis.html#cb419-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> sin.xy, <span class="at">alpha=</span><span class="fl">0.5</span>) <span class="sc">+</span> </span>
<span id="cb419-15"><a href="principal-components-analysis.html#cb419-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> Ax.df, <span class="at">color=</span><span class="st">&#39;red&#39;</span>, <span class="at">alpha=</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb419-16"><a href="principal-components-analysis.html#cb419-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_ab_or_vline</span>(<span class="at">slope =</span> eigenvec.A.slope1,<span class="at">intercept=</span><span class="dv">0</span>,</span>
<span id="cb419-17"><a href="principal-components-analysis.html#cb419-17" aria-hidden="true" tabindex="-1"></a>              <span class="at">color=</span><span class="st">&#39;red&#39;</span>,<span class="at">linetype=</span><span class="st">&#39;dashed&#39;</span>) <span class="sc">+</span></span>
<span id="cb419-18"><a href="principal-components-analysis.html#cb419-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_ab_or_vline</span>(<span class="at">slope =</span> eigenvec.A.slope2,<span class="at">intercept=</span><span class="dv">0</span>,</span>
<span id="cb419-19"><a href="principal-components-analysis.html#cb419-19" aria-hidden="true" tabindex="-1"></a>              <span class="at">color=</span><span class="st">&#39;red&#39;</span>,<span class="at">linetype=</span><span class="st">&#39;dashed&#39;</span>) <span class="sc">+</span> </span>
<span id="cb419-20"><a href="principal-components-analysis.html#cb419-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>()</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-416-1.png" width="576" /></p>
<p>Referring back to the eigenvectors, we see that it’s the second eigenvector that is represented by the horizontal line:</p>
<div class="sourceCode" id="cb420"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb420-1"><a href="principal-components-analysis.html#cb420-1" aria-hidden="true" tabindex="-1"></a>eigen.A<span class="sc">$</span>vectors</span>
<span id="cb420-2"><a href="principal-components-analysis.html#cb420-2" aria-hidden="true" tabindex="-1"></a><span class="do">##           [,1] [,2]</span></span>
<span id="cb420-3"><a href="principal-components-analysis.html#cb420-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1,] 0.8320503    1</span></span>
<span id="cb420-4"><a href="principal-components-analysis.html#cb420-4" aria-hidden="true" tabindex="-1"></a><span class="do">## [2,] 0.5547002    0</span></span></code></pre></div>
<p>The implies that lines pointing in the horizontal are unchanged in the direction by the transformation represented by A. To drive home this point, here’s another configuration of points and the corresponding points under the same transformation:</p>
<div class="sourceCode" id="cb421"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb421-1"><a href="principal-components-analysis.html#cb421-1" aria-hidden="true" tabindex="-1"></a>horiz.x <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="dv">1</span>,<span class="dv">1</span>,<span class="at">by=</span><span class="fl">0.1</span>)</span>
<span id="cb421-2"><a href="principal-components-analysis.html#cb421-2" aria-hidden="true" tabindex="-1"></a>horiz.y <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, <span class="fu">length</span>(horiz.x))</span>
<span id="cb421-3"><a href="principal-components-analysis.html#cb421-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb421-4"><a href="principal-components-analysis.html#cb421-4" aria-hidden="true" tabindex="-1"></a>horiz.x2 <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="fl">0.5</span>,<span class="fl">0.5</span>,<span class="at">by=</span><span class="fl">0.1</span>)</span>
<span id="cb421-5"><a href="principal-components-analysis.html#cb421-5" aria-hidden="true" tabindex="-1"></a>horiz.y2 <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="fl">0.5</span>, <span class="fu">length</span>(horiz.x2))</span>
<span id="cb421-6"><a href="principal-components-analysis.html#cb421-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb421-7"><a href="principal-components-analysis.html#cb421-7" aria-hidden="true" tabindex="-1"></a>horiz.x3 <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="fl">0.5</span>,<span class="fl">0.5</span>,<span class="at">by=</span><span class="fl">0.1</span>)</span>
<span id="cb421-8"><a href="principal-components-analysis.html#cb421-8" aria-hidden="true" tabindex="-1"></a>horiz.y3 <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="sc">-</span><span class="fl">0.5</span>, <span class="fu">length</span>(horiz.x3))</span>
<span id="cb421-9"><a href="principal-components-analysis.html#cb421-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb421-10"><a href="principal-components-analysis.html#cb421-10" aria-hidden="true" tabindex="-1"></a>vert.y <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="dv">1</span>,<span class="dv">1</span>,<span class="at">by=</span><span class="fl">0.1</span>)</span>
<span id="cb421-11"><a href="principal-components-analysis.html#cb421-11" aria-hidden="true" tabindex="-1"></a>vert.x <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, <span class="fu">length</span>(vert.y))</span>
<span id="cb421-12"><a href="principal-components-analysis.html#cb421-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb421-13"><a href="principal-components-analysis.html#cb421-13" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">c</span>(horiz.x, horiz.x2, horiz.x3, vert.x)</span>
<span id="cb421-14"><a href="principal-components-analysis.html#cb421-14" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">c</span>(horiz.y, horiz.y2, horiz.y3, vert.y)</span>
<span id="cb421-15"><a href="principal-components-analysis.html#cb421-15" aria-hidden="true" tabindex="-1"></a>cross.df <span class="ot">&lt;-</span> <span class="fu">data_frame</span>(<span class="at">x =</span> x, <span class="at">y =</span> y)</span>
<span id="cb421-16"><a href="principal-components-analysis.html#cb421-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb421-17"><a href="principal-components-analysis.html#cb421-17" aria-hidden="true" tabindex="-1"></a>Ax <span class="ot">&lt;-</span> A <span class="sc">%*%</span> <span class="fu">t</span>(cross.df)</span>
<span id="cb421-18"><a href="principal-components-analysis.html#cb421-18" aria-hidden="true" tabindex="-1"></a>Ax.df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">t</span>(Ax)) <span class="sc">%&gt;%</span> <span class="fu">rename</span>(<span class="at">x =</span> V1, <span class="at">y =</span> V2)</span>
<span id="cb421-19"><a href="principal-components-analysis.html#cb421-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb421-20"><a href="principal-components-analysis.html#cb421-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb421-21"><a href="principal-components-analysis.html#cb421-21" aria-hidden="true" tabindex="-1"></a>plot.A <span class="ot">&lt;-</span> </span>
<span id="cb421-22"><a href="principal-components-analysis.html#cb421-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(cross.df, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">y=</span>y)) <span class="sc">+</span> </span>
<span id="cb421-23"><a href="principal-components-analysis.html#cb421-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb421-24"><a href="principal-components-analysis.html#cb421-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lims</span>(<span class="at">x=</span><span class="fu">c</span>(<span class="sc">-</span><span class="fl">1.5</span>,<span class="fl">1.5</span>), <span class="at">y=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">2</span>,<span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb421-25"><a href="principal-components-analysis.html#cb421-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>() <span class="sc">+</span> </span>
<span id="cb421-26"><a href="principal-components-analysis.html#cb421-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;X&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Y&quot;</span>, <span class="at">title=</span><span class="st">&quot;Points prior to transformation&quot;</span>)</span>
<span id="cb421-27"><a href="principal-components-analysis.html#cb421-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb421-28"><a href="principal-components-analysis.html#cb421-28" aria-hidden="true" tabindex="-1"></a>plot.B <span class="ot">&lt;-</span> </span>
<span id="cb421-29"><a href="principal-components-analysis.html#cb421-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(Ax.df, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x=</span>x,<span class="at">y=</span>y)) <span class="sc">+</span> </span>
<span id="cb421-30"><a href="principal-components-analysis.html#cb421-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">color=</span><span class="st">&#39;red&#39;</span>) <span class="sc">+</span></span>
<span id="cb421-31"><a href="principal-components-analysis.html#cb421-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lims</span>(<span class="at">x=</span><span class="fu">c</span>(<span class="sc">-</span><span class="fl">1.5</span>,<span class="fl">1.5</span>), <span class="at">y=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">2</span>,<span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb421-32"><a href="principal-components-analysis.html#cb421-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>() <span class="sc">+</span> </span>
<span id="cb421-33"><a href="principal-components-analysis.html#cb421-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;X&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Y&quot;</span>, <span class="at">title=</span><span class="st">&quot;Points after transformation&quot;</span>)</span>
<span id="cb421-34"><a href="principal-components-analysis.html#cb421-34" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb421-35"><a href="principal-components-analysis.html#cb421-35" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(plot.A, plot.B)</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-418-1.png" width="576" /></p>
</div>
<div id="principal-components-analysis-in-r" class="section level2 hasAnchor" number="13.4">
<h2><span class="header-section-number">13.4</span> Principal Components Analysis in R<a href="principal-components-analysis.html#principal-components-analysis-in-r" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>There are two functions in R for carrying out PCA - <code>princomp()</code> and
<code>prcomp()</code>. The <code>princomp()</code> function uses the
<code>eigen()</code> function to carry out the analysis on the covariance
matrix or correlation matrix, while carries out an
equivalent analysis, starting from a data matrix, using a technique called
singular value decomposition (SVD). The SVD routine has greater numerical
accuracy, so the <code>prcomp()</code> function should generally be preferred. The
<code>princomp()</code> function is useful when you don’t have access to
the original data, but you do have a covariance or correlation matrix (a
frequent situation when re-analyzing data from the literature). We’ll concentrate on using the <code>prcomp()</code> function.</p>
<div id="bioenv-dataset" class="section level3 hasAnchor" number="13.4.1">
<h3><span class="header-section-number">13.4.1</span> Bioenv dataset<a href="principal-components-analysis.html#bioenv-dataset" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To demonstrate PCA we’ll use a dataset called `bioenv.txt’ (see class wiki), obtained from a book called “Biplots in Practice” (M. Greenacre, 2010). Here is Greenacre’s description of the dataset:</p>
<blockquote>
<p>The context is in marine biology and the data consist of two sets of variables observed at the same locations on the sea-bed: the first is a set of biological variables, the counts of five groups of species, and the second is a set of four environmental variables. The data set, called “bioenv”, is shown in Exhibit 2.1. The species groups are abbreviated as “a” to “e”. The environmental variables are “pollution”, a composite index of pollution combining measurements of heavy metal concentrations and hydrocarbons; depth, the depth in metres of the sea-bed where the sample was taken; “temperature”, the temperature of the water at the sampling point; and “sediment”, a classification of the substrate of the sample into one of three sediment categories.</p>
</blockquote>
<p>We’ll start by reading the <code>bioenv.txt</code> data set from the Github repository:</p>
<div class="sourceCode" id="cb422"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb422-1"><a href="principal-components-analysis.html#cb422-1" aria-hidden="true" tabindex="-1"></a>bioenv <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">&#39;https://github.com/Bio723-class/example-datasets/raw/master/bioenv.txt&#39;</span>) </span>
<span id="cb422-2"><a href="principal-components-analysis.html#cb422-2" aria-hidden="true" tabindex="-1"></a><span class="do">## New names:</span></span>
<span id="cb422-3"><a href="principal-components-analysis.html#cb422-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Rows: 30 Columns: 10</span></span>
<span id="cb422-4"><a href="principal-components-analysis.html#cb422-4" aria-hidden="true" tabindex="-1"></a><span class="do">## ── Column specification</span></span>
<span id="cb422-5"><a href="principal-components-analysis.html#cb422-5" aria-hidden="true" tabindex="-1"></a><span class="do">## ──────────────────────────────────── Delimiter: &quot;\t&quot; chr</span></span>
<span id="cb422-6"><a href="principal-components-analysis.html#cb422-6" aria-hidden="true" tabindex="-1"></a><span class="do">## (2): ...1, Sediment dbl (8): a, b, c, d, e, Pollution,</span></span>
<span id="cb422-7"><a href="principal-components-analysis.html#cb422-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Depth, Temperature</span></span>
<span id="cb422-8"><a href="principal-components-analysis.html#cb422-8" aria-hidden="true" tabindex="-1"></a><span class="do">## ℹ Use `spec()` to retrieve the full column specification</span></span>
<span id="cb422-9"><a href="principal-components-analysis.html#cb422-9" aria-hidden="true" tabindex="-1"></a><span class="do">## for this data. ℹ Specify the column types or set</span></span>
<span id="cb422-10"><a href="principal-components-analysis.html#cb422-10" aria-hidden="true" tabindex="-1"></a><span class="do">## `show_col_types = FALSE` to quiet this message.</span></span>
<span id="cb422-11"><a href="principal-components-analysis.html#cb422-11" aria-hidden="true" tabindex="-1"></a><span class="do">## • `` -&gt; `...1`</span></span>
<span id="cb422-12"><a href="principal-components-analysis.html#cb422-12" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(bioenv)</span>
<span id="cb422-13"><a href="principal-components-analysis.html#cb422-13" aria-hidden="true" tabindex="-1"></a><span class="do">##  [1] &quot;...1&quot;        &quot;a&quot;           &quot;b&quot;           &quot;c&quot;           &quot;d&quot;          </span></span>
<span id="cb422-14"><a href="principal-components-analysis.html#cb422-14" aria-hidden="true" tabindex="-1"></a><span class="do">##  [6] &quot;e&quot;           &quot;Pollution&quot;   &quot;Depth&quot;       &quot;Temperature&quot; &quot;Sediment&quot;</span></span></code></pre></div>
<p>Notice that the first column got assigned the generic name <code>...1</code>. This is because there is a missing column header in the <code>bioenv.txt</code> file. This first column corresponds to the sampling sites. Before we move on let’s give this column a more meaningful name:</p>
<div class="sourceCode" id="cb423"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb423-1"><a href="principal-components-analysis.html#cb423-1" aria-hidden="true" tabindex="-1"></a>bioenv <span class="ot">&lt;-</span></span>
<span id="cb423-2"><a href="principal-components-analysis.html#cb423-2" aria-hidden="true" tabindex="-1"></a>  bioenv <span class="sc">%&gt;%</span> <span class="fu">rename</span>(<span class="at">Site =</span> <span class="st">&quot;...1&quot;</span>)</span>
<span id="cb423-3"><a href="principal-components-analysis.html#cb423-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb423-4"><a href="principal-components-analysis.html#cb423-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(bioenv)</span>
<span id="cb423-5"><a href="principal-components-analysis.html#cb423-5" aria-hidden="true" tabindex="-1"></a><span class="do">##  [1] &quot;Site&quot;        &quot;a&quot;           &quot;b&quot;           &quot;c&quot;           &quot;d&quot;          </span></span>
<span id="cb423-6"><a href="principal-components-analysis.html#cb423-6" aria-hidden="true" tabindex="-1"></a><span class="do">##  [6] &quot;e&quot;           &quot;Pollution&quot;   &quot;Depth&quot;       &quot;Temperature&quot; &quot;Sediment&quot;</span></span></code></pre></div>
<p>The columns labeled <code>a</code> to <code>e</code> contain the counts of the five species at each site, while the remaining columns give additional information about the physical properties of each sampling site. For our purposes today we’ll confine our attention to the abundance data.</p>
<div class="sourceCode" id="cb424"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb424-1"><a href="principal-components-analysis.html#cb424-1" aria-hidden="true" tabindex="-1"></a>abundance <span class="ot">&lt;-</span> </span>
<span id="cb424-2"><a href="principal-components-analysis.html#cb424-2" aria-hidden="true" tabindex="-1"></a>  bioenv <span class="sc">%&gt;%</span></span>
<span id="cb424-3"><a href="principal-components-analysis.html#cb424-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(Site, a, b, c, d, e)</span>
<span id="cb424-4"><a href="principal-components-analysis.html#cb424-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb424-5"><a href="principal-components-analysis.html#cb424-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(abundance)</span>
<span id="cb424-6"><a href="principal-components-analysis.html#cb424-6" aria-hidden="true" tabindex="-1"></a><span class="do">## # A tibble: 6 × 6</span></span>
<span id="cb424-7"><a href="principal-components-analysis.html#cb424-7" aria-hidden="true" tabindex="-1"></a><span class="do">##   Site      a     b     c     d     e</span></span>
<span id="cb424-8"><a href="principal-components-analysis.html#cb424-8" aria-hidden="true" tabindex="-1"></a><span class="do">##   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span id="cb424-9"><a href="principal-components-analysis.html#cb424-9" aria-hidden="true" tabindex="-1"></a><span class="do">## 1 s1        0     2     9    14     2</span></span>
<span id="cb424-10"><a href="principal-components-analysis.html#cb424-10" aria-hidden="true" tabindex="-1"></a><span class="do">## 2 s2       26     4    13    11     0</span></span>
<span id="cb424-11"><a href="principal-components-analysis.html#cb424-11" aria-hidden="true" tabindex="-1"></a><span class="do">## 3 s3        0    10     9     8     0</span></span>
<span id="cb424-12"><a href="principal-components-analysis.html#cb424-12" aria-hidden="true" tabindex="-1"></a><span class="do">## 4 s4        0     0    15     3     0</span></span>
<span id="cb424-13"><a href="principal-components-analysis.html#cb424-13" aria-hidden="true" tabindex="-1"></a><span class="do">## 5 s5       13     5     3    10     7</span></span>
<span id="cb424-14"><a href="principal-components-analysis.html#cb424-14" aria-hidden="true" tabindex="-1"></a><span class="do">## 6 s6       31    21    13    16     5</span></span></code></pre></div>
<p>The data is currently in a “wide” format. For the purposes of plotting it will be more convenient to generate a “long” version of the data using functions from the tidyr library (see the Data Wrangling chapter).</p>
<div class="sourceCode" id="cb425"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb425-1"><a href="principal-components-analysis.html#cb425-1" aria-hidden="true" tabindex="-1"></a>long.abundance <span class="ot">&lt;-</span></span>
<span id="cb425-2"><a href="principal-components-analysis.html#cb425-2" aria-hidden="true" tabindex="-1"></a>  abundance <span class="sc">%&gt;%</span></span>
<span id="cb425-3"><a href="principal-components-analysis.html#cb425-3" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">gather</span>(Species, Count, <span class="sc">-</span>Site)</span>
<span id="cb425-4"><a href="principal-components-analysis.html#cb425-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb425-5"><a href="principal-components-analysis.html#cb425-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(long.abundance)</span>
<span id="cb425-6"><a href="principal-components-analysis.html#cb425-6" aria-hidden="true" tabindex="-1"></a><span class="do">## # A tibble: 6 × 3</span></span>
<span id="cb425-7"><a href="principal-components-analysis.html#cb425-7" aria-hidden="true" tabindex="-1"></a><span class="do">##   Site  Species Count</span></span>
<span id="cb425-8"><a href="principal-components-analysis.html#cb425-8" aria-hidden="true" tabindex="-1"></a><span class="do">##   &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;</span></span>
<span id="cb425-9"><a href="principal-components-analysis.html#cb425-9" aria-hidden="true" tabindex="-1"></a><span class="do">## 1 s1    a           0</span></span>
<span id="cb425-10"><a href="principal-components-analysis.html#cb425-10" aria-hidden="true" tabindex="-1"></a><span class="do">## 2 s2    a          26</span></span>
<span id="cb425-11"><a href="principal-components-analysis.html#cb425-11" aria-hidden="true" tabindex="-1"></a><span class="do">## 3 s3    a           0</span></span>
<span id="cb425-12"><a href="principal-components-analysis.html#cb425-12" aria-hidden="true" tabindex="-1"></a><span class="do">## 4 s4    a           0</span></span>
<span id="cb425-13"><a href="principal-components-analysis.html#cb425-13" aria-hidden="true" tabindex="-1"></a><span class="do">## 5 s5    a          13</span></span>
<span id="cb425-14"><a href="principal-components-analysis.html#cb425-14" aria-hidden="true" tabindex="-1"></a><span class="do">## 6 s6    a          31</span></span></code></pre></div>
<div class="sourceCode" id="cb426"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb426-1"><a href="principal-components-analysis.html#cb426-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(long.abundance, <span class="fu">aes</span>(<span class="at">x =</span> Species, <span class="at">y =</span> Count)) <span class="sc">+</span></span>
<span id="cb426-2"><a href="principal-components-analysis.html#cb426-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb426-3"><a href="principal-components-analysis.html#cb426-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Species&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Count&quot;</span>, <span class="at">title=</span><span class="st">&quot;Distribution of</span><span class="sc">\n</span><span class="st">Species Counts per Site&quot;</span>)</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-423-1.png" width="576" /></p>
<p>From the boxplot it looks like the counts for species `e’ are smaller on average, and less variable. The mean and variance functions confirm that.</p>
<div class="sourceCode" id="cb427"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb427-1"><a href="principal-components-analysis.html#cb427-1" aria-hidden="true" tabindex="-1"></a>long.abundance <span class="sc">%&gt;%</span></span>
<span id="cb427-2"><a href="principal-components-analysis.html#cb427-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(Species) <span class="sc">%&gt;%</span></span>
<span id="cb427-3"><a href="principal-components-analysis.html#cb427-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="fu">mean</span>(Count), <span class="fu">var</span>(Count))</span>
<span id="cb427-4"><a href="principal-components-analysis.html#cb427-4" aria-hidden="true" tabindex="-1"></a><span class="do">## # A tibble: 5 × 3</span></span>
<span id="cb427-5"><a href="principal-components-analysis.html#cb427-5" aria-hidden="true" tabindex="-1"></a><span class="do">##   Species `mean(Count)` `var(Count)`</span></span>
<span id="cb427-6"><a href="principal-components-analysis.html#cb427-6" aria-hidden="true" tabindex="-1"></a><span class="do">##   &lt;chr&gt;           &lt;dbl&gt;        &lt;dbl&gt;</span></span>
<span id="cb427-7"><a href="principal-components-analysis.html#cb427-7" aria-hidden="true" tabindex="-1"></a><span class="do">## 1 a               13.5         158. </span></span>
<span id="cb427-8"><a href="principal-components-analysis.html#cb427-8" aria-hidden="true" tabindex="-1"></a><span class="do">## 2 b                8.73         83.4</span></span>
<span id="cb427-9"><a href="principal-components-analysis.html#cb427-9" aria-hidden="true" tabindex="-1"></a><span class="do">## 3 c                8.4          73.6</span></span>
<span id="cb427-10"><a href="principal-components-analysis.html#cb427-10" aria-hidden="true" tabindex="-1"></a><span class="do">## 4 d               10.9          44.4</span></span>
<span id="cb427-11"><a href="principal-components-analysis.html#cb427-11" aria-hidden="true" tabindex="-1"></a><span class="do">## 5 e                2.97         15.7</span></span></code></pre></div>
<p>A correlation matrix suggests weak to moderate associations between the variables:</p>
<div class="sourceCode" id="cb428"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb428-1"><a href="principal-components-analysis.html#cb428-1" aria-hidden="true" tabindex="-1"></a>abundance.only <span class="ot">&lt;-</span></span>
<span id="cb428-2"><a href="principal-components-analysis.html#cb428-2" aria-hidden="true" tabindex="-1"></a>  abundance <span class="sc">%&gt;%</span></span>
<span id="cb428-3"><a href="principal-components-analysis.html#cb428-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>Site) <span class="co"># drop the Site column</span></span>
<span id="cb428-4"><a href="principal-components-analysis.html#cb428-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb428-5"><a href="principal-components-analysis.html#cb428-5" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(abundance.only)</span>
<span id="cb428-6"><a href="principal-components-analysis.html#cb428-6" aria-hidden="true" tabindex="-1"></a><span class="do">##            a           b           c            d            e</span></span>
<span id="cb428-7"><a href="principal-components-analysis.html#cb428-7" aria-hidden="true" tabindex="-1"></a><span class="do">## a  1.0000000  0.67339954 -0.23992888  0.358192050  0.273522301</span></span>
<span id="cb428-8"><a href="principal-components-analysis.html#cb428-8" aria-hidden="true" tabindex="-1"></a><span class="do">## b  0.6733995  1.00000000 -0.08041947  0.501834036  0.036914702</span></span>
<span id="cb428-9"><a href="principal-components-analysis.html#cb428-9" aria-hidden="true" tabindex="-1"></a><span class="do">## c -0.2399289 -0.08041947  1.00000000  0.081504483 -0.343540453</span></span>
<span id="cb428-10"><a href="principal-components-analysis.html#cb428-10" aria-hidden="true" tabindex="-1"></a><span class="do">## d  0.3581921  0.50183404  0.08150448  1.000000000 -0.004048517</span></span>
<span id="cb428-11"><a href="principal-components-analysis.html#cb428-11" aria-hidden="true" tabindex="-1"></a><span class="do">## e  0.2735223  0.03691470 -0.34354045 -0.004048517  1.000000000</span></span></code></pre></div>
<p>However a scatterplot matrix generated by the <code>GGally::ggapirs()</code> function suggests that many of the relationships have a strong non-linear element.</p>
<div class="sourceCode" id="cb429"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb429-1"><a href="principal-components-analysis.html#cb429-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggpairs</span>(abundance.only)</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-426-1.png" width="576" /></p>
</div>
<div id="pca-of-the-bioenv-dataset" class="section level3 hasAnchor" number="13.4.2">
<h3><span class="header-section-number">13.4.2</span> PCA of the Bioenv dataset<a href="principal-components-analysis.html#pca-of-the-bioenv-dataset" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Linearity is not a requirement for PCA, as it’s simply a rigid rotation of the original data. So we’ll continue with our analysis after taking a moment to read the help on the <code>prcomp()</code> function that is used to carry-out PCA in R.</p>
<div class="sourceCode" id="cb430"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb430-1"><a href="principal-components-analysis.html#cb430-1" aria-hidden="true" tabindex="-1"></a>abundance.pca <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(abundance.only, <span class="at">center=</span><span class="cn">TRUE</span>, <span class="at">retx=</span><span class="cn">TRUE</span>)</span>
<span id="cb430-2"><a href="principal-components-analysis.html#cb430-2" aria-hidden="true" tabindex="-1"></a>    <span class="co"># center=TRUE mean centers the data</span></span>
<span id="cb430-3"><a href="principal-components-analysis.html#cb430-3" aria-hidden="true" tabindex="-1"></a>    <span class="co"># retx=TRUE returns the PC scores</span></span>
<span id="cb430-4"><a href="principal-components-analysis.html#cb430-4" aria-hidden="true" tabindex="-1"></a>    <span class="co"># if you want to do PCA on the correlation matrix set scale.=TRUE</span></span>
<span id="cb430-5"><a href="principal-components-analysis.html#cb430-5" aria-hidden="true" tabindex="-1"></a>    <span class="co">#    -- notice the period after scale!</span></span>
<span id="cb430-6"><a href="principal-components-analysis.html#cb430-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb430-7"><a href="principal-components-analysis.html#cb430-7" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(abundance.pca)</span>
<span id="cb430-8"><a href="principal-components-analysis.html#cb430-8" aria-hidden="true" tabindex="-1"></a><span class="do">## Importance of components:</span></span>
<span id="cb430-9"><a href="principal-components-analysis.html#cb430-9" aria-hidden="true" tabindex="-1"></a><span class="do">##                            PC1    PC2    PC3     PC4     PC5</span></span>
<span id="cb430-10"><a href="principal-components-analysis.html#cb430-10" aria-hidden="true" tabindex="-1"></a><span class="do">## Standard deviation     14.8653 8.8149 6.2193 5.03477 3.48231</span></span>
<span id="cb430-11"><a href="principal-components-analysis.html#cb430-11" aria-hidden="true" tabindex="-1"></a><span class="do">## Proportion of Variance  0.5895 0.2073 0.1032 0.06763 0.03235</span></span>
<span id="cb430-12"><a href="principal-components-analysis.html#cb430-12" aria-hidden="true" tabindex="-1"></a><span class="do">## Cumulative Proportion   0.5895 0.7968 0.9000 0.96765 1.00000</span></span></code></pre></div>
<p>We see that approximately 59% of the variance in the data is capture by the first PC, and approximately 90% by the first three PCs.</p>
<p>Let’s compare the values return by PCA to what we would get if we carried out eigenanalysis of the covariance matrix that corresponds to our data.</p>
<p>First the list object return by <code>prcomp()</code>:</p>
<div class="sourceCode" id="cb431"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb431-1"><a href="principal-components-analysis.html#cb431-1" aria-hidden="true" tabindex="-1"></a>abundance.pca</span>
<span id="cb431-2"><a href="principal-components-analysis.html#cb431-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Standard deviations (1, .., p=5):</span></span>
<span id="cb431-3"><a href="principal-components-analysis.html#cb431-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 14.865306  8.814912  6.219250  5.034774  3.482308</span></span>
<span id="cb431-4"><a href="principal-components-analysis.html#cb431-4" aria-hidden="true" tabindex="-1"></a><span class="do">## </span></span>
<span id="cb431-5"><a href="principal-components-analysis.html#cb431-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Rotation (n x k) = (5 x 5):</span></span>
<span id="cb431-6"><a href="principal-components-analysis.html#cb431-6" aria-hidden="true" tabindex="-1"></a><span class="do">##           PC1         PC2         PC3         PC4         PC5</span></span>
<span id="cb431-7"><a href="principal-components-analysis.html#cb431-7" aria-hidden="true" tabindex="-1"></a><span class="do">## a  0.81064462  0.07052882 -0.53108427  0.18442140 -0.14771336</span></span>
<span id="cb431-8"><a href="principal-components-analysis.html#cb431-8" aria-hidden="true" tabindex="-1"></a><span class="do">## b  0.51264394 -0.27799671  0.47711910 -0.63418946  0.17342177</span></span>
<span id="cb431-9"><a href="principal-components-analysis.html#cb431-9" aria-hidden="true" tabindex="-1"></a><span class="do">## c -0.16235135 -0.88665551 -0.40897655 -0.01149647  0.14173943</span></span>
<span id="cb431-10"><a href="principal-components-analysis.html#cb431-10" aria-hidden="true" tabindex="-1"></a><span class="do">## d  0.22207108 -0.31665237  0.56250980  0.72941223 -0.04422938</span></span>
<span id="cb431-11"><a href="principal-components-analysis.html#cb431-11" aria-hidden="true" tabindex="-1"></a><span class="do">## e  0.06616623  0.17696554 -0.08141111  0.17781482  0.96231977</span></span></code></pre></div>
<p>And now the corresponding values returned by eigenanaysis of the covariance matrix generated from the abundance data:</p>
<div class="sourceCode" id="cb432"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb432-1"><a href="principal-components-analysis.html#cb432-1" aria-hidden="true" tabindex="-1"></a>eig.abundance <span class="ot">&lt;-</span> <span class="fu">eigen</span>(<span class="fu">cov</span>(abundance.only))</span>
<span id="cb432-2"><a href="principal-components-analysis.html#cb432-2" aria-hidden="true" tabindex="-1"></a>eig.abundance<span class="sc">$</span>vectors  <span class="co"># compare to rotation matrix of PCA</span></span>
<span id="cb432-3"><a href="principal-components-analysis.html#cb432-3" aria-hidden="true" tabindex="-1"></a><span class="do">##             [,1]        [,2]        [,3]        [,4]        [,5]</span></span>
<span id="cb432-4"><a href="principal-components-analysis.html#cb432-4" aria-hidden="true" tabindex="-1"></a><span class="do">## [1,]  0.81064462 -0.07052882  0.53108427  0.18442140 -0.14771336</span></span>
<span id="cb432-5"><a href="principal-components-analysis.html#cb432-5" aria-hidden="true" tabindex="-1"></a><span class="do">## [2,]  0.51264394  0.27799671 -0.47711910 -0.63418946  0.17342177</span></span>
<span id="cb432-6"><a href="principal-components-analysis.html#cb432-6" aria-hidden="true" tabindex="-1"></a><span class="do">## [3,] -0.16235135  0.88665551  0.40897655 -0.01149647  0.14173943</span></span>
<span id="cb432-7"><a href="principal-components-analysis.html#cb432-7" aria-hidden="true" tabindex="-1"></a><span class="do">## [4,]  0.22207108  0.31665237 -0.56250980  0.72941223 -0.04422938</span></span>
<span id="cb432-8"><a href="principal-components-analysis.html#cb432-8" aria-hidden="true" tabindex="-1"></a><span class="do">## [5,]  0.06616623 -0.17696554  0.08141111  0.17781482  0.96231977</span></span>
<span id="cb432-9"><a href="principal-components-analysis.html#cb432-9" aria-hidden="true" tabindex="-1"></a><span class="fu">sqrt</span>(eig.abundance<span class="sc">$</span>values) <span class="co"># compare to sdev of PCA</span></span>
<span id="cb432-10"><a href="principal-components-analysis.html#cb432-10" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 14.865306  8.814912  6.219250  5.034774  3.482308</span></span></code></pre></div>
<p>Notice that the <code>rotation</code> object returned by the <code>prcomp()</code> represents the scaled eigenvectors (scaled to have length 1). The standard deviations of the PCA are the square roots of the eigenvalues of the covariance matrix.</p>
</div>
<div id="calculating-factor-loadings" class="section level3 hasAnchor" number="13.4.3">
<h3><span class="header-section-number">13.4.3</span> Calculating Factor Loadings<a href="principal-components-analysis.html#calculating-factor-loadings" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Let’s calculate the “factor loadings” associated with the PCs:</p>
<div class="sourceCode" id="cb433"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb433-1"><a href="principal-components-analysis.html#cb433-1" aria-hidden="true" tabindex="-1"></a>V <span class="ot">&lt;-</span> abundance.pca<span class="sc">$</span>rotation <span class="co"># eigenvectors</span></span>
<span id="cb433-2"><a href="principal-components-analysis.html#cb433-2" aria-hidden="true" tabindex="-1"></a>L <span class="ot">&lt;-</span> <span class="fu">diag</span>(abundance.pca<span class="sc">$</span>sdev) <span class="co"># diag mtx w/sqrts of eigenvalues on diag.</span></span>
<span id="cb433-3"><a href="principal-components-analysis.html#cb433-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb433-4"><a href="principal-components-analysis.html#cb433-4" aria-hidden="true" tabindex="-1"></a>abundance.loadings <span class="ot">&lt;-</span> V <span class="sc">%*%</span> L</span>
<span id="cb433-5"><a href="principal-components-analysis.html#cb433-5" aria-hidden="true" tabindex="-1"></a>abundance.loadings</span>
<span id="cb433-6"><a href="principal-components-analysis.html#cb433-6" aria-hidden="true" tabindex="-1"></a><span class="do">##         [,1]       [,2]       [,3]        [,4]       [,5]</span></span>
<span id="cb433-7"><a href="principal-components-analysis.html#cb433-7" aria-hidden="true" tabindex="-1"></a><span class="do">## a 12.0504801  0.6217053 -3.3029460  0.92852016 -0.5143835</span></span>
<span id="cb433-8"><a href="principal-components-analysis.html#cb433-8" aria-hidden="true" tabindex="-1"></a><span class="do">## b  7.6206090 -2.4505164  2.9673232 -3.19300085  0.6039081</span></span>
<span id="cb433-9"><a href="principal-components-analysis.html#cb433-9" aria-hidden="true" tabindex="-1"></a><span class="do">## c -2.4134024 -7.8157898 -2.5435276 -0.05788214  0.4935804</span></span>
<span id="cb433-10"><a href="principal-components-analysis.html#cb433-10" aria-hidden="true" tabindex="-1"></a><span class="do">## d  3.3011545 -2.7912626  3.4983893  3.67242602 -0.1540203</span></span>
<span id="cb433-11"><a href="principal-components-analysis.html#cb433-11" aria-hidden="true" tabindex="-1"></a><span class="do">## e  0.9835813  1.5599356 -0.5063161  0.89525751  3.3510942</span></span></code></pre></div>
<p>The magnitude of the factor loadings is what you want to focus on. For example, species <code>a</code> and <code>b</code> contribute most to the first PC, while species <code>c</code> has the largest influence on PC2.</p>
<p>You can think of the factor loadings, as defined above, as the components (i.e lengths of the projected vectors) of the original variables with respect to the PC basis vectors. Since vector length is proportional to the standard deviation of the variables they represent, you can think of the loadings as giving the standard deviation of the original variables with respect the PC axes. This implies that the loadings squared sum to the total variance in the original data, as illustrated below.</p>
<div class="sourceCode" id="cb434"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb434-1"><a href="principal-components-analysis.html#cb434-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(abundance.loadings<span class="sc">**</span><span class="dv">2</span>)</span>
<span id="cb434-2"><a href="principal-components-analysis.html#cb434-2" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 374.8345</span></span>
<span id="cb434-3"><a href="principal-components-analysis.html#cb434-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb434-4"><a href="principal-components-analysis.html#cb434-4" aria-hidden="true" tabindex="-1"></a>abundance.only <span class="sc">%&gt;%</span> purrr<span class="sc">::</span><span class="fu">map_dbl</span>(var) <span class="sc">%&gt;%</span> sum</span>
<span id="cb434-5"><a href="principal-components-analysis.html#cb434-5" aria-hidden="true" tabindex="-1"></a><span class="do">## [1] 374.8345</span></span></code></pre></div>
</div>
</div>
<div id="drawing-figures-to-represent-pca" class="section level2 hasAnchor" number="13.5">
<h2><span class="header-section-number">13.5</span> Drawing Figures to Represent PCA<a href="principal-components-analysis.html#drawing-figures-to-represent-pca" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="pc-score-plots" class="section level3 hasAnchor" number="13.5.1">
<h3><span class="header-section-number">13.5.1</span> PC Score Plots<a href="principal-components-analysis.html#pc-score-plots" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The simplest PCA figure is to depict the PC scores, i.e. the projection of the observations into the space defined by the PC axes. Let’s make a figure with three subplots, depicting PC1 vs PC2, PC1 vs PC3, and PC2 vs. PC3.</p>
<div class="sourceCode" id="cb435"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb435-1"><a href="principal-components-analysis.html#cb435-1" aria-hidden="true" tabindex="-1"></a>pca.scores.df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(abundance.pca<span class="sc">$</span>x)</span>
<span id="cb435-2"><a href="principal-components-analysis.html#cb435-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb435-3"><a href="principal-components-analysis.html#cb435-3" aria-hidden="true" tabindex="-1"></a>coord.system <span class="ot">&lt;-</span> <span class="fu">coord_fixed</span>(<span class="at">ratio=</span><span class="dv">1</span>, <span class="at">xlim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">30</span>,<span class="dv">30</span>),<span class="at">ylim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">30</span>,<span class="dv">30</span>))</span>
<span id="cb435-4"><a href="principal-components-analysis.html#cb435-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb435-5"><a href="principal-components-analysis.html#cb435-5" aria-hidden="true" tabindex="-1"></a>pc1v2 <span class="ot">&lt;-</span></span>
<span id="cb435-6"><a href="principal-components-analysis.html#cb435-6" aria-hidden="true" tabindex="-1"></a>  pca.scores.df <span class="sc">%&gt;%</span></span>
<span id="cb435-7"><a href="principal-components-analysis.html#cb435-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y=</span> PC2)) <span class="sc">+</span></span>
<span id="cb435-8"><a href="principal-components-analysis.html#cb435-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb435-9"><a href="principal-components-analysis.html#cb435-9" aria-hidden="true" tabindex="-1"></a>  coord.system</span>
<span id="cb435-10"><a href="principal-components-analysis.html#cb435-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb435-11"><a href="principal-components-analysis.html#cb435-11" aria-hidden="true" tabindex="-1"></a>pc1v3 <span class="ot">&lt;-</span></span>
<span id="cb435-12"><a href="principal-components-analysis.html#cb435-12" aria-hidden="true" tabindex="-1"></a>  pca.scores.df <span class="sc">%&gt;%</span></span>
<span id="cb435-13"><a href="principal-components-analysis.html#cb435-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y=</span> PC3)) <span class="sc">+</span></span>
<span id="cb435-14"><a href="principal-components-analysis.html#cb435-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb435-15"><a href="principal-components-analysis.html#cb435-15" aria-hidden="true" tabindex="-1"></a>  coord.system</span>
<span id="cb435-16"><a href="principal-components-analysis.html#cb435-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb435-17"><a href="principal-components-analysis.html#cb435-17" aria-hidden="true" tabindex="-1"></a>pc2v3 <span class="ot">&lt;-</span></span>
<span id="cb435-18"><a href="principal-components-analysis.html#cb435-18" aria-hidden="true" tabindex="-1"></a>  pca.scores.df <span class="sc">%&gt;%</span></span>
<span id="cb435-19"><a href="principal-components-analysis.html#cb435-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> PC2, <span class="at">y=</span> PC3)) <span class="sc">+</span></span>
<span id="cb435-20"><a href="principal-components-analysis.html#cb435-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb435-21"><a href="principal-components-analysis.html#cb435-21" aria-hidden="true" tabindex="-1"></a>  coord.system</span>
<span id="cb435-22"><a href="principal-components-analysis.html#cb435-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb435-23"><a href="principal-components-analysis.html#cb435-23" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(pc1v2, pc1v3, pc2v3, <span class="at">align=</span><span class="st">&quot;hv&quot;</span>)</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-432-1.png" width="576" /></p>
<p>When plotting PC scores, it is <strong>very important</strong> to keep the aspect ratio with a value of 1, so that the distance between points in the plot is an accurate representation of the distance in the PC space. Note too that I used the <code>xlim</code> and <code>ylim</code> arguments to keep the axis limits the same in all plots; comparable scaling of axes is important when comparing plots. Also note the use of the <code>align="hv"</code> argument to <code>plot_grid()</code> to keep my plots the same size when combining them into a single figure.</p>
</div>
<div id="simultaneous-depiction-of-observations-and-variables-in-the-pc-space" class="section level3 hasAnchor" number="13.5.2">
<h3><span class="header-section-number">13.5.2</span> Simultaneous Depiction of Observations and Variables in the PC Space<a href="principal-components-analysis.html#simultaneous-depiction-of-observations-and-variables-in-the-pc-space" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Let’s return to our simple PC score plot. As we discussed above, the loadings are components of the original variables in the space of the PCs. This implies we can depict those loadings in the same PC basis that we use to depict the scores.</p>
<p>First let’s create a data frame with the loadings from the first two PCs as well as a column representation the variable names:</p>
<div class="sourceCode" id="cb436"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb436-1"><a href="principal-components-analysis.html#cb436-1" aria-hidden="true" tabindex="-1"></a>loadings<span class="fl">.1</span>and2 <span class="ot">&lt;-</span></span>
<span id="cb436-2"><a href="principal-components-analysis.html#cb436-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>(abundance.loadings[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]) <span class="sc">%&gt;%</span></span>
<span id="cb436-3"><a href="principal-components-analysis.html#cb436-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">PC1.loading =</span> X1, <span class="at">PC2.loading =</span> X2) <span class="sc">%&gt;%</span></span>
<span id="cb436-4"><a href="principal-components-analysis.html#cb436-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">variable =</span> <span class="fu">row.names</span>(abundance.loadings))</span>
<span id="cb436-5"><a href="principal-components-analysis.html#cb436-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb436-6"><a href="principal-components-analysis.html#cb436-6" aria-hidden="true" tabindex="-1"></a>loadings<span class="fl">.1</span>and2</span>
<span id="cb436-7"><a href="principal-components-analysis.html#cb436-7" aria-hidden="true" tabindex="-1"></a><span class="do">## # A tibble: 5 × 3</span></span>
<span id="cb436-8"><a href="principal-components-analysis.html#cb436-8" aria-hidden="true" tabindex="-1"></a><span class="do">##   PC1.loading PC2.loading variable</span></span>
<span id="cb436-9"><a href="principal-components-analysis.html#cb436-9" aria-hidden="true" tabindex="-1"></a><span class="do">##         &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;   </span></span>
<span id="cb436-10"><a href="principal-components-analysis.html#cb436-10" aria-hidden="true" tabindex="-1"></a><span class="do">## 1      12.1         0.622 a       </span></span>
<span id="cb436-11"><a href="principal-components-analysis.html#cb436-11" aria-hidden="true" tabindex="-1"></a><span class="do">## 2       7.62       -2.45  b       </span></span>
<span id="cb436-12"><a href="principal-components-analysis.html#cb436-12" aria-hidden="true" tabindex="-1"></a><span class="do">## 3      -2.41       -7.82  c       </span></span>
<span id="cb436-13"><a href="principal-components-analysis.html#cb436-13" aria-hidden="true" tabindex="-1"></a><span class="do">## 4       3.30       -2.79  d       </span></span>
<span id="cb436-14"><a href="principal-components-analysis.html#cb436-14" aria-hidden="true" tabindex="-1"></a><span class="do">## 5       0.984       1.56  e</span></span></code></pre></div>
<p>With this data frame in hand, we can now draw a set of vector to represent the original variables projected into the subspace defined by PCs 1 and 2:</p>
<div class="sourceCode" id="cb437"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb437-1"><a href="principal-components-analysis.html#cb437-1" aria-hidden="true" tabindex="-1"></a>pc1v2.biplot <span class="ot">&lt;-</span></span>
<span id="cb437-2"><a href="principal-components-analysis.html#cb437-2" aria-hidden="true" tabindex="-1"></a>  pc1v2 <span class="sc">+</span></span>
<span id="cb437-3"><a href="principal-components-analysis.html#cb437-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="at">data=</span>loadings<span class="fl">.1</span>and2,</span>
<span id="cb437-4"><a href="principal-components-analysis.html#cb437-4" aria-hidden="true" tabindex="-1"></a>               <span class="fu">aes</span>(<span class="at">x =</span> <span class="dv">0</span>, <span class="at">y =</span> <span class="dv">0</span>, <span class="at">xend =</span> PC1.loading, <span class="at">yend =</span> PC2.loading),</span>
<span id="cb437-5"><a href="principal-components-analysis.html#cb437-5" aria-hidden="true" tabindex="-1"></a>               <span class="at">color=</span><span class="st">&#39;red&#39;</span>, </span>
<span id="cb437-6"><a href="principal-components-analysis.html#cb437-6" aria-hidden="true" tabindex="-1"></a>               <span class="at">arrow =</span> <span class="fu">arrow</span>(<span class="at">angle=</span><span class="dv">15</span>, <span class="at">length=</span><span class="fu">unit</span>(<span class="fl">0.1</span>,<span class="st">&quot;inches&quot;</span>))) <span class="sc">+</span></span>
<span id="cb437-7"><a href="principal-components-analysis.html#cb437-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">data=</span>loadings<span class="fl">.1</span>and2,</span>
<span id="cb437-8"><a href="principal-components-analysis.html#cb437-8" aria-hidden="true" tabindex="-1"></a>            <span class="fu">aes</span>(<span class="at">x =</span> PC1.loading, <span class="at">y =</span> PC2.loading, <span class="at">label=</span>variable),</span>
<span id="cb437-9"><a href="principal-components-analysis.html#cb437-9" aria-hidden="true" tabindex="-1"></a>            <span class="at">color=</span><span class="st">&#39;red&#39;</span>, <span class="at">nudge_x =</span> <span class="dv">1</span>, <span class="at">nudge_y =</span> <span class="dv">1</span>)</span>
<span id="cb437-10"><a href="principal-components-analysis.html#cb437-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb437-11"><a href="principal-components-analysis.html#cb437-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb437-12"><a href="principal-components-analysis.html#cb437-12" aria-hidden="true" tabindex="-1"></a>pc1v2.biplot</span></code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-434-1.png" width="768" /></p>
<p>The figure above is called a “biplot”“, as it simultaneously depicts both the observations and variables in the same space. From this biplot we can immediately see that variable <code>a</code> is highly correlated with PC1, but only weakly associated with PC2. Conversely, variable <code>c</code> is strongly correlated with PC2 but only weakly so with PC1. We can also approximate the correlations among the variables themselves – for example <code>b</code> and <code>d</code> are fairly strongly correlated, but weakly correlated with <code>c</code>. Keep in mind however that with respect to the relationships among the variables, this visualization is a 2D projection of a 5D space so the geometry is approximate.</p>
<p>The biplot is a generally useful tool for multivariate analyses and there are a number of different ways to define biplots. We’ll study biplots more formally in a few weeks after we’ve covered singular value decomposition.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="linear-regression-models-ii.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="canonical-variates-analysis.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
},
"toolbar": {
"position": "fixed"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
