<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Biology 723: Statistical Computing for Biologists</title>
  <meta name="description" content="Hands-on materials for Bio 723 course at Duke University.">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Biology 723: Statistical Computing for Biologists" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Biology 723: Statistical Computing for Biologists" />
  
  <meta name="twitter:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

<meta name="author" content="Paul M. Magwene">


<meta name="date" content="2018-04-04">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="simulating-confidence-intervals.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#how-to-use-these-lecture-notes"><i class="fa fa-check"></i><b>1.1</b> How to use these lecture notes</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html"><i class="fa fa-check"></i><b>2</b> Getting Started with R</a><ul>
<li class="chapter" data-level="2.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-r"><i class="fa fa-check"></i><b>2.1</b> What is R?</a></li>
<li class="chapter" data-level="2.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-rstudio"><i class="fa fa-check"></i><b>2.2</b> What is RStudio?</a></li>
<li class="chapter" data-level="2.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#entering-commands-in-the-console"><i class="fa fa-check"></i><b>2.3</b> Entering commands in the console</a></li>
<li class="chapter" data-level="2.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#comments"><i class="fa fa-check"></i><b>2.4</b> Comments</a></li>
<li class="chapter" data-level="2.5" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#using-r-as-a-calculator"><i class="fa fa-check"></i><b>2.5</b> Using R as a Calculator</a><ul>
<li class="chapter" data-level="2.5.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#common-mathematical-functions"><i class="fa fa-check"></i><b>2.5.1</b> Common mathematical functions</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#numeric-data-types-in-r"><i class="fa fa-check"></i><b>2.6</b> Numeric data types in R</a></li>
<li class="chapter" data-level="2.7" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#variable-assignment-in-r"><i class="fa fa-check"></i><b>2.7</b> Variable assignment in R</a><ul>
<li class="chapter" data-level="2.7.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#valid-variable-names"><i class="fa fa-check"></i><b>2.7.1</b> Valid variable names</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#working-with-vectors-in-r"><i class="fa fa-check"></i><b>2.8</b> Working with Vectors in R</a><ul>
<li class="chapter" data-level="2.8.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#vector-arithmetic"><i class="fa fa-check"></i><b>2.8.1</b> Vector Arithmetic</a></li>
<li class="chapter" data-level="2.8.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#vector-recycling"><i class="fa fa-check"></i><b>2.8.2</b> Vector recycling</a></li>
<li class="chapter" data-level="2.8.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#simple-statistical-functions-for-numeric-vectors"><i class="fa fa-check"></i><b>2.8.3</b> Simple statistical functions for numeric vectors</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#the-r-help-system"><i class="fa fa-check"></i><b>2.9</b> The R Help System</a><ul>
<li class="chapter" data-level="2.9.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#getting-help-from-the-console"><i class="fa fa-check"></i><b>2.9.1</b> Getting help from the console</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#packages"><i class="fa fa-check"></i><b>2.10</b> Packages</a><ul>
<li class="chapter" data-level="2.10.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-console"><i class="fa fa-check"></i><b>2.10.1</b> Installing packages from the console</a></li>
<li class="chapter" data-level="2.10.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#install-the-tidyverse-package"><i class="fa fa-check"></i><b>2.10.2</b> Install the tidyverse package</a></li>
<li class="chapter" data-level="2.10.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-rstudio-dialog"><i class="fa fa-check"></i><b>2.10.3</b> Installing packages from the RStudio dialog</a></li>
<li class="chapter" data-level="2.10.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#install-the-stringr-package"><i class="fa fa-check"></i><b>2.10.4</b> Install the stringr package</a></li>
<li class="chapter" data-level="2.10.5" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#loading-packages-with-the-library-function"><i class="fa fa-check"></i><b>2.10.5</b> Loading packages with the <code>library()</code> function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html"><i class="fa fa-check"></i><b>3</b> R Basics: Data types and data structures</a><ul>
<li class="chapter" data-level="3.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-boolean-values"><i class="fa fa-check"></i><b>3.1</b> Logical (Boolean) values</a><ul>
<li class="chapter" data-level="3.1.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#comparison-operators-applied-to-numeric-vectors"><i class="fa fa-check"></i><b>3.1.1</b> Comparison operators applied to numeric vectors</a></li>
<li class="chapter" data-level="3.1.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-vectors"><i class="fa fa-check"></i><b>3.1.2</b> Logical vectors</a></li>
<li class="chapter" data-level="3.1.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-operators"><i class="fa fa-check"></i><b>3.1.3</b> Logical operators</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#character-strings"><i class="fa fa-check"></i><b>3.2</b> Character strings</a><ul>
<li class="chapter" data-level="3.2.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#the-stringr-package"><i class="fa fa-check"></i><b>3.2.1</b> The <code>stringr</code> package</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#vectors"><i class="fa fa-check"></i><b>3.3</b> Vectors</a><ul>
<li class="chapter" data-level="3.3.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#indexing-vectors"><i class="fa fa-check"></i><b>3.3.1</b> Indexing Vectors</a></li>
<li class="chapter" data-level="3.3.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#combining-indexing-and-comparison-of-vectors"><i class="fa fa-check"></i><b>3.3.2</b> Combining Indexing and Comparison of Vectors</a></li>
<li class="chapter" data-level="3.3.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#vector-manipulation"><i class="fa fa-check"></i><b>3.3.3</b> Vector manipulation</a></li>
<li class="chapter" data-level="3.3.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#vectors-from-regular-sequences"><i class="fa fa-check"></i><b>3.3.4</b> Vectors from regular sequences</a></li>
<li class="chapter" data-level="3.3.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#additional-functions-for-working-with-vectors"><i class="fa fa-check"></i><b>3.3.5</b> Additional functions for working with vectors</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#lists"><i class="fa fa-check"></i><b>3.4</b> Lists</a><ul>
<li class="chapter" data-level="3.4.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#length-and-type-of-lists"><i class="fa fa-check"></i><b>3.4.1</b> Length and type of lists</a></li>
<li class="chapter" data-level="3.4.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#indexing-lists"><i class="fa fa-check"></i><b>3.4.2</b> Indexing lists</a></li>
<li class="chapter" data-level="3.4.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#naming-list-elements"><i class="fa fa-check"></i><b>3.4.3</b> Naming list elements</a></li>
<li class="chapter" data-level="3.4.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#the-operator"><i class="fa fa-check"></i><b>3.4.4</b> The <code>$</code> operator</a></li>
<li class="chapter" data-level="3.4.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#changing-and-adding-lists-items"><i class="fa fa-check"></i><b>3.4.5</b> Changing and adding lists items</a></li>
<li class="chapter" data-level="3.4.6" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#combining-lists"><i class="fa fa-check"></i><b>3.4.6</b> Combining lists</a></li>
<li class="chapter" data-level="3.4.7" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#converting-lists-to-vectors"><i class="fa fa-check"></i><b>3.4.7</b> Converting lists to vectors</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#data-frames"><i class="fa fa-check"></i><b>3.5</b> Data frames</a><ul>
<li class="chapter" data-level="3.5.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#creating-a-data-frame"><i class="fa fa-check"></i><b>3.5.1</b> Creating a data frame</a></li>
<li class="chapter" data-level="3.5.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#type-and-class-for-data-frames"><i class="fa fa-check"></i><b>3.5.2</b> Type and class for data frames</a></li>
<li class="chapter" data-level="3.5.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#length-and-dimension-for-data-frames"><i class="fa fa-check"></i><b>3.5.3</b> Length and dimension for data frames</a></li>
<li class="chapter" data-level="3.5.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#indexing-and-accessing-data-frames"><i class="fa fa-check"></i><b>3.5.4</b> Indexing and accessing data frames</a></li>
<li class="chapter" data-level="3.5.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-indexing-of-data-frames"><i class="fa fa-check"></i><b>3.5.5</b> Logical indexing of data frames</a></li>
<li class="chapter" data-level="3.5.6" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#adding-columns-to-a-data-frame"><i class="fa fa-check"></i><b>3.5.6</b> Adding columns to a data frame</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html"><i class="fa fa-check"></i><b>4</b> R Markdown and R Notebooks</a><ul>
<li class="chapter" data-level="4.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#r-notebooks"><i class="fa fa-check"></i><b>4.1</b> R Notebooks</a></li>
<li class="chapter" data-level="4.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#creating-an-r-notebook"><i class="fa fa-check"></i><b>4.2</b> Creating an R Notebook</a></li>
<li class="chapter" data-level="4.3" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#the-default-r-notebook-template"><i class="fa fa-check"></i><b>4.3</b> The default R Notebook template</a></li>
<li class="chapter" data-level="4.4" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-and-non-code-blocks"><i class="fa fa-check"></i><b>4.4</b> Code and Non-code blocks</a><ul>
<li class="chapter" data-level="4.4.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#non-code-blocks"><i class="fa fa-check"></i><b>4.4.1</b> Non-code blocks</a></li>
<li class="chapter" data-level="4.4.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-blocks"><i class="fa fa-check"></i><b>4.4.2</b> Code blocks</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-a-code-chunk"><i class="fa fa-check"></i><b>4.5</b> Running a code chunk</a></li>
<li class="chapter" data-level="4.6" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-all-code-chunks-above"><i class="fa fa-check"></i><b>4.6</b> Running all code chunks above</a></li>
<li class="chapter" data-level="4.7" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#knitting-an-r-markdown-to-html"><i class="fa fa-check"></i><b>4.7</b> “Knitting an” R Markdown to HTML</a></li>
<li class="chapter" data-level="4.8" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#sharing-your-reproducible-r-notebook"><i class="fa fa-check"></i><b>4.8</b> Sharing your reproducible R Notebook</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html"><i class="fa fa-check"></i><b>5</b> Introduction to ggplot2</a><ul>
<li class="chapter" data-level="5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#loading-ggplot2"><i class="fa fa-check"></i><b>5.1</b> Loading ggplot2</a></li>
<li class="chapter" data-level="5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#example-data-set-andersons-iris-data"><i class="fa fa-check"></i><b>5.2</b> Example data set: Anderson’s Iris Data</a></li>
<li class="chapter" data-level="5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#template-for-single-layer-plots-in-ggplot2"><i class="fa fa-check"></i><b>5.3</b> Template for single layer plots in ggplot2</a></li>
<li class="chapter" data-level="5.4" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#an-aside-about-function-arguments"><i class="fa fa-check"></i><b>5.4</b> An aside about function arguments</a></li>
<li class="chapter" data-level="5.5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#strip-plots"><i class="fa fa-check"></i><b>5.5</b> Strip plots</a><ul>
<li class="chapter" data-level="5.5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#jittering-data"><i class="fa fa-check"></i><b>5.5.1</b> Jittering data</a></li>
<li class="chapter" data-level="5.5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-categorical-information"><i class="fa fa-check"></i><b>5.5.2</b> Adding categorical information</a></li>
<li class="chapter" data-level="5.5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#rotating-plot-coordinates"><i class="fa fa-check"></i><b>5.5.3</b> Rotating plot coordinates</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#histograms"><i class="fa fa-check"></i><b>5.6</b> Histograms</a><ul>
<li class="chapter" data-level="5.6.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#variations-on-histograms-when-considering-categorical-data"><i class="fa fa-check"></i><b>5.6.1</b> Variations on histograms when considering categorical data</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#faceting-to-depict-categorical-information"><i class="fa fa-check"></i><b>5.7</b> Faceting to depict categorical information</a></li>
<li class="chapter" data-level="5.8" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots"><i class="fa fa-check"></i><b>5.8</b> Density plots</a></li>
<li class="chapter" data-level="5.9" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-or-beanplot"><i class="fa fa-check"></i><b>5.9</b> Violin or Beanplot</a></li>
<li class="chapter" data-level="5.10" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplots"><i class="fa fa-check"></i><b>5.10</b> Boxplots</a></li>
<li class="chapter" data-level="5.11" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#building-complex-visualizations-with-layers"><i class="fa fa-check"></i><b>5.11</b> Building complex visualizations with layers</a></li>
<li class="chapter" data-level="5.12" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#useful-combination-plots"><i class="fa fa-check"></i><b>5.12</b> Useful combination plots</a><ul>
<li class="chapter" data-level="5.12.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplot-plus-strip-plot"><i class="fa fa-check"></i><b>5.12.1</b> Boxplot plus strip plot</a></li>
<li class="chapter" data-level="5.12.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#setting-shared-aesthetics"><i class="fa fa-check"></i><b>5.12.2</b> Setting shared aesthetics</a></li>
</ul></li>
<li class="chapter" data-level="5.13" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot-layers-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>5.13</b> ggplot layers can be assigned to variables</a><ul>
<li class="chapter" data-level="5.13.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-plot-plus-strip-plot"><i class="fa fa-check"></i><b>5.13.1</b> Violin plot plus strip plot</a></li>
</ul></li>
<li class="chapter" data-level="5.14" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-titles-and-tweaking-axis-labels"><i class="fa fa-check"></i><b>5.14</b> Adding titles and tweaking axis labels</a></li>
<li class="chapter" data-level="5.15" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot2-themes"><i class="fa fa-check"></i><b>5.15</b> ggplot2 themes</a><ul>
<li class="chapter" data-level="5.15.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#further-customization-with-ggplot2theme"><i class="fa fa-check"></i><b>5.15.1</b> Further customization with <code>ggplot2::theme</code></a></li>
</ul></li>
<li class="chapter" data-level="5.16" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#other-aspects-of-ggplots-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>5.16</b> Other aspects of ggplots can be assigned to variables</a></li>
<li class="chapter" data-level="5.17" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-plots"><i class="fa fa-check"></i><b>5.17</b> Bivariate plots</a><ul>
<li class="chapter" data-level="5.17.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#scatter-plots"><i class="fa fa-check"></i><b>5.17.1</b> Scatter plots</a></li>
</ul></li>
<li class="chapter" data-level="5.18" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-density-plots"><i class="fa fa-check"></i><b>5.18</b> Bivariate density plots</a></li>
<li class="chapter" data-level="5.19" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#combining-scatter-plots-and-density-plots-with-categorical-information"><i class="fa fa-check"></i><b>5.19</b> Combining Scatter Plots and Density Plots with Categorical Information</a></li>
<li class="chapter" data-level="5.20" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots-with-fill"><i class="fa fa-check"></i><b>5.20</b> Density plots with fill</a></li>
<li class="chapter" data-level="5.21" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#d-bin-and-hex-plots"><i class="fa fa-check"></i><b>5.21</b> 2D bin and hex plots</a></li>
<li class="chapter" data-level="5.22" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#the-cowplot-package"><i class="fa fa-check"></i><b>5.22</b> The <code>cowplot</code> package</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html"><i class="fa fa-check"></i><b>6</b> Introduction to the <code>dplyr</code> package</a><ul>
<li class="chapter" data-level="6.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#libraries"><i class="fa fa-check"></i><b>6.1</b> Libraries</a></li>
<li class="chapter" data-level="6.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#reading-data-with-the-readr-package"><i class="fa fa-check"></i><b>6.2</b> Reading data with the <code>readr</code> package</a><ul>
<li class="chapter" data-level="6.2.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#example-data-nc-births"><i class="fa fa-check"></i><b>6.2.1</b> Example data: NC Births</a></li>
<li class="chapter" data-level="6.2.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#reading-excel-files"><i class="fa fa-check"></i><b>6.2.2</b> Reading Excel files</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#a-note-on-tibbles"><i class="fa fa-check"></i><b>6.3</b> A note on “tibbles”</a></li>
<li class="chapter" data-level="6.4" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#data-filtering-and-transformation-with-dplyr"><i class="fa fa-check"></i><b>6.4</b> Data filtering and transformation with <code>dplyr</code></a></li>
<li class="chapter" data-level="6.5" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#dplyrs-verbs"><i class="fa fa-check"></i><b>6.5</b> dplyr’s “verbs”</a><ul>
<li class="chapter" data-level="6.5.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#select"><i class="fa fa-check"></i><b>6.5.1</b> <code>select</code></a></li>
<li class="chapter" data-level="6.5.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#filter"><i class="fa fa-check"></i><b>6.5.2</b> <code>filter</code></a></li>
<li class="chapter" data-level="6.5.3" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#mutate"><i class="fa fa-check"></i><b>6.5.3</b> <code>mutate</code></a></li>
<li class="chapter" data-level="6.5.4" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#arrange"><i class="fa fa-check"></i><b>6.5.4</b> <code>arrange</code></a></li>
<li class="chapter" data-level="6.5.5" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#summarize"><i class="fa fa-check"></i><b>6.5.5</b> <code>summarize</code></a></li>
<li class="chapter" data-level="6.5.6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#group_by"><i class="fa fa-check"></i><b>6.5.6</b> <code>group_by</code></a></li>
<li class="chapter" data-level="6.5.7" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#combining-grouping-and-summarizing"><i class="fa fa-check"></i><b>6.5.7</b> Combining grouping and summarizing</a></li>
<li class="chapter" data-level="6.5.8" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#scoped-variants-of-mutate-and-summarize"><i class="fa fa-check"></i><b>6.5.8</b> Scoped variants of <code>mutate</code> and <code>summarize</code></a></li>
<li class="chapter" data-level="6.5.9" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#combining-summarize-with-grouping-aesthetics-in-ggplot2"><i class="fa fa-check"></i><b>6.5.9</b> Combining summarize with grouping aesthetics in <code>ggplot2</code></a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#pipes"><i class="fa fa-check"></i><b>6.6</b> Pipes</a><ul>
<li class="chapter" data-level="6.6.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#install-and-load-magrittr"><i class="fa fa-check"></i><b>6.6.1</b> Install and load <code>magrittr</code></a></li>
<li class="chapter" data-level="6.6.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-basic-pipe-operator"><i class="fa fa-check"></i><b>6.6.2</b> The basic pipe operator</a></li>
<li class="chapter" data-level="6.6.3" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#an-example-without-pipes"><i class="fa fa-check"></i><b>6.6.3</b> An example without pipes</a></li>
<li class="chapter" data-level="6.6.4" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-same-example-using-pipes"><i class="fa fa-check"></i><b>6.6.4</b> The same example using pipes</a></li>
<li class="chapter" data-level="6.6.5" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#assigning-the-output-of-a-statement-involving-pipes-to-a-variable"><i class="fa fa-check"></i><b>6.6.5</b> Assigning the output of a statement involving pipes to a variable</a></li>
<li class="chapter" data-level="6.6.6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#compound-assignment-pipe-operator"><i class="fa fa-check"></i><b>6.6.6</b> Compound assignment pipe operator</a></li>
<li class="chapter" data-level="6.6.7" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-dot-operator-with-pipes"><i class="fa fa-check"></i><b>6.6.7</b> The dot operator with pipes</a></li>
<li class="chapter" data-level="6.6.8" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-exposition-pipe-operator"><i class="fa fa-check"></i><b>6.6.8</b> The exposition pipe operator</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>7</b> Data wrangling</a><ul>
<li class="chapter" data-level="7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#libraries-1"><i class="fa fa-check"></i><b>7.1</b> Libraries</a></li>
<li class="chapter" data-level="7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#data"><i class="fa fa-check"></i><b>7.2</b> Data</a></li>
<li class="chapter" data-level="7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#renaming-data-frame-columms"><i class="fa fa-check"></i><b>7.3</b> Renaming data frame columms</a></li>
<li class="chapter" data-level="7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-unneeded-columns"><i class="fa fa-check"></i><b>7.4</b> Dropping unneeded columns</a><ul>
<li class="chapter" data-level="7.4.1" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-all-empty-columns"><i class="fa fa-check"></i><b>7.4.1</b> Finding all empty columns</a></li>
<li class="chapter" data-level="7.4.2" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-columns-by-matching-names"><i class="fa fa-check"></i><b>7.4.2</b> Dropping columns by matching names</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="data-wrangling.html"><a href="data-wrangling.html#reshaping-data-with-tidyr"><i class="fa fa-check"></i><b>7.5</b> Reshaping data with tidyr</a><ul>
<li class="chapter" data-level="7.5.1" data-path="data-wrangling.html"><a href="data-wrangling.html#wide-to-long-conversions-using-tidyrgather"><i class="fa fa-check"></i><b>7.5.1</b> Wide to long conversions using <code>tidyr::gather</code></a></li>
<li class="chapter" data-level="7.5.2" data-path="data-wrangling.html"><a href="data-wrangling.html#extracting-information-from-combined-variables-using-tidyrextract"><i class="fa fa-check"></i><b>7.5.2</b> Extracting information from combined variables using <code>tidyr::extract</code></a></li>
<li class="chapter" data-level="7.5.3" data-path="data-wrangling.html"><a href="data-wrangling.html#subsetting-rows"><i class="fa fa-check"></i><b>7.5.3</b> Subsetting rows</a></li>
<li class="chapter" data-level="7.5.4" data-path="data-wrangling.html"><a href="data-wrangling.html#splitting-columns"><i class="fa fa-check"></i><b>7.5.4</b> Splitting columns</a></li>
<li class="chapter" data-level="7.5.5" data-path="data-wrangling.html"><a href="data-wrangling.html#combining-data-frame-rows"><i class="fa fa-check"></i><b>7.5.5</b> Combining data frame rows</a></li>
<li class="chapter" data-level="7.5.6" data-path="data-wrangling.html"><a href="data-wrangling.html#sorting-data-frame-rows"><i class="fa fa-check"></i><b>7.5.6</b> Sorting data frame rows</a></li>
<li class="chapter" data-level="7.5.7" data-path="data-wrangling.html"><a href="data-wrangling.html#collapsing-columns-using-tidyrunite"><i class="fa fa-check"></i><b>7.5.7</b> Collapsing columns using <code>tidyr::unite</code></a></li>
<li class="chapter" data-level="7.5.8" data-path="data-wrangling.html"><a href="data-wrangling.html#splitting-columns-using-tidyrseparate"><i class="fa fa-check"></i><b>7.5.8</b> Splitting columns using <code>tidyr::separate</code></a></li>
<li class="chapter" data-level="7.5.9" data-path="data-wrangling.html"><a href="data-wrangling.html#long-to-wide-conversion-using-tidyrspread"><i class="fa fa-check"></i><b>7.5.9</b> Long-to-wide conversion using <code>tidyr::spread</code></a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="data-wrangling.html"><a href="data-wrangling.html#using-your-tidy-data"><i class="fa fa-check"></i><b>7.6</b> Using your tidy data</a><ul>
<li class="chapter" data-level="7.6.1" data-path="data-wrangling.html"><a href="data-wrangling.html#visualizing-gene-expression-time-series"><i class="fa fa-check"></i><b>7.6.1</b> Visualizing gene expression time series</a></li>
<li class="chapter" data-level="7.6.2" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-the-most-variable-genes"><i class="fa fa-check"></i><b>7.6.2</b> Finding the most variable genes</a></li>
<li class="chapter" data-level="7.6.3" data-path="data-wrangling.html"><a href="data-wrangling.html#exploring-bivariate-relationships-using-wide-data"><i class="fa fa-check"></i><b>7.6.3</b> Exploring bivariate relationships using “wide” data</a></li>
<li class="chapter" data-level="7.6.4" data-path="data-wrangling.html"><a href="data-wrangling.html#calculating-correlations-between-genes"><i class="fa fa-check"></i><b>7.6.4</b> Calculating correlations between genes</a></li>
<li class="chapter" data-level="7.6.5" data-path="data-wrangling.html"><a href="data-wrangling.html#adding-new-columns-and-combining-filtered-data-frames"><i class="fa fa-check"></i><b>7.6.5</b> Adding new columns and combining filtered data frames</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="data-wrangling.html"><a href="data-wrangling.html#heat-maps"><i class="fa fa-check"></i><b>7.7</b> Heat maps</a><ul>
<li class="chapter" data-level="7.7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#rcolorbrewer"><i class="fa fa-check"></i><b>7.7.1</b> RColorBrewer</a></li>
<li class="chapter" data-level="7.7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#creating-a-heat-map-using-geom_tile"><i class="fa fa-check"></i><b>7.7.2</b> Creating a heat map using <code>geom_tile</code></a></li>
<li class="chapter" data-level="7.7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#a-fancy-figure"><i class="fa fa-check"></i><b>7.7.3</b> A “fancy” figure</a></li>
<li class="chapter" data-level="7.7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#sorting-and-visualizing-genes-by-time-of-maximum-expression"><i class="fa fa-check"></i><b>7.7.4</b> Sorting and visualizing genes by time of maximum expression</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html"><i class="fa fa-check"></i><b>8</b> Vectors and Functions</a><ul>
<li class="chapter" data-level="8.1" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#libraries-2"><i class="fa fa-check"></i><b>8.1</b> Libraries</a></li>
<li class="chapter" data-level="8.2" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#vector-mathematics-in-r"><i class="fa fa-check"></i><b>8.2</b> Vector Mathematics in R</a></li>
<li class="chapter" data-level="8.3" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#writing-your-own-functions-in-r"><i class="fa fa-check"></i><b>8.3</b> Writing Your Own Functions in R</a></li>
<li class="chapter" data-level="8.4" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#function-arguments"><i class="fa fa-check"></i><b>8.4</b> Function arguments</a></li>
<li class="chapter" data-level="8.5" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#writing-functions-with-optional-arguments"><i class="fa fa-check"></i><b>8.5</b> Writing functions with optional arguments</a></li>
<li class="chapter" data-level="8.6" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#putting-r-functions-in-scripts"><i class="fa fa-check"></i><b>8.6</b> Putting R functions in Scripts</a></li>
<li class="chapter" data-level="8.7" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#simple-statistics-in-vector-form"><i class="fa fa-check"></i><b>8.7</b> Simple statistics in vector form</a><ul>
<li class="chapter" data-level="8.7.1" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#mean"><i class="fa fa-check"></i><b>8.7.1</b> Mean</a></li>
<li class="chapter" data-level="8.7.2" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#variance-and-standard-deviation"><i class="fa fa-check"></i><b>8.7.2</b> Variance and standard deviation</a></li>
<li class="chapter" data-level="8.7.3" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#covariance-and-correlation"><i class="fa fa-check"></i><b>8.7.3</b> Covariance and correlation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html"><i class="fa fa-check"></i><b>9</b> Bivariate Linear Regression</a><ul>
<li class="chapter" data-level="9.1" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#regression-terminology"><i class="fa fa-check"></i><b>9.1</b> Regression terminology</a></li>
<li class="chapter" data-level="9.2" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#libraries-3"><i class="fa fa-check"></i><b>9.2</b> Libraries</a></li>
<li class="chapter" data-level="9.3" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#illustrating-linear-regression-with-simulated-data"><i class="fa fa-check"></i><b>9.3</b> Illustrating linear regression with simulated data</a></li>
<li class="chapter" data-level="9.4" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#specifying-regression-models-in-r"><i class="fa fa-check"></i><b>9.4</b> Specifying Regression Models in R</a><ul>
<li class="chapter" data-level="9.4.1" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#fitted-values"><i class="fa fa-check"></i><b>9.4.1</b> Fitted values</a></li>
<li class="chapter" data-level="9.4.2" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#getting-the-model-coefficients"><i class="fa fa-check"></i><b>9.4.2</b> Getting the model coefficients</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#residuals"><i class="fa fa-check"></i><b>9.5</b> Residuals</a></li>
<li class="chapter" data-level="9.6" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#regression-as-sum-of-squares-decomposition"><i class="fa fa-check"></i><b>9.6</b> Regression as sum-of-squares decomposition</a></li>
<li class="chapter" data-level="9.7" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#variance-explained-by-a-regression-model"><i class="fa fa-check"></i><b>9.7</b> Variance “explained” by a regression model</a></li>
<li class="chapter" data-level="9.8" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broom-a-library-for-converting-model-results-into-data-frames"><i class="fa fa-check"></i><b>9.8</b> Broom: a library for converting model results into data frames</a><ul>
<li class="chapter" data-level="9.8.1" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broomtidy"><i class="fa fa-check"></i><b>9.8.1</b> <code>broom::tidy</code></a></li>
<li class="chapter" data-level="9.8.2" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broomaugment"><i class="fa fa-check"></i><b>9.8.2</b> <code>broom::augment</code></a></li>
<li class="chapter" data-level="9.8.3" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broomglance"><i class="fa fa-check"></i><b>9.8.3</b> <code>broom::glance</code></a></li>
</ul></li>
<li class="chapter" data-level="9.9" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#bivariate-linear-regression-on-a-real-world-data-set"><i class="fa fa-check"></i><b>9.9</b> Bivariate linear regression on a real world data set</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="matrices-in-r.html"><a href="matrices-in-r.html"><i class="fa fa-check"></i><b>10</b> Matrices in R</a><ul>
<li class="chapter" data-level="10.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#creating-matrices-in-r"><i class="fa fa-check"></i><b>10.1</b> Creating matrices in R</a></li>
<li class="chapter" data-level="10.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-arithmetic-operations-in-r"><i class="fa fa-check"></i><b>10.2</b> Matrix arithmetic operations in R</a></li>
<li class="chapter" data-level="10.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#descriptive-statistics-as-matrix-functions"><i class="fa fa-check"></i><b>10.3</b> Descriptive statistics as matrix functions</a><ul>
<li class="chapter" data-level="10.3.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#mean-vector-and-matrix"><i class="fa fa-check"></i><b>10.3.1</b> Mean vector and matrix</a></li>
<li class="chapter" data-level="10.3.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#deviation-matrix"><i class="fa fa-check"></i><b>10.3.2</b> Deviation matrix</a></li>
<li class="chapter" data-level="10.3.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#covariance-matrix"><i class="fa fa-check"></i><b>10.3.3</b> Covariance matrix</a></li>
<li class="chapter" data-level="10.3.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#correlation-matrix"><i class="fa fa-check"></i><b>10.3.4</b> Correlation matrix</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-inverse"><i class="fa fa-check"></i><b>10.4</b> Matrix Inverse</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="multiple-regression.html"><a href="multiple-regression.html"><i class="fa fa-check"></i><b>11</b> Multiple Regression</a><ul>
<li class="chapter" data-level="11.1" data-path="multiple-regression.html"><a href="multiple-regression.html#new-libraries-to-install"><i class="fa fa-check"></i><b>11.1</b> New Libraries to install</a></li>
<li class="chapter" data-level="11.2" data-path="multiple-regression.html"><a href="multiple-regression.html#libraries-4"><i class="fa fa-check"></i><b>11.2</b> Libraries</a></li>
<li class="chapter" data-level="11.3" data-path="multiple-regression.html"><a href="multiple-regression.html#examplar-data"><i class="fa fa-check"></i><b>11.3</b> Examplar data</a></li>
<li class="chapter" data-level="11.4" data-path="multiple-regression.html"><a href="multiple-regression.html#data-exploration"><i class="fa fa-check"></i><b>11.4</b> Data exploration</a></li>
<li class="chapter" data-level="11.5" data-path="multiple-regression.html"><a href="multiple-regression.html#d-plots"><i class="fa fa-check"></i><b>11.5</b> 3D Plots</a><ul>
<li class="chapter" data-level="11.5.1" data-path="multiple-regression.html"><a href="multiple-regression.html#scatterplot3d"><i class="fa fa-check"></i><b>11.5.1</b> scatterplot3d</a></li>
<li class="chapter" data-level="11.5.2" data-path="multiple-regression.html"><a href="multiple-regression.html#rgl"><i class="fa fa-check"></i><b>11.5.2</b> rgl</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="multiple-regression.html"><a href="multiple-regression.html#fitting-the-regression-model"><i class="fa fa-check"></i><b>11.6</b> Fitting the regression model</a><ul>
<li class="chapter" data-level="11.6.1" data-path="multiple-regression.html"><a href="multiple-regression.html#visualizing-the-regression-model-in-scatterplot3d"><i class="fa fa-check"></i><b>11.6.1</b> Visualizing the regression model in scatterplot3d</a></li>
</ul></li>
<li class="chapter" data-level="11.7" data-path="multiple-regression.html"><a href="multiple-regression.html#interpretting-the-regression-model"><i class="fa fa-check"></i><b>11.7</b> Interpretting the regression model</a></li>
<li class="chapter" data-level="11.8" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-vector-geometry-of-the-regression-model"><i class="fa fa-check"></i><b>11.8</b> Exploring the Vector Geometry of the Regression Model</a></li>
<li class="chapter" data-level="11.9" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-residuals-from-the-model-fit"><i class="fa fa-check"></i><b>11.9</b> Exploring the Residuals from the Model Fit</a></li>
<li class="chapter" data-level="11.10" data-path="multiple-regression.html"><a href="multiple-regression.html#an-alternate-model"><i class="fa fa-check"></i><b>11.10</b> An alternate model</a></li>
<li class="chapter" data-level="11.11" data-path="multiple-regression.html"><a href="multiple-regression.html#fitting-a-curvilinear-model-using-lm"><i class="fa fa-check"></i><b>11.11</b> Fitting a curvilinear model using lm()</a></li>
<li class="chapter" data-level="11.12" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-impact-of-nearly-collinear-predictors-on-regression"><i class="fa fa-check"></i><b>11.12</b> Exploring the impact of nearly collinear predictors on regression</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html"><i class="fa fa-check"></i><b>12</b> Principal Components Analysis</a><ul>
<li class="chapter" data-level="12.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#libraries-5"><i class="fa fa-check"></i><b>12.1</b> Libraries</a></li>
<li class="chapter" data-level="12.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#matrices-as-linear-transformations"><i class="fa fa-check"></i><b>12.2</b> Matrices as linear transformations</a></li>
<li class="chapter" data-level="12.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#eigenanalysis-in-r"><i class="fa fa-check"></i><b>12.3</b> Eigenanalysis in R</a></li>
<li class="chapter" data-level="12.4" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#principal-components-analysis-in-r"><i class="fa fa-check"></i><b>12.4</b> Principal Components Analysis in R</a><ul>
<li class="chapter" data-level="12.4.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#bioenv-dataset"><i class="fa fa-check"></i><b>12.4.1</b> Bioenv dataset</a></li>
<li class="chapter" data-level="12.4.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pca-of-the-bioenv-dataset"><i class="fa fa-check"></i><b>12.4.2</b> PCA of the Bioenv dataset</a></li>
<li class="chapter" data-level="12.4.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#calculating-factor-loadings"><i class="fa fa-check"></i><b>12.4.3</b> Calculating Factor Loadings</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#drawing-figures-to-represent-pca"><i class="fa fa-check"></i><b>12.5</b> Drawing Figures to Represent PCA</a><ul>
<li class="chapter" data-level="12.5.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pc-score-plots"><i class="fa fa-check"></i><b>12.5.1</b> PC Score Plots</a></li>
<li class="chapter" data-level="12.5.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#simultaneous-depiction-of-observations-and-variables-in-the-pc-space"><i class="fa fa-check"></i><b>12.5.2</b> Simultaneous Depiction of Observations and Variables in the PC Space</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html"><i class="fa fa-check"></i><b>13</b> Simulating Sampling Distributions</a><ul>
<li class="chapter" data-level="13.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#libraries-6"><i class="fa fa-check"></i><b>13.1</b> Libraries</a></li>
<li class="chapter" data-level="13.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#data-set-simulated-male-heights"><i class="fa fa-check"></i><b>13.2</b> Data set: Simulated male heights</a><ul>
<li class="chapter" data-level="13.2.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#seeding-the-pseudo-random-number-generator"><i class="fa fa-check"></i><b>13.2.1</b> Seeding the pseudo-random number generator</a></li>
<li class="chapter" data-level="13.2.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#generating-the-simulated-population"><i class="fa fa-check"></i><b>13.2.2</b> Generating the simulated population</a></li>
<li class="chapter" data-level="13.2.3" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#properties-of-the-simulated-population"><i class="fa fa-check"></i><b>13.2.3</b> Properties of the simulated population</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#random-sampling-from-the-simulated-population"><i class="fa fa-check"></i><b>13.3</b> Random sampling from the simulated population</a><ul>
<li class="chapter" data-level="13.3.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#another-random-sample"><i class="fa fa-check"></i><b>13.3.1</b> Another random sample</a></li>
<li class="chapter" data-level="13.3.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#simulating-the-generation-of-many-random-samples"><i class="fa fa-check"></i><b>13.3.2</b> Simulating the generation of many random samples</a></li>
<li class="chapter" data-level="13.3.3" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#a-function-to-estimate-statistics-of-interest-in-a-random-sample"><i class="fa fa-check"></i><b>13.3.3</b> A function to estimate statistics of interest in a random sample</a></li>
<li class="chapter" data-level="13.3.4" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#generating-statistics-for-many-random-samples"><i class="fa fa-check"></i><b>13.3.4</b> Generating statistics for many random samples</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#simulated-sampling-distribution-of-the-mean"><i class="fa fa-check"></i><b>13.4</b> Simulated sampling distribution of the mean</a><ul>
<li class="chapter" data-level="13.4.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#sampling-distributions-for-different-sample-sizes"><i class="fa fa-check"></i><b>13.4.1</b> Sampling distributions for different sample sizes</a></li>
<li class="chapter" data-level="13.4.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#discussion-of-trends-for-sampling-distributions-of-different-sample-sizes"><i class="fa fa-check"></i><b>13.4.2</b> Discussion of trends for sampling distributions of different sample sizes</a></li>
</ul></li>
<li class="chapter" data-level="13.5" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#standard-error-of-the-mean"><i class="fa fa-check"></i><b>13.5</b> Standard Error of the Mean</a></li>
<li class="chapter" data-level="13.6" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#sampling-distribution-of-the-standard-deviation"><i class="fa fa-check"></i><b>13.6</b> Sampling Distribution of the Standard Deviation</a></li>
<li class="chapter" data-level="13.7" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#what-happens-to-the-sampling-distribution-of-the-mean-and-standard-deviation-when-our-sample-size-is-small"><i class="fa fa-check"></i><b>13.7</b> What happens to the sampling distribution of the mean and standard deviation when our sample size is small?</a><ul>
<li class="chapter" data-level="13.7.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#for-small-samples-sample-standard-deviations-systematically-underestimate-the-population-standard-deviation"><i class="fa fa-check"></i><b>13.7.1</b> For small samples, sample standard deviations systematically underestimate the population standard deviation</a></li>
<li class="chapter" data-level="13.7.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#underestimates-of-the-standard-deviation-given-small-n-lead-to-understimates-of-the-se-of-the-mean"><i class="fa fa-check"></i><b>13.7.2</b> Underestimates of the standard deviation given small <span class="math inline">\(n\)</span> lead to understimates of the SE of the mean</a></li>
</ul></li>
<li class="chapter" data-level="13.8" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#the-t-distribution-is-the-appropriate-distribution-for-describing-the-sampling-distribution-of-the-mean-when-n-is-small"><i class="fa fa-check"></i><b>13.8</b> The t-distribution is the appropriate distribution for describing the sampling distribution of the mean when <span class="math inline">\(n\)</span> is small</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html"><i class="fa fa-check"></i><b>14</b> Simulating confidence intervals</a><ul>
<li class="chapter" data-level="14.1" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#libraries-7"><i class="fa fa-check"></i><b>14.1</b> Libraries</a></li>
<li class="chapter" data-level="14.2" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#confidence-intervals"><i class="fa fa-check"></i><b>14.2</b> Confidence Intervals</a></li>
<li class="chapter" data-level="14.3" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#calibrating-confidence-intervals"><i class="fa fa-check"></i><b>14.3</b> Calibrating confidence intervals</a></li>
<li class="chapter" data-level="14.4" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#standard-formulation-for-confidence-intervals"><i class="fa fa-check"></i><b>14.4</b> Standard formulation for confidence intervals</a></li>
<li class="chapter" data-level="14.5" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#simulating-confidence-intervals-for-the-mean"><i class="fa fa-check"></i><b>14.5</b> Simulating confidence intervals for the mean</a></li>
<li class="chapter" data-level="14.6" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#sample-estimate-of-the-standard-error-of-the-mean"><i class="fa fa-check"></i><b>14.6</b> Sample estimate of the standard error of the mean</a></li>
<li class="chapter" data-level="14.7" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#ci-of-mean-simulation"><i class="fa fa-check"></i><b>14.7</b> CI of mean simulation</a></li>
<li class="chapter" data-level="14.8" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#generating-a-table-of-cis-and-corresponding-margins-of-error"><i class="fa fa-check"></i><b>14.8</b> Generating a table of CIs and corresponding margins of error</a></li>
<li class="chapter" data-level="14.9" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#interpreting-our-simulation"><i class="fa fa-check"></i><b>14.9</b> Interpreting our simulation</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html"><i class="fa fa-check"></i><b>15</b> Randomization, Jackknife, and Bootstrap</a><ul>
<li class="chapter" data-level="15.1" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#libraries-8"><i class="fa fa-check"></i><b>15.1</b> Libraries</a></li>
<li class="chapter" data-level="15.2" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#randomization"><i class="fa fa-check"></i><b>15.2</b> Randomization</a></li>
<li class="chapter" data-level="15.3" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#using-randomization-to-test-for-a-difference-in-means"><i class="fa fa-check"></i><b>15.3</b> Using randomization to test for a difference in means</a></li>
<li class="chapter" data-level="15.4" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#using-randomization-to-test-for-equality-of-variances"><i class="fa fa-check"></i><b>15.4</b> Using randomization to test for equality of variances</a></li>
<li class="chapter" data-level="15.5" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#jackknifing-in-r"><i class="fa fa-check"></i><b>15.5</b> Jackknifing in R</a></li>
<li class="chapter" data-level="15.6" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#bootstrapping-in-r"><i class="fa fa-check"></i><b>15.6</b> Bootstrapping in R</a><ul>
<li class="chapter" data-level="15.6.1" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#a-more-sophisticated-application-of-the-bootstrap"><i class="fa fa-check"></i><b>15.6.1</b> A more sophisticated application of the bootstrap</a></li>
<li class="chapter" data-level="15.6.2" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#about-logistic-regression"><i class="fa fa-check"></i><b>15.6.2</b> About logistic regression</a></li>
<li class="chapter" data-level="15.6.3" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#bumpus-house-sparrow-data-set"><i class="fa fa-check"></i><b>15.6.3</b> Bumpus house sparrow data set</a></li>
<li class="chapter" data-level="15.6.4" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#predicting-survival-as-a-function-of-body-weight"><i class="fa fa-check"></i><b>15.6.4</b> Predicting survival as a function of body weight</a></li>
<li class="chapter" data-level="15.6.5" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#fitting-the-logistic-regression"><i class="fa fa-check"></i><b>15.6.5</b> Fitting the logistic regression</a></li>
<li class="chapter" data-level="15.6.6" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#defining-an-appropriate-function-for-boot"><i class="fa fa-check"></i><b>15.6.6</b> Defining an appropriate function for <code>boot</code></a></li>
<li class="chapter" data-level="15.6.7" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#calculating-bootstrap-confidence-intervals"><i class="fa fa-check"></i><b>15.6.7</b> Calculating bootstrap confidence intervals</a></li>
<li class="chapter" data-level="15.6.8" data-path="randomization-jackknife-and-bootstrap.html"><a href="randomization-jackknife-and-bootstrap.html#visualizing-bootstrap-confidence-intervals"><i class="fa fa-check"></i><b>15.6.8</b> Visualizing bootstrap confidence intervals</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Biology 723: Statistical Computing for Biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="randomization-jackknife-and-bootstrap" class="section level1">
<h1><span class="header-section-number">Chapter 15</span> Randomization, Jackknife, and Bootstrap</h1>
<div id="libraries-8" class="section level2">
<h2><span class="header-section-number">15.1</span> Libraries</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)
<span class="kw">library</span>(broom)
<span class="kw">set.seed</span>(<span class="dv">20180404</span>)</code></pre></div>
</div>
<div id="randomization" class="section level2">
<h2><span class="header-section-number">15.2</span> Randomization</h2>
<p>There are a number of packages (e.g. <code>coin</code>) that include functions for carrying out randomization/permutation tests in R. However, it’s often just as easy to write a quick set of functions to carry out such tests yourself. We’ll illustrate a simple example of this using an example from Manly (2007).</p>
<p>Consider the following data set composed of measures of mandible lengths (in mm) for male and female golden jackals. This set of measurements was taken from a set of skeletons in the collections of the British Museum of Natural History.</p>
<ul>
<li><strong>Females</strong>: 110, 111, 107, 108, 110, 105, 107, 106, 111, 111</li>
<li><strong>Males</strong>: 120, 107, 110, 116, 114, 111, 113, 117, 114, 112</li>
</ul>
<p>Let’s first two vectors to represent this set of measurements, and then create a data frame to hold this information:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">male.mandibles &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">120</span>,<span class="dv">107</span>,<span class="dv">110</span>,<span class="dv">116</span>,<span class="dv">114</span>,<span class="dv">111</span>,<span class="dv">113</span>,<span class="dv">117</span>,<span class="dv">114</span>,<span class="dv">112</span>)
female.mandibles &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">110</span>,<span class="dv">111</span>,<span class="dv">107</span>,<span class="dv">108</span>,<span class="dv">110</span>,<span class="dv">105</span>,<span class="dv">107</span>,<span class="dv">106</span>,<span class="dv">111</span>,<span class="dv">111</span>)

jackals &lt;-<span class="st"> </span><span class="kw">data_frame</span>(<span class="dt">sex =</span> <span class="kw">c</span>(<span class="kw">rep</span>(<span class="st">&quot;m&quot;</span>,<span class="dv">10</span>),<span class="kw">rep</span>(<span class="st">&quot;f&quot;</span>,<span class="dv">10</span>)), 
                      <span class="dt">mandible.length =</span> <span class="kw">c</span>(male.mandibles,female.mandibles))</code></pre></div>
<p>And we’ll create a simple strip chart to visualize the data</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(jackals, <span class="kw">aes</span>(<span class="dt">x =</span> sex, <span class="dt">y =</span> mandible.length, <span class="dt">color =</span> sex)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_jitter</span>(<span class="dt">width=</span><span class="fl">0.1</span>, <span class="dt">height=</span><span class="dv">0</span>,<span class="dt">size=</span><span class="dv">2</span>,<span class="dt">alpha=</span><span class="fl">0.5</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-406-1.png" width="576" /></p>
<p>Having done so, let’s look at the male and female means and the difference between them:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">jackals <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(sex) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">mean.mandible.length =</span> <span class="kw">mean</span>(mandible.length))
<span class="co"># A tibble: 2 x 2</span>
  sex   mean.mandible.length
  <span class="op">&lt;</span>chr<span class="op">&gt;</span><span class="st">                </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="dv">1</span> f                     <span class="dv">109</span>.
<span class="dv">2</span> m                     <span class="dv">113</span>.</code></pre></div>
<p>The difference in the mean mandible length is:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">mean</span>(male.mandibles) <span class="op">-</span><span class="st"> </span><span class="kw">mean</span>(female.mandibles) 
[<span class="dv">1</span>] <span class="fl">4.8</span></code></pre></div>
</div>
<div id="using-randomization-to-test-for-a-difference-in-means" class="section level2">
<h2><span class="header-section-number">15.3</span> Using randomization to test for a difference in means</h2>
<p>The hypothesis we want to test is that male jackals have, on average, larger mandibles than female jackals. The strip plot we constructed and difference in the means would seem to suggest so but let’s carry out some more formal tests. The obvious way to compare this set of measurements would be to carry out a t-test, which is approropriate if the samples are normally distributed with approximately equal variance. We have small samples here, so it’s hard to know if the normal distribution holds. Instead we’ll use a randomization test to compare the observed difference in the means to the distribution of differences we would expect to observe if the labels <code>male' and</code>female’ were randomly applied to samples of equal size from the data at hand.</p>
<p>Let’s create a function that takes a sample and randomly assigns the observations to two groups of a specified size. The function takes as input a vector of values (size <span class="math inline">\(N\)</span>) and two integers representing the sample sizes (<span class="math inline">\(n_1\)</span> and <span class="math inline">\(n_2\)</span> where <span class="math inline">\(n_1 + n_2 = N\)</span>) of the two groups to be compared.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">randomize.two.groups &lt;-<span class="st"> </span><span class="cf">function</span>(x, n1, n2){
  <span class="co"># sample w/out replacement</span>
  reordered &lt;-<span class="st"> </span><span class="kw">sample</span>(x, <span class="kw">length</span>(x)) <span class="co"># see help(sample) for more info</span>
  g1 &lt;-<span class="st"> </span>reordered[<span class="kw">seq</span>(<span class="dv">1</span>,n1)]  <span class="co"># take the first n1 items as group 1</span>
  g2 &lt;-<span class="st"> </span>reordered[<span class="kw">seq</span>(n1<span class="op">+</span><span class="dv">1</span>,n1<span class="op">+</span>n2)] <span class="co"># take the the remaining items as group 2</span>
  <span class="kw">list</span>(g1,g2)
}</code></pre></div>
<p>Test out this function by calling it repeatedly as shown below. You’ll see that it returns a random reordering of the original data, split into two groups:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">randomize.two.groups</span>(jackals<span class="op">$</span>mandible.length, <span class="dv">10</span>, <span class="dv">10</span>)
[[<span class="dv">1</span>]]
 [<span class="dv">1</span>] <span class="dv">116</span> <span class="dv">107</span> <span class="dv">110</span> <span class="dv">120</span> <span class="dv">110</span> <span class="dv">114</span> <span class="dv">107</span> <span class="dv">110</span> <span class="dv">117</span> <span class="dv">105</span>

[[<span class="dv">2</span>]]
 [<span class="dv">1</span>] <span class="dv">114</span> <span class="dv">106</span> <span class="dv">108</span> <span class="dv">111</span> <span class="dv">112</span> <span class="dv">107</span> <span class="dv">111</span> <span class="dv">113</span> <span class="dv">111</span> <span class="dv">111</span></code></pre></div>
<p>Everytime you call this function you get a different random reordering:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">randomize.two.groups</span>(jackals<span class="op">$</span>mandible.length,<span class="dv">10</span>,<span class="dv">10</span>) <span class="co"># call it again to get a different sample</span>
[[<span class="dv">1</span>]]
 [<span class="dv">1</span>] <span class="dv">111</span> <span class="dv">108</span> <span class="dv">110</span> <span class="dv">110</span> <span class="dv">111</span> <span class="dv">113</span> <span class="dv">107</span> <span class="dv">111</span> <span class="dv">110</span> <span class="dv">105</span>

[[<span class="dv">2</span>]]
 [<span class="dv">1</span>] <span class="dv">112</span> <span class="dv">106</span> <span class="dv">111</span> <span class="dv">116</span> <span class="dv">114</span> <span class="dv">117</span> <span class="dv">107</span> <span class="dv">120</span> <span class="dv">107</span> <span class="dv">114</span></code></pre></div>
<p>Now let’s write a simple function that returns the mean difference between two samples:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mean.diff &lt;-<span class="st"> </span><span class="cf">function</span>(x1, x2) {
  <span class="kw">mean</span>(x1) <span class="op">-</span><span class="st"> </span><span class="kw">mean</span>(x2)
}</code></pre></div>
<p>Now let’s write a generic randomization function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">randomization &lt;-<span class="st"> </span><span class="cf">function</span>(x1, x2, fxn,<span class="dt">nsamples=</span><span class="dv">100</span>){
  stats &lt;-<span class="st"> </span><span class="kw">c</span>()  <span class="co"># will hold our randomized stats</span>
  orig &lt;-<span class="st"> </span><span class="kw">c</span>(x1,x2)
  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>nsamples){
    g &lt;-<span class="st"> </span><span class="kw">randomize.two.groups</span>(orig, <span class="kw">length</span>(x1), <span class="kw">length</span>(x2))
    stats &lt;-<span class="st"> </span><span class="kw">c</span>(stats, <span class="kw">fxn</span>(g[[<span class="dv">1</span>]],g[[<span class="dv">2</span>]]))
  }
  <span class="kw">return</span> (stats)
}</code></pre></div>
<p>We can then use the <code>randomization()</code> function we wrote as follows to evaluate the significance of the observed difference in means in the original sample. First we generate 1000 random samples, where the group membership has been randomized with respect to the measurements, and calculate the mean difference between the groups for each of those randomized samples:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># generate 1000  samples of the mean.diff for randomized data</span>
rsample &lt;-<span class="st"> </span><span class="kw">randomization</span>(male.mandibles, 
                         female.mandibles, 
                         mean.diff, <span class="dv">1000</span>)</code></pre></div>
<p>Now let’s examine the distribution, using a function called <code>quickplot()</code> which is a convenience function included with <code>ggplot2</code> that enables us to avoid creating data frames when we just want to create some quick visualizations:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># examine the distribution</span>
<span class="kw">quickplot</span>(rsample, <span class="dt">geom=</span><span class="st">&#39;histogram&#39;</span>,
          <span class="dt">bins=</span><span class="dv">20</span>,
          <span class="dt">main=</span><span class="st">&quot;Histogram of randomized differences</span><span class="ch">\n</span><span class="st">in male and female means&quot;</span>,
          <span class="dt">xlab =</span> <span class="st">&quot;mean(males) - mean(females)&quot;</span>)  </code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-415-1.png" width="576" /></p>
<p>We then pose the questions “In how many of the random samples is the mean difference between the two groups as great or larger than the observed difference in our original samples?”</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ngreater &lt;-<span class="st"> </span><span class="kw">sum</span>(rsample <span class="op">&gt;=</span><span class="st"> </span><span class="fl">4.8</span>)
ngreater
[<span class="dv">1</span>] <span class="dv">1</span></code></pre></div>
<p>So our conclusion is that the probability of getting a mean difference between samples of this size is about 1/1000 = 0.001. Note that we can’t generalize this to golden jackals as a whole because we know nothing about whether these samples actually represent random samples of the golden jackal population or biases that might have been imposed on the collection (e.g. maybe the collectors liked to single out particularly large males). However, if we saw a similar trend (males larger than females) in multiple museum collections we might see this as supporting evidence that the trend holds true in general.</p>
</div>
<div id="using-randomization-to-test-for-equality-of-variances" class="section level2">
<h2><span class="header-section-number">15.4</span> Using randomization to test for equality of variances</h2>
<p>Note that we wrote our <code>randomization()</code> function to take an arbitrary function that takes as its input two vectors of data. That means we can use it to estimate the randomized distribution of arbitrary statistics of interest. Here we illustrate that with another function that calculates the ratio of variances. This for example could be used to assess the null hypothesis that male and female jackals have similar variation in mandible length. If the null hypothesis was true, we would expect that the ratio of variances should be approximately 1.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ratio.var &lt;-<span class="st"> </span><span class="cf">function</span>(x1, x2){
    <span class="kw">var</span>(x1)<span class="op">/</span><span class="kw">var</span>(x2)
}

<span class="kw">ratio.var</span>(male.mandibles,
          female.mandibles)  <span class="co"># ratio of variances for the original samples</span>
[<span class="dv">1</span>] <span class="fl">2.681034</span></code></pre></div>
<p>Here we see that the variance of the male sample is more than 2.5 times that of the female sample. But would this ratio of variances be unexpected under a null hypothesis of equal variances between the groups?</p>
<p>Again, our approach is to randomize the assignment of group labels associated with each measured value, and calculate the statistic of interest (ratio of variances in this case), for each randomized instance:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">vsample &lt;-<span class="st"> </span><span class="kw">randomization</span>(male.mandibles, female.mandibles, ratio.var, <span class="dv">1000</span>)
<span class="kw">quickplot</span>(vsample, <span class="dt">bins=</span><span class="dv">20</span>,
          <span class="dt">main=</span><span class="st">&quot;Histogram of randomized ratios</span><span class="ch">\n</span><span class="st">of male and female variances&quot;</span>,
          <span class="dt">xlab =</span> <span class="st">&quot;var(males)/var(females)&quot;</span>)  </code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-418-1.png" width="576" /></p>
<p>The mean ratio of variances in our randomized sample is:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">mean</span>(vsample)
[<span class="dv">1</span>] <span class="fl">1.277209</span></code></pre></div>
<p>The number and fraction of times our randomized samples are greater than that observed for our actual samples is:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sum</span>(vsample <span class="op">&gt;=</span><span class="st"> </span><span class="fl">2.68</span>)  <span class="co"># number of times </span>
[<span class="dv">1</span>] <span class="dv">79</span>
<span class="kw">sum</span>(vsample <span class="op">&gt;=</span><span class="st"> </span><span class="fl">2.68</span>)<span class="op">/</span><span class="dv">1000</span>. <span class="co"># fraction of times </span>
[<span class="dv">1</span>] <span class="fl">0.079</span></code></pre></div>
<p>In this case the observed ratio of variances is at roughly the 92.5% percentile of the distribution. This doesn’t quite meet the conventional 95% threshold for “significance” but it is suggests some support to doubt the null hypothesis of equal variances.</p>
<p>Let’s make one more comparison. From your intro biostats course you should be recall that ratios of variances have an <span class="math inline">\(F\)</span>-distribution so let’s compare the distribution of ratios of variances from our randomized sample to that of a sample of the same size drawn from the <span class="math inline">\(F\)</span>-distribution with the same degrees of freedom.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">randomF &lt;-<span class="st"> </span><span class="kw">rf</span>(<span class="dv">1000</span>, <span class="dv">9</span>, <span class="dv">9</span>) <span class="co"># see help(rf)</span>

<span class="kw">ggplot</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_density</span>(<span class="kw">aes</span>(vsample, <span class="dt">color=</span><span class="st">&#39;randomization&#39;</span>)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_density</span>(<span class="kw">aes</span>(randomF, <span class="dt">color=</span><span class="st">&#39;theoretical&#39;</span>)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Ratio of Variances&quot;</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">scale_color_manual</span>(<span class="st">&quot;Type&quot;</span>, <span class="dt">values =</span> <span class="kw">c</span>(<span class="st">&quot;steelblue&quot;</span>, <span class="st">&quot;firebrick&quot;</span>))</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-421-1.png" width="576" /></p>
</div>
<div id="jackknifing-in-r" class="section level2">
<h2><span class="header-section-number">15.5</span> Jackknifing in R</h2>
<p>Jackknife estimates of simple statistics are also relatively straightforward to calculate in R. Here we implement our own simple jackknife function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">jknife &lt;-<span class="st"> </span><span class="cf">function</span>(x, fxn, <span class="dt">ci=</span><span class="fl">0.95</span>) {
    theta &lt;-<span class="st"> </span><span class="kw">fxn</span>(x)
    n &lt;-<span class="st"> </span><span class="kw">length</span>(x)
    partials &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">0</span>,n)  
    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n){
       partials[i] &lt;-<span class="st"> </span><span class="kw">fxn</span>(x[<span class="op">-</span>i])
    }
    pseudos &lt;-<span class="st"> </span>(n<span class="op">*</span>theta) <span class="op">-</span><span class="st"> </span>(n<span class="op">-</span><span class="dv">1</span>)<span class="op">*</span>partials
    jack.est &lt;-<span class="st"> </span><span class="kw">mean</span>(pseudos)
    jack.se &lt;-<span class="st"> </span><span class="kw">sqrt</span>(<span class="kw">var</span>(pseudos)<span class="op">/</span>n)
    alpha =<span class="st"> </span><span class="dv">1</span><span class="op">-</span>ci
    CI &lt;-<span class="st"> </span><span class="kw">qt</span>(alpha<span class="op">/</span><span class="dv">2</span>,n<span class="op">-</span><span class="dv">1</span>,<span class="dt">lower.tail=</span><span class="ot">FALSE</span>)<span class="op">*</span>jack.se
    jack.ci &lt;-<span class="st"> </span><span class="kw">c</span>(jack.est <span class="op">-</span><span class="st"> </span>CI, jack.est <span class="op">+</span><span class="st"> </span>CI)
    <span class="kw">list</span>(<span class="dt">est=</span>jack.est, <span class="dt">se=</span>jack.se, <span class="dt">ci=</span>jack.ci)
}</code></pre></div>
<p>The <code>bootstrap</code> package contains a very similar implementation of a jackknife function (<code>jackknife()</code>).</p>
<p>Let’s illustrate our jackknife function using samples drawn from a Poisson distribution. The Poisson is a discrete probability distribution that is often used to describe the probability of a number of events occuring in a fixed period of time, where the events are independent and occur with an average rate <span class="math inline">\(\lambda\)</span>. The Poisson distribution is used to model processes like mutations in DNA sequences or atomic decay. Both the mean and variance of a Poisson distribution are equal to <span class="math inline">\(\lambda\)</span>. Let’s see how well the jackknife does at estimating confidence intervals for the mean and variance of a modest number of samples drawn from a Poisson.</p>
<p>First we draw a random sample from Poisson distribution with <span class="math inline">\(\lambda = 4\)</span>. Since we’re simulating a random sample we know what the “true” value is.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">psample &lt;-<span class="st"> </span><span class="kw">rpois</span>(<span class="dv">25</span>,<span class="dv">4</span>) <span class="co"># 25 observations from poisson with lambda = 4</span>
psample  <span class="co"># your sample will be different unless you used the same seed as I did</span>
 [<span class="dv">1</span>] <span class="dv">9</span> <span class="dv">6</span> <span class="dv">8</span> <span class="dv">6</span> <span class="dv">2</span> <span class="dv">3</span> <span class="dv">4</span> <span class="dv">3</span> <span class="dv">8</span> <span class="dv">3</span> <span class="dv">1</span> <span class="dv">4</span> <span class="dv">5</span> <span class="dv">2</span> <span class="dv">7</span> <span class="dv">2</span> <span class="dv">2</span> <span class="dv">4</span> <span class="dv">3</span> <span class="dv">4</span> <span class="dv">6</span> <span class="dv">6</span> <span class="dv">3</span> <span class="dv">8</span> <span class="dv">2</span></code></pre></div>
<p>Our sample estimates of the mean and variance are:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">mean</span>(psample)
[<span class="dv">1</span>] <span class="fl">4.44</span>
<span class="kw">var</span>(psample)
[<span class="dv">1</span>] <span class="fl">5.34</span></code></pre></div>
<p>We now jackknifing to estimate standard errors and confidence intervals for the mean and variance:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">jack.means &lt;-<span class="st"> </span><span class="kw">jknife</span>(psample, mean)
jack.vars &lt;-<span class="st"> </span><span class="kw">jknife</span>(psample, var)</code></pre></div>
<p>Here is the 95% CI for the means and variances:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">jack.means<span class="op">$</span>cin <span class="co"># 95% bootstreap CI for mean</span>
<span class="ot">NULL</span>
jack.vars<span class="op">$</span>ci <span class="co"># 95% boostrap CI for vars</span>
[<span class="dv">1</span>] <span class="fl">2.982004</span> <span class="fl">7.697996</span></code></pre></div>
<p>In both cases above, the true mean and variance were contained within the 95% confidence intervals estimated by the jackknife.</p>
<p>Let’s do a little experiment to see how often that’s true for samples of this size:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># create 500 samples of size 25 drawn from Poisson w/lambda=4</span>
psamples &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rpois</span>(<span class="dv">25</span><span class="op">*</span><span class="dv">500</span>,<span class="dv">4</span>),<span class="dt">ncol=</span><span class="dv">25</span>,<span class="dt">byrow=</span>T)
<span class="kw">dim</span>(psamples)
[<span class="dv">1</span>] <span class="dv">500</span>  <span class="dv">25</span>

<span class="co"># create a convenience function</span>
get.ci &lt;-<span class="st"> </span><span class="cf">function</span>(x) { <span class="kw">return</span>(x<span class="op">$</span>ci) }  <span class="co">#x$ci gives confidence interval</span>

<span class="co"># generate jackknife estimates for mean</span>
j.mean &lt;-<span class="st"> </span><span class="kw">apply</span>(psamples, <span class="dv">1</span>, jknife, mean)

<span class="co"># make matrix that holds 95% confidence intervals of mean</span>
mean.ci &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">sapply</span>(j.mean, get.ci))
mean.ci[<span class="dv">1</span>,]
[<span class="dv">1</span>] <span class="fl">3.684721</span> <span class="fl">5.515279</span>
mean.ci[<span class="dv">2</span>,]
[<span class="dv">1</span>] <span class="fl">2.844251</span> <span class="fl">4.275749</span>

<span class="co"># check how often true mean is w/in CI</span>
includes.true.mean &lt;-<span class="st"> </span><span class="kw">sum</span>(mean.ci[,<span class="dv">1</span>] <span class="op">&lt;=</span><span class="dv">4</span>, mean.ci[,<span class="dv">2</span>] <span class="op">&gt;=</span><span class="st"> </span><span class="dv">4</span>)
includes.true.mean
[<span class="dv">1</span>] <span class="dv">983</span>
includes.true.mean<span class="op">/</span><span class="dv">500</span>  <span class="co"># true mean is w/in estimated 95% CI about 93% of the time.</span>
[<span class="dv">1</span>] <span class="fl">1.966</span>


<span class="co"># now the same for variances</span>
j.var &lt;-<span class="st"> </span><span class="kw">apply</span>(psamples, <span class="dv">1</span>, jknife, var)
var.ci &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">sapply</span>(j.var, get.ci))
includes.true.var &lt;-<span class="st"> </span><span class="kw">sum</span>(var.ci[,<span class="dv">1</span>] <span class="op">&lt;=</span><span class="dv">4</span>, var.ci[,<span class="dv">2</span>] <span class="op">&gt;=</span><span class="st"> </span><span class="dv">4</span>)
includes.true.var
[<span class="dv">1</span>] <span class="dv">950</span>
includes.true.var<span class="op">/</span><span class="dv">500</span>  <span class="co"># true variance is w/in 95% CI only 88% of time</span>
[<span class="dv">1</span>] <span class="fl">1.9</span></code></pre></div>
<p>In the case of the confidence intervals for the mean, the jacknife estimator did a decent job – the true mean is with the 95% confidence interval about 93% of the time. In the case of the variance it did less well. The jackknife confidence intervals work well when the estimator is normally distributed. This suggests that one way we might improve the jackknife CIs is by using a normalizing transformation, like the logarithm function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">log.var &lt;-<span class="st"> </span><span class="cf">function</span>(x){<span class="kw">log</span>(<span class="kw">var</span>(x))}
j.log.var &lt;-<span class="st"> </span><span class="kw">apply</span>(psamples, <span class="dv">1</span>, jknife, log.var)
log.var.ci &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">sapply</span>(j.log.var, get.ci))
includes.true.var.transformed &lt;-<span class="st"> </span><span class="kw">sum</span>(log.var.ci[,<span class="dv">1</span>] <span class="op">&lt;=</span><span class="kw">log</span>(<span class="dv">4</span>), log.var.ci[,<span class="dv">2</span>] <span class="op">&gt;=</span><span class="st"> </span><span class="kw">log</span>(<span class="dv">4</span>))

<span class="co"># an improvement in the performance of the 95% CIs</span>
includes.true.var.transformed<span class="op">/</span><span class="dv">500</span>  
[<span class="dv">1</span>] <span class="fl">1.944</span></code></pre></div>
<p>This illustrates the type of simulation study you might do to check the robustness of the jackknife for a statistic of interest for a given class of distributions.</p>
</div>
<div id="bootstrapping-in-r" class="section level2">
<h2><span class="header-section-number">15.6</span> Bootstrapping in R</h2>
<p>There are several packages that provide functions for doing bootstrapping in R. These include <code>bootstrap</code> and <code>boot</code>. We’ll take a quick look at the functions in <code>bootstrap</code>.</p>
<p>Install <code>bootstrap</code> using the standard package installation mechanism and then load it.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(bootstrap)

Attaching package<span class="op">:</span><span class="st"> &#39;bootstrap&#39;</span>
The following object is masked from <span class="st">&#39;package:broom&#39;</span><span class="op">:</span>

<span class="st">    </span>bootstrap</code></pre></div>
<p>We’ll start with the same set of samples from the Poisson that we used before to illustrate the jackknife.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># generate 1000 bootstrap sample estimate of var</span>
<span class="co"># using full name here to avoid naming conflict with broom</span>
b &lt;-<span class="st"> </span>bootstrap<span class="op">::</span><span class="kw">bootstrap</span>(psample, <span class="dv">1000</span>, var)</code></pre></div>
<p>The <code>bootstrap::bootstrap()</code> function returns a list:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(b)
List of <span class="dv">5</span>
 <span class="op">$</span><span class="st"> </span>thetastar     <span class="op">:</span><span class="st"> </span>num [<span class="dv">1</span><span class="op">:</span><span class="dv">1000</span>] <span class="fl">3.42</span> <span class="fl">6.06</span> <span class="fl">4.06</span> <span class="fl">3.56</span> <span class="fl">4.79</span> ...
 <span class="op">$</span><span class="st"> </span>func.thetastar<span class="op">:</span><span class="st"> </span><span class="ot">NULL</span>
 <span class="op">$</span><span class="st"> </span>jack.boot.val <span class="op">:</span><span class="st"> </span><span class="ot">NULL</span>
 <span class="op">$</span><span class="st"> </span>jack.boot.se  <span class="op">:</span><span class="st"> </span><span class="ot">NULL</span>
 <span class="op">$</span><span class="st"> </span>call          <span class="op">:</span><span class="st"> </span>language bootstrap<span class="op">::</span><span class="kw">bootstrap</span>(<span class="dt">x =</span> psample, <span class="dt">nboot =</span> <span class="dv">1000</span>, <span class="dt">theta =</span> var)</code></pre></div>
<p>The list item <code>thetastar</code> (a vector) contains each of the bootrap estimates of the statistic of interest (variance in the present example). It’s always good to plot a histogram of the bootstrap distribution:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">quickplot</span>(b<span class="op">$</span>thetastar, <span class="dt">geom=</span><span class="st">&quot;histogram&quot;</span>, <span class="dt">bins=</span><span class="dv">25</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-432-1.png" width="576" /></p>
<p>We’ll examine two different bootstrap confidence interval. The standard boostrap CIs are based on the assumption of the approximate normality of the sampling distribution of interest:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># standard bootstrap confidence limits</span>
<span class="co"># based on assumption of approximate normality of the bootstrap distn</span>
bstar &lt;-<span class="st"> </span>b<span class="op">$</span>thetastar
ci.multiplier =<span class="st"> </span><span class="kw">abs</span>(<span class="kw">qt</span>(<span class="fl">0.025</span>, <span class="dt">df=</span><span class="dv">24</span>))  <span class="co"># cutoff of t-distn w/24 df</span>
<span class="kw">c</span>(<span class="kw">mean</span>(bstar)<span class="op">-</span>ci.multiplier<span class="op">*</span><span class="kw">sd</span>(bstar),
  <span class="kw">mean</span>(bstar)<span class="op">+</span>ci.multiplier<span class="op">*</span><span class="kw">sd</span>(bstar))
[<span class="dv">1</span>] <span class="fl">2.905821</span> <span class="fl">7.357725</span></code></pre></div>
<p>Another approach is to use “percentile confidence limits” based on percentiles of the bootstrap distribution itself:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># estimate the bootstrap percentile confidence limits</span>
<span class="kw">quantile</span>(b<span class="op">$</span>thetastar,<span class="kw">c</span>(<span class="fl">0.025</span>,<span class="fl">0.975</span>))
    <span class="fl">2.5</span><span class="op">%    97.5%</span><span class="st"> </span>
<span class="fl">3.009167</span> <span class="fl">7.177500</span> </code></pre></div>
<div id="a-more-sophisticated-application-of-the-bootstrap" class="section level3">
<h3><span class="header-section-number">15.6.1</span> A more sophisticated application of the bootstrap</h3>
<p>Now let’s use the bootstrap to look at the distribution of a more complicated pair of statistics – the intercept and coefficients of a logistic regression. This time we’ll use the <code>boot</code> package to do the analysis, which is more flexible than the <code>bootstrap</code> package. <strong>Install the <code>boot</code> package as necessary.</strong></p>
</div>
<div id="about-logistic-regression" class="section level3">
<h3><span class="header-section-number">15.6.2</span> About logistic regression</h3>
<p>Logistic regression is used when the dependent variable is discrete (often binary). The explanatory variables may be either continuous or discrete.</p>
<p>Examples:</p>
<ul>
<li>Whether a gene is turned off (=0) or on (=1) as a function of levels of various proteins</li>
<li>Whether an individual is healthy (=0) or diseased (=1) as a function of various risk factors.</li>
<li>Whether an individual died (=0) or survived (=1) some selective event as a function of behavior, morphology, etc.</li>
</ul>
<p>We model the binary response variable, <span class="math inline">\(Y\)</span>, as a function of the predictor variables, <span class="math inline">\(X_1\)</span>, <span class="math inline">\(X_2\)</span>, etc as :</p>
<p><span class="math display">\[
P(Y = 1|X_1,\ldots,X_p) = f(\beta_0 + \beta_1 X_1 + \beta_2 X_2 + \cdots + \beta_p X_p)
\]</span></p>
<p>So we’re modeling the <em>probability of the state of Y as a function of a linear combination of the predictor variables</em>.</p>
<p>For logistic regression, <span class="math inline">\(f\)</span> is the logistic function: <span class="math display">\[
f(z) = \frac{e^z}{1+e^z} = \frac{1}{1 + e^{-z}}
\]</span></p>
<p>Therefore, the bivariate logistic regression is given by: <span class="math display">\[
P(Y = 1|X) = \frac{1}{1 + e^{-(\beta_0 + \beta_1 X)}}
\]</span></p>
<p>Note that <span class="math inline">\(\beta_0\)</span> here is akin to the intercept in our standard linear regression.</p>
</div>
<div id="bumpus-house-sparrow-data-set" class="section level3">
<h3><span class="header-section-number">15.6.3</span> Bumpus house sparrow data set</h3>
<p>Bumpus (1898) described a sample of house sparrows which he collected after a very severe storm. The sample included 136 birds, sixty four of which perished during the storm. Also included in his description were a variety of morphological measurements on the birds and information about their sex and age (for male birds). This data set has become a benchmark in the evolutionary biology literature for demonstrating methods for analyzing natural selection.</p>
<ul>
<li>Bumpus, H. C. 1898. The elimination of the unfit as illustrated by the introduced sparrow, Passer domesticus. (A fourth contribution to the study of variation.) <em>Biol. Lectures: Woods Hole Marine Biological Laboratory</em>, 209–225.</li>
</ul>
<p>The bumpus data set is available from the class website as a tab-delimited file <a href="https://github.com/bio304-class/bio304-fall2017/raw/master/datasets/bumpus-data.txt">bumpus-data.txt</a>.</p>
</div>
<div id="predicting-survival-as-a-function-of-body-weight" class="section level3">
<h3><span class="header-section-number">15.6.4</span> Predicting survival as a function of body weight</h3>
<p>Let’s first carry out a logistic regression predicting survival as a function of body weight for the Bumpus bird data set we looked at previously.</p>
<p>First we’ll load the data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bumpus &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;https://raw.githubusercontent.com/bio304-class/bio304-fall2017/master/datasets/bumpus-data.txt&quot;</span>)
Parsed with column specification<span class="op">:</span>
<span class="kw">cols</span>(
  <span class="st">`</span><span class="dt">line number</span><span class="st">`</span> =<span class="st"> </span><span class="kw">col_integer</span>(),
  <span class="dt">sex =</span> <span class="kw">col_character</span>(),
  <span class="dt">age.a_adult.y_young =</span> <span class="kw">col_character</span>(),
  <span class="dt">survived =</span> <span class="kw">col_logical</span>(),
  <span class="dt">total.length.mm =</span> <span class="kw">col_integer</span>(),
  <span class="dt">alar.extent.mm =</span> <span class="kw">col_integer</span>(),
  <span class="dt">weight.g =</span> <span class="kw">col_double</span>(),
  <span class="dt">length.beak.and.head =</span> <span class="kw">col_double</span>(),
  <span class="dt">length.humerus.in =</span> <span class="kw">col_double</span>(),
  <span class="dt">length.femur.in =</span> <span class="kw">col_double</span>(),
  <span class="dt">length.tibiotarsus.in =</span> <span class="kw">col_double</span>(),
  <span class="dt">width.skull.in =</span> <span class="kw">col_double</span>(),
  <span class="dt">length.sternal.keel.in =</span> <span class="kw">col_double</span>()
)</code></pre></div>
</div>
<div id="fitting-the-logistic-regression" class="section level3">
<h3><span class="header-section-number">15.6.5</span> Fitting the logistic regression</h3>
<p>Now let’s fit the logistic regression model and visualize the regression using ggplot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># convert TRUE/FALSE values to 1/0 values, to satisfy geom_smooth</span>
bumpus<span class="op">$</span>survived &lt;-<span class="st"> </span><span class="kw">as.integer</span>(bumpus<span class="op">$</span>survived) 

<span class="co"># fit the model</span>
fit.survival &lt;-<span class="st"> </span><span class="kw">glm</span>(survived <span class="op">~</span><span class="st"> </span>weight.g, <span class="dt">family =</span> binomial, 
                    <span class="dt">data =</span> bumpus)

<span class="co"># draw the regression plot</span>
<span class="kw">ggplot</span>(bumpus, <span class="kw">aes</span>(<span class="dt">x=</span>weight.g, <span class="dt">y=</span>survived)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_jitter</span>(<span class="dt">width =</span> <span class="dv">0</span>, <span class="dt">height =</span> <span class="fl">0.1</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_smooth</span>(<span class="dt">method=</span><span class="st">&quot;glm&quot;</span>, <span class="dt">method.args =</span> <span class="kw">list</span>(<span class="dt">family=</span><span class="st">&quot;binomial&quot;</span>), <span class="dt">se=</span><span class="ot">FALSE</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Weight (g)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Prob. Survival&quot;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-436-1.png" width="576" /></p>
<p>The logistic regression suggests that larger birds were <em>less likely</em> to survive the storm. Let’s look at a summary of the logistic regression model:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(fit.survival)

Call<span class="op">:</span>
<span class="kw">glm</span>(<span class="dt">formula =</span> survived <span class="op">~</span><span class="st"> </span>weight.g, <span class="dt">family =</span> binomial, <span class="dt">data =</span> bumpus)

Deviance Residuals<span class="op">:</span><span class="st"> </span>
<span class="st">    </span>Min       1Q   Median       3Q      Max  
<span class="op">-</span><span class="fl">1.6331</span>  <span class="op">-</span><span class="fl">1.1654</span>   <span class="fl">0.8579</span>   <span class="fl">1.0791</span>   <span class="fl">1.4626</span>  

Coefficients<span class="op">:</span>
<span class="st">            </span>Estimate Std. Error z value <span class="kw">Pr</span>(<span class="op">&gt;</span><span class="er">|</span>z<span class="op">|</span>)  
(Intercept)   <span class="fl">8.0456</span>     <span class="fl">3.2516</span>   <span class="fl">2.474</span>   <span class="fl">0.0133</span> <span class="op">*</span>
weight.g     <span class="op">-</span><span class="fl">0.3105</span>     <span class="fl">0.1272</span>  <span class="op">-</span><span class="fl">2.441</span>   <span class="fl">0.0146</span> <span class="op">*</span>
<span class="op">---</span>
Signif. codes<span class="op">:</span><span class="st">  </span><span class="dv">0</span> <span class="st">&#39;***&#39;</span> <span class="fl">0.001</span> <span class="st">&#39;**&#39;</span> <span class="fl">0.01</span> <span class="st">&#39;*&#39;</span> <span class="fl">0.05</span> <span class="st">&#39;.&#39;</span> <span class="fl">0.1</span> <span class="st">&#39; &#39;</span> <span class="dv">1</span>

(Dispersion parameter <span class="cf">for</span> binomial family taken to be <span class="dv">1</span>)

    Null deviance<span class="op">:</span><span class="st"> </span><span class="fl">188.07</span>  on <span class="dv">135</span>  degrees of freedom
Residual deviance<span class="op">:</span><span class="st"> </span><span class="fl">181.58</span>  on <span class="dv">134</span>  degrees of freedom
AIC<span class="op">:</span><span class="st"> </span><span class="fl">185.58</span>

Number of Fisher Scoring iterations<span class="op">:</span><span class="st"> </span><span class="dv">4</span></code></pre></div>
<p>We’re most interested in the estimated intercept and regression coefficients.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit.survival<span class="op">$</span>coefficients
(Intercept)    weight.g 
  <span class="fl">8.0455585</span>  <span class="op">-</span><span class="fl">0.3105315</span> </code></pre></div>
</div>
<div id="defining-an-appropriate-function-for-boot" class="section level3">
<h3><span class="header-section-number">15.6.6</span> Defining an appropriate function for <code>boot</code></h3>
<p>The <code>boot()</code> function defined in the <code>boot</code> package requires at least three arguments: <code>boot(data, statistic, R,...)</code>.</p>
<ol style="list-style-type: decimal">
<li><code>data</code> is the data frame, matrix or vector you want to resample from</li>
<li><code>statistic</code> is a function which when applied to data returns a vector containing the statistic of interest;</li>
<li><code>R</code> is the number of bootstrap replicates to generate.</li>
</ol>
<p>The function passed as the <code>statistic</code> argument to <code>boot</code> must take at least two arguments – the first is the original data, and the second is a vector of indices defining the observations in the bootstrap sample.</p>
<p>So to do bootstrapping of the logistic regression model we have to define a suitable “wrapper function” that will carry out the logistic regression on each bootstrap sample and return the coefficients that we want. The code block that follow illustrates this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(boot)

logistic.reg.coeffs &lt;-<span class="st"> </span><span class="cf">function</span>(x, indices) {
  fit.model &lt;-<span class="st"> </span><span class="kw">glm</span>(survived <span class="op">~</span><span class="st"> </span>weight.g, <span class="dt">family =</span> binomial, x[indices,])
  reg.b0 &lt;-<span class="st"> </span>fit.model<span class="op">$</span>coefficients[[<span class="dv">1</span>]]  <span class="co"># intercept</span>
  reg.b1 &lt;-<span class="st"> </span>fit.model<span class="op">$</span>coefficients[[<span class="dv">2</span>]]  <span class="co"># regression coefficient</span>
  <span class="kw">return</span>(<span class="kw">c</span>(reg.b0, reg.b1))
}</code></pre></div>
<p>Having defined this function, we can carry out the bootstrap as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># generate 500 bootstrap replicates</span>
nreps &lt;-<span class="st"> </span><span class="dv">500</span>
reg.boot &lt;-<span class="st"> </span><span class="kw">boot</span>(bumpus, logistic.reg.coeffs, nreps)</code></pre></div>
<p>The object returned by the <code>boot</code> function has various components. First, let’s look at the bootstrap replicates of our statistic(s) of interest, which are stored in a matrix called <code>t</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># the first column of t corresponds to the intercept</span>
<span class="co"># the second column to the coefficient w/respect to the explanatory</span>
<span class="co"># variable.</span>
<span class="kw">head</span>(reg.boot<span class="op">$</span>t)
         [,<span class="dv">1</span>]       [,<span class="dv">2</span>]
[<span class="dv">1</span>,] <span class="fl">8.669596</span> <span class="op">-</span><span class="fl">0.3460843</span>
[<span class="dv">2</span>,] <span class="fl">6.868043</span> <span class="op">-</span><span class="fl">0.2724601</span>
[<span class="dv">3</span>,] <span class="fl">6.721286</span> <span class="op">-</span><span class="fl">0.2632590</span>
[<span class="dv">4</span>,] <span class="fl">5.428487</span> <span class="op">-</span><span class="fl">0.2139940</span>
[<span class="dv">5</span>,] <span class="fl">9.621277</span> <span class="op">-</span><span class="fl">0.3587585</span>
[<span class="dv">6</span>,] <span class="fl">8.132526</span> <span class="op">-</span><span class="fl">0.3169114</span></code></pre></div>
<p>We can look at histograms of the bootstrap estimates of the intercept and regression coefficient. First the intercept:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">quickplot</span>(reg.boot<span class="op">$</span>t[,<span class="dv">1</span>], <span class="dt">bins =</span> <span class="dv">25</span>, <span class="dt">xlab=</span><span class="st">&quot;Regression Intercept&quot;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-442-1.png" width="576" /></p>
<p>And now the boostrap estimate of the regression coefficient:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">quickplot</span>(reg.boot<span class="op">$</span>t[,<span class="dv">2</span>], <span class="dt">bins =</span> <span class="dv">25</span>, <span class="dt">xlab=</span><span class="st">&quot;Regression Coefficient&quot;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-443-1.png" width="576" /></p>
</div>
<div id="calculating-bootstrap-confidence-intervals" class="section level3">
<h3><span class="header-section-number">15.6.7</span> Calculating bootstrap confidence intervals</h3>
<p>The <code>boot</code> package makes it easy to calculate confidence intervals, using the output of the <code>boot::boot()</code> function. The <code>boot::boot.ci()</code> handles the underlying calculations for us. To use <code>boot.ci()</code> we pass the <code>boot</code> object returned by the</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">boot.ci</span>(reg.boot, <span class="dt">index =</span> <span class="dv">1</span>, <span class="dt">conf =</span> <span class="fl">0.95</span>)
BOOTSTRAP CONFIDENCE INTERVAL CALCULATIONS
Based on <span class="dv">500</span> bootstrap replicates

CALL <span class="op">:</span><span class="st"> </span>
<span class="kw">boot.ci</span>(<span class="dt">boot.out =</span> reg.boot, <span class="dt">conf =</span> <span class="fl">0.95</span>, <span class="dt">index =</span> <span class="dv">1</span>)

Intervals <span class="op">:</span><span class="st"> </span>
Level      Normal              Basic         
<span class="dv">95</span>%   ( <span class="fl">1.378</span>, <span class="fl">14.564</span> )   ( <span class="fl">1.359</span>, <span class="fl">14.759</span> )  

Level     Percentile            BCa          
<span class="dv">95</span>%   ( <span class="fl">1.332</span>, <span class="fl">14.733</span> )   ( <span class="fl">1.311</span>, <span class="fl">14.699</span> )  
Calculations and Intervals on Original Scale</code></pre></div>
<p>The <code>boot.ci()</code> function returns information on Bootstrap CI’s calculated using four different approaches. The “Basic” CIs are based on using bootstrap standard errors and the t-distribution to estimate CIs. The “Normal” CIs are based on a normal approximation of the bootstrap distribution. The “Percentile” CIs are based on simple percentile intervals of the bootstrap distribution (as we did earlier), and the “BCa” CIs are based on an approach called “bias corrected and accelerated” (we won’t go into the details about this).</p>
<p>Which one of these CIs should you use? The answer unfortunately depends on the class of problem, sample sizes, and the form of the underlying distribution, etc. In general, my recommendation is to use bootstrap CIs (similarly Jackknife CIs) as “guides” about the nature of uncertainty in statistical estimates, not hard and fast rules. For example, when the original sample size is small, bootstrap CIs often tend to be too narrow for the desired level of confidence, and so it’s wise to consider that the CIs are “at least this wide”, and interpret your results accordingly.</p>
</div>
<div id="visualizing-bootstrap-confidence-intervals" class="section level3">
<h3><span class="header-section-number">15.6.8</span> Visualizing bootstrap confidence intervals</h3>
<p>One of the reasons for calculating confidence intervals is to provide insight into the range of plausible parameters associated with a statistical estimate or model. This suggests when we fit a model that we should really think of a cloud of other plausible models, representing the uncertainty associated with sampling. Let’s see how we can illustrate that uncertainty by visualizing 95% confidence intervals for our logistic regression.</p>
<p>First, we’ll write a helper function that given a range of set of points “x” (the predictor variable) and the coefficients of a logistic regression, yields the predicted values of “y” (the dependent variable):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">predicted.y &lt;-<span class="st"> </span><span class="cf">function</span>(x, coeffs){
  <span class="fl">1.0</span><span class="op">/</span>(<span class="fl">1.0</span> <span class="op">+</span><span class="st"> </span><span class="kw">exp</span>(<span class="op">-</span>(coeffs[<span class="dv">1</span>] <span class="op">+</span><span class="st"> </span>coeffs[<span class="dv">2</span>]<span class="op">*</span>x)))
}</code></pre></div>
<p>For our particular example, the range of x-values we’ll make predictions over is the roughly the range of the observed values of the sparrow’s body mass:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">range</span>(bumpus<span class="op">$</span>weight.g)
[<span class="dv">1</span>] <span class="fl">22.6</span> <span class="fl">31.0</span>

<span class="co"># setup x-values over which to make predictions</span>
nx &lt;-<span class="st"> </span><span class="dv">200</span>
x &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dv">22</span>, <span class="dv">32</span>, <span class="dt">length.out =</span> nx)</code></pre></div>
<p>We then create an empty matrix to hold the predicted values for each bootstrap sample, and then iterate over the bootstrap samples calculating the predictions for each:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># create empty matrix to hold model predictions for each bootstrap sample</span>
predicted.mtx &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="dt">nrow=</span>nreps, <span class="dt">ncol =</span> nx)

<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>nreps) {
  predicted.mtx[i,] &lt;-<span class="st"> </span><span class="kw">predicted.y</span>(x, reg.boot<span class="op">$</span>t[i,])
}

<span class="co"># cast the matrix of predictions as a data frame</span>
predicted.mtx.df &lt;-<span class="st"> </span><span class="kw">as_data_frame</span>(predicted.mtx)</code></pre></div>
<p>For reference, we also create the prediction for the original sample we have in had. Here we can use the built-in <code>predict()</code> function along with the original logistic regression fit object:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sample.prediction &lt;-<span class="st"> </span><span class="kw">predict</span>(fit.survival, 
                             <span class="kw">data.frame</span>(<span class="dt">weight.g =</span> x), 
                             <span class="dt">type =</span> <span class="st">&quot;response&quot;</span>)</code></pre></div>
<p>Finally we create a couple of convenience functions to calculate the 2.5% and 97.5% percentile points given a set of values:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">quantile.<span class="dv">975</span> &lt;-<span class="st"> </span><span class="cf">function</span>(x){ <span class="kw">quantile</span>(x, <span class="fl">0.975</span>) }
quantile.<span class="dv">025</span> &lt;-<span class="st"> </span><span class="cf">function</span>(x){ <span class="kw">quantile</span>(x, <span class="fl">0.025</span>) }</code></pre></div>
<p>With the various pieces in hand, we’re now ready to create a visual representation of the 95% bootstrap percentile CIs for the logistic regression:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>() <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_jitter</span>(<span class="kw">aes</span>(<span class="dt">x=</span>bumpus<span class="op">$</span>weight.g, <span class="dt">y=</span>bumpus<span class="op">$</span>survived), 
              <span class="dt">width =</span> <span class="dv">0</span>, <span class="dt">height =</span> <span class="fl">0.1</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="dt">x =</span> x, <span class="dt">y =</span> sample.prediction), <span class="dt">color=</span><span class="st">&#39;red&#39;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="dt">x =</span> x, <span class="dt">y =</span> <span class="kw">map_dbl</span>(predicted.mtx.df, quantile.<span class="dv">975</span>))) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="dt">x =</span> x, <span class="dt">y =</span> <span class="kw">map_dbl</span>(predicted.mtx.df, quantile.<span class="dv">025</span>))) <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Weight (g)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Prob. Survival&quot;</span>,
       <span class="dt">title=</span><span class="st">&quot;Bumpus Survival Data</span><span class="ch">\n</span><span class="st">Logistic Regression and Bootstrap 95% CI&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">lims</span>(<span class="dt">x =</span> <span class="kw">c</span>(<span class="dv">22</span>,<span class="dv">32</span>), <span class="dt">y =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>))</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-450-1.png" width="576" /></p>
<p>In the next plot we directly compare our bootstrap 95% CIs to the confidence intervals calculated by <code>geom_smooth</code>, which uses an asymptotic approximation to calculate confidence intervals for the logistic regression.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># draw the regression plot</span>
<span class="kw">ggplot</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_jitter</span>(<span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x=</span>weight.g, <span class="dt">y=</span>survived), <span class="dt">data =</span> bumpus, <span class="dt">width =</span> <span class="dv">0</span>, <span class="dt">height =</span> <span class="fl">0.1</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_smooth</span>(<span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x=</span>weight.g, <span class="dt">y=</span>survived), <span class="dt">data =</span> bumpus, <span class="dt">method=</span><span class="st">&quot;glm&quot;</span>, <span class="dt">method.args =</span> <span class="kw">list</span>(<span class="dt">family=</span><span class="st">&quot;binomial&quot;</span>), <span class="dt">se=</span><span class="ot">TRUE</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_line</span>(<span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> x, <span class="dt">y =</span> <span class="kw">map_dbl</span>(predicted.mtx.df, quantile.<span class="dv">975</span>)), <span class="dt">color=</span><span class="st">&#39;red&#39;</span>, <span class="dt">linetype=</span><span class="st">&#39;dashed&#39;</span>, <span class="dt">data =</span> <span class="ot">NULL</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_line</span>(<span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">x =</span> x, <span class="dt">y =</span> <span class="kw">map_dbl</span>(predicted.mtx.df, quantile.<span class="dv">025</span>)), <span class="dt">color=</span><span class="st">&#39;red&#39;</span>, <span class="dt">linetype=</span><span class="st">&#39;dashed&#39;</span>, <span class="dt">data =</span> <span class="ot">NULL</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Weight (g)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Prob. Survival&quot;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-451-1.png" width="576" /></p>
<p>Our plot indicates that the asymptotic estimates of the logistic regression CIs and the bootstrap estimates of the CIs are very similar.</p>

</div>
</div>
</div>






            </section>

          </div>
        </div>
      </div>
<a href="simulating-confidence-intervals.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
