<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Biology 723: Statistical Computing for Biologists</title>
  <meta name="description" content="Hands-on materials for Bio 723 course at Duke University.">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Biology 723: Statistical Computing for Biologists" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Biology 723: Statistical Computing for Biologists" />
  
  <meta name="twitter:description" content="Hands-on materials for Bio 723 course at Duke University." />
  

<meta name="author" content="Paul M. Magwene">


<meta name="date" content="2018-03-28">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="principal-components-analysis.html">
<link rel="next" href="simulating-confidence-intervals.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#how-to-use-these-lecture-notes"><i class="fa fa-check"></i><b>1.1</b> How to use these lecture notes</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html"><i class="fa fa-check"></i><b>2</b> Getting Started with R</a><ul>
<li class="chapter" data-level="2.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-r"><i class="fa fa-check"></i><b>2.1</b> What is R?</a></li>
<li class="chapter" data-level="2.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#what-is-rstudio"><i class="fa fa-check"></i><b>2.2</b> What is RStudio?</a></li>
<li class="chapter" data-level="2.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#entering-commands-in-the-console"><i class="fa fa-check"></i><b>2.3</b> Entering commands in the console</a></li>
<li class="chapter" data-level="2.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#comments"><i class="fa fa-check"></i><b>2.4</b> Comments</a></li>
<li class="chapter" data-level="2.5" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#using-r-as-a-calculator"><i class="fa fa-check"></i><b>2.5</b> Using R as a Calculator</a><ul>
<li class="chapter" data-level="2.5.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#common-mathematical-functions"><i class="fa fa-check"></i><b>2.5.1</b> Common mathematical functions</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#numeric-data-types-in-r"><i class="fa fa-check"></i><b>2.6</b> Numeric data types in R</a></li>
<li class="chapter" data-level="2.7" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#variable-assignment-in-r"><i class="fa fa-check"></i><b>2.7</b> Variable assignment in R</a><ul>
<li class="chapter" data-level="2.7.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#valid-variable-names"><i class="fa fa-check"></i><b>2.7.1</b> Valid variable names</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#working-with-vectors-in-r"><i class="fa fa-check"></i><b>2.8</b> Working with Vectors in R</a><ul>
<li class="chapter" data-level="2.8.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#vector-arithmetic"><i class="fa fa-check"></i><b>2.8.1</b> Vector Arithmetic</a></li>
<li class="chapter" data-level="2.8.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#vector-recycling"><i class="fa fa-check"></i><b>2.8.2</b> Vector recycling</a></li>
<li class="chapter" data-level="2.8.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#simple-statistical-functions-for-numeric-vectors"><i class="fa fa-check"></i><b>2.8.3</b> Simple statistical functions for numeric vectors</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#the-r-help-system"><i class="fa fa-check"></i><b>2.9</b> The R Help System</a><ul>
<li class="chapter" data-level="2.9.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#getting-help-from-the-console"><i class="fa fa-check"></i><b>2.9.1</b> Getting help from the console</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#packages"><i class="fa fa-check"></i><b>2.10</b> Packages</a><ul>
<li class="chapter" data-level="2.10.1" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-console"><i class="fa fa-check"></i><b>2.10.1</b> Installing packages from the console</a></li>
<li class="chapter" data-level="2.10.2" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#install-the-tidyverse-package"><i class="fa fa-check"></i><b>2.10.2</b> Install the tidyverse package</a></li>
<li class="chapter" data-level="2.10.3" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#installing-packages-from-the-rstudio-dialog"><i class="fa fa-check"></i><b>2.10.3</b> Installing packages from the RStudio dialog</a></li>
<li class="chapter" data-level="2.10.4" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#install-the-stringr-package"><i class="fa fa-check"></i><b>2.10.4</b> Install the stringr package</a></li>
<li class="chapter" data-level="2.10.5" data-path="getting-started-with-r.html"><a href="getting-started-with-r.html#loading-packages-with-the-library-function"><i class="fa fa-check"></i><b>2.10.5</b> Loading packages with the <code>library()</code> function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html"><i class="fa fa-check"></i><b>3</b> R Basics: Data types and data structures</a><ul>
<li class="chapter" data-level="3.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-boolean-values"><i class="fa fa-check"></i><b>3.1</b> Logical (Boolean) values</a><ul>
<li class="chapter" data-level="3.1.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#comparison-operators-applied-to-numeric-vectors"><i class="fa fa-check"></i><b>3.1.1</b> Comparison operators applied to numeric vectors</a></li>
<li class="chapter" data-level="3.1.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-vectors"><i class="fa fa-check"></i><b>3.1.2</b> Logical vectors</a></li>
<li class="chapter" data-level="3.1.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-operators"><i class="fa fa-check"></i><b>3.1.3</b> Logical operators</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#character-strings"><i class="fa fa-check"></i><b>3.2</b> Character strings</a><ul>
<li class="chapter" data-level="3.2.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#the-stringr-package"><i class="fa fa-check"></i><b>3.2.1</b> The <code>stringr</code> package</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#vectors"><i class="fa fa-check"></i><b>3.3</b> Vectors</a><ul>
<li class="chapter" data-level="3.3.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#indexing-vectors"><i class="fa fa-check"></i><b>3.3.1</b> Indexing Vectors</a></li>
<li class="chapter" data-level="3.3.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#combining-indexing-and-comparison-of-vectors"><i class="fa fa-check"></i><b>3.3.2</b> Combining Indexing and Comparison of Vectors</a></li>
<li class="chapter" data-level="3.3.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#vector-manipulation"><i class="fa fa-check"></i><b>3.3.3</b> Vector manipulation</a></li>
<li class="chapter" data-level="3.3.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#vectors-from-regular-sequences"><i class="fa fa-check"></i><b>3.3.4</b> Vectors from regular sequences</a></li>
<li class="chapter" data-level="3.3.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#additional-functions-for-working-with-vectors"><i class="fa fa-check"></i><b>3.3.5</b> Additional functions for working with vectors</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#lists"><i class="fa fa-check"></i><b>3.4</b> Lists</a><ul>
<li class="chapter" data-level="3.4.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#length-and-type-of-lists"><i class="fa fa-check"></i><b>3.4.1</b> Length and type of lists</a></li>
<li class="chapter" data-level="3.4.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#indexing-lists"><i class="fa fa-check"></i><b>3.4.2</b> Indexing lists</a></li>
<li class="chapter" data-level="3.4.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#naming-list-elements"><i class="fa fa-check"></i><b>3.4.3</b> Naming list elements</a></li>
<li class="chapter" data-level="3.4.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#the-operator"><i class="fa fa-check"></i><b>3.4.4</b> The <code>$</code> operator</a></li>
<li class="chapter" data-level="3.4.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#changing-and-adding-lists-items"><i class="fa fa-check"></i><b>3.4.5</b> Changing and adding lists items</a></li>
<li class="chapter" data-level="3.4.6" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#combining-lists"><i class="fa fa-check"></i><b>3.4.6</b> Combining lists</a></li>
<li class="chapter" data-level="3.4.7" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#converting-lists-to-vectors"><i class="fa fa-check"></i><b>3.4.7</b> Converting lists to vectors</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#data-frames"><i class="fa fa-check"></i><b>3.5</b> Data frames</a><ul>
<li class="chapter" data-level="3.5.1" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#creating-a-data-frame"><i class="fa fa-check"></i><b>3.5.1</b> Creating a data frame</a></li>
<li class="chapter" data-level="3.5.2" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#type-and-class-for-data-frames"><i class="fa fa-check"></i><b>3.5.2</b> Type and class for data frames</a></li>
<li class="chapter" data-level="3.5.3" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#length-and-dimension-for-data-frames"><i class="fa fa-check"></i><b>3.5.3</b> Length and dimension for data frames</a></li>
<li class="chapter" data-level="3.5.4" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#indexing-and-accessing-data-frames"><i class="fa fa-check"></i><b>3.5.4</b> Indexing and accessing data frames</a></li>
<li class="chapter" data-level="3.5.5" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#logical-indexing-of-data-frames"><i class="fa fa-check"></i><b>3.5.5</b> Logical indexing of data frames</a></li>
<li class="chapter" data-level="3.5.6" data-path="r-basics-data-types-and-data-structures.html"><a href="r-basics-data-types-and-data-structures.html#adding-columns-to-a-data-frame"><i class="fa fa-check"></i><b>3.5.6</b> Adding columns to a data frame</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html"><i class="fa fa-check"></i><b>4</b> R Markdown and R Notebooks</a><ul>
<li class="chapter" data-level="4.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#r-notebooks"><i class="fa fa-check"></i><b>4.1</b> R Notebooks</a></li>
<li class="chapter" data-level="4.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#creating-an-r-notebook"><i class="fa fa-check"></i><b>4.2</b> Creating an R Notebook</a></li>
<li class="chapter" data-level="4.3" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#the-default-r-notebook-template"><i class="fa fa-check"></i><b>4.3</b> The default R Notebook template</a></li>
<li class="chapter" data-level="4.4" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-and-non-code-blocks"><i class="fa fa-check"></i><b>4.4</b> Code and Non-code blocks</a><ul>
<li class="chapter" data-level="4.4.1" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#non-code-blocks"><i class="fa fa-check"></i><b>4.4.1</b> Non-code blocks</a></li>
<li class="chapter" data-level="4.4.2" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#code-blocks"><i class="fa fa-check"></i><b>4.4.2</b> Code blocks</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-a-code-chunk"><i class="fa fa-check"></i><b>4.5</b> Running a code chunk</a></li>
<li class="chapter" data-level="4.6" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#running-all-code-chunks-above"><i class="fa fa-check"></i><b>4.6</b> Running all code chunks above</a></li>
<li class="chapter" data-level="4.7" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#knitting-an-r-markdown-to-html"><i class="fa fa-check"></i><b>4.7</b> “Knitting an” R Markdown to HTML</a></li>
<li class="chapter" data-level="4.8" data-path="r-markdown-and-r-notebooks.html"><a href="r-markdown-and-r-notebooks.html#sharing-your-reproducible-r-notebook"><i class="fa fa-check"></i><b>4.8</b> Sharing your reproducible R Notebook</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html"><i class="fa fa-check"></i><b>5</b> Introduction to ggplot2</a><ul>
<li class="chapter" data-level="5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#loading-ggplot2"><i class="fa fa-check"></i><b>5.1</b> Loading ggplot2</a></li>
<li class="chapter" data-level="5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#example-data-set-andersons-iris-data"><i class="fa fa-check"></i><b>5.2</b> Example data set: Anderson’s Iris Data</a></li>
<li class="chapter" data-level="5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#template-for-single-layer-plots-in-ggplot2"><i class="fa fa-check"></i><b>5.3</b> Template for single layer plots in ggplot2</a></li>
<li class="chapter" data-level="5.4" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#an-aside-about-function-arguments"><i class="fa fa-check"></i><b>5.4</b> An aside about function arguments</a></li>
<li class="chapter" data-level="5.5" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#strip-plots"><i class="fa fa-check"></i><b>5.5</b> Strip plots</a><ul>
<li class="chapter" data-level="5.5.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#jittering-data"><i class="fa fa-check"></i><b>5.5.1</b> Jittering data</a></li>
<li class="chapter" data-level="5.5.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-categorical-information"><i class="fa fa-check"></i><b>5.5.2</b> Adding categorical information</a></li>
<li class="chapter" data-level="5.5.3" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#rotating-plot-coordinates"><i class="fa fa-check"></i><b>5.5.3</b> Rotating plot coordinates</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#histograms"><i class="fa fa-check"></i><b>5.6</b> Histograms</a><ul>
<li class="chapter" data-level="5.6.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#variations-on-histograms-when-considering-categorical-data"><i class="fa fa-check"></i><b>5.6.1</b> Variations on histograms when considering categorical data</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#faceting-to-depict-categorical-information"><i class="fa fa-check"></i><b>5.7</b> Faceting to depict categorical information</a></li>
<li class="chapter" data-level="5.8" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots"><i class="fa fa-check"></i><b>5.8</b> Density plots</a></li>
<li class="chapter" data-level="5.9" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-or-beanplot"><i class="fa fa-check"></i><b>5.9</b> Violin or Beanplot</a></li>
<li class="chapter" data-level="5.10" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplots"><i class="fa fa-check"></i><b>5.10</b> Boxplots</a></li>
<li class="chapter" data-level="5.11" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#building-complex-visualizations-with-layers"><i class="fa fa-check"></i><b>5.11</b> Building complex visualizations with layers</a></li>
<li class="chapter" data-level="5.12" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#useful-combination-plots"><i class="fa fa-check"></i><b>5.12</b> Useful combination plots</a><ul>
<li class="chapter" data-level="5.12.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#boxplot-plus-strip-plot"><i class="fa fa-check"></i><b>5.12.1</b> Boxplot plus strip plot</a></li>
<li class="chapter" data-level="5.12.2" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#setting-shared-aesthetics"><i class="fa fa-check"></i><b>5.12.2</b> Setting shared aesthetics</a></li>
</ul></li>
<li class="chapter" data-level="5.13" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot-layers-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>5.13</b> ggplot layers can be assigned to variables</a><ul>
<li class="chapter" data-level="5.13.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#violin-plot-plus-strip-plot"><i class="fa fa-check"></i><b>5.13.1</b> Violin plot plus strip plot</a></li>
</ul></li>
<li class="chapter" data-level="5.14" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#adding-titles-and-tweaking-axis-labels"><i class="fa fa-check"></i><b>5.14</b> Adding titles and tweaking axis labels</a></li>
<li class="chapter" data-level="5.15" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#ggplot2-themes"><i class="fa fa-check"></i><b>5.15</b> ggplot2 themes</a><ul>
<li class="chapter" data-level="5.15.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#further-customization-with-ggplot2theme"><i class="fa fa-check"></i><b>5.15.1</b> Further customization with <code>ggplot2::theme</code></a></li>
</ul></li>
<li class="chapter" data-level="5.16" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#other-aspects-of-ggplots-can-be-assigned-to-variables"><i class="fa fa-check"></i><b>5.16</b> Other aspects of ggplots can be assigned to variables</a></li>
<li class="chapter" data-level="5.17" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-plots"><i class="fa fa-check"></i><b>5.17</b> Bivariate plots</a><ul>
<li class="chapter" data-level="5.17.1" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#scatter-plots"><i class="fa fa-check"></i><b>5.17.1</b> Scatter plots</a></li>
</ul></li>
<li class="chapter" data-level="5.18" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#bivariate-density-plots"><i class="fa fa-check"></i><b>5.18</b> Bivariate density plots</a></li>
<li class="chapter" data-level="5.19" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#combining-scatter-plots-and-density-plots-with-categorical-information"><i class="fa fa-check"></i><b>5.19</b> Combining Scatter Plots and Density Plots with Categorical Information</a></li>
<li class="chapter" data-level="5.20" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#density-plots-with-fill"><i class="fa fa-check"></i><b>5.20</b> Density plots with fill</a></li>
<li class="chapter" data-level="5.21" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#d-bin-and-hex-plots"><i class="fa fa-check"></i><b>5.21</b> 2D bin and hex plots</a></li>
<li class="chapter" data-level="5.22" data-path="introduction-to-ggplot2.html"><a href="introduction-to-ggplot2.html#the-cowplot-package"><i class="fa fa-check"></i><b>5.22</b> The <code>cowplot</code> package</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html"><i class="fa fa-check"></i><b>6</b> Introduction to the <code>dplyr</code> package</a><ul>
<li class="chapter" data-level="6.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#libraries"><i class="fa fa-check"></i><b>6.1</b> Libraries</a></li>
<li class="chapter" data-level="6.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#reading-data-with-the-readr-package"><i class="fa fa-check"></i><b>6.2</b> Reading data with the <code>readr</code> package</a><ul>
<li class="chapter" data-level="6.2.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#example-data-nc-births"><i class="fa fa-check"></i><b>6.2.1</b> Example data: NC Births</a></li>
<li class="chapter" data-level="6.2.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#reading-excel-files"><i class="fa fa-check"></i><b>6.2.2</b> Reading Excel files</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#a-note-on-tibbles"><i class="fa fa-check"></i><b>6.3</b> A note on “tibbles”</a></li>
<li class="chapter" data-level="6.4" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#data-filtering-and-transformation-with-dplyr"><i class="fa fa-check"></i><b>6.4</b> Data filtering and transformation with <code>dplyr</code></a></li>
<li class="chapter" data-level="6.5" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#dplyrs-verbs"><i class="fa fa-check"></i><b>6.5</b> dplyr’s “verbs”</a><ul>
<li class="chapter" data-level="6.5.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#select"><i class="fa fa-check"></i><b>6.5.1</b> <code>select</code></a></li>
<li class="chapter" data-level="6.5.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#filter"><i class="fa fa-check"></i><b>6.5.2</b> <code>filter</code></a></li>
<li class="chapter" data-level="6.5.3" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#mutate"><i class="fa fa-check"></i><b>6.5.3</b> <code>mutate</code></a></li>
<li class="chapter" data-level="6.5.4" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#arrange"><i class="fa fa-check"></i><b>6.5.4</b> <code>arrange</code></a></li>
<li class="chapter" data-level="6.5.5" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#summarize"><i class="fa fa-check"></i><b>6.5.5</b> <code>summarize</code></a></li>
<li class="chapter" data-level="6.5.6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#group_by"><i class="fa fa-check"></i><b>6.5.6</b> <code>group_by</code></a></li>
<li class="chapter" data-level="6.5.7" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#combining-grouping-and-summarizing"><i class="fa fa-check"></i><b>6.5.7</b> Combining grouping and summarizing</a></li>
<li class="chapter" data-level="6.5.8" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#scoped-variants-of-mutate-and-summarize"><i class="fa fa-check"></i><b>6.5.8</b> Scoped variants of <code>mutate</code> and <code>summarize</code></a></li>
<li class="chapter" data-level="6.5.9" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#combining-summarize-with-grouping-aesthetics-in-ggplot2"><i class="fa fa-check"></i><b>6.5.9</b> Combining summarize with grouping aesthetics in <code>ggplot2</code></a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#pipes"><i class="fa fa-check"></i><b>6.6</b> Pipes</a><ul>
<li class="chapter" data-level="6.6.1" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#install-and-load-magrittr"><i class="fa fa-check"></i><b>6.6.1</b> Install and load <code>magrittr</code></a></li>
<li class="chapter" data-level="6.6.2" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-basic-pipe-operator"><i class="fa fa-check"></i><b>6.6.2</b> The basic pipe operator</a></li>
<li class="chapter" data-level="6.6.3" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#an-example-without-pipes"><i class="fa fa-check"></i><b>6.6.3</b> An example without pipes</a></li>
<li class="chapter" data-level="6.6.4" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-same-example-using-pipes"><i class="fa fa-check"></i><b>6.6.4</b> The same example using pipes</a></li>
<li class="chapter" data-level="6.6.5" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#assigning-the-output-of-a-statement-involving-pipes-to-a-variable"><i class="fa fa-check"></i><b>6.6.5</b> Assigning the output of a statement involving pipes to a variable</a></li>
<li class="chapter" data-level="6.6.6" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#compound-assignment-pipe-operator"><i class="fa fa-check"></i><b>6.6.6</b> Compound assignment pipe operator</a></li>
<li class="chapter" data-level="6.6.7" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-dot-operator-with-pipes"><i class="fa fa-check"></i><b>6.6.7</b> The dot operator with pipes</a></li>
<li class="chapter" data-level="6.6.8" data-path="introduction-to-the-dplyr-package.html"><a href="introduction-to-the-dplyr-package.html#the-exposition-pipe-operator"><i class="fa fa-check"></i><b>6.6.8</b> The exposition pipe operator</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="data-wrangling.html"><a href="data-wrangling.html"><i class="fa fa-check"></i><b>7</b> Data wrangling</a><ul>
<li class="chapter" data-level="7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#libraries-1"><i class="fa fa-check"></i><b>7.1</b> Libraries</a></li>
<li class="chapter" data-level="7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#data"><i class="fa fa-check"></i><b>7.2</b> Data</a></li>
<li class="chapter" data-level="7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#renaming-data-frame-columms"><i class="fa fa-check"></i><b>7.3</b> Renaming data frame columms</a></li>
<li class="chapter" data-level="7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-unneeded-columns"><i class="fa fa-check"></i><b>7.4</b> Dropping unneeded columns</a><ul>
<li class="chapter" data-level="7.4.1" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-all-empty-columns"><i class="fa fa-check"></i><b>7.4.1</b> Finding all empty columns</a></li>
<li class="chapter" data-level="7.4.2" data-path="data-wrangling.html"><a href="data-wrangling.html#dropping-columns-by-matching-names"><i class="fa fa-check"></i><b>7.4.2</b> Dropping columns by matching names</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="data-wrangling.html"><a href="data-wrangling.html#reshaping-data-with-tidyr"><i class="fa fa-check"></i><b>7.5</b> Reshaping data with tidyr</a><ul>
<li class="chapter" data-level="7.5.1" data-path="data-wrangling.html"><a href="data-wrangling.html#wide-to-long-conversions-using-tidyrgather"><i class="fa fa-check"></i><b>7.5.1</b> Wide to long conversions using <code>tidyr::gather</code></a></li>
<li class="chapter" data-level="7.5.2" data-path="data-wrangling.html"><a href="data-wrangling.html#extracting-information-from-combined-variables-using-tidyrextract"><i class="fa fa-check"></i><b>7.5.2</b> Extracting information from combined variables using <code>tidyr::extract</code></a></li>
<li class="chapter" data-level="7.5.3" data-path="data-wrangling.html"><a href="data-wrangling.html#subsetting-rows"><i class="fa fa-check"></i><b>7.5.3</b> Subsetting rows</a></li>
<li class="chapter" data-level="7.5.4" data-path="data-wrangling.html"><a href="data-wrangling.html#splitting-columns"><i class="fa fa-check"></i><b>7.5.4</b> Splitting columns</a></li>
<li class="chapter" data-level="7.5.5" data-path="data-wrangling.html"><a href="data-wrangling.html#combining-data-frame-rows"><i class="fa fa-check"></i><b>7.5.5</b> Combining data frame rows</a></li>
<li class="chapter" data-level="7.5.6" data-path="data-wrangling.html"><a href="data-wrangling.html#sorting-data-frame-rows"><i class="fa fa-check"></i><b>7.5.6</b> Sorting data frame rows</a></li>
<li class="chapter" data-level="7.5.7" data-path="data-wrangling.html"><a href="data-wrangling.html#collapsing-columns-using-tidyrunite"><i class="fa fa-check"></i><b>7.5.7</b> Collapsing columns using <code>tidyr::unite</code></a></li>
<li class="chapter" data-level="7.5.8" data-path="data-wrangling.html"><a href="data-wrangling.html#splitting-columns-using-tidyrseparate"><i class="fa fa-check"></i><b>7.5.8</b> Splitting columns using <code>tidyr::separate</code></a></li>
<li class="chapter" data-level="7.5.9" data-path="data-wrangling.html"><a href="data-wrangling.html#long-to-wide-conversion-using-tidyrspread"><i class="fa fa-check"></i><b>7.5.9</b> Long-to-wide conversion using <code>tidyr::spread</code></a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="data-wrangling.html"><a href="data-wrangling.html#using-your-tidy-data"><i class="fa fa-check"></i><b>7.6</b> Using your tidy data</a><ul>
<li class="chapter" data-level="7.6.1" data-path="data-wrangling.html"><a href="data-wrangling.html#visualizing-gene-expression-time-series"><i class="fa fa-check"></i><b>7.6.1</b> Visualizing gene expression time series</a></li>
<li class="chapter" data-level="7.6.2" data-path="data-wrangling.html"><a href="data-wrangling.html#finding-the-most-variable-genes"><i class="fa fa-check"></i><b>7.6.2</b> Finding the most variable genes</a></li>
<li class="chapter" data-level="7.6.3" data-path="data-wrangling.html"><a href="data-wrangling.html#exploring-bivariate-relationships-using-wide-data"><i class="fa fa-check"></i><b>7.6.3</b> Exploring bivariate relationships using “wide” data</a></li>
<li class="chapter" data-level="7.6.4" data-path="data-wrangling.html"><a href="data-wrangling.html#calculating-correlations-between-genes"><i class="fa fa-check"></i><b>7.6.4</b> Calculating correlations between genes</a></li>
<li class="chapter" data-level="7.6.5" data-path="data-wrangling.html"><a href="data-wrangling.html#adding-new-columns-and-combining-filtered-data-frames"><i class="fa fa-check"></i><b>7.6.5</b> Adding new columns and combining filtered data frames</a></li>
</ul></li>
<li class="chapter" data-level="7.7" data-path="data-wrangling.html"><a href="data-wrangling.html#heat-maps"><i class="fa fa-check"></i><b>7.7</b> Heat maps</a><ul>
<li class="chapter" data-level="7.7.1" data-path="data-wrangling.html"><a href="data-wrangling.html#rcolorbrewer"><i class="fa fa-check"></i><b>7.7.1</b> RColorBrewer</a></li>
<li class="chapter" data-level="7.7.2" data-path="data-wrangling.html"><a href="data-wrangling.html#creating-a-heat-map-using-geom_tile"><i class="fa fa-check"></i><b>7.7.2</b> Creating a heat map using <code>geom_tile</code></a></li>
<li class="chapter" data-level="7.7.3" data-path="data-wrangling.html"><a href="data-wrangling.html#a-fancy-figure"><i class="fa fa-check"></i><b>7.7.3</b> A “fancy” figure</a></li>
<li class="chapter" data-level="7.7.4" data-path="data-wrangling.html"><a href="data-wrangling.html#sorting-and-visualizing-genes-by-time-of-maximum-expression"><i class="fa fa-check"></i><b>7.7.4</b> Sorting and visualizing genes by time of maximum expression</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html"><i class="fa fa-check"></i><b>8</b> Vectors and Functions</a><ul>
<li class="chapter" data-level="8.1" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#libraries-2"><i class="fa fa-check"></i><b>8.1</b> Libraries</a></li>
<li class="chapter" data-level="8.2" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#vector-mathematics-in-r"><i class="fa fa-check"></i><b>8.2</b> Vector Mathematics in R</a></li>
<li class="chapter" data-level="8.3" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#writing-your-own-functions-in-r"><i class="fa fa-check"></i><b>8.3</b> Writing Your Own Functions in R</a></li>
<li class="chapter" data-level="8.4" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#function-arguments"><i class="fa fa-check"></i><b>8.4</b> Function arguments</a></li>
<li class="chapter" data-level="8.5" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#writing-functions-with-optional-arguments"><i class="fa fa-check"></i><b>8.5</b> Writing functions with optional arguments</a></li>
<li class="chapter" data-level="8.6" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#putting-r-functions-in-scripts"><i class="fa fa-check"></i><b>8.6</b> Putting R functions in Scripts</a></li>
<li class="chapter" data-level="8.7" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#simple-statistics-in-vector-form"><i class="fa fa-check"></i><b>8.7</b> Simple statistics in vector form</a><ul>
<li class="chapter" data-level="8.7.1" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#mean"><i class="fa fa-check"></i><b>8.7.1</b> Mean</a></li>
<li class="chapter" data-level="8.7.2" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#variance-and-standard-deviation"><i class="fa fa-check"></i><b>8.7.2</b> Variance and standard deviation</a></li>
<li class="chapter" data-level="8.7.3" data-path="vectors-and-functions.html"><a href="vectors-and-functions.html#covariance-and-correlation"><i class="fa fa-check"></i><b>8.7.3</b> Covariance and correlation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html"><i class="fa fa-check"></i><b>9</b> Bivariate Linear Regression</a><ul>
<li class="chapter" data-level="9.1" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#regression-terminology"><i class="fa fa-check"></i><b>9.1</b> Regression terminology</a></li>
<li class="chapter" data-level="9.2" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#libraries-3"><i class="fa fa-check"></i><b>9.2</b> Libraries</a></li>
<li class="chapter" data-level="9.3" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#illustrating-linear-regression-with-simulated-data"><i class="fa fa-check"></i><b>9.3</b> Illustrating linear regression with simulated data</a></li>
<li class="chapter" data-level="9.4" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#specifying-regression-models-in-r"><i class="fa fa-check"></i><b>9.4</b> Specifying Regression Models in R</a><ul>
<li class="chapter" data-level="9.4.1" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#fitted-values"><i class="fa fa-check"></i><b>9.4.1</b> Fitted values</a></li>
<li class="chapter" data-level="9.4.2" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#getting-the-model-coefficients"><i class="fa fa-check"></i><b>9.4.2</b> Getting the model coefficients</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#residuals"><i class="fa fa-check"></i><b>9.5</b> Residuals</a></li>
<li class="chapter" data-level="9.6" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#regression-as-sum-of-squares-decomposition"><i class="fa fa-check"></i><b>9.6</b> Regression as sum-of-squares decomposition</a></li>
<li class="chapter" data-level="9.7" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#variance-explained-by-a-regression-model"><i class="fa fa-check"></i><b>9.7</b> Variance “explained” by a regression model</a></li>
<li class="chapter" data-level="9.8" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broom-a-library-for-converting-model-results-into-data-frames"><i class="fa fa-check"></i><b>9.8</b> Broom: a library for converting model results into data frames</a><ul>
<li class="chapter" data-level="9.8.1" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broomtidy"><i class="fa fa-check"></i><b>9.8.1</b> <code>broom::tidy</code></a></li>
<li class="chapter" data-level="9.8.2" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broomaugment"><i class="fa fa-check"></i><b>9.8.2</b> <code>broom::augment</code></a></li>
<li class="chapter" data-level="9.8.3" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#broomglance"><i class="fa fa-check"></i><b>9.8.3</b> <code>broom::glance</code></a></li>
</ul></li>
<li class="chapter" data-level="9.9" data-path="bivariate-linear-regression.html"><a href="bivariate-linear-regression.html#bivariate-linear-regression-on-a-real-world-data-set"><i class="fa fa-check"></i><b>9.9</b> Bivariate linear regression on a real world data set</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="matrices-in-r.html"><a href="matrices-in-r.html"><i class="fa fa-check"></i><b>10</b> Matrices in R</a><ul>
<li class="chapter" data-level="10.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#creating-matrices-in-r"><i class="fa fa-check"></i><b>10.1</b> Creating matrices in R</a></li>
<li class="chapter" data-level="10.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-arithmetic-operations-in-r"><i class="fa fa-check"></i><b>10.2</b> Matrix arithmetic operations in R</a></li>
<li class="chapter" data-level="10.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#descriptive-statistics-as-matrix-functions"><i class="fa fa-check"></i><b>10.3</b> Descriptive statistics as matrix functions</a><ul>
<li class="chapter" data-level="10.3.1" data-path="matrices-in-r.html"><a href="matrices-in-r.html#mean-vector-and-matrix"><i class="fa fa-check"></i><b>10.3.1</b> Mean vector and matrix</a></li>
<li class="chapter" data-level="10.3.2" data-path="matrices-in-r.html"><a href="matrices-in-r.html#deviation-matrix"><i class="fa fa-check"></i><b>10.3.2</b> Deviation matrix</a></li>
<li class="chapter" data-level="10.3.3" data-path="matrices-in-r.html"><a href="matrices-in-r.html#covariance-matrix"><i class="fa fa-check"></i><b>10.3.3</b> Covariance matrix</a></li>
<li class="chapter" data-level="10.3.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#correlation-matrix"><i class="fa fa-check"></i><b>10.3.4</b> Correlation matrix</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="matrices-in-r.html"><a href="matrices-in-r.html#matrix-inverse"><i class="fa fa-check"></i><b>10.4</b> Matrix Inverse</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="multiple-regression.html"><a href="multiple-regression.html"><i class="fa fa-check"></i><b>11</b> Multiple Regression</a><ul>
<li class="chapter" data-level="11.1" data-path="multiple-regression.html"><a href="multiple-regression.html#new-libraries-to-install"><i class="fa fa-check"></i><b>11.1</b> New Libraries to install</a></li>
<li class="chapter" data-level="11.2" data-path="multiple-regression.html"><a href="multiple-regression.html#libraries-4"><i class="fa fa-check"></i><b>11.2</b> Libraries</a></li>
<li class="chapter" data-level="11.3" data-path="multiple-regression.html"><a href="multiple-regression.html#examplar-data"><i class="fa fa-check"></i><b>11.3</b> Examplar data</a></li>
<li class="chapter" data-level="11.4" data-path="multiple-regression.html"><a href="multiple-regression.html#data-exploration"><i class="fa fa-check"></i><b>11.4</b> Data exploration</a></li>
<li class="chapter" data-level="11.5" data-path="multiple-regression.html"><a href="multiple-regression.html#d-plots"><i class="fa fa-check"></i><b>11.5</b> 3D Plots</a><ul>
<li class="chapter" data-level="11.5.1" data-path="multiple-regression.html"><a href="multiple-regression.html#scatterplot3d"><i class="fa fa-check"></i><b>11.5.1</b> scatterplot3d</a></li>
<li class="chapter" data-level="11.5.2" data-path="multiple-regression.html"><a href="multiple-regression.html#rgl"><i class="fa fa-check"></i><b>11.5.2</b> rgl</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="multiple-regression.html"><a href="multiple-regression.html#fitting-the-regression-model"><i class="fa fa-check"></i><b>11.6</b> Fitting the regression model</a><ul>
<li class="chapter" data-level="11.6.1" data-path="multiple-regression.html"><a href="multiple-regression.html#visualizing-the-regression-model-in-scatterplot3d"><i class="fa fa-check"></i><b>11.6.1</b> Visualizing the regression model in scatterplot3d</a></li>
</ul></li>
<li class="chapter" data-level="11.7" data-path="multiple-regression.html"><a href="multiple-regression.html#interpretting-the-regression-model"><i class="fa fa-check"></i><b>11.7</b> Interpretting the regression model</a></li>
<li class="chapter" data-level="11.8" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-vector-geometry-of-the-regression-model"><i class="fa fa-check"></i><b>11.8</b> Exploring the Vector Geometry of the Regression Model</a></li>
<li class="chapter" data-level="11.9" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-residuals-from-the-model-fit"><i class="fa fa-check"></i><b>11.9</b> Exploring the Residuals from the Model Fit</a></li>
<li class="chapter" data-level="11.10" data-path="multiple-regression.html"><a href="multiple-regression.html#an-alternate-model"><i class="fa fa-check"></i><b>11.10</b> An alternate model</a></li>
<li class="chapter" data-level="11.11" data-path="multiple-regression.html"><a href="multiple-regression.html#fitting-a-curvilinear-model-using-lm"><i class="fa fa-check"></i><b>11.11</b> Fitting a curvilinear model using lm()</a></li>
<li class="chapter" data-level="11.12" data-path="multiple-regression.html"><a href="multiple-regression.html#exploring-the-impact-of-nearly-collinear-predictors-on-regression"><i class="fa fa-check"></i><b>11.12</b> Exploring the impact of nearly collinear predictors on regression</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html"><i class="fa fa-check"></i><b>12</b> Principal Components Analysis</a><ul>
<li class="chapter" data-level="12.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#libraries-5"><i class="fa fa-check"></i><b>12.1</b> Libraries</a></li>
<li class="chapter" data-level="12.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#matrices-as-linear-transformations"><i class="fa fa-check"></i><b>12.2</b> Matrices as linear transformations</a></li>
<li class="chapter" data-level="12.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#eigenanalysis-in-r"><i class="fa fa-check"></i><b>12.3</b> Eigenanalysis in R</a></li>
<li class="chapter" data-level="12.4" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#principal-components-analysis-in-r"><i class="fa fa-check"></i><b>12.4</b> Principal Components Analysis in R</a><ul>
<li class="chapter" data-level="12.4.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#bioenv-dataset"><i class="fa fa-check"></i><b>12.4.1</b> Bioenv dataset</a></li>
<li class="chapter" data-level="12.4.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pca-of-the-bioenv-dataset"><i class="fa fa-check"></i><b>12.4.2</b> PCA of the Bioenv dataset</a></li>
<li class="chapter" data-level="12.4.3" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#calculating-factor-loadings"><i class="fa fa-check"></i><b>12.4.3</b> Calculating Factor Loadings</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#drawing-figures-to-represent-pca"><i class="fa fa-check"></i><b>12.5</b> Drawing Figures to Represent PCA</a><ul>
<li class="chapter" data-level="12.5.1" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#pc-score-plots"><i class="fa fa-check"></i><b>12.5.1</b> PC Score Plots</a></li>
<li class="chapter" data-level="12.5.2" data-path="principal-components-analysis.html"><a href="principal-components-analysis.html#simultaneous-depiction-of-observations-and-variables-in-the-pc-space"><i class="fa fa-check"></i><b>12.5.2</b> Simultaneous Depiction of Observations and Variables in the PC Space</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html"><i class="fa fa-check"></i><b>13</b> Simulating Sampling Distributions</a><ul>
<li class="chapter" data-level="13.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#libraries-6"><i class="fa fa-check"></i><b>13.1</b> Libraries</a></li>
<li class="chapter" data-level="13.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#data-set-simulated-male-heights"><i class="fa fa-check"></i><b>13.2</b> Data set: Simulated male heights</a><ul>
<li class="chapter" data-level="13.2.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#seeding-the-pseudo-random-number-generator"><i class="fa fa-check"></i><b>13.2.1</b> Seeding the pseudo-random number generator</a></li>
<li class="chapter" data-level="13.2.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#generating-the-simulated-population"><i class="fa fa-check"></i><b>13.2.2</b> Generating the simulated population</a></li>
<li class="chapter" data-level="13.2.3" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#properties-of-the-simulated-population"><i class="fa fa-check"></i><b>13.2.3</b> Properties of the simulated population</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#random-sampling-from-the-simulated-population"><i class="fa fa-check"></i><b>13.3</b> Random sampling from the simulated population</a><ul>
<li class="chapter" data-level="13.3.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#another-random-sample"><i class="fa fa-check"></i><b>13.3.1</b> Another random sample</a></li>
<li class="chapter" data-level="13.3.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#simulating-the-generation-of-many-random-samples"><i class="fa fa-check"></i><b>13.3.2</b> Simulating the generation of many random samples</a></li>
<li class="chapter" data-level="13.3.3" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#a-function-to-estimate-statistics-of-interest-in-a-random-sample"><i class="fa fa-check"></i><b>13.3.3</b> A function to estimate statistics of interest in a random sample</a></li>
<li class="chapter" data-level="13.3.4" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#generating-statistics-for-many-random-samples"><i class="fa fa-check"></i><b>13.3.4</b> Generating statistics for many random samples</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#simulated-sampling-distribution-of-the-mean"><i class="fa fa-check"></i><b>13.4</b> Simulated sampling distribution of the mean</a><ul>
<li class="chapter" data-level="13.4.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#sampling-distributions-for-different-sample-sizes"><i class="fa fa-check"></i><b>13.4.1</b> Sampling distributions for different sample sizes</a></li>
<li class="chapter" data-level="13.4.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#discussion-of-trends-for-sampling-distributions-of-different-sample-sizes"><i class="fa fa-check"></i><b>13.4.2</b> Discussion of trends for sampling distributions of different sample sizes</a></li>
</ul></li>
<li class="chapter" data-level="13.5" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#standard-error-of-the-mean"><i class="fa fa-check"></i><b>13.5</b> Standard Error of the Mean</a></li>
<li class="chapter" data-level="13.6" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#sampling-distribution-of-the-standard-deviation"><i class="fa fa-check"></i><b>13.6</b> Sampling Distribution of the Standard Deviation</a></li>
<li class="chapter" data-level="13.7" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#what-happens-to-the-sampling-distribution-of-the-mean-and-standard-deviation-when-our-sample-size-is-small"><i class="fa fa-check"></i><b>13.7</b> What happens to the sampling distribution of the mean and standard deviation when our sample size is small?</a><ul>
<li class="chapter" data-level="13.7.1" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#for-small-samples-sample-standard-deviations-systematically-underestimate-the-population-standard-deviation"><i class="fa fa-check"></i><b>13.7.1</b> For small samples, sample standard deviations systematically underestimate the population standard deviation</a></li>
<li class="chapter" data-level="13.7.2" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#underestimates-of-the-standard-deviation-given-small-n-lead-to-understimates-of-the-se-of-the-mean"><i class="fa fa-check"></i><b>13.7.2</b> Underestimates of the standard deviation given small <span class="math inline">\(n\)</span> lead to understimates of the SE of the mean</a></li>
</ul></li>
<li class="chapter" data-level="13.8" data-path="simulating-sampling-distributions.html"><a href="simulating-sampling-distributions.html#the-t-distribution-is-the-appropriate-distribution-for-describing-the-sampling-distribution-of-the-mean-when-n-is-small"><i class="fa fa-check"></i><b>13.8</b> The t-distribution is the appropriate distribution for describing the sampling distribution of the mean when <span class="math inline">\(n\)</span> is small</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html"><i class="fa fa-check"></i><b>14</b> Simulating confidence intervals</a><ul>
<li class="chapter" data-level="14.1" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#libraries-7"><i class="fa fa-check"></i><b>14.1</b> Libraries</a></li>
<li class="chapter" data-level="14.2" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#confidence-intervals"><i class="fa fa-check"></i><b>14.2</b> Confidence Intervals</a></li>
<li class="chapter" data-level="14.3" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#calibrating-confidence-intervals"><i class="fa fa-check"></i><b>14.3</b> Calibrating confidence intervals</a></li>
<li class="chapter" data-level="14.4" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#standard-formulation-for-confidence-intervals"><i class="fa fa-check"></i><b>14.4</b> Standard formulation for confidence intervals</a></li>
<li class="chapter" data-level="14.5" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#simulating-confidence-intervals-for-the-mean"><i class="fa fa-check"></i><b>14.5</b> Simulating confidence intervals for the mean</a></li>
<li class="chapter" data-level="14.6" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#sample-estimate-of-the-standard-error-of-the-mean"><i class="fa fa-check"></i><b>14.6</b> Sample estimate of the standard error of the mean</a></li>
<li class="chapter" data-level="14.7" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#ci-of-mean-simulation"><i class="fa fa-check"></i><b>14.7</b> CI of mean simulation</a></li>
<li class="chapter" data-level="14.8" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#generating-a-table-of-cis-and-corresponding-margins-of-error"><i class="fa fa-check"></i><b>14.8</b> Generating a table of CIs and corresponding margins of error</a></li>
<li class="chapter" data-level="14.9" data-path="simulating-confidence-intervals.html"><a href="simulating-confidence-intervals.html#interpreting-our-simulation"><i class="fa fa-check"></i><b>14.9</b> Interpreting our simulation</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Biology 723: Statistical Computing for Biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="simulating-sampling-distributions" class="section level1">
<h1><span class="header-section-number">Chapter 13</span> Simulating Sampling Distributions</h1>
<p>Usually when we collect biological data, it’s because we’re trying to learn about some underlying “population” of interest. Population here could refer to an actual population (e.g. all males over 20 in the United States; brushtail possums in the state of Victoria, Australia), an abstract population (e.g. corn plants grown from Monsanto “round up ready” seed; yeast cells with genotypes identical to the reference strain S288c), outcomes of a stochastic process we can observe and measure (e.g. meiotic recombination in flies; hadrons detected at the LHC during a particle collision experiment), etc.</p>
<p>It is often impractical or impossible to measure all objects/individuals in a population of interest, so we take a <strong>sample</strong> from the population and make measurements on the variables of interest in that sample. We do so with the hope that the various statistics we calculate on the variables of interest in that sample will be useful estimates of those same statistics in the underlying population.</p>
<p>However, we must always keep in mind that the statistics we calculate from our sample will almost never exactly match those of the underlying population. That is when we collect a sample, and measure a statistic (e.g. mean) on variable X in the sample, there is a degree of <em>uncertainty</em> about how well our estimate matches the true value of that statistic for X in the underlying population.</p>
<p><em>Statistical inference</em> is about quantifying the uncertainty associated with statistics and using that information to test hypotheses and evaluate models.</p>
<p>Today we’re going to review a fundamental concept in statistical inference, the notion of a <em>sampling distribution</em> for a statistic of interest. A sampling distribution is the probability distribution of a given statistic for samples of a given size. Traditionally sampling distributions were derived analytically. In this class session we’ll see how to approximate sampling distributions for any a statistic using computer simulation.</p>
<div id="libraries-6" class="section level2">
<h2><span class="header-section-number">13.1</span> Libraries</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)
<span class="kw">library</span>(magrittr)
<span class="kw">library</span>(stringr)</code></pre></div>
</div>
<div id="data-set-simulated-male-heights" class="section level2">
<h2><span class="header-section-number">13.2</span> Data set: Simulated male heights</h2>
<p>To illustrate the concept of sampling distributions, we’ll use a simulated data set to represent the underlying population we’re trying to estimate statistics for. This will allow us to compare the various statistics we calculate and their sampling distributions to their “true” values.</p>
<p>Let’s simulate a population consisting of 25,000 individuals with a single trait of interest – height (measured in centimeters). We will simulate this data set based on information about the distribution of the heights of adult males in the US estimated in a study carried out from 2011-2014 by the US Department of Health and Human Services<a href="#fn1" class="footnoteRef" id="fnref1"><sup>1</sup></a>.</p>
<div id="seeding-the-pseudo-random-number-generator" class="section level3">
<h3><span class="header-section-number">13.2.1</span> Seeding the pseudo-random number generator</h3>
<p>When carrying out simulations, we employ random number generators (e.g. to choose random samples). Most computers can not generate true random numbers – instead they use algorithms that approximate the generation of random numbers (pseudo-random number generators). One important difference between a true random number generator and a pseudo-random number generator is that we can regenerate a series of pseudo-random numbers if we know the “seed” value that initialized the algorithm. We can specifically set this seed value, so that we can guarantee that two different people evaluating this notebook get the same results, even though we’re using (pseudo)random numbers in our simulation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># make our simulation repeatable by seeding RNG</span>
<span class="kw">set.seed</span>(<span class="dv">20180321</span>)</code></pre></div>
</div>
<div id="generating-the-simulated-population" class="section level3">
<h3><span class="header-section-number">13.2.2</span> Generating the simulated population</h3>
<p>Having seeded our RNG, we then generate our simulated population using the <code>rnorm()</code> function. <code>rnorm()</code> draws random values from a normal distribution with the given parameters of mean and standard deviation. We didn’t have to use a normal distribution for our simulations, but many biological variables are approximately normally distributed, so this is a common assumption.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># male mean height and sd in centimeters from USDHHS report</span>
mean.ht &lt;-<span class="st"> </span><span class="fl">175.7</span>
sd.ht &lt;-<span class="st"> </span><span class="fl">15.19</span>

height.data &lt;-<span class="st"> </span><span class="kw">data_frame</span>(<span class="dt">height =</span> <span class="kw">rnorm</span>(<span class="dv">25000</span>, <span class="dt">mean =</span> mean.ht, <span class="dt">sd =</span> sd.ht))</code></pre></div>
</div>
<div id="properties-of-the-simulated-population" class="section level3">
<h3><span class="header-section-number">13.2.3</span> Properties of the simulated population</h3>
<p>Let’s take a moment to visualize the distribution of heights in our population of interest:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(height.data, <span class="kw">aes</span>(<span class="dt">x =</span> height)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_histogram</span>(<span class="kw">aes</span>(<span class="dt">y =</span> ..density..), <span class="dt">bins=</span><span class="dv">50</span>, <span class="dt">alpha=</span><span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_density</span>() <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Height (cm)&quot;</span>, <span class="dt">title =</span> <span class="st">&quot;Distribution of Heights in the Population of Interest&quot;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-370-1.png" width="576" /></p>
<p>When we generated this simulated population, I specified that the heights should be drawn from a normal distribution. As one would expect, the histogram and density plot look fairly close to the classic bell-shaped curve of a normal distribution.</p>
<p>Let’s turn some key summary statistics of the population. The values of these summary statistics are those we’re trying to estimate from our sample(s). We’ll refer to these as the “true” values.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">true.values &lt;-<span class="st"> </span>
<span class="st">  </span>height.data <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">mean.height =</span> <span class="kw">mean</span>(height), 
            <span class="dt">sd.height =</span> <span class="kw">sd</span>(height))

true.values
<span class="co"># A tibble: 1 x 2</span>
  mean.height sd.height
        <span class="op">&lt;</span>dbl<span class="op">&gt;</span><span class="st">     </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="dv">1</span>        <span class="dv">176</span>.      <span class="fl">15.2</span></code></pre></div>
<p>In mathematical notation our population is normally distributed with a mean height of 175.7201895cm and standard deviation of 15.1963663cm. Note that these values are close to, but not exactly the parametes we passed to <code>rnorm()</code>.</p>
</div>
</div>
<div id="random-sampling-from-the-simulated-population" class="section level2">
<h2><span class="header-section-number">13.3</span> Random sampling from the simulated population</h2>
<p>Let’s simulate the process of taking a single sample of 30 individuals from our population, using the <code>dplyr::sample_n()</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sample.a &lt;-
<span class="st">  </span>height.data <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">sample_n</span>(<span class="dv">30</span>)</code></pre></div>
<p>Now we’ll create a histogram of the height variable in our sample. For reference we’ll also plot the histogram for the true population in the background (but remember, in the typical case you don’t know what the true population looks like)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sample.a <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> height)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_histogram</span>(<span class="dt">data=</span>height.data, <span class="kw">aes</span>(<span class="dt">x =</span> height, <span class="dt">y =</span> ..density..), <span class="dt">alpha=</span><span class="fl">0.25</span>, <span class="dt">bins=</span><span class="dv">50</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_histogram</span>(<span class="kw">aes</span>(<span class="dt">y =</span> ..density..), <span class="dt">fill =</span> <span class="st">&#39;steelblue&#39;</span>, <span class="dt">alpha=</span><span class="fl">0.75</span>, <span class="dt">bins=</span><span class="dv">9</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> true.values<span class="op">$</span>mean.height, <span class="dt">linetype =</span> <span class="st">&quot;solid&quot;</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="kw">mean</span>(sample.a<span class="op">$</span>height), <span class="dt">linetype =</span> <span class="st">&quot;dashed&quot;</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Height (cm)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Density&quot;</span>, 
       <span class="dt">title =</span> <span class="st">&quot;Distribution of heights in the underlying population (grey)</span><span class="ch">\n</span><span class="st">and a single sample of size 30 (blue)&quot;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-373-1.png" width="576" /></p>
<p>The solid vertical line represent the true mean of the population, the dashed line represents the sample mean. Comparing the two distributions we see that while our sample of 30 observations is relatively small,its location (center) and spread that are roughly similar to those of the underlying population.</p>
<p>Let’s create a table giving the estimates of the mean and standard deviation in our sample:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sample.a <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">sample.mean =</span> <span class="kw">mean</span>(height), 
            <span class="dt">sample.sd =</span> <span class="kw">sd</span>(height))
<span class="co"># A tibble: 1 x 2</span>
  sample.mean sample.sd
        <span class="op">&lt;</span>dbl<span class="op">&gt;</span><span class="st">     </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="dv">1</span>        <span class="dv">173</span>.      <span class="fl">13.2</span></code></pre></div>
<p>Based on our sample, we estimate that the mean height of males in our population of interest is 173.4747577cm with a standard deviation of 13.1761523cm.</p>
<div id="another-random-sample" class="section level3">
<h3><span class="header-section-number">13.3.1</span> Another random sample</h3>
<p>Let’s step back and think about our experiment. We took a random sample of 30 indiviuals from the population. The very nature of a “random sample” means we could just as well have gotten a different collection of individuals in our sample. Let’s take a second random sample of 25 individuals and see what the data looks like this time:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sample.b &lt;-
<span class="st">  </span>height.data <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">sample_n</span>(<span class="dv">30</span>)

<span class="kw">ggplot</span>(sample.b, <span class="kw">aes</span>(<span class="dt">x =</span> height)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_histogram</span>(<span class="dt">data=</span>height.data, <span class="kw">aes</span>(<span class="dt">x =</span> height, <span class="dt">y =</span> ..density..), <span class="dt">alpha=</span><span class="fl">0.25</span>, <span class="dt">bins=</span><span class="dv">50</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_histogram</span>(<span class="kw">aes</span>(<span class="dt">y =</span> ..density..), <span class="dt">fill =</span> <span class="st">&#39;steelblue&#39;</span>, <span class="dt">alpha=</span><span class="fl">0.75</span>, <span class="dt">bins=</span><span class="dv">9</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> true.values<span class="op">$</span>mean.height, <span class="dt">linetype =</span> <span class="st">&quot;solid&quot;</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="kw">mean</span>(sample.b<span class="op">$</span>height), <span class="dt">linetype =</span> <span class="st">&quot;dashed&quot;</span>)  <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Height (cm)&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Density&quot;</span>, 
       <span class="dt">title =</span> <span class="st">&quot;Distribution of heights in the underlying population (grey)</span><span class="ch">\n</span><span class="st">and a single sample of size 30 (blue)&quot;</span>)  </code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-375-1.png" width="576" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sample.b <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">sample.mean =</span> <span class="kw">mean</span>(height), 
            <span class="dt">sample.sd =</span> <span class="kw">sd</span>(height))
<span class="co"># A tibble: 1 x 2</span>
  sample.mean sample.sd
        <span class="op">&lt;</span>dbl<span class="op">&gt;</span><span class="st">     </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="dv">1</span>        <span class="dv">175</span>.      <span class="fl">14.8</span></code></pre></div>
<p>This time we estimated the mean height to be 175.1897271 cm and the standard deviation to be 14.8025475 cm.</p>
</div>
<div id="simulating-the-generation-of-many-random-samples" class="section level3">
<h3><span class="header-section-number">13.3.2</span> Simulating the generation of many random samples</h3>
<p>When we estimate population parameters, like the mean and standard deviation, based on a <em>sample</em>, our estimates will differ from the true population values by some amount. Any given random sample might provide better or worse estimates than another sample.</p>
<p>We can’t know how good our estimates of statistics like the mean and standard deviation are from any specific sample, but we we can study the behavior of such estimates <em>across many simulated samples</em> and learn something about how well our estimates do on average, as well the spread of these estimates.</p>
</div>
<div id="a-function-to-estimate-statistics-of-interest-in-a-random-sample" class="section level3">
<h3><span class="header-section-number">13.3.3</span> A function to estimate statistics of interest in a random sample</h3>
<p>First we’re going to write a function called <code>rsample.stats</code> that to carries out the following steps:</p>
<ul>
<li>Given a data frame <code>x</code></li>
<li>Take a random sample of size <code>n</code></li>
<li>For the variable specified by the character string, <code>var.name</code>, calculate the mean and standard deviation of that variable in the random sample</li>
<li>Return a table giving the sample size, sample mean, and sample standard deviation, represented as a data frame</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rsample.stats &lt;-<span class="st"> </span><span class="cf">function</span>(x, n, var.name) {
  sample_x &lt;-<span class="st"> </span><span class="kw">sample_n</span>(x, n)
  <span class="kw">data_frame</span>(<span class="dt">sample.size =</span> n, 
             <span class="dt">sample.mean =</span> <span class="kw">mean</span>(sample_x[[var.name]]), 
             <span class="dt">sample.sd =</span> <span class="kw">sd</span>(sample_x[[var.name]]))
}</code></pre></div>
<p>Let’s test <code>rsample.stats</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rsample.stats</span>(height.data, <span class="dv">30</span>, <span class="st">&quot;height&quot;</span>)
<span class="co"># A tibble: 1 x 3</span>
  sample.size sample.mean sample.sd
        <span class="op">&lt;</span>dbl<span class="op">&gt;</span><span class="st">       </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">     </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="dv">1</span>         <span class="dv">30</span>.        <span class="dv">174</span>.      <span class="fl">16.7</span></code></pre></div>
</div>
<div id="generating-statistics-for-many-random-samples" class="section level3">
<h3><span class="header-section-number">13.3.4</span> Generating statistics for many random samples</h3>
<p>Now we’ll see how to combine <code>rsample.stats</code> with two additional functions to repeatedly run the <code>rsample.stats</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df.samples.of.<span class="dv">30</span> &lt;-
<span class="st">  </span><span class="kw">rerun</span>(<span class="dv">1000</span>,  <span class="kw">rsample.stats</span>(height.data, <span class="dv">30</span>, <span class="st">&quot;height&quot;</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">bind_rows</span>()</code></pre></div>
<p>The function <code>rerun</code> is defined in the <code>purrr</code> library (automatically loaded with tidyverse). <code>purrr:rerun()</code> re-runs an expression(s) multiple times. The first argument to <code>rerun()</code> is the number of times you want to re-run, and the following arguments are the expressions to be re-run. Thus the second line of the code block above re-runs the <code>rsample.stats</code> function 1000 times using <code>height.data</code> as the input, generating sample statistics for samples of size 30 each time it’s run. <code>rerun</code> returns a list whose length is the specified number of runs.</p>
<p>The third line includes a call the <code>dplyr::bind_rows()</code>. This simply takes the list that <code>rerun</code> returns and collapses the list into a single data frame. <code>df.samples.of.30</code> is thus a data frame in which each row gives the sample size, sample mean, and sample standard deviation for a random sample of 30 individuals drawn from our underlying population (<code>height.data</code>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df.samples.of.<span class="dv">30</span>
<span class="co"># A tibble: 1,000 x 3</span>
   sample.size sample.mean sample.sd
         <span class="op">&lt;</span>dbl<span class="op">&gt;</span><span class="st">       </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">     </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="st"> </span><span class="dv">1</span>         <span class="dv">30</span>.        <span class="dv">177</span>.      <span class="fl">18.0</span>
 <span class="dv">2</span>         <span class="dv">30</span>.        <span class="dv">173</span>.      <span class="fl">14.6</span>
 <span class="dv">3</span>         <span class="dv">30</span>.        <span class="dv">178</span>.      <span class="fl">15.1</span>
 <span class="dv">4</span>         <span class="dv">30</span>.        <span class="dv">171</span>.      <span class="fl">16.9</span>
 <span class="dv">5</span>         <span class="dv">30</span>.        <span class="dv">176</span>.      <span class="fl">12.7</span>
 <span class="dv">6</span>         <span class="dv">30</span>.        <span class="dv">176</span>.      <span class="fl">13.5</span>
 <span class="dv">7</span>         <span class="dv">30</span>.        <span class="dv">177</span>.      <span class="fl">15.2</span>
 <span class="dv">8</span>         <span class="dv">30</span>.        <span class="dv">178</span>.      <span class="fl">19.3</span>
 <span class="dv">9</span>         <span class="dv">30</span>.        <span class="dv">171</span>.      <span class="fl">15.5</span>
<span class="dv">10</span>         <span class="dv">30</span>.        <span class="dv">175</span>.      <span class="fl">13.0</span>
<span class="co"># ... with 990 more rows</span></code></pre></div>
</div>
</div>
<div id="simulated-sampling-distribution-of-the-mean" class="section level2">
<h2><span class="header-section-number">13.4</span> Simulated sampling distribution of the mean</h2>
<p>Let’s review what we just did:</p>
<ul>
<li>We generated 1000 samples of size 30</li>
<li>For each of the samples we calculated the mean and standard deviation of the height variable <em>in that sample</em></li>
<li>We combined each of those estimates of the mean and standard deviation into a data frame</li>
</ul>
<p>The 1000 estimates of the mean we generated represents a new distribution – what we will call a <strong>sampling distribution of the mean for samples of size 30</strong>. Let’s plot this sampling distribution:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(df.samples.of.<span class="dv">30</span>, <span class="kw">aes</span>(<span class="dt">x =</span> sample.mean, <span class="dt">y =</span> ..density..)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_histogram</span>(<span class="dt">bins=</span><span class="dv">25</span>, <span class="dt">fill =</span> <span class="st">&#39;firebrick&#39;</span>, <span class="dt">alpha=</span><span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> true.values<span class="op">$</span>mean.height, <span class="dt">linetype =</span> <span class="st">&quot;dashed&quot;</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Sample means&quot;</span>, <span class="dt">y  =</span> <span class="st">&quot;Density&quot;</span>,
       <span class="dt">title =</span> <span class="st">&quot;Distribution of mean heights for 1000 samples of size 30&quot;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-381-1.png" width="576" /></p>
<p>This particular sampling distribution of the mean is a probability distribution that we can use to estimate the probability that a sample mean falls within a given interval, assuming our sample is a random sample of size 30 drawn from our underlying population.</p>
<p>From our visualization, we see that the distribution of sample mean heights is approximately centered around the true mean height. Most of the sample estiamtes of the mean height are within 5 cm of the true population mean height (175.6cm), but a small number of estimates of the sample mean as off by nearly 10cm.</p>
<p>Let’s make this more precise by calculating the mean and standard deviation of the sampling distribution of means (I included the min and max as well).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df.samples.of.<span class="dv">30</span> <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">mean.of.means =</span> <span class="kw">mean</span>(sample.mean),
            <span class="dt">sd.of.means =</span> <span class="kw">sd</span>(sample.mean),
            <span class="dt">min.of.means =</span> <span class="kw">min</span>(sample.mean),
            <span class="dt">max.of.means =</span> <span class="kw">max</span>(sample.mean))
<span class="co"># A tibble: 1 x 4</span>
  mean.of.means sd.of.means min.of.means max.of.means
          <span class="op">&lt;</span>dbl<span class="op">&gt;</span><span class="st">       </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">        </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">        </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="dv">1</span>          <span class="dv">176</span>.        <span class="fl">2.85</span>         <span class="dv">167</span>.         <span class="dv">185</span>.</code></pre></div>
<div id="sampling-distributions-for-different-sample-sizes" class="section level3">
<h3><span class="header-section-number">13.4.1</span> Sampling distributions for different sample sizes</h3>
<p>In the example above we simulated the sampling distribution of the mean for samples of size 30. How would the sampling distribution change if we increased the sample size? In the next code block we generate sampling distributions of the mean (and standard deviation) for samples of size 50, 100, 250, and 500.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df.samples.of.<span class="dv">50</span> &lt;-
<span class="st">  </span><span class="kw">rerun</span>(<span class="dv">1000</span>,  height.data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">rsample.stats</span>(<span class="dv">50</span>, <span class="st">&quot;height&quot;</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">bind_rows</span>()

df.samples.of.<span class="dv">100</span> &lt;-
<span class="st">  </span><span class="kw">rerun</span>(<span class="dv">1000</span>,  height.data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">rsample.stats</span>(<span class="dv">100</span>, <span class="st">&quot;height&quot;</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">bind_rows</span>()

df.samples.of.<span class="dv">250</span> &lt;-
<span class="st">  </span><span class="kw">rerun</span>(<span class="dv">1000</span>,  height.data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">rsample.stats</span>(<span class="dv">250</span>, <span class="st">&quot;height&quot;</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">bind_rows</span>()

df.samples.of.<span class="dv">500</span> &lt;-
<span class="st">  </span><span class="kw">rerun</span>(<span class="dv">1000</span>,  height.data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">rsample.stats</span>(<span class="dv">500</span>, <span class="st">&quot;height&quot;</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">bind_rows</span>()</code></pre></div>
<p>To make plotting and comparison easier we will combine each of the individual data frames, representing the different sampling distributions for samples of a given size, into a single data frame.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df.combined &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw">bind_rows</span>(df.samples.of.<span class="dv">30</span>,
            df.samples.of.<span class="dv">50</span>,
            df.samples.of.<span class="dv">100</span>,
            df.samples.of.<span class="dv">250</span>,
            df.samples.of.<span class="dv">500</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># create a factor version of sample size to facilitate plotting</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">sample.sz =</span> <span class="kw">as.factor</span>(sample.size))</code></pre></div>
<p>We then plot each of the individual sampling distributions, faceting on sample size.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(df.combined, <span class="kw">aes</span>(<span class="dt">x =</span> sample.mean, <span class="dt">y =</span> ..density.., <span class="dt">fill =</span> sample.sz)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_histogram</span>(<span class="dt">bins=</span><span class="dv">25</span>, <span class="dt">alpha=</span><span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> true.values<span class="op">$</span>mean.height, <span class="dt">linetype =</span> <span class="st">&quot;dashed&quot;</span>)  <span class="op">+</span>
<span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>sample.sz, <span class="dt">nrow =</span> <span class="dv">1</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">scale_fill_brewer</span>(<span class="dt">palette=</span><span class="st">&quot;Set1&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="co"># change color palette</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Sample means&quot;</span>, <span class="dt">y  =</span> <span class="st">&quot;Density&quot;</span>,
       <span class="dt">title =</span> <span class="st">&quot;Distribution of mean heights for samples of varying size&quot;</span>)  </code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-385-1.png" width="1440" /></p>
</div>
<div id="discussion-of-trends-for-sampling-distributions-of-different-sample-sizes" class="section level3">
<h3><span class="header-section-number">13.4.2</span> Discussion of trends for sampling distributions of different sample sizes</h3>
<p>The key trend we see when comparing the sampling distributions of the mean for samples of different size is that as the sample size gets larger, the spread of the sampling distribution of the mean becomes narrower around the true mean. This means that <em>as sample size increases, the uncertainty associated with our estimates of the mean decreases</em>.</p>
<p>Let’s create a table, grouped by sample size, to help quantify this pattern:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sampling.distn.mean.table &lt;-
<span class="st">  </span>df.combined <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(sample.size) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">mean.of.means =</span> <span class="kw">mean</span>(sample.mean),
            <span class="dt">sd.of.means =</span> <span class="kw">sd</span>(sample.mean),
            <span class="dt">min.of.means =</span> <span class="kw">min</span>(sample.mean),
            <span class="dt">max.of.means =</span> <span class="kw">max</span>(sample.mean))

sampling.distn.mean.table
<span class="co"># A tibble: 5 x 5</span>
  sample.size mean.of.means sd.of.means min.of.means max.of.means
        <span class="op">&lt;</span>dbl<span class="op">&gt;</span><span class="st">         </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">       </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">        </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">        </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="dv">1</span>         <span class="dv">30</span>.          <span class="dv">176</span>.       <span class="fl">2.85</span>          <span class="dv">167</span>.         <span class="dv">185</span>.
<span class="dv">2</span>         <span class="dv">50</span>.          <span class="dv">176</span>.       <span class="fl">2.16</span>          <span class="dv">169</span>.         <span class="dv">182</span>.
<span class="dv">3</span>        <span class="dv">100</span>.          <span class="dv">176</span>.       <span class="fl">1.47</span>          <span class="dv">171</span>.         <span class="dv">180</span>.
<span class="dv">4</span>        <span class="dv">250</span>.          <span class="dv">176</span>.       <span class="fl">0.967</span>         <span class="dv">173</span>.         <span class="dv">179</span>.
<span class="dv">5</span>        <span class="dv">500</span>.          <span class="dv">176</span>.       <span class="fl">0.665</span>         <span class="dv">174</span>.         <span class="dv">178</span>.</code></pre></div>
</div>
</div>
<div id="standard-error-of-the-mean" class="section level2">
<h2><span class="header-section-number">13.5</span> Standard Error of the Mean</h2>
<p>We see from the graph and table above that our estimates of the mean cluster more tightly about the true mean as our sample size increases. This is obvious when we compare the standard deviation of our mean estimates as a function of sample size.</p>
<p>The standard deviation of the sampling distribution of a statistic of interest is called the <strong>Standard Error</strong> of that statistic. Here, through simulation, we are approximating the <strong>Standard Error of the Mean</strong>.</p>
<p>When sample sizes are large (&gt;30 observations), one can show mathematically that for normally distributed data the expected Standard Error of the Mean as a function of sample size is approximately: <span class="math display">\[
\mbox{Standard Error of Mean} \approx \frac{\sigma}{\sqrt{n}}
\]</span> where <span class="math inline">\(\sigma\)</span> is the population standard deviation (i.e. the “true” standard deviation), and <span class="math inline">\(n\)</span> is the sample size.</p>
<p>Let’s compare that theoretical expectation to our simulated results:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">se.mean.theory &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">seq</span>(<span class="dv">10</span>,<span class="dv">500</span>,<span class="dv">10</span>), 
                         <span class="cf">function</span>(n){ true.values<span class="op">$</span>sd.height<span class="op">/</span><span class="kw">sqrt</span>(n) })

df.se.mean.theory &lt;-<span class="st"> </span><span class="kw">data_frame</span>(<span class="dt">sample.size =</span> <span class="kw">seq</span>(<span class="dv">10</span>,<span class="dv">500</span>,<span class="dv">10</span>),
                                <span class="dt">std.error =</span> se.mean.theory)

<span class="kw">ggplot</span>(sampling.distn.mean.table, <span class="kw">aes</span>(<span class="dt">x =</span> sample.size, <span class="dt">y =</span> sd.of.means)) <span class="op">+</span>
<span class="st">   </span><span class="co"># plot standard errors of mean based on our simulations</span>
<span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st">  </span>
<span class="st">  </span><span class="co"># plot standard errors of the mean based on theory</span>
<span class="st">  </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="dt">x =</span> sample.size, <span class="dt">y =</span> std.error), <span class="dt">data =</span> df.se.mean.theory, <span class="dt">color=</span><span class="st">&quot;red&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Sample size&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Std Error of Mean&quot;</span>,
       <span class="dt">title =</span> <span class="st">&quot;A comparison of theoretical (red line) and simulated (points) estimates of</span><span class="ch">\n</span><span class="st">the standard error of the mean for samples of different size&quot;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-387-1.png" width="576" /></p>
<p>We see that as sample sizes increase, the standard error of the mean decreases. This means that as our samples get larger, our uncertainty in our sample estimate of the mean (our best guess for the population mean) gets smaller.</p>
</div>
<div id="sampling-distribution-of-the-standard-deviation" class="section level2">
<h2><span class="header-section-number">13.6</span> Sampling Distribution of the Standard Deviation</h2>
<p>Above we explored how the sampling distribution of the mean changes with sample size. We can similarly explore the sampling distribution of any other statistic, such as the standard deviation, or the median, or the the range, etc.</p>
<p>Recall that when we drew random samples we calculated the standard deviation of each of those samples in addition to the mean. This means we can immediately visualize the sampling distribution of the standard deviation as shown below:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(df.combined, <span class="kw">aes</span>(<span class="dt">x =</span> sample.sd, <span class="dt">y =</span> ..density.., <span class="dt">fill =</span> sample.sz)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_histogram</span>(<span class="dt">bins=</span><span class="dv">25</span>, <span class="dt">alpha=</span><span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> true.values<span class="op">$</span>sd.height, <span class="dt">linetype =</span> <span class="st">&quot;dashed&quot;</span>)  <span class="op">+</span>
<span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>sample.sz, <span class="dt">nrow =</span> <span class="dv">1</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">scale_fill_brewer</span>(<span class="dt">palette=</span><span class="st">&quot;Set1&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Sample standard deviations&quot;</span>, <span class="dt">y  =</span> <span class="st">&quot;Density&quot;</span>,
       <span class="dt">title =</span> <span class="st">&quot;Sampling distribution of standard deviation of height for samples of varying size&quot;</span>)  </code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-388-1.png" width="1440" /></p>
<p>The key trend we saw when examining the sampling distribution of the mean is also apparent for standard deviation – bigger samples lead to tighter sampling distributions and hence less uncertainty in the sample estimates of the standard deviation.</p>
<p>As before, we summarize key statistics of the sampling distribution in a table:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sampling.distn.sd.table &lt;-
<span class="st">  </span>df.combined <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(sample.size) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">mean.of.sds =</span> <span class="kw">mean</span>(sample.sd),
            <span class="dt">sd.of.sds =</span> <span class="kw">sd</span>(sample.sd),
            <span class="dt">min.of.sds =</span> <span class="kw">min</span>(sample.sd),
            <span class="dt">max.of.sds =</span> <span class="kw">max</span>(sample.sd))

sampling.distn.sd.table
<span class="co"># A tibble: 5 x 5</span>
  sample.size mean.of.sds sd.of.sds min.of.sds max.of.sds
        <span class="op">&lt;</span>dbl<span class="op">&gt;</span><span class="st">       </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">     </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">      </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">      </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="dv">1</span>         <span class="dv">30</span>.        <span class="fl">15.2</span>     <span class="fl">2.02</span>        <span class="fl">9.58</span>       <span class="fl">21.3</span>
<span class="dv">2</span>         <span class="dv">50</span>.        <span class="fl">15.2</span>     <span class="fl">1.56</span>       <span class="fl">10.1</span>        <span class="fl">20.9</span>
<span class="dv">3</span>        <span class="dv">100</span>.        <span class="fl">15.1</span>     <span class="fl">1.11</span>       <span class="fl">11.9</span>        <span class="fl">18.7</span>
<span class="dv">4</span>        <span class="dv">250</span>.        <span class="fl">15.2</span>     <span class="fl">0.690</span>      <span class="fl">13.1</span>        <span class="fl">17.8</span>
<span class="dv">5</span>        <span class="dv">500</span>.        <span class="fl">15.2</span>     <span class="fl">0.465</span>      <span class="fl">13.6</span>        <span class="fl">17.0</span></code></pre></div>
<p>For normally distributed data the expected Standard Error of the Standard Deviation (i.e. the standard deviation of standard deviations!) is approximately:</p>
<p><span class="math display">\[
\mbox{Standard Error of Standard Deviation} \approx \frac{\sigma}{\sqrt{2(n-1)}}
\]</span> where <span class="math inline">\(\sigma\)</span> is the population standard deviation, and <span class="math inline">\(n\)</span> is the sample size.</p>
<p>As before, let’s visually compare the theoretical expectation to our simulated estimates.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">se.sd.theory &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">seq</span>(<span class="dv">10</span>, <span class="dv">500</span>, <span class="dv">10</span>), 
                       <span class="cf">function</span>(n){ true.values<span class="op">$</span>sd.height<span class="op">/</span><span class="kw">sqrt</span>(<span class="dv">2</span><span class="op">*</span>(n<span class="op">-</span><span class="dv">1</span>))})

df.se.sd.theory &lt;-<span class="st"> </span><span class="kw">data_frame</span>(<span class="dt">sample.size =</span> <span class="kw">seq</span>(<span class="dv">10</span>,<span class="dv">500</span>,<span class="dv">10</span>), 
                              <span class="dt">std.error =</span> se.sd.theory)

<span class="kw">ggplot</span>(sampling.distn.sd.table, <span class="kw">aes</span>(<span class="dt">x =</span> sample.size, <span class="dt">y =</span> sd.of.sds)) <span class="op">+</span>
<span class="st">   </span><span class="co"># plot standard errors of mean based on our simulations</span>
<span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st">  </span>
<span class="st">  </span><span class="co"># plot standard errors of the mean based on theory</span>
<span class="st">  </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="dt">x =</span> sample.size, <span class="dt">y =</span> std.error), <span class="dt">data =</span> df.se.sd.theory, <span class="dt">color=</span><span class="st">&quot;red&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Sample size&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Std Error of Standard Deviation&quot;</span>,
       <span class="dt">title =</span> <span class="st">&quot;A comparison of theoretical (red line) and simulated (points) estimates of</span><span class="ch">\n</span><span class="st">the standard error of the standard deviation for samples of different size&quot;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-390-1.png" width="576" /></p>
</div>
<div id="what-happens-to-the-sampling-distribution-of-the-mean-and-standard-deviation-when-our-sample-size-is-small" class="section level2">
<h2><span class="header-section-number">13.7</span> What happens to the sampling distribution of the mean and standard deviation when our sample size is small?</h2>
<p>We would hope that, regardless of sample size, the sampling distributions of both the mean and standard deviation should be centered around the true population value, <span class="math inline">\(\mu\)</span> and <span class="math inline">\(\sigma\)</span> respectively. That seemed to be the case for the modest to large sample sizes we’ve looked at so far (30 to 500 observations). Does this also hold for small samples? Let’s use simulation to explore how well this is expectation is met for small samples.</p>
<p>As we’ve done before, we simulate the sampling distribution of the mean and standard deviation for samples of varying size (we also calculate some other values which will be useful for our exposition below).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># we&#39;ll use the same mean and sd we&#39;ve been using previously</span>
mu =<span class="st"> </span>mean.ht
sigma =<span class="st"> </span>sd.ht

<span class="co"># list of sample sizes we&#39;ll generate</span>
ssizes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">7</span>, <span class="dv">10</span>, <span class="dv">20</span>, <span class="dv">30</span>)

<span class="co"># number of simulations to carry out *for each sample size*</span>
nsims &lt;-<span class="st"> </span><span class="dv">2500</span>

df.combined.small &lt;-<span class="st"> </span><span class="kw">data_frame</span>(<span class="dt">sample.size =</span> <span class="kw">double</span>(), 
                          <span class="dt">sample.mean =</span> <span class="kw">double</span>(), 
                          <span class="dt">sample.sd =</span> <span class="kw">double</span>(),
                          <span class="dt">estimated.SE =</span> <span class="kw">double</span>(), 
                          <span class="dt">sample.zscore =</span> <span class="kw">double</span>())

<span class="cf">for</span> (i <span class="cf">in</span> ssizes) {
  <span class="co"># create empty vectors to hold simulation stats (for efficiency)</span>
  s.means &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="ot">NA</span>, nsims)
  s.sds &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="ot">NA</span>, nsims)
  s.SEs &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="ot">NA</span>, nsims)
  s.zscores &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="ot">NA</span>, nsims)
  
  <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>nsims) {
    s &lt;-<span class="st"> </span><span class="kw">rnorm</span>(i, <span class="dt">mean =</span> mu, <span class="dt">sd =</span> sigma)  <span class="co"># draw random sample</span>
    s.means[j] &lt;-<span class="st"> </span><span class="kw">mean</span>(s) <span class="co"># calculate mean of that sample</span>
    s.sds[j] &lt;-<span class="st"> </span><span class="kw">sd</span>(s) <span class="co"># calculate sd of that sample</span>
    SE &lt;-<span class="st"> </span><span class="kw">sd</span>(s)<span class="op">/</span><span class="kw">sqrt</span>(i)  <span class="co"># cacluate estimated SE of that sample</span>
    s.SEs[j] &lt;-<span class="st"> </span>SE
    s.zscores[j] &lt;-<span class="st"> </span>(<span class="kw">mean</span>(s) <span class="op">-</span><span class="st"> </span>mu)<span class="op">/</span>SE
  }
  df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">sample.size =</span> i, <span class="dt">sample.mean =</span> s.means, <span class="dt">sample.sd =</span> s.sds, 
                   <span class="dt">estimated.SE =</span> s.SEs, <span class="dt">sample.zscore =</span> s.zscores)
  
  df.combined.small &lt;-<span class="st"> </span><span class="kw">bind_rows</span>(df.combined.small, df)
}

df.combined.small <span class="op">%&lt;&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">sample.sz =</span> <span class="kw">as.factor</span>(sample.size))</code></pre></div>
<div id="for-small-samples-sample-standard-deviations-systematically-underestimate-the-population-standard-deviation" class="section level3">
<h3><span class="header-section-number">13.7.1</span> For small samples, sample standard deviations systematically underestimate the population standard deviation</h3>
<p>Let’s examine how the well centered the sampling distributions of the mean and standard deviation are around their true values, as a function of sample size.</p>
<p>First a table summarizing this information:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">by.sample.size &lt;-
<span class="st">  </span>df.combined.small <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(sample.size) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">mean.of.means =</span> <span class="kw">mean</span>(sample.mean),
            <span class="dt">mean.of.sds =</span> <span class="kw">mean</span>(sample.sd))

by.sample.size
<span class="co"># A tibble: 8 x 3</span>
  sample.size mean.of.means mean.of.sds
        <span class="op">&lt;</span>dbl<span class="op">&gt;</span><span class="st">         </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">       </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span>
<span class="dv">1</span>          <span class="dv">2</span>.          <span class="dv">175</span>.        <span class="fl">12.2</span>
<span class="dv">2</span>          <span class="dv">3</span>.          <span class="dv">175</span>.        <span class="fl">13.5</span>
<span class="dv">3</span>          <span class="dv">4</span>.          <span class="dv">176</span>.        <span class="fl">13.8</span>
<span class="dv">4</span>          <span class="dv">5</span>.          <span class="dv">176</span>.        <span class="fl">14.2</span>
<span class="dv">5</span>          <span class="dv">7</span>.          <span class="dv">176</span>.        <span class="fl">14.7</span>
<span class="dv">6</span>         <span class="dv">10</span>.          <span class="dv">176</span>.        <span class="fl">14.8</span>
<span class="dv">7</span>         <span class="dv">20</span>.          <span class="dv">176</span>.        <span class="fl">15.0</span>
<span class="dv">8</span>         <span class="dv">30</span>.          <span class="dv">176</span>.        <span class="fl">15.1</span></code></pre></div>
<p>We see that the sampling distributions of means are well centered around the true mean, and there is no systematic bias one way or the other. By contrast the sampling distribution of standard deviations tends to underestimate the true standard deviation when the samples are small (less than 30 observations).</p>
<p>We can visualize this bias as shown here:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(by.sample.size, <span class="kw">aes</span>(<span class="dt">x =</span> sample.size, <span class="dt">y =</span> mean.of.sds)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">color =</span> <span class="st">&#39;red&#39;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_line</span>(<span class="dt">color =</span> <span class="st">&#39;red&#39;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> sigma, <span class="dt">color =</span> <span class="st">&#39;black&#39;</span>, <span class="dt">linetype=</span><span class="st">&#39;dashed&#39;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Sample Size&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Mean of Sampling Distn of Std Dev&quot;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-393-1.png" width="576" /></p>
<p>The source of this bias is clear if we look at the sampling distribution of the standard deviation for samples of size 3, 5, and 30.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filtered.df &lt;-<span class="st"> </span>
<span class="st">  </span>df.combined.small <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(sample.size <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="dv">3</span>, <span class="dv">5</span>, <span class="dv">30</span>))

<span class="kw">ggplot</span>(filtered.df, <span class="kw">aes</span>(<span class="dt">x =</span> sample.sd, <span class="dt">y =</span> ..density.., <span class="dt">fill =</span> sample.sz)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_histogram</span>(<span class="dt">bins=</span><span class="dv">50</span>, <span class="dt">alpha=</span><span class="fl">0.65</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span>sample.size, <span class="dt">nrow =</span> <span class="dv">1</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> sigma, <span class="dt">linetype =</span> <span class="st">&#39;dashed&#39;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Std Deviations&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Density&quot;</span>,
       <span class="dt">title =</span> <span class="st">&quot;Sampling distributions of the standard deviation</span><span class="ch">\n</span><span class="st">As a function of sample size&quot;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-394-1.png" width="1152" /></p>
<p>There’s very clear indication that the the sampling distribution of standard deviations is not centered around the true value for <span class="math inline">\(n=3\)</span> and for <span class="math inline">\(n=5\)</span>, however with samples of size 30 the sampling distribution of the standard deviation appears fairly well centered around the true value of the underlying population.</p>
</div>
<div id="underestimates-of-the-standard-deviation-given-small-n-lead-to-understimates-of-the-se-of-the-mean" class="section level3">
<h3><span class="header-section-number">13.7.2</span> Underestimates of the standard deviation given small <span class="math inline">\(n\)</span> lead to understimates of the SE of the mean</h3>
<p>When sample sizes are small, sample estimates of the standard deviation, <span class="math inline">\(s_x\)</span>, tend to underestimate the true standard deviation, <span class="math inline">\(\sigma\)</span>, then it follows that sample estimates of the standard error of the mean, <span class="math inline">\(SE_\overline{x} = \frac{s_x}{\sqrt{n}}\)</span>, must tend to understimate the true standard error of the mean, <span class="math inline">\(SE_\mu = \frac{\sigma}{\sqrt{n}}\)</span>.</p>
</div>
</div>
<div id="the-t-distribution-is-the-appropriate-distribution-for-describing-the-sampling-distribution-of-the-mean-when-n-is-small" class="section level2">
<h2><span class="header-section-number">13.8</span> The t-distribution is the appropriate distribution for describing the sampling distribution of the mean when <span class="math inline">\(n\)</span> is small</h2>
<p>The problem associated with estimating the standard error of the mean for small sample sizes was recognized in the early 20th century by William Gosset, an employee at the Guinness Brewing Company. He published a paper, under the pseudonym “Student”, giving the appropriate distribution for describing the standard error of the mean as a function of the sample size <span class="math inline">\(n\)</span>. Gosset’s distribution is known as the “t-distribution” or “Student’s t-distribution”.</p>
<p>The t-distribution is specified by a single parameter, called degrees of freedom (<span class="math inline">\(df\)</span>) where <span class="math inline">\({df} = n - 1\)</span>. As <span class="math inline">\(df\)</span> increases, the t-distribution becomes more and more like the normal such that when <span class="math inline">\(n \geq 30\)</span> it’s nearly indistinguishable from the standard normal distribution.</p>
<p>In the figures below we compare the t-distribution and the standard normal distribution for sample sizes <span class="math inline">\(n = {4, 8, 32}\)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="op">-</span><span class="dv">6</span>, <span class="dv">6</span>, <span class="dt">length.out =</span> <span class="dv">200</span>)
n &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">5</span>, <span class="dv">10</span>, <span class="dv">30</span>) <span class="co"># sample sizes</span>

distns.df &lt;-<span class="st"> </span><span class="kw">data_frame</span>(<span class="dt">sample.size =</span> <span class="kw">double</span>(), <span class="dt">z.or.t =</span> <span class="kw">double</span>(),
                        <span class="dt">norm.density =</span> <span class="kw">double</span>(), <span class="dt">t.density =</span> <span class="kw">double</span>())
<span class="cf">for</span> (i <span class="cf">in</span> n) {
  norm.density &lt;-<span class="st"> </span><span class="kw">dnorm</span>(x, <span class="dt">mean =</span> <span class="dv">0</span>, <span class="dt">sd =</span> <span class="dv">1</span>)
  t.density &lt;-<span class="st"> </span><span class="kw">dt</span>(x, <span class="dt">df =</span> i <span class="op">-</span><span class="st"> </span><span class="dv">1</span>)
  df.temp &lt;-<span class="st"> </span><span class="kw">data_frame</span>(<span class="dt">sample.size =</span> i, <span class="dt">z.or.t =</span> x,
                        <span class="dt">norm.density =</span> norm.density, <span class="dt">t.density  =</span> t.density)
  distns.df &lt;-<span class="st"> </span><span class="kw">bind_rows</span>(distns.df, df.temp)
}

distns.df <span class="op">%&lt;&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">df =</span> <span class="kw">as.factor</span>(sample.size <span class="op">-</span><span class="st"> </span><span class="dv">1</span>))
  
<span class="kw">ggplot</span>(distns.df, <span class="kw">aes</span>(<span class="dt">x =</span> z.or.t, <span class="dt">y =</span> t.density, <span class="dt">color =</span> df)) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">geom_line</span>() <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="dt">y =</span> norm.density), <span class="dt">color=</span><span class="st">&#39;black&#39;</span>, <span class="dt">linetype=</span><span class="st">&quot;dotted&quot;</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;z or t value&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Probablity density&quot;</span>,
       <span class="dt">title =</span> <span class="st">&quot;Standard normal distribution (black dotted line)</span><span class="ch">\n</span><span class="st">versus t-distributions for different degrees of freedom&quot;</span>)</code></pre></div>
<p><img src="bio723-book_files/figure-html/unnamed-chunk-395-1.png" width="576" /></p>

</div>
</div>
<div class="footnotes">
<hr />
<ol start="1">
<li id="fn1"><p>US Dept. of Health and Human Services; et al. (August 2016). “Anthropometric Reference Data for Children and Adults: United States, 2011–2014” (PDF). National Health Statistics Reports. 11. <a href="https://www.cdc.gov/nchs/data/series/sr_03/sr03_039.pdf" class="uri">https://www.cdc.gov/nchs/data/series/sr_03/sr03_039.pdf</a><a href="simulating-sampling-distributions.html#fnref1">↩</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="principal-components-analysis.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="simulating-confidence-intervals.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
